// @c8y/ngx-components/esm2015/core/forms/ip-range-input-list.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './form-group.component';
import * as ɵngcc2 from './input-group-list-container.directive';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from './input-group-list.component';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from './required-input-placeholder.directive';
import * as ɵngcc7 from './min-validation.directive';
import * as ɵngcc8 from '../i18n/c8y-translate.pipe';

function IpRangeInputListComponent_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 3);
    ɵngcc0.ɵɵelementStart(1, "c8y-input-group-list", 4);
    ɵngcc0.ɵɵlistener("onAdd", function IpRangeInputListComponent_li_2_Template_c8y_input_group_list_onAdd_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.add(); })("onRemove", function IpRangeInputListComponent_li_2_Template_c8y_input_group_list_onRemove_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.remove($event); });
    ɵngcc0.ɵɵelementStart(2, "input", 5, 6);
    ɵngcc0.ɵɵlistener("ngModelChange", function IpRangeInputListComponent_li_2_Template_input_ngModelChange_2_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); const item_r1 = restoredCtx.$implicit; return item_r1.ip = $event; });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "input", 7, 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function IpRangeInputListComponent_li_2_Template_input_ngModelChange_5_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); const item_r1 = restoredCtx.$implicit; return item_r1.cidr = $event; });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r1 = ctx.$implicit;
    const i_r2 = ctx.index;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("index", i_r2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(4, 8, "IP range start, e.g."), " 192.168.0.1");
    ɵngcc0.ɵɵproperty("ngModel", item_r1.ip)("required", true);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(7, 10, "e.g."), " 20");
    ɵngcc0.ɵɵproperty("min", 0)("ngModel", item_r1.cidr)("required", true);
} }
export class IpRangeInputListComponent {
    set data(list) {
        this.ipRangeList = list;
        if (this.ipRangeList.length === 0) {
            this.add();
        }
    }
    get data() {
        return this.ipRangeList;
    }
    add() {
        this.ipRangeList.push({ ip: '', cidr: undefined });
    }
    remove(index) {
        this.ipRangeList.splice(index, 1);
    }
}
IpRangeInputListComponent.ɵfac = function IpRangeInputListComponent_Factory(t) { return new (t || IpRangeInputListComponent)(); };
IpRangeInputListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: IpRangeInputListComponent, selectors: [["c8y-ip-range-input-list"]], inputs: { data: "data" }, decls: 3, vars: 1, consts: [[1, "m-t-8", "m-b-8"], ["c8yInputGroupListContainer", "", 1, "list-unstyled"], ["class", "m-b-8", 4, "ngFor", "ngForOf"], [1, "m-b-8"], [3, "index", "onAdd", "onRemove"], ["type", "text", "name", "ipAddress", 1, "form-control", "m-r-4", 2, "flex-grow", "2", 3, "placeholder", "ngModel", "required", "ngModelChange"], ["ipAddress", ""], ["type", "number", "name", "cidr", 1, "form-control", 3, "min", "placeholder", "ngModel", "required", "ngModelChange"], ["cidr", ""]], template: function IpRangeInputListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, IpRangeInputListComponent_li_2_Template, 8, 12, "li", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.data);
    } }, directives: [ɵngcc1.FormGroupComponent, ɵngcc2.InputGroupListContainerDirective, ɵngcc3.NgForOf, ɵngcc4.InputGroupListComponent, ɵngcc5.DefaultValueAccessor, ɵngcc6.RequiredInputPlaceholderDirective, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc5.RequiredValidator, ɵngcc5.MinValidator, ɵngcc5.NumberValueAccessor, ɵngcc7.MinValidationDirective], pipes: [ɵngcc8.C8yTranslatePipe], encapsulation: 2 });
IpRangeInputListComponent.propDecorators = {
    data: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(IpRangeInputListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-ip-range-input-list',
                template: "<c8y-form-group class=\"m-t-8 m-b-8\">\n  <ul c8yInputGroupListContainer class=\"list-unstyled\">\n    <li class=\"m-b-8\" *ngFor=\"let item of data; let i = index\">\n      <c8y-input-group-list [index]=\"i\" (onAdd)=\"add()\" (onRemove)=\"remove($event)\">\n        <input\n          type=\"text\"\n          name=\"ipAddress\"\n          class=\"form-control m-r-4\"\n          placeholder=\"{{ 'IP range start, e.g.' | translate }} 192.168.0.1\"\n          style=\"flex-grow: 2;\"\n          [(ngModel)]=\"item.ip\"\n          [required]=\"true\"\n          #ipAddress\n        />\n        <input\n          type=\"number\"\n          name=\"cidr\"\n          [min]=\"0\"\n          class=\"form-control\"\n          placeholder=\"{{ 'e.g.' | translate }} 20\"\n          [(ngModel)]=\"item.cidr\"\n          [required]=\"true\"\n          #cidr\n        />\n      </c8y-input-group-list>\n    </li>\n  </ul>\n</c8y-form-group>\n"
            }]
    }], null, { data: [{
            type: Input
        }] }); })();
