// @c8y/ngx-components/esm2015/core/forms/range.component.js
import { Component, ContentChild } from '@angular/core';
import { RangeDirective } from './range.directive';
/**
 * An component which renders a range slider in Cumulocity layout. You need
 * to wrap the default range input in this component.
 *
 * Example:
 * ```
 * <c8y-range>
 *   <input id="range" type="range" min="0" max="100000" value="0" step="1">
 * </c8y-range>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["c8yRangeValue"];
function RangeComponent_span_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.value);
} }
function RangeComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c1 = ["*"];
export class RangeComponent {
    constructor() {
        this.changeEventListener = () => this.setValue();
    }
    ngAfterContentInit() {
        this.setValue();
        this.range.elementRef.nativeElement.addEventListener('input', this.changeEventListener);
    }
    ngOnDestroy() {
        this.range.elementRef.nativeElement.removeEventListener('input', this.changeEventListener);
    }
    setValue() {
        const range = this.range.elementRef.nativeElement;
        const newValue = Number(((range.value - range.min) * 100) / (range.max - range.min));
        const newPosition = 12 - newValue * 0.24;
        this.value = range.value;
        this.left = `calc(${newValue}% + (${newPosition}px))`;
    }
}
RangeComponent.ɵfac = function RangeComponent_Factory(t) { return new (t || RangeComponent)(); };
RangeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RangeComponent, selectors: [["c8y-range"]], contentQueries: function RangeComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, RangeDirective, 7);
        ɵngcc0.ɵɵcontentQuery(dirIndex, _c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.range = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.rangeValue = _t.first);
    } }, ngContentSelectors: _c1, decls: 5, vars: 7, consts: [[1, "range-wrap"], [1, "range-value"], [4, "ngIf"], [4, "ngTemplateOutlet"]], template: function RangeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, RangeComponent_span_2_Template, 2, 1, "span", 2);
        ɵngcc0.ɵɵtemplate(3, RangeComponent_ng_container_3_Template, 1, 0, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵstyleMapInterpolate1("--track-width: ", ctx.left, "");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("left", ctx.left);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.rangeValue);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.rangeValue);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
RangeComponent.propDecorators = {
    range: [{ type: ContentChild, args: [RangeDirective, { static: true },] }],
    rangeValue: [{ type: ContentChild, args: ['c8yRangeValue', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RangeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-range',
                template: "<div class=\"range-wrap\" style=\"--track-width: {{left}}\">\n  <div class=\"range-value\" [style.left]=\"left\">\n    <span *ngIf=\"!rangeValue\">{{ value }}</span>\n    <ng-container *ngTemplateOutlet=\"rangeValue\"></ng-container>\n  </div>\n  <ng-content></ng-content>\n</div>\n"
            }]
    }], function () { return []; }, { range: [{
            type: ContentChild,
            args: [RangeDirective, { static: true }]
        }], rangeValue: [{
            type: ContentChild,
            args: ['c8yRangeValue', { static: true }]
        }] }); })();
