// @c8y/ngx-components/esm2015/binary-file-download/binary-file-download.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@c8y/ngx-components';
import { BinaryFileDownloadService } from './binary-file-download.service';
import { RouterModule } from '@angular/router';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './binary-file-download.service';
export class BinaryFileDownloadModule {
    constructor(binaryFileDownloadService) {
        binaryFileDownloadService.run();
    }
}
BinaryFileDownloadModule.ɵfac = function BinaryFileDownloadModule_Factory(t) { return new (t || BinaryFileDownloadModule)(ɵngcc0.ɵɵinject(ɵngcc1.BinaryFileDownloadService)); };
BinaryFileDownloadModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BinaryFileDownloadModule });
BinaryFileDownloadModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [BinaryFileDownloadService], imports: [[CommonModule, RouterModule]] });
BinaryFileDownloadModule.ctorParameters = () => [
    { type: BinaryFileDownloadService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BinaryFileDownloadModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, RouterModule],
                providers: [BinaryFileDownloadService]
            }]
    }], function () { return [{ type: ɵngcc1.BinaryFileDownloadService }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BinaryFileDownloadModule, { imports: function () { return [CommonModule, RouterModule]; } }); })();
