// @c8y/ngx-components/esm2015/auth-configuration/factories/tabs.factory.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class TabsFactory {
    constructor(router) {
        this.router = router;
    }
    get() {
        const tabs = [];
        if (this.router.url.match(/auth-configuration/g)) {
            tabs.push({
                path: 'auth-configuration/basic_settings',
                label: gettext('Basic settings'),
                icon: 'c8y-icon dlt-c8y-icon-unlock-alt',
                priority: 1100,
                orientation: 'horizontal'
            });
            tabs.push({
                path: 'auth-configuration/single_sign-on',
                label: gettext('Single sign-on'),
                icon: 'c8y-icon dlt-c8y-icon-sign-in',
                priority: 1050,
                orientation: 'horizontal'
            });
        }
        return tabs;
    }
}
TabsFactory.ɵfac = function TabsFactory_Factory(t) { return new (t || TabsFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Router)); };
TabsFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TabsFactory, factory: TabsFactory.ɵfac });
TabsFactory.ctorParameters = () => [
    { type: Router }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabsFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }]; }, null); })();
