// @c8y/ngx-components/esm2015/operations/bulk-operations-list/bulk-operations-list.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterModule } from '@c8y/ngx-components/operations/status-filter';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { BulkOperationDetailsService } from './bulk-operation-details.service';
import { BulkOperationsListComponent } from './bulk-operations-list.component';
import { BulkOperationListItemComponent } from './bulk-operation-list-item.component';
import { BulkOperationDetailsTabsComponent } from './bulk-operation-details-tabs.component';
import { BulkOperationDetailsComponent } from './bulk-operation-details.component';
import { SingleOperationsListComponent } from './single-operations-list.component';
import { SingleOperationItemComponent } from './single-operation-item.component';
import { BulkOperationWrapperComponent } from './bulk-operation-wrapper.component';
import { BulkOperationTypeListComponent } from './type-list/bulk-operation-type-list.component';
import { BulkOperationModalsService } from './modals/bulk-operation-modals.service';
import { BulkOperationsModalComponent } from './modals/bulk-operations-modal.component';
import { BulkOperationsRescheduleModalComponent } from './modals/bulk-operations-reschedule-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components';
export const bulkOperationListRoutes = [
    {
        path: 'devicecontrol/bulk',
        component: BulkOperationsListComponent
    }
];
/**
 * This module allows for displaying the list of bulk operations and their details.
 */
export class BulkOperationsListModule {
}
BulkOperationsListModule.ɵfac = function BulkOperationsListModule_Factory(t) { return new (t || BulkOperationsListModule)(); };
BulkOperationsListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationsListModule });
BulkOperationsListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [BulkOperationDetailsService, BulkOperationModalsService], imports: [[
            CoreModule,
            PopoverModule,
            TooltipModule,
            StatusFilterModule,
            OperationSummaryModule,
            BulkOperationSchedulerModule,
            RouterModule.forChild(bulkOperationListRoutes)
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsListModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    PopoverModule,
                    TooltipModule,
                    StatusFilterModule,
                    OperationSummaryModule,
                    BulkOperationSchedulerModule,
                    RouterModule.forChild(bulkOperationListRoutes)
                ],
                providers: [BulkOperationDetailsService, BulkOperationModalsService],
                declarations: [
                    BulkOperationsListComponent,
                    BulkOperationListItemComponent,
                    BulkOperationDetailsTabsComponent,
                    BulkOperationDetailsComponent,
                    SingleOperationsListComponent,
                    SingleOperationItemComponent,
                    BulkOperationWrapperComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent,
                    BulkOperationsRescheduleModalComponent
                ],
                entryComponents: [
                    BulkOperationsListComponent,
                    BulkOperationListItemComponent,
                    BulkOperationDetailsTabsComponent,
                    BulkOperationDetailsComponent,
                    SingleOperationsListComponent,
                    SingleOperationItemComponent,
                    BulkOperationWrapperComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent,
                    BulkOperationsRescheduleModalComponent
                ],
                exports: [
                    BulkOperationsListComponent,
                    BulkOperationListItemComponent,
                    BulkOperationDetailsTabsComponent,
                    BulkOperationDetailsComponent,
                    SingleOperationsListComponent,
                    SingleOperationItemComponent,
                    BulkOperationWrapperComponent,
                    BulkOperationTypeListComponent,
                    BulkOperationsModalComponent,
                    BulkOperationsRescheduleModalComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationsListModule, { declarations: function () { return [BulkOperationsListComponent,
        BulkOperationListItemComponent,
        BulkOperationDetailsTabsComponent,
        BulkOperationDetailsComponent,
        SingleOperationsListComponent,
        SingleOperationItemComponent,
        BulkOperationWrapperComponent,
        BulkOperationTypeListComponent,
        BulkOperationsModalComponent,
        BulkOperationsRescheduleModalComponent]; }, imports: function () { return [CoreModule,
        PopoverModule,
        TooltipModule,
        StatusFilterModule,
        OperationSummaryModule,
        BulkOperationSchedulerModule, ɵngcc1.RouterModule]; }, exports: function () { return [BulkOperationsListComponent,
        BulkOperationListItemComponent,
        BulkOperationDetailsTabsComponent,
        BulkOperationDetailsComponent,
        SingleOperationsListComponent,
        SingleOperationItemComponent,
        BulkOperationWrapperComponent,
        BulkOperationTypeListComponent,
        BulkOperationsModalComponent,
        BulkOperationsRescheduleModalComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(BulkOperationsModalComponent, [ɵngcc2.C8yTranslateDirective, ɵngcc2.GuideDocsComponent, ɵngcc2.GuideHrefDirective, BulkOperationTypeListComponent], [ɵngcc2.C8yTranslatePipe]);
