// @c8y/ngx-components/esm2015/operations/bulk-operations-list/bulk-operation-list-item.model.js
import { OperationBulkGeneralStatus } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
export const BULK_OPERATION_STATUS_LABELS = {
    SCHEDULED: gettext('Scheduled'),
    EXECUTING: gettext('Executing'),
    CANCELED: gettext('Canceled'),
    FAILED: gettext('Completed with failures'),
    SUCCESSFUL: gettext('Completed successfully'),
    DELETED: gettext('Deleted')
};
export const BULK_OPERATION_STATUS_OPTIONS = [
    {
        label: BULK_OPERATION_STATUS_LABELS.SCHEDULED,
        generalStatus: [OperationBulkGeneralStatus.SCHEDULED],
        iconClass: 'dlt-c8y-icon-calendar text-info'
    },
    {
        label: BULK_OPERATION_STATUS_LABELS.EXECUTING,
        generalStatus: [
            OperationBulkGeneralStatus.EXECUTING,
            OperationBulkGeneralStatus.EXECUTING_WITH_ERROR
        ],
        iconClass: 'dlt-c8y-icon-refresh text-primary'
    },
    {
        label: BULK_OPERATION_STATUS_LABELS.DELETED,
        generalStatus: [OperationBulkGeneralStatus.DELETED],
        iconClass: 'dlt-c8y-icon-delete-document text-muted'
    },
    {
        label: BULK_OPERATION_STATUS_LABELS.CANCELED,
        generalStatus: [OperationBulkGeneralStatus.CANCELED],
        iconClass: 'dlt-c8y-icon-ban text-muted'
    },
    {
        label: BULK_OPERATION_STATUS_LABELS.FAILED,
        generalStatus: [OperationBulkGeneralStatus.FAILED],
        iconClass: 'dlt-c8y-icon-exclamation-triangle text-danger'
    },
    {
        label: BULK_OPERATION_STATUS_LABELS.SUCCESSFUL,
        generalStatus: [OperationBulkGeneralStatus.SUCCESSFUL],
        iconClass: 'dlt-c8y-icon-check-circle text-success'
    }
];
export const BULK_OPERATION_FALLBACK_STATUS_ICON = 'dlt-c8y-icon-ask-question text-muted';
