// @c8y/ngx-components/esm2015/upgrade/dashboard/widget.component.js
import { __awaiter } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Optional } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetsDashboardComponent } from '@c8y/ngx-components';
export class WidgetComponent {
    constructor(dashboard, dashboardBridgeService, el) {
        this.dashboard = dashboard;
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.config) {
                this.ng1Scope = yield this.dashboardBridgeService.instantiateComponent(this.config, this.el.nativeElement);
                if (this.dashboard) {
                    this.resizeSubscription = this.dashboard.onChangeDashboard.subscribe(() => {
                        this.ng1Scope.$broadcast('dashboardResize');
                    });
                }
            }
        });
    }
    ngOnDestroy() {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
        if (this.resizeSubscription) {
            this.resizeSubscription.unsubscribe();
        }
    }
}
WidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'c8y-widget-legacy',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
WidgetComponent.ctorParameters = () => [
    { type: WidgetsDashboardComponent, decorators: [{ type: Optional }] },
    { type: DashboardBridgeService },
    { type: ElementRef }
];
