// @c8y/ngx-components/esm2015/core/alert/alert-text.component.js
import { Component } from '@angular/core';
/**
 * The content of `c8y-alert-text` component will be displayed as alert's text.
 * This component must be placed inside `c8y-alert` component.
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class AlertTextComponent {
}
AlertTextComponent.ɵfac = function AlertTextComponent_Factory(t) { return new (t || AlertTextComponent)(); };
AlertTextComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AlertTextComponent, selectors: [["c8y-alert-text"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function AlertTextComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlertTextComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-alert-text',
                template: `<ng-content></ng-content>`
            }]
    }], null, null); })();
