// @c8y/ngx-components/esm2015/core/alert/alert.module.js
import { NgModule } from '@angular/core';
import { AlertDetailsComponent } from './alert-details.component';
import { AlertOutletComponent } from './alert-outlet.component';
import { AlertTextComponent } from './alert-text.component';
import { AlertComponent } from './alert.component';
import { CommonModule } from '../common/common.module';
/**
 * Alert module is used for displaying status messages to the user e.g. danger, success, warning and info alerts.
 */
import * as ɵngcc0 from '@angular/core';
export class AlertModule {
}
AlertModule.ɵfac = function AlertModule_Factory(t) { return new (t || AlertModule)(); };
AlertModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AlertModule });
AlertModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlertModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                exports: [
                    AlertOutletComponent,
                    AlertComponent,
                    AlertDetailsComponent,
                    AlertTextComponent
                ],
                declarations: [
                    AlertOutletComponent,
                    AlertComponent,
                    AlertDetailsComponent,
                    AlertTextComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AlertModule, { declarations: function () { return [AlertOutletComponent,
        AlertComponent,
        AlertDetailsComponent,
        AlertTextComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [AlertOutletComponent,
        AlertComponent,
        AlertDetailsComponent,
        AlertTextComponent]; } }); })();
