// @c8y/ngx-components/esm2015/api/api.service.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { FetchClient } from '@c8y/client';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/client/lib/src/core/FetchClient";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ApiService {
    constructor(client) {
        this.client = client;
        this.callsSubject = new Subject();
        this.calls = this.callsSubject.asObservable();
        this.hookIntoClientFetch();
    }
    hookResponse(hookFilter) {
        return this.callsSubject.pipe(filter(({ phase }) => phase === 'finish'), filter(hookFilter));
    }
    hookRequest(hookFilter) {
        return this.callsSubject.pipe(filter(({ phase }) => phase === 'start'), filter(hookFilter));
    }
    onFinish(call) {
        return __awaiter(this, void 0, void 0, function* () {
            this.callsSubject.next(Object.assign({ phase: 'finish' }, call));
        });
    }
    onStart(call) {
        this.callsSubject.next(Object.assign({ phase: 'start' }, call));
    }
    resolveData(call) {
        return __awaiter(this, void 0, void 0, function* () {
            const { response, method } = call;
            if ('data' in response) {
                return Promise.resolve({ data: response.data, method });
            }
            else {
                const cb = (data) => ({ data, method });
                return response.clone().json().then(cb, cb);
            }
        });
    }
    hookIntoClientFetch() {
        const fetch = this.client.fetch.bind(this.client);
        this.client.fetch = (url, options = { method: 'GET' }) => __awaiter(this, void 0, void 0, function* () {
            const { method } = options;
            this.onStart({ options, method, url });
            let fetchPromise = fetch(url, options);
            if (typeof options.responseInterceptor === 'function') {
                fetchPromise = fetchPromise.then(options.responseInterceptor);
            }
            fetchPromise.then((response) => this.onFinish({ response, url, options, method }), (response) => this.onFinish({ response, url, options, method }));
            return fetchPromise;
        });
    }
}
ApiService.ɵfac = function ApiService_Factory(t) { return new (t || ApiService)(ɵngcc0.ɵɵinject(ɵngcc1.FetchClient)); };
ApiService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ApiService_Factory() { return new ApiService(i0.ɵɵinject(i1.FetchClient)); }, token: ApiService, providedIn: "root" });
ApiService.ctorParameters = () => [
    { type: FetchClient }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApiService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.FetchClient }]; }, null); })();
