// @c8y/ngx-components/esm2015/repository/configuration-device-tab/configuration-preview.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { DeviceConfigurationOperation } from '../repository.model';
import { OperationService, OperationStatus, UserService } from '@c8y/client';
import { DeviceConfigurationService } from './device-configuration.service';
import { map } from 'rxjs/operators';
import { saveAs } from 'file-saver';
import { BsModalService } from 'ngx-bootstrap/modal';
import { SaveToRepositoryComponent } from './save-to-repository.component';
import { cloneDeep } from 'lodash-es';
import { AlertService, AppStateService, OperationRealtimeService } from '@c8y/ngx-components';
import { RepositoryService } from '../repository.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './device-configuration.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@c8y/client';
import * as ɵngcc5 from '../repository.service';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from './source-code-preview.component';
import * as ɵngcc8 from '@c8y/ngx-components/operations/single-operation-details';

function ConfigurationPreviewComponent_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "strong");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.configSnapshot.name);
} }
function ConfigurationPreviewComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0, " --- ");
} }
function ConfigurationPreviewComponent_small_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r3.configSnapshot.time), " ");
} }
function ConfigurationPreviewComponent_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0, " --- ");
} }
function ConfigurationPreviewComponent_button_15_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵlistener("click", function ConfigurationPreviewComponent_button_15_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.createDeviceOperation(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 12);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 4, ctx_r6.actionButtonText));
    ɵngcc0.ɵɵproperty("disabled", ctx_r6.isCreateOperationDisabled());
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r6.actionButtonIcon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 6, ctx_r6.actionButtonText), " ");
} }
function ConfigurationPreviewComponent_div_16_small_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 14);
    ɵngcc0.ɵɵtext(1, "The file is not available.");
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationPreviewComponent_div_16_ng_template_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 14);
    ɵngcc0.ɵɵtext(1, "The file is empty.");
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationPreviewComponent_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelement(1, "h1", 12);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 14);
    ɵngcc0.ɵɵtext(4, "No preview available.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "br");
    ɵngcc0.ɵɵtemplate(6, ConfigurationPreviewComponent_div_16_small_6_Template, 2, 0, "small", 15);
    ɵngcc0.ɵɵtemplate(7, ConfigurationPreviewComponent_div_16_ng_template_7_Template, 2, 0, "ng-template", null, 16, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r13 = ɵngcc0.ɵɵreference(8);
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "file-image-o");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r7.configSnapshot == null ? null : ctx_r7.configSnapshot.binary) !== "")("ngIfElse", _r13);
} }
function ConfigurationPreviewComponent_div_17_div_2_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 23);
    ɵngcc0.ɵɵlistener("click", function ConfigurationPreviewComponent_div_17_div_2_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(3); return ctx_r17.saveToRepository(); });
    ɵngcc0.ɵɵtext(1, " Save to repository ");
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationPreviewComponent_div_17_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelementStart(1, "button", 21);
    ɵngcc0.ɵɵlistener("click", function ConfigurationPreviewComponent_div_17_div_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r20); const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.download(); });
    ɵngcc0.ɵɵtext(2, " Download ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, ConfigurationPreviewComponent_div_17_div_2_button_3_Template, 2, 0, "button", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r15.hasPermission());
} }
function ConfigurationPreviewComponent_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵelement(1, "c8y-source-code-preview", 18);
    ɵngcc0.ɵɵtemplate(2, ConfigurationPreviewComponent_div_17_div_2_Template, 4, 1, "div", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", ctx_r8.configSnapshot.binary)("isDisabled", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r8.canSaveSnapshot);
} }
function ConfigurationPreviewComponent_div_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-single-operation", 24);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("operation", ctx_r9.operation);
} }
export class ConfigurationPreviewComponent {
    constructor(deviceConfigurationService, operationRealtime, bsModal, user, appState, repositoryService, operationService, alertService) {
        this.deviceConfigurationService = deviceConfigurationService;
        this.operationRealtime = operationRealtime;
        this.bsModal = bsModal;
        this.user = user;
        this.appState = appState;
        this.repositoryService = repositoryService;
        this.operationService = operationService;
        this.alertService = alertService;
        this.isLegacy = false;
        this.canCallAction = true;
        this.deviceConfigurationOperation = DeviceConfigurationOperation;
    }
    set configurationType(type) {
        this._configurationType = type;
        this.setOperation(type);
    }
    get configurationType() {
        return this._configurationType;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.setCanCallAction();
            this.setOperation(this._configurationType);
            this.operationsSubscription = this.operationRealtime
                .onAll$(this.device.id)
                .pipe(map(({ data }) => data))
                .subscribe(operation => {
                this.updatePreview(operation);
            });
        });
    }
    setOperation(configType) {
        return __awaiter(this, void 0, void 0, function* () {
            const operationList = yield this.repositoryService.getConfigFileOperationList(this.device.id, this.operationToTrigger);
            const operation = this.isLegacy
                ? operationList.find(op => op[this.operationToTrigger] && !op[this.operationToTrigger].type)
                : operationList.find(op => op[this.operationToTrigger].type === configType);
            this.operation =
                operation && operation.status !== OperationStatus.SUCCESSFUL ? operation : undefined;
        });
    }
    setCanCallAction() {
        this.canCallAction = this.deviceConfigurationService.hasAnySupportedOperation(this.device, this.operationToTrigger);
    }
    createDeviceOperation() {
        return __awaiter(this, void 0, void 0, function* () {
            let operationCfg;
            if (this.operationToTrigger === DeviceConfigurationOperation.DOWNLOAD_CONFIG) {
                operationCfg = this.repositoryService.getDownloadConfigurationFileOperation(this.device, this._configurationType, this.configSnapshot, this.isLegacy);
            }
            if (this.operationToTrigger === DeviceConfigurationOperation.UPLOAD_CONFIG) {
                operationCfg = this.repositoryService.getUploadConfigurationFileOperation(this.device, this._configurationType, this.isLegacy);
            }
            try {
                this.operation = (yield this.operationService.create(operationCfg)).data;
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    showOperation() {
        if (this.operationToTrigger === DeviceConfigurationOperation.DOWNLOAD_CONFIG) {
            return !!this.operation;
        }
        return (this.operation &&
            [OperationStatus.PENDING, OperationStatus.EXECUTING, OperationStatus.FAILED].includes(this.operation.status));
    }
    showBinary() {
        if (this.operationToTrigger === DeviceConfigurationOperation.DOWNLOAD_CONFIG) {
            return true;
        }
        return !this.showOperation();
    }
    isCreateOperationDisabled() {
        return (this.operation &&
            [OperationStatus.PENDING, OperationStatus.EXECUTING].includes(this.operation.status));
    }
    updatePreview(operation) {
        if (operation &&
            operation[this.operationToTrigger] &&
            (this.isLegacy ||
                (operation[this.operationToTrigger].type &&
                    operation[this.operationToTrigger].type === this.configurationType))) {
            this.operation = operation;
            this.updateSnapshotsOnConfigUpload(operation);
        }
    }
    download() {
        const blob = new Blob([this.configSnapshot.binary], { type: this.configSnapshot.binaryType });
        let fileName = this.configSnapshot.name;
        switch (this.configSnapshot.binaryType) {
            case 'text/csv':
            case 'application/csv':
                fileName = fileName.concat('.csv');
                break;
            case 'text/yaml':
            case 'application/x-yaml':
                fileName = fileName.concat('.yaml');
                break;
            case 'application/json':
                fileName = fileName.concat('.json');
                break;
        }
        saveAs(blob, fileName);
    }
    saveToRepository() {
        return __awaiter(this, void 0, void 0, function* () {
            const initialState = {
                configSnapshot: cloneDeep(this.configSnapshot)
            };
            const modal = this.bsModal.show(SaveToRepositoryComponent, {
                class: 'modal-sm',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                yield modal.result;
                this.deviceConfigurationService.updateConfigurations(true);
                modal.close();
            }
            catch (ex) {
                // do nothing
            }
        });
    }
    hasPermission() {
        return this.user.hasAnyRole(this.appState.currentUser.value, [
            'ROLE_INVENTORY_ADMIN',
            'ROLE_INVENTORY_CREATE'
        ]);
    }
    ngOnDestroy() {
        if (this.operationsSubscription) {
            this.operationsSubscription.unsubscribe();
        }
    }
    updateSnapshotsOnConfigUpload(operation) {
        return __awaiter(this, void 0, void 0, function* () {
            if (operation[DeviceConfigurationOperation.UPLOAD_CONFIG] &&
                operation.status === OperationStatus.SUCCESSFUL) {
                this.deviceConfigurationService.updateConfigurations();
            }
        });
    }
}
ConfigurationPreviewComponent.ɵfac = function ConfigurationPreviewComponent_Factory(t) { return new (t || ConfigurationPreviewComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DeviceConfigurationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OperationRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.OperationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
ConfigurationPreviewComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ConfigurationPreviewComponent, selectors: [["c8y-device-configuration-preview"]], inputs: { isLegacy: "isLegacy", configurationType: "configurationType", device: "device", configSnapshot: "configSnapshot", canSaveSnapshot: "canSaveSnapshot", actionButtonText: "actionButtonText", actionButtonIcon: "actionButtonIcon", operationToTrigger: "operationToTrigger" }, decls: 19, vars: 8, consts: [[1, "content-flex-55", "p-b-16"], [1, "col-7", "p-t-4"], ["translate", "", 1, "text-label-small", "text-uppercase", "m-r-4"], [4, "ngIf", "ngIfElse"], ["emptyText", ""], ["emptyDate", ""], [1, "col-5"], ["id", "action-btn", "class", "btn btn-default btn-sm pull-right", "type", "button", 3, "title", "disabled", "click", 4, "ngIf"], ["class", "c8y-empty-state text-left", 4, "ngIf"], ["class", "flex-grow d-flex d-col", 4, "ngIf"], [4, "ngIf"], ["id", "action-btn", "type", "button", 1, "btn", "btn-default", "btn-sm", "pull-right", 3, "title", "disabled", "click"], [3, "c8yIcon"], [1, "c8y-empty-state", "text-left"], ["translate", ""], ["translate", "", 4, "ngIf", "ngIfElse"], ["emptyFile", ""], [1, "flex-grow", "d-flex", "d-col"], [1, "d-contents", 3, "text", "isDisabled"], ["class", "p-t-16", 4, "ngIf"], [1, "p-t-16"], ["type", "button", "translate", "", 1, "btn", "btn-primary", "btn-sm", "pull-right", "m-l-8", 3, "click"], ["type", "button", "class", "btn btn-default btn-sm pull-right", "translate", "", 3, "click", 4, "ngIf"], ["type", "button", "translate", "", 1, "btn", "btn-default", "btn-sm", "pull-right", 3, "click"], [3, "operation"]], template: function ConfigurationPreviewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "p");
        ɵngcc0.ɵɵelementStart(3, "span", 2);
        ɵngcc0.ɵɵtext(4, "Configuration");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, ConfigurationPreviewComponent_span_5_Template, 3, 1, "span", 3);
        ɵngcc0.ɵɵtemplate(6, ConfigurationPreviewComponent_ng_template_6_Template, 1, 0, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "p");
        ɵngcc0.ɵɵelementStart(9, "span", 2);
        ɵngcc0.ɵɵtext(10, "Last updated");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, ConfigurationPreviewComponent_small_11_Template, 3, 3, "small", 3);
        ɵngcc0.ɵɵtemplate(12, ConfigurationPreviewComponent_ng_template_12_Template, 1, 0, "ng-template", null, 5, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 6);
        ɵngcc0.ɵɵtemplate(15, ConfigurationPreviewComponent_button_15_Template, 5, 8, "button", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(16, ConfigurationPreviewComponent_div_16_Template, 9, 3, "div", 8);
        ɵngcc0.ɵɵtemplate(17, ConfigurationPreviewComponent_div_17_Template, 3, 3, "div", 9);
        ɵngcc0.ɵɵtemplate(18, ConfigurationPreviewComponent_div_18_Template, 2, 1, "div", 10);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(7);
        const _r4 = ɵngcc0.ɵɵreference(13);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.configSnapshot == null ? null : ctx.configSnapshot.name)("ngIfElse", _r1);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngIf", ctx.configSnapshot == null ? null : ctx.configSnapshot.time)("ngIfElse", _r4);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.canCallAction);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !(ctx.configSnapshot == null ? null : ctx.configSnapshot.binary) && ctx.showBinary());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.configSnapshot == null ? null : ctx.configSnapshot.binary) && ctx.showBinary());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showOperation());
    } }, directives: [ɵngcc2.C8yTranslateDirective, ɵngcc6.NgIf, ɵngcc2.IconDirective, ɵngcc7.SourceCodePreviewComponent, ɵngcc8.SingleOperationComponent], pipes: [ɵngcc2.DatePipe, ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ConfigurationPreviewComponent.ctorParameters = () => [
    { type: DeviceConfigurationService },
    { type: OperationRealtimeService },
    { type: BsModalService },
    { type: UserService },
    { type: AppStateService },
    { type: RepositoryService },
    { type: OperationService },
    { type: AlertService }
];
ConfigurationPreviewComponent.propDecorators = {
    device: [{ type: Input }],
    configurationType: [{ type: Input }],
    configSnapshot: [{ type: Input }],
    canSaveSnapshot: [{ type: Input }],
    actionButtonText: [{ type: Input }],
    actionButtonIcon: [{ type: Input }],
    isLegacy: [{ type: Input }],
    operationToTrigger: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfigurationPreviewComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-configuration-preview',
                template: "<div class=\"content-flex-55 p-b-16\">\n  <div class=\"col-7 p-t-4\">\n    <p>\n      <span class=\"text-label-small text-uppercase m-r-4\" translate>Configuration</span>\n      <span *ngIf=\"configSnapshot?.name; else emptyText\">\n        <strong>{{ configSnapshot.name }}</strong>\n      </span>\n      <ng-template #emptyText> --- </ng-template>\n    </p>\n    <p>\n      <span class=\"text-label-small text-uppercase m-r-4\" translate>Last updated</span>\n      <small *ngIf=\"configSnapshot?.time; else emptyDate\">\n        {{ configSnapshot.time | c8yDate }}\n      </small>\n      <ng-template #emptyDate> --- </ng-template>\n    </p>\n  </div>\n  <div class=\"col-5\">\n    <button\n      id=\"action-btn\"\n      class=\"btn btn-default btn-sm pull-right\"\n      type=\"button\"\n      title=\"{{ actionButtonText | translate }}\"\n      (click)=\"createDeviceOperation()\"\n      [disabled]=\"isCreateOperationDisabled()\"\n      *ngIf=\"canCallAction\"\n    >\n      <i [c8yIcon]=\"actionButtonIcon\"></i> {{ actionButtonText | translate }}\n    </button>\n  </div>\n</div>\n<div class=\"c8y-empty-state text-left\" *ngIf=\"!configSnapshot?.binary && showBinary()\">\n  <h1 [c8yIcon]=\"'file-image-o'\"></h1>\n  <p>\n    <strong translate>No preview available.</strong><br />\n    <small *ngIf=\"configSnapshot?.binary !== ''; else emptyFile\" translate\n      >The file is not available.</small\n    >\n    <ng-template #emptyFile>\n      <small translate>The file is empty.</small>\n    </ng-template>\n  </p>\n</div>\n<div *ngIf=\"configSnapshot?.binary && showBinary()\" class=\"flex-grow d-flex d-col\">\n  <c8y-source-code-preview\n    [text]=\"configSnapshot.binary\"\n    [isDisabled]=\"true\"\n    class=\"d-contents\"\n  ></c8y-source-code-preview>\n  <div *ngIf=\"canSaveSnapshot\" class=\"p-t-16\">\n    <button\n      type=\"button\"\n      class=\"btn btn-primary btn-sm pull-right m-l-8\"\n      (click)=\"download()\"\n      translate\n    >\n      Download\n    </button>\n    <button\n      *ngIf=\"hasPermission()\"\n      type=\"button\"\n      class=\"btn btn-default btn-sm pull-right\"\n      (click)=\"saveToRepository()\"\n      translate\n    >\n      Save to repository\n    </button>\n  </div>\n</div>\n<div *ngIf=\"showOperation()\">\n  <c8y-single-operation [operation]=\"operation\"></c8y-single-operation>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.DeviceConfigurationService }, { type: ɵngcc2.OperationRealtimeService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc4.UserService }, { type: ɵngcc2.AppStateService }, { type: ɵngcc5.RepositoryService }, { type: ɵngcc4.OperationService }, { type: ɵngcc2.AlertService }]; }, { isLegacy: [{
            type: Input
        }], configurationType: [{
            type: Input
        }], device: [{
            type: Input
        }], configSnapshot: [{
            type: Input
        }], canSaveSnapshot: [{
            type: Input
        }], actionButtonText: [{
            type: Input
        }], actionButtonIcon: [{
            type: Input
        }], operationToTrigger: [{
            type: Input
        }] }); })();
