// @c8y/ngx-components/esm2015/repository/configuration-device-tab/configuration-filter.pipe.js
import { Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class ConfigurationFilterPipe {
    transform(items, filterTerm) {
        return filterTerm.trim().length === 0
            ? items
            : items.filter((item) => this.filterContainString(item.name, filterTerm) ||
                this.filterContainString(item.deviceType, filterTerm));
    }
    filterContainString(name, filterTerm) {
        const term = filterTerm.toLowerCase().trim();
        return name && name.toLowerCase().indexOf(term) > -1;
    }
}
ConfigurationFilterPipe.ɵfac = function ConfigurationFilterPipe_Factory(t) { return new (t || ConfigurationFilterPipe)(); };
ConfigurationFilterPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "configurationFilterPipe", type: ConfigurationFilterPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfigurationFilterPipe, [{
        type: Pipe,
        args: [{
                name: 'configurationFilterPipe'
            }]
    }], null, null); })();
