// @c8y/ngx-components/esm2015/ecosystem/applications/application-plugins/install-plugin.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { ModalService, gettext, Status } from '@c8y/ngx-components';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { map } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from './plugin-list.component';
export class InstallPluginComponent {
    constructor(bsModalRef, modal) {
        this.bsModalRef = bsModalRef;
        this.modal = modal;
        this.filterTerm$ = new BehaviorSubject('');
        this.filteredPlugins$ = new BehaviorSubject([]);
        this.selectedPlugins = [];
        this.result = new Promise((resolve, reject) => {
            this._install = resolve;
            this._cancel = reject;
        });
    }
    ngOnInit() {
        this.filteredPlugins$ = combineLatest(this.plugins$, this.filterTerm$).pipe(map(([plugins, filterTerm]) => filterTerm.trim().length === 0
            ? plugins
            : plugins.filter((plugin) => this.filterContainString(plugin.name, filterTerm))));
    }
    cancel() {
        this.bsModalRef.hide();
        this._cancel();
    }
    install() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.modal.confirm(gettext('Install plugins'), gettext('The operation may take several minutes. During this time the application may be unavailable. Do you want to proceed?'), Status.INFO, { ok: gettext('Continue') });
            this._install(this.selectedPlugins);
            this.bsModalRef.hide();
        });
    }
    filterContainString(name, filterTerm) {
        const term = filterTerm.toLowerCase().trim();
        return name && name.toLowerCase().indexOf(term) > -1;
    }
}
InstallPluginComponent.ɵfac = function InstallPluginComponent_Factory(t) { return new (t || InstallPluginComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ModalService)); };
InstallPluginComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: InstallPluginComponent, selectors: [["c8y-install-plugin"]], inputs: { plugins$: "plugins$" }, decls: 21, vars: 17, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], ["translate", ""], [1, "p-t-8", "p-16", "text-center", "separator-bottom"], ["translate", "", 1, "text-medium", "m-b-8"], [3, "onSearch"], [1, "modal-inner-scroll"], [1, "modal-body", "p-t-0", "p-b-0"], [1, "m-t-16", 3, "emptyListText", "plugins$", "selectable", "selectedItems"], [1, "modal-footer"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "title", "click"]], template: function InstallPluginComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, "Available plugins");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "p", 5);
        ɵngcc0.ɵɵtext(7, "Select the compatible plugins to install");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "c8y-filter", 6);
        ɵngcc0.ɵɵlistener("onSearch", function InstallPluginComponent_Template_c8y_filter_onSearch_8_listener($event) { return ctx.filterTerm$.next($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "c8y-plugin-list", 9);
        ɵngcc0.ɵɵlistener("selectedItems", function InstallPluginComponent_Template_c8y_plugin_list_selectedItems_11_listener($event) { return ctx.selectedPlugins = $event; });
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "div", 10);
        ɵngcc0.ɵɵelementStart(14, "button", 11);
        ɵngcc0.ɵɵlistener("click", function InstallPluginComponent_Template_button_click_14_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵtext(16, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "button", 12);
        ɵngcc0.ɵɵlistener("click", function InstallPluginComponent_Template_button_click_17_listener() { return ctx.install(); });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵtext(19);
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "plugin");
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("emptyListText", ɵngcc0.ɵɵpipeBind1(12, 9, "No plugins available"))("plugins$", ctx.filteredPlugins$)("selectable", true);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 11, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 13, "Install"));
        ɵngcc0.ɵɵproperty("disabled", ctx.selectedPlugins.length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate2(" ", ɵngcc0.ɵɵpipeBind1(20, 15, "Install"), " (", ctx.selectedPlugins.length, ") ");
    } }, directives: [ɵngcc2.IconDirective, ɵngcc2.C8yTranslateDirective, ɵngcc2.FilterInputComponent, ɵngcc3.PluginListComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
InstallPluginComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: ModalService }
];
InstallPluginComponent.propDecorators = {
    plugins$: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(InstallPluginComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-install-plugin',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'plugin'\"></i>\n    <h4 translate>Available plugins</h4>\n  </div>\n  <div class=\"p-t-8 p-16 text-center separator-bottom\">\n    <p class=\"text-medium m-b-8\" translate>Select the compatible plugins to install</p>\n    <c8y-filter (onSearch)=\"this.filterTerm$.next($event)\"></c8y-filter>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body p-t-0 p-b-0\">\n      <c8y-plugin-list\n        (selectedItems)=\"selectedPlugins = $event\"\n        [emptyListText]=\"'No plugins available' | translate\"\n        [plugins$]=\"filteredPlugins$\"\n        [selectable]=\"true\"\n        class=\"m-t-16\"\n      ></c8y-plugin-list>\n    </div>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      (click)=\"cancel()\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n      translate\n      type=\"button\"\n    >\n      Cancel\n    </button>\n    <button\n      (click)=\"install()\"\n      [disabled]=\"selectedPlugins.length === 0\"\n      class=\"btn btn-primary\"\n      title=\"{{ 'Install' | translate }}\"\n      type=\"button\"\n    >\n      {{ 'Install' | translate }} ({{ selectedPlugins.length }})\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.ModalService }]; }, { plugins$: [{
            type: Input
        }] }); })();
