// @c8y/ngx-components/esm2015/ecosystem/applications/application-plugins/plugin-list.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './plugin-list-item.component';

function PluginListComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-plugin-list-item", 4);
    ɵngcc0.ɵɵlistener("isItemSelected", function PluginListComponent_ng_container_1_ng_container_1_Template_c8y_plugin_list_item_isItemSelected_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); const plugin_r4 = restoredCtx.$implicit; const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.updateSelectedItems($event, plugin_r4.id); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const plugin_r4 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("plugin", plugin_r4)("selectable", ctx_r3.selectable);
} }
function PluginListComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, PluginListComponent_ng_container_1_ng_container_1_Template, 2, 2, "ng-container", 3);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r0.plugins$));
} }
function PluginListComponent_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelement(1, "h1", 6);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 1, ctx_r2.emptyListText), " ");
} }
export class PluginListComponent {
    constructor() {
        this.emptyListText = '';
        this.selectedItems = new EventEmitter();
        this.selectedPlugins = {};
    }
    updateSelectedItems(value, name) {
        this.selectedPlugins[name] = value;
        this.selectedItems.emit(Object.keys(this.selectedPlugins).filter((k) => this.selectedPlugins[k]));
    }
}
PluginListComponent.ɵfac = function PluginListComponent_Factory(t) { return new (t || PluginListComponent)(); };
PluginListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PluginListComponent, selectors: [["c8y-plugin-list"]], inputs: { emptyListText: "emptyListText", plugins$: "plugins$", selectable: "selectable" }, outputs: { selectedItems: "selectedItems" }, decls: 5, vars: 4, consts: [[1, "bg-inherit"], [4, "ngIf", "ngIfElse"], ["emptyList", ""], [4, "ngFor", "ngForOf"], [3, "plugin", "selectable", "isItemSelected"], [1, "c8y-empty-state", "text-left"], ["c8yIcon", "plugin"]], template: function PluginListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 0);
        ɵngcc0.ɵɵtemplate(1, PluginListComponent_ng_container_1_Template, 3, 3, "ng-container", 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, PluginListComponent_ng_template_3_Template, 5, 3, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(4);
        let tmp_0_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 2, ctx.plugins$)) == null ? null : tmp_0_0.length) !== 0)("ngIfElse", _r1);
    } }, directives: [ɵngcc1.ListGroupComponent, ɵngcc2.NgIf, ɵngcc2.NgForOf, ɵngcc3.PluginListItemComponent, ɵngcc1.IconDirective], pipes: [ɵngcc2.AsyncPipe, ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
PluginListComponent.propDecorators = {
    plugins$: [{ type: Input }],
    emptyListText: [{ type: Input }],
    selectable: [{ type: Input }],
    selectedItems: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PluginListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-plugin-list',
                template: "<c8y-list-group class=\"bg-inherit\">\n  <ng-container *ngIf=\"(plugins$ | async)?.length !== 0; else emptyList\">\n    <ng-container *ngFor=\"let plugin of plugins$ | async\">\n      <c8y-plugin-list-item\n        (isItemSelected)=\"updateSelectedItems($event, plugin.id)\"\n        [plugin]=\"plugin\"\n        [selectable]=\"selectable\"\n      ></c8y-plugin-list-item>\n    </ng-container>\n  </ng-container>\n</c8y-list-group>\n<ng-template #emptyList>\n  <div class=\"c8y-empty-state text-left\">\n    <h1 c8yIcon=\"plugin\"></h1>\n    <p>\n      {{ emptyListText | translate }}\n    </p>\n  </div>\n</ng-template>\n"
            }]
    }], function () { return []; }, { emptyListText: [{
            type: Input
        }], selectedItems: [{
            type: Output
        }], plugins$: [{
            type: Input
        }], selectable: [{
            type: Input
        }] }); })();
