// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-bulk-operation-stepper.js
import { Directive, TemplateRef, Input, EventEmitter, Component, Output, ContentChildren, ViewChild, NgModule } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { __awaiter } from 'tslib';
import { gettext, Status, ModalService, AlertService, C8yStepper, CoreModule } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { OperationDetailsComponent, OperationDetailsModule } from '@c8y/ngx-components/operations/details';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { DeviceSelectorModule } from '@c8y/ngx-components/operations/device-selector';
import { OperationSummaryModule } from '@c8y/ngx-components/operations/operation-summary';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/cdk/stepper';
import * as ɵngcc5 from '@c8y/ngx-components/operations/device-selector';
import * as ɵngcc6 from '@c8y/ngx-components/operations/operation-summary';
import * as ɵngcc7 from '@c8y/ngx-components/operations/details';

function BulkOperationStepper_c8y_stepper_1_cdk_step_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function BulkOperationStepper_c8y_stepper_1_cdk_step_1_c8y_stepper_buttons_3_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper-buttons", 23);
    ɵngcc0.ɵɵlistener("onNext", function BulkOperationStepper_c8y_stepper_1_cdk_step_1_c8y_stepper_buttons_3_Template_c8y_stepper_buttons_onNext_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const step_r4 = ɵngcc0.ɵɵnextContext().$implicit; return step_r4.onNext($event); })("onCancel", function BulkOperationStepper_c8y_stepper_1_cdk_step_1_c8y_stepper_buttons_3_Template_c8y_stepper_buttons_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r10 = ɵngcc0.ɵɵnextContext(3); return ctx_r10.cancel(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r4 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("disabled", step_r4.buttonsDisabled);
} }
function BulkOperationStepper_c8y_stepper_1_cdk_step_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "cdk-step", 20);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, BulkOperationStepper_c8y_stepper_1_cdk_step_1_ng_container_2_Template, 1, 0, "ng-container", 21);
    ɵngcc0.ɵɵtemplate(3, BulkOperationStepper_c8y_stepper_1_cdk_step_1_c8y_stepper_buttons_3_Template, 1, 1, "c8y-stepper-buttons", 22);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r4 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(1, 4, step_r4.label))("completed", step_r4.completed);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", step_r4.templateRef);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showButtons);
} }
function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_12_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper-buttons", 24);
    ɵngcc0.ɵɵlistener("onNext", function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_12_Template_c8y_stepper_buttons_onNext_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.confirmDeviceSelection($event); })("onCancel", function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_12_Template_c8y_stepper_buttons_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r14 = ɵngcc0.ɵɵnextContext(2); return ctx_r14.cancel(); });
    ɵngcc0.ɵɵelementEnd();
} }
function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_31_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper-buttons", 25);
    ɵngcc0.ɵɵlistener("onCancel", function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_31_Template_c8y_stepper_buttons_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.cancel(); })("onCustom", function BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_31_Template_c8y_stepper_buttons_onCustom_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.scheduleBulkOperation(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("labels", ctx_r3.stepperButtonsLabels)("pending", ctx_r3.pendingStatus)("disabled", ctx_r3.operationDetailsForm == null ? null : ctx_r3.operationDetailsForm.invalid);
} }
const _c0 = function () { return { edit: true, done: false }; };
const _c1 = function () { return ["col-md-6", "col-md-offset-3", "m-t-24", "m-b-40", "p-0", "flex-no-shrink"]; };
function BulkOperationStepper_c8y_stepper_1_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 2);
    ɵngcc0.ɵɵtemplate(1, BulkOperationStepper_c8y_stepper_1_cdk_step_1_Template, 4, 6, "cdk-step", 3);
    ɵngcc0.ɵɵelementStart(2, "cdk-step", 4);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵelementStart(5, "div", 6);
    ɵngcc0.ɵɵelementStart(6, "div", 7);
    ɵngcc0.ɵɵelementStart(7, "h4", 8);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "div", 9);
    ɵngcc0.ɵɵelementStart(11, "c8y-device-selector", 10);
    ɵngcc0.ɵɵlistener("onDeviceQueryStringChange", function BulkOperationStepper_c8y_stepper_1_Template_c8y_device_selector_onDeviceQueryStringChange_11_listener($event) { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.deviceQueryString = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(12, BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_12_Template, 1, 0, "c8y-stepper-buttons", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "cdk-step", 4);
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementStart(15, "div", 5);
    ɵngcc0.ɵɵelementStart(16, "div", 6);
    ɵngcc0.ɵɵelementStart(17, "div", 7);
    ɵngcc0.ɵɵelementStart(18, "h4", 8);
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 9);
    ɵngcc0.ɵɵelementStart(22, "div", 12);
    ɵngcc0.ɵɵelementStart(23, "div", 13);
    ɵngcc0.ɵɵelementStart(24, "div", 14);
    ɵngcc0.ɵɵelement(25, "c8y-operation-summary", 15);
    ɵngcc0.ɵɵpipe(26, "translate");
    ɵngcc0.ɵɵpipe(27, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "div", 16);
    ɵngcc0.ɵɵelementStart(29, "div", 17);
    ɵngcc0.ɵɵelement(30, "c8y-operation-details", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(31, BulkOperationStepper_c8y_stepper_1_c8y_stepper_buttons_31_Template, 1, 3, "c8y-stepper-buttons", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(26, _c0))("customClasses", ɵngcc0.ɵɵpureFunction0(27, _c1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.steps);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(3, 14, "Select target devices"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 16, "Select target devices"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("deviceTypes", ctx_r0.deviceTypes$);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.showButtons);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(14, 18, "Confirm and schedule bulk operation"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 20, "Confirm and schedule bulk operation"), " ");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("name", ɵngcc0.ɵɵpipeBind1(26, 22, ctx_r0.operationDetails == null ? null : ctx_r0.operationDetails.name))("description", ɵngcc0.ɵɵpipeBind1(27, 24, ctx_r0.operationDetails == null ? null : ctx_r0.operationDetails.description))("deviceQueryString", ctx_r0.deviceQueryString);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("bulkOperationType", ctx_r0.bulkOperationType);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.showButtons);
} }
class CustomStep {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.buttonsDisabled = false;
        this.onNext = ({ stepper, step }) => {
            // steps without own `onNext` handler, e.g. preview steps need to mark themselves as `completed`,
            // otherwise stepper will not allow to move forth from them as soon as the user navigates back
            // and the `c8y-stepper-buttons` component marks the step as incomplete.
            step.completed = true;
            stepper.next();
        };
    }
}
CustomStep.ɵfac = function CustomStep_Factory(t) { return new (t || CustomStep)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
CustomStep.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CustomStep, selectors: [["", "customStep", ""]], inputs: { buttonsDisabled: ["customStepButtonsDisabled", "buttonsDisabled"], onNext: ["customStepOnNext", "onNext"], label: ["customStep", "label"], completed: ["customStepCompleted", "completed"] } });
CustomStep.ctorParameters = () => [
    { type: TemplateRef }
];
CustomStep.propDecorators = {
    label: [{ type: Input, args: ['customStep',] }],
    completed: [{ type: Input, args: ['customStepCompleted',] }],
    buttonsDisabled: [{ type: Input, args: ['customStepButtonsDisabled',] }],
    onNext: [{ type: Input, args: ['customStepOnNext',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CustomStep, [{
        type: Directive,
        args: [{ selector: '[customStep]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { buttonsDisabled: [{
            type: Input,
            args: ['customStepButtonsDisabled']
        }], onNext: [{
            type: Input,
            args: ['customStepOnNext']
        }], label: [{
            type: Input,
            args: ['customStep']
        }], completed: [{
            type: Input,
            args: ['customStepCompleted']
        }] }); })();

class BulkOperationStepper {
    constructor(bulkOperationService, modal, alert) {
        this.bulkOperationService = bulkOperationService;
        this.modal = modal;
        this.alert = alert;
        this.selectionChange = new EventEmitter();
        this.steps = [];
        this.showStepper = false;
        this.showButtons = false;
        this.stepperButtonsLabels = { custom: gettext('Schedule bulk operation') };
        this.deviceTypesSubject$ = new Subject();
        this.endSubscriptions = new Subject();
        this.deviceTypes$ = this.deviceTypesSubject$.asObservable();
    }
    ngAfterViewInit() {
        setTimeout(() => {
            // wait for the next event loop turn as `steps` has already been checked in this CD cycle
            this.steps = this.customSteps.toArray();
            this.showStepper = true;
            setTimeout(() => {
                // postpone rendering of buttons for custom steps to the point where custom steps have already been rendered
                this.showButtons = true;
                if (this.stepper) {
                    this.stepper.selectionChange.pipe(takeUntil(this.endSubscriptions)).subscribe(event => {
                        this.selectionChange.next(event);
                    });
                    this.operationDetailsForm = this.operationDetailsComponent.fgOperationDescription;
                }
            });
        });
    }
    changeDeviceTypes(deviceTypes) {
        if (deviceTypes) {
            this.deviceTypesSubject$.next(Array.isArray(deviceTypes) ? deviceTypes : [deviceTypes]);
        }
        else {
            this.deviceTypesSubject$.next([]);
        }
    }
    confirmDeviceSelection($event) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.deviceQueryString) {
                try {
                    yield this.modal.confirm(gettext('All devices selected'), gettext('You are about to schedule the bulk operation to be executed for all devices. Do you want to proceed?'), Status.WARNING, { ok: gettext('Schedule for all devices'), cancel: gettext('Cancel and select devices') });
                    $event.step.completed = true;
                    $event.stepper.next();
                    this.operationDetails = this.retrieveOperationDetails
                        ? yield this.retrieveOperationDetails()
                        : undefined;
                }
                catch (ex) {
                    // Intentionally empty
                }
            }
            else {
                $event.step.completed = true;
                $event.stepper.next();
                this.operationDetails = this.retrieveOperationDetails
                    ? yield this.retrieveOperationDetails()
                    : undefined;
            }
            this.bulkOperationType = this.bulkOperationService.retrieveBulkOperationType(get(this.operationDetails, 'prototype'));
            if (this.operationDetailsForm &&
                get(this.operationDetailsForm, 'controls.description.pristine') &&
                this.operationDetails) {
                this.operationDetailsForm.patchValue({
                    description: get(this.operationDetails, 'prototype.description')
                });
            }
        });
    }
    cancel() {
        this.close();
    }
    scheduleBulkOperation() {
        return __awaiter(this, void 0, void 0, function* () {
            this.pendingStatus = true;
            try {
                this.operationDetails.prototype.description = get(this.operationDetailsForm, 'controls.description.value');
                this.operationDetails.note = get(this.operationDetailsForm, 'controls.note.value');
                this.operationDetails.schedule = get(this.operationDetailsForm, 'controls.schedule.value');
                yield this.bulkOperationService.scheduleBulkOperation(this.deviceQueryString, this.operationDetails);
                this.alert.success(gettext('New bulk operation scheduled.'));
                this.close();
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
            this.pendingStatus = false;
        });
    }
    ngOnDestroy() {
        this.endSubscriptions.next();
        this.endSubscriptions.complete();
    }
    close() {
        this.stepper.reset();
        this.bulkOperationService.returnToBulkOperationOverview();
    }
}
BulkOperationStepper.ɵfac = function BulkOperationStepper_Factory(t) { return new (t || BulkOperationStepper)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
BulkOperationStepper.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationStepper, selectors: [["c8y-bulk-operation-stepper"]], contentQueries: function BulkOperationStepper_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, CustomStep, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.customSteps = _t);
    } }, viewQuery: function BulkOperationStepper_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 5);
        ɵngcc0.ɵɵviewQuery(OperationDetailsComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationDetailsComponent = _t.first);
    } }, outputs: { selectionChange: "selectionChange" }, decls: 2, vars: 1, consts: [[1, "fit-h"], ["class", "flex-col no-align-items fit-h c8y-stepper--no-btns", "linear", "", 3, "disableDefaultIcons", "customClasses", 4, "ngIf"], ["linear", "", 1, "flex-col", "no-align-items", "fit-h", "c8y-stepper--no-btns", 3, "disableDefaultIcons", "customClasses"], [3, "label", "completed", 4, "ngFor", "ngForOf"], [3, "label"], [1, "card-block", "p-t-0", "flex-no-shrink", "separator-bottom", "col-xs-12"], [1, "row", "p-b-16"], [1, "col-md-6", "col-md-offset-3", "col-lg-4", "col-lg-offset-4"], [1, "text-center", "m-b-16"], [1, "col-xs-12", "flex-grow", "no-gutter"], [3, "deviceTypes", "onDeviceQueryStringChange"], ["class", "d-block card-footer p-24 separator", 3, "onNext", "onCancel", 4, "ngIf"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-b-0"], [1, "text-center", "p-b-16"], [3, "name", "description", "deviceQueryString"], [1, "row"], [1, "col-md-4", "col-md-offset-4"], [3, "bulkOperationType"], ["class", "d-block card-footer p-24 separator", 3, "labels", "pending", "disabled", "onCancel", "onCustom", 4, "ngIf"], [3, "label", "completed"], [4, "ngTemplateOutlet"], ["class", "d-block card-footer p-24 separator", 3, "disabled", "onNext", "onCancel", 4, "ngIf"], [1, "d-block", "card-footer", "p-24", "separator", 3, "disabled", "onNext", "onCancel"], [1, "d-block", "card-footer", "p-24", "separator", 3, "onNext", "onCancel"], [1, "d-block", "card-footer", "p-24", "separator", 3, "labels", "pending", "disabled", "onCancel", "onCustom"]], template: function BulkOperationStepper_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, BulkOperationStepper_c8y_stepper_1_Template, 32, 28, "c8y-stepper", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showStepper);
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.C8yStepper, ɵngcc3.NgForOf, ɵngcc4.CdkStep, ɵngcc5.DeviceSelectorComponent, ɵngcc6.OperationSummaryComponent, ɵngcc7.OperationDetailsComponent, ɵngcc3.NgTemplateOutlet, ɵngcc2.C8yStepperButtons], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
BulkOperationStepper.ctorParameters = () => [
    { type: BulkOperationsService },
    { type: ModalService },
    { type: AlertService }
];
BulkOperationStepper.propDecorators = {
    selectionChange: [{ type: Output }],
    customSteps: [{ type: ContentChildren, args: [CustomStep,] }],
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: false },] }],
    operationDetailsComponent: [{ type: ViewChild, args: [OperationDetailsComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationStepper, [{
        type: Component,
        args: [{
                selector: 'c8y-bulk-operation-stepper',
                template: "<div class=\"fit-h\">\n  <c8y-stepper\n    class=\"flex-col no-align-items fit-h c8y-stepper--no-btns\"\n    linear\n    [disableDefaultIcons]=\"{ edit: true, done: false }\"\n    [customClasses]=\"['col-md-6', 'col-md-offset-3', 'm-t-24', 'm-b-40', 'p-0', 'flex-no-shrink']\"\n    *ngIf=\"showStepper\"\n  >\n    <!-- CUSTOM STEPS 1 to N-2 -->\n    <cdk-step\n      *ngFor=\"let step of steps\"\n      [label]=\"step.label | translate\"\n      [completed]=\"step.completed\"\n    >\n      <ng-container *ngTemplateOutlet=\"step.templateRef\"></ng-container>\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        *ngIf=\"showButtons\"\n        [disabled]=\"step.buttonsDisabled\"\n        (onNext)=\"step.onNext($event)\"\n        (onCancel)=\"cancel()\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n    <!-- STEP N-1 - Data-grid -->\n    <cdk-step [label]=\"'Select target devices' | translate\">\n      <div class=\"card-block p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row p-b-16\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center m-b-16\">\n              {{ 'Select target devices' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"col-xs-12 flex-grow no-gutter\">\n        <c8y-device-selector\n          [deviceTypes]=\"deviceTypes$\"\n          (onDeviceQueryStringChange)=\"deviceQueryString = $event\"\n        ></c8y-device-selector>\n      </div>\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        *ngIf=\"showButtons\"\n        (onNext)=\"confirmDeviceSelection($event)\"\n        (onCancel)=\"cancel()\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n\n    <!-- STEP N - Scheduler -->\n    <cdk-step [label]=\"'Confirm and schedule bulk operation' | translate\">\n      <div class=\"card-block p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row p-b-16\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center m-b-16\">\n              {{ 'Confirm and schedule bulk operation' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"col-xs-12 flex-grow no-gutter\">\n        <div class=\"card-inner-scroll fit-h\">\n          <div class=\"card-block p-b-0\">\n            <div class=\"text-center p-b-16\">\n              <c8y-operation-summary\n                [name]=\"operationDetails?.name | translate\"\n                [description]=\"operationDetails?.description | translate\"\n                [deviceQueryString]=\"deviceQueryString\"\n              ></c8y-operation-summary>\n            </div>\n            <div class=\"row\">\n              <div class=\"col-md-4 col-md-offset-4\">\n                <c8y-operation-details\n                  [bulkOperationType]=\"bulkOperationType\"\n                ></c8y-operation-details>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        *ngIf=\"showButtons\"\n        [labels]=\"stepperButtonsLabels\"\n        [pending]=\"pendingStatus\"\n        [disabled]=\"operationDetailsForm?.invalid\"\n        (onCancel)=\"cancel()\"\n        (onCustom)=\"scheduleBulkOperation()\"\n      >\n      </c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkOperationsService }, { type: ɵngcc2.ModalService }, { type: ɵngcc2.AlertService }]; }, { selectionChange: [{
            type: Output
        }], customSteps: [{
            type: ContentChildren,
            args: [CustomStep]
        }], stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: false }]
        }], operationDetailsComponent: [{
            type: ViewChild,
            args: [OperationDetailsComponent, { static: false }]
        }] }); })();

class BaseStepperComponent {
    constructor() {
        /**
         * A map holding step data. The order of properties need to match the order of
         * the steps they hold data for as the index of the property is used to clear
         * step data when navigating forth after changing data at an earlier step.
         */
        this.stepData = {};
        this.endSubscriptions = new Subject();
    }
    set deviceTypes(deviceTypes) {
        if (this.operationStepper) {
            this.operationStepper.changeDeviceTypes(deviceTypes);
        }
    }
    ngOnInit() {
        this.operationStepper.retrieveOperationDetails = this.retrieveOperationPrototype.bind(this);
        this.operationStepper.selectionChange
            .pipe(takeUntil(this.endSubscriptions))
            .subscribe(this.onSelectionChange.bind(this));
    }
    ngOnDestroy() {
        this.endSubscriptions.next();
        this.endSubscriptions.complete();
    }
    onSelectionChange(event) {
        const { selectedIndex, previouslySelectedIndex } = event;
        if (selectedIndex > previouslySelectedIndex &&
            selectedIndex < Object.keys(this.stepData).length) {
            // TODO clear step data only if previous step is "dirty"
            this.stepData[this.getStepDataKeyByIndex(selectedIndex)] = undefined;
        }
    }
    getStepDataKeyByIndex(index) {
        return Object.keys(this.stepData)[index];
    }
}
BaseStepperComponent.ɵfac = function BaseStepperComponent_Factory(t) { return new (t || BaseStepperComponent)(); };
BaseStepperComponent.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: BaseStepperComponent, viewQuery: function BaseStepperComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(BulkOperationStepper, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.operationStepper = _t.first);
    } } });
BaseStepperComponent.propDecorators = {
    operationStepper: [{ type: ViewChild, args: [BulkOperationStepper, { static: true },] }]
};


/**
 * This module provides base stepper class and stepper wrapper component.
 */
class BulkOperationStepperModule {
}
BulkOperationStepperModule.ɵfac = function BulkOperationStepperModule_Factory(t) { return new (t || BulkOperationStepperModule)(); };
BulkOperationStepperModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationStepperModule });
BulkOperationStepperModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, DeviceSelectorModule, OperationDetailsModule, OperationSummaryModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationStepperModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, DeviceSelectorModule, OperationDetailsModule, OperationSummaryModule],
                declarations: [BulkOperationStepper, CustomStep],
                exports: [BulkOperationStepper, CustomStep]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationStepperModule, { declarations: function () { return [BulkOperationStepper, CustomStep]; }, imports: function () { return [CoreModule, DeviceSelectorModule, OperationDetailsModule, OperationSummaryModule]; }, exports: function () { return [BulkOperationStepper, CustomStep]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { BaseStepperComponent, BulkOperationStepper, BulkOperationStepperModule, CustomStep };
