// @c8y/ngx-components/fesm2015/c8y-ngx-components-operations-operation-summary.js
import { Component, Input, NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { InventoryService } from '@c8y/client';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@c8y/ngx-components';

function OperationSummaryComponent_span_11_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1, "No devices selected");
    ɵngcc0.ɵɵelementEnd();
} }
function OperationSummaryComponent_span_11_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1, "1 device selected");
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { deviceCount: a0 }; };
function OperationSummaryComponent_span_11_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 10);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, " {{ deviceCount }} devices selected");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c0, ctx_r3.deviceCount));
} }
function OperationSummaryComponent_span_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 5);
    ɵngcc0.ɵɵtemplate(1, OperationSummaryComponent_span_11_ng_template_1_Template, 2, 0, "ng-template", 6);
    ɵngcc0.ɵɵtemplate(2, OperationSummaryComponent_span_11_ng_template_2_Template, 2, 0, "ng-template", 7);
    ɵngcc0.ɵɵtemplate(3, OperationSummaryComponent_span_11_ng_template_3_Template, 2, 3, "ng-template", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngPlural", ctx_r0.deviceCount);
} }
class OperationSummaryComponent {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
        this.deviceCount = 0;
        this.displayDeviceCount = false;
    }
    set deviceQueryString(query) {
        this.displayDeviceCount = true;
        const filters = {
            q: query || '',
            pageSize: 1,
            currentPage: 1,
            withTotalPages: true
        };
        this.inventoryService
            .list(filters)
            .then((result) => (this.deviceCount = result.paging.totalPages));
    }
}
OperationSummaryComponent.ɵfac = function OperationSummaryComponent_Factory(t) { return new (t || OperationSummaryComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.InventoryService)); };
OperationSummaryComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OperationSummaryComponent, selectors: [["c8y-operation-summary"]], inputs: { deviceQueryString: "deviceQueryString", name: "name", description: "description" }, decls: 12, vars: 10, consts: [[1, "row"], [1, "col-md-12", "text-center"], [1, "col-md-4", "col-md-offset-4", "text-center"], [1, "m-t-16", "m-b-8", "text-truncate", 3, "title"], ["class", "text-muted", 3, "ngPlural", 4, "ngIf"], [1, "text-muted", 3, "ngPlural"], ["ngPluralCase", "=0"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], ["translate", ""], ["translate", "", 3, "translateParams"]], template: function OperationSummaryComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "p");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 0);
        ɵngcc0.ɵɵelementStart(6, "div", 2);
        ɵngcc0.ɵɵelementStart(7, "h4", 3);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, OperationSummaryComponent_span_11_Template, 4, 1, "span", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, ctx.name));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(8, 6, ctx.description));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(10, 8, ctx.description));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.displayDeviceCount && ctx.deviceCount !== undefined && ctx.deviceCount !== null);
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgPlural, ɵngcc2.NgPluralCase, ɵngcc3.C8yTranslateDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
OperationSummaryComponent.ctorParameters = () => [
    { type: InventoryService }
];
OperationSummaryComponent.propDecorators = {
    name: [{ type: Input }],
    description: [{ type: Input }],
    deviceQueryString: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationSummaryComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-operation-summary',
                template: "<div class=\"row\">\n  <div class=\"col-md-12 text-center\">\n    <p>{{ name | translate }}</p>\n    <div class=\"row\">\n      <div class=\"col-md-4 col-md-offset-4 text-center\">\n        <h4 class=\"m-t-16 m-b-8 text-truncate\" title=\"{{ description | translate }}\">{{ description | translate }}</h4>\n      </div>\n    </div>\n    <span\n      class=\"text-muted\"\n      [ngPlural]=\"deviceCount\"\n      *ngIf=\"displayDeviceCount && deviceCount !== undefined && deviceCount !== null\"\n    >\n      <ng-template ngPluralCase=\"=0\">\n        <span translate>No devices selected</span>\n      </ng-template>\n      <ng-template ngPluralCase=\"=1\">\n        <span translate>1 device selected</span>\n      </ng-template>\n      <ng-template ngPluralCase=\"other\">\n        <span translate ngNonBindable [translateParams]=\"{ deviceCount: deviceCount }\">\n          {{ deviceCount }} devices selected</span\n        >\n      </ng-template>\n    </span>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.InventoryService }]; }, { deviceQueryString: [{
            type: Input
        }], name: [{
            type: Input
        }], description: [{
            type: Input
        }] }); })();

/**
 * This module provides a component for displaying the summary of given bulk operation.
 */
class OperationSummaryModule {
}
OperationSummaryModule.ɵfac = function OperationSummaryModule_Factory(t) { return new (t || OperationSummaryModule)(); };
OperationSummaryModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationSummaryModule });
OperationSummaryModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationSummaryModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule],
                declarations: [OperationSummaryComponent],
                entryComponents: [OperationSummaryComponent],
                exports: [OperationSummaryComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationSummaryModule, { declarations: function () { return [OperationSummaryComponent]; }, imports: function () { return [CoreModule]; }, exports: function () { return [OperationSummaryComponent]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { OperationSummaryComponent, OperationSummaryModule };
