// @c8y/ngx-components/esm2015/ecosystem/applications/application-list/application-list.component.js
import { Component } from '@angular/core';
import { WizardService } from '@c8y/ngx-components';
import { BehaviorSubject } from 'rxjs';
import { shareReplay, switchMap, tap } from 'rxjs/operators';
import { EcosystemService } from './../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './../../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../../shared/application-card.component';

function ApplicationListComponent_div_25_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelement(1, "h1", 13);
    ɵngcc0.ɵɵelementStart(2, "h3", 14);
    ɵngcc0.ɵɵtext(3, "No applications to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 14);
    ɵngcc0.ɵɵtext(5, "Add your first application by clicking below.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "p");
    ɵngcc0.ɵɵelementStart(7, "button", 15);
    ɵngcc0.ɵɵlistener("click", function ApplicationListComponent_div_25_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.addApplication(); });
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(8, 2, "Add application"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 4, "Add application"), " ");
} }
function ApplicationListComponent_div_28_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "div", 17);
    ɵngcc0.ɵɵelement(2, "div", 18);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 19);
    ɵngcc0.ɵɵelementStart(6, "div", 20);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 21);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(12, "div", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, "Application"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 5, "Description"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 7, "Type"));
} }
function ApplicationListComponent_div_30_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelementStart(1, "c8y-application-card", 23);
    ɵngcc0.ɵɵlistener("onAppDeleted", function ApplicationListComponent_div_30_Template_c8y_application_card_onAppDeleted_1_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.loadApplications(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const app_r5 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("app", app_r5);
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
export class ApplicationListComponent {
    constructor(ecosystemService, wizardService) {
        this.ecosystemService = ecosystemService;
        this.wizardService = wizardService;
        this.reloading = false;
        this.reload$ = new BehaviorSubject(null);
        this.apps$ = this.reload$.pipe(tap(() => (this.reloading = true)), switchMap(() => this.ecosystemService.getWebApplications()), tap(() => (this.reloading = false)), shareReplay());
    }
    ngOnInit() {
        this.loadApplications();
    }
    loadApplications() {
        this.reload$.next();
    }
    addApplication() {
        const wizardConfig = {
            headerText: 'Add Application',
            headerIcon: 'c8y-atom'
        };
        const initialState = {
            wizardConfig,
            id: 'uploadApplication'
        };
        const modalOptions = { initialState };
        const modalRef = this.wizardService.show(modalOptions);
        modalRef.content.onClose.subscribe(() => {
            this.loadApplications();
        });
    }
}
ApplicationListComponent.ɵfac = function ApplicationListComponent_Factory(t) { return new (t || ApplicationListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.WizardService)); };
ApplicationListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ApplicationListComponent, selectors: [["c8y-application-list"]], decls: 32, vars: 47, consts: [[3, "icon", "label"], [3, "icon", "label", "path"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "plus-circle"], ["c8yIcon", "refresh", 3, "ngClass"], ["itemClass", "navbar-form hidden-xs", 3, "placement"], [3, "listLength", "onListClassChange"], ["class", "c8y-empty-state text-center", 4, "ngIf"], [1, "card-group", 3, "ngClass"], ["class", "page-sticky-header hidden-xs flex-row", 4, "ngIf"], ["class", "col-xs-12 col-sm-4 col-md-3", 4, "ngFor", "ngForOf"], [1, "c8y-empty-state", "text-center"], [1, "c8y-icon", "c8y-icon-modules", "c8y-icon-duocolor"], ["translate", ""], [1, "btn", "btn-primary", 3, "title", "click"], [1, "page-sticky-header", "hidden-xs", "flex-row"], [1, "card-block", "card-column-40"], [1, "card-appicon", "p-l-32", "p-r-16", "m-r-0", "m-l-4"], [1, "card-block", "p-0", "card-column-80"], [1, "card-block", "card-column-80"], [1, "card-block", "card-column-20"], [1, "col-xs-12", "col-sm-4", "col-md-3"], [1, "d-contents", 3, "app", "onAppDeleted"]], template: function ApplicationListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(4, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelement(8, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "c8y-action-bar-item", 2);
        ɵngcc0.ɵɵelementStart(11, "button", 3);
        ɵngcc0.ɵɵlistener("click", function ApplicationListComponent_Template_button_click_11_listener() { return ctx.addApplication(); });
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelement(13, "i", 4);
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-action-bar-item", 2);
        ɵngcc0.ɵɵelementStart(17, "button", 3);
        ɵngcc0.ɵɵlistener("click", function ApplicationListComponent_Template_button_click_17_listener() { return ctx.loadApplications(); });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelement(19, "i", 5);
        ɵngcc0.ɵɵtext(20);
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "c8y-action-bar-item", 6);
        ɵngcc0.ɵɵelementStart(23, "c8y-list-display-switch", 7);
        ɵngcc0.ɵɵlistener("onListClassChange", function ApplicationListComponent_Template_c8y_list_display_switch_onListClassChange_23_listener($event) { return ctx.listClass = $event; });
        ɵngcc0.ɵɵpipe(24, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(25, ApplicationListComponent_div_25_Template, 11, 6, "div", 8);
        ɵngcc0.ɵɵpipe(26, "async");
        ɵngcc0.ɵɵelementStart(27, "div", 9);
        ɵngcc0.ɵɵtemplate(28, ApplicationListComponent_div_28_Template, 13, 9, "div", 10);
        ɵngcc0.ɵɵpipe(29, "async");
        ɵngcc0.ɵɵtemplate(30, ApplicationListComponent_div_30_Template, 2, 1, "div", 11);
        ɵngcc0.ɵɵpipe(31, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_16_0;
        let tmp_17_0;
        let tmp_19_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 21, "All applications"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(5, 23, "Ecosystem"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(7, 25, "Applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(9, 27, "All applications"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 29, "Add application"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 31, "Add application"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 33, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(45, _c0, ctx.reloading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(21, 35, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "left");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("listLength", (tmp_16_0 = ɵngcc0.ɵɵpipeBind1(24, 37, ctx.apps$)) == null ? null : tmp_16_0.length);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_17_0 = ɵngcc0.ɵɵpipeBind1(26, 39, ctx.apps$)) == null ? null : tmp_17_0.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.listClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_19_0 = ɵngcc0.ɵɵpipeBind1(29, 41, ctx.apps$)) == null ? null : tmp_19_0.length) > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(31, 43, ctx.apps$));
    } }, directives: [ɵngcc2.TitleComponent, ɵngcc2.BreadcrumbComponent, ɵngcc2.BreadcrumbItemComponent, ɵngcc2.ActionBarItemComponent, ɵngcc2.IconDirective, ɵngcc3.NgClass, ɵngcc2.ɵbh, ɵngcc3.NgIf, ɵngcc3.NgForOf, ɵngcc2.C8yTranslateDirective, ɵngcc4.ApplicationCardComponent], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.AsyncPipe], encapsulation: 2 });
ApplicationListComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: WizardService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApplicationListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-application-list',
                template: "<c8y-title>{{ 'All applications' | translate }}</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'All applications' | translate\"\n  ></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    (click)=\"addApplication()\"\n    class=\"btn btn-link\"\n    title=\"{{ 'Add application' | translate }}\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add application' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button (click)=\"loadApplications()\" class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\">\n    <i [ngClass]=\"{ 'icon-spin': reloading }\" c8yIcon=\"refresh\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form hidden-xs\">\n  <c8y-list-display-switch\n    (onListClassChange)=\"listClass = $event\"\n    [listLength]=\"(apps$ | async)?.length\"\n  ></c8y-list-display-switch>\n</c8y-action-bar-item>\n\n<div *ngIf=\"(apps$ | async)?.length === 0\" class=\"c8y-empty-state text-center\">\n  <h1 class=\"c8y-icon c8y-icon-modules c8y-icon-duocolor\"></h1>\n  <h3 translate>No applications to display.</h3>\n  <p translate>Add your first application by clicking below.</p>\n  <p>\n    <button\n      (click)=\"addApplication()\"\n      class=\"btn btn-primary\"\n      title=\"{{ 'Add application' | translate }}\"\n    >\n      {{ 'Add application' | translate }}\n    </button>\n  </p>\n</div>\n\n<div [ngClass]=\"listClass\" class=\"card-group\">\n  <div class=\"page-sticky-header hidden-xs flex-row\" *ngIf=\"(apps$ | async)?.length > 0\">\n    <div class=\"card-block card-column-40\">\n      <div class=\"card-appicon p-l-32 p-r-16 m-r-0 m-l-4\"></div>\n      {{ 'Application' | translate }}\n    </div>\n    <div class=\"card-block p-0 card-column-80\">\n      <div class=\"card-block card-column-80\">{{ 'Description' | translate }}</div>\n      <div class=\"card-block card-column-20\">{{ 'Type' | translate }}</div>\n      <div class=\"card-block card-column-20\"></div>\n    </div>\n  </div>\n  <div *ngFor=\"let app of apps$ | async\" class=\"col-xs-12 col-sm-4 col-md-3\">\n    <c8y-application-card\n      (onAppDeleted)=\"loadApplications()\"\n      [app]=\"app\"\n      class=\"d-contents\"\n    ></c8y-application-card>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.WizardService }]; }, null); })();
