// @c8y/ngx-components/esm2015/core/select/select.module.js
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { SelectComponent } from './select.component';
import { CommonModule } from '../common/common.module';
import { TypeaheadComponent } from './typeahead.component';
import { ListGroupModule } from '../list-group/list-group.module';
import { FormsModule } from '../forms/forms.module';
/**
 * The angular module definition for a select.
 * @exports SelectComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
export class SelectModule {
}
SelectModule.ɵfac = function SelectModule_Factory(t) { return new (t || SelectModule)(); };
SelectModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SelectModule });
SelectModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            BsDropdownModule.forRoot(),
            CommonModule,
            ListGroupModule,
            FormsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BsDropdownModule.forRoot(),
                    CommonModule,
                    ListGroupModule,
                    FormsModule
                ],
                exports: [
                    SelectComponent,
                    TypeaheadComponent
                ],
                declarations: [
                    SelectComponent,
                    TypeaheadComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SelectModule, { declarations: function () { return [SelectComponent,
        TypeaheadComponent]; }, imports: function () { return [ɵngcc1.BsDropdownModule, CommonModule,
        ListGroupModule,
        FormsModule]; }, exports: function () { return [SelectComponent,
        TypeaheadComponent]; } }); })();
