// @c8y/ngx-components/esm2015/core/user/user-menu-item.component.js
import { Component, Input, ViewChild, EventEmitter, Output } from '@angular/core';
import { UserMenuService } from './user-menu.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './user-menu.service';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';

const _c0 = ["template"];
function UserMenuItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 1);
    ɵngcc0.ɵɵlistener("click", function UserMenuItemComponent_ng_template_0_Template_li_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.onClick(); });
    ɵngcc0.ɵɵelementStart(1, "a", 2);
    ɵngcc0.ɵɵelement(2, "i", 3);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵprojection(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵattribute("href", ctx_r1.link, ɵngcc0.ɵɵsanitizeUrl)("target", ctx_r1.target);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r1.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r1.label), " ");
} }
const _c1 = ["*"];
export class UserMenuItemComponent {
    constructor(userService) {
        this.userService = userService;
        this.priority = 0;
        this.click = new EventEmitter();
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.userService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.userService.remove(this);
    }
    onClick() {
        this.click.emit(this);
    }
}
UserMenuItemComponent.ɵfac = function UserMenuItemComponent_Factory(t) { return new (t || UserMenuItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UserMenuService)); };
UserMenuItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UserMenuItemComponent, selectors: [["c8y-user-menu-item"]], viewQuery: function UserMenuItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { priority: "priority", icon: "icon", label: "label", link: "link", target: "target" }, outputs: { click: "click" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""], [3, "click"], [1, "interact"], [3, "c8yIcon"]], template: function UserMenuItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, UserMenuItemComponent_ng_template_0_Template, 6, 6, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc2.IconDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
UserMenuItemComponent.ctorParameters = () => [
    { type: UserMenuService }
];
UserMenuItemComponent.propDecorators = {
    icon: [{ type: Input }],
    label: [{ type: Input }],
    link: [{ type: Input }],
    target: [{ type: Input }],
    priority: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: false },] }],
    click: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserMenuItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-user-menu-item',
                template: "<ng-template #template>\n  <li (click)=\"onClick()\">\n    <a class=\"interact\" [attr.href]=\"link\" [attr.target]=\"target\">\n      <i [c8yIcon]=\"icon\"></i>\n      {{ label | translate }}\n      <ng-content></ng-content>\n    </a>\n  </li>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.UserMenuService }]; }, { priority: [{
            type: Input
        }], click: [{
            type: Output
        }], icon: [{
            type: Input
        }], label: [{
            type: Input
        }], link: [{
            type: Input
        }], target: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }] }); })();
