// @c8y/ngx-components/esm2015/core/user/user-menu-outlet.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { TenantService, UserService } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
import { AppStateService } from '../common/ui-state.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { LoginService } from '../login/login.service';
import { ModalService } from '../modal/modal.service';
import { OptionsService } from '../common/options.service';
import { Status } from '../common/status.model';
import { TranslateService } from '@ngx-translate/core';
import { UserEditModalComponent } from './user-edit-modal.component';
import { gettext } from '../i18n/gettext';
import { sortBy } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/ui-state.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from '../modal/modal.service';
import * as ɵngcc4 from '../login/login.service';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from '@c8y/client';
import * as ɵngcc7 from '../alert/alert.service';
import * as ɵngcc8 from '../common/options.service';
import * as ɵngcc9 from 'ngx-bootstrap/dropdown';
import * as ɵngcc10 from '../common/icon.directive';
import * as ɵngcc11 from './user-menu-item.component';
import * as ɵngcc12 from '@angular/common';
import * as ɵngcc13 from '../common/outlet.directive';
import * as ɵngcc14 from '../common/shorten-user-name.pipe';
import * as ɵngcc15 from '../i18n/c8y-translate.pipe';

function UserMenuOutletComponent_ul_9_ng_container_1_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function UserMenuOutletComponent_ul_9_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, UserMenuOutletComponent_ul_9_ng_container_1_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 14);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r7.template);
} }
function UserMenuOutletComponent_ul_9_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    const _r14 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "li", 15);
    ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_ul_9_ng_container_1_ng_container_2_Template_li_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r14); const item_r7 = ɵngcc0.ɵɵnextContext().$implicit; return item_r7.click(); });
    ɵngcc0.ɵɵelementStart(2, "a", 16);
    ɵngcc0.ɵɵelement(3, "i", 17);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("href", item_r7.link, ɵngcc0.ɵɵsanitizeUrl)("target", item_r7.target);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", item_r7.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, item_r7.label), " ");
} }
function UserMenuOutletComponent_ul_9_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, UserMenuOutletComponent_ul_9_ng_container_1_ng_container_1_Template, 2, 1, "ng-container", 13);
    ɵngcc0.ɵɵtemplate(2, UserMenuOutletComponent_ul_9_ng_container_1_ng_container_2_Template, 6, 6, "ng-container", 13);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r7 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r7.template);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r7.template);
} }
function UserMenuOutletComponent_ul_9_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 18);
} }
function UserMenuOutletComponent_ul_9_li_4_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 19);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵelement(2, "i", 21);
    ɵngcc0.ɵɵelementStart(3, "span", 22);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementStart(6, "strong");
    ɵngcc0.ɵɵelementStart(7, "span", 23);
    ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_ul_9_li_4_Template_span_click_7_listener($event) { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(2); $event.stopPropagation(); return ctx_r16.copyIt(ctx_r16.ui.currentTenant.value.name); });
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵelement(9, "i", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(10, "br");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementStart(13, "strong");
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(16, "br");
    ɵngcc0.ɵɵtext(17);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementStart(19, "strong");
    ɵngcc0.ɵɵtext(20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "info-circle");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 8, "Tenant ID"), ": ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.ui.currentTenant.value.name, "\u00A0 ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "clipboard");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 10, "Backend"), ": ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(15, 12, ctx_r6.ui.state$).versions.backend);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(18, 14, "UI"), ": ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ctx_r6.ui.uiVersion);
} }
function UserMenuOutletComponent_ul_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 9);
    ɵngcc0.ɵɵtemplate(1, UserMenuOutletComponent_ul_9_ng_container_1_Template, 3, 2, "ng-container", 10);
    ɵngcc0.ɵɵtemplate(2, UserMenuOutletComponent_ul_9_li_2_Template, 1, 0, "li", 11);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵtemplate(4, UserMenuOutletComponent_ul_9_li_4_Template, 21, 16, "li", 12);
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.getSortedItems());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(3, 3, ctx_r0.ui.state$).hidePowered);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(5, 5, ctx_r0.ui.state$).hidePowered);
} }
function UserMenuOutletComponent_c8y_user_menu_item_14_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-user-menu-item", 6);
    ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_c8y_user_menu_item_14_Template_c8y_user_menu_item_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.activateSupportAccess(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("icon", "c8y-c8y-support")("label", ɵngcc0.ɵɵpipeBind1(1, 2, "Activate support"));
} }
function UserMenuOutletComponent_c8y_user_menu_item_17_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-user-menu-item", 6);
    ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_c8y_user_menu_item_17_Template_c8y_user_menu_item_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(); return ctx_r20.deactivateSupportAccess(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("icon", "c8y-c8y-support")("label", ɵngcc0.ɵɵpipeBind1(1, 2, "Deactivate support"));
} }
function UserMenuOutletComponent_c8y_user_menu_item_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-user-menu-item", 24);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵpipe(2, "translate");
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("icon", "question-circle")("link", ɵngcc0.ɵɵpipeBind1(1, 4, ctx_r3.ui.state$).supportUrl)("target", "_blank")("label", ɵngcc0.ɵɵpipeBind1(2, 6, "Request support"));
} }
export class UserMenuOutletComponent {
    constructor(ui, bsModalService, modalService, loginService, translateService, tenantService, alertService, user, optionsService) {
        this.ui = ui;
        this.bsModalService = bsModalService;
        this.modalService = modalService;
        this.loginService = loginService;
        this.translateService = translateService;
        this.tenantService = tenantService;
        this.alertService = alertService;
        this.user = user;
        this.optionsService = optionsService;
    }
    copyIt(text) {
        const handler = {
            handleEvent: (e) => {
                e.clipboardData.setData('text/plain', text);
                e.preventDefault();
            }
        };
        document.addEventListener('copy', handler);
        let copied;
        try {
            copied = document.execCommand('copy');
        }
        catch (e) {
            copied = false;
        }
        if (copied) {
            this.alertService.addByText('success', gettext('Copied to clipboard.'));
        }
        else {
            this.alertService.addByText('danger', gettext('Could not copy to clipboard.'));
        }
        document.removeEventListener('copy', handler);
    }
    editUser() {
        return __awaiter(this, void 0, void 0, function* () {
            this.bsModalService.show(UserEditModalComponent);
        });
    }
    logout() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.loginService.logout();
        });
    }
    activateSupportAccess() {
        return __awaiter(this, void 0, void 0, function* () {
            const title = gettext('Activate support user access');
            const companyName = this.optionsService.get('companyName', 'Cumulocity');
            const textWithCompany = gettext(
            // tslint:disable-next-line:max-line-length
            'You are about to allow a support user from {{companyName}} to access your tenant to help you with your issue.');
            const textWithoutCompany = gettext(
            // tslint:disable-next-line:max-line-length
            'You are about to allow a support user to access your tenant to help you with your issue.');
            const finalQuestion = gettext('Do you want to proceed?');
            const body = [
                this.translateService.instant(companyName ? textWithCompany : textWithoutCompany, {
                    companyName
                }),
                this.translateService.instant(finalQuestion)
            ].join(' ');
            const labels = {
                ok: gettext('Activate access'),
                cancel: gettext('Cancel')
            };
            const successMsg = gettext('Support user access activated.');
            try {
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.tenantService.enableSupportUser();
                yield this.refreshCurrentUser();
                this.alertService.success(successMsg);
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    deactivateSupportAccess() {
        return __awaiter(this, void 0, void 0, function* () {
            const title = gettext('Deactivate support user access');
            const companyName = this.optionsService.get('companyName', 'Cumulocity');
            const textWithCompany = gettext(
            // tslint:disable-next-line:max-line-length
            'You are about to block a support user from {{companyName}} from accessing your tenant to help you with your issue.');
            const textWithoutCompany = gettext(
            // tslint:disable-next-line:max-line-length
            'You are about to block a support user from accessing your tenant to help you with your issue.');
            const { data: currentUser } = yield this.user.current();
            const isTenantAdmin = yield this.user.hasRole(currentUser, 'ROLE_TENANT_ADMIN');
            const tenantAdminNote = gettext(
            // tslint:disable-next-line:max-line-length
            'Deactivating support access as tenant admin will disable all other support requests on your tenant.');
            const finalQuestion = gettext('Do you want to proceed?');
            const body = [
                this.translateService.instant(companyName ? textWithCompany : textWithoutCompany, {
                    companyName
                }),
                isTenantAdmin ? this.translateService.instant(tenantAdminNote) : '',
                this.translateService.instant(finalQuestion)
            ]
                .filter(Boolean)
                .join(' ');
            const labels = {
                ok: gettext('Deactivate access'),
                cancel: gettext('Cancel')
            };
            const successMsg = gettext('Support user access deactivated.');
            try {
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.tenantService.disableSupportUser();
                yield this.refreshCurrentUser();
                this.alertService.success(successMsg);
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    getSortedItems() {
        return sortBy(Array.from(this.items), this.byPriority);
    }
    refreshCurrentUser() {
        return __awaiter(this, void 0, void 0, function* () {
            const currentUserResult = yield this.user.current();
            this.ui.currentUser.next(currentUserResult.data);
        });
    }
    byPriority(item) {
        return -item.priority;
    }
}
UserMenuOutletComponent.ɵfac = function UserMenuOutletComponent_Factory(t) { return new (t || UserMenuOutletComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.TenantService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc8.OptionsService)); };
UserMenuOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UserMenuOutletComponent, selectors: [["c8y-user-menu-outlet"]], inputs: { items: "items" }, decls: 22, vars: 33, consts: [["dropdown", "", 1, "dropdown"], ["dropdownToggle", "", 1, "main-header-button", "c8y-dropdown", "dropdown-toggle", "text-nowrap"], [1, "d-inline-block", "hidden-xs", "text-truncate", "m-r-8", 2, "vertical-align", "text-bottom", "max-width", "104px", 3, "title"], [1, "icon-2x", 3, "c8yIcon"], ["class", "dropdown-menu dropdown-menu-right", "style", "max-width: 240px;", 4, "dropdownMenu"], [3, "icon", "label", "priority", "click"], [3, "icon", "label", "click"], [3, "icon", "label", "click", 4, "ngIf"], [3, "icon", "link", "target", "label", 4, "ngIf"], [1, "dropdown-menu", "dropdown-menu-right", 2, "max-width", "240px"], [4, "ngFor", "ngForOf"], ["role", "separator", "class", "divider", 4, "ngIf"], ["class", "dropdown-header bg-gray-white text-pre-normal", "style", "margin-top: -1px;", 4, "ngIf"], [4, "ngIf"], [4, "c8yOutlet"], [3, "click"], [1, "interact"], [3, "c8yIcon"], ["role", "separator", 1, "divider"], [1, "dropdown-header", "bg-gray-white", "text-pre-normal", 2, "margin-top", "-1px"], [1, "flex-row"], [1, "text-info", "flex-item-v-start", "text-14", 2, "margin", "1px 6px 0 -3px", 3, "c8yIcon"], [1, "text-muted", "text-truncate"], [1, "text-primary", "interact", 3, "click"], [3, "icon", "link", "target", "label"]], template: function UserMenuOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵelementStart(2, "span", 2);
        ɵngcc0.ɵɵpipe(3, "shortenUserName");
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵpipe(6, "shortenUserName");
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(8, "i", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, UserMenuOutletComponent_ul_9_Template, 6, 7, "ul", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "c8y-user-menu-item", 5);
        ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_Template_c8y_user_menu_item_click_10_listener() { return ctx.editUser(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "c8y-user-menu-item", 6);
        ɵngcc0.ɵɵlistener("click", function UserMenuOutletComponent_Template_c8y_user_menu_item_click_12_listener() { return ctx.logout(); });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, UserMenuOutletComponent_c8y_user_menu_item_14_Template, 2, 4, "c8y-user-menu-item", 7);
        ɵngcc0.ɵɵpipe(15, "async");
        ɵngcc0.ɵɵpipe(16, "async");
        ɵngcc0.ɵɵtemplate(17, UserMenuOutletComponent_c8y_user_menu_item_17_Template, 2, 4, "c8y-user-menu-item", 7);
        ɵngcc0.ɵɵpipe(18, "async");
        ɵngcc0.ɵɵpipe(19, "async");
        ɵngcc0.ɵɵtemplate(20, UserMenuOutletComponent_c8y_user_menu_item_20_Template, 3, 8, "c8y-user-menu-item", 8);
        ɵngcc0.ɵɵpipe(21, "async");
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 11, ɵngcc0.ɵɵpipeBind1(4, 13, ctx.ui.currentUser)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 15, ɵngcc0.ɵɵpipeBind1(7, 17, ctx.ui.currentUser)), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-user");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "user-menu-male")("label", ɵngcc0.ɵɵpipeBind1(11, 19, "User settings"))("priority", 20);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "sign-out")("label", ɵngcc0.ɵɵpipeBind1(13, 21, "Logout"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(15, 23, ctx.ui.currentUser).supportUserEnabled && ɵngcc0.ɵɵpipeBind1(16, 25, ctx.ui.state$).activateSupportUserAvailable);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(18, 27, ctx.ui.currentUser).supportUserEnabled && ɵngcc0.ɵɵpipeBind1(19, 29, ctx.ui.state$).activateSupportUserAvailable);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(21, 31, ctx.ui.state$).supportUrl);
    } }, directives: [ɵngcc9.BsDropdownDirective, ɵngcc9.BsDropdownToggleDirective, ɵngcc10.IconDirective, ɵngcc9.BsDropdownMenuDirective, ɵngcc11.UserMenuItemComponent, ɵngcc12.NgIf, ɵngcc12.NgForOf, ɵngcc13.OutletDirective], pipes: [ɵngcc14.ShortenUserNamePipe, ɵngcc12.AsyncPipe, ɵngcc15.C8yTranslatePipe], encapsulation: 2 });
UserMenuOutletComponent.ctorParameters = () => [
    { type: AppStateService },
    { type: BsModalService },
    { type: ModalService },
    { type: LoginService },
    { type: TranslateService },
    { type: TenantService },
    { type: AlertService },
    { type: UserService },
    { type: OptionsService }
];
UserMenuOutletComponent.propDecorators = {
    items: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserMenuOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-user-menu-outlet',
                template: "<div\n  dropdown\n  class=\"dropdown\"\n>\n  <button\n    class=\"main-header-button c8y-dropdown dropdown-toggle text-nowrap\"\n    dropdownToggle\n  >\n    <span\n      class=\"d-inline-block hidden-xs text-truncate m-r-8\"\n      style=\"vertical-align: text-bottom; max-width: 104px;\"\n      title=\"{{ui.currentUser | async | shortenUserName}}\"\n    >\n      {{ui.currentUser | async | shortenUserName}}\n    </span>\n    <i\n      [c8yIcon]=\"'c8y-user'\"\n      class=\"icon-2x\"\n    ></i>\n  </button>\n  <ul\n    *dropdownMenu\n    class=\"dropdown-menu dropdown-menu-right\"\n    style=\"max-width: 240px;\"\n  >\n    <ng-container *ngFor=\"let item of getSortedItems()\">\n      <ng-container *ngIf=\"item.template\">\n        <ng-container *c8yOutlet=\"item.template\"></ng-container>\n      </ng-container>\n      <ng-container *ngIf=\"!item.template\">\n        <li (click)=\"item.click()\">\n          <a class=\"interact\" [attr.href]=\"item.link\" [attr.target]=\"item.target\">\n            <i [c8yIcon]=\"item.icon\"></i>\n            {{item.label | translate}}\n          </a>\n        </li>\n      </ng-container>\n    </ng-container>\n    <li\n      *ngIf=\"!(ui.state$ | async).hidePowered\"\n      role=\"separator\"\n      class=\"divider\"\n    ></li>\n    <li\n      class=\"dropdown-header bg-gray-white text-pre-normal\"\n      style=\"margin-top: -1px;\"\n      *ngIf=\"!(ui.state$ | async).hidePowered\"\n    >\n      <div class=\"flex-row\">\n        <i\n          [c8yIcon]=\"'info-circle'\"\n          class=\"text-info flex-item-v-start text-14\"\n          style=\"margin: 1px 6px 0 -3px;\"\n        ></i>\n        <span class=\"text-muted text-truncate\">\n          {{'Tenant ID' | translate}}: <strong>\n            <span class=\"text-primary interact\" (click)=\"$event.stopPropagation(); copyIt(ui.currentTenant.value.name)\">\n              {{ui.currentTenant.value.name}}&nbsp;\n              <i [c8yIcon]=\"'clipboard'\"></i>\n            </span>\n            </strong><br>\n          {{'Backend' | translate}}: <strong>{{(ui.state$ | async).versions.backend}}</strong><br>\n          {{'UI' | translate }}: <strong>{{ui.uiVersion}}</strong>\n        </span>\n      </div>\n    </li>\n  </ul>\n</div>\n\n<!-- the default items -->\n<c8y-user-menu-item\n  [icon]=\"'user-menu-male'\"\n  [label]=\"'User settings' | translate\"\n  [priority]=\"20\"\n  (click)=\"editUser()\"\n></c8y-user-menu-item>\n<c8y-user-menu-item\n  [icon]=\"'sign-out'\"\n  [label]=\"'Logout' | translate\"\n  (click)=\"logout()\"\n></c8y-user-menu-item>\n<c8y-user-menu-item\n  *ngIf=\"!(ui.currentUser | async).supportUserEnabled && ((ui.state$ | async).activateSupportUserAvailable)\"\n  [icon]=\"'c8y-c8y-support'\"\n  [label]=\"'Activate support' | translate\"\n  (click)=\"activateSupportAccess()\"\n></c8y-user-menu-item>\n<c8y-user-menu-item\n  *ngIf=\"(ui.currentUser | async).supportUserEnabled && ((ui.state$ | async).activateSupportUserAvailable)\"\n  [icon]=\"'c8y-c8y-support'\"\n  [label]=\"'Deactivate support' | translate\"\n  (click)=\"deactivateSupportAccess()\"\n></c8y-user-menu-item>\n<c8y-user-menu-item\n  *ngIf=\"(ui.state$ | async).supportUrl\"\n  [icon]=\"'question-circle'\"\n  [link]=\"(ui.state$ | async).supportUrl\"\n  [target]=\"'_blank'\"\n  [label]=\"'Request support' | translate\"\n></c8y-user-menu-item>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: ɵngcc2.BsModalService }, { type: ɵngcc3.ModalService }, { type: ɵngcc4.LoginService }, { type: ɵngcc5.TranslateService }, { type: ɵngcc6.TenantService }, { type: ɵngcc7.AlertService }, { type: ɵngcc6.UserService }, { type: ɵngcc8.OptionsService }]; }, { items: [{
            type: Input
        }] }); })();
