// @c8y/ngx-components/esm2015/core/user/user.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { ModalModule } from '../modal/modal.module';
import { UserMenuItemComponent } from './user-menu-item.component';
import { UserEditModalComponent } from './user-edit-modal.component';
import { UserEditComponent } from './user-edit.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { UserMenuOutletComponent } from './user-menu-outlet.component';
import { AuthenticationModule } from '../authentication/authentication.module';
import { AlertModule } from '../alert/alert.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { RouterModule } from '@angular/router';
import { FormsModule } from '../forms/forms.module';
import { UserTotpSetupComponent } from './user-totp-setup.component';
import { UserTotpRevokeComponent } from './user-totp-revoke.component';
/**
 * The angular module definition for the user menu.
 * @exports UserMenuComponent
 * @exports UserMenuItemComponent
 * @exports UserEditComponent
 * @exports UserEditModalComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/tooltip';
export class UserModule {
}
UserModule.ɵfac = function UserModule_Factory(t) { return new (t || UserModule)(); };
UserModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: UserModule });
UserModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            CommonModule,
            ModalModule,
            BsDropdownModule,
            AuthenticationModule,
            AlertModule,
            RouterModule,
            TooltipModule.forRoot(),
            PopoverModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    UserMenuOutletComponent,
                    UserEditComponent,
                    UserEditModalComponent,
                    UserMenuItemComponent,
                    UserTotpSetupComponent,
                    UserTotpRevokeComponent,
                ],
                imports: [
                    FormsModule,
                    CommonModule,
                    ModalModule,
                    BsDropdownModule,
                    AuthenticationModule,
                    AlertModule,
                    RouterModule,
                    TooltipModule.forRoot(),
                    PopoverModule
                ],
                exports: [
                    UserMenuOutletComponent,
                    UserMenuItemComponent,
                    UserEditComponent,
                    UserEditModalComponent,
                    UserTotpSetupComponent,
                    UserTotpRevokeComponent
                ],
                entryComponents: [
                    UserEditModalComponent,
                    UserTotpSetupComponent,
                    UserTotpRevokeComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(UserModule, { declarations: function () { return [UserMenuOutletComponent,
        UserEditComponent,
        UserEditModalComponent,
        UserMenuItemComponent,
        UserTotpSetupComponent,
        UserTotpRevokeComponent]; }, imports: function () { return [FormsModule,
        CommonModule,
        ModalModule,
        BsDropdownModule,
        AuthenticationModule,
        AlertModule,
        RouterModule, ɵngcc1.TooltipModule, PopoverModule]; }, exports: function () { return [UserMenuOutletComponent,
        UserMenuItemComponent,
        UserEditComponent,
        UserEditModalComponent,
        UserTotpSetupComponent,
        UserTotpRevokeComponent]; } }); })();
