// @c8y/ngx-components/esm2015/core/user/user-totp-setup.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { UserService } from '@c8y/client';
import { Status } from '../common/status.model';
import { gettext } from '../i18n/gettext';
import { ModalService } from '../modal/modal.service';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { LoginService } from '../login/login.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../modal/modal.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '../login/login.service';
import * as ɵngcc5 from '../modal/modal.component';
import * as ɵngcc6 from '../i18n/c8y-translate.directive';
import * as ɵngcc7 from '../authentication/totp-setup.component';
import * as ɵngcc8 from '../authentication/totp-challenge.component';
import * as ɵngcc9 from '../i18n/c8y-translate.pipe';
export class UserTotpSetupComponent {
    constructor(user, modalService, modal, loginService) {
        this.user = user;
        this.modalService = modalService;
        this.modal = modal;
        this.loginService = loginService;
    }
    totpSetupVerified(token) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.user.activateTotp();
            this.modal.hide();
            try {
                yield this.modalService.acknowledge(gettext('Logout required'), gettext('You must log out in order to apply your changes'), Status.WARNING, gettext('Log out'));
                yield this.loginService.logout();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    close() {
        this.modal.hide();
    }
}
UserTotpSetupComponent.ɵfac = function UserTotpSetupComponent_Factory(t) { return new (t || UserTotpSetupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.LoginService)); };
UserTotpSetupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UserTotpSetupComponent, selectors: [["c8y-user-totp-setup"]], decls: 10, vars: 4, consts: [[3, "customFooter"], ["c8y-modal-title", ""], ["type", "button", 1, "close", 3, "title", "click"], ["aria-hidden", "true"], ["translate", "", 1, "text-center"], [3, "onSuccess"]], template: function UserTotpSetupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-modal", 0);
        ɵngcc0.ɵɵelementContainerStart(1, 1);
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵlistener("click", function UserTotpSetupComponent_Template_button_click_2_listener() { return ctx.close(); });
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵelementStart(4, "span", 3);
        ɵngcc0.ɵɵtext(5, "\u00D7");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "h3", 4);
        ɵngcc0.ɵɵtext(7, "Set up two-factor authentication");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelement(8, "c8y-totp-setup");
        ɵngcc0.ɵɵelementStart(9, "c8y-totp-challenge", 5);
        ɵngcc0.ɵɵlistener("onSuccess", function UserTotpSetupComponent_Template_c8y_totp_challenge_onSuccess_9_listener($event) { return ctx.totpSetupVerified($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("customFooter", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 2, "Close"));
    } }, directives: [ɵngcc5.ModalComponent, ɵngcc6.C8yTranslateDirective, ɵngcc7.TotpSetupComponent, ɵngcc8.TotpChallengeComponent], pipes: [ɵngcc9.C8yTranslatePipe], encapsulation: 2 });
UserTotpSetupComponent.ctorParameters = () => [
    { type: UserService },
    { type: ModalService },
    { type: BsModalRef },
    { type: LoginService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserTotpSetupComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-user-totp-setup',
                template: "<c8y-modal [customFooter]=\"true\">\n  <ng-container c8y-modal-title>\n    <button\n      title=\"{{ 'Close' | translate }}\"\n      (click)=\"close()\"\n      type=\"button\"\n      class=\"close\"\n    ><span aria-hidden=\"true\">\u00D7</span></button>\n    <h3 class=\"text-center\" translate>Set up two-factor authentication</h3>\n  </ng-container>\n  <c8y-totp-setup></c8y-totp-setup>\n  <c8y-totp-challenge (onSuccess)=\"totpSetupVerified($event)\"></c8y-totp-challenge>\n</c8y-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.UserService }, { type: ɵngcc2.ModalService }, { type: ɵngcc3.BsModalRef }, { type: ɵngcc4.LoginService }]; }, null); })();
