// @c8y/ngx-components/esm2015/core/user/user-totp-revoke.component.js
import { __awaiter } from "tslib";
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { UserService } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../alert/alert.service';
import * as ɵngcc3 from 'ngx-bootstrap/popover';
import * as ɵngcc4 from '../common/icon.directive';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../i18n/c8y-translate.pipe';

function UserTotpRevokeComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "p", 4);
    ɵngcc0.ɵɵtext(1, " This action will revoke the user's stored TOTP secret which will require a new setup. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "button", 5);
    ɵngcc0.ɵɵlistener("click", function UserTotpRevokeComponent_ng_template_6_Template_button_click_2_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); const _r0 = ɵngcc0.ɵɵreference(1); ctx_r3.revoke(); return _r0.hide(); });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtext(4, " Revoke ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "button", 6);
    ɵngcc0.ɵɵlistener("click", function UserTotpRevokeComponent_ng_template_6_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r4); ɵngcc0.ɵɵnextContext(); const _r0 = ɵngcc0.ɵɵreference(1); return _r0.hide(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵtext(7, " Cancel ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 2, "Revoke"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 4, "Cancel"));
} }
export class UserTotpRevokeComponent {
    constructor(users, alert) {
        this.users = users;
        this.alert = alert;
        this.revokeEmitter = new EventEmitter();
    }
    revoke() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.users.totpRevokeSecret(this.user);
                this.alert.success(gettext('User TOTP secret revoked.'));
                this.revokeEmitter.next();
            }
            catch (e) {
                this.alert.addServerFailure(e);
            }
        });
    }
}
UserTotpRevokeComponent.ɵfac = function UserTotpRevokeComponent_Factory(t) { return new (t || UserTotpRevokeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
UserTotpRevokeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UserTotpRevokeComponent, selectors: [["c8y-user-totp-revoke"]], inputs: { user: "user" }, outputs: { revokeEmitter: "revoke" }, decls: 8, vars: 7, consts: [["type", "button", "placement", "right", 1, "btn", "btn-default", 3, "popover", "popoverTitle"], ["popover", "bs-popover"], ["c8yIcon", "key"], ["confirm", ""], ["translate", "", 1, "m-b-16"], ["type", "button", "translate", "", 1, "btn", "btn-danger", 3, "title", "click"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "click"]], template: function UserTotpRevokeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0, 1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelement(3, "i", 2);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, UserTotpRevokeComponent_ng_template_6_Template, 8, 6, "ng-template", null, 3, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(7);
        ɵngcc0.ɵɵproperty("popover", _r1)("popoverTitle", ɵngcc0.ɵɵpipeBind1(2, 3, "Revoke TOTP secret"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 5, "Revoke TOTP secret"), "\n");
    } }, directives: [ɵngcc3.PopoverDirective, ɵngcc4.IconDirective, ɵngcc5.C8yTranslateDirective], pipes: [ɵngcc6.C8yTranslatePipe], encapsulation: 2 });
UserTotpRevokeComponent.ctorParameters = () => [
    { type: UserService },
    { type: AlertService }
];
UserTotpRevokeComponent.propDecorators = {
    user: [{ type: Input }],
    revokeEmitter: [{ type: Output, args: ['revoke',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserTotpRevokeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-user-totp-revoke',
                template: "<button\n  type=\"button\"\n  class=\"btn btn-default\"\n  [popover]=\"confirm\"\n  [popoverTitle]=\"'Revoke TOTP secret' | translate\"\n  #popover=\"bs-popover\"\n  placement=\"right\"\n>\n  <i c8yIcon=\"key\"></i> {{ 'Revoke TOTP secret' | translate }}\n</button>\n\n<ng-template #confirm>\n  <p class=\"m-b-16\" translate>\n    This action will revoke the user's stored TOTP secret which will require a new setup.\n  </p>\n  <button\n    title=\"{{ 'Revoke' | translate }}\"\n    type=\"button\"\n    class=\"btn btn-danger\"\n    (click)=\"revoke(); popover.hide()\"\n    translate\n  >\n    Revoke\n  </button>\n  <button\n    title=\"{{ 'Cancel' | translate }}\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    (click)=\"popover.hide()\"\n    translate\n  >\n    Cancel\n  </button>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.UserService }, { type: ɵngcc2.AlertService }]; }, { revokeEmitter: [{
            type: Output,
            args: ['revoke']
        }], user: [{
            type: Input
        }] }); })();
