// @c8y/ngx-components/esm2015/core/user/user-menu.service.js
import { Injectable } from '@angular/core';
import { StateService } from '../common/state-service.abstract';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class UserMenuService extends StateService {
    constructor() {
        super(...arguments);
        this.state$ = new BehaviorSubject({
            userMenuItems: new Set()
        });
    }
    /**
     * Returns the current state.
     * @readonly
     * @returns The current set of actions.
     */
    get state() {
        return this.state$.value;
    }
    /**
     * Adds a new item to the search in the header.
     * @param item The item to add.
     */
    add(item) {
        this.state.userMenuItems.add(item);
        this.emitNewState();
    }
    /**
     * Removes an search from the header.
     * @param item The item to remove.
     */
    remove(item) {
        this.state.userMenuItems.delete(item);
        this.emitNewState();
    }
}
UserMenuService.ɵfac = /*@__PURE__*/ function () { let ɵUserMenuService_BaseFactory; return function UserMenuService_Factory(t) { return (ɵUserMenuService_BaseFactory || (ɵUserMenuService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(UserMenuService)))(t || UserMenuService); }; }();
UserMenuService.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserMenuService_Factory() { return new UserMenuService(); }, token: UserMenuService, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserMenuService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
