// @c8y/ngx-components/esm2015/upgrade/dashboard/component.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dashboard-bridge.service';
export class Ng1ComponentFactoryService {
    constructor(bridge) {
        this.bridge = bridge;
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            const components = (yield this.bridge.ng1Components.list()).map(cmp => {
                const data = {
                    settings: Object.assign(Object.assign({}, cmp.options), { configTemplateUrl: cmp.configTemplateUrl, widgetComponent: cmp.widgetComponent, configComponent: cmp.configComponent, transformConfigWithContext: cmp.transformConfigWithContext || cmp.componentTransformConfigWithContext, upgrade: true })
                };
                data.settings.templateUrl = cmp.templateUrl; // workaround as in object it creates build error
                return {
                    id: cmp.name,
                    label: cmp.nameDisplay || cmp.name,
                    description: cmp.description,
                    component: WidgetComponent,
                    configComponent: WidgetComponent,
                    previewImage: cmp.previewImage,
                    data
                };
            });
            // add the device selector legacy component
            components.push({
                id: 'device.selector.legacy',
                label: 'Device selector',
                description: '',
                data: { settings: { noNewWidgets: true, upgrade: true } },
                component: DeviceSelectorComponent
            });
            return components;
        });
    }
}
Ng1ComponentFactoryService.ɵfac = function Ng1ComponentFactoryService_Factory(t) { return new (t || Ng1ComponentFactoryService)(ɵngcc0.ɵɵinject(ɵngcc1.DashboardBridgeService)); };
Ng1ComponentFactoryService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Ng1ComponentFactoryService, factory: Ng1ComponentFactoryService.ɵfac });
Ng1ComponentFactoryService.ctorParameters = () => [
    { type: DashboardBridgeService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng1ComponentFactoryService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DashboardBridgeService }]; }, null); })();
