// @c8y/ngx-components/esm2015/upgrade/dashboard/dashboard-actions.factory.js
import { Injectable } from '@angular/core';
import { ViewContext } from '@c8y/ngx-components';
import { CopyActionComponent } from './copy-action.component';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { PasteActionComponent } from './paste-action.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dashboard-bridge.service';
export class Ng1DashboardActionFactoryService {
    constructor(dashboardBridgeService) {
        this.dashboardBridgeService = dashboardBridgeService;
    }
    get(activeRoute) {
        const actions = [];
        const data = !activeRoute.parent || activeRoute.snapshot.data.context
            ? activeRoute.snapshot.data
            : activeRoute.parent.snapshot.data;
        const { contextData } = data;
        const canCopy = contextData &&
            activeRoute.snapshot.data.dashboard &&
            [ViewContext.Device, ViewContext.Group].includes(data.context);
        const canPaste = contextData &&
            [ViewContext.Device, ViewContext.Group].includes(data.context) &&
            this.dashboardBridgeService.dashboardClipboard;
        if (canCopy) {
            actions.push({
                priority: -20,
                placement: 'more',
                template: CopyActionComponent
            });
        }
        if (canPaste) {
            actions.push({
                priority: -10,
                placement: 'more',
                template: PasteActionComponent
            });
        }
        return actions;
    }
}
Ng1DashboardActionFactoryService.ɵfac = function Ng1DashboardActionFactoryService_Factory(t) { return new (t || Ng1DashboardActionFactoryService)(ɵngcc0.ɵɵinject(ɵngcc1.DashboardBridgeService)); };
Ng1DashboardActionFactoryService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Ng1DashboardActionFactoryService, factory: Ng1DashboardActionFactoryService.ɵfac });
Ng1DashboardActionFactoryService.ctorParameters = () => [
    { type: DashboardBridgeService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng1DashboardActionFactoryService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DashboardBridgeService }]; }, null); })();
