// @c8y/ngx-components/esm2015/upgrade/dashboard/dashboard-upgrade.module.js
import { CommonModule } from '@angular/common';
import { NgModule, NgZone, Optional } from '@angular/core';
import { Router } from '@angular/router';
import { ActionBarService, HOOK_ACTION_BAR, HOOK_COMPONENTS } from '@c8y/ngx-components';
import { ContextDashboardService } from '@c8y/ngx-components/context-dashboard';
import { Ng1ComponentFactoryService } from './component.factory';
import { CopyActionComponent } from './copy-action.component';
import { Ng1DashboardActionFactoryService } from './dashboard-actions.factory';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { PasteActionComponent } from './paste-action.component';
import { WidgetComponent } from './widget.component';
import { DeviceSelectorComponent } from './device-selector.component';
import { C8yTranslateModule } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
export class DashboardUpgradeModule {
}
DashboardUpgradeModule.ɵfac = function DashboardUpgradeModule_Factory(t) { return new (t || DashboardUpgradeModule)(); };
DashboardUpgradeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DashboardUpgradeModule });
DashboardUpgradeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: DashboardBridgeService,
            useFactory: dashboardBridgeFactory,
            deps: ['$injector', NgZone, Router, [new Optional(), ContextDashboardService], ActionBarService]
        },
        { provide: HOOK_COMPONENTS, useClass: Ng1ComponentFactoryService, multi: true },
        { provide: HOOK_ACTION_BAR, useClass: Ng1DashboardActionFactoryService, multi: true }
    ], imports: [[CommonModule, C8yTranslateModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardUpgradeModule, [{
        type: NgModule,
        args: [{
                declarations: [DeviceSelectorComponent, WidgetComponent, CopyActionComponent, PasteActionComponent],
                entryComponents: [DeviceSelectorComponent, WidgetComponent, CopyActionComponent, PasteActionComponent],
                imports: [CommonModule, C8yTranslateModule],
                providers: [
                    {
                        provide: DashboardBridgeService,
                        useFactory: dashboardBridgeFactory,
                        deps: ['$injector', NgZone, Router, [new Optional(), ContextDashboardService], ActionBarService]
                    },
                    { provide: HOOK_COMPONENTS, useClass: Ng1ComponentFactoryService, multi: true },
                    { provide: HOOK_ACTION_BAR, useClass: Ng1DashboardActionFactoryService, multi: true }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DashboardUpgradeModule, { declarations: function () { return [DeviceSelectorComponent, WidgetComponent, CopyActionComponent, PasteActionComponent]; }, imports: function () { return [CommonModule, C8yTranslateModule]; } }); })();
export function dashboardBridgeFactory(injector, ngZone, router, contextDashboardService, actionBarService) {
    return new DashboardBridgeService(injector, ngZone, router, contextDashboardService, actionBarService);
}
