// @c8y/ngx-components/esm2015/upgrade/dashboard/device-selector.component.js
import { __awaiter } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dashboard-bridge.service';
export class DeviceSelectorComponent {
    constructor(dashboardBridgeService, el) {
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.ng1Scope = this.dashboardBridgeService.instantiateDeviceSelector(this.el.nativeElement, this.config);
        });
    }
    ngOnDestroy() {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
    }
}
DeviceSelectorComponent.ɵfac = function DeviceSelectorComponent_Factory(t) { return new (t || DeviceSelectorComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DashboardBridgeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
DeviceSelectorComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceSelectorComponent, selectors: [["c8y-device-selector-legacy"]], decls: 0, vars: 0, template: function DeviceSelectorComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
DeviceSelectorComponent.ctorParameters = () => [
    { type: DashboardBridgeService },
    { type: ElementRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceSelectorComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-selector-legacy',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.DashboardBridgeService }, { type: ɵngcc0.ElementRef }]; }, null); })();
