// @c8y/ngx-components/esm2015/upgrade/dashboard/widget.component.js
import { __awaiter } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Optional } from '@angular/core';
import { DashboardBridgeService } from './dashboard-bridge.service';
import { WidgetsDashboardComponent } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './dashboard-bridge.service';
export class WidgetComponent {
    constructor(dashboard, dashboardBridgeService, el) {
        this.dashboard = dashboard;
        this.dashboardBridgeService = dashboardBridgeService;
        this.el = el;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.config) {
                this.ng1Scope = yield this.dashboardBridgeService.instantiateComponent(this.config, this.el.nativeElement);
                if (this.dashboard) {
                    this.resizeSubscription = this.dashboard.onChangeDashboard.subscribe(() => {
                        this.ng1Scope.$broadcast('dashboardResize');
                    });
                }
            }
        });
    }
    ngOnDestroy() {
        if (this.ng1Scope) {
            this.ng1Scope.$destroy();
        }
        if (this.resizeSubscription) {
            this.resizeSubscription.unsubscribe();
        }
    }
}
WidgetComponent.ɵfac = function WidgetComponent_Factory(t) { return new (t || WidgetComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WidgetsDashboardComponent, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DashboardBridgeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
WidgetComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WidgetComponent, selectors: [["c8y-widget-legacy"]], decls: 0, vars: 0, template: function WidgetComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
WidgetComponent.ctorParameters = () => [
    { type: WidgetsDashboardComponent, decorators: [{ type: Optional }] },
    { type: DashboardBridgeService },
    { type: ElementRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WidgetComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-widget-legacy',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: ɵngcc1.WidgetsDashboardComponent, decorators: [{
                type: Optional
            }] }, { type: ɵngcc2.DashboardBridgeService }, { type: ɵngcc0.ElementRef }]; }, null); })();
