// @c8y/ngx-components/esm2015/core/properties-list/properties-list.component.js
import { Component, Input } from '@angular/core';
import { sortBy, get } from 'lodash-es';
/**
 * Renders a list of properties of an object.
 *
 * @example
 * ```html
 * <c8y-properties-list
 *   icon="info"
 *   [properties]="properties"
 *   [values]="options"
 *   [emptyLabel]="'-'"
 *   [title]="'Application properties' | translate"
 * ></c8y-properties-list>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';

function PropertiesListComponent_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 5);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r0.icon);
} }
function PropertiesListComponent_li_6_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(prop_r2.value);
} }
function PropertiesListComponent_li_6_a_6_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 13);
    ɵngcc0.ɵɵlistener("click", function PropertiesListComponent_li_6_a_6_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const prop_r2 = ɵngcc0.ɵɵnextContext().$implicit; return prop_r2.action($event, prop_r2); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(prop_r2.value);
} }
const _c0 = function (a0) { return { pointer: a0 }; };
function PropertiesListComponent_li_6_span_7_span_1_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span", 15);
    ɵngcc0.ɵɵlistener("click", function PropertiesListComponent_li_6_span_7_span_1_Template_span_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r16); const propTag_r13 = restoredCtx.$implicit; const prop_r2 = ɵngcc0.ɵɵnextContext(2).$implicit; return prop_r2.action && prop_r2.action($event, propTag_r13); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const propTag_r13 = ctx.$implicit;
    const prop_r2 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c0, prop_r2.action));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(propTag_r13);
} }
function PropertiesListComponent_li_6_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, PropertiesListComponent_li_6_span_7_span_1_Template, 2, 4, "span", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", prop_r2.value);
} }
const _c1 = function (a0, a1) { return { "separator-top-bottom": a0, "separator-bottom": a1 }; };
const _c2 = function (a0, a1, a2, a3, a4) { return { "m-l-16": a0, legend: a1, "form-block": a2, "m-b-0": a3, "m-t-4": a4 }; };
function PropertiesListComponent_li_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 6);
    ɵngcc0.ɵɵelementStart(1, "div", 7);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span", 8);
    ɵngcc0.ɵɵtemplate(5, PropertiesListComponent_li_6_span_5_Template, 2, 1, "span", 9);
    ɵngcc0.ɵɵtemplate(6, PropertiesListComponent_li_6_a_6_Template, 2, 1, "a", 10);
    ɵngcc0.ɵɵtemplate(7, PropertiesListComponent_li_6_span_7_Template, 2, 1, "span", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const prop_r2 = ctx.$implicit;
    const i_r3 = ctx.index;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(9, _c1, i_r3 === 0, i_r3 > 0));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction5(12, _c2, ctx_r1.hasGroup(prop_r2), prop_r2.type === "group", prop_r2.type === "group", prop_r2.type === "group", prop_r2.type === "group"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 7, prop_r2.label), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngSwitch", prop_r2.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "string");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "link");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "array");
} }
export class PropertiesListComponent {
    constructor() {
        /**
         * A string array of groups that are shown. If noParse is set to false,
         * each complex key will form a group.
         */
        this.groups = [];
        /**
         * The component tries to parse the properties and resolve keys and types. You can
         * avoid this by setting this property to true.
         */
        this.noParse = false;
        this._data = {};
    }
    /**
     * The properties that this list should display.
     */
    set properties(items) {
        this._properties = this.parseProperties(items);
    }
    /**
     * @ignore
     */
    get properties() {
        if (this.noParse) {
            return this._properties;
        }
        const propsWithGroups = [
            ...this._properties,
            ...this.groups.map(group => ({
                key: group,
                label: group,
                value: group,
                type: 'group'
            }))
        ].filter(item => !!item.value);
        return sortBy(propsWithGroups, ['key']);
    }
    /**
     * An object where the properties keys are resolved from.
     */
    set data(data) {
        this._data = data;
        this._properties = this.parseProperties(this._properties);
    }
    /**
     * Checks if a certain property has an group associated.
     * @param item The property to verify.
     */
    hasGroup(item) {
        if (!item.key) {
            return false;
        }
        const keyPath = item.key.split('.');
        return keyPath.length > 1 && this.groups.includes(keyPath[0]);
    }
    /**
     * @ignore
     */
    ngOnInit() {
        this._properties = this.parseProperties(this._properties);
    }
    /**
     * Used in trackBy to avoid recalculation all the time.
     * @ignore
     */
    identity(index, item) {
        return item.value;
    }
    parseProperties(items = []) {
        return items.map(item => this.parsePropertyItem(item));
    }
    parsePropertyItem(item) {
        if (this.noParse) {
            return item;
        }
        item.value = this.resolveValueFromKey(item);
        item.value = item.transform && item.value ? item.transform(item.value) : item.value;
        item.type = this.resolveType(item);
        item.value = this.attachEmptyLabel(item);
        return item;
    }
    resolveValueFromKey(item) {
        if (item.key && this._data) {
            const keyPath = item.key.split('.');
            const rootGroup = keyPath[0];
            if (keyPath.length > 1 && !this.groups.includes(rootGroup)) {
                this.groups.push(rootGroup);
            }
            return get(this._data, item.key);
        }
        return item.value;
    }
    resolveType(item) {
        return Array.isArray(item.value) ? 'array' : item.action && item.value ? 'link' : 'string';
    }
    attachEmptyLabel(item) {
        if (!item.value) {
            return this.emptyLabel;
        }
        return item.value;
    }
}
PropertiesListComponent.ɵfac = function PropertiesListComponent_Factory(t) { return new (t || PropertiesListComponent)(); };
PropertiesListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PropertiesListComponent, selectors: [["c8y-properties-list"]], inputs: { groups: "groups", noParse: "noParse", properties: "properties", data: "data", title: "title", icon: "icon", emptyLabel: "emptyLabel" }, decls: 7, vars: 6, consts: [[1, "m-b-8"], ["class", "text-info m-r-8", 3, "c8yIcon", 4, "ngIf"], [1, "text-label-small"], [1, "list-unstyled", "small"], ["class", "p-t-4 p-b-4 d-flex", 3, "ngClass", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "text-info", "m-r-8", 3, "c8yIcon"], [1, "p-t-4", "p-b-4", "d-flex", 3, "ngClass"], [1, "small", "text-medium", "text-nowrap", "m-r-4", 3, "ngClass"], [1, "flex-item-right", 3, "ngSwitch"], ["class", "flex-item-right", 4, "ngSwitchCase"], ["class", "flex-item-right pointer text-truncate m-l-4", 3, "click", 4, "ngSwitchCase"], [4, "ngSwitchCase"], [1, "flex-item-right"], [1, "flex-item-right", "pointer", "text-truncate", "m-l-4", 3, "click"], ["class", "label label-default m-l-4", 3, "ngClass", "click", 4, "ngFor", "ngForOf"], [1, "label", "label-default", "m-l-4", 3, "ngClass", "click"]], template: function PropertiesListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p", 0);
        ɵngcc0.ɵɵtemplate(1, PropertiesListComponent_i_1_Template, 1, 1, "i", 1);
        ɵngcc0.ɵɵelementStart(2, "span", 2);
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "ul", 3);
        ɵngcc0.ɵɵtemplate(6, PropertiesListComponent_li_6_Template, 8, 18, "li", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, ctx.title));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.properties)("ngForTrackBy", ctx.identity);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc2.IconDirective, ɵngcc1.NgClass, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
PropertiesListComponent.propDecorators = {
    properties: [{ type: Input }],
    title: [{ type: Input }],
    icon: [{ type: Input }],
    data: [{ type: Input }],
    groups: [{ type: Input }],
    noParse: [{ type: Input }],
    emptyLabel: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PropertiesListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-properties-list',
                template: "<p class=\"m-b-8\">\n  <i *ngIf=\"icon\" [c8yIcon]=\"icon\" class=\"text-info m-r-8\"></i>\n  <span class=\"text-label-small\">{{ title | translate }}</span>\n</p>\n<ul class=\"list-unstyled small\">\n  <li class=\"p-t-4 p-b-4 d-flex\" \n    *ngFor=\"let prop of properties; let i = index; trackBy: identity\"\n    [ngClass]=\"{'separator-top-bottom': i === 0, 'separator-bottom': i > 0}\"\n    >\n    <div\n      [ngClass]=\"{\n        'm-l-16': hasGroup(prop),\n        legend: prop.type === 'group',\n        'form-block': prop.type === 'group',\n        'm-b-0': prop.type === 'group',\n        'm-t-4': prop.type === 'group'\n      }\"\n      class=\"small text-medium text-nowrap m-r-4\"\n    >\n      {{ prop.label | translate }}\n    </div>\n    <span [ngSwitch]=\"prop.type\" class=\"flex-item-right\">\n      <span *ngSwitchCase=\"'string'\" class=\"flex-item-right\">{{ prop.value }}</span>\n      <a\n        *ngSwitchCase=\"'link'\"\n        (click)=\"prop.action($event, prop)\"\n        class=\"flex-item-right pointer text-truncate m-l-4\"\n        >{{ prop.value }}</a\n      >\n      <span *ngSwitchCase=\"'array'\">\n        <span\n          class=\"label label-default m-l-4\"\n          *ngFor=\"let propTag of prop.value\"\n          (click)=\"prop.action && prop.action($event, propTag)\"\n          [ngClass]=\"{\n            pointer: prop.action\n          }\"\n          >{{ propTag }}</span\n        >\n      </span>\n    </span>\n  </li>\n</ul>\n"
            }]
    }], function () { return []; }, { groups: [{
            type: Input
        }], noParse: [{
            type: Input
        }], properties: [{
            type: Input
        }], data: [{
            type: Input
        }], title: [{
            type: Input
        }], icon: [{
            type: Input
        }], emptyLabel: [{
            type: Input
        }] }); })();
