// @c8y/ngx-components/esm2015/ecosystem/ecosystem-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { Permissions } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class EcosystemNavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
        this.navs = [];
        this.showNavigatorNodes = this.permissions.hasAnyRole([
            'ROLE_APPLICATION_MANAGEMENT_READ',
            'ROLE_APPLICATION_MANAGEMENT_ADMIN'
        ]);
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            /*
            The NavigatorNodes are added in AngularJS code to keep the Quick links widget working.
            After the Quick links widget is migrated to Angular, the logic of adding the NavigatorNodes can be migrated as well.
                if (this.navs.length === 0 && this.showNavigatorNodes) {
                  const applicationsNode: NavigatorNode = new NavigatorNode({
                    label: gettext('Applications'),
                    icon: 'c8y-modules',
                    path: '/ecosystem',
                    priority: 300,
                    routerLinkExact: false
                  });
            
                  const microservicesNode = new NavigatorNode({
                    label: gettext('Microservices'),
                    icon: 'microchip',
                    path: '/microservices',
                    priority: 200,
                    routerLinkExact: false
                  });
            
                  this.navs.push(
                    new NavigatorNode({
                      label: gettext('Ecosystem'),
                      icon: 'c8y-atom',
                      priority: 3200,
                      children: [applicationsNode, microservicesNode]
                    })
                  );
                }
            */
            return this.navs;
        });
    }
}
EcosystemNavigationFactory.ɵfac = function EcosystemNavigationFactory_Factory(t) { return new (t || EcosystemNavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions)); };
EcosystemNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: EcosystemNavigationFactory, factory: EcosystemNavigationFactory.ɵfac });
EcosystemNavigationFactory.ctorParameters = () => [
    { type: Permissions }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EcosystemNavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Permissions }]; }, null); })();
