// @c8y/ngx-components/esm2015/core/dashboard/dashboard-child.component.js
import { Component, ContentChildren, ElementRef, EventEmitter, HostBinding, Input, Output } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { DashboardChildActionComponent } from './dashboard-child-action.component';
import { DashboardChildChange } from './dashboard-child-change';
import { DashboardComponent } from './dashboard.component';
/**
 * A dashboard child allows to position elements
 * correctly on a grid. The user can then resize and
 * rearrange the elements, as long as they are not `frozen`.
 *
 * By setting `c8y-dashboard-child-actions` and
 * `c8y-dashboard-child-title` on the element you can add
 * custom actions or a custom title to the current child.
 *
 * By adding the correct branded classes, you can define
 * the look and feel of the child. By default it is displayed
 * as a card.
 *
 * Example:
 *
 * ```html
 *   <c8y-dashboard-child
 *     #cpWidget3
 *     [isFrozen]="isFrozen"
 *     [x]="0"
 *     [y]="3"
 *     [width]="4"
 *     [height]="4"
 *     [class]="'card-dashboard panel-content-transparent'"
 *   >
 *     <c8y-dashboard-child-title *ngIf="showTitle">
 *       <span>Transparent!</span>
 *     </c8y-dashboard-child-title>
 *     <c8y-dashboard-child-action>
 *       <a href="" (click)="showTitle = !showTitle; (false)">
 *         <i [c8yIcon]="'heading'"></i> Hide/show title
 *       </a>
 *     </c8y-dashboard-child-action>
 *     <c8y-dashboard-child-action>
 *       <a href="" (click)="cpWidget3.isFrozen = !cpWidget3.isFrozen; (false)">
 *         <i [c8yIcon]="cpWidget3.isFrozen ? 'lock' : 'unlock'"></i> Toggle freeze
 *       </a>
 *     </c8y-dashboard-child-action>
 *     x: {{ cpWidget3.x }}<br />
 *     y: {{ cpWidget3.y }}<br />
 *     width: {{ cpWidget3.width }}<br />
 *     height: {{ cpWidget3.height }}<br />
 *   </c8y-dashboard-child>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dashboard.component';
import * as ɵngcc2 from '@angular/platform-browser';
import * as ɵngcc3 from '@angular/cdk/drag-drop';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from 'ngx-bootstrap/dropdown';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from '../i18n/c8y-translate.pipe';

function DashboardChildComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 9);
} }
function DashboardChildComponent_div_5_ul_5_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function DashboardChildComponent_div_5_ul_5_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DashboardChildComponent_div_5_ul_5_ng_container_1_ng_container_1_Template, 1, 0, "ng-container", 17);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const action_r7 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", action_r7.template);
} }
function DashboardChildComponent_div_5_ul_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 15);
    ɵngcc0.ɵɵtemplate(1, DashboardChildComponent_div_5_ul_5_ng_container_1_Template, 2, 1, "ng-container", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r5.actions);
} }
function DashboardChildComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelementStart(1, "div", 11);
    ɵngcc0.ɵɵelementStart(2, "a", 12);
    ɵngcc0.ɵɵlistener("click", function DashboardChildComponent_div_5_Template_a_click_2_listener() { return false; });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelement(4, "i", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, DashboardChildComponent_div_5_ul_5_Template, 2, 1, "ul", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 2, "Settings"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "cog");
} }
function DashboardChildComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵlistener("cdkDragStarted", function DashboardChildComponent_div_8_Template_div_cdkDragStarted_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.resizeStarted($event); })("cdkDragEnded", function DashboardChildComponent_div_8_Template_div_cdkDragEnded_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.reset($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("cdkDragDisabled", ctx_r2.isFrozen);
} }
function DashboardChildComponent_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 19);
} }
function DashboardChildComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 9);
} }
const _c0 = [[["c8y-dashboard-child-title"]], "*"];
const _c1 = function (a0, a1) { return { width: a0, height: a1 }; };
const _c2 = function (a0, a1) { return { "drag-handle": a0, draggableCursor: a1 }; };
const _c3 = ["c8y-dashboard-child-title", "*"];
export class DashboardChildComponent {
    constructor(dashboard, sanitizer, element) {
        this.dashboard = dashboard;
        this.sanitizer = sanitizer;
        this.element = element;
        this.actions = [];
        this.isResize = false;
        this.isDragging = false;
        this.klasses = {};
        this._pxWidth = '100%';
        this._pxHeight = '100%';
        /**
         * The width of the component in grid-columns.
         */
        this.width = 1;
        /**
         * The height of the component in grid-rows.
         */
        this.height = 1;
        /**
         * The margin of the child in pixel.
         */
        this.margin = 12;
        /**
         * If a dashboard is frozen, all children cannot be moved
         * or resized.
         */
        this.isFrozen = false;
        /**
         * The child content is initialized, as soon it is scrolled into viewport
         */
        this.useIntersection = false;
        /**
         * An event fired if a child change is started (dragging or resizing)
         */
        this.changeStart = new EventEmitter();
        /**
         * An event fired if a child change is ended
         */
        this.changeEnd = new EventEmitter();
        /**
         * All classes added to this child
         */
        this.class = {};
        /**
         * An indicator if the child is intersected (that mean visible for the user)
         */
        this.intersected = false;
    }
    /**
     * Updates the pixel width of the child (used for resizing)
     */
    set pxWidth(value) {
        this._pxWidth = `${value}px`;
    }
    /**
     * Updates the pixel height of the child (used for resizing)
     */
    set pxHeight(value) {
        this._pxHeight = `${value}px`;
    }
    /**
     * nasty workaround for that issue:
     * https://github.com/angular/angular/issues/9343
     */
    get inlineStyle() {
        return this.sanitizer.bypassSecurityTrustStyle(`
    grid-column-start: ${this.x + 1};
    grid-row-start: ${this.y + 1};
    grid-column-end: span ${this.width};
    grid-row-end: span ${this.height};
    display: block;
    margin: ${this.margin || 12}px;
    order: ${this.getOrder()};
    `);
    }
    ngOnChanges() {
        this.klasses = Object.assign({ card: true, 'card-dashboard': true, disabled: this.isFrozen, 'on-resize': this.isResize }, this.class);
    }
    ngOnInit() {
        if (this.x === undefined || this.y === undefined) {
            setTimeout(() => this.setDynamicDimension());
        }
        if (this.useIntersection && 'IntersectionObserver' in window) {
            const intersectionObserver = new IntersectionObserver(event => (this.intersected = this.childInView(event[0], intersectionObserver)));
            intersectionObserver.observe(this.element.nativeElement);
        }
        else {
            this.intersected = true;
        }
    }
    ngAfterViewInit() {
        this.dashboard.children.push(this);
    }
    setDynamicDimension() {
        const ds = new DashboardChildChange(this);
        const { x, y } = ds.findFreeDimension();
        this.x = x;
        this.y = y;
        this.dashboard.emitChange(this);
    }
    resizeStarted($event) {
        this.isResize = true;
        this.dashboard.updateRectSize();
        this.dragSource = $event.source;
        const positioning = new DashboardChildChange(this);
        this.changeSubscription = positioning.resize$.subscribe();
        this.changeStart.emit(this);
        this.ngOnChanges();
    }
    dragStarted($event) {
        this.isDragging = true;
        this.dashboard.updateRectSize();
        this.dragSource = $event.source;
        const positioning = new DashboardChildChange(this);
        this.changeSubscription = positioning.drag$.subscribe();
        this.changeStart.emit(this);
    }
    reset($event) {
        this.isResize = false;
        this.isDragging = false;
        this._pxWidth = '100%';
        this._pxHeight = '100%';
        this.ngOnChanges();
        if ($event) {
            $event.source.reset();
        }
        if (this.changeSubscription) {
            this.changeSubscription.unsubscribe();
            this.dashboard.emitChange(this);
            this.changeEnd.emit(this);
        }
    }
    ngOnDestroy() {
        if (this.changeSubscription) {
            this.changeSubscription.unsubscribe();
        }
        this.removeSelfFromDashboard();
    }
    removeSelfFromDashboard() {
        const i = this.dashboard.children.indexOf(this);
        if (i >= 0) {
            this.dashboard.children.splice(i, 1);
        }
    }
    getOrder() {
        return `${Math.round((this.y + (this.x + 1) / 100) * 100)}`;
    }
    childInView(event, observer) {
        if (event.isIntersecting) {
            observer.unobserve(event.target);
            return true;
        }
        return false;
    }
}
DashboardChildComponent.ɵfac = function DashboardChildComponent_Factory(t) { return new (t || DashboardChildComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DashboardComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DomSanitizer), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
DashboardChildComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DashboardChildComponent, selectors: [["c8y-dashboard-child"]], contentQueries: function DashboardChildComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, DashboardChildActionComponent, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.actions = _t);
    } }, hostAttrs: [1, "dashboard-grid-child"], hostVars: 1, hostBindings: function DashboardChildComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵattribute("style", ctx.inlineStyle, ɵngcc0.ɵɵsanitizeStyle);
    } }, inputs: { width: "width", height: "height", margin: "margin", isFrozen: "isFrozen", useIntersection: "useIntersection", class: "class", x: "x", y: "y", data: "data" }, outputs: { changeStart: "changeStart", changeEnd: "changeEnd" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c3, decls: 11, vars: 14, consts: [["cdkDropList", ""], ["class", "card-placeholder", 4, "ngIf"], ["cdkDrag", "", 3, "ngClass", "ngStyle", "cdkDragDisabled", "cdkDragStarted", "cdkDragEnded"], ["cdkDragHandle", "", 1, "card-header-actions", "card-header-grid", 3, "ngClass"], ["class", "header-actions", 4, "ngIf"], [1, "card-inner-scroll"], ["class", "resize-handle", "cdkDrag", "", 3, "cdkDragDisabled", "cdkDragStarted", "cdkDragEnded", 4, "ngIf"], ["class", "resize-icon", 4, "ngIf"], ["class", "card-placeholder", 4, "cdkDragPlaceholder"], [1, "card-placeholder"], [1, "header-actions"], ["dropdown", "", "container", "body", "placement", "bottom right", 1, "optionsBtn", "dropdown"], ["href", "", "dropdownToggle", "", 1, "btnIcon", "c8y-dropdown", 3, "title", "click"], [3, "c8yIcon"], ["class", "dropdown-menu dropdown-menu-right", "style", "right:-1px;", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu-right", 2, "right", "-1px"], [4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["cdkDrag", "", 1, "resize-handle", 3, "cdkDragDisabled", "cdkDragStarted", "cdkDragEnded"], [1, "resize-icon"]], template: function DashboardChildComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, DashboardChildComponent_div_1_Template, 1, 0, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵlistener("cdkDragStarted", function DashboardChildComponent_Template_div_cdkDragStarted_2_listener($event) { return ctx.dragStarted($event); })("cdkDragEnded", function DashboardChildComponent_Template_div_cdkDragEnded_2_listener($event) { return ctx.reset($event); });
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵprojection(4);
        ɵngcc0.ɵɵtemplate(5, DashboardChildComponent_div_5_Template, 6, 4, "div", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵprojection(7, 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, DashboardChildComponent_div_8_Template, 1, 1, "div", 6);
        ɵngcc0.ɵɵtemplate(9, DashboardChildComponent_div_9_Template, 1, 0, "div", 7);
        ɵngcc0.ɵɵtemplate(10, DashboardChildComponent_div_10_Template, 1, 0, "div", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isResize);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ctx.klasses)("ngStyle", ɵngcc0.ɵɵpureFunction2(8, _c1, ctx._pxWidth, ctx._pxHeight))("cdkDragDisabled", ctx.isFrozen);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(11, _c2, !ctx.isFrozen, !ctx.isFrozen));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isFrozen && ctx.actions.length > 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isFrozen && !ctx.isDragging);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isFrozen && !ctx.isDragging);
    } }, directives: [ɵngcc3.CdkDropList, ɵngcc4.NgIf, ɵngcc3.CdkDrag, ɵngcc4.NgClass, ɵngcc4.NgStyle, ɵngcc3.CdkDragHandle, ɵngcc3.CdkDragPlaceholder, ɵngcc5.BsDropdownDirective, ɵngcc5.BsDropdownToggleDirective, ɵngcc6.IconDirective, ɵngcc5.BsDropdownMenuDirective, ɵngcc4.NgForOf, ɵngcc4.NgTemplateOutlet], pipes: [ɵngcc7.C8yTranslatePipe], encapsulation: 2 });
DashboardChildComponent.ctorParameters = () => [
    { type: DashboardComponent },
    { type: DomSanitizer },
    { type: ElementRef }
];
DashboardChildComponent.propDecorators = {
    actions: [{ type: ContentChildren, args: [DashboardChildActionComponent,] }],
    x: [{ type: Input }],
    y: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    data: [{ type: Input }],
    margin: [{ type: Input }],
    isFrozen: [{ type: Input }],
    useIntersection: [{ type: Input }],
    changeStart: [{ type: Output }],
    changeEnd: [{ type: Output }],
    class: [{ type: Input }],
    inlineStyle: [{ type: HostBinding, args: ['attr.style',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardChildComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-dashboard-child',
                template: "<div cdkDropList>\n  <div *ngIf=\"isResize\" class=\"card-placeholder\"></div>\n  <div\n    [ngClass]=\"klasses\"\n    cdkDrag\n    [ngStyle]=\"{ width: _pxWidth, height: _pxHeight }\"\n    (cdkDragStarted)=\"dragStarted($event)\"\n    (cdkDragEnded)=\"reset($event)\"\n    [cdkDragDisabled]=\"isFrozen\"\n  >\n    <div\n      class=\"card-header-actions card-header-grid\"\n      [ngClass]=\"{ 'drag-handle': !isFrozen, draggableCursor: !isFrozen }\"\n      cdkDragHandle\n    >\n      <ng-content select=\"c8y-dashboard-child-title\"></ng-content>\n      <div class=\"header-actions\" *ngIf=\"!isFrozen && actions.length > 0\">\n        <div class=\"optionsBtn dropdown\" dropdown container=\"body\" placement=\"bottom right\">\n          <a\n            title=\"{{ 'Settings' | translate }}\"\n            href=\"\"\n            class=\"btnIcon c8y-dropdown\"\n            (click)=\"(false)\"\n            dropdownToggle\n          >\n            <i [c8yIcon]=\"'cog'\"></i>\n          </a>\n          <ul\n            class=\"dropdown-menu dropdown-menu-right\"\n            style=\"right:-1px;\"\n            *dropdownMenu\n          >\n            <ng-container *ngFor=\"let action of actions\">\n              <ng-container *ngTemplateOutlet=\"action.template\"></ng-container>\n            </ng-container>\n          </ul>\n        </div>\n      </div>\n    </div>\n    <div class=\"card-inner-scroll\">\n      <ng-content></ng-content>\n    </div>\n    <div\n      *ngIf=\"!isFrozen && !isDragging\"\n      class=\"resize-handle\"\n      cdkDrag\n      [cdkDragDisabled]=\"isFrozen\"\n      (cdkDragStarted)=\"resizeStarted($event)\"\n      (cdkDragEnded)=\"reset($event)\"\n    ></div>\n    <div class=\"resize-icon\" *ngIf=\"!isFrozen && !isDragging\"></div>\n\n    <div *cdkDragPlaceholder class=\"card-placeholder\"></div>\n  </div>\n</div>\n",
                host: {
                    class: 'dashboard-grid-child'
                }
            }]
    }], function () { return [{ type: ɵngcc1.DashboardComponent }, { type: ɵngcc2.DomSanitizer }, { type: ɵngcc0.ElementRef }]; }, { actions: [{
            type: ContentChildren,
            args: [DashboardChildActionComponent]
        }], width: [{
            type: Input
        }], height: [{
            type: Input
        }], margin: [{
            type: Input
        }], isFrozen: [{
            type: Input
        }], useIntersection: [{
            type: Input
        }], changeStart: [{
            type: Output
        }], changeEnd: [{
            type: Output
        }], class: [{
            type: Input
        }], inlineStyle: [{
            type: HostBinding,
            args: ['attr.style']
        }], x: [{
            type: Input
        }], y: [{
            type: Input
        }], data: [{
            type: Input
        }] }); })();
