// @c8y/ngx-components/esm2015/core/dashboard/dashboard.module.js
import { DragDropModule } from '@angular/cdk/drag-drop';
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { ActionBarModule } from '../action-bar/action-bar.module';
import { CommonModule } from '../common/common.module';
import { DynamicComponentModule } from '../dynamic-component/dynamic-component.module';
import { DashboardChildActionComponent } from './dashboard-child-action.component';
import { DashboardChildTitleComponent } from './dashboard-child-title.component';
import { DashboardChildComponent } from './dashboard-child.component';
import { DashboardComponent } from './dashboard.component';
import { WidgetsDashboardComponent } from './widgets-dashboard.component';
import { HeaderModule } from '../header/header.module';
import { DocsModule } from '../docs/docs.module';
import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
export class DashboardModule {
}
DashboardModule.ɵfac = function DashboardModule_Factory(t) { return new (t || DashboardModule)(); };
DashboardModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DashboardModule });
DashboardModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            ActionBarModule,
            DynamicComponentModule,
            DragDropModule,
            BsDropdownModule.forRoot(),
            HeaderModule,
            DocsModule,
            BreadcrumbModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DashboardChildComponent,
                    DashboardComponent,
                    DashboardChildTitleComponent,
                    DashboardChildActionComponent,
                    WidgetsDashboardComponent
                ],
                imports: [
                    CommonModule,
                    ActionBarModule,
                    DynamicComponentModule,
                    DragDropModule,
                    BsDropdownModule.forRoot(),
                    HeaderModule,
                    DocsModule,
                    BreadcrumbModule
                ],
                exports: [
                    DashboardComponent,
                    DashboardChildComponent,
                    DashboardChildTitleComponent,
                    DashboardChildActionComponent,
                    WidgetsDashboardComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DashboardModule, { declarations: function () { return [DashboardChildComponent,
        DashboardComponent,
        DashboardChildTitleComponent,
        DashboardChildActionComponent,
        WidgetsDashboardComponent]; }, imports: function () { return [CommonModule,
        ActionBarModule,
        DynamicComponentModule,
        DragDropModule, ɵngcc1.BsDropdownModule, HeaderModule,
        DocsModule,
        BreadcrumbModule]; }, exports: function () { return [DashboardComponent,
        DashboardChildComponent,
        DashboardChildTitleComponent,
        DashboardChildActionComponent,
        WidgetsDashboardComponent]; } }); })();
