// @c8y/ngx-components/esm2015/core/dashboard/dashboard-child-action.component.js
import { Component, ViewChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["template"];
function DashboardChildActionComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = ["*"];
export class DashboardChildActionComponent {
}
DashboardChildActionComponent.ɵfac = function DashboardChildActionComponent_Factory(t) { return new (t || DashboardChildActionComponent)(); };
DashboardChildActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DashboardChildActionComponent, selectors: [["c8y-dashboard-child-action"]], viewQuery: function DashboardChildActionComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""]], template: function DashboardChildActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, DashboardChildActionComponent_ng_template_0_Template, 2, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
DashboardChildActionComponent.propDecorators = {
    template: [{ type: ViewChild, args: ['template', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardChildActionComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-dashboard-child-action',
                template: "<ng-template #template>\n  <li>\n    <ng-content></ng-content>\n  </li>\n</ng-template>\n"
            }]
    }], null, { template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }] }); })();
