// @c8y/ngx-components/esm2015/core/dashboard/dashboard-child-title.component.js
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class DashboardChildTitleComponent {
}
DashboardChildTitleComponent.ɵfac = function DashboardChildTitleComponent_Factory(t) { return new (t || DashboardChildTitleComponent)(); };
DashboardChildTitleComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DashboardChildTitleComponent, selectors: [["c8y-dashboard-child-title"]], ngContentSelectors: _c0, decls: 2, vars: 0, consts: [[1, "card-title"]], template: function DashboardChildTitleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardChildTitleComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-dashboard-child-title',
                template: "<div class=\"card-title\">\n  <ng-content></ng-content>\n</div>\n"
            }]
    }], null, null); })();
