// @c8y/cli/dist/webpack/config/angular.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const path = require("path");
const webpack_1 = require("@ngtools/webpack");
function config(env) {
    let angularCompilerPluginInstance;
    const plugins = [];
    const rules = [];
    const tsRule = {
        test: /(?:\.ngfactory\.js|\.ngstyle\.js|\.ts)$/,
        use: ['@ngtools/webpack']
    };
    try {
        angularCompilerPluginInstance = new webpack_1.AngularWebpackPlugin({
            tsconfig: path.resolve(env.tsConfigPath || env.app.tsConfigPath || 'tsconfig.json'),
            compilerOptions: {
                entryModule: env.app.entryModule && path.resolve(env.app.entryModule),
                skipCodeGeneration: true,
                sourceMap: env.mode !== 'production' && env.sourceMap !== 'false'
            },
            jitMode: true
        });
    }
    catch (e) {
        // probably we should log this
        // well..., probably we should.
        console.error(e);
    }
    if (angularCompilerPluginInstance) {
        plugins.push(angularCompilerPluginInstance);
        rules.push(tsRule);
    }
    return {
        module: { rules },
        plugins
    };
}
exports.config = config;
