// @c8y/ngx-components/esm2015/core/authentication/new-password.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
import { LoginService } from '../login/login.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../login/login.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../forms/form-group.component';
import * as ɵngcc4 from '../i18n/c8y-translate.directive';
import * as ɵngcc5 from '../forms/required-input-placeholder.directive';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from '../forms/default-validation.directive';
import * as ɵngcc8 from './password-confirm.directive';
import * as ɵngcc9 from './password-strength.component';
import * as ɵngcc10 from './password-check-list.component';
import * as ɵngcc11 from '../i18n/c8y-translate.pipe';

function NewPasswordComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Change password"), " ");
} }
function NewPasswordComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Cancel password change"), " ");
} }
function NewPasswordComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelementStart(1, "div", 5);
    ɵngcc0.ɵɵelementStart(2, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(3, "label", 6);
    ɵngcc0.ɵɵtext(4, "Password");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "input", 7, 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function NewPasswordComponent_div_4_Template_input_ngModelChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.model.newPassword = $event; })("change", function NewPasswordComponent_div_4_Template_input_change_5_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.newPasswordChanged(); })("input", function NewPasswordComponent_div_4_Template_input_input_5_listener() { ɵngcc0.ɵɵrestoreView(_r6); const _r4 = ɵngcc0.ɵɵreference(11); return _r4.control.updateValueAndValidity(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(8, "label", 9);
    ɵngcc0.ɵɵtext(9, "Confirm password");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "input", 10, 11);
    ɵngcc0.ɵɵlistener("ngModelChange", function NewPasswordComponent_div_4_Template_input_ngModelChange_10_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.model.newPasswordConfirm = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 12);
    ɵngcc0.ɵɵelementStart(13, "c8y-password-strength", 13);
    ɵngcc0.ɵɵlistener("strength", function NewPasswordComponent_div_4_Template_c8y_password_strength_strength_13_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.model.strength = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(14, "c8y-password-check-list", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r2.model.newPassword);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r2.model.newPasswordConfirm);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("password", ctx_r2.model.newPassword);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("password", ctx_r2.model.newPassword)("hidden", !ctx_r2.passwordEnforced);
} }
export class NewPasswordComponent {
    constructor(loginService) {
        this.loginService = loginService;
        this.password = new EventEmitter();
        this.model = {};
        this.changePassword = false;
        this.passwordEnforced = false;
    }
    ngOnInit() {
        this.loadPasswordStrengthSettings();
        this.loginService.getGreenMinLength().then(value => {
            this.minlength = value;
        });
    }
    newPasswordChanged() {
        this.password.emit({
            password: this.model.newPassword,
            passwordStrength: this.model.strength
        });
    }
    loadPasswordStrengthSettings() {
        return __awaiter(this, void 0, void 0, function* () {
            this.passwordEnforced = yield this.loginService.getPasswordStrengthEnforced({ refresh: true });
        });
    }
    showPasswordStrength() {
        return this.model.newPassword && this.model.newPassword.length >= this.minlength;
    }
    toggleChangePassword() {
        this.changePassword = !this.changePassword;
        if (!this.changePassword) {
            this.password.emit({});
            this.model = {};
        }
    }
}
NewPasswordComponent.ɵfac = function NewPasswordComponent_Factory(t) { return new (t || NewPasswordComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService)); };
NewPasswordComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NewPasswordComponent, selectors: [["c8y-new-password"]], outputs: { password: "password" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgForm }])], decls: 5, vars: 3, consts: [[1, "form-group"], ["type", "button", 1, "btn", "btn-default", 3, "click"], [4, "ngIf"], ["class", "row content-flex-50", 4, "ngIf"], [1, "row", "content-flex-50"], [1, "col-7"], ["for", "newPassword", "translate", ""], ["id", "newPassword", "name", "newPassword", "type", "password", "required", "", "c8yDefaultValidation", "password", "autocomplete", "new-password", 1, "form-control", 3, "ngModel", "ngModelChange", "change", "input"], ["newPassword", "ngModel"], ["for", "newConfirmPassword", "translate", ""], ["id", "newConfirmPassword", "name", "newPasswordConfirm", "type", "password", "required", "", "passwordConfirm", "newPassword", "autocomplete", "new-password", 1, "form-control", 3, "ngModel", "ngModelChange"], ["newPasswordConfirm", "ngModel"], [1, "col-5"], [3, "password", "strength"], [1, "m-t-32", 3, "password", "hidden"]], template: function NewPasswordComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function NewPasswordComponent_Template_button_click_1_listener() { return ctx.toggleChangePassword(); });
        ɵngcc0.ɵɵtemplate(2, NewPasswordComponent_ng_container_2_Template, 3, 3, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(3, NewPasswordComponent_ng_container_3_Template, 3, 3, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, NewPasswordComponent_div_4_Template, 15, 5, "div", 3);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.changePassword);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.changePassword);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.changePassword);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.FormGroupComponent, ɵngcc4.C8yTranslateDirective, ɵngcc5.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.RequiredValidator, ɵngcc7.DefaultValidationDirective, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc8.PasswordConfirm, ɵngcc9.PasswordStrengthComponent, ɵngcc10.PasswordCheckListComponent], pipes: [ɵngcc11.C8yTranslatePipe], encapsulation: 2 });
NewPasswordComponent.ctorParameters = () => [
    { type: LoginService }
];
NewPasswordComponent.propDecorators = {
    password: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NewPasswordComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-new-password',
                template: "<div class=\"form-group\">\n  <button type=\"button\" class=\"btn btn-default\" (click)=\"toggleChangePassword()\">\n    <ng-container *ngIf=\"!changePassword\">\n      {{ 'Change password' | translate }}\n    </ng-container>\n    <ng-container *ngIf=\"changePassword\">\n      {{ 'Cancel password change' | translate }}\n    </ng-container>\n  </button>\n</div>\n\n<div *ngIf=\"changePassword\" class=\"row content-flex-50\">\n  <div class=\"col-7\">\n    <c8y-form-group>\n      <label for=\"newPassword\" translate>Password</label>\n      <input\n        id=\"newPassword\"\n        name=\"newPassword\"\n        type=\"password\"\n        #newPassword=\"ngModel\"\n        class=\"form-control\"\n        [(ngModel)]=\"model.newPassword\"\n        (change)=\"newPasswordChanged()\"\n        (input)=\"newPasswordConfirm.control.updateValueAndValidity()\"\n        required\n        c8yDefaultValidation=\"password\"\n        autocomplete=\"new-password\"\n      />\n    </c8y-form-group>\n\n    <c8y-form-group>\n      <label for=\"newConfirmPassword\" translate>Confirm password</label>\n      <input\n        id=\"newConfirmPassword\"\n        name=\"newPasswordConfirm\"\n        type=\"password\"\n        #newPasswordConfirm=\"ngModel\"\n        class=\"form-control\"\n        [(ngModel)]=\"model.newPasswordConfirm\"\n        required\n        passwordConfirm=\"newPassword\"\n        autocomplete=\"new-password\"\n      />\n    </c8y-form-group>\n  </div>\n  <div class=\"col-5\">\n    <c8y-password-strength\n      [password]=\"model.newPassword\"\n      (strength)=\"model.strength = $event\"\n    ></c8y-password-strength>\n\n    <c8y-password-check-list\n      class=\"m-t-32\"\n      [password]=\"model.newPassword\"\n      [hidden]=\"!passwordEnforced\"\n    ></c8y-password-check-list>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }]; }, { password: [{
            type: Output
        }] }); })();
