// @c8y/ngx-components/esm2015/core/authentication/provide-phone-number.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { LoginService } from '../login/login.service';
import { AlertService } from '../alert/alert.service';
import { LoginViews } from '../login/login.model';
import { UserService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../login/login.service';
import * as ɵngcc2 from '../alert/alert.service';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../i18n/c8y-translate.directive';
import * as ɵngcc6 from '../forms/form-group.component';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '../forms/required-input-placeholder.directive';
import * as ɵngcc9 from '../forms/phone-validation.directive';
import * as ɵngcc10 from '../forms/default-validation.directive';
import * as ɵngcc11 from '../i18n/c8y-translate.pipe';
export class ProvidePhoneNumberComponent {
    constructor(loginService, alert, userService) {
        this.loginService = loginService;
        this.alert = alert;
        this.userService = userService;
        this.onCancel = new EventEmitter();
        this.onChangeView = new EventEmitter();
        this.requestInProgress = false;
        this.sendTfa = '0';
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.requestInProgress = true;
                yield this.userService.savePhoneNumber(this.phoneNumber);
                yield this.sendTFASms();
                this.onChangeView.emit({
                    view: LoginViews.SmsChallenge,
                    credentials: this.credentials
                });
            }
            catch (e) {
                this.alert.addServerFailure(e);
            }
            finally {
                this.requestInProgress = false;
            }
        });
    }
    sendTFASms() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.userService.verifyTFACode(this.sendTfa);
            }
            catch (e) {
                if (e.res.status === 403) {
                    this.loginService.cleanMessages();
                    this.loginService.addSuccessMessage('send_sms');
                }
                else {
                    throw e;
                }
            }
        });
    }
}
ProvidePhoneNumberComponent.ɵfac = function ProvidePhoneNumberComponent_Factory(t) { return new (t || ProvidePhoneNumberComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.UserService)); };
ProvidePhoneNumberComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ProvidePhoneNumberComponent, selectors: [["c8y-provide-phone-number"]], inputs: { credentials: "credentials" }, outputs: { onCancel: "onCancel", onChangeView: "onChangeView" }, decls: 19, vars: 18, consts: [["role", "form", "novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["twoFactorForm", "ngForm"], ["translate", "", 1, "legend", "form-block", "center"], [3, "ngClass"], ["translate", ""], ["type", "text", "name", "phone", "autocomplete", "off", "c8yPhoneValidation", "", "c8yDefaultValidation", "phoneNumber", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["contactPhone", "ngModel"], ["type", "submit", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"], [1, "d-flex", "m-t-8"], ["href", "#", 1, "small", "pointer", "m-l-auto", 3, "title", "click"]], template: function ProvidePhoneNumberComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function ProvidePhoneNumberComponent_Template_form_ngSubmit_0_listener() { return ctx.save(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Two-factor authentication ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-form-group", 3);
        ɵngcc0.ɵɵelementStart(5, "label", 4);
        ɵngcc0.ɵɵtext(6, "Provide your phone number");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "input", 5, 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function ProvidePhoneNumberComponent_Template_input_ngModelChange_7_listener($event) { return ctx.phoneNumber = $event; });
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "button", 7);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 8);
        ɵngcc0.ɵɵelementStart(15, "a", 9);
        ɵngcc0.ɵɵlistener("click", function ProvidePhoneNumberComponent_Template_a_click_15_listener() { return ctx.onCancel.emit(); });
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵtext(17);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngClass", ctx.requestInProgress || _r0.invalid ? "p-b-8" : "");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(9, 8, "e.g. +49 9 876 543 210`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.phoneNumber);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 10, "Save and continue"));
        ɵngcc0.ɵɵproperty("disabled", ctx.requestInProgress || _r0.invalid);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 12, "Save and continue"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(16, 14, "Login"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(18, 16, "Login"), " ");
    } }, directives: [ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc5.C8yTranslateDirective, ɵngcc6.FormGroupComponent, ɵngcc7.NgClass, ɵngcc8.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc9.PhoneValidationDirective, ɵngcc10.DefaultValidationDirective, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel], pipes: [ɵngcc11.C8yTranslatePipe], encapsulation: 2 });
ProvidePhoneNumberComponent.ctorParameters = () => [
    { type: LoginService },
    { type: AlertService },
    { type: UserService }
];
ProvidePhoneNumberComponent.propDecorators = {
    credentials: [{ type: Input }],
    onCancel: [{ type: Output }],
    onChangeView: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProvidePhoneNumberComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-provide-phone-number',
                template: "<form #twoFactorForm=\"ngForm\" role=\"form\" class=\"loginForm\" (ngSubmit)=\"save()\" novalidate>\n  <div class=\"legend form-block center\" translate>\n    Two-factor authentication\n  </div>\n\n  <c8y-form-group [ngClass]=\"requestInProgress || twoFactorForm.invalid ? 'p-b-8' : ''\">\n    <label translate>Provide your phone number</label>\n\n    <input\n      class=\"form-control\"\n      [(ngModel)]=\"phoneNumber\"\n      #contactPhone=\"ngModel\"\n      type=\"text\"\n      name=\"phone\"\n      autocomplete=\"off\"\n      placeholder=\"{{ 'e.g. +49 9 876 543 210`LOCALIZE`' | translate }}\"\n      c8yPhoneValidation\n      c8yDefaultValidation=\"phoneNumber\"\n      required\n    />\n  </c8y-form-group>\n\n  <button\n    title=\"{{ 'Save and continue' | translate }}\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n    [disabled]=\"requestInProgress || twoFactorForm.invalid\"\n  >\n    {{ 'Save and continue' | translate }}\n  </button>\n\n  <div class=\"d-flex m-t-8\">\n    <a\n      title=\"{{ 'Login' | translate }}\"\n      class=\"small pointer m-l-auto\"\n      href=\"#\"\n      (click)=\"onCancel.emit()\"\n    >\n      {{ 'Login' | translate }}\n    </a>\n  </div>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.AlertService }, { type: ɵngcc3.UserService }]; }, { onCancel: [{
            type: Output
        }], onChangeView: [{
            type: Output
        }], credentials: [{
            type: Input
        }] }); })();
