// @c8y/ngx-components/esm2015/core/authentication/totp-challenge.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { ControlContainer, NgForm } from '@angular/forms';
import { LoginService } from '../login/login.service';
import { UserService } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../login/login.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../alert/alert.service';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../forms/form-group.component';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from '../i18n/c8y-translate.directive';
import * as ɵngcc8 from '../forms/required-input-placeholder.directive';
import * as ɵngcc9 from '../forms/messages.component';
import * as ɵngcc10 from '../forms/message.directive';
import * as ɵngcc11 from '../i18n/c8y-translate.pipe';

function TotpChallengeComponent_c8y_message_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 9);
    ɵngcc0.ɵɵtext(1, " Invalid verification code. In case of key loss, please contact your platform administrator. ");
    ɵngcc0.ɵɵelementEnd();
} }
function TotpChallengeComponent_p_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 10);
    ɵngcc0.ɵɵtext(1, " In case of key loss, please contact your platform administrator. ");
    ɵngcc0.ɵɵelementEnd();
} }
function TotpChallengeComponent_button_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 11);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r0 = ɵngcc0.ɵɵreference(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Verify"));
    ɵngcc0.ɵɵproperty("disabled", !_r0.form.valid);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 5, "Verify"), " ");
} }
function TotpChallengeComponent_button_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 12);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Verifying\u2026"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Verifying\u2026"), " ");
} }
export class TotpChallengeComponent {
    constructor(loginService, users, alert) {
        this.loginService = loginService;
        this.users = users;
        this.alert = alert;
        /**
         * Calls the verify endpoint if set to true (default true)
         */
        this.verify = true;
        /**
         * Emits the token on success.
         */
        this.onSuccess = new EventEmitter();
        this.model = {
            token: ''
        };
        this.loading = false;
        this.hasError = false;
    }
    verifyCode() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.loading = true;
                this.hasError = false;
                if (this.verify) {
                    yield this.users.verifyTotpCode(this.model.token);
                }
                this.onSuccess.emit(this.model.token);
            }
            catch (e) {
                this.hasError = true;
                this.alert.removeLastDanger();
                this.loading = false;
            }
        });
    }
}
TotpChallengeComponent.ɵfac = function TotpChallengeComponent_Factory(t) { return new (t || TotpChallengeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
TotpChallengeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TotpChallengeComponent, selectors: [["c8y-totp-challenge"]], inputs: { verify: "verify", loading: "loading", hasError: "hasError" }, outputs: { onSuccess: "onSuccess" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgForm }])], decls: 12, vars: 11, consts: [["role", "form", "novalidate", "", 1, "loginForm", 3, "ngSubmit"], ["totpForm", "ngForm"], [3, "hasError", "novalidation", "ngClass"], ["translate", "", "for", "totpToken"], ["id", "totpToken", "name", "totpToken", "type", "text", "autocapitalize", "off", "autocorrect", "off", "autocomplete", "off", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["translate", "", 4, "ngIf"], ["id", "helpinput", "class", "help-block", "translate", "", 4, "ngIf"], ["type", "submit", "class", "btn btn-primary btn-lg btn-block form-group", 3, "title", "disabled", 4, "ngIf"], ["type", "submit", "class", "btn btn-primary btn-lg btn-block btn-pending", 3, "title", 4, "ngIf"], ["translate", ""], ["id", "helpinput", "translate", "", 1, "help-block"], ["type", "submit", 1, "btn", "btn-primary", "btn-lg", "btn-block", "form-group", 3, "title", "disabled"], ["type", "submit", 1, "btn", "btn-primary", "btn-lg", "btn-block", "btn-pending", 3, "title"]], template: function TotpChallengeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0, 1);
        ɵngcc0.ɵɵlistener("ngSubmit", function TotpChallengeComponent_Template_form_ngSubmit_0_listener() { return ctx.verifyCode(); });
        ɵngcc0.ɵɵelementStart(2, "c8y-form-group", 2);
        ɵngcc0.ɵɵelementStart(3, "label", 3);
        ɵngcc0.ɵɵtext(4, " Verification code ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "input", 4);
        ɵngcc0.ɵɵlistener("ngModelChange", function TotpChallengeComponent_Template_input_ngModelChange_5_listener($event) { return ctx.model.token = $event; });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "c8y-messages");
        ɵngcc0.ɵɵtemplate(8, TotpChallengeComponent_c8y_message_8_Template, 2, 0, "c8y-message", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, TotpChallengeComponent_p_9_Template, 2, 0, "p", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, TotpChallengeComponent_button_10_Template, 4, 7, "button", 7);
        ɵngcc0.ɵɵtemplate(11, TotpChallengeComponent_button_11_Template, 4, 6, "button", 8);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("hasError", ctx.hasError)("novalidation", true)("ngClass", ctx.hasError ? "p-b-24" : "");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(6, 9, "e.g."), " 624327");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.token);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.hasError);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.loading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loading);
    } }, directives: [ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc5.FormGroupComponent, ɵngcc6.NgClass, ɵngcc7.C8yTranslateDirective, ɵngcc8.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc9.MessagesComponent, ɵngcc6.NgIf, ɵngcc10.MessageDirective], pipes: [ɵngcc11.C8yTranslatePipe], encapsulation: 2 });
TotpChallengeComponent.ctorParameters = () => [
    { type: LoginService },
    { type: UserService },
    { type: AlertService }
];
TotpChallengeComponent.propDecorators = {
    verify: [{ type: Input }],
    onSuccess: [{ type: Output }],
    loading: [{ type: Input }],
    hasError: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TotpChallengeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-totp-challenge',
                template: "<form #totpForm=\"ngForm\" role=\"form\" class=\"loginForm\" (ngSubmit)=\"verifyCode()\" novalidate>\n  <c8y-form-group [hasError]=\"hasError\" [novalidation]=\"true\" [ngClass]=\"hasError ? 'p-b-24' : ''\">\n    <label translate for=\"totpToken\">\n      Verification code\n    </label>\n\n    <input\n      id=\"totpToken\"\n      [(ngModel)]=\"model.token\"\n      name=\"totpToken\"\n      type=\"text\"\n      autocapitalize=\"off\"\n      autocorrect=\"off\"\n      autocomplete=\"off\"\n      class=\"form-control\"\n      placeholder=\"{{ 'e.g.' | translate }} 624327\"\n      required\n    />\n\n    <c8y-messages>\n      <c8y-message *ngIf=\"hasError\" translate>\n        Invalid verification code. In case of key loss, please contact your platform administrator.\n      </c8y-message>\n    </c8y-messages>\n    <p id=\"helpinput\" *ngIf=\"!hasError\" class=\"help-block\" translate>\n      In case of key loss, please contact your platform administrator.\n    </p>\n  </c8y-form-group>\n  <button *ngIf=\"!loading\"\n    title=\"{{ 'Verify' | translate }}\"\n    [disabled]=\"!totpForm.form.valid\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block form-group\"\n  >\n    {{ 'Verify' | translate }}\n  </button>\n\n  <button *ngIf=\"loading\"\n    title=\"{{ 'Verifying\u2026' | translate }}\"\n    type=\"submit\"\n    class=\"btn btn-primary btn-lg btn-block btn-pending\"\n  >\n    {{ 'Verifying\u2026' | translate }}\n  </button>\n</form>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.UserService }, { type: ɵngcc3.AlertService }]; }, { verify: [{
            type: Input
        }], onSuccess: [{
            type: Output
        }], loading: [{
            type: Input
        }], hasError: [{
            type: Input
        }] }); })();
