// @c8y/ngx-components/esm2015/core/authentication/totp-setup.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { UserService } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
import { gettext } from '../i18n';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../alert/alert.service';
import * as ɵngcc3 from '../i18n/c8y-translate.directive';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from 'angularx-qrcode';

function TotpSetupComponent_qrcode_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "qrcode", 4);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("width", 300)("qrdata", ctx_r0.qrData)("cssClass", "m-b-16")("errorCorrectionLevel", "M");
} }
export class TotpSetupComponent {
    constructor(user, alert) {
        this.user = user;
        this.alert = alert;
    }
    get qrCodeImage() {
        return this.totpSecret ? this.totpSecret.secretQrUrl : '';
    }
    get secret() {
        return this.totpSecret ? this.totpSecret.rawSecret : '';
    }
    get qrData() {
        // TODO: waiting for BE:, now we need extract it from secretQrUrl
        // https://cumulocity.atlassian.net/browse/MTM-36387
        // return this.totpSecret ? this.totpSecret.qrData : '';
        const otpAuth = this.qrCodeImage.split('chl=')[1];
        if (!otpAuth) {
            this.alert.danger(gettext('Failed to generate a QR code.'));
            return '';
        }
        return decodeURIComponent(otpAuth);
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const { data } = yield this.user.generateTotpSecret();
                this.totpSecret = data;
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
            }
        });
    }
}
TotpSetupComponent.ɵfac = function TotpSetupComponent_Factory(t) { return new (t || TotpSetupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
TotpSetupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TotpSetupComponent, selectors: [["c8y-totp-setup"]], decls: 6, vars: 2, consts: [[1, "text-center"], ["translate", "", 1, "lead", "m-b-16"], [3, "width", "qrdata", "cssClass", "errorCorrectionLevel", 4, "ngIf"], [1, "text-center", "text-muted"], [3, "width", "qrdata", "cssClass", "errorCorrectionLevel"]], template: function TotpSetupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "p", 1);
        ɵngcc0.ɵɵtext(2, " Scan this QR code with your smartphone using the Google Authenticator app. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, TotpSetupComponent_qrcode_3_Template, 1, 4, "qrcode", 2);
        ɵngcc0.ɵɵelementStart(4, "p", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.totpSecret);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.secret, " ");
    } }, directives: [ɵngcc3.C8yTranslateDirective, ɵngcc4.NgIf, ɵngcc5.QRCodeComponent], encapsulation: 2 });
TotpSetupComponent.ctorParameters = () => [
    { type: UserService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TotpSetupComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-totp-setup',
                template: "<div class=\"text-center\">\n  <p\n    class=\"lead m-b-16\"\n    translate\n  >\n    Scan this QR code with your smartphone using the Google Authenticator app.\n  </p>\n\n\n  <qrcode\n    *ngIf=\"totpSecret\"\n    [width]=\"300\"\n    [qrdata]=\"qrData\"\n    [cssClass]=\"'m-b-16'\"\n    [errorCorrectionLevel]=\"'M'\"\n  ></qrcode>\n  <p class=\"text-center text-muted\">\n    {{secret}}\n  </p>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.UserService }, { type: ɵngcc2.AlertService }]; }, null); })();
