// @c8y/ngx-components/esm2015/core/file-picker/file-picker.module.js
import { NgModule } from '@angular/core';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { FilePickerComponent } from './file-picker.component';
import { DropAreaModule } from '../drop-area/drop-area.module';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { PopoverModule } from 'ngx-bootstrap/popover';
/**
 * The angular module definition for a drop-area.
 * @exports FilePickerComponent
 */
import * as ɵngcc0 from '@angular/core';
export class FilePickerModule {
}
FilePickerModule.ɵfac = function FilePickerModule_Factory(t) { return new (t || FilePickerModule)(); };
FilePickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: FilePickerModule });
FilePickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            C8yFormsModule,
            CommonModule,
            DropAreaModule,
            FormsModule,
            C8yCommonModule,
            PopoverModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FilePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    C8yFormsModule,
                    CommonModule,
                    DropAreaModule,
                    FormsModule,
                    C8yCommonModule,
                    PopoverModule
                ],
                exports: [FilePickerComponent],
                declarations: [FilePickerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FilePickerModule, { declarations: function () { return [FilePickerComponent]; }, imports: function () { return [C8yFormsModule,
        CommonModule,
        DropAreaModule,
        FormsModule,
        C8yCommonModule,
        PopoverModule]; }, exports: function () { return [FilePickerComponent]; } }); })();
