// @c8y/ngx-components/esm2015/sensor-phone/connect-smartphone.factory.js
import { Injectable } from '@angular/core';
import { gettext, OptionsService } from '@c8y/ngx-components';
import { BsModalService } from 'ngx-bootstrap/modal';
import { of } from 'rxjs';
import { SensorPhoneModalComponent } from './sensor-phone-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
export class ConnectSmartphoneFactory {
    constructor(optionsService, bsModalService) {
        this.optionsService = optionsService;
        this.bsModalService = bsModalService;
    }
    get() {
        if (this.optionsService.get('sensorPhone', false)) {
            return of([
                {
                    icon: 'c8y-icon c8y-icon-mobile-add',
                    type: 'quicklink',
                    label: gettext('Connect smartphone'),
                    url: undefined,
                    click: () => {
                        this.bsModalService.show(SensorPhoneModalComponent, {
                            backdrop: 'static',
                            class: 'lg'
                        });
                    }
                }
            ]);
        }
    }
}
ConnectSmartphoneFactory.ɵfac = function ConnectSmartphoneFactory_Factory(t) { return new (t || ConnectSmartphoneFactory)(ɵngcc0.ɵɵinject(ɵngcc1.OptionsService), ɵngcc0.ɵɵinject(ɵngcc2.BsModalService)); };
ConnectSmartphoneFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ConnectSmartphoneFactory, factory: ConnectSmartphoneFactory.ɵfac });
ConnectSmartphoneFactory.ctorParameters = () => [
    { type: OptionsService },
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConnectSmartphoneFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.OptionsService }, { type: ɵngcc2.BsModalService }]; }, null); })();
