// @c8y/ngx-components/esm2015/sensor-phone/sensor-phone.module.js
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_DOCS } from '@c8y/ngx-components';
import { ConnectSmartphoneFactory } from './connect-smartphone.factory';
import { SensorPhoneModalComponent } from './sensor-phone-modal.component';
import { QRCodeModule } from 'angularx-qrcode';
import { SensorPhoneService } from './sensor-phone.service';
import * as ɵngcc0 from '@angular/core';
export class SensorPhoneModule {
}
SensorPhoneModule.ɵfac = function SensorPhoneModule_Factory(t) { return new (t || SensorPhoneModule)(); };
SensorPhoneModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SensorPhoneModule });
SensorPhoneModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_DOCS,
            useClass: ConnectSmartphoneFactory,
            multi: true
        },
        SensorPhoneService
    ], imports: [[CoreModule, QRCodeModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SensorPhoneModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, QRCodeModule],
                declarations: [SensorPhoneModalComponent],
                entryComponents: [SensorPhoneModalComponent],
                providers: [
                    {
                        provide: HOOK_DOCS,
                        useClass: ConnectSmartphoneFactory,
                        multi: true
                    },
                    SensorPhoneService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SensorPhoneModule, { declarations: function () { return [SensorPhoneModalComponent]; }, imports: function () { return [CoreModule, QRCodeModule]; } }); })();
