// @c8y/ngx-components/esm2015/core/forms/default-validation.directive.js
import { NG_VALIDATORS } from '@angular/forms';
import { Directive, Input } from '@angular/core';
import { ValidationPattern } from './validation-pattern';
import * as ɵngcc0 from '@angular/core';
export class DefaultValidationDirective {
    constructor() {
        this.c8yDefaultValidation = '';
    }
    validate(control) {
        if (!control.value) {
            return null;
        }
        const validation = ValidationPattern.get(this.c8yDefaultValidation);
        if (!validation) {
            throw new Error(`Pattern ${this.c8yDefaultValidation} not found.`);
        }
        let valid = true;
        if (validation.pattern) {
            valid = validation.pattern.test(control.value);
        }
        if (validation.maxLength) {
            valid = control.value.length < validation.maxLength;
        }
        if (validation.minLength) {
            valid = control.value.length >= validation.minLength;
        }
        return valid ? null : { [this.c8yDefaultValidation]: true };
    }
}
DefaultValidationDirective.ɵfac = function DefaultValidationDirective_Factory(t) { return new (t || DefaultValidationDirective)(); };
DefaultValidationDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: DefaultValidationDirective, selectors: [["", "c8yDefaultValidation", ""]], inputs: { c8yDefaultValidation: "c8yDefaultValidation" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: DefaultValidationDirective, multi: true }])] });
DefaultValidationDirective.propDecorators = {
    c8yDefaultValidation: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DefaultValidationDirective, [{
        type: Directive,
        args: [{
                selector: '[c8yDefaultValidation]',
                providers: [{ provide: NG_VALIDATORS, useExisting: DefaultValidationDirective, multi: true }]
            }]
    }], function () { return []; }, { c8yDefaultValidation: [{
            type: Input
        }] }); })();
