// @c8y/ngx-components/esm2015/core/forms/emails-validator.directive.js
import { Directive, Input } from '@angular/core';
import { FormControl, NG_VALIDATORS, Validators } from '@angular/forms';
/**
 * A directive that adds the `emails` validator to controls marked with the
 * `emails` attribute. The directive is provided with the `NG_VALIDATORS` multi-provider list.
 *
 * **Example**
 *
 * ```html
 * <input type="text" name="emails" ngModel emails>
 * <input type="text" name="emails" ngModel emails="true">
 * <input type="text" name="emails" ngModel [emails]="true">
 * ```
 */
import * as ɵngcc0 from '@angular/core';
export class EmailsValidatorDirective {
    constructor() {
        this._enabled = false;
        this.separator = ',';
    }
    /**
     * Tracks changes to the emails attribute bound to this directive.
     */
    set emails(value) {
        this._enabled = value === '' || value === true || value === 'true';
        if (this._onChange) {
            this._onChange();
        }
    }
    /**
     * Method that validates whether email addresses are valid.
     * Returns the validation result if enabled, otherwise null.
     */
    validate(control) {
        if (!this._enabled) {
            return null;
        }
        const value = control.value || '';
        const emails = value.split(this.separator).map(email => email.trim());
        const invalid = emails.some(email => Validators.email(new FormControl(email)));
        return invalid ? { emails: true } : null;
    }
    /**
     * Registers a callback function to call when the validator inputs change.
     */
    registerOnValidatorChange(fn) {
        this._onChange = fn;
    }
}
EmailsValidatorDirective.ɵfac = function EmailsValidatorDirective_Factory(t) { return new (t || EmailsValidatorDirective)(); };
EmailsValidatorDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: EmailsValidatorDirective, selectors: [["", "emails", "", "formControlName", ""], ["", "emails", "", "formControl", ""], ["", "emails", "", "ngModel", ""]], inputs: { emails: "emails" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: EmailsValidatorDirective, multi: true }])] });
EmailsValidatorDirective.propDecorators = {
    emails: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EmailsValidatorDirective, [{
        type: Directive,
        args: [{
                selector: '[emails][formControlName],[emails][formControl],[emails][ngModel]',
                providers: [{ provide: NG_VALIDATORS, useExisting: EmailsValidatorDirective, multi: true }]
            }]
    }], function () { return []; }, { emails: [{
            type: Input
        }] }); })();
