// @c8y/ngx-components/esm2015/core/forms/filter-input.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Subject, timer } from 'rxjs';
import { debounce, distinctUntilChanged, map } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/icon.directive';
import * as ɵngcc2 from '../i18n/c8y-translate.pipe';
export class FilterInputComponent {
    constructor() {
        this.onSearch = new EventEmitter();
        this.filterChange$ = new Subject();
        this.filterTerm = '';
        this.CARRIAGE_RETURN_KEY = 13;
        this.CARRIAGE_RETURN_CODE = 'Enter';
    }
    ngOnInit() {
        this.reset();
        this.filterChange$
            .pipe(debounce((event) => event.code === this.CARRIAGE_RETURN_CODE || event.keyCode === this.CARRIAGE_RETURN_KEY
            ? timer(10)
            : timer(300)), map((e) => e.target.value), distinctUntilChanged())
            .subscribe(filterTerm => {
            this.filterTerm = filterTerm;
            this.onSearch.emit(filterTerm);
        });
    }
    reset() {
        this.filterTerm = '';
        this.onSearch.emit('');
    }
    ngOnDestroy() {
        this.filterChange$.complete();
    }
}
FilterInputComponent.ɵfac = function FilterInputComponent_Factory(t) { return new (t || FilterInputComponent)(); };
FilterInputComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FilterInputComponent, selectors: [["c8y-filter"]], inputs: { icon: "icon" }, outputs: { onSearch: "onSearch" }, decls: 6, vars: 5, consts: [[1, "input-group", "input-group-search"], ["type", "text", 1, "form-control", 3, "placeholder", "value", "keyup"], [1, "input-group-btn"], [1, "btn", "btn-clean", 3, "click"], [3, "c8yIcon"]], template: function FilterInputComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "input", 1);
        ɵngcc0.ɵɵlistener("keyup", function FilterInputComponent_Template_input_keyup_1_listener($event) { return ctx.filterChange$.next($event); });
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "span", 2);
        ɵngcc0.ɵɵelementStart(4, "button", 3);
        ɵngcc0.ɵɵlistener("click", function FilterInputComponent_Template_button_click_4_listener() { return ctx.reset(); });
        ɵngcc0.ɵɵelement(5, "i", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(2, 3, "Filter\u2026"));
        ɵngcc0.ɵɵproperty("value", ctx.filterTerm);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.filterTerm.length === 0 ? "search" : "close");
    } }, directives: [ɵngcc1.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
FilterInputComponent.propDecorators = {
    icon: [{ type: Input }],
    onSearch: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FilterInputComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-filter',
                template: "<div class=\"input-group input-group-search\">\n  <input\n    class=\"form-control\"\n    placeholder=\"{{ 'Filter\u2026' | translate }}\"\n    type=\"text\"\n    [value]=\"filterTerm\"\n    (keyup)=\"filterChange$.next($event)\"\n  />\n  <span class=\"input-group-btn\">\n    <button class=\"btn btn-clean\" (click)=\"reset()\">\n      <i [c8yIcon]=\"filterTerm.length === 0 ? 'search' : 'close'\"></i>\n    </button>\n  </span>\n</div>\n"
            }]
    }], function () { return []; }, { onSearch: [{
            type: Output
        }], icon: [{
            type: Input
        }] }); })();
