// @c8y/ngx-components/esm2015/core/forms/input-group-list-container.directive.js
import { Directive, ContentChildren } from '@angular/core';
import { InputGroupListComponent } from './input-group-list.component';
import * as ɵngcc0 from '@angular/core';
export class InputGroupListContainerDirective {
    ngAfterContentInit() {
        this.listProjections.map(element => {
            if (element !== this.listProjections.last) {
                setTimeout(() => {
                    element.plus = false;
                });
            }
        });
        this.listProjections.changes.subscribe(elements => {
            if (elements.length === 1) {
                setTimeout(() => {
                    elements.first.plus = true;
                    elements.first.minus = false;
                });
                return;
            }
            elements.map(element => {
                setTimeout(() => {
                    element.minus = true;
                });
                if (element !== elements.last) {
                    setTimeout(() => {
                        element.plus = false;
                    });
                }
                if (element === elements.last) {
                    setTimeout(() => {
                        element.plus = true;
                    });
                }
            });
        });
    }
}
InputGroupListContainerDirective.ɵfac = function InputGroupListContainerDirective_Factory(t) { return new (t || InputGroupListContainerDirective)(); };
InputGroupListContainerDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: InputGroupListContainerDirective, selectors: [["", "c8yInputGroupListContainer", ""]], contentQueries: function InputGroupListContainerDirective_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, InputGroupListComponent, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listProjections = _t);
    } } });
InputGroupListContainerDirective.propDecorators = {
    listProjections: [{ type: ContentChildren, args: [InputGroupListComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(InputGroupListContainerDirective, [{
        type: Directive,
        args: [{
                selector: '[c8yInputGroupListContainer]'
            }]
    }], null, { listProjections: [{
            type: ContentChildren,
            args: [InputGroupListComponent]
        }] }); })();
