// @c8y/ngx-components/esm2015/core/forms/json-validation-prettifier.directive.js
import { Directive, EventEmitter, Output } from '@angular/core';
import { NgControl } from '@angular/forms';
import { debounceTime, distinctUntilChanged, tap } from 'rxjs/operators';
import { gettext } from '../i18n/index';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
export class JsonValidationPrettifierDirective {
    constructor(ngCtrl) {
        this.invalidJSON = new EventEmitter();
        this.message = gettext('Must be a valid JSON object.');
        this.debounceTimeInMs = 1000;
        this.abstractCtrl = ngCtrl.control;
    }
    ngOnInit() {
        this.subscription = this.abstractCtrl.valueChanges
            .pipe(tap(() => this.clearValidationMessage()), debounceTime(this.debounceTimeInMs), distinctUntilChanged())
            .subscribe(value => {
            this.validateInputAndPrettify(value);
        });
    }
    ngOnDestroy() {
        if (this.subscription && !this.subscription.closed) {
            this.subscription.unsubscribe();
        }
    }
    validateInputAndPrettify(value) {
        if (value) {
            try {
                const validJsonObject = JSON.parse(value);
                const pretty = JSON.stringify(validJsonObject, undefined, 2);
                this.abstractCtrl.setValue(pretty);
            }
            catch (ex) {
                this.setErrorAndValidationMessage();
            }
        }
    }
    setErrorAndValidationMessage() {
        this.abstractCtrl.setErrors({ invalidBodyTemplate: true });
        this.invalidJSON.emit(this.message);
    }
    clearValidationMessage() {
        this.invalidJSON.emit(undefined);
    }
}
JsonValidationPrettifierDirective.ɵfac = function JsonValidationPrettifierDirective_Factory(t) { return new (t || JsonValidationPrettifierDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgControl)); };
JsonValidationPrettifierDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: JsonValidationPrettifierDirective, selectors: [["textarea", "prettyValidJson", ""]], outputs: { invalidJSON: "invalidJSON" } });
JsonValidationPrettifierDirective.ctorParameters = () => [
    { type: NgControl }
];
JsonValidationPrettifierDirective.propDecorators = {
    invalidJSON: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(JsonValidationPrettifierDirective, [{
        type: Directive,
        args: [{
                selector: 'textarea[prettyValidJson]'
            }]
    }], function () { return [{ type: ɵngcc1.NgControl }]; }, { invalidJSON: [{
            type: Output
        }] }); })();
