// @c8y/ngx-components/esm2015/core/forms/text-area-row-height.directive.js
import { Directive, ElementRef, Input, HostBinding, Renderer2 } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class TextAreaRowHeightDirective {
    constructor(elem, renderer) {
        this.elem = elem;
        this.renderer = renderer;
        this.rows = 1;
        this.minHeightPerRow = 20;
    }
    ngAfterViewInit() {
        const convertedValue = Number(this.heightPerRow);
        if (this.heightPerRow && convertedValue && convertedValue > this.minHeightPerRow) {
            this.setHeight(`${this.rows * convertedValue}px`);
        }
        else {
            this.setHeight(`${this.rows * this.minHeightPerRow}px`);
        }
    }
    setHeight(value) {
        this.renderer.setStyle(this.elem.nativeElement, 'height', value);
    }
}
TextAreaRowHeightDirective.ɵfac = function TextAreaRowHeightDirective_Factory(t) { return new (t || TextAreaRowHeightDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
TextAreaRowHeightDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: TextAreaRowHeightDirective, selectors: [["textarea", "heightPerRow", ""]], hostVars: 1, hostBindings: function TextAreaRowHeightDirective_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("rows", ctx.rows);
    } }, inputs: { rows: "rows", heightPerRow: "heightPerRow" } });
TextAreaRowHeightDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TextAreaRowHeightDirective.propDecorators = {
    rows: [{ type: Input }, { type: HostBinding, args: ['rows',] }],
    heightPerRow: [{ type: Input, args: ['heightPerRow',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TextAreaRowHeightDirective, [{
        type: Directive,
        args: [{
                selector: 'textarea[heightPerRow]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { rows: [{
            type: Input
        }, {
            type: HostBinding,
            args: ['rows']
        }], heightPerRow: [{
            type: Input,
            args: ['heightPerRow']
        }] }); })();
