// @c8y/ngx-components/esm2015/ecosystem/ecosystem-wizard.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { EcosystemGuard } from './ecosystem.guard';
import { InstallFromPackageComponent } from './applications/install-from-package/install-from-package.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ecosystem.guard';
export class EcosystemWizardFactory {
    constructor(ecosystemGuard) {
        this.ecosystemGuard = ecosystemGuard;
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.ecosystemGuard.isFeatureEnabled()) {
                return;
            }
            else {
                return {
                    wizardId: 'uploadApplication',
                    component: InstallFromPackageComponent,
                    name: 'Install from available packages',
                    c8yIcon: 'big-parcel'
                };
            }
        });
    }
}
EcosystemWizardFactory.ɵfac = function EcosystemWizardFactory_Factory(t) { return new (t || EcosystemWizardFactory)(ɵngcc0.ɵɵinject(ɵngcc1.EcosystemGuard)); };
EcosystemWizardFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: EcosystemWizardFactory, factory: EcosystemWizardFactory.ɵfac });
EcosystemWizardFactory.ctorParameters = () => [
    { type: EcosystemGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EcosystemWizardFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.EcosystemGuard }]; }, null); })();
