// @c8y/ngx-components/esm2015/ecosystem/ecosystem.guard.js
import { Injectable } from '@angular/core';
import { OptionsService } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class EcosystemGuard {
    constructor(optionsService) {
        this.optionsService = optionsService;
    }
    canActivate(route, state) {
        return this.isFeatureEnabled();
    }
    isFeatureEnabled() {
        return Boolean(this.optionsService.get('beta', false));
    }
}
EcosystemGuard.ɵfac = function EcosystemGuard_Factory(t) { return new (t || EcosystemGuard)(ɵngcc0.ɵɵinject(ɵngcc1.OptionsService)); };
EcosystemGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: EcosystemGuard, factory: EcosystemGuard.ɵfac });
EcosystemGuard.ctorParameters = () => [
    { type: OptionsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EcosystemGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.OptionsService }]; }, null); })();
