// @c8y/ngx-components/esm2015/ecosystem/ecosystem-tabs.factory.js
import { Injectable } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { Router } from '@angular/router';
import { EcosystemGuard } from './ecosystem.guard';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './ecosystem.guard';
export class EcosystemTabs {
    constructor(router, ecosystemGuard) {
        this.router = router;
        this.ecosystemGuard = ecosystemGuard;
    }
    get() {
        const tabs = [];
        if (this.router.url === '/ecosystem/applications' ||
            this.router.url === '/ecosystem/package-list' ||
            this.router.url === '/ecosystem/features') {
            tabs.push({
                icon: 'c8y-modules',
                priority: 1000,
                label: gettext('All applications'),
                path: 'ecosystem/applications'
            });
            if (this.ecosystemGuard.isFeatureEnabled()) {
                tabs.push({
                    icon: 'big-parcel',
                    priority: 999,
                    label: gettext('Packages'),
                    path: 'ecosystem/package-list'
                });
            }
            tabs.push({
                icon: 'tab',
                priority: 998,
                label: gettext('Features'),
                path: '/ecosystem/features'
            });
        }
        return tabs;
    }
}
EcosystemTabs.ɵfac = function EcosystemTabs_Factory(t) { return new (t || EcosystemTabs)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.EcosystemGuard)); };
EcosystemTabs.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: EcosystemTabs, factory: EcosystemTabs.ɵfac });
EcosystemTabs.ctorParameters = () => [
    { type: Router },
    { type: EcosystemGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(EcosystemTabs, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.EcosystemGuard }]; }, null); })();
