// @c8y/ngx-components/esm2015/core/audit-log/audit-log.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { AuditService } from '@c8y/client';
/**
 * Audit log component allows to show audits list. Component fetches audit records for the source object given as input.
 *
 * **Example**
 *
 * ```html
 *  <c8y-audit-log [source]="sourceId"></c8y-audit-log>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '../common/forOf.directive';
import * as ɵngcc3 from '../common/date.pipe';
import * as ɵngcc4 from '../i18n/c8y-translate.pipe';

function AuditLogComponent_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 3);
    ɵngcc0.ɵɵelementStart(1, "small", 4);
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 5);
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵelementStart(7, "div", 7);
    ɵngcc0.ɵɵelementStart(8, "div", 8);
    ɵngcc0.ɵɵelementStart(9, "span", 9);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "small", 10);
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "div", 11);
    ɵngcc0.ɵɵelementStart(15, "div", 12);
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const record_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, record_r1.time));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 6, record_r1.activity));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(record_r1.user);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(17, 8, record_r1.text));
} }
export class AuditLogComponent {
    /**
     * @ignore
     */
    constructor(audit) {
        this.audit = audit;
    }
    /**
     * After page view is initialized, component gets audit records from auditRecords endpoint.
     */
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.source) {
                this.records = yield this.audit.list({
                    withTotalPages: true,
                    source: this.source
                });
            }
        });
    }
}
AuditLogComponent.ɵfac = function AuditLogComponent_Factory(t) { return new (t || AuditLogComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AuditService)); };
AuditLogComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AuditLogComponent, selectors: [["c8y-audit-log"]], inputs: { source: "source" }, decls: 3, vars: 1, consts: [[1, "inner-scroll"], [1, "timeline-list", "list-condensed", "m-t-16"], ["class", "timeline-list-item flex-row", 4, "c8yFor", "c8yForOf"], [1, "timeline-list-item", "flex-row"], [1, "timeline-item-date", "text-right"], [1, "timeline-item-content", "flex-row"], [1, "list-item-body"], [1, "row", "text-break-word"], [1, "col-md-5"], [1, "text-truncate"], [1, "text-muted"], [1, "col-md-7"], [1, "small", "text-truncate-wrap"]], template: function AuditLogComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, AuditLogComponent_li_2_Template, 18, 10, "li", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.records);
    } }, directives: [ɵngcc2.ForOfDirective], pipes: [ɵngcc3.DatePipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
AuditLogComponent.ctorParameters = () => [
    { type: AuditService }
];
AuditLogComponent.propDecorators = {
    source: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuditLogComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-audit-log',
                template: "<div class=\"inner-scroll\">\n  <ul class=\"timeline-list list-condensed m-t-16\">\n    <li class=\"timeline-list-item flex-row\" *c8yFor=\"let record of records\">\n      <small class=\"timeline-item-date text-right\">\n        <span>{{ record.time | c8yDate }}</span>\n      </small>\n      <div class=\"timeline-item-content flex-row\">\n        <div class=\"list-item-body\">\n          <div class=\"row text-break-word\">\n            <div class=\"col-md-5\">\n              <span class=\"text-truncate\">{{ record.activity | translate }}</span>\n              <small class=\"text-muted\">{{ record.user }}</small>\n            </div>\n            <div class=\"col-md-7\">\n              <div class=\"small text-truncate-wrap\">{{ record.text | translate }}</div>\n            </div>\n          </div>\n        </div>\n      </div>\n    </li>\n  </ul>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AuditService }]; }, { source: [{
            type: Input
        }] }); })();
