// @c8y/ngx-components/esm2015/tenants/tenant-list/status.filtering-form-renderer.component.js
import { Component } from '@angular/core';
import { TenantStatus } from '@c8y/client';
import { FilteringFormRendererContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/forms';
export class StatusFilteringFormRendererComponent {
    constructor(context) {
        this.context = context;
        this.model = (this.context.property.externalFilterQuery || {}).model || {};
    }
    applyFilter() {
        this.context.applyFilter({
            externalFilterQuery: {
                model: this.model
            },
            filterPredicate: (tenant) => Boolean((!this.model.active && !this.model.suspended) ||
                (this.model.active && tenant.status === TenantStatus.ACTIVE) ||
                (this.model.suspended && tenant.status === TenantStatus.SUSPENDED))
        });
    }
    resetFilter() {
        this.context.resetFilter();
    }
}
StatusFilteringFormRendererComponent.ɵfac = function StatusFilteringFormRendererComponent_Factory(t) { return new (t || StatusFilteringFormRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FilteringFormRendererContext)); };
StatusFilteringFormRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StatusFilteringFormRendererComponent, selectors: [["ng-component"]], decls: 29, vars: 24, consts: [["filterForm", "ngForm"], [1, "m-b-8", "p-t-8"], [1, "m-b-0"], [1, "c8y-checkbox"], ["type", "checkbox", "name", "active", 3, "ngModel", "ngModelChange"], ["type", "checkbox", "name", "suspended", 3, "ngModel", "ngModelChange"], [1, "data-grid__dropdown__footer", "d-flex", "separator-top"], [1, "btn", "btn-default", "btn-sm", "m-r-8", "flex-grow", 3, "title", "click"], [1, "btn", "btn-primary", "btn-sm", "flex-grow", 3, "disabled", "title", "click"]], template: function StatusFilteringFormRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", null, 0);
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵelementStart(3, "label");
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-form-group", 2);
        ɵngcc0.ɵɵelementStart(7, "label", 3);
        ɵngcc0.ɵɵelementStart(8, "input", 4);
        ɵngcc0.ɵɵlistener("ngModelChange", function StatusFilteringFormRendererComponent_Template_input_ngModelChange_8_listener($event) { return ctx.model.active = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(9, "span");
        ɵngcc0.ɵɵelementStart(10, "span");
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "c8y-form-group", 2);
        ɵngcc0.ɵɵelementStart(14, "label", 3);
        ɵngcc0.ɵɵelementStart(15, "input", 5);
        ɵngcc0.ɵɵlistener("ngModelChange", function StatusFilteringFormRendererComponent_Template_input_ngModelChange_15_listener($event) { return ctx.model.suspended = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(16, "span");
        ɵngcc0.ɵɵelementStart(17, "span");
        ɵngcc0.ɵɵtext(18);
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "div", 6);
        ɵngcc0.ɵɵelementStart(21, "button", 7);
        ɵngcc0.ɵɵlistener("click", function StatusFilteringFormRendererComponent_Template_button_click_21_listener() { return ctx.resetFilter(); });
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵtext(23);
        ɵngcc0.ɵɵpipe(24, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "button", 8);
        ɵngcc0.ɵɵlistener("click", function StatusFilteringFormRendererComponent_Template_button_click_25_listener() { return ctx.applyFilter(); });
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵtext(27);
        ɵngcc0.ɵɵpipe(28, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(1);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 10, "Filter by status"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.active);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(12, 12, "Active`tenant`"));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.suspended);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(19, 14, "Suspended`tenant`"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(22, 16, "Reset"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(24, 18, "Reset"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(26, 20, "Apply"));
        ɵngcc0.ɵɵproperty("disabled", _r0.invalid);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(28, 22, "Apply"), " ");
    } }, directives: [ɵngcc2.ɵNgNoValidate, ɵngcc2.NgControlStatusGroup, ɵngcc2.NgForm, ɵngcc1.FormGroupComponent, ɵngcc2.CheckboxControlValueAccessor, ɵngcc2.NgControlStatus, ɵngcc2.NgModel], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
StatusFilteringFormRendererComponent.ctorParameters = () => [
    { type: FilteringFormRendererContext }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StatusFilteringFormRendererComponent, [{
        type: Component,
        args: [{
                template: "<form #filterForm=\"ngForm\">\n  <div class=\"m-b-8 p-t-8\">\n    <label>{{ 'Filter by status' | translate }}</label>\n    <c8y-form-group class=\"m-b-0\">\n      <label class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"active\" [(ngModel)]=\"model.active\" />\n        <span></span>\n        <span>{{ 'Active`tenant`' | translate }}</span>\n      </label>\n    </c8y-form-group>\n    <c8y-form-group class=\"m-b-0\">\n      <label class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"suspended\" [(ngModel)]=\"model.suspended\" />\n        <span></span>\n        <span>{{ 'Suspended`tenant`' | translate }}</span>\n      </label>\n    </c8y-form-group>\n  </div>\n</form>\n\n<div class=\"data-grid__dropdown__footer d-flex separator-top\">\n  <button\n    class=\"btn btn-default btn-sm m-r-8 flex-grow\"\n    (click)=\"resetFilter()\"\n    title=\"{{ 'Reset' | translate }}\"\n  >\n    {{ 'Reset' | translate }}\n  </button>\n  <button\n    class=\"btn btn-primary btn-sm flex-grow\"\n    [disabled]=\"filterForm.invalid\"\n    (click)=\"applyFilter()\"\n    title=\"{{ 'Apply' | translate }}\"\n  >\n    {{ 'Apply' | translate }}\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FilteringFormRendererContext }]; }, null); })();
