// @c8y/ngx-components/esm2015/tenants/tenant-list/tenant-list.guard.js
import { Injectable } from '@angular/core';
import { TenantUiService } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class TenantListGuard {
    constructor(tenantUiService) {
        this.tenantUiService = tenantUiService;
    }
    /**
     * Checks if tenant list should be active,
     * i.e. whether the current tenant can read other tenants.
     * **Note: the check is executed only once in the runtime.**
     *
     * @returns True, if the feature should be active.
     */
    canActivate() {
        if (this.active === undefined) {
            this.active = this.tenantUiService.canReadTenants();
        }
        return this.active;
    }
}
TenantListGuard.ɵfac = function TenantListGuard_Factory(t) { return new (t || TenantListGuard)(ɵngcc0.ɵɵinject(ɵngcc1.TenantUiService)); };
TenantListGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TenantListGuard, factory: TenantListGuard.ɵfac });
TenantListGuard.ctorParameters = () => [
    { type: TenantUiService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantListGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.TenantUiService }]; }, null); })();
