// @c8y/ngx-components/esm2015/core/breadcrumb/breadcrumb-item.component.js
import { Component, Input, ViewChild } from '@angular/core';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["template"];
function BreadcrumbItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c1 = ["*"];
export class BreadcrumbItemComponent {
}
BreadcrumbItemComponent.ɵfac = function BreadcrumbItemComponent_Factory(t) { return new (t || BreadcrumbItemComponent)(); };
BreadcrumbItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BreadcrumbItemComponent, selectors: [["c8y-breadcrumb-item"]], viewQuery: function BreadcrumbItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { icon: "icon", translate: "translate", label: "label", path: "path" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""]], template: function BreadcrumbItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, BreadcrumbItemComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
BreadcrumbItemComponent.propDecorators = {
    icon: [{ type: Input }],
    translate: [{ type: Input }],
    label: [{ type: Input }],
    path: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BreadcrumbItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-breadcrumb-item',
                template: "<ng-template #template>\n  <ng-content></ng-content>\n</ng-template>"
            }]
    }], null, { icon: [{
            type: Input
        }], translate: [{
            type: Input
        }], label: [{
            type: Input
        }], path: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }] }); })();
