// @c8y/ngx-components/esm2015/core/wizard/wizard-header.component.js
import { Component, ViewChild } from '@angular/core';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './wizard.component';
import * as ɵngcc2 from './wizard.service';

const _c0 = ["headerContent"];
function WizardHeaderComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = ["*"];
export class WizardHeaderComponent {
    constructor(wizard, wizardService) {
        this.wizard = wizard;
        this.wizardService = wizardService;
    }
    ngOnInit() {
        setTimeout(() => {
            this.wizardService.headerTemplate$.next({
                template: this.headerContent,
                wizardId: this.wizard.id
            });
        });
    }
}
WizardHeaderComponent.ɵfac = function WizardHeaderComponent_Factory(t) { return new (t || WizardHeaderComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WizardComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.WizardService)); };
WizardHeaderComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WizardHeaderComponent, selectors: [["c8y-wizard-header"]], viewQuery: function WizardHeaderComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerContent = _t.first);
    } }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["headerContent", ""], [1, "modal-header", "dialog-header", "animated", "fadeIn"]], template: function WizardHeaderComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, WizardHeaderComponent_ng_template_0_Template, 2, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
WizardHeaderComponent.ctorParameters = () => [
    { type: WizardComponent },
    { type: WizardService }
];
WizardHeaderComponent.propDecorators = {
    headerContent: [{ type: ViewChild, args: ['headerContent', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WizardHeaderComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-wizard-header',
                template: `
    <ng-template #headerContent>
      <div class="modal-header dialog-header animated fadeIn">
        <ng-content></ng-content>
      </div>
    </ng-template>
  `
            }]
    }], function () { return [{ type: ɵngcc1.WizardComponent }, { type: ɵngcc2.WizardService }]; }, { headerContent: [{
            type: ViewChild,
            args: ['headerContent', { static: true }]
        }] }); })();
