// @c8y/ngx-components/esm2015/core/wizard/wizard.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { BsModalRef, ModalModule as BsModalModule } from 'ngx-bootstrap/modal';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { WizardBodyComponent } from './wizard-body.component';
import { WizardFooterComponent } from './wizard-footer.component';
import { WizardHeaderComponent } from './wizard-header.component';
import { WizardOutletComponent } from './wizard-outlet/wizard-outlet.component';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.directive';
import * as ɵngcc4 from '../i18n/c8y-translate.pipe';
export class WizardModule {
}
WizardModule.ɵfac = function WizardModule_Factory(t) { return new (t || WizardModule)(); };
WizardModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: WizardModule });
WizardModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [WizardService, BsModalRef], imports: [[BsModalModule, CommonModule, C8yCommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WizardModule, [{
        type: NgModule,
        args: [{
                imports: [BsModalModule, CommonModule, C8yCommonModule],
                exports: [
                    WizardComponent,
                    WizardOutletComponent,
                    WizardHeaderComponent,
                    WizardBodyComponent,
                    WizardFooterComponent
                ],
                declarations: [
                    WizardComponent,
                    WizardOutletComponent,
                    WizardHeaderComponent,
                    WizardBodyComponent,
                    WizardFooterComponent
                ],
                entryComponents: [
                    WizardComponent,
                    WizardHeaderComponent,
                    WizardBodyComponent,
                    WizardFooterComponent
                ],
                providers: [WizardService, BsModalRef]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(WizardModule, { declarations: function () { return [WizardComponent,
        WizardOutletComponent,
        WizardHeaderComponent,
        WizardBodyComponent,
        WizardFooterComponent]; }, imports: function () { return [BsModalModule, CommonModule, C8yCommonModule]; }, exports: function () { return [WizardComponent,
        WizardOutletComponent,
        WizardHeaderComponent,
        WizardBodyComponent,
        WizardFooterComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(WizardComponent, [ɵngcc1.NgTemplateOutlet, ɵngcc2.IconDirective, ɵngcc1.NgIf, WizardOutletComponent, ɵngcc3.C8yTranslateDirective], [ɵngcc1.AsyncPipe, ɵngcc4.C8yTranslatePipe]);
