// @c8y/ngx-components/esm2015/core/data-grid/column/cell-renderer.js
/* tslint:disable:max-classes-per-file */
import { Directive, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class CellRendererContext {
}
export class CellRendererDefDirective {
    constructor(template) {
        this.template = template;
    }
}
CellRendererDefDirective.ɵfac = function CellRendererDefDirective_Factory(t) { return new (t || CellRendererDefDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
CellRendererDefDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CellRendererDefDirective, selectors: [["", "c8yCellRendererDef", ""]] });
CellRendererDefDirective.ctorParameters = () => [
    { type: TemplateRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CellRendererDefDirective, [{
        type: Directive,
        args: [{ selector: '[c8yCellRendererDef]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, null); })();
export class HeaderCellRendererDefDirective {
    constructor(template) {
        this.template = template;
    }
}
HeaderCellRendererDefDirective.ɵfac = function HeaderCellRendererDefDirective_Factory(t) { return new (t || HeaderCellRendererDefDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
HeaderCellRendererDefDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: HeaderCellRendererDefDirective, selectors: [["", "c8yHeaderCellRendererDef", ""]] });
HeaderCellRendererDefDirective.ctorParameters = () => [
    { type: TemplateRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HeaderCellRendererDefDirective, [{
        type: Directive,
        args: [{ selector: '[c8yHeaderCellRendererDef]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, null); })();
