// @c8y/ngx-components/esm2015/core/data-grid/column/column.directive.js
import { ContentChild, Directive, Input } from '@angular/core';
import { CellRendererDefDirective, HeaderCellRendererDefDirective } from './cell-renderer';
import { FilteringFormRendererDefDirective } from './filtering-form-renderer';
import * as ɵngcc0 from '@angular/core';
export class ColumnDirective {
}
ColumnDirective.ɵfac = function ColumnDirective_Factory(t) { return new (t || ColumnDirective)(); };
ColumnDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ColumnDirective, selectors: [["c8y-column"]], contentQueries: function ColumnDirective_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, CellRendererDefDirective, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, HeaderCellRendererDefDirective, 5);
        ɵngcc0.ɵɵcontentQuery(dirIndex, FilteringFormRendererDefDirective, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.cellRendererDef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerCellRendererDef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.filteringFormRendererDef = _t.first);
    } }, inputs: { name: "name" } });
ColumnDirective.propDecorators = {
    name: [{ type: Input }],
    cellRendererDef: [{ type: ContentChild, args: [CellRendererDefDirective, { static: false },] }],
    headerCellRendererDef: [{ type: ContentChild, args: [HeaderCellRendererDefDirective, { static: false },] }],
    filteringFormRendererDef: [{ type: ContentChild, args: [FilteringFormRendererDefDirective, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ColumnDirective, [{
        type: Directive,
        args: [{ selector: 'c8y-column' }]
    }], null, { name: [{
            type: Input
        }], cellRendererDef: [{
            type: ContentChild,
            args: [CellRendererDefDirective, { static: false }]
        }], headerCellRendererDef: [{
            type: ContentChild,
            args: [HeaderCellRendererDefDirective, { static: false }]
        }], filteringFormRendererDef: [{
            type: ContentChild,
            args: [FilteringFormRendererDefDirective, { static: false }]
        }] }); })();
