// @c8y/ngx-components/esm2015/core/data-grid/column/filtering-form-renderer.js
/* tslint:disable:max-classes-per-file */
import { Directive, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class FilteringFormRendererContext {
}
export class FilteringFormRendererDefDirective {
    constructor(template) {
        this.template = template;
    }
}
FilteringFormRendererDefDirective.ɵfac = function FilteringFormRendererDefDirective_Factory(t) { return new (t || FilteringFormRendererDefDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
FilteringFormRendererDefDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: FilteringFormRendererDefDirective, selectors: [["", "c8yFilteringFormRendererDef", ""]] });
FilteringFormRendererDefDirective.ctorParameters = () => [
    { type: TemplateRef }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FilteringFormRendererDefDirective, [{
        type: Directive,
        args: [{ selector: '[c8yFilteringFormRendererDef]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, null); })();
