// @c8y/ngx-components/esm2015/device-grid-example/device-grid-example.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, CommonModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { DeviceGridExampleNavigationFactory } from './device-grid-example-navigation.factory';
import { DeviceGridExampleGuard } from './device-grid-example-guard';
import { DeviceGridExampleComponent } from './device-grid-example.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const routes = [
    {
        path: 'grids/device-grid',
        component: DeviceGridExampleComponent,
        canActivate: [DeviceGridExampleGuard]
    }
];
export class DeviceGridExampleModule {
}
DeviceGridExampleModule.ɵfac = function DeviceGridExampleModule_Factory(t) { return new (t || DeviceGridExampleModule)(); };
DeviceGridExampleModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceGridExampleModule });
DeviceGridExampleModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: DeviceGridExampleNavigationFactory,
            multi: true
        },
        DeviceGridExampleGuard
    ], imports: [[
            CoreModule,
            CommonModule,
            DeviceGridModule,
            RouterModule.forRoot(routes, { useHash: true })
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceGridExampleModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    CommonModule,
                    DeviceGridModule,
                    RouterModule.forRoot(routes, { useHash: true })
                ],
                declarations: [DeviceGridExampleComponent],
                entryComponents: [DeviceGridExampleComponent],
                exports: [],
                providers: [
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: DeviceGridExampleNavigationFactory,
                        multi: true
                    },
                    DeviceGridExampleGuard
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceGridExampleModule, { declarations: function () { return [DeviceGridExampleComponent]; }, imports: function () { return [CoreModule,
        CommonModule,
        DeviceGridModule, ɵngcc1.RouterModule]; } }); })();
