// @c8y/ngx-components/esm2015/operations/bulk-operation-from-single/bulk-operation-from-single.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { get } from 'lodash-es';
import { BaseStepperComponent } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc3 from '@c8y/ngx-components/operations/bulk-operation-stepper';
export class BulkOperationFromSingleComponent extends BaseStepperComponent {
    constructor(activatedRoute, bulkOperationService) {
        super();
        this.activatedRoute = activatedRoute;
        this.bulkOperationService = bulkOperationService;
    }
    retrieveOperationPrototype() {
        return __awaiter(this, void 0, void 0, function* () {
            const prototype = (yield this.bulkOperationService.getOperation(this.activatedRoute.snapshot.params.operationId)).data;
            return {
                description: get(prototype, 'description'),
                prototype
            };
        });
    }
}
BulkOperationFromSingleComponent.ɵfac = function BulkOperationFromSingleComponent_Factory(t) { return new (t || BulkOperationFromSingleComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BulkOperationsService)); };
BulkOperationFromSingleComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationFromSingleComponent, selectors: [["bulk-operation-from-single"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 0, template: function BulkOperationFromSingleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-bulk-operation-stepper");
    } }, directives: [ɵngcc3.BulkOperationStepper], encapsulation: 2 });
BulkOperationFromSingleComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: BulkOperationsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationFromSingleComponent, [{
        type: Component,
        args: [{
                selector: 'bulk-operation-from-single',
                template: "<c8y-bulk-operation-stepper> </c8y-bulk-operation-stepper>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.BulkOperationsService }]; }, null); })();
