// @c8y/ngx-components/esm2015/core/breadcrumb/breadcrumb.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { BreadcrumbItemComponent } from './breadcrumb-item.component';
import { BreadcrumbOutletComponent } from './breadcrumb-outlet.component';
import { BreadcrumbComponent } from './breadcrumb.component';
/**
 * The angular module definition for breadcrumbs.
 * @exports BreadcrumbComponent
 * @exports BreadcrumbItemComponent
 * @exports BreadcrumbOutletComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from 'ngx-bootstrap/tooltip';
export class BreadcrumbModule {
}
BreadcrumbModule.ɵfac = function BreadcrumbModule_Factory(t) { return new (t || BreadcrumbModule)(); };
BreadcrumbModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BreadcrumbModule });
BreadcrumbModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            RouterModule.forChild([]),
            CommonModule,
            TooltipModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BreadcrumbModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    BreadcrumbComponent,
                    BreadcrumbItemComponent,
                    BreadcrumbOutletComponent
                ],
                imports: [
                    RouterModule.forChild([]),
                    CommonModule,
                    TooltipModule.forRoot()
                ],
                exports: [
                    BreadcrumbComponent,
                    BreadcrumbItemComponent,
                    BreadcrumbOutletComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BreadcrumbModule, { declarations: function () { return [BreadcrumbComponent,
        BreadcrumbItemComponent,
        BreadcrumbOutletComponent]; }, imports: function () { return [ɵngcc1.RouterModule, CommonModule, ɵngcc2.TooltipModule]; }, exports: function () { return [BreadcrumbComponent,
        BreadcrumbItemComponent,
        BreadcrumbOutletComponent]; } }); })();
