// @c8y/ngx-components/esm2015/core/breadcrumb/breadcrumb-outlet.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from 'ngx-bootstrap/tooltip';
import * as ɵngcc4 from '../common/outlet.directive';
import * as ɵngcc5 from '@angular/router';
import * as ɵngcc6 from '../i18n/c8y-translate.pipe';

function BreadcrumbOutletComponent_div_0_button_1_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 6);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("tooltip", ɵngcc0.ɵɵpipeBind1(1, 2, "Show all breadcrumbs"));
    ɵngcc0.ɵɵproperty("c8yIcon", "caret-right");
} }
function BreadcrumbOutletComponent_div_0_button_1_i_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 6);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("tooltip", ɵngcc0.ɵɵpipeBind1(1, 2, "Collapse breadcrumbs"));
    ɵngcc0.ɵɵproperty("c8yIcon", "caret-down");
} }
function BreadcrumbOutletComponent_div_0_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 4);
    ɵngcc0.ɵɵlistener("click", function BreadcrumbOutletComponent_div_0_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(2); return ctx_r5.showAll = !ctx_r5.showAll; });
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_button_1_i_1_Template, 2, 4, "i", 5);
    ɵngcc0.ɵɵtemplate(2, BreadcrumbOutletComponent_div_0_button_1_i_2_Template, 2, 4, "i", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.showAll);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.showAll);
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1, " > ");
    ɵngcc0.ɵɵelementEnd();
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_3_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    const item_r11 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("c8yIcon", item_r11.icon);
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 12);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtemplate(3, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_3_i_3_Template, 1, 1, "i", 13);
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r19 = ɵngcc0.ɵɵnextContext();
    const item_r11 = ctx_r19.$implicit;
    const firstItem_r12 = ctx_r19.first;
    const ctx_r15 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 4, item_r11.label));
    ɵngcc0.ɵɵproperty("routerLink", ctx_r15.normalizePath(item_r11.path));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", firstItem_r12);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 6, item_r11.label));
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_4_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 14);
} if (rf & 2) {
    const item_r11 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("c8yIcon", item_r11.icon);
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_4_i_1_Template, 1, 1, "i", 13);
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r22 = ɵngcc0.ɵɵnextContext();
    const firstItem_r12 = ctx_r22.first;
    const item_r11 = ctx_r22.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", firstItem_r12);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 2, item_r11.label));
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 10);
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_span_1_Template, 2, 0, "span", 8);
    ɵngcc0.ɵɵtemplate(2, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_2_Template, 1, 0, "ng-container", 11);
    ɵngcc0.ɵɵtemplate(3, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_3_Template, 7, 8, "ng-container", 8);
    ɵngcc0.ɵɵtemplate(4, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_ng_container_4_Template, 5, 4, "ng-container", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r11 = ctx.$implicit;
    const firstItem_r12 = ctx.first;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !firstItem_r12);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r11.template);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r11.label && item_r11.path);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r11.label && !item_r11.path);
} }
function BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_li_1_Template, 5, 4, "li", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const breadcrumb_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", breadcrumb_r7.items);
} }
function BreadcrumbOutletComponent_div_0_ul_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 7);
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_ul_2_ng_container_1_Template, 2, 1, "ng-container", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const first_r8 = ctx.first;
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", first_r8 || ctx_r2.showAll);
} }
const _c0 = function (a0, a1) { return { multiple: a0, open: a1 }; };
function BreadcrumbOutletComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵtemplate(1, BreadcrumbOutletComponent_div_0_button_1_Template, 3, 2, "button", 2);
    ɵngcc0.ɵɵtemplate(2, BreadcrumbOutletComponent_div_0_ul_2_Template, 2, 1, "ul", 3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c0, ctx_r0.breadcrumbs.length > 1, ctx_r0.showAll));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.breadcrumbs.length > 1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.breadcrumbs);
} }
export class BreadcrumbOutletComponent {
    constructor() {
        this.showAll = false;
        this.breadcrumbs = [];
    }
    /**
     * For upgrade only. Old angularjs routes start with hash, new ones not.
     */
    normalizePath(path) {
        return path === null || path === void 0 ? void 0 : path.replace(/^#\/?/, '');
    }
}
BreadcrumbOutletComponent.ɵfac = function BreadcrumbOutletComponent_Factory(t) { return new (t || BreadcrumbOutletComponent)(); };
BreadcrumbOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BreadcrumbOutletComponent, selectors: [["c8y-breadcrumb-outlet"]], inputs: { breadcrumbs: "breadcrumbs" }, decls: 1, vars: 1, consts: [["class", "breadcrumbs-container", 3, "ngClass", 4, "ngIf"], [1, "breadcrumbs-container", 3, "ngClass"], ["class", "btn-show-all-breadcrumbs pull-left", 3, "click", 4, "ngIf"], ["class", "breadcrumbs text-muted", 4, "ngFor", "ngForOf"], [1, "btn-show-all-breadcrumbs", "pull-left", 3, "click"], ["container", "body", "placement", "top", 3, "c8yIcon", "tooltip", 4, "ngIf"], ["container", "body", "placement", "top", 3, "c8yIcon", "tooltip"], [1, "breadcrumbs", "text-muted"], [4, "ngIf"], ["class", "text-muted", 4, "ngFor", "ngForOf"], [1, "text-muted"], [4, "c8yOutlet"], [1, "word-break", 3, "routerLink", "title"], ["class", "m-r-4", 3, "c8yIcon", 4, "ngIf"], [1, "m-r-4", 3, "c8yIcon"]], template: function BreadcrumbOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, BreadcrumbOutletComponent_div_0_Template, 3, 6, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.breadcrumbs && ctx.breadcrumbs.length > 0);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc2.IconDirective, ɵngcc3.TooltipDirective, ɵngcc4.OutletDirective, ɵngcc5.RouterLinkWithHref], pipes: [ɵngcc6.C8yTranslatePipe], encapsulation: 2 });
BreadcrumbOutletComponent.propDecorators = {
    breadcrumbs: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BreadcrumbOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-breadcrumb-outlet',
                template: "<div\n  class=\"breadcrumbs-container\"\n  [ngClass]=\"{\n    multiple: breadcrumbs.length > 1,\n    open: showAll\n  }\"\n  *ngIf=\"breadcrumbs && breadcrumbs.length > 0\"\n>\n  <button\n    *ngIf=\"breadcrumbs.length > 1\"\n    class=\"btn-show-all-breadcrumbs pull-left\"\n    (click)=\"showAll = !showAll\"\n  >\n    <i\n      *ngIf=\"!showAll\"\n      [c8yIcon]=\"'caret-right'\"\n      tooltip=\"{{ 'Show all breadcrumbs' | translate }}\"\n      container=\"body\"\n      placement=\"top\"\n    ></i>\n    <i\n      *ngIf=\"showAll\"\n      [c8yIcon]=\"'caret-down'\"\n      tooltip=\"{{ 'Collapse breadcrumbs' | translate }}\"\n      container=\"body\"\n      placement=\"top\"\n    ></i>\n  </button>\n\n  <ul class=\"breadcrumbs text-muted\" *ngFor=\"let breadcrumb of breadcrumbs; let first = first\">\n    <ng-container *ngIf=\"first || showAll\">\n      <li class=\"text-muted\" *ngFor=\"let item of breadcrumb.items; let firstItem = first\">\n        <span *ngIf=\"!firstItem\"> > </span>\n        <ng-container *c8yOutlet=\"item.template\"></ng-container>\n        <ng-container *ngIf=\"item.label && item.path\">\n          <a\n            [routerLink]=\"normalizePath(item.path)\"\n            class=\"word-break\"\n            title=\"{{ item.label | translate }}\"\n          >\n            <i [c8yIcon]=\"item.icon\" *ngIf=\"firstItem\" class=\"m-r-4\"></i>\n            <span>{{ item.label | translate }}</span>\n          </a>\n        </ng-container>\n        <ng-container *ngIf=\"item.label && !item.path\">\n          <i [c8yIcon]=\"item.icon\" *ngIf=\"firstItem\" class=\"m-r-4\"></i>\n          <span>{{ item.label | translate }}</span>\n        </ng-container>\n      </li>\n    </ng-container>\n  </ul>\n</div>\n"
            }]
    }], function () { return []; }, { breadcrumbs: [{
            type: Input
        }] }); })();
