// @c8y/ngx-components/esm2015/core/action-bar/action-bar-item.component.js
import { Component, Input, ViewChild } from '@angular/core';
import { ActionBarService } from './action-bar.service';
/**
 * A <c8y-action-bar-item> can be added with content projection.
 * It allows to show items in three positions (left, right
 * or more).
 *
 * @example
 * ```html
 * <c8y-action-bar-item [placement]="'right'">
 *   <button class="btn btn-link" (click)="addRandom()"><i c8yIcon="plus-circle"></i> Add random</button>
 * </c8y-action-bar-item>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './action-bar.service';

const _c0 = ["template"];
function ActionBarItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(ctx_r1.itemClass);
} }
const _c1 = ["*"];
export class ActionBarItemComponent {
    /**
     * @ignore
     */
    constructor(actionBarService) {
        this.actionBarService = actionBarService;
        /**
         * Where the item should be shown. Use "more" to place
         * it in a dropdown on the right.
         */
        this.placement = 'left';
        /**
         * Priority for sorting the list. Higher priority means
         * an earlier position in the list.
         */
        this.priority = 0;
        /**
         * Adds additional classes to the item.
         */
        this.itemClass = '';
    }
    /**
     * @ignore
     */
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.actionBarService.add(this));
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.actionBarService.remove(this);
    }
}
ActionBarItemComponent.ɵfac = function ActionBarItemComponent_Factory(t) { return new (t || ActionBarItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActionBarService)); };
ActionBarItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActionBarItemComponent, selectors: [["c8y-action-bar-item"]], viewQuery: function ActionBarItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { placement: "placement", priority: "priority", itemClass: "itemClass" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""]], template: function ActionBarItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ActionBarItemComponent_ng_template_0_Template, 2, 3, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ActionBarItemComponent.ctorParameters = () => [
    { type: ActionBarService }
];
ActionBarItemComponent.propDecorators = {
    placement: [{ type: Input }],
    priority: [{ type: Input }],
    itemClass: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionBarItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-action-bar-item',
                template: "<ng-template #template>\n  <li class=\"{{itemClass}}\">\n    <ng-content></ng-content>\n  </li>\n</ng-template>"
            }]
    }], function () { return [{ type: ɵngcc1.ActionBarService }]; }, { placement: [{
            type: Input
        }], priority: [{
            type: Input
        }], itemClass: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }] }); })();
