// @c8y/ngx-components/esm2015/core/action-bar/action-bar.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ActionBarItemComponent } from './action-bar-item.component';
import { ActionBarComponent } from './action-bar.component';
/**
 * An action bar can be added with content projection or multi provider
 * extension. It allows to show them in three positions (left, right
 * or more). The actions placed here are always local actions to the
 * current route like "add device" or "lock dashboard".
 *
 * @example
 * ```html
 * <c8y-action-bar-item [placement]="'right'">
 *   <button class="btn btn-link" (click)="addRandom()"><i c8yIcon="plus-circle"></i> Add random</button>
 * </c8y-action-bar-item>
 * ```
 * @exports ActionBarComponent The outlet component
 * @exports ActionBarItemComponent The items to add with cp.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class ActionBarModule {
}
ActionBarModule.ɵfac = function ActionBarModule_Factory(t) { return new (t || ActionBarModule)(); };
ActionBarModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ActionBarModule });
ActionBarModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[BsDropdownModule, CommonModule, CollapseModule, RouterModule.forChild([])]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionBarModule, [{
        type: NgModule,
        args: [{
                declarations: [ActionBarComponent, ActionBarItemComponent],
                imports: [BsDropdownModule, CommonModule, CollapseModule, RouterModule.forChild([])],
                exports: [ActionBarComponent, ActionBarItemComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ActionBarModule, { declarations: function () { return [ActionBarComponent, ActionBarItemComponent]; }, imports: function () { return [BsDropdownModule, CommonModule, CollapseModule, ɵngcc1.RouterModule]; }, exports: function () { return [ActionBarComponent, ActionBarItemComponent]; } }); })();
