// @c8y/ngx-components/esm2015/core/properties-list/properties-list.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { PropertiesListComponent } from './properties-list.component';
import * as ɵngcc0 from '@angular/core';
export class PropertiesListModule {
}
PropertiesListModule.ɵfac = function PropertiesListModule_Factory(t) { return new (t || PropertiesListModule)(); };
PropertiesListModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: PropertiesListModule });
PropertiesListModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [], imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PropertiesListModule, [{
        type: NgModule,
        args: [{
                declarations: [PropertiesListComponent],
                imports: [CommonModule],
                exports: [PropertiesListComponent],
                providers: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(PropertiesListModule, { declarations: function () { return [PropertiesListComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [PropertiesListComponent]; } }); })();
