// @c8y/ngx-components/esm2015/device-grid/columns/alarms.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import { ColumnUtilService } from './column-util.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './column-util.service';
import * as ɵngcc3 from '@angular/common';

function AlarmsCellRendererComponent_small_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 5);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "i", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r0.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx_r0.context.item.c8y_ActiveAlarmsStatus.critical, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "warning");
} }
function AlarmsCellRendererComponent_small_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 7);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "i", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx_r1.context.item.c8y_ActiveAlarmsStatus.major, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "exclamation-circle");
} }
function AlarmsCellRendererComponent_small_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "i", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r2.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx_r2.context.item.c8y_ActiveAlarmsStatus.minor, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "exclamation-circle");
} }
function AlarmsCellRendererComponent_small_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 9);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelement(2, "i", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r3.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx_r3.context.item.c8y_ActiveAlarmsStatus.warning, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "circle");
} }
export class AlarmsCellRendererComponent {
    constructor(context, columnUtilService) {
        this.context = context;
        this.columnUtilService = columnUtilService;
    }
}
AlarmsCellRendererComponent.ɵfac = function AlarmsCellRendererComponent_Factory(t) { return new (t || AlarmsCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ColumnUtilService)); };
AlarmsCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AlarmsCellRendererComponent, selectors: [["ng-component"]], decls: 5, vars: 5, consts: [[1, "listLink", 3, "href"], ["class", "status critical text-nowrap", 4, "ngIf"], ["class", "status major text-nowrap", 4, "ngIf"], ["class", "status minor text-nowrap", 4, "ngIf"], ["class", "status warning text-nowrap", 4, "ngIf"], [1, "status", "critical", "text-nowrap"], [3, "c8yIcon"], [1, "status", "major", "text-nowrap"], [1, "status", "minor", "text-nowrap"], [1, "status", "warning", "text-nowrap"]], template: function AlarmsCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "a", 0);
        ɵngcc0.ɵɵtemplate(1, AlarmsCellRendererComponent_small_1_Template, 3, 2, "small", 1);
        ɵngcc0.ɵɵtemplate(2, AlarmsCellRendererComponent_small_2_Template, 3, 2, "small", 2);
        ɵngcc0.ɵɵtemplate(3, AlarmsCellRendererComponent_small_3_Template, 3, 2, "small", 3);
        ɵngcc0.ɵɵtemplate(4, AlarmsCellRendererComponent_small_4_Template, 3, 2, "small", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("href", ctx.columnUtilService.getAlarmsHref(ctx.context.item), ɵngcc0.ɵɵsanitizeUrl);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx.context.item.c8y_ActiveAlarmsStatus.critical);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx.context.item.c8y_ActiveAlarmsStatus.major);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx.context.item.c8y_ActiveAlarmsStatus.minor);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.context.item.c8y_ActiveAlarmsStatus == null ? null : ctx.context.item.c8y_ActiveAlarmsStatus.warning);
    } }, directives: [ɵngcc3.NgIf, ɵngcc1.IconDirective], encapsulation: 2 });
AlarmsCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext },
    { type: ColumnUtilService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlarmsCellRendererComponent, [{
        type: Component,
        args: [{
                template: "<a class=\"listLink\" [href]=\"columnUtilService.getAlarmsHref(context.item)\">\n  <small class=\"status critical text-nowrap\" *ngIf=\"context.item.c8y_ActiveAlarmsStatus?.critical\">\n    {{ context.item.c8y_ActiveAlarmsStatus?.critical }}\n    <i [c8yIcon]=\"'warning'\"></i>\n  </small>\n  <small class=\"status major text-nowrap\" *ngIf=\"context.item.c8y_ActiveAlarmsStatus?.major\">\n    {{ context.item.c8y_ActiveAlarmsStatus?.major }}\n    <i [c8yIcon]=\"'exclamation-circle'\"></i>\n  </small>\n  <small class=\"status minor text-nowrap\" *ngIf=\"context.item.c8y_ActiveAlarmsStatus?.minor\">\n    {{ context.item.c8y_ActiveAlarmsStatus?.minor }}\n    <i [c8yIcon]=\"'exclamation-circle'\"></i>\n  </small>\n  <small class=\"status warning text-nowrap\" *ngIf=\"context.item.c8y_ActiveAlarmsStatus?.warning\">\n    {{ context.item.c8y_ActiveAlarmsStatus?.warning }}\n    <i [c8yIcon]=\"'circle'\"></i>\n  </small>\n</a>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }, { type: ɵngcc2.ColumnUtilService }]; }, null); })();
