// @c8y/ngx-components/esm2015/device-grid/columns/registration-date.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class RegistrationDateCellRendererComponent {
    constructor(context) {
        this.context = context;
    }
}
RegistrationDateCellRendererComponent.ɵfac = function RegistrationDateCellRendererComponent_Factory(t) { return new (t || RegistrationDateCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext)); };
RegistrationDateCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RegistrationDateCellRendererComponent, selectors: [["ng-component"]], decls: 2, vars: 3, template: function RegistrationDateCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtext(0);
        ɵngcc0.ɵɵpipe(1, "c8yDate");
    } if (rf & 2) {
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(1, 1, ctx.context.value), " ");
    } }, pipes: [ɵngcc1.DatePipe], encapsulation: 2 });
RegistrationDateCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RegistrationDateCellRendererComponent, [{
        type: Component,
        args: [{
                template: `
    {{ context.value | c8yDate }}
  `
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }]; }, null); })();
