// @c8y/ngx-components/esm2015/core/data-grid/data-grid.component.js
import { CdkHeaderCell } from '@angular/cdk/table';
import { ChangeDetectorRef, Component, ComponentFactoryResolver, ContentChildren, ElementRef, EventEmitter, Input, Output, ViewChild, ViewChildren, ViewContainerRef } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { flow, get, parseInt as _parseInt, union, without, uniqBy } from 'lodash-es';
import { BehaviorSubject, combineLatest, Subject } from 'rxjs';
import { debounceTime, delay, distinctUntilChanged, first, map, mergeMap, take, takeUntil, tap } from 'rxjs/operators';
import { LoadMoreComponent } from '../common/load-more.component';
import { gettext } from '../i18n/gettext';
import { ColumnDirective } from './column/column.directive';
import { FilteringActionType, minColumnGridTrackSize, ratiosByColumnTypes } from './data-grid.model';
import { GridDataSource } from './grid-data-source';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@angular/cdk/table';
import * as ɵngcc4 from '../common/progress-bar.component';
import * as ɵngcc5 from 'ngx-bootstrap/dropdown';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from 'ngx-bootstrap/popover';
import * as ɵngcc8 from '../i18n/c8y-translate.directive';
import * as ɵngcc9 from '@angular/cdk/drag-drop';
import * as ɵngcc10 from '@angular/forms';
import * as ɵngcc11 from 'ngx-bootstrap/tooltip';
import * as ɵngcc12 from './column/cell-renderer.component';
import * as ɵngcc13 from './column/filtering-form-renderer.component';
import * as ɵngcc14 from 'ngx-bootstrap/pagination';
import * as ɵngcc15 from '../i18n/c8y-translate.pipe';
import * as ɵngcc16 from './is-control-visible.pipe';
import * as ɵngcc17 from '../common/map-function.pipe';

const _c0 = ["scroll"];
const _c1 = ["infiniteScrollContainer"];
function DataGridComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelement(2, "c8y-progress-bar", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("message", ctx_r1.loadingItemsLabel);
} }
function DataGridComponent_div_5_ng_container_4_span_1_small_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 35);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, "{{ filteredItemsCount }} of {{ allItemsCount }} items");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r18 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵproperty("translateParams", ctx_r18.filteringLabelsParams);
} }
function DataGridComponent_div_5_ng_container_4_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 33);
    ɵngcc0.ɵɵtext(2, "No filters");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_div_5_ng_container_4_span_1_small_3_Template, 2, 1, "small", 34);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !!ctx_r15.filteringLabelsParams.allItemsCount);
} }
function DataGridComponent_div_5_ng_container_4_span_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "span", 37);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "small", 38);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(5, "of {{ allItemsCount }} items");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r19 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 2, ctx_r19.dataSource.stats$).filteredSize, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ctx_r19.filteringLabelsParams);
} }
function DataGridComponent_div_5_ng_container_4_span_2_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_ng_container_4_span_2_ng_container_1_Template, 6, 4, "ng-container", 10);
    ɵngcc0.ɵɵelementStart(2, "button", 36);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_ng_container_4_span_2_Template_button_click_2_listener() { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.clearFilters(); });
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !!ctx_r16.filteringLabelsParams.allItemsCount);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(3, 3, "Clear filters"), "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 5, "Clear filters"), " ");
} }
function DataGridComponent_div_5_ng_container_4_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 29);
    ɵngcc0.ɵɵelementStart(1, "button", 30);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 39);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(2, 1, "The counter for the total number of items might be inaccurate."));
} }
function DataGridComponent_div_5_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_ng_container_4_span_1_Template, 4, 1, "span", 10);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_div_5_ng_container_4_span_2_Template, 6, 7, "span", 10);
    ɵngcc0.ɵɵelementStart(3, "span", 29);
    ɵngcc0.ɵɵelementStart(4, "button", 30);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, DataGridComponent_div_5_ng_container_4_span_7_Template, 4, 3, "span", 32);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r10.filteringApplied);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r10.filteringApplied);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(5, 4, "Click the column headers to apply filters."));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r10.showCounterWarning);
} }
function DataGridComponent_div_5_ng_container_7_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 42);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_ng_container_7_ng_container_1_ng_container_1_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r29); const headerActionControl_r22 = ɵngcc0.ɵɵnextContext(2).$implicit; return headerActionControl_r22.callback(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 43);
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const headerActionControl_r22 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, headerActionControl_r22.text));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 4, headerActionControl_r22.text));
} }
function DataGridComponent_div_5_ng_container_7_ng_container_1_ng_template_2_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c2 = function (a0) { return { headerActionControl: a0 }; };
function DataGridComponent_div_5_ng_container_7_ng_container_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DataGridComponent_div_5_ng_container_7_ng_container_1_ng_template_2_ng_container_0_Template, 1, 0, "ng-container", 44);
} if (rf & 2) {
    const headerActionControl_r22 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", headerActionControl_r22.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c2, headerActionControl_r22));
} }
function DataGridComponent_div_5_ng_container_7_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_ng_container_7_ng_container_1_ng_container_1_Template, 7, 6, "ng-container", 40);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_div_5_ng_container_7_ng_container_1_ng_template_2_Template, 1, 4, "ng-template", null, 41, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r25 = ɵngcc0.ɵɵreference(3);
    const headerActionControl_r22 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !headerActionControl_r22.template)("ngIfElse", _r25);
} }
function DataGridComponent_div_5_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_ng_container_7_ng_container_1_Template, 4, 2, "ng-container", 10);
    ɵngcc0.ɵɵpipe(2, "isControlVisible");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const headerActionControl_r22 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 1, headerActionControl_r22));
} }
function DataGridComponent_div_5_ul_15_div_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 50);
    ɵngcc0.ɵɵelementStart(2, "label", 51);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementStart(4, "input", 52);
    ɵngcc0.ɵɵlistener("ngModelChange", function DataGridComponent_div_5_ul_15_div_3_ng_container_1_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r39); const column_r35 = ɵngcc0.ɵɵnextContext().$implicit; return column_r35.visible = $event; })("change", function DataGridComponent_div_5_ul_15_div_3_ng_container_1_Template_input_change_4_listener() { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r40 = ɵngcc0.ɵɵnextContext(4); return ctx_r40.updateGridColumnsSize(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "span");
    ɵngcc0.ɵɵelementStart(6, "span");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r35 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 3, column_r35.header) || column_r35.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", column_r35.visible);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 5, column_r35.header) || column_r35.name);
} }
function DataGridComponent_div_5_ul_15_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 49);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_ul_15_div_3_ng_container_1_Template, 9, 7, "ng-container", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r35 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !column_r35.positionFixed);
} }
function DataGridComponent_div_5_ul_15_Template(rf, ctx) { if (rf & 1) {
    const _r44 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "ul", 45);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_ul_15_Template_ul_click_0_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelementStart(1, "li", 46);
    ɵngcc0.ɵɵelementStart(2, "div", 47);
    ɵngcc0.ɵɵlistener("cdkDropListDropped", function DataGridComponent_div_5_ul_15_Template_div_cdkDropListDropped_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r44); const ctx_r43 = ɵngcc0.ɵɵnextContext(2); return ctx_r43.onColumnDrop($event); });
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_div_5_ul_15_div_3_Template, 2, 1, "div", 48);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r12.columns);
} }
function DataGridComponent_div_5_div_23_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 58);
} }
function DataGridComponent_div_5_div_23_i_5_Template(rf, ctx) { if (rf & 1) {
    const _r48 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 59);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_23_i_5_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r48); const ctx_r47 = ɵngcc0.ɵɵnextContext(3); ctx_r47.searchText = ""; return ctx_r47.searchText$.emit(""); });
    ɵngcc0.ɵɵelementEnd();
} }
function DataGridComponent_div_5_div_23_Template(rf, ctx) { if (rf & 1) {
    const _r50 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 53);
    ɵngcc0.ɵɵelementStart(1, "input", 54);
    ɵngcc0.ɵɵlistener("input", function DataGridComponent_div_5_div_23_Template_input_input_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r50); const ctx_r49 = ɵngcc0.ɵɵnextContext(2); return ctx_r49.searchText$.emit($event.target.value); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 55);
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_div_5_div_23_i_4_Template, 1, 0, "i", 56);
    ɵngcc0.ɵɵtemplate(5, DataGridComponent_div_5_div_23_i_5_Template, 1, 0, "i", 57);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(2, 4, "Search\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r13.searchText);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r13.searchText.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r13.searchText.length > 0);
} }
const _c3 = function (a0) { return { count: a0 }; };
function DataGridComponent_div_5_div_24_small_5_Template(rf, ctx) { if (rf & 1) {
    const _r54 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵelementStart(1, "a", 65);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_24_small_5_Template_a_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r54); const ctx_r53 = ɵngcc0.ɵɵnextContext(3); return ctx_r53.setAllItemsSelected(true); });
    ɵngcc0.ɵɵelementStart(2, "span", 38);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵtext(4, " Select all {{ count }} items ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r51 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(3, _c3, ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r51.dataSource.stats$).filteredSize));
} }
function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r61 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 63);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r61); const bulkActionControl_r55 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r60 = ɵngcc0.ɵɵnextContext(3); return bulkActionControl_r55.callback(ctx_r60.selectedItemIds, ctx_r60.reload.bind(ctx_r60)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 69);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Export"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 4, "Export"));
} }
function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r64 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 63);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r64); const bulkActionControl_r55 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r63 = ɵngcc0.ɵɵnextContext(3); return bulkActionControl_r55.callback(ctx_r63.selectedItemIds, ctx_r63.reload.bind(ctx_r63)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 70);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Delete"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 4, "Delete"));
} }
function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r67 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 63);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r67); const bulkActionControl_r55 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r66 = ɵngcc0.ɵɵnextContext(3); return bulkActionControl_r55.callback(ctx_r66.selectedItemIds, ctx_r66.reload.bind(ctx_r66)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 71);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const bulkActionControl_r55 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, bulkActionControl_r55.text));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", bulkActionControl_r55.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 5, bulkActionControl_r55.text));
} }
function DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_1_Template, 6, 6, "button", 67);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_2_Template, 6, 6, "button", 67);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_button_3_Template, 6, 7, "button", 68);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r56 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r56.builtInActionType.Export);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r56.builtInActionType.Delete);
} }
function DataGridComponent_div_5_div_24_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 66);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_div_5_div_24_ng_container_8_ng_container_1_Template, 4, 2, "ng-container", 10);
    ɵngcc0.ɵɵpipe(2, "isControlVisible");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const bulkActionControl_r55 = ctx.$implicit;
    const ctx_r52 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngSwitch", bulkActionControl_r55.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind2(2, 2, bulkActionControl_r55, ctx_r52.selectedItemIds));
} }
function DataGridComponent_div_5_div_24_Template(rf, ctx) { if (rf & 1) {
    const _r71 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 60);
    ɵngcc0.ɵɵelementStart(1, "h4");
    ɵngcc0.ɵɵelementStart(2, "span", 38);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ count }} selected items. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "br", 61);
    ɵngcc0.ɵɵtemplate(5, DataGridComponent_div_5_div_24_small_5_Template, 5, 5, "small", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 18);
    ɵngcc0.ɵɵelementStart(7, "div", 19);
    ɵngcc0.ɵɵtemplate(8, DataGridComponent_div_5_div_24_ng_container_8_Template, 3, 5, "ng-container", 62);
    ɵngcc0.ɵɵelementStart(9, "button", 63);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_div_24_Template_button_click_9_listener() { ɵngcc0.ɵɵrestoreView(_r71); const ctx_r70 = ɵngcc0.ɵɵnextContext(2); return ctx_r70.cancel(); });
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelement(11, "i", 64);
    ɵngcc0.ɵɵelementStart(12, "span");
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(9, _c3, ctx_r14.selectedItemIds.length));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r14.selectedItemIds.length >= ctx_r14.pagination.pageSize);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r14.bulkActionControls);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 5, "Cancel"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(14, 7, "Cancel"));
} }
const _c4 = function (a0) { return { "m-r-16": a0 }; };
function DataGridComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    const _r73 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "h4", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_div_5_ng_container_4_Template, 8, 6, "ng-container", 10);
    ɵngcc0.ɵɵelementStart(5, "div", 18);
    ɵngcc0.ɵɵelementStart(6, "div", 19);
    ɵngcc0.ɵɵtemplate(7, DataGridComponent_div_5_ng_container_7_Template, 3, 3, "ng-container", 20);
    ɵngcc0.ɵɵelementStart(8, "div", 21);
    ɵngcc0.ɵɵelementStart(9, "button", 22);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelement(11, "i", 23);
    ɵngcc0.ɵɵelementStart(12, "span");
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(15, DataGridComponent_div_5_ul_15_Template, 4, 1, "ul", 24);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "button", 25);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_div_5_Template_button_click_16_listener() { ɵngcc0.ɵɵrestoreView(_r73); const ctx_r72 = ɵngcc0.ɵɵnextContext(); return ctx_r72.clickReload(); });
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵpipe(18, "async");
    ɵngcc0.ɵɵelement(19, "i", 26);
    ɵngcc0.ɵɵelementStart(20, "span");
    ɵngcc0.ɵɵtext(21);
    ɵngcc0.ɵɵpipe(22, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(23, DataGridComponent_div_5_div_23_Template, 6, 6, "div", 27);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(24, DataGridComponent_div_5_div_24_Template, 15, 11, "div", 28);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(24, _c4, !!ctx_r2.title));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 12, ctx_r2.title));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.displayOptions.filter);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.headerActionControls);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("insideClick", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 14, "Configure columns"));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(14, 16, "Configure columns"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(17, 18, "Reload"));
    ɵngcc0.ɵɵproperty("disabled", ɵngcc0.ɵɵpipeBind1(18, 20, ctx_r2.dataSource.loading$));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(22, 22, "Reload"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.serverSideDataCallback || ctx_r2.showSearch);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.selectedItemIds.length !== 0);
} }
function DataGridComponent_ng_container_7_ng_container_2_th_1_Template(rf, ctx) { if (rf & 1) {
    const _r81 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "th", 77);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementStart(2, "label", 78);
    ɵngcc0.ɵɵelementStart(3, "input", 79);
    ɵngcc0.ɵɵlistener("change", function DataGridComponent_ng_container_7_ng_container_2_th_1_Template_input_change_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r81); const ctx_r80 = ɵngcc0.ɵɵnextContext(3); return ctx_r80.setAllItemsInCurrentPageSelected($event.target.checked); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r78 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("checked", ctx_r78.currentPageSelectionState.allSelected)("indeterminate", !(ctx_r78.currentPageSelectionState.allSelected || ctx_r78.currentPageSelectionState.allDeselected));
} }
function DataGridComponent_ng_container_7_ng_container_2_td_2_Template(rf, ctx) { if (rf & 1) {
    const _r84 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "td", 80);
    ɵngcc0.ɵɵelementStart(1, "label", 78);
    ɵngcc0.ɵɵelementStart(2, "input", 81);
    ɵngcc0.ɵɵlistener("change", function DataGridComponent_ng_container_7_ng_container_2_td_2_Template_input_change_2_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r84); const row_r82 = restoredCtx.$implicit; const ctx_r83 = ɵngcc0.ɵɵnextContext(3); return ctx_r83.setItemsSelected([row_r82], $event.target.checked); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const row_r82 = ctx.$implicit;
    const ctx_r79 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("checked", ctx_r79.isItemSelected(row_r82));
} }
function DataGridComponent_ng_container_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_2_th_1_Template, 5, 2, "th", 75);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_2_td_2_Template, 4, 1, "td", 76);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function DataGridComponent_ng_container_7_ng_container_3_th_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "th", 77);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r98 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 86);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r98); const actionControl_r91 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r96 = ɵngcc0.ɵɵnextContext(3); return actionControl_r91.callback(row_r87, ctx_r96.reload.bind(ctx_r96)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 87);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Edit"));
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r102 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 88);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r102); const actionControl_r91 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r100 = ɵngcc0.ɵɵnextContext(3); return actionControl_r91.callback(row_r87, ctx_r100.reload.bind(ctx_r100)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 89);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Delete"));
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r106 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 90);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r106); const actionControl_r91 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r104 = ɵngcc0.ɵɵnextContext(3); return actionControl_r91.callback(row_r87, ctx_r104.reload.bind(ctx_r104)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 71);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const actionControl_r91 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, actionControl_r91.text));
    ɵngcc0.ɵɵpropertyInterpolate("tooltip", ɵngcc0.ɵɵpipeBind1(2, 5, actionControl_r91.text));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", actionControl_r91.icon);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_1_Template, 3, 3, "button", 83);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_2_Template, 3, 3, "button", 84);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_button_3_Template, 4, 7, "button", 85);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r92 = ɵngcc0.ɵɵnextContext(6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r92.builtInActionType.Edit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r92.builtInActionType.Delete);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 66);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_ng_container_1_Template, 4, 2, "ng-container", 10);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const actionControl_r91 = ctx.$implicit;
    const row_r87 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("ngSwitch", actionControl_r91.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", actionControl_r91.showIf ? actionControl_r91.showIf(row_r87) : true);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_ng_container_1_Template, 2, 2, "ng-container", 62);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r88 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r88.actionControls);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r120 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 98);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r120); const actionControl_r112 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r118 = ɵngcc0.ɵɵnextContext(3); return actionControl_r112.callback(row_r87, ctx_r118.reload.bind(ctx_r118)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 87);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Edit"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Edit"), " ");
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r124 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 98);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r124); const actionControl_r112 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r122 = ɵngcc0.ɵɵnextContext(3); return actionControl_r112.callback(row_r87, ctx_r122.reload.bind(ctx_r122)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 70);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Delete"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Delete"), " ");
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_3_Template(rf, ctx) { if (rf & 1) {
    const _r128 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 98);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_3_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r128); const actionControl_r112 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r126 = ɵngcc0.ɵɵnextContext(3); return actionControl_r112.callback(row_r87, ctx_r126.reload.bind(ctx_r126)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 99);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Export"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Export"), " ");
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r132 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 98);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_4_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r132); const actionControl_r112 = ɵngcc0.ɵɵnextContext(2).$implicit; const row_r87 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r130 = ɵngcc0.ɵɵnextContext(3); return actionControl_r112.callback(row_r87, ctx_r130.reload.bind(ctx_r130)); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 71);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const actionControl_r112 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, actionControl_r112.text));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", actionControl_r112.icon);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 5, actionControl_r112.text), " ");
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_1_Template, 5, 6, "button", 96);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_2_Template, 5, 6, "button", 96);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_3_Template, 5, 6, "button", 96);
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_button_4_Template, 5, 7, "button", 97);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r113 = ɵngcc0.ɵɵnextContext(7);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r113.builtInActionType.Edit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r113.builtInActionType.Delete);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", ctx_r113.builtInActionType.Export);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 66);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_ng_container_1_Template, 5, 3, "ng-container", 10);
    ɵngcc0.ɵɵpipe(2, "isControlVisible");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const actionControl_r112 = ctx.$implicit;
    const row_r87 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngSwitch", actionControl_r112.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind2(2, 2, actionControl_r112, row_r87));
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 95);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_li_1_Template, 3, 5, "li", 62);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r110 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r110.actionControls);
} }
function DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 91);
    ɵngcc0.ɵɵelementStart(1, "button", 92);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 93);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_ul_4_Template, 2, 1, "ul", 94);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Actions"));
} }
const _c5 = function (a0) { return { "m-l-auto overflow-visible": a0 }; };
function DataGridComponent_ng_container_7_ng_container_3_td_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 80);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_td_2_ng_container_1_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵelementStart(2, "div", 17);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_3_td_2_div_3_Template, 5, 3, "div", 82);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r86 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r86.actionControls.length <= 2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c5, ctx_r86.actionControls.length > 2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r86.actionControls.length > 2);
} }
function DataGridComponent_ng_container_7_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_3_th_1_Template, 1, 0, "th", 75);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_3_td_2_Template, 4, 5, "td", 76);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-cell-renderer", 108);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const cellRendererSpec_r142 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("spec", cellRendererSpec_r142);
} }
const _c6 = function (a0, a1) { return { columnName: a0, value: a1 }; };
const _c7 = function (a0) { return [a0]; };
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 107);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_4_th_1_div_2_ng_container_2_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵpipe(3, "map");
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r74 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const ctx_r138 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 2, column_r74.header) || column_r74.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind3(3, 4, ɵngcc0.ɵɵpureFunction1(13, _c7, ɵngcc0.ɵɵpureFunction2(10, _c6, column_r74.name, ɵngcc0.ɵɵpipeBind1(4, 8, column_r74.header) || column_r74.name)), ctx_r138.getHeaderCellRendererSpec, ctx_r138));
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-cell-renderer", 108);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const cellRendererSpec_r147 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("spec", cellRendererSpec_r147);
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ul_9_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-filtering-form-renderer", 108);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const filteringFormRendererSpec_r149 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("spec", filteringFormRendererSpec_r149);
} }
const _c8 = function (a0, a1) { return { column: a0, dropdown: a1 }; };
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ul_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 45);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ul_9_Template_ul_click_0_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelementStart(1, "li", 112);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ul_9_ng_container_2_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵpipe(3, "map");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r144 = ɵngcc0.ɵɵreference(1);
    const column_r74 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const ctx_r146 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind3(3, 1, ɵngcc0.ɵɵpureFunction1(8, _c7, ɵngcc0.ɵɵpureFunction2(5, _c8, column_r74, _r144)), ctx_r146.getFilteringFormRendererSpec, ctx_r146));
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 21, 109);
    ɵngcc0.ɵɵelementStart(2, "button", 110);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ng_container_4_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵpipe(5, "map");
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelement(7, "i", 111);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(9, DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_ul_9_Template, 4, 10, "ul", 24);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r74 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const ctx_r139 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("insideClick", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(3, 4, column_r74.header) || column_r74.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind3(5, 6, ɵngcc0.ɵɵpureFunction1(17, _c7, ɵngcc0.ɵɵpureFunction2(14, _c6, column_r74.name, ɵngcc0.ɵɵpipeBind1(6, 10, column_r74.header) || column_r74.name)), ctx_r139.getHeaderCellRendererSpec, ctx_r139));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(8, 12, "Filter"));
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 117);
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 118);
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 119);
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_Template(rf, ctx) { if (rf & 1) {
    const _r158 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 113);
    ɵngcc0.ɵɵlistener("click", function DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r158); const column_r74 = ɵngcc0.ɵɵnextContext(3).$implicit; const ctx_r156 = ɵngcc0.ɵɵnextContext(); return ctx_r156.changeSortOrder(column_r74.name); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementContainerStart(2, 66);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_3_Template, 1, 0, "i", 114);
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_4_Template, 1, 0, "i", 115);
    ɵngcc0.ɵɵtemplate(5, DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_i_5_Template, 1, 0, "i", 116);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r74 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 4, "Sort"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngSwitch", column_r74.sortOrder);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "asc");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "desc");
} }
function DataGridComponent_ng_container_7_ng_container_4_th_1_Template(rf, ctx) { if (rf & 1) {
    const _r162 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "th", 102);
    ɵngcc0.ɵɵpipe(1, "map");
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_4_th_1_div_2_Template, 5, 15, "div", 103);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_4_th_1_div_3_Template, 10, 19, "div", 104);
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_4_th_1_button_4_Template, 6, 6, "button", 105);
    ɵngcc0.ɵɵelementStart(5, "span", 106);
    ɵngcc0.ɵɵlistener("mousedown", function DataGridComponent_ng_container_7_ng_container_4_th_1_Template_span_mousedown_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r162); const column_r74 = ɵngcc0.ɵɵnextContext(2).$implicit; const ctx_r160 = ɵngcc0.ɵɵnextContext(); return ctx_r160.resizeHandleMouseDown$.emit({ event: $event, targetColumnName: column_r74.name }); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const column_r74 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r136 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("sorted", column_r74.sortOrder)("filtered", ɵngcc0.ɵɵpipeBind2(1, 11, column_r74, ctx_r136.isColumnFilteringApplied))("hidden", !column_r74.visible);
    ɵngcc0.ɵɵproperty("ngClass", column_r74.headerCSSClassName);
    ɵngcc0.ɵɵattribute("data-type", column_r74.dataType);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !column_r74.filterable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r74.filterable);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", column_r74.sortable);
} }
function DataGridComponent_ng_container_7_ng_container_4_td_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-cell-renderer", 108);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const cellRendererSpec_r166 = ctx.ngIf;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("spec", cellRendererSpec_r166);
} }
const _c9 = function (a0, a1, a2) { return { value: a0, row: a1, columnName: a2 }; };
function DataGridComponent_ng_container_7_ng_container_4_td_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 120);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_4_td_2_ng_container_2_Template, 2, 1, "ng-container", 10);
    ɵngcc0.ɵɵpipe(3, "map");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const row_r164 = ctx.$implicit;
    const column_r74 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r137 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("hidden", !column_r74.visible);
    ɵngcc0.ɵɵproperty("ngClass", column_r74.cellCSSClassName);
    ɵngcc0.ɵɵattribute("data-cell-title", ɵngcc0.ɵɵpipeBind1(1, 6, column_r74.header))("data-type", column_r74.dataType);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind3(3, 8, ɵngcc0.ɵɵpureFunction1(16, _c7, ɵngcc0.ɵɵpureFunction3(12, _c9, ctx_r137.resolveCellValue(row_r164, column_r74.path), row_r164, column_r74.name)), ctx_r137.getCellRendererSpec, ctx_r137));
} }
function DataGridComponent_ng_container_7_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_7_ng_container_4_th_1_Template, 6, 14, "th", 100);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_4_td_2_Template, 4, 18, "td", 101);
    ɵngcc0.ɵɵelementContainerEnd();
} }
function DataGridComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 72);
    ɵngcc0.ɵɵelementContainerStart(1, 66);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_ng_container_7_ng_container_2_Template, 3, 0, "ng-container", 73);
    ɵngcc0.ɵɵtemplate(3, DataGridComponent_ng_container_7_ng_container_3_Template, 3, 0, "ng-container", 73);
    ɵngcc0.ɵɵtemplate(4, DataGridComponent_ng_container_7_ng_container_4_Template, 3, 0, "ng-container", 74);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const column_r74 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("cdkColumnDef", column_r74.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", column_r74.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "checkbox");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "actions");
} }
function DataGridComponent_td_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 121);
    ɵngcc0.ɵɵelement(1, "template", null, 122);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵstyleProp("grid-column", ctx_r4.styles.gridInfiniteScrollColumn);
} }
function DataGridComponent_tr_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "tr", 123);
} }
function DataGridComponent_tr_11_Template(rf, ctx) { if (rf & 1) {
    const _r171 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 124);
    ɵngcc0.ɵɵlistener("mouseover", function DataGridComponent_tr_11_Template_tr_mouseover_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r171); const row_r169 = restoredCtx.$implicit; const ctx_r170 = ɵngcc0.ɵɵnextContext(); return ctx_r170.rowMouseOver.emit(row_r169); })("mouseleave", function DataGridComponent_tr_11_Template_tr_mouseleave_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r171); const row_r169 = restoredCtx.$implicit; const ctx_r172 = ɵngcc0.ɵɵnextContext(); return ctx_r172.rowMouseLeave.emit(row_r169); })("click", function DataGridComponent_tr_11_Template_tr_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r171); const row_r169 = restoredCtx.$implicit; const ctx_r173 = ɵngcc0.ɵɵnextContext(); return ctx_r173.rowClick.emit(row_r169); });
    ɵngcc0.ɵɵelementEnd();
} }
function DataGridComponent_ng_container_12_tr_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "tr", 126);
} }
const _c10 = function () { return ["infiniteScrollFooter"]; };
function DataGridComponent_ng_container_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DataGridComponent_ng_container_12_tr_1_Template, 1, 0, "tr", 125);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkFooterRowDef", ɵngcc0.ɵɵpureFunction0(1, _c10));
} }
function DataGridComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 127);
    ɵngcc0.ɵɵelementStart(1, "div", 128);
    ɵngcc0.ɵɵprojection(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function DataGridComponent_div_16_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 136);
    ɵngcc0.ɵɵelementStart(1, "span", 137);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(2, "{{ pageFirstItemIdx }} - {{ pageLastItemIdx }} of {{ itemsTotal }}");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r175 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("translateParams", ctx_r175.paginationLabelParams);
} }
function DataGridComponent_div_16_div_5_option_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 143);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const pageSize_r178 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngValue", pageSize_r178);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", pageSize_r178, " ");
} }
function DataGridComponent_div_16_div_5_Template(rf, ctx) { if (rf & 1) {
    const _r180 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 138);
    ɵngcc0.ɵɵelementStart(1, "label", 139);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 140);
    ɵngcc0.ɵɵelementStart(5, "select", 141);
    ɵngcc0.ɵɵlistener("ngModelChange", function DataGridComponent_div_16_div_5_Template_select_ngModelChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r180); const ctx_r179 = ɵngcc0.ɵɵnextContext(2); return ctx_r179.updatePagination({ itemsPerPage: $event, page: ctx_r179.pagination.currentPage }); });
    ɵngcc0.ɵɵtemplate(6, DataGridComponent_div_16_div_5_option_6_Template, 2, 2, "option", 142);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r176 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 3, "Items per page"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r176.pagination.pageSize);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r176.possiblePageSizes);
} }
function DataGridComponent_div_16_Template(rf, ctx) { if (rf & 1) {
    const _r182 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 129);
    ɵngcc0.ɵɵelementStart(1, "div", 130);
    ɵngcc0.ɵɵtemplate(2, DataGridComponent_div_16_div_2_Template, 3, 1, "div", 131);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 132);
    ɵngcc0.ɵɵtemplate(5, DataGridComponent_div_16_div_5_Template, 7, 5, "div", 133);
    ɵngcc0.ɵɵpipe(6, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 134);
    ɵngcc0.ɵɵelementStart(8, "pagination", 135);
    ɵngcc0.ɵɵlistener("pageChanged", function DataGridComponent_div_16_Template_pagination_pageChanged_8_listener($event) { ɵngcc0.ɵɵrestoreView(_r182); const ctx_r181 = ɵngcc0.ɵɵnextContext(); return ctx_r181.updatePagination($event); })("numPages", function DataGridComponent_div_16_Template_pagination_numPages_8_listener($event) { ɵngcc0.ɵɵrestoreView(_r182); const ctx_r183 = ɵngcc0.ɵɵnextContext(); return ctx_r183.totalPagesCount$.next($event); });
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵpipe(10, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(3, 9, ctx_r9.dataSource.stats$).currentPageSize > 0);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(6, 11, ctx_r9.dataSource.stats$).filteredSize > ctx_r9.minPossiblePageSize);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵclassProp("hidden", ɵngcc0.ɵɵpipeBind1(9, 13, ctx_r9.hidePagination$));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r9.pagination.currentPage)("totalItems", ɵngcc0.ɵɵpipeBind1(10, 15, ctx_r9.dataSource.stats$).filteredSize)("itemsPerPage", ctx_r9.pagination.pageSize)("maxSize", 5)("boundaryLinks", false);
} }
const _c11 = [[["", 8, "c8y-empty-state"]]];
const _c12 = function (a0) { return { "table-data-grid__overlay": a0 }; };
const _c13 = [".c8y-empty-state"];
export class DataGridComponent {
    constructor(sanitizer, hostRef, cd, cfr) {
        this.sanitizer = sanitizer;
        this.hostRef = hostRef;
        this.cd = cd;
        this.cfr = cfr;
        /** The title for the data grid, it's displayed in the grid's header. */
        this.title = gettext('Items');
        /** The label for load more button. */
        this.loadMoreItemsLabel = gettext('Load more items');
        /** The label for loading indicator. */
        this.loadingItemsLabel = gettext('Loading items…');
        /** Determines whether text search input is shown in the grid's header. */
        this.showSearch = false;
        this.dataSource = new GridDataSource();
        this.filteringLabelsParams = {
            filteredItemsCount: 0,
            allItemsCount: 0
        };
        this.paginationLabelParams = {
            pageFirstItemIdx: 0,
            pageLastItemIdx: 0,
            itemsTotal: 0
        };
        this.possiblePageSizes = [10, 15, 20, 30, 50];
        this.minPossiblePageSize = Math.min(...this.possiblePageSizes);
        this.selectable = false;
        this.selectionPrimaryKey = 'id';
        this.displayOptions = {
            striped: true,
            bordered: false,
            gridHeader: true,
            filter: true
        };
        /** Sets initial search text. */
        this.searchText = '';
        /** Shows the warning for the sub-assets counter */
        this.showCounterWarning = false;
        /** Emits an event when mouse is over a row. */
        this.rowMouseOver = new EventEmitter();
        /** Emits an event when mouse leaves a row. */
        this.rowMouseLeave = new EventEmitter();
        /** Emits an event when a row is clicked. */
        this.rowClick = new EventEmitter();
        /** Emits an event when grid's configuration is changed. */
        this.onConfigChange = new EventEmitter();
        /** Emits an event when a filter is applied in a column. */
        this.onFilter = new EventEmitter();
        /** Emits an event when items selection changes. The array contains keys of selected items (key property is defined by `selectionPrimaryKey`). */
        this.itemsSelect = new EventEmitter();
        /** Emits an event when reload button is clicked. */
        this.onReload = new EventEmitter();
        this.columnNames = [];
        this.styles = {
            tableCursor: 'auto',
            gridTemplateColumns: undefined,
            gridInfiniteScrollColumn: undefined
        };
        this.searchText$ = new EventEmitter();
        this.filteringApplied = false;
        this.totalPagesCount$ = new BehaviorSubject(Infinity);
        this.hidePagination$ = this.totalPagesCount$.pipe(map(totalPagesCount => totalPagesCount <= 1), delay(0) // prevents ExpressionChangedAfterItHasBeenCheckedError
        );
        this.selectedItemIds = [];
        this.currentPageSelectionState = {
            allSelected: false,
            allDeselected: true
        };
        this.builtInActionType = {
            Edit: "EDIT" /* Edit */,
            Delete: "DELETE" /* Delete */,
            Export: "EXPORT" /* Export */
        };
        this.resizeHandleMouseDown$ = new EventEmitter();
        this.resizeHandleContainerMouseMove$ = new EventEmitter();
        this.resizeHandleContainerMouseUp$ = new EventEmitter();
        this.unsubscribe$ = new Subject();
        /**
         * Event emitter used for loading data grid data with debounce.
         * This is used to avoid having multiple this.loadData() function calls.
         */
        this.triggerLoadData = new EventEmitter();
        this.triggerLoadData.pipe(debounceTime(1), takeUntil(this.unsubscribe$)).subscribe(() => {
            this.loadData();
        });
    }
    /** The list of columns to be displayed in the grid. */
    set _columns(columns) {
        this.columns = columns || [];
    }
    /** The list of rows to be displayed in the grid (used for client side data). */
    set _rows(rows) {
        this.rows = rows || [];
    }
    /** Pagination settings, e.g. allows for setting current page or page size. */
    set _pagination(pagination) {
        this.pagination = pagination;
        const pageSize = get(this.pagination, 'pageSize');
        if (pageSize &&
            !this.possiblePageSizes.find(possiblePageSize => possiblePageSize === pageSize)) {
            this.pagination = Object.assign(Object.assign({}, this.pagination), { pageSize: 10 });
        }
    }
    /** Sets load more mode. */
    set _infiniteScroll(infiniteScroll) {
        this.infiniteScroll = infiniteScroll;
    }
    /**
     * Sets a callback function which will be invoked whenever data needs to be loaded from server.
     * The function should take [[DataSourceModifier]] and return [[ServerSideDataResult]].
     */
    set _serverSideDataCallback(serverSideDataCallback) {
        this.serverSideDataCallback = serverSideDataCallback;
    }
    /** Determines whether items can be selected by clicking a checkbox in the first column. */
    set _selectable(selectable) {
        this.selectable = selectable;
    }
    /** Determines which item's property will be used to distinguish selection status. */
    set _selectionPrimaryKey(selectionPrimaryKey) {
        this.selectionPrimaryKey = selectionPrimaryKey;
    }
    /** Sets display options. */
    set _displayOptions(displayOptions) {
        this.displayOptions = Object.assign(Object.assign({}, this.displayOptions), displayOptions);
    }
    /** Sets action controls (actions available for individual items). */
    set _actionControls(actionControls) {
        this.actionControls = actionControls || [];
    }
    /** Sets bulk action controls (actions available for items selected by user). */
    set _bulkActionControls(bulkActionControls) {
        this.bulkActionControls = bulkActionControls || [];
    }
    /** Sets header action controls (actions available from data grid header). */
    set _headerActionControls(headerActionControls) {
        this.headerActionControls = headerActionControls || [];
    }
    ngOnInit() {
        this.searchText$
            .pipe(takeUntil(this.unsubscribe$), debounceTime(300), distinctUntilChanged(), tap(searchText => {
            this.searchText = searchText;
        }))
            .subscribe(() => {
            this.reload();
        });
        if (this.selectable) {
            combineLatest(this.dataSource.data$, this.itemsSelect.asObservable())
                .pipe(takeUntil(this.unsubscribe$))
                .subscribe(([data]) => {
                const currentPageEmpty = data.length === 0;
                this.currentPageSelectionState = {
                    allSelected: currentPageEmpty ? false : data.every(item => this.isItemSelected(item)),
                    allDeselected: currentPageEmpty ? true : data.every(item => !this.isItemSelected(item))
                };
            });
        }
        this.dataSource.stats$.pipe(takeUntil(this.unsubscribe$)).subscribe(stats => {
            this.createLoadMoreComponent(stats);
            this.updateFilteringLabelsParams(stats);
            this.updatePaginationLabelParams(stats);
        });
        if (this.refresh) {
            this.refresh.pipe(takeUntil(this.unsubscribe$)).subscribe(() => {
                this.cancel();
                this.reload();
            });
        }
    }
    ngOnChanges(event) {
        if (!event._actionControls || event._actionControls.firstChange) {
            this.triggerLoadData.emit();
        }
        this.updateColumns();
    }
    ngAfterViewInit() {
        this.updateGridColumnsSize();
        this.thEls = this.thRefs.toArray().map(({ nativeElement }) => nativeElement);
        this.setupResizeHandle();
    }
    ngOnDestroy() {
        this.unsubscribe$.next();
        this.unsubscribe$.complete();
    }
    trackByName(index, item) {
        return item.name;
    }
    resolveCellValue(row, path) {
        return flow([
            x => this.dataSource.resolveValue(x, path),
            this.dataSource.resolveFunction,
            this.dataSource.normalizeNil
        ])(row);
    }
    changeSortOrder(columnName) {
        const column = this.columns.find(({ name }) => name === columnName);
        if (column) {
            const { sortOrder } = column;
            if (!sortOrder) {
                this.updateSorting([columnName], 'asc');
            }
            else if (sortOrder === 'asc') {
                this.updateSorting([columnName], 'desc');
            }
            else {
                this.updateSorting([columnName], '');
            }
        }
    }
    updateSorting(columnNames, sortOrder) {
        this.columns = this.columns.map((column) => {
            if (columnNames.includes(column.name)) {
                return Object.assign(Object.assign({}, column), { sortOrder });
            }
            return column;
        });
        this.emitConfigChange();
        this.reload();
    }
    applyFilter(columnName, dropdown, filteringModifier) {
        this.updateFiltering([columnName], {
            type: FilteringActionType.ApplyFilter,
            payload: { filteringModifier }
        });
        dropdown.hide();
        this.onFilter.emit({ columnName, dropdown, filteringModifier });
    }
    resetFilter(columnName, dropdown) {
        this.updateFiltering([columnName], { type: FilteringActionType.ResetFilter });
        dropdown.hide();
    }
    clearFilters() {
        this.updateFiltering(this.columns.map(({ name }) => name), {
            type: FilteringActionType.ResetFilter
        });
    }
    updateFiltering(columnNames, action, reload = true) {
        this.columns = this.columns.map(column => {
            if (columnNames.includes(column.name)) {
                return Object.assign(Object.assign({}, column), (action.type === FilteringActionType.ApplyFilter
                    ? action.payload.filteringModifier
                    : { filterPredicate: undefined, externalFilterQuery: undefined }));
            }
            return column;
        });
        this.updateFilteringApplied();
        if (reload) {
            this.reload();
        }
    }
    updateFilteringApplied() {
        this.filteringApplied = this.columns.some(this.isColumnFilteringApplied);
    }
    isColumnFilteringApplied(column) {
        const { filterable, filterPredicate, externalFilterQuery } = column;
        return !!(filterable && (filterPredicate || externalFilterQuery));
    }
    updatePagination({ itemsPerPage, page }) {
        this.pagination = Object.assign(Object.assign({}, this.pagination), { pageSize: itemsPerPage, currentPage: page });
        this.loadData();
        this.emitConfigChange();
    }
    clickReload() {
        this.searchText = '';
        this.reload();
        this.onReload.next();
    }
    reload() {
        this.pagination = Object.assign(Object.assign({}, this.pagination), { currentPage: 1 });
        this.recreateLoadMoreComponent = true;
        this.loadData(true);
        this.scrollToTop();
    }
    loadNextPage() {
        this.pagination = Object.assign(Object.assign({}, this.pagination), { currentPage: this.pagination.nextPage });
        this.loadData();
        return this.dataSource.resultList$
            .pipe(take(1)) // in order for `toPromise` to work, the observable needs to complete
            .toPromise()
            .then(result => {
            return Object.assign(Object.assign({}, result), { paging: Object.assign(Object.assign({}, result.paging), { next: this.loadNextPage.bind(this) }) });
        });
    }
    getCellRendererSpec({ value, row, columnName }) {
        return this._getCellRendererSpec({ type: 'CELL', value, row, columnName });
    }
    getHeaderCellRendererSpec({ value, columnName }) {
        return this._getCellRendererSpec({ type: 'HEADER', value, row: undefined, columnName });
    }
    getFilteringFormRendererSpec({ column, dropdown }) {
        return {
            renderer: get(this.getColumnRenderer(column), 'filteringFormRendererDef.template') ||
                column.filteringFormRendererComponent,
            context: {
                property: column,
                applyFilter: this.applyFilter.bind(this, column.name, dropdown),
                resetFilter: this.resetFilter.bind(this, column.name, dropdown)
            }
        };
    }
    setAllItemsSelected(selected) {
        this.dataSource.selection$
            .pipe(first())
            .subscribe(({ filteredDataIds }) => this.setItemsSelected(filteredDataIds, selected));
    }
    setAllItemsInCurrentPageSelected(selected) {
        this.dataSource.data$.pipe(first()).subscribe(data => this.setItemsSelected(data, selected));
    }
    setItemsSelected(items, selected) {
        const itemIds = items.map((item) => typeof item === 'object' ? item[this.selectionPrimaryKey] : item);
        this.selectedItemIds = selected
            ? union(this.selectedItemIds, itemIds)
            : without(this.selectedItemIds, ...itemIds);
        this.itemsSelect.emit(this.selectedItemIds);
    }
    cancel() {
        this.selectedItemIds = [];
        this.itemsSelect.emit(this.selectedItemIds);
    }
    isItemSelected(item) {
        return this.selectedItemIds.includes(item[this.selectionPrimaryKey]);
    }
    onColumnDrop({ previousIndex, currentIndex }) {
        const differentIndex = previousIndex !== currentIndex;
        if (differentIndex) {
            const column = this.columns.splice(previousIndex, 1);
            this.columns.splice(currentIndex, 0, column[0]);
        }
        this.updateColumnNames();
        this.updateGridColumnsSize();
    }
    updateGridColumnsSize() {
        this.styles = Object.assign(Object.assign({}, this.styles), { gridTemplateColumns: this.sanitizer.bypassSecurityTrustStyle(this.columns
                .filter(column => column.visible)
                .map(({ gridTrackSize }) => gridTrackSize)
                .join(' ')), gridInfiniteScrollColumn: this.sanitizer.bypassSecurityTrustStyle(`1 / span ${this.columns.filter(column => column.visible).length}`) });
        this.emitConfigChange();
    }
    loadData(reload = false) {
        const { rows, columns, pagination, searchText, serverSideDataCallback, selectable, selectionPrimaryKey, infiniteScroll } = this;
        this.dataSource.loadData({
            rows,
            columns,
            pagination,
            searchText,
            serverSideDataCallback,
            selectable,
            selectionPrimaryKey,
            infiniteScroll,
            reload
        });
    }
    emitConfigChange() {
        const columns = this.columns.map(({ visible, sortOrder, name }) => ({
            visible,
            sortOrder,
            name
        }));
        this.onConfigChange.emit({ columns, pagination: this.pagination });
    }
    updateColumns() {
        const specialColumn = {
            gridTrackSize: 'minmax(42px, 82px)',
            sortable: false,
            positionFixed: true
        };
        const columns = [
            this.selectable ? Object.assign(Object.assign({}, specialColumn), { name: "checkbox" /* Checkbox */ }) : undefined,
            ...this.columns,
            this.actionControls.length > 0
                ? Object.assign(Object.assign({}, specialColumn), { name: "actions" /* Actions */ }) : undefined
        ]
            .filter(Boolean)
            .map(this.withColumnDefaults);
        this.columns = uniqBy(columns, 'name');
        this.updateColumnNames();
        this.updateGridColumnsSize();
        this.updateFilteringApplied();
    }
    withColumnDefaults(column) {
        const dataType = column.dataType || "text-short" /* TextShort */;
        const { headerCSSClassName, cellCSSClassName } = column;
        return Object.assign(Object.assign({ visible: true, positionFixed: false, sortable: true, sortOrder: '', filterable: false }, column), { dataType, gridTrackSize: column.gridTrackSize ||
                `minmax(${minColumnGridTrackSize}px, ${ratiosByColumnTypes[dataType]}fr)`, headerCSSClassName: (typeof headerCSSClassName === 'string'
                ? headerCSSClassName.split(' ')
                : headerCSSClassName) || [], cellCSSClassName: (typeof cellCSSClassName === 'string' ? cellCSSClassName.split(' ') : cellCSSClassName) ||
                [] });
    }
    updateColumnNames() {
        this.columnNames = this.columns.map(({ name }) => name);
    }
    setupResizeHandle() {
        const resizeHandleDrag$ = this.resizeHandleMouseDown$.pipe(takeUntil(this.unsubscribe$), tap(() => this.clearMouseHighlights()), mergeMap(({ event, targetColumnName }) => {
            this.columns = this.columns.map(column => {
                if (column.name === targetColumnName) {
                    return Object.assign(Object.assign({}, column), { headerCSSClassName: union(column.headerCSSClassName, ['header--being-resized']) });
                }
                return column;
            });
            this.headerBeingResized = {
                columnName: targetColumnName,
                el: event.target.parentNode
            };
            this.styles = Object.assign(Object.assign({}, this.styles), { tableCursor: 'col-resize' });
            return this.resizeHandleContainerMouseMove$.pipe(tap(() => this.clearMouseHighlights()), takeUntil(this.resizeHandleContainerMouseUp$));
        }));
        resizeHandleDrag$.subscribe(event => {
            requestAnimationFrame(() => {
                this.columns = this.columns.map((column, i) => {
                    if (column.name === this.headerBeingResized.columnName) {
                        const wrapper = this.hostRef.nativeElement;
                        // Read wrapper's offset left relative to the document.
                        const horizontalOffset = wrapper.getBoundingClientRect().left;
                        // Adjust with the wrapper horizontal scroll position.
                        const horizontalScrollOffset = wrapper.scrollLeft - horizontalOffset;
                        // Calculate the desired width.
                        const width = horizontalScrollOffset + event.clientX - (this.headerBeingResized.el.offsetLeft || 0);
                        return Object.assign(Object.assign({}, column), { 
                            // Update the column object with the new size value, enforce our minimum size.
                            gridTrackSize: `${Math.max(minColumnGridTrackSize, width)}px` });
                    }
                    // For the other headers which don't have a set width, fix it to their computed width.
                    if (column.gridTrackSize.startsWith('minmax')) {
                        return Object.assign(Object.assign({}, column), { 
                            // isn't fixed yet (it would be a px value)
                            gridTrackSize: `${_parseInt(this.thEls[i].clientWidth)}px` });
                    }
                    return column;
                });
                /*
                 * Update the column sizes.
                 * Note: grid-template-columns sets the width for all columns in one value.
                 */
                this.updateGridColumnsSize();
            });
        });
        this.resizeHandleContainerMouseUp$.pipe(takeUntil(this.unsubscribe$)).subscribe(() => {
            if (this.headerBeingResized) {
                this.columns = this.columns.map(column => {
                    if (column.name === this.headerBeingResized.columnName) {
                        return Object.assign(Object.assign({}, column), { headerCSSClassName: without(column.headerCSSClassName, 'header--being-resized') });
                    }
                    return column;
                });
                this.headerBeingResized = undefined;
                this.styles = Object.assign(Object.assign({}, this.styles), { tableCursor: 'auto' });
            }
        });
    }
    clearMouseHighlights() {
        if (window.getSelection) {
            window.getSelection().removeAllRanges();
        }
    }
    _getCellRendererSpec({ type, value, row, columnName }) {
        const column = this.columns.find(({ name }) => name === columnName);
        const columnRenderer = this.getColumnRenderer(column);
        const rendererTemplate = get(columnRenderer, `${type === 'HEADER' ? 'headerCellRendererDef' : 'cellRendererDef'}.template`);
        const { cellRendererComponent, headerCellRendererComponent } = column;
        const rendererComponent = type === 'HEADER' ? headerCellRendererComponent : cellRendererComponent;
        return {
            renderer: rendererTemplate || rendererComponent,
            context: {
                value,
                item: row,
                property: column
            }
        };
    }
    getColumnRenderer(column) {
        return this.columnRenderers.toArray().find(({ name }) => name === column.name);
    }
    updateFilteringLabelsParams(stats) {
        this.filteringLabelsParams = {
            filteredItemsCount: stats.filteredSize,
            allItemsCount: stats.size
        };
    }
    updatePaginationLabelParams(stats) {
        this.pagination.nextPage = stats.nextPage;
        const pageFirstItemIdx = (stats.currentPage - 1) * stats.firstPageSize + 1;
        this.paginationLabelParams = {
            pageFirstItemIdx,
            pageLastItemIdx: pageFirstItemIdx + (stats.currentPageSize - 1),
            itemsTotal: stats.filteredSize
        };
    }
    createLoadMoreComponent(stats) {
        if (this.infiniteScroll &&
            stats &&
            stats.nextPage &&
            (!this.loadMoreComponent || this.recreateLoadMoreComponent)) {
            this.recreateLoadMoreComponent = false;
            this.infiniteScrollContainer.clear();
            const componentFactory = this.cfr.resolveComponentFactory(LoadMoreComponent);
            const componentRef = this.infiniteScrollContainer.createComponent(componentFactory);
            const instance = componentRef.instance;
            instance.useIntersection = this.infiniteScroll === 'auto' || this.infiniteScroll === 'hidden';
            instance.hidden = this.infiniteScroll === 'hidden';
            instance.paging = {
                nextPage: stats.nextPage,
                next: this.loadNextPage.bind(this)
            };
            instance.loadNextLabel = this.loadMoreItemsLabel;
            instance.loadingLabel = this.loadingItemsLabel;
            this.loadMoreComponent = instance;
        }
        else if (this.loadMoreComponent && !stats.nextPage) {
            this.loadMoreComponent.paging = {
                nextPage: null
            };
        }
    }
    scrollToTop() {
        if (this.infiniteScroll) {
            this.scrollContainer.nativeElement.scrollTop = 0;
        }
    }
}
DataGridComponent.ɵfac = function DataGridComponent_Factory(t) { return new (t || DataGridComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomSanitizer), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver)); };
DataGridComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DataGridComponent, selectors: [["c8y-data-grid"]], contentQueries: function DataGridComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ColumnDirective, 4);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.columnRenderers = _t);
    } }, viewQuery: function DataGridComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 5, ViewContainerRef);
        ɵngcc0.ɵɵviewQuery(CdkHeaderCell, 5, ElementRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.scrollContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.infiniteScrollContainer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.thRefs = _t);
    } }, inputs: { title: "title", loadMoreItemsLabel: "loadMoreItemsLabel", loadingItemsLabel: "loadingItemsLabel", showSearch: "showSearch", searchText: "searchText", showCounterWarning: "showCounterWarning", _columns: ["columns", "_columns"], _rows: ["rows", "_rows"], _pagination: ["pagination", "_pagination"], _infiniteScroll: ["infiniteScroll", "_infiniteScroll"], _serverSideDataCallback: ["serverSideDataCallback", "_serverSideDataCallback"], _selectable: ["selectable", "_selectable"], _selectionPrimaryKey: ["selectionPrimaryKey", "_selectionPrimaryKey"], _displayOptions: ["displayOptions", "_displayOptions"], _actionControls: ["actionControls", "_actionControls"], _bulkActionControls: ["bulkActionControls", "_bulkActionControls"], _headerActionControls: ["headerActionControls", "_headerActionControls"], refresh: "refresh" }, outputs: { rowMouseOver: "rowMouseOver", rowMouseLeave: "rowMouseLeave", rowClick: "rowClick", onConfigChange: "onConfigChange", onFilter: "onFilter", itemsSelect: "itemsSelect", onReload: "onReload" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c13, decls: 17, vars: 32, consts: [[1, "table-data-grid-scroll", 3, "ngClass"], ["scroll", ""], ["class", "table-data-grid__loading--wrapper", 4, "ngIf"], ["class", "table-data-grid-header separator large-padding", 4, "ngIf"], ["cdk-table", "", 1, "table", "table-hover", "table-filtered-sorted", "table-data-grid", "large-padding", 3, "dataSource", "trackBy", "mousemove", "mouseup"], [3, "cdkColumnDef", 4, "ngFor", "ngForOf", "ngForTrackBy"], ["cdkColumnDef", "infiniteScrollFooter"], ["cdk-footer-cell", "", 3, "grid-column", 4, "cdkFooterCellDef"], ["cdk-header-row", "", 4, "cdkHeaderRowDef"], ["cdk-row", "", 3, "mouseover", "mouseleave", "click", 4, "cdkRowDef", "cdkRowDefColumns"], [4, "ngIf"], ["class", "row m-0 p-t-40 p-b-40", 4, "ngIf"], ["class", "table-data-grid-footer separator large-padding", 4, "ngIf"], [1, "table-data-grid__loading--wrapper"], [1, "table-data-grid__loading--loader"], [3, "message"], [1, "table-data-grid-header", "separator", "large-padding"], [3, "ngClass"], [1, "flex-item-right"], [1, "btnbar", "flex-row"], [4, "ngFor", "ngForOf"], ["dropdown", "", "container", "body", "placement", "bottom left", 1, "dropdown", 3, "insideClick"], ["dropdownToggle", "", 1, "btnbar-btn", "c8y-dropdown", 3, "title"], ["c8yIcon", "columns", 1, "m-r-4"], ["class", "dropdown-menu", 3, "click", 4, "dropdownMenu"], [1, "btnbar-btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "refresh", 1, "m-r-4"], ["class", "input-group input-group-search m-l-sm-16", 4, "ngIf"], ["class", "table-data-grid-header-bulk-actions animated slideInDown fast", 4, "ngIf"], [1, "hidden-xs", "hidden-sm"], ["placement", "right", "triggers", "focus", 1, "btn-clean", "text-primary", 3, "popover"], ["c8yIcon", "question-circle-o"], ["class", "hidden-xs hidden-sm", 4, "ngIf"], ["translate", "", 1, "label", "label-default", "m-r-4"], ["class", "m-r-4", "ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], ["translate", "", 1, "m-r-4", 3, "translateParams"], [1, "m-l-8", "btn", "btn-xs", "btn-default", "m-r-4", 3, "title", "click"], [1, "badge", "badge-info", "m-r-4"], ["translate", "", 3, "translateParams"], ["c8yIcon", "warning"], [4, "ngIf", "ngIfElse"], ["customTemplate", ""], [1, "btnbar-btn", "btn-link", 3, "title", "click"], ["c8yIcon", "headerActionControl.icon", 1, "m-r-4"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "dropdown-menu", 3, "click"], [1, "bg-white"], ["cdkDropList", "", 1, "list-group", "m-0", 3, "cdkDropListDropped"], ["cdkDrag", "", "cdkDragLockAxis", "y", 4, "ngFor", "ngForOf"], ["cdkDrag", "", "cdkDragLockAxis", "y"], [1, "list-group-item", "draggable-after", "p-0"], [1, "c8y-checkbox", "p-l-16", "p-r-32", "p-t-8", "p-b-8", "flex-item-v-stretch", 3, "title"], ["type", "checkbox", 3, "ngModel", "ngModelChange", "change"], [1, "input-group", "input-group-search", "m-l-sm-16"], ["type", "search", 1, "form-control", 3, "placeholder", "ngModel", "input"], [1, "input-group-addon"], ["c8yIcon", "search", 4, "ngIf"], ["c8yIcon", "times", "class", "pointer", 3, "click", 4, "ngIf"], ["c8yIcon", "search"], ["c8yIcon", "times", 1, "pointer", 3, "click"], [1, "table-data-grid-header-bulk-actions", "animated", "slideInDown", "fast"], [1, "visible-xs"], [3, "ngSwitch", 4, "ngFor", "ngForOf"], [1, "btnbar-btn", 3, "title", "click"], ["c8yIcon", "times"], [1, "interact", 3, "click"], [3, "ngSwitch"], ["class", "btnbar-btn", 3, "title", "click", 4, "ngSwitchCase"], ["class", "btnbar-btn", 3, "title", "click", 4, "ngSwitchDefault"], ["c8yIcon", "sign-out"], ["c8yIcon", "trash"], [3, "c8yIcon"], [3, "cdkColumnDef"], [4, "ngSwitchCase"], [4, "ngSwitchDefault"], ["cdk-header-cell", "", "data-type", "icon", 4, "cdkHeaderCellDef"], ["cdk-cell", "", "data-type", "icon", 4, "cdkCellDef"], ["cdk-header-cell", "", "data-type", "icon"], [1, "c8y-checkbox"], ["type", "checkbox", 3, "checked", "indeterminate", "change"], ["cdk-cell", "", "data-type", "icon"], ["type", "checkbox", 3, "checked", "change"], ["class", "dropdown", "dropdown", "", "container", "body", "placement", "bottom right", 4, "ngIf"], ["class", "btn btn-icon btn-xs btn-default", 3, "title", "click", 4, "ngSwitchCase"], ["class", "btn btn-dot showOnHover", 3, "title", "click", 4, "ngSwitchCase"], ["class", "btn btn-icon btn-xs btn-default", "container", "body", 3, "title", "tooltip", "click", 4, "ngSwitchDefault"], [1, "btn", "btn-icon", "btn-xs", "btn-default", 3, "title", "click"], ["c8yIcon", "pencil"], [1, "btn", "btn-dot", "showOnHover", 3, "title", "click"], ["c8yIcon", "minus-circle", 1, "text-danger"], ["container", "body", 1, "btn", "btn-icon", "btn-xs", "btn-default", 3, "title", "tooltip", "click"], ["dropdown", "", "container", "body", "placement", "bottom right", 1, "dropdown"], ["dropdownToggle", "", 1, "dropdown-toggle", "c8y-dropdown", 3, "title"], ["c8yIcon", "ellipsis-v"], ["class", "dropdown-menu dropdown-menu-right", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu-right"], [3, "title", "click", 4, "ngSwitchCase"], [3, "title", "click", 4, "ngSwitchDefault"], [3, "title", "click"], ["c8yIcon", "download"], ["cdk-header-cell", "", 3, "ngClass", "sorted", "filtered", "hidden", 4, "cdkHeaderCellDef"], ["cdk-cell", "", 3, "ngClass", "hidden", 4, "cdkCellDef"], ["cdk-header-cell", "", 3, "ngClass"], [3, "title", 4, "ngIf"], ["class", "dropdown", "dropdown", "", "container", "body", "placement", "bottom left", 3, "insideClick", 4, "ngIf"], ["class", "btn-sort", 3, "title", "click", 4, "ngIf"], [1, "resize-handle", 3, "mousedown"], [3, "title"], [3, "spec"], ["dropdown", "bs-dropdown"], ["dropdownToggle", "", 1, "btn-header", "c8y-dropdown", 3, "title"], ["c8yIcon", "filter", 3, "title"], [1, "data-grid__dropdown"], [1, "btn-sort", 3, "title", "click"], ["c8yIcon", "long-arrow-up", 4, "ngSwitchCase"], ["c8yIcon", "long-arrow-down", 4, "ngSwitchCase"], ["c8yIcon", "exchange", 4, "ngSwitchDefault"], ["c8yIcon", "long-arrow-up"], ["c8yIcon", "long-arrow-down"], ["c8yIcon", "exchange"], ["cdk-cell", "", 3, "ngClass"], ["cdk-footer-cell", ""], ["infiniteScrollContainer", ""], ["cdk-header-row", ""], ["cdk-row", "", 3, "mouseover", "mouseleave", "click"], ["cdk-footer-row", "", 4, "cdkFooterRowDef"], ["cdk-footer-row", ""], [1, "row", "m-0", "p-t-40", "p-b-40"], [1, "col-md-4", "col-md-offset-4"], [1, "table-data-grid-footer", "separator", "large-padding"], [1, "col-sm-4", "no-gutter"], ["class", "counter", 4, "ngIf"], [1, "col-sm-4", "no-gutter", "text-center"], ["class", "form-group form-inline", 4, "ngIf"], [1, "col-sm-4", "no-gutter", "text-right"], ["previousText", " ", "nextText", " ", 3, "ngModel", "totalItems", "itemsPerPage", "maxSize", "boundaryLinks", "pageChanged", "numPages"], [1, "counter"], ["translate", "", 1, "text-muted", 3, "translateParams"], [1, "form-group", "form-inline"], [1, "m-r-4"], [1, "c8y-select-wrapper"], [1, "form-control", 3, "ngModel", "ngModelChange"], [3, "ngValue", 4, "ngFor", "ngForOf"], [3, "ngValue"]], template: function DataGridComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c11);
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵtemplate(3, DataGridComponent_div_3_Template, 3, 1, "div", 2);
        ɵngcc0.ɵɵpipe(4, "async");
        ɵngcc0.ɵɵtemplate(5, DataGridComponent_div_5_Template, 25, 26, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "table", 4);
        ɵngcc0.ɵɵlistener("mousemove", function DataGridComponent_Template_table_mousemove_6_listener($event) { return ctx.resizeHandleContainerMouseMove$.emit($event); })("mouseup", function DataGridComponent_Template_table_mouseup_6_listener($event) { return ctx.resizeHandleContainerMouseUp$.emit($event); });
        ɵngcc0.ɵɵtemplate(7, DataGridComponent_ng_container_7_Template, 5, 4, "ng-container", 5);
        ɵngcc0.ɵɵelementContainerStart(8, 6);
        ɵngcc0.ɵɵtemplate(9, DataGridComponent_td_9_Template, 3, 2, "td", 7);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(10, DataGridComponent_tr_10_Template, 1, 0, "tr", 8);
        ɵngcc0.ɵɵtemplate(11, DataGridComponent_tr_11_Template, 1, 0, "tr", 9);
        ɵngcc0.ɵɵtemplate(12, DataGridComponent_ng_container_12_Template, 2, 2, "ng-container", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(13, DataGridComponent_div_13_Template, 3, 0, "div", 11);
        ɵngcc0.ɵɵpipe(14, "async");
        ɵngcc0.ɵɵpipe(15, "async");
        ɵngcc0.ɵɵtemplate(16, DataGridComponent_div_16_Template, 11, 17, "div", 12);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(30, _c12, ɵngcc0.ɵɵpipeBind1(2, 22, ctx.dataSource.loading$) && !ctx.loadMoreComponent));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 24, ctx.dataSource.loading$) && !ctx.loadMoreComponent);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.displayOptions.gridHeader);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("grid-template-columns", ctx.styles.gridTemplateColumns);
        ɵngcc0.ɵɵclassProp("table-striped", ctx.displayOptions.striped)("table-bordered", ctx.displayOptions.bordered)("table-data-grid-with-checkboxes", ctx.selectable)("table-data-grid-with-actions", ctx.actionControls.length > 0);
        ɵngcc0.ɵɵproperty("dataSource", ctx.dataSource)("trackBy", ctx.trackByName);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns)("ngForTrackBy", ctx.trackByName);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("cdkHeaderRowDef", ctx.columnNames);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("cdkRowDefColumns", ctx.columnNames);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.infiniteScroll);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(14, 26, ctx.dataSource.loading$) && ɵngcc0.ɵɵpipeBind1(15, 28, ctx.dataSource.stats$).filteredSize === 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.pagination && !ctx.infiniteScroll);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgIf, ɵngcc3.CdkTable, ɵngcc2.NgForOf, ɵngcc3.CdkColumnDef, ɵngcc3.CdkFooterCellDef, ɵngcc3.CdkHeaderRowDef, ɵngcc3.CdkRowDef, ɵngcc4.ProgressBarComponent, ɵngcc5.BsDropdownDirective, ɵngcc5.BsDropdownToggleDirective, ɵngcc6.IconDirective, ɵngcc5.BsDropdownMenuDirective, ɵngcc7.PopoverDirective, ɵngcc8.C8yTranslateDirective, ɵngcc2.NgTemplateOutlet, ɵngcc9.CdkDropList, ɵngcc9.CdkDrag, ɵngcc10.CheckboxControlValueAccessor, ɵngcc10.NgControlStatus, ɵngcc10.NgModel, ɵngcc10.DefaultValueAccessor, ɵngcc2.NgSwitch, ɵngcc2.NgSwitchCase, ɵngcc2.NgSwitchDefault, ɵngcc3.CdkHeaderCellDef, ɵngcc3.CdkCellDef, ɵngcc3.CdkHeaderCell, ɵngcc3.CdkCell, ɵngcc11.TooltipDirective, ɵngcc12.CellRendererComponent, ɵngcc13.FilteringFormRendererComponent, ɵngcc3.CdkFooterCell, ɵngcc3.CdkHeaderRow, ɵngcc3.CdkRow, ɵngcc3.CdkFooterRowDef, ɵngcc3.CdkFooterRow, ɵngcc14.PaginationComponent, ɵngcc10.SelectControlValueAccessor, ɵngcc10.NgSelectOption, ɵngcc10.ɵNgSelectMultipleOption], pipes: [ɵngcc2.AsyncPipe, ɵngcc15.C8yTranslatePipe, ɵngcc16.IsControlVisiblePipe, ɵngcc17.MapFunctionPipe], encapsulation: 2 });
DataGridComponent.ctorParameters = () => [
    { type: DomSanitizer },
    { type: ElementRef },
    { type: ChangeDetectorRef },
    { type: ComponentFactoryResolver }
];
DataGridComponent.propDecorators = {
    title: [{ type: Input }],
    loadMoreItemsLabel: [{ type: Input }],
    loadingItemsLabel: [{ type: Input }],
    showSearch: [{ type: Input }],
    refresh: [{ type: Input }],
    _columns: [{ type: Input, args: ['columns',] }],
    _rows: [{ type: Input, args: ['rows',] }],
    _pagination: [{ type: Input, args: ['pagination',] }],
    _infiniteScroll: [{ type: Input, args: ['infiniteScroll',] }],
    _serverSideDataCallback: [{ type: Input, args: ['serverSideDataCallback',] }],
    _selectable: [{ type: Input, args: ['selectable',] }],
    _selectionPrimaryKey: [{ type: Input, args: ['selectionPrimaryKey',] }],
    _displayOptions: [{ type: Input, args: ['displayOptions',] }],
    _actionControls: [{ type: Input, args: ['actionControls',] }],
    _bulkActionControls: [{ type: Input, args: ['bulkActionControls',] }],
    _headerActionControls: [{ type: Input, args: ['headerActionControls',] }],
    searchText: [{ type: Input }],
    showCounterWarning: [{ type: Input }],
    rowMouseOver: [{ type: Output }],
    rowMouseLeave: [{ type: Output }],
    rowClick: [{ type: Output }],
    onConfigChange: [{ type: Output }],
    onFilter: [{ type: Output }],
    itemsSelect: [{ type: Output }],
    onReload: [{ type: Output }],
    columnRenderers: [{ type: ContentChildren, args: [ColumnDirective,] }],
    scrollContainer: [{ type: ViewChild, args: ['scroll', { static: true },] }],
    infiniteScrollContainer: [{ type: ViewChild, args: ['infiniteScrollContainer', { static: false, read: ViewContainerRef },] }],
    thRefs: [{ type: ViewChildren, args: [CdkHeaderCell, { read: ElementRef },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DataGridComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-data-grid',
                template: "<div\n  #scroll\n  class=\"table-data-grid-scroll\"\n  [ngClass]=\"{ 'table-data-grid__overlay': (dataSource.loading$ | async) && !loadMoreComponent }\"\n>\n  <div  class=\"table-data-grid__loading--wrapper\" *ngIf=\"(dataSource.loading$ | async) && !loadMoreComponent\">\n    <div class=\"table-data-grid__loading--loader \">\n      <c8y-progress-bar [message]=\"loadingItemsLabel\"></c8y-progress-bar>\n    </div>\n  </div>\n\n  <div *ngIf=\"displayOptions.gridHeader\" class=\"table-data-grid-header separator large-padding\">\n    <h4 [ngClass]=\"{ 'm-r-16': !!title }\">{{ title | translate }}</h4>\n\n    <ng-container *ngIf=\"displayOptions.filter\">\n      <span *ngIf=\"!filteringApplied\">\n        <span class=\"label label-default m-r-4\" translate>No filters</span>\n        <small\n          *ngIf=\"!!filteringLabelsParams.allItemsCount\"\n          class=\"m-r-4\"\n          ngNonBindable\n          translate\n          [translateParams]=\"filteringLabelsParams\"\n          >{{ filteredItemsCount }} of {{ allItemsCount }} items</small\n        >\n      </span>\n      <span *ngIf=\"filteringApplied\">\n        <ng-container *ngIf=\"!!filteringLabelsParams.allItemsCount\">\n          <span class=\"badge badge-info m-r-4\">\n            {{ (dataSource.stats$ | async).filteredSize }}\n          </span>\n          <small ngNonBindable translate [translateParams]=\"filteringLabelsParams\"\n            >of {{ allItemsCount }} items</small\n          >\n        </ng-container>\n        <button\n          title=\" {{ 'Clear filters' | translate }}\"\n          class=\"m-l-8 btn btn-xs btn-default m-r-4\"\n          (click)=\"clearFilters()\"\n        >\n          {{ 'Clear filters' | translate }}\n        </button>\n      </span>\n      <span class=\"hidden-xs hidden-sm\">\n        <button\n          class=\"btn-clean text-primary\"\n          popover=\"{{ 'Click the column headers to apply filters.' | translate }}\"\n          placement=\"right\"\n          triggers=\"focus\"\n        >\n          <i c8yIcon=\"question-circle-o\"></i>\n        </button>\n      </span>\n\n      <span class=\"hidden-xs hidden-sm\" *ngIf=\"showCounterWarning\">\n        <button\n          class=\"btn-clean text-primary\"\n          popover=\"{{\n            'The counter for the total number of items might be inaccurate.' | translate\n          }}\"\n          placement=\"right\"\n          triggers=\"focus\"\n        >\n          <i c8yIcon=\"warning\"></i>\n        </button>\n      </span>\n    </ng-container>\n\n    <div class=\"flex-item-right\">\n      <div class=\"btnbar flex-row\">\n        <ng-container *ngFor=\"let headerActionControl of headerActionControls\">\n          <ng-container *ngIf=\"headerActionControl | isControlVisible\">\n            <ng-container *ngIf=\"!headerActionControl.template; else customTemplate\">\n              <button\n                title=\"{{ headerActionControl.text | translate }}\"\n                class=\"btnbar-btn btn-link\"\n                (click)=\"headerActionControl.callback()\"\n              >\n                <i c8yIcon=\"headerActionControl.icon\" class=\"m-r-4\"></i>\n                <span>{{ headerActionControl.text | translate }}</span>\n              </button>\n            </ng-container>\n            <ng-template #customTemplate>\n              <ng-container\n                *ngTemplateOutlet=\"\n                  headerActionControl.template;\n                  context: { headerActionControl: headerActionControl }\n                \"\n              ></ng-container>\n            </ng-template>\n          </ng-container>\n        </ng-container>\n\n        <div\n          class=\"dropdown\"\n          dropdown\n          [insideClick]=\"true\"\n          container=\"body\"\n          placement=\"bottom left\"\n        >\n          <button\n            title=\"{{ 'Configure columns' | translate }}\"\n            class=\"btnbar-btn c8y-dropdown\"\n            dropdownToggle\n          >\n            <i c8yIcon=\"columns\" class=\"m-r-4\"></i>\n            <span>{{ 'Configure columns' | translate }}</span>\n          </button>\n\n          <ul *dropdownMenu class=\"dropdown-menu\" (click)=\"$event.stopPropagation()\">\n            <li class=\"bg-white\">\n              <div cdkDropList (cdkDropListDropped)=\"onColumnDrop($event)\" class=\"list-group m-0\">\n                <div *ngFor=\"let column of columns\" cdkDrag cdkDragLockAxis=\"y\">\n                  <ng-container *ngIf=\"!column.positionFixed\">\n                    <div class=\"list-group-item draggable-after p-0\">\n                      <label\n                        title=\"{{ (column.header | translate) || column.name }}\"\n                        class=\"c8y-checkbox p-l-16 p-r-32 p-t-8 p-b-8 flex-item-v-stretch\"\n                      >\n                        <input\n                          type=\"checkbox\"\n                          [(ngModel)]=\"column.visible\"\n                          (change)=\"updateGridColumnsSize()\"\n                        />\n                        <span></span>\n                        <span>{{ (column.header | translate) || column.name }}</span>\n                      </label>\n                    </div>\n                  </ng-container>\n                </div>\n              </div>\n            </li>\n          </ul>\n        </div>\n\n        <button\n          title=\"{{ 'Reload' | translate }}\"\n          class=\"btnbar-btn btn-link\"\n          [disabled]=\"dataSource.loading$ | async\"\n          (click)=\"clickReload()\"\n        >\n          <i c8yIcon=\"refresh\" class=\"m-r-4\"></i>\n          <span>{{ 'Reload' | translate }}</span>\n        </button>\n\n        <div\n          *ngIf=\"!serverSideDataCallback || showSearch\"\n          class=\"input-group input-group-search m-l-sm-16\"\n        >\n          <input\n            type=\"search\"\n            class=\"form-control\"\n            placeholder=\"{{ 'Search\u2026' | translate }}\"\n            [ngModel]=\"searchText\"\n            (input)=\"searchText$.emit($event.target.value)\"\n          />\n          <div class=\"input-group-addon\">\n            <i c8yIcon=\"search\" *ngIf=\"searchText.length === 0\"></i>\n            <i\n              c8yIcon=\"times\"\n              class=\"pointer\"\n              *ngIf=\"searchText.length > 0\"\n              (click)=\"searchText = ''; searchText$.emit('')\"\n            ></i>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div\n      class=\"table-data-grid-header-bulk-actions animated slideInDown fast\"\n      *ngIf=\"selectedItemIds.length !== 0\"\n    >\n      <h4>\n        <span ngNonBindable translate [translateParams]=\"{ count: selectedItemIds.length }\">\n          {{ count }} selected items.\n        </span>\n        <br class=\"visible-xs\" />\n        <small *ngIf=\"selectedItemIds.length >= pagination.pageSize\">\n          <a (click)=\"setAllItemsSelected(true)\" class=\"interact\">\n            <span\n              ngNonBindable\n              translate\n              [translateParams]=\"{ count: (dataSource.stats$ | async).filteredSize }\"\n            >\n              Select all {{ count }} items\n            </span>\n          </a>\n        </small>\n      </h4>\n      <div class=\"flex-item-right\">\n        <div class=\"btnbar flex-row\">\n          <ng-container\n            *ngFor=\"let bulkActionControl of bulkActionControls\"\n            [ngSwitch]=\"bulkActionControl.type\"\n          >\n            <ng-container *ngIf=\"bulkActionControl | isControlVisible: selectedItemIds\">\n              <button\n                title=\"{{ 'Export' | translate }}\"\n                *ngSwitchCase=\"builtInActionType.Export\"\n                class=\"btnbar-btn\"\n                (click)=\"bulkActionControl.callback(selectedItemIds, reload.bind(this))\"\n              >\n                <i c8yIcon=\"sign-out\"></i>\n                <span>{{ 'Export' | translate }}</span>\n              </button>\n\n              <button\n                title=\"{{ 'Delete' | translate }}\"\n                *ngSwitchCase=\"builtInActionType.Delete\"\n                class=\"btnbar-btn\"\n                (click)=\"bulkActionControl.callback(selectedItemIds, reload.bind(this))\"\n              >\n                <i c8yIcon=\"trash\"></i>\n                <span>{{ 'Delete' | translate }}</span>\n              </button>\n\n              <button\n                title=\"{{ bulkActionControl.text | translate }}\"\n                *ngSwitchDefault\n                class=\"btnbar-btn\"\n                (click)=\"bulkActionControl.callback(selectedItemIds, reload.bind(this))\"\n              >\n                <i c8yIcon=\"{{ bulkActionControl.icon }}\"></i>\n                <span>{{ bulkActionControl.text | translate }}</span>\n              </button>\n            </ng-container>\n          </ng-container>\n\n          <button title=\"{{ 'Cancel' | translate }}\" class=\"btnbar-btn\" (click)=\"cancel()\">\n            <i c8yIcon=\"times\"></i>\n            <span>{{ 'Cancel' | translate }}</span>\n          </button>\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <table\n    class=\"table table-hover table-filtered-sorted table-data-grid large-padding\"\n    cdk-table\n    [dataSource]=\"dataSource\"\n    [trackBy]=\"trackByName\"\n    [class.table-striped]=\"displayOptions.striped\"\n    [class.table-bordered]=\"displayOptions.bordered\"\n    [class.table-data-grid-with-checkboxes]=\"selectable\"\n    [class.table-data-grid-with-actions]=\"actionControls.length > 0\"\n    [style.grid-template-columns]=\"styles.gridTemplateColumns\"\n    (mousemove)=\"resizeHandleContainerMouseMove$.emit($event)\"\n    (mouseup)=\"resizeHandleContainerMouseUp$.emit($event)\"\n  >\n    <ng-container *ngFor=\"let column of columns; trackBy: trackByName\" [cdkColumnDef]=\"column.name\">\n      <ng-container [ngSwitch]=\"column.name\">\n        <ng-container *ngSwitchCase=\"'checkbox'\">\n          <th cdk-header-cell *cdkHeaderCellDef data-type=\"icon\">\n            <div>\n              <label class=\"c8y-checkbox\">\n                <input\n                  type=\"checkbox\"\n                  [checked]=\"currentPageSelectionState.allSelected\"\n                  [indeterminate]=\"\n                    !(\n                      currentPageSelectionState.allSelected ||\n                      currentPageSelectionState.allDeselected\n                    )\n                  \"\n                  (change)=\"setAllItemsInCurrentPageSelected($event.target.checked)\"\n                />\n                <span></span>\n              </label>\n            </div>\n          </th>\n\n          <td cdk-cell *cdkCellDef=\"let row\" data-type=\"icon\">\n            <label class=\"c8y-checkbox\">\n              <input\n                type=\"checkbox\"\n                [checked]=\"isItemSelected(row)\"\n                (change)=\"setItemsSelected([row], $event.target.checked)\"\n              />\n              <span></span>\n            </label>\n          </td>\n        </ng-container>\n\n        <ng-container *ngSwitchCase=\"'actions'\">\n          <th cdk-header-cell *cdkHeaderCellDef data-type=\"icon\"></th>\n\n          <td cdk-cell *cdkCellDef=\"let row\" data-type=\"icon\">\n            <ng-container *ngIf=\"actionControls.length <= 2\">\n              <ng-container\n                *ngFor=\"let actionControl of actionControls\"\n                [ngSwitch]=\"actionControl.type\"\n              >\n                <ng-container *ngIf=\"actionControl.showIf ? actionControl.showIf(row) : true\">\n                  <button\n                    *ngSwitchCase=\"builtInActionType.Edit\"\n                    class=\"btn btn-icon btn-xs btn-default\"\n                    title=\"{{ 'Edit' | translate }}\"\n                    (click)=\"actionControl.callback(row, reload.bind(this))\"\n                  >\n                    <i c8yIcon=\"pencil\"></i>\n                  </button>\n                  <button\n                    *ngSwitchCase=\"builtInActionType.Delete\"\n                    class=\"btn btn-dot showOnHover\"\n                    title=\"{{ 'Delete' | translate }}\"\n                    (click)=\"actionControl.callback(row, reload.bind(this))\"\n                  >\n                    <i class=\"text-danger\" c8yIcon=\"minus-circle\"></i>\n                  </button>\n                  <button\n                    *ngSwitchDefault\n                    class=\"btn btn-icon btn-xs btn-default\"\n                    title=\"{{ actionControl.text | translate }}\"\n                    (click)=\"actionControl.callback(row, reload.bind(this))\"\n                    tooltip=\"{{ actionControl.text | translate }}\"\n                    container=\"body\"\n                  >\n                    <i c8yIcon=\"{{ actionControl.icon }}\"></i>\n                  </button>\n                </ng-container>\n              </ng-container>\n            </ng-container>\n\n            <div [ngClass]=\"{ 'm-l-auto overflow-visible': actionControls.length > 2 }\">\n              <div *ngIf=\"actionControls.length > 2\" class=\"dropdown\" dropdown container=\"body\" placement=\"bottom right\">\n                <button\n                  title=\"{{ 'Actions' | translate }}\"\n                  class=\"dropdown-toggle c8y-dropdown\"\n                  dropdownToggle\n                >\n                  <i c8yIcon=\"ellipsis-v\"></i>\n                </button>\n                <ul class=\"dropdown-menu dropdown-menu-right\" *dropdownMenu>\n                  <li *ngFor=\"let actionControl of actionControls\" [ngSwitch]=\"actionControl.type\">\n                    <ng-container *ngIf=\"actionControl | isControlVisible: row\">\n                      <button\n                        title=\"{{ 'Edit' | translate }}\"\n                        *ngSwitchCase=\"builtInActionType.Edit\"\n                        (click)=\"actionControl.callback(row, reload.bind(this))\"\n                      >\n                        <i c8yIcon=\"pencil\"></i>\n                        {{ 'Edit' | translate }}\n                      </button>\n                      <button\n                        title=\"{{ 'Delete' | translate }}\"\n                        *ngSwitchCase=\"builtInActionType.Delete\"\n                        (click)=\"actionControl.callback(row, reload.bind(this))\"\n                      >\n                        <i c8yIcon=\"trash\"></i>\n                        {{ 'Delete' | translate }}\n                      </button>\n                      <button\n                        title=\"{{ 'Export' | translate }}\"\n                        *ngSwitchCase=\"builtInActionType.Export\"\n                        (click)=\"actionControl.callback(row, reload.bind(this))\"\n                      >\n                        <i c8yIcon=\"download\"></i>\n                        {{ 'Export' | translate }}\n                      </button>\n                      <button\n                        title=\"{{ actionControl.text | translate }}\"\n                        *ngSwitchDefault\n                        (click)=\"actionControl.callback(row, reload.bind(this))\"\n                      >\n                        <i c8yIcon=\"{{ actionControl.icon }}\"></i>\n                        {{ actionControl.text | translate }}\n                      </button>\n                    </ng-container>\n                  </li>\n                </ul>\n              </div>\n            </div>\n          </td>\n        </ng-container>\n\n        <ng-container *ngSwitchDefault>\n          <th\n            cdk-header-cell\n            *cdkHeaderCellDef\n            [ngClass]=\"column.headerCSSClassName\"\n            [attr.data-type]=\"column.dataType\"\n            [class.sorted]=\"column.sortOrder\"\n            [class.filtered]=\"column | map: isColumnFilteringApplied\"\n            [class.hidden]=\"!column.visible\"\n          >\n            <div *ngIf=\"!column.filterable\" [title]=\"(column.header | translate) || column.name\">\n              <ng-container\n                *ngIf=\"\n                  [\n                    {\n                      columnName: column.name,\n                      value: (column.header | translate) || column.name\n                    }\n                  ] | map: getHeaderCellRendererSpec:this as cellRendererSpec\n                \"\n              >\n                <c8y-cell-renderer [spec]=\"cellRendererSpec\"></c8y-cell-renderer>\n              </ng-container>\n            </div>\n\n            <div\n              *ngIf=\"column.filterable\"\n              class=\"dropdown\"\n              dropdown\n              #dropdown=\"bs-dropdown\"\n              container=\"body\"\n              placement=\"bottom left\"\n              [insideClick]=\"true\"\n            >\n              <button\n                class=\"btn-header c8y-dropdown\"\n                dropdownToggle\n                [title]=\"(column.header | translate) || column.name\"\n              >\n                <ng-container\n                  *ngIf=\"\n                    [\n                      {\n                        columnName: column.name,\n                        value: (column.header | translate) || column.name\n                      }\n                    ] | map: getHeaderCellRendererSpec:this as cellRendererSpec\n                  \"\n                >\n                  <c8y-cell-renderer [spec]=\"cellRendererSpec\"></c8y-cell-renderer>\n                </ng-container>\n                <i c8yIcon=\"filter\" title=\"{{ 'Filter' | translate }}\"></i>\n              </button>\n\n              <ul *dropdownMenu class=\"dropdown-menu\" (click)=\"$event.stopPropagation()\">\n                <li class=\"data-grid__dropdown\">\n                  <ng-container\n                    *ngIf=\"\n                      [\n                        {\n                          column: column,\n                          dropdown: dropdown\n                        }\n                      ] | map: getFilteringFormRendererSpec:this as filteringFormRendererSpec\n                    \"\n                  >\n                    <c8y-filtering-form-renderer [spec]=\"filteringFormRendererSpec\">\n                    </c8y-filtering-form-renderer>\n                  </ng-container>\n                </li>\n              </ul>\n            </div>\n\n            <button\n              *ngIf=\"column.sortable\"\n              class=\"btn-sort\"\n              title=\"{{ 'Sort' | translate }}\"\n              (click)=\"changeSortOrder(column.name)\"\n            >\n              <ng-container [ngSwitch]=\"column.sortOrder\">\n                <i *ngSwitchCase=\"'asc'\" c8yIcon=\"long-arrow-up\"></i>\n                <i *ngSwitchCase=\"'desc'\" c8yIcon=\"long-arrow-down\"></i>\n                <i *ngSwitchDefault c8yIcon=\"exchange\"></i>\n              </ng-container>\n            </button>\n\n            <span\n              class=\"resize-handle\"\n              (mousedown)=\"\n                resizeHandleMouseDown$.emit({ event: $event, targetColumnName: column.name })\n              \"\n            >\n            </span>\n          </th>\n\n          <td\n            cdk-cell\n            *cdkCellDef=\"let row\"\n            [ngClass]=\"column.cellCSSClassName\"\n            [attr.data-cell-title]=\"column.header | translate\"\n            [attr.data-type]=\"column.dataType\"\n            [class.hidden]=\"!column.visible\"\n          >\n            <ng-container\n              *ngIf=\"\n                [\n                  {\n                    value: resolveCellValue(row, column.path),\n                    row: row,\n                    columnName: column.name\n                  }\n                ] | map: getCellRendererSpec:this as cellRendererSpec\n              \"\n            >\n              <c8y-cell-renderer [spec]=\"cellRendererSpec\"></c8y-cell-renderer>\n            </ng-container>\n          </td>\n        </ng-container>\n      </ng-container>\n    </ng-container>\n\n    <ng-container cdkColumnDef=\"infiniteScrollFooter\">\n      <td cdk-footer-cell *cdkFooterCellDef [style.grid-column]=\"styles.gridInfiniteScrollColumn\">\n        <template #infiniteScrollContainer></template>\n      </td>\n    </ng-container>\n\n    <tr cdk-header-row *cdkHeaderRowDef=\"columnNames\"></tr>\n\n    <tr\n      cdk-row\n      *cdkRowDef=\"let row; columns: columnNames\"\n      (mouseover)=\"rowMouseOver.emit(row)\"\n      (mouseleave)=\"rowMouseLeave.emit(row)\"\n      (click)=\"rowClick.emit(row)\"\n    ></tr>\n\n    <ng-container *ngIf=\"infiniteScroll\">\n      <tr cdk-footer-row *cdkFooterRowDef=\"['infiniteScrollFooter']\"></tr>\n    </ng-container>\n  </table>\n\n  <div\n    class=\"row m-0 p-t-40 p-b-40\"\n    *ngIf=\"!(dataSource.loading$ | async) && (dataSource.stats$ | async).filteredSize === 0\"\n  >\n    <div class=\"col-md-4 col-md-offset-4\">\n      <ng-content select=\".c8y-empty-state\"></ng-content>\n    </div>\n  </div>\n\n  <div *ngIf=\"pagination && !infiniteScroll\" class=\"table-data-grid-footer separator large-padding\">\n    <div class=\"col-sm-4 no-gutter\">\n      <div *ngIf=\"(dataSource.stats$ | async).currentPageSize > 0\" class=\"counter\">\n        <span class=\"text-muted\" ngNonBindable translate [translateParams]=\"paginationLabelParams\"\n          >{{ pageFirstItemIdx }} - {{ pageLastItemIdx }} of {{ itemsTotal }}</span\n        >\n      </div>\n    </div>\n\n    <div class=\"col-sm-4 no-gutter text-center\">\n      <div\n        *ngIf=\"(dataSource.stats$ | async).filteredSize > minPossiblePageSize\"\n        class=\"form-group form-inline\"\n      >\n        <label class=\"m-r-4\">{{ 'Items per page' | translate }}</label>\n        <div class=\"c8y-select-wrapper\">\n          <select\n            class=\"form-control\"\n            [ngModel]=\"pagination.pageSize\"\n            (ngModelChange)=\"\n              updatePagination({ itemsPerPage: $event, page: pagination.currentPage })\n            \"\n          >\n            <option *ngFor=\"let pageSize of possiblePageSizes\" [ngValue]=\"pageSize\">\n              {{ pageSize }}\n            </option>\n          </select>\n        </div>\n      </div>\n    </div>\n\n    <div class=\"col-sm-4 no-gutter text-right\">\n      <pagination\n        [class.hidden]=\"hidePagination$ | async\"\n        [ngModel]=\"pagination.currentPage\"\n        (pageChanged)=\"updatePagination($event)\"\n        [totalItems]=\"(dataSource.stats$ | async).filteredSize\"\n        [itemsPerPage]=\"pagination.pageSize\"\n        (numPages)=\"totalPagesCount$.next($event)\"\n        [maxSize]=\"5\"\n        [boundaryLinks]=\"false\"\n        previousText=\" \"\n        nextText=\" \"\n      ></pagination>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.DomSanitizer }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.ComponentFactoryResolver }]; }, { title: [{
            type: Input
        }], loadMoreItemsLabel: [{
            type: Input
        }], loadingItemsLabel: [{
            type: Input
        }], showSearch: [{
            type: Input
        }], searchText: [{
            type: Input
        }], showCounterWarning: [{
            type: Input
        }], rowMouseOver: [{
            type: Output
        }], rowMouseLeave: [{
            type: Output
        }], rowClick: [{
            type: Output
        }], onConfigChange: [{
            type: Output
        }], onFilter: [{
            type: Output
        }], itemsSelect: [{
            type: Output
        }], onReload: [{
            type: Output
        }], _columns: [{
            type: Input,
            args: ['columns']
        }], _rows: [{
            type: Input,
            args: ['rows']
        }], _pagination: [{
            type: Input,
            args: ['pagination']
        }], _infiniteScroll: [{
            type: Input,
            args: ['infiniteScroll']
        }], _serverSideDataCallback: [{
            type: Input,
            args: ['serverSideDataCallback']
        }], _selectable: [{
            type: Input,
            args: ['selectable']
        }], _selectionPrimaryKey: [{
            type: Input,
            args: ['selectionPrimaryKey']
        }], _displayOptions: [{
            type: Input,
            args: ['displayOptions']
        }], _actionControls: [{
            type: Input,
            args: ['actionControls']
        }], _bulkActionControls: [{
            type: Input,
            args: ['bulkActionControls']
        }], _headerActionControls: [{
            type: Input,
            args: ['headerActionControls']
        }], refresh: [{
            type: Input
        }], columnRenderers: [{
            type: ContentChildren,
            args: [ColumnDirective]
        }], scrollContainer: [{
            type: ViewChild,
            args: ['scroll', { static: true }]
        }], infiniteScrollContainer: [{
            type: ViewChild,
            args: ['infiniteScrollContainer', { static: false, read: ViewContainerRef }]
        }], thRefs: [{
            type: ViewChildren,
            args: [CdkHeaderCell, { read: ElementRef }]
        }] }); })();
