// @c8y/ngx-components/esm2015/core/dynamic-forms/dynamic-forms.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule } from '@ngx-formly/core';
import { DynamicFormsUIThemeModule } from './dynamic-forms-ui-theme.module';
import { I18nModule } from './i18n/i18n.module';
import { C8yJSONSchema } from './json-schema/c8y-json-schema.service';
import { ValidationModule } from './validation/validation.module';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-formly/core';
export class DynamicFormsModule {
}
DynamicFormsModule.ɵfac = function DynamicFormsModule_Factory(t) { return new (t || DynamicFormsModule)(); };
DynamicFormsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicFormsModule });
DynamicFormsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [C8yJSONSchema], imports: [[
            CommonModule,
            ReactiveFormsModule,
            I18nModule,
            ValidationModule,
            DynamicFormsUIThemeModule,
            FormlyModule.forRoot()
        ], FormlyModule, ReactiveFormsModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicFormsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    I18nModule,
                    ValidationModule,
                    DynamicFormsUIThemeModule,
                    FormlyModule.forRoot()
                ],
                providers: [C8yJSONSchema],
                exports: [FormlyModule, ReactiveFormsModule]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicFormsModule, { imports: function () { return [CommonModule,
        ReactiveFormsModule,
        I18nModule,
        ValidationModule,
        DynamicFormsUIThemeModule, ɵngcc1.FormlyModule]; }, exports: function () { return [FormlyModule, ReactiveFormsModule]; } }); })();
