// @c8y/ngx-components/esm2015/ecosystem/activity-log/activity-log.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { AlertService } from '@c8y/ngx-components';
import { EcosystemService } from '../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../shared/upload-archive.component';

function ActivityLogComponent_c8y_li_timeline_3_small_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 17);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const archive_r2 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(archive_r2.description);
} }
function ActivityLogComponent_c8y_li_timeline_3_i_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 18);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Activating"));
    ɵngcc0.ɵɵproperty("c8yIcon", "circle-o-notch");
} }
function ActivityLogComponent_c8y_li_timeline_3_span_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 19);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, "Active"));
} }
function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_14_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 20);
    ɵngcc0.ɵɵlistener("click", function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_14_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r12); const archive_r2 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.setActive(archive_r2); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Set as active`archive`"), " ");
} }
function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_18_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 21);
    ɵngcc0.ɵɵlistener("click", function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_18_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r15); const archive_r2 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.deleteArchive(archive_r2); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Delete`archive`"), " ");
} }
function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_19_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 22);
    ɵngcc0.ɵɵlistener("click", function ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_19_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(2); return ctx_r16.reactivateArchive(); });
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Reactivate archive"), " ");
} }
const _c0 = function (a0) { return { active: a0 }; };
function ActivityLogComponent_c8y_li_timeline_3_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-timeline", 5);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "date");
    ɵngcc0.ɵɵpipe(3, "date");
    ɵngcc0.ɵɵelementStart(4, "c8y-li");
    ɵngcc0.ɵɵelement(5, "c8y-li-icon", 6);
    ɵngcc0.ɵɵelementStart(6, "c8y-li-body");
    ɵngcc0.ɵɵelementStart(7, "div", 7);
    ɵngcc0.ɵɵelementStart(8, "div", 8);
    ɵngcc0.ɵɵelementStart(9, "span", 9);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(11, ActivityLogComponent_c8y_li_timeline_3_small_11_Template, 2, 1, "small", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(12, ActivityLogComponent_c8y_li_timeline_3_i_12_Template, 2, 4, "i", 11);
    ɵngcc0.ɵɵtemplate(13, ActivityLogComponent_c8y_li_timeline_3_span_13_Template, 3, 3, "span", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(14, ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_14_Template, 3, 3, "c8y-li-action", 13);
    ɵngcc0.ɵɵelementStart(15, "c8y-li-action", 14);
    ɵngcc0.ɵɵlistener("click", function ActivityLogComponent_c8y_li_timeline_3_Template_c8y_li_action_click_15_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r19); const archive_r2 = restoredCtx.$implicit; const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.downloadArchive(archive_r2); });
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(18, ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_18_Template, 3, 3, "c8y-li-action", 15);
    ɵngcc0.ɵɵtemplate(19, ActivityLogComponent_c8y_li_timeline_3_c8y_li_action_19_Template, 3, 3, "c8y-li-action", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const archive_r2 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(21, _c0, ctx_r0.isActive(archive_r2)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate2(" ", ɵngcc0.ɵɵpipeBind2(2, 13, archive_r2.created, "d MMM YYYY"), " ", ɵngcc0.ɵɵpipeBind2(3, 16, archive_r2.created, "shortTime"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("icon", ctx_r0.checkIfLast(archive_r2) ? "flag-checkered" : "file-zip-o");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", archive_r2.description || archive_r2.name, "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", archive_r2.description || archive_r2.name, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", archive_r2.description);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isLoading && ctx_r0.toActivate(archive_r2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isActive(archive_r2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.isLoading && !ctx_r0.isActive(archive_r2));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 19, "Download`archive`"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.archives.length > 1 && !ctx_r0.checkIfLast(archive_r2) && !ctx_r0.isActive(archive_r2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.canReactivate && ctx_r0.isActive(archive_r2));
} }
function ActivityLogComponent_c8y_form_group_5_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 23);
    ɵngcc0.ɵɵelementStart(1, "c8y-upload-archive", 24);
    ɵngcc0.ɵɵlistener("applicationChange", function ActivityLogComponent_c8y_form_group_5_Template_c8y_upload_archive_applicationChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(); return ctx_r20.application = $event; })("refresh", function ActivityLogComponent_c8y_form_group_5_Template_c8y_upload_archive_refresh_1_listener() { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.onRefresh(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("application", ctx_r1.application);
} }
export class ActivityLogComponent {
    constructor(ecosystemService, alertService) {
        this.ecosystemService = ecosystemService;
        this.alertService = alertService;
        this.archives = [];
        this.canReactivate = false;
    }
    get uploadProgress() {
        return this.ecosystemService.progress;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.canReactivate = this.showReactivate();
            this.refresh();
        });
    }
    isActive(archive) {
        return this.application.activeVersionId === archive.id;
    }
    toActivate(archive) {
        return this.toActivateVersionId === archive.id;
    }
    checkIfLast(archive) {
        return archive.id === this.last.id;
    }
    showReactivate() {
        return this.ecosystemService.isApplication(this.application);
    }
    setActive(archive) {
        return __awaiter(this, void 0, void 0, function* () {
            const id = archive.id || archive;
            this.toActivateVersionId = id;
            this.isLoading = true;
            try {
                this.application = (yield this.ecosystemService.setAppActiveVersion(this.application, id)).data;
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
            this.isLoading = false;
            this.refresh();
        });
    }
    deleteArchive(archive) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.deleteArchive(archive, this.application);
            this.refresh();
        });
    }
    downloadArchive(archive) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.downloadArchive(this.application, archive);
        });
    }
    reactivateArchive() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.reactivateArchive(this.application);
        });
    }
    onRefresh() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.refresh();
        });
    }
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.archives = yield this.ecosystemService.listArchives(this.application.id);
            this.archives.sort((a, b) => {
                return new Date(b.created) - new Date(a.created);
            });
            this.last = this.archives[this.archives.length - 1];
            this.isLoading = false;
        });
    }
}
ActivityLogComponent.ɵfac = function ActivityLogComponent_Factory(t) { return new (t || ActivityLogComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
ActivityLogComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActivityLogComponent, selectors: [["c8y-activity-log"]], inputs: { application: "application" }, decls: 6, vars: 2, consts: [[1, "inner-scroll", "bg-gray-white", "flex-grow"], [1, "card-block", "overflow-visible"], [3, "ngClass", 4, "ngFor", "ngForOf"], [1, "card-footer"], ["class", "m-auto", 4, "ngIf"], [3, "ngClass"], [3, "icon"], [1, "d-flex", "a-i-start"], [2, "min-width", "0", "flex", "1"], [1, "text-truncate-wrap", 3, "title"], ["class", "text-muted", 4, "ngIf"], ["class", "icon-spin", 3, "c8yIcon", "title", 4, "ngIf"], ["class", "label label-primary m-l-auto m-t-4", 4, "ngIf"], ["icon", "check-square-o", 3, "click", 4, "ngIf"], ["icon", "download", 3, "click"], ["icon", "trash", 3, "click", 4, "ngIf"], ["icon", "undo", 3, "click", 4, "ngIf"], [1, "text-muted"], [1, "icon-spin", 3, "c8yIcon", "title"], [1, "label", "label-primary", "m-l-auto", "m-t-4"], ["icon", "check-square-o", 3, "click"], ["icon", "trash", 3, "click"], ["icon", "undo", 3, "click"], [1, "m-auto"], [3, "application", "applicationChange", "refresh"]], template: function ActivityLogComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "c8y-list-group");
        ɵngcc0.ɵɵtemplate(3, ActivityLogComponent_c8y_li_timeline_3_Template, 20, 23, "c8y-li-timeline", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵtemplate(5, ActivityLogComponent_c8y_form_group_5_Template, 2, 1, "c8y-form-group", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.archives);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
    } }, directives: [ɵngcc2.ListGroupComponent, ɵngcc3.NgForOf, ɵngcc3.NgIf, ɵngcc2.ListItemTimelineComponent, ɵngcc3.NgClass, ɵngcc2.ListItemComponent, ɵngcc2.ListItemIconComponent, ɵngcc2.ListItemBodyComponent, ɵngcc2.ListItemActionComponent, ɵngcc2.IconDirective, ɵngcc2.FormGroupComponent, ɵngcc4.UploadArchiveComponent], pipes: [ɵngcc3.DatePipe, ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ActivityLogComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: AlertService }
];
ActivityLogComponent.propDecorators = {
    application: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActivityLogComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-activity-log',
                template: "<div class=\"inner-scroll bg-gray-white flex-grow\">\n  <div class=\"card-block overflow-visible\">\n    <c8y-list-group>\n      <c8y-li-timeline *ngFor=\"let archive of archives\" [ngClass]=\"{ active: isActive(archive) }\">\n        {{ archive.created | date: 'd MMM YYYY' }}\n        {{ archive.created | date: 'shortTime' }}\n        <c8y-li>\n          <c8y-li-icon\n            [icon]=\"checkIfLast(archive) ? 'flag-checkered' : 'file-zip-o'\"\n          ></c8y-li-icon>\n          <c8y-li-body>\n            <div class=\"d-flex a-i-start\">\n              <div style=\"min-width: 0; flex: 1\">\n                <span class=\"text-truncate-wrap\" title=\" {{ archive.description || archive.name }}\">\n                  {{ archive.description || archive.name }}\n                </span>\n                <small *ngIf=\"archive.description\" class=\"text-muted\">{{\n                  archive.description\n                }}</small>\n              </div>\n              <i\n                *ngIf=\"isLoading && toActivate(archive)\"\n                [c8yIcon]=\"'circle-o-notch'\"\n                class=\"icon-spin\"\n                title=\"{{ 'Activating' | translate }}\"\n              ></i>\n\n              <span *ngIf=\"isActive(archive)\" class=\"label label-primary m-l-auto m-t-4\">{{\n                'Active' | translate\n              }}</span>\n            </div>\n          </c8y-li-body>\n          <c8y-li-action\n            (click)=\"setActive(archive)\"\n            *ngIf=\"!isLoading && !isActive(archive)\"\n            icon=\"check-square-o\"\n          >\n            {{ 'Set as active`archive`' | translate }}\n          </c8y-li-action>\n          <c8y-li-action (click)=\"downloadArchive(archive)\" icon=\"download\">\n            {{ 'Download`archive`' | translate }}\n          </c8y-li-action>\n          <c8y-li-action\n            (click)=\"deleteArchive(archive)\"\n            *ngIf=\"archives.length > 1 && !checkIfLast(archive) && !isActive(archive)\"\n            icon=\"trash\"\n          >\n            {{ 'Delete`archive`' | translate }}\n          </c8y-li-action>\n          <c8y-li-action\n            (click)=\"reactivateArchive()\"\n            *ngIf=\"canReactivate && isActive(archive)\"\n            icon=\"undo\"\n          >\n            {{ 'Reactivate archive' | translate }}\n          </c8y-li-action>\n        </c8y-li>\n      </c8y-li-timeline>\n    </c8y-list-group>\n  </div>\n</div>\n<div class=\"card-footer\">\n  <c8y-form-group *ngIf=\"!isLoading\" class=\"m-auto\">\n    <c8y-upload-archive [(application)]=\"application\" (refresh)=\"onRefresh()\"></c8y-upload-archive>\n  </c8y-form-group>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.AlertService }]; }, { application: [{
            type: Input
        }] }); })();
