// @c8y/ngx-components/esm2015/ecosystem/applications/duplicate-application/duplicate-application.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { C8yStepper, WizardComponent } from '@c8y/ngx-components';
import { BehaviorSubject } from 'rxjs';
import { EcosystemService } from './../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './../../ecosystem.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/cdk/stepper';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from './duplicate-application-list/duplicate-application-list.component';
import * as ɵngcc7 from './duplicate-application-properties/duplicate-application-properties.component';

function DuplicateApplicationComponent_c8y_duplicate_application_list_10_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-duplicate-application-list", 7);
    ɵngcc0.ɵɵlistener("onSelectedApp", function DuplicateApplicationComponent_c8y_duplicate_application_list_10_Template_c8y_duplicate_application_list_onSelectedApp_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.onSelectedApp($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("existingApps", ctx_r0.existingApps)("stepper", ctx_r0.stepper);
} }
function DuplicateApplicationComponent_c8y_duplicate_application_properties_12_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-duplicate-application-properties", 8);
    ɵngcc0.ɵɵlistener("duplicatedApp", function DuplicateApplicationComponent_c8y_duplicate_application_properties_12_Template_c8y_duplicate_application_properties_duplicatedApp_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.duplicatedApp = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("existingApps", ctx_r1.existingApps)("selectedApp", ctx_r1.selectedApp)("stepper", ctx_r1.stepper);
} }
function DuplicateApplicationComponent_ng_container_14_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 12);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, "Application created"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function DuplicateApplicationComponent_ng_container_14_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 13);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, "Application creation failed"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function DuplicateApplicationComponent_ng_container_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DuplicateApplicationComponent_ng_container_14_div_1_Template, 3, 5, "div", 9);
    ɵngcc0.ɵɵtemplate(2, DuplicateApplicationComponent_ng_container_14_ng_template_2_Template, 3, 5, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r9 = ɵngcc0.ɵɵreference(3);
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.duplicatedApp)("ngIfElse", _r9);
} }
function DuplicateApplicationComponent_c8y_wizard_footer_15_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-wizard-footer");
    ɵngcc0.ɵɵelementStart(1, "button", 14);
    ɵngcc0.ɵɵlistener("click", function DuplicateApplicationComponent_c8y_wizard_footer_15_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.done(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "a", 15);
    ɵngcc0.ɵɵlistener("click", function DuplicateApplicationComponent_c8y_wizard_footer_15_Template_a_click_5_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelement(7, "i", 16);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 5, "Done"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, "Done"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 9, "Open"));
    ɵngcc0.ɵɵproperty("href", ctx_r3.getHref(ctx_r3.duplicatedApp), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 11, "Open"), " ");
} }
const _c0 = function () { return ["hidden"]; };
const _c1 = function () { return { edit: true, done: false }; };
export class DuplicateApplicationComponent {
    constructor(ecosystemService, applicationService, wizardComponent) {
        this.ecosystemService = ecosystemService;
        this.applicationService = applicationService;
        this.wizardComponent = wizardComponent;
        this.selectedApp = new BehaviorSubject(null);
        this.existingApps = [];
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.existingApps = yield this.ecosystemService.getWebApplications();
        });
    }
    onSelectedApp(app) {
        this.selectedApp.next(app);
    }
    getHref(app) {
        return this.applicationService.getHref(app);
    }
    done() {
        this.wizardComponent.close();
    }
}
DuplicateApplicationComponent.ɵfac = function DuplicateApplicationComponent_Factory(t) { return new (t || DuplicateApplicationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardComponent)); };
DuplicateApplicationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DuplicateApplicationComponent, selectors: [["c8y-duplicate-application"]], viewQuery: function DuplicateApplicationComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
    } }, decls: 16, vars: 12, consts: [[3, "c8yIcon"], ["translate", ""], ["linear", "", 3, "customClasses", "disableDefaultIcons"], [1, "p-16", "text-center", "text-medium", "separator-bottom", "sticky-top", "bg-white"], ["class", "d-block", "style", "min-height: 257px", 3, "existingApps", "stepper", "onSelectedApp", 4, "ngIf"], ["class", "d-flex d-col j-c-center a-i-center", "style", "min-height: 257px", 3, "existingApps", "selectedApp", "stepper", "duplicatedApp", 4, "ngIf"], [4, "ngIf"], [1, "d-block", 2, "min-height", "257px", 3, "existingApps", "stepper", "onSelectedApp"], [1, "d-flex", "d-col", "j-c-center", "a-i-center", 2, "min-height", "257px", 3, "existingApps", "selectedApp", "stepper", "duplicatedApp"], ["style", "min-height: 257px", "class", "d-flex a-i-center j-c-center", 4, "ngIf", "ngIfElse"], ["duplicateAppFailed", ""], [1, "d-flex", "a-i-center", "j-c-center", 2, "min-height", "257px"], ["type", "success", 3, "text", "size", "vertical"], ["type", "error", 3, "text", "size", "vertical"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["target", "_blank", 1, "btn", "btn-primary", 3, "href", "title", "click"], ["c8yIcon", "external-link", 1, "m-r-4"]], template: function DuplicateApplicationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-wizard-header");
        ɵngcc0.ɵɵelement(1, "h1", 0);
        ɵngcc0.ɵɵelementStart(2, "h4", 1);
        ɵngcc0.ɵɵtext(3, "Duplicate application");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-wizard-body");
        ɵngcc0.ɵɵelementStart(5, "c8y-stepper", 2);
        ɵngcc0.ɵɵelementStart(6, "cdk-step");
        ɵngcc0.ɵɵelementStart(7, "p", 3);
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, DuplicateApplicationComponent_c8y_duplicate_application_list_10_Template, 1, 2, "c8y-duplicate-application-list", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "cdk-step");
        ɵngcc0.ɵɵtemplate(12, DuplicateApplicationComponent_c8y_duplicate_application_properties_12_Template, 1, 3, "c8y-duplicate-application-properties", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(13, "cdk-step");
        ɵngcc0.ɵɵtemplate(14, DuplicateApplicationComponent_ng_container_14_Template, 4, 2, "ng-container", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(15, DuplicateApplicationComponent_c8y_wizard_footer_15_Template, 10, 13, "c8y-wizard-footer", 6);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("c8yIcon", "copy");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("customClasses", ɵngcc0.ɵɵpureFunction0(10, _c0))("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(11, _c1));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 8, "Select application"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepper.selectedIndex === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepper.selectedIndex === 1);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepper.selectedIndex === 2);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.stepper.selectedIndex === 2);
    } }, directives: [ɵngcc3.WizardHeaderComponent, ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc3.WizardBodyComponent, ɵngcc3.C8yStepper, ɵngcc4.CdkStep, ɵngcc5.NgIf, ɵngcc6.DuplicateApplicationListComponent, ɵngcc7.DuplicateApplicationPropertiesComponent, ɵngcc3.ɵe, ɵngcc3.WizardFooterComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
DuplicateApplicationComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: ApplicationService },
    { type: WizardComponent }
];
DuplicateApplicationComponent.propDecorators = {
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DuplicateApplicationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-duplicate-application',
                template: "<c8y-wizard-header>\n  <h1 [c8yIcon]=\"'copy'\"></h1>\n  <h4 translate>Duplicate application</h4>\n</c8y-wizard-header>\n\n<c8y-wizard-body>\n  <c8y-stepper\n    [customClasses]=\"['hidden']\"\n    [disableDefaultIcons]=\"{ edit: true, done: false }\"\n    linear\n  >\n    <!-- 1st step -->\n    <cdk-step>\n      <p class=\"p-16 text-center text-medium separator-bottom sticky-top bg-white\">\n        {{ 'Select application' | translate }}\n      </p>\n      <c8y-duplicate-application-list\n        (onSelectedApp)=\"onSelectedApp($event)\"\n        *ngIf=\"stepper.selectedIndex === 0\"\n        [existingApps]=\"existingApps\"\n        [stepper]=\"stepper\"\n        class=\"d-block\"\n        style=\"min-height: 257px\"\n      ></c8y-duplicate-application-list>\n    </cdk-step>\n    <!-- 2nd step -->\n    <cdk-step>\n      <c8y-duplicate-application-properties\n        class=\"d-flex d-col j-c-center a-i-center\"\n        style=\"min-height: 257px\"\n        (duplicatedApp)=\"duplicatedApp = $event\"\n        *ngIf=\"stepper.selectedIndex === 1\"\n        [existingApps]=\"existingApps\"\n        [selectedApp]=\"selectedApp\"\n        [stepper]=\"stepper\"\n      ></c8y-duplicate-application-properties>\n    </cdk-step>\n    <!-- Final step -->\n    <cdk-step>\n      <ng-container *ngIf=\"stepper.selectedIndex === 2\">\n        <div\n          *ngIf=\"duplicatedApp; else duplicateAppFailed\"\n          style=\"min-height: 257px\"\n          class=\"d-flex a-i-center j-c-center\"\n        >\n          <c8y-operation-result\n            text=\"{{ 'Application created' | translate }}\"\n            [size]=\"84\"\n            [vertical]=\"true\"\n            type=\"success\"\n          ></c8y-operation-result>\n        </div>\n        <ng-template #duplicateAppFailed>\n          <div style=\"min-height: 257px\" class=\"d-flex a-i-center j-c-center\">\n            <c8y-operation-result\n              text=\"{{ 'Application creation failed' | translate }}\"\n              [size]=\"84\"\n              [vertical]=\"true\"\n              type=\"error\"\n            ></c8y-operation-result>\n          </div>\n        </ng-template>\n      </ng-container>\n    </cdk-step>\n  </c8y-stepper>\n</c8y-wizard-body>\n\n<c8y-wizard-footer *ngIf=\"stepper.selectedIndex === 2\">\n  <button type=\"button\" (click)=\"done()\" class=\"btn btn-default\" title=\"{{ 'Done' | translate }}\">\n    {{ 'Done' | translate }}\n  </button>\n  <a\n    (click)=\"$event.stopPropagation()\"\n    [href]=\"getHref(duplicatedApp)\"\n    target=\"_blank\"\n    class=\"btn btn-primary\"\n    title=\"{{ 'Open' | translate }}\"\n  >\n    <i c8yIcon=\"external-link\" class=\"m-r-4\"></i>\n    {{ 'Open' | translate }}\n  </a>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.ApplicationService }, { type: ɵngcc3.WizardComponent }]; }, { stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: true }]
        }] }); })();
