// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-mapping-customaction.component.js
import { Component, Input } from '@angular/core';
import { NgModelGroup, ControlContainer } from '@angular/forms';
import { get } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from 'ngx-bootstrap/collapse';
import * as ɵngcc4 from '@angular/common';

function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_19_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 35);
    ɵngcc0.ɵɵelementStart(1, "i", 36);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_19_Template_i_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r11); const i_r5 = ɵngcc0.ɵɵnextContext().index; const ctx_r9 = ɵngcc0.ɵɵnextContext(2); return ctx_r9.removeCustomAction(i_r5); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_20_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 35);
    ɵngcc0.ɵɵelementStart(1, "i", 37);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_20_Template_i_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(3); return ctx_r12.addHeader(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelementStart(1, "div", 27);
    ɵngcc0.ɵɵelementStart(2, "c8y-form-group", 20);
    ɵngcc0.ɵɵelementStart(3, "label", 28);
    ɵngcc0.ɵɵtext(4, "Key");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 21);
    ɵngcc0.ɵɵelementStart(6, "input", 29);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_Template_input_ngModelChange_6_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const header_r4 = restoredCtx.$implicit; return header_r4.key = $event; });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 27);
    ɵngcc0.ɵɵelementStart(9, "c8y-form-group", 20);
    ɵngcc0.ɵɵelementStart(10, "label", 28);
    ɵngcc0.ɵɵtext(11, "Value");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 21);
    ɵngcc0.ɵɵelementStart(13, "input", 30);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_Template_input_ngModelChange_13_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const header_r4 = restoredCtx.$implicit; return header_r4.value = $event; });
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 31);
    ɵngcc0.ɵɵelementStart(16, "div", 32);
    ɵngcc0.ɵɵelementStart(17, "label", 33);
    ɵngcc0.ɵɵtext(18, "\u00A0");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(19, OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_19_Template, 2, 0, "button", 34);
    ɵngcc0.ɵɵtemplate(20, OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_button_20_Template, 2, 0, "button", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const header_r4 = ctx.$implicit;
    const i_r5 = ctx.index;
    const isLast_r6 = ctx.last;
    const ctx_r1 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngModelGroup", i_r5);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(7, 7, "e.g."), " Content-Type");
    ɵngcc0.ɵɵproperty("ngModel", header_r4.key);
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(14, 9, "e.g."), " application/json");
    ɵngcc0.ɵɵproperty("ngModel", header_r4.value);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.customAction.headers.length !== 1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", isLast_r6);
} }
function OpcuaDeviceProtocolMappingCustomAction_div_12_div_30_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate2(" ", ɵngcc0.ɵɵpipeBind1(2, 2, "The following placeholders are available:"), " ", ctx_r3.placeholders, " ");
} }
function OpcuaDeviceProtocolMappingCustomAction_div_12_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelementStart(1, "fieldset", 12);
    ɵngcc0.ɵɵelementStart(2, "div", 13);
    ɵngcc0.ɵɵtext(3, "Headers");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 14);
    ɵngcc0.ɵɵelementStart(5, "div", 15);
    ɵngcc0.ɵɵelementStart(6, "label", 16);
    ɵngcc0.ɵɵtext(7, "Key");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 15);
    ɵngcc0.ɵɵelementStart(9, "label", 16);
    ɵngcc0.ɵɵtext(10, "Value");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(11, OpcuaDeviceProtocolMappingCustomAction_div_12_div_11_Template, 21, 11, "div", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 13);
    ɵngcc0.ɵɵtext(13, "Action");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "div", 18);
    ɵngcc0.ɵɵelementStart(15, "div", 19);
    ɵngcc0.ɵɵelementStart(16, "c8y-form-group", 20);
    ɵngcc0.ɵɵelementStart(17, "label");
    ɵngcc0.ɵɵelementStart(18, "span", 16);
    ɵngcc0.ɵɵtext(19, "Endpoint");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(20, "div", 21);
    ɵngcc0.ɵɵelementStart(21, "input", 22);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMappingCustomAction_div_12_Template_input_ngModelChange_21_listener($event) { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(); return ctx_r17.customAction.endpoint = $event; });
    ɵngcc0.ɵɵpipe(22, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(23, "div", 19);
    ɵngcc0.ɵɵelementStart(24, "c8y-form-group", 20);
    ɵngcc0.ɵɵelementStart(25, "label", 16);
    ɵngcc0.ɵɵtext(26, "Body template");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(27, "textarea", 23, 24);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMappingCustomAction_div_12_Template_textarea_ngModelChange_27_listener($event) { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.customAction.bodyTemplate = $event; });
    ɵngcc0.ɵɵpipe(29, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(30, OpcuaDeviceProtocolMappingCustomAction_div_12_div_30_Template, 3, 4, "div", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r2 = ɵngcc0.ɵɵreference(28);
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(11);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.headers)("ngForTrackBy", ctx_r0.trackByIndex);
    ɵngcc0.ɵɵadvance(10);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(22, 8, "e.g."), " http://localhost:1080/cucumber");
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.customAction.endpoint);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "(", ɵngcc0.ɵɵpipeBind1(29, 10, "required"), ")");
    ɵngcc0.ɵɵproperty("heightPerRow", 40)("ngModel", ctx_r0.customAction.bodyTemplate);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", _r2.untouched || !_r2.errors);
} }
export class OpcuaDeviceProtocolMappingCustomAction {
    constructor() {
        this.hasCustomAction = false;
        // no translation needed
        this.placeholders = '${value}, ${serverId}, ${nodeId}, ${deviceId}';
    }
    getHeaderPair() {
        return { key: '', value: '' };
    }
    get headers() {
        return this.customAction.headers;
    }
    addHeader() {
        const { headers } = this.customAction;
        headers.push(this.getHeaderPair());
    }
    updateCustomAction() {
        this.hasCustomAction = !this.hasCustomAction;
    }
    removeCustomAction(index) {
        const { headers } = this.customAction;
        headers.splice(index, 1);
    }
    trackByIndex(index) {
        return index;
    }
    ngOnInit() {
        if (get(this.customAction, 'endpoint')) {
            this.hasCustomAction = true;
        }
    }
}
OpcuaDeviceProtocolMappingCustomAction.ɵfac = function OpcuaDeviceProtocolMappingCustomAction_Factory(t) { return new (t || OpcuaDeviceProtocolMappingCustomAction)(); };
OpcuaDeviceProtocolMappingCustomAction.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaDeviceProtocolMappingCustomAction, selectors: [["opcua-device-protocol-mapping-customaction"]], inputs: { customAction: "customAction" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgModelGroup }])], decls: 13, vars: 4, consts: [[1, "list-group"], [1, "list-group-item"], [1, "flex-row"], [1, "list-item-switch"], [1, "c8y-switch"], ["type", "checkbox", "name", "hasCustomAction", 3, "ngModel", "ngModelChange", "click"], [1, "list-item-icon"], ["c8yIcon", "wrench"], ["translate", "", 1, "list-item-title"], [3, "collapse", "isAnimated"], ["class", "p-t-8", 4, "ngIf"], [1, "p-t-8"], ["ngModelGroup", "headers"], ["translate", "", 1, "legend", "form-block"], [1, "tight-grid", "flex-row", "hidden-xs"], [1, "col-sm-4"], ["translate", ""], ["class", "flex-row tight-grid row", 3, "ngModelGroup", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "row"], [1, "col-md-4"], [1, "form-group-sm"], [1, "input-group"], ["type", "text", "name", "endpoint", "required", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["rows", "3", "name", "bodyTemplate", "required", "", 1, "form-control", 3, "heightPerRow", "placeholder", "ngModel", "ngModelChange"], ["model", "ngModel"], ["class", "help-block", 4, "ngIf"], [1, "flex-row", "tight-grid", "row", 3, "ngModelGroup"], [1, "col-xs-5", "col-sm-4"], ["translate", "", 1, "visible-xs"], ["type", "text", "name", "key", "required", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["type", "text", "name", "value", "required", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], [1, "col-xs-2"], [1, "form-group"], [1, "visible-xs"], ["class", "btn btn-clean", 4, "ngIf"], [1, "btn", "btn-clean"], ["c8yIcon", "minus-circle", 1, "text-danger", 3, "click"], ["c8yIcon", "plus-circle", 1, "text-primary", 3, "click"], [1, "help-block"]], template: function OpcuaDeviceProtocolMappingCustomAction_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "label", 4);
        ɵngcc0.ɵɵelementStart(5, "input", 5);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMappingCustomAction_Template_input_ngModelChange_5_listener($event) { return ctx.hasCustomAction = $event; })("click", function OpcuaDeviceProtocolMappingCustomAction_Template_input_click_5_listener() { return ctx.updateCustomAction(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "span");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 6);
        ɵngcc0.ɵɵelement(8, "i", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 8);
        ɵngcc0.ɵɵtext(10, " Custom action ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵtemplate(12, OpcuaDeviceProtocolMappingCustomAction_div_12_Template, 31, 12, "div", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngModel", ctx.hasCustomAction);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("collapse", !ctx.hasCustomAction)("isAnimated", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.hasCustomAction);
    } }, directives: [ɵngcc1.CheckboxControlValueAccessor, ɵngcc1.NgControlStatus, ɵngcc1.NgModel, ɵngcc2.IconDirective, ɵngcc2.C8yTranslateDirective, ɵngcc3.CollapseDirective, ɵngcc4.NgIf, ɵngcc1.NgControlStatusGroup, ɵngcc1.NgModelGroup, ɵngcc4.NgForOf, ɵngcc2.FormGroupComponent, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc1.DefaultValueAccessor, ɵngcc1.RequiredValidator, ɵngcc2.TextAreaRowHeightDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
OpcuaDeviceProtocolMappingCustomAction.propDecorators = {
    customAction: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolMappingCustomAction, [{
        type: Component,
        args: [{
                selector: 'opcua-device-protocol-mapping-customaction',
                template: "<div class=\"list-group\">\n  <div class=\"list-group-item\">\n    <div class=\"flex-row\">\n      <div class=\"list-item-switch\">\n        <label class=\"c8y-switch\">\n          <input\n            type=\"checkbox\"\n            name=\"hasCustomAction\"\n            [(ngModel)]=\"hasCustomAction\"\n            (click)=\"updateCustomAction()\"\n          />\n          <span></span>\n        </label>\n      </div>\n      <div class=\"list-item-icon\">\n        <i c8yIcon=\"wrench\"></i>\n      </div>\n      <div class=\"list-item-title\" translate>\n        Custom action\n      </div>\n    </div>\n    <div [collapse]=\"!hasCustomAction\" [isAnimated]=\"true\">\n      <div class=\"p-t-8\" *ngIf=\"hasCustomAction\">\n        <fieldset ngModelGroup=\"headers\">\n          <div class=\"legend form-block\" translate>Headers</div>\n          <div class=\"tight-grid flex-row hidden-xs\">\n            <div class=\"col-sm-4\">\n              <label translate>Key</label>\n            </div>\n            <div class=\"col-sm-4\">\n              <label translate>Value</label>\n            </div>\n          </div>\n          <div\n            *ngFor=\"let header of headers; trackBy: trackByIndex; let i = index; last as isLast\"\n            class=\"flex-row tight-grid row\"\n            [ngModelGroup]=\"i\"\n          >\n            <div class=\"col-xs-5 col-sm-4\">\n              <c8y-form-group class=\"form-group-sm\">\n                <label class=\"visible-xs\" translate>Key</label>\n                <div class=\"input-group\">\n                  <input\n                    class=\"form-control\"\n                    type=\"text\"\n                    name=\"key\"\n                    placeholder=\"{{ 'e.g.' | translate }} Content-Type\"\n                    required\n                    [(ngModel)]=\"header.key\"\n                  />\n                </div>\n              </c8y-form-group>\n            </div>\n            <div class=\"col-xs-5 col-sm-4\">\n              <c8y-form-group class=\"form-group-sm\">\n                <label class=\"visible-xs\" translate>Value</label>\n                <div class=\"input-group\">\n                  <input\n                    class=\"form-control\"\n                    type=\"text\"\n                    name=\"value\"\n                    placeholder=\"{{ 'e.g.' | translate }} application/json\"\n                    required\n                    [(ngModel)]=\"header.value\"\n                  />\n                </div>\n              </c8y-form-group>\n            </div>\n            <div class=\"col-xs-2\">\n              <div class=\"form-group\">\n                <label class=\"visible-xs\">&nbsp;</label>\n                <button class=\"btn btn-clean\" *ngIf=\"customAction.headers.length !== 1\">\n                  <i c8yIcon=\"minus-circle\" class=\"text-danger\" (click)=\"removeCustomAction(i)\"></i>\n                </button>\n                <button class=\"btn btn-clean\" *ngIf=\"isLast\">\n                  <i c8yIcon=\"plus-circle\" class=\"text-primary\" (click)=\"addHeader()\"></i>\n                </button>\n              </div>\n            </div>\n          </div>\n        </fieldset>\n        <div class=\"legend form-block\" translate>Action</div>\n        <div class=\"row\">\n          <div class=\"col-md-4\">\n            <c8y-form-group class=\"form-group-sm\">\n              <label>\n                <span translate>Endpoint</span>\n              </label>\n              <div class=\"input-group\">\n                <input\n                  class=\"form-control\"\n                  type=\"text\"\n                  name=\"endpoint\"\n                  placeholder=\"{{ 'e.g.' | translate }} http://localhost:1080/cucumber\"\n                  required\n                  [(ngModel)]=\"customAction.endpoint\"\n                />\n              </div>\n            </c8y-form-group>\n          </div>\n          <div class=\"col-md-4\">\n            <c8y-form-group class=\"form-group-sm\">\n              <label translate>Body template</label>\n              <textarea\n                #model='ngModel'\n                [heightPerRow]=\"40\"\n                class=\"form-control\"\n                rows=\"3\"\n                name=\"bodyTemplate\"\n                placeholder=\"({{'required' | translate}})\"\n                [(ngModel)]=\"customAction.bodyTemplate\"\n                required\n              ></textarea>\n              <div class=\"help-block\" *ngIf=\"model.untouched || !model.errors\">\n                {{ 'The following placeholders are available:' | translate }} {{ placeholders }}\n              </div>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgModelGroup }]
            }]
    }], function () { return []; }, { customAction: [{
            type: Input
        }] }); })();
