// @c8y/ngx-components/esm2015/protocol-opcua/opcua-auto-apply-settings.component.js
import { __awaiter } from "tslib";
import { Component, Input, EventEmitter } from '@angular/core';
import { InventoryService } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { reject } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from 'ngx-bootstrap/collapse';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';

function OpcuaAutoApplySettingsComponent_c8y_li_19_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 20);
    ɵngcc0.ɵɵelementStart(1, "c8y-filter", 21);
    ɵngcc0.ɵɵlistener("onSearch", function OpcuaAutoApplySettingsComponent_c8y_li_19_Template_c8y_filter_onSearch_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.filterItems($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "search");
} }
function OpcuaAutoApplySettingsComponent_c8y_li_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-body");
    ɵngcc0.ɵɵelementStart(2, "div", 22);
    ɵngcc0.ɵɵelementStart(3, "p");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 1, "No items to display."));
} }
function OpcuaAutoApplySettingsComponent_c8y_li_21_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-checkbox", 23);
    ɵngcc0.ɵɵlistener("onSelect", function OpcuaAutoApplySettingsComponent_c8y_li_21_Template_c8y_li_checkbox_onSelect_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r8); const item_r6 = restoredCtx.$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onSelect($event, item_r6); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-body");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelement(5, "br");
    ɵngcc0.ɵɵelementStart(6, "span", 24);
    ɵngcc0.ɵɵtext(7, "ID");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "small");
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r6 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selected", ctx_r2.isChecked(item_r6));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 3, item_r6.name));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(item_r6.id);
} }
function OpcuaAutoApplySettingsComponent_li_56_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 25);
    ɵngcc0.ɵɵelementStart(1, "c8y-input-group-list", 26);
    ɵngcc0.ɵɵlistener("onAdd", function OpcuaAutoApplySettingsComponent_li_56_Template_c8y_input_group_list_onAdd_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.add(); })("onRemove", function OpcuaAutoApplySettingsComponent_li_56_Template_c8y_input_group_list_onRemove_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.remove($event); });
    ɵngcc0.ɵɵelementStart(2, "c8y-form-group", 27);
    ɵngcc0.ɵɵelementStart(3, "input", 28);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaAutoApplySettingsComponent_li_56_Template_input_ngModelChange_3_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const i_r10 = restoredCtx.index; const ctx_r14 = ɵngcc0.ɵɵnextContext(); return (ctx_r14.constraints.matchesNodeIds[i_r10] = $event); });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const i_r10 = ctx.index;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("index", i_r10);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(4, 4, "e.g."), " nodeId");
    ɵngcc0.ɵɵproperty("ngModel", ctx_r3.constraints.matchesNodeIds[i_r10])("required", true);
} }
export class OpcuaAutoApplySettingsComponent {
    constructor(inventoryService) {
        this.inventoryService = inventoryService;
        this.opcuaServers = [];
        this.selectedItems = [];
        this.filteredList = [];
        this.checked = {};
        this.sizeToShowFilter = 5;
        this.constraints = {
            browsePathMatchesRegex: '',
            matchesNodeIds: [],
            serverObjectHasFragment: '',
            matchesServerIds: []
        };
        this.placeholderSelectServerIds = gettext('Select server IDs from list');
        this.updateSelectedItem = new EventEmitter();
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.inventoryService.list({
                pageSize: 1000,
                withTotalPages: true,
                type: 'c8y_OpcuaServer'
            });
            this.opcuaServers = data;
            this.filteredList = data;
            this.selectedItems = [];
            const { matchesServerIds } = this.constraints;
            data.forEach(server => {
                if (matchesServerIds &&
                    matchesServerIds.length > 0 &&
                    matchesServerIds.find(itemId => itemId === server.id)) {
                    this.selectedItems.push(server);
                    this.checked[server.id] = true;
                    this.showServerIds = true;
                }
            });
        });
    }
    set model(model) {
        if (model && model.applyConstraints) {
            this.constraints = model.applyConstraints;
        }
        this._model = model;
    }
    get model() {
        return this._model;
    }
    serverIdsSelected(items) {
        if (this.constraints) {
            this.constraints.matchesServerIds = items.map((item) => item.id);
        }
        this.selectedItems = items;
    }
    onChangeNodeId(event) {
        if (event.target.checked) {
            this.showRootNodes = true;
            this.add();
        }
        else {
            this.showRootNodes = false;
            this.constraints.matchesNodeIds = [];
        }
    }
    onChangeShowServerIds(event) {
        if (!event.target.checked) {
            this.constraints.matchesServerIds = [];
            this.showServerIds = false;
            this.selectedItems = [];
            this.checked = {};
        }
        else {
            this.showServerIds = true;
        }
    }
    onChangeShowBrowsePath(event) {
        if (!event.target.checked) {
            this.constraints.browsePathMatchesRegex = '';
            this.showBrowsePath = false;
        }
        else {
            this.showBrowsePath = true;
        }
    }
    onChangeShowServerFragment(event) {
        if (!event.target.checked) {
            this.constraints.serverObjectHasFragment = '';
            this.showServerFragment = false;
        }
        else {
            this.showServerFragment = true;
        }
    }
    add() {
        this.constraints.matchesNodeIds.push('');
    }
    remove(index) {
        this.constraints.matchesNodeIds.splice(index, 1);
    }
    trackByFn(index, item) {
        return index;
    }
    updateConstraints(items) {
        if (this.constraints) {
            this.constraints.matchesServerIds = items.map((item) => item.id);
        }
    }
    filterItems(filterText) {
        if (filterText.length !== 0) {
            const search = new RegExp(filterText, 'i');
            this.filteredList = this.opcuaServers.filter(({ name, id }) => {
                return search.test(name) || search.test(id);
            });
        }
        else {
            this.filteredList = this.opcuaServers;
        }
    }
    isChecked(item) {
        return this.checked[item.id];
    }
    onSelect(selected, item) {
        if (!selected) {
            this.selectedItems = reject(this.selectedItems, { id: item.id });
            delete this.checked[item.id];
        }
        else {
            this.selectedItems.push(item);
            this.checked[item.id] = selected;
        }
        this.updateSelectedItem.emit(true);
        this.updateConstraints(this.selectedItems);
    }
}
OpcuaAutoApplySettingsComponent.ɵfac = function OpcuaAutoApplySettingsComponent_Factory(t) { return new (t || OpcuaAutoApplySettingsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.InventoryService)); };
OpcuaAutoApplySettingsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaAutoApplySettingsComponent, selectors: [["opcua-auto-apply"]], inputs: { model: "model" }, decls: 57, vars: 52, consts: [[1, "row"], [1, "col-md-4"], ["translate", ""], [1, "col-md-6"], [1, "list-group", 2, "box-shadow", "none"], [1, "list-group-item"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", 3, "checked", "change"], [1, "m-l-8"], [1, "collapse", 3, "collapse"], [3, "items", "selected", "placeholder", "updateItems", "onChange"], ["class", "sticky-top", 4, "ngIf"], [4, "ngIf"], [4, "ngFor", "ngForOf"], [1, "collapse", 3, "isAnimated", "collapse"], [1, "m-t-8", "m-b-8"], ["name", "browsePath", "type", "text", "ngDefaultControl", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["name", "serverFragment", "type", "text", "ngDefaultControl", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["c8yInputGroupListContainer", "", 1, "list-unstyled", "p-t-16"], ["class", "m-b-8", 4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "sticky-top"], [3, "icon", "onSearch"], [1, "c8y-empty-state", "text-left"], [3, "selected", "onSelect"], ["translate", "", 1, "text-label-small", "m-r-4"], [1, "m-b-8"], [3, "index", "onAdd", "onRemove"], [1, "form-group--tooltip-validation"], ["type", "text", 1, "form-control", 3, "placeholder", "ngModel", "required", "ngModelChange"]], template: function OpcuaAutoApplySettingsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "p", 2);
        ɵngcc0.ɵɵtext(3, " Specifying auto-apply constraints allows you to limit the scope where the device protocols are applied, for example by specifying a set of possible servers or node IDs. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "p", 2);
        ɵngcc0.ɵɵtext(5, " If no constraints are set, device protocols are applied at any fitting location on the OPC UA server. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 3);
        ɵngcc0.ɵɵelementStart(7, "ul", 4);
        ɵngcc0.ɵɵelementStart(8, "li", 5);
        ɵngcc0.ɵɵelementStart(9, "label", 6);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementStart(11, "input", 7);
        ɵngcc0.ɵɵlistener("change", function OpcuaAutoApplySettingsComponent_Template_input_change_11_listener($event) { return ctx.onChangeShowServerIds($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(12, "span");
        ɵngcc0.ɵɵelementStart(13, "span", 8);
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 9);
        ɵngcc0.ɵɵelementStart(17, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(18, "c8y-select", 10);
        ɵngcc0.ɵɵlistener("onChange", function OpcuaAutoApplySettingsComponent_Template_c8y_select_onChange_18_listener($event) { return ctx.serverIdsSelected($event); });
        ɵngcc0.ɵɵtemplate(19, OpcuaAutoApplySettingsComponent_c8y_li_19_Template, 2, 1, "c8y-li", 11);
        ɵngcc0.ɵɵtemplate(20, OpcuaAutoApplySettingsComponent_c8y_li_20_Template, 6, 3, "c8y-li", 12);
        ɵngcc0.ɵɵtemplate(21, OpcuaAutoApplySettingsComponent_c8y_li_21_Template, 10, 5, "c8y-li", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "li", 5);
        ɵngcc0.ɵɵelementStart(23, "label", 6);
        ɵngcc0.ɵɵpipe(24, "translate");
        ɵngcc0.ɵɵelementStart(25, "input", 7);
        ɵngcc0.ɵɵlistener("change", function OpcuaAutoApplySettingsComponent_Template_input_change_25_listener($event) { return ctx.onChangeShowBrowsePath($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(26, "span");
        ɵngcc0.ɵɵelementStart(27, "span", 8);
        ɵngcc0.ɵɵtext(28);
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "div", 14);
        ɵngcc0.ɵɵelementStart(31, "c8y-form-group", 15);
        ɵngcc0.ɵɵelementStart(32, "input", 16);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaAutoApplySettingsComponent_Template_input_ngModelChange_32_listener($event) { return ctx.constraints.browsePathMatchesRegex = $event; });
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(34, "li", 5);
        ɵngcc0.ɵɵelementStart(35, "label", 6);
        ɵngcc0.ɵɵpipe(36, "translate");
        ɵngcc0.ɵɵelementStart(37, "input", 7);
        ɵngcc0.ɵɵlistener("change", function OpcuaAutoApplySettingsComponent_Template_input_change_37_listener($event) { return ctx.onChangeShowServerFragment($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(38, "span");
        ɵngcc0.ɵɵelementStart(39, "span", 8);
        ɵngcc0.ɵɵtext(40);
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 14);
        ɵngcc0.ɵɵelementStart(43, "c8y-form-group", 15);
        ɵngcc0.ɵɵelementStart(44, "input", 17);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaAutoApplySettingsComponent_Template_input_ngModelChange_44_listener($event) { return ctx.constraints.serverObjectHasFragment = $event; });
        ɵngcc0.ɵɵpipe(45, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(46, "li", 5);
        ɵngcc0.ɵɵelementStart(47, "label", 6);
        ɵngcc0.ɵɵpipe(48, "translate");
        ɵngcc0.ɵɵelementStart(49, "input", 7);
        ɵngcc0.ɵɵlistener("change", function OpcuaAutoApplySettingsComponent_Template_input_change_49_listener($event) { return ctx.onChangeNodeId($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(50, "span");
        ɵngcc0.ɵɵelementStart(51, "span", 8);
        ɵngcc0.ɵɵtext(52);
        ɵngcc0.ɵɵpipe(53, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(54, "div", 14);
        ɵngcc0.ɵɵelementStart(55, "ul", 18);
        ɵngcc0.ɵɵtemplate(56, OpcuaAutoApplySettingsComponent_li_56_Template, 5, 6, "li", 19);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 32, "Limit device protocol to a set of servers"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("checked", ctx.constraints.matchesServerIds !== null && ctx.constraints.matchesServerIds.length > 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 34, "Limit device protocol to a set of servers"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("collapse", !ctx.showServerIds && (ctx.constraints.matchesServerIds !== null && ctx.constraints.matchesServerIds.length < 1) || !ctx.showServerIds && ctx.constraints.matchesServerIds === null);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("items", ctx.opcuaServers)("selected", ctx.selectedItems)("placeholder", ctx.placeholderSelectServerIds)("updateItems", ctx.updateSelectedItem);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.opcuaServers.length > ctx.sizeToShowFilter);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.filteredList && ctx.filteredList.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.filteredList);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(24, 36, "Limit device protocol scope in the address space"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("checked", ctx.constraints.browsePathMatchesRegex !== null && ctx.constraints.browsePathMatchesRegex.length > 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(29, 38, "Limit device protocol scope in the address space"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("isAnimated", true)("collapse", !ctx.showBrowsePath && ctx.constraints.browsePathMatchesRegex !== null && ctx.constraints.browsePathMatchesRegex.length < 1 || !ctx.showBrowsePath && ctx.constraints.browsePathMatchesRegex === null);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(33, 40, "e.g."), " /objects/devices/.*");
        ɵngcc0.ɵɵproperty("ngModel", ctx.constraints.browsePathMatchesRegex);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(36, 42, "Limit device protocol to servers with a certain fragment"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("checked", ctx.constraints.serverObjectHasFragment !== null && ctx.constraints.serverObjectHasFragment.length > 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(41, 44, "Limit device protocol to servers with a certain fragment"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("isAnimated", true)("collapse", !ctx.showServerFragment && ctx.constraints.serverObjectHasFragment !== null && ctx.constraints.serverObjectHasFragment.length < 1 || !ctx.showServerFragment && ctx.constraints.serverObjectHasFragment === null);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(45, 46, "e.g."), " c8y_SomeServerMarker");
        ɵngcc0.ɵɵproperty("ngModel", ctx.constraints.serverObjectHasFragment);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(48, 48, "Limit device protocol to specific root nodes ID"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("checked", ctx.constraints.matchesNodeIds !== null && ctx.constraints.matchesNodeIds.length > 0);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(53, 50, "Limit device protocol to specific root nodes ID"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("isAnimated", true)("collapse", !ctx.showRootNodes && ctx.constraints.matchesNodeIds !== null && ctx.constraints.matchesNodeIds.length < 1 || !ctx.showRootNodes && ctx.constraints.matchesNodeIds === null);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.constraints.matchesNodeIds)("ngForTrackBy", ctx.trackByFn);
    } }, directives: [ɵngcc2.C8yTranslateDirective, ɵngcc3.CollapseDirective, ɵngcc2.FormGroupComponent, ɵngcc2.SelectComponent, ɵngcc4.NgIf, ɵngcc4.NgForOf, ɵngcc5.DefaultValueAccessor, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc2.InputGroupListContainerDirective, ɵngcc2.ListItemComponent, ɵngcc2.FilterInputComponent, ɵngcc2.ListItemBodyComponent, ɵngcc2.ListItemCheckboxComponent, ɵngcc2.InputGroupListComponent, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc5.RequiredValidator], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
OpcuaAutoApplySettingsComponent.ctorParameters = () => [
    { type: InventoryService }
];
OpcuaAutoApplySettingsComponent.propDecorators = {
    model: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaAutoApplySettingsComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-auto-apply',
                template: "<div class=\"row\">\n  <div class=\"col-md-4\">\n    <p translate>\n      Specifying auto-apply constraints allows you to limit the scope where the device protocols are\n      applied, for example by specifying a set of possible servers or node IDs.\n    </p>\n    <p translate>\n      If no constraints are set, device protocols are applied at any fitting location on the OPC UA\n      server.\n    </p>\n  </div>\n\n  <div class=\"col-md-6\">\n    <ul class=\"list-group\" style=\"box-shadow: none;\">\n      <!-- Limit device type to a set of servers -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to a set of servers' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"\n              constraints.matchesServerIds !== null && constraints.matchesServerIds.length > 0\n            \"\n            (change)=\"onChangeShowServerIds($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to a set of servers' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [collapse]=\"\n            (!showServerIds &&\n              (constraints.matchesServerIds !== null && constraints.matchesServerIds.length < 1)) ||\n            (!showServerIds && constraints.matchesServerIds === null)\n          \"\n        >\n          <c8y-form-group>\n            <c8y-select\n              [items]=\"opcuaServers\"\n              [selected]=\"selectedItems\"\n              [placeholder]=\"placeholderSelectServerIds\"\n              (onChange)=\"serverIdsSelected($event)\"\n              [updateItems]=\"updateSelectedItem\"\n            >\n            \n              <c8y-li *ngIf=\"opcuaServers.length > sizeToShowFilter\" class=\"sticky-top\">\n                <c8y-filter [icon]=\"'search'\"  (onSearch)=\"filterItems($event)\"></c8y-filter>\n              </c8y-li>\n              <c8y-li *ngIf=\"filteredList && filteredList.length === 0\">\n                <c8y-li-body> \n                  <div class=\"c8y-empty-state text-left\">\n                  <p>{{ 'No items to display.' | translate }}</p>\n                </div>\n                </c8y-li-body>\n              </c8y-li>\n              <c8y-li *ngFor=\"let item of filteredList\">\n                <c8y-li-checkbox\n                  [selected]=\"isChecked(item)\"\n                  (onSelect)=\"onSelect($event, item)\"\n                ></c8y-li-checkbox>\n                <c8y-li-body>{{ item.name | translate }}<br>\n                  <span class=\"text-label-small m-r-4\" translate>ID</span> <small>{{ item.id }}</small>\n                </c8y-li-body>\n              </c8y-li>\n            </c8y-select>\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type scope in the address space -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol scope in the address space' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"\n              constraints.browsePathMatchesRegex !== null &&\n              constraints.browsePathMatchesRegex.length > 0\n            \"\n            (change)=\"onChangeShowBrowsePath($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol scope in the address space' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"\n            (!showBrowsePath &&\n              constraints.browsePathMatchesRegex !== null &&\n              constraints.browsePathMatchesRegex.length < 1) ||\n            (!showBrowsePath && constraints.browsePathMatchesRegex === null)\n          \"\n        >\n          <c8y-form-group class=\"m-t-8 m-b-8\">\n            <input\n              name=\"browsePath\"\n              type=\"text\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} /objects/devices/.*\"\n              [(ngModel)]=\"constraints.browsePathMatchesRegex\"\n              ngDefaultControl\n            />\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type to servers with a certain fragment-->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to servers with a certain fragment' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"constraints.serverObjectHasFragment !== null && constraints.serverObjectHasFragment.length > 0\"\n            (change)=\"onChangeShowServerFragment($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to servers with a certain fragment' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"(!showServerFragment && constraints.serverObjectHasFragment !== null && constraints.serverObjectHasFragment.length < 1) || (!showServerFragment && constraints.serverObjectHasFragment === null)\"\n        >\n          <c8y-form-group class=\"m-t-8 m-b-8\">\n            <input\n              name=\"serverFragment\"\n              type=\"text\"\n              class=\"form-control\"\n              placeholder=\"{{ 'e.g.' | translate }} c8y_SomeServerMarker\"\n              [(ngModel)]=\"constraints.serverObjectHasFragment\"\n              ngDefaultControl\n            />\n          </c8y-form-group>\n        </div>\n      </li>\n      <!-- Limit device type to a specific root node ID -->\n      <li class=\"list-group-item\">\n        <label\n          title=\"{{ 'Limit device protocol to specific root nodes ID' | translate }}\"\n          class=\"c8y-checkbox\"\n        >\n          <input\n            type=\"checkbox\"\n            [checked]=\"constraints.matchesNodeIds !== null && constraints.matchesNodeIds.length > 0\"\n            (change)=\"onChangeNodeId($event)\"\n          />\n          <span></span>\n          <span class=\"m-l-8\">\n            {{ 'Limit device protocol to specific root nodes ID' | translate }}\n          </span>\n        </label>\n        <div\n          class=\"collapse\"\n          [isAnimated]=\"true\"\n          [collapse]=\"(!showRootNodes && constraints.matchesNodeIds !== null && constraints.matchesNodeIds.length < 1) || ( !showRootNodes && constraints.matchesNodeIds === null)\"\n        >\n            <ul c8yInputGroupListContainer class=\"list-unstyled p-t-16\">\n              <li\n                class=\"m-b-8\"\n                *ngFor=\"let item of constraints.matchesNodeIds; let i = index; trackBy: trackByFn\"\n              >\n                <c8y-input-group-list [index]=\"i\" (onAdd)=\"add()\" (onRemove)=\"remove($event)\">\n                  <c8y-form-group class=\"form-group--tooltip-validation\">\n                    <input\n                      type=\"text\"\n                      class=\"form-control\"\n                      placeholder=\"{{ 'e.g.' | translate }} nodeId\"\n                      [(ngModel)]=\"constraints.matchesNodeIds[i]\"\n                      [required]=\"true\"\n                    />\n                  </c8y-form-group>\n                </c8y-input-group-list>\n              </li>\n            </ul> \n        </div>\n      </li>\n    </ul>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.InventoryService }]; }, { model: [{
            type: Input
        }] }); })();
