// @c8y/ngx-components/esm2015/protocol-opcua/opcua-server-list.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';

function OpcuaServerListComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerListComponent_div_0_Template_div_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r3); const server_r1 = restoredCtx.$implicit; const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.presentConfig(server_r1); });
    ɵngcc0.ɵɵelement(1, "i", 2);
    ɵngcc0.ɵɵelementStart(2, "div", 3);
    ɵngcc0.ɵɵelementStart(3, "p", 4);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "small", 5);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵelement(8, "device-status", 7);
    ɵngcc0.ɵɵelementStart(9, "a", 8);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelement(11, "i", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const server_r1 = ctx.$implicit;
    ɵngcc0.ɵɵclassProp("active", server_r1.active);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "server");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(server_r1.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(server_r1.config.serverUrl);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("mo", server_r1);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 8, server_r1.quickInfo.padlockMsg));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", server_r1.quickInfo.padlock);
} }
export class OpcuaServerListComponent {
    constructor() {
        this.serverList = [];
        this.present = new EventEmitter();
    }
    presentConfig(server) {
        server.active = true;
        this.present.emit(server);
    }
}
OpcuaServerListComponent.ɵfac = function OpcuaServerListComponent_Factory(t) { return new (t || OpcuaServerListComponent)(); };
OpcuaServerListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaServerListComponent, selectors: [["opcua-server-list"]], inputs: { serverList: "serverList" }, outputs: { present: "present" }, decls: 1, vars: 1, consts: [["class", "c8y-stacked-item", 3, "active", "click", 4, "ngFor", "ngForOf"], [1, "c8y-stacked-item", 3, "click"], [1, "icon-20", 3, "c8yIcon"], [1, "p-l-16", "text-truncate"], ["id", "serverName", 1, "text-truncate"], ["id", "serverUrl", 1, "text-truncate", "text-muted"], [1, "flex-item-right", "p-l-8", "d-flex", "p-r-16"], [1, "p-r-8", 3, "mo"], ["placement", "left", 1, "text-gray-darker", 3, "title"], [3, "c8yIcon"]], template: function OpcuaServerListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, OpcuaServerListComponent_div_0_Template, 12, 10, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.serverList);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc2.IconDirective, ɵngcc2.DeviceStatusComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
OpcuaServerListComponent.propDecorators = {
    serverList: [{ type: Input }],
    present: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaServerListComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-server-list',
                template: "<!-- add active class to c8y-stacked-item when clicked -->\n<div\n  class=\"c8y-stacked-item\"\n  [class.active]=\"server.active\"\n  *ngFor=\"let server of serverList\"\n  (click)=\"presentConfig(server)\"\n>\n  <i [c8yIcon]=\"'server'\" class=\"icon-20 \"></i>\n  <div class=\"p-l-16 text-truncate\">\n    <p id=\"serverName\" class=\"text-truncate\">{{ server.name }}</p>\n    <small id=\"serverUrl\" class=\"text-truncate text-muted\">{{ server.config.serverUrl }}</small>\n  </div>\n  <div class=\"flex-item-right p-l-8 d-flex p-r-16\">\n    <device-status class=\"p-r-8\" [mo]=\"server\"></device-status>\n    <a\n      title=\"{{ server.quickInfo.padlockMsg | translate }}\"\n      placement=\"left\"\n      class=\"text-gray-darker\"\n    >\n      <i [c8yIcon]=\"server.quickInfo.padlock\"></i>\n    </a>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { serverList: [{
            type: Input
        }], present: [{
            type: Output
        }] }); })();
