// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-detail.component.js
import { __awaiter } from "tslib";
import { Component, ViewChildren, ChangeDetectorRef } from '@angular/core';
import { Router } from '@angular/router';
import { OpcuaService } from './opcuaService';
import { AlertService, gettext } from '@c8y/ngx-components';
import { find, assign, omit, findIndex, pick, get } from 'lodash';
import { OpcuaDeviceProtocolMapping } from './opcua-device-protocol-mapping.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './opcuaService';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/router';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from './opcua-device-protocol-description.component';
import * as ɵngcc7 from './opcua-device-protocol-data-reporting.component';
import * as ɵngcc8 from './opcua-auto-apply-settings.component';
import * as ɵngcc9 from './opcua-device-protocol-mapping.component';

function OpcuaDeviceProtocolDetailComponent_c8y_title_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-title");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.model.name);
} }
function OpcuaDeviceProtocolDetailComponent_form_2_div_7_opcua_device_protocol_mapping_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "opcua-device-protocol-mapping", 22);
    ɵngcc0.ɵɵlistener("onAction", function OpcuaDeviceProtocolDetailComponent_form_2_div_7_opcua_device_protocol_mapping_1_Template_opcua_device_protocol_mapping_onAction_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(3); return ctx_r8.actionHandler($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const resource_r6 = ctx.$implicit;
    const i_r7 = ctx.index;
    const ctx_r5 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("index", i_r7)("referencedServerId", ctx_r5.model.referencedServerId)("referencedRootNodeId", ctx_r5.model.referencedRootNodeId)("resource", ctx_r5.getStructuredResource(resource_r6))("getParentAttr", ctx_r5.getParentAttr);
} }
function OpcuaDeviceProtocolDetailComponent_form_2_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵtemplate(1, OpcuaDeviceProtocolDetailComponent_form_2_div_7_opcua_device_protocol_mapping_1_Template, 1, 5, "opcua-device-protocol-mapping", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.getMapping())("ngForTrackBy", ctx_r3.trackByIndex);
} }
function OpcuaDeviceProtocolDetailComponent_form_2_div_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 23);
    ɵngcc0.ɵɵelement(1, "h1", 24);
    ɵngcc0.ɵɵelementStart(2, "p", 7);
    ɵngcc0.ɵɵtext(3, "No variables to display. Click below to add.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function OpcuaDeviceProtocolDetailComponent_form_2_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 2, 3);
    ɵngcc0.ɵɵelement(2, "opcua-device-protocol-description", 4);
    ɵngcc0.ɵɵelementStart(3, "div", 5);
    ɵngcc0.ɵɵelementStart(4, "div", 6);
    ɵngcc0.ɵɵelementStart(5, "h4", 7);
    ɵngcc0.ɵɵtext(6, "Variables");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, OpcuaDeviceProtocolDetailComponent_form_2_div_7_Template, 2, 2, "div", 8);
    ɵngcc0.ɵɵelementStart(8, "div", 9);
    ɵngcc0.ɵɵtemplate(9, OpcuaDeviceProtocolDetailComponent_form_2_div_9_Template, 4, 0, "div", 10);
    ɵngcc0.ɵɵelementStart(10, "button", 11);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolDetailComponent_form_2_Template_button_click_10_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.addVariable(); });
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelement(12, "i", 12);
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 5);
    ɵngcc0.ɵɵelementStart(16, "div", 6);
    ɵngcc0.ɵɵelementStart(17, "h4", 7);
    ɵngcc0.ɵɵtext(18, "Data reporting");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "div", 13);
    ɵngcc0.ɵɵelement(20, "opcua-device-protocol-data-reporting", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 15);
    ɵngcc0.ɵɵelementStart(22, "div", 6);
    ɵngcc0.ɵɵelementStart(23, "h4", 7);
    ɵngcc0.ɵɵtext(24, "Auto apply constraints");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 16);
    ɵngcc0.ɵɵelement(26, "opcua-auto-apply", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(27, "span");
    ɵngcc0.ɵɵelementStart(28, "div", 17);
    ɵngcc0.ɵɵelementStart(29, "div", 18);
    ɵngcc0.ɵɵelementStart(30, "button", 19);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolDetailComponent_form_2_Template_button_click_30_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.save(); });
    ɵngcc0.ɵɵpipe(31, "translate");
    ɵngcc0.ɵɵtext(32, " Save ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r2 = ɵngcc0.ɵɵreference(1);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("model", ctx_r1.model);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.model.mappings.length > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.model.mappings.length === 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 10, "Add variable"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(14, 12, "Add variable"), " ");
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵproperty("groupName", "subscription")("model", ctx_r1.model);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("model", ctx_r1.model);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(31, 14, "Save"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r1.canSave(_r2));
} }
export class OpcuaDeviceProtocolDetailComponent {
    constructor(changeDetectorRef, opcuaService, alertService, router) {
        this.changeDetectorRef = changeDetectorRef;
        this.opcuaService = opcuaService;
        this.alertService = alertService;
        this.router = router;
        this.initialModel = {
            id: '',
            fieldbusType: 'opcuaV2',
            description: '',
            unit: '',
            fieldbusVersion: 4,
            name: '',
            referencedServerId: '',
            referencedRootNodeId: '',
            subscriptionType: {
                type: 'None'
            },
            mappings: [],
            overriddenSubscriptions: [],
            applyConstraints: {
                browsePathMatchesRegex: '',
                matchesNodeIds: [],
                serverObjectHasFragment: '',
                matchesServerIds: []
            },
            enabled: ''
        };
        this.isLoaded = true;
        this.getParentAttr = key => get(this.model, key);
    }
    ngAfterContentChecked() {
        this.changeDetectorRef.detectChanges();
    }
    getMapping() {
        return this.model.mappings;
    }
    getEmptyMappingObject() {
        return {
            id: 'new',
            browsePath: []
        };
    }
    getOverriddenSubscriptionsByPath(browsePath) {
        return find(this.model.overriddenSubscriptions, { browsePath });
    }
    getStructuredResource(resource) {
        const overriddenSubscriptions = this.getOverriddenSubscriptionsByPath(resource.browsePath);
        let result = assign({}, resource);
        if (overriddenSubscriptions) {
            result = assign({}, resource, { subscriptionType: overriddenSubscriptions.subscriptionType });
        }
        return result;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const id = this.opcuaService.getId();
            if (id) {
                const res = yield this.opcuaService.getDeviceProtocol(id);
                if (res && res.status !== 200) {
                    const data = res.json ? yield res.json() : undefined;
                    this.alertService.addServerFailure({ data, res });
                    this.isLoaded = false;
                }
                else {
                    const data = yield res.json();
                    if (data && data.applyConstraints === null) {
                        delete data.applyConstraints;
                    }
                    if (data && data.subscriptionType === null) {
                        delete data.subscriptionType;
                    }
                    this.model = assign(this.initialModel, data);
                    if (!this.model.mappings) {
                        this.model.mappings = [];
                    }
                    this.model = assign(this.initialModel, this.updateViableMapping(data));
                    this.isLoaded = false;
                }
            }
        });
    }
    updateViableMapping(model) {
        const { mappings } = model;
        let result = [];
        if (mappings) {
            result = mappings.map((item, i) => {
                return assign(item, { id: i });
            });
        }
        return assign(model, { mappings: result });
    }
    trackByIndex(index) {
        return index;
    }
    addVariable() {
        this.model.mappings.push(this.getEmptyMappingObject());
    }
    updateVariable(mappingObject) {
        const { mappings } = this.model;
        const index = findIndex(mappings, { id: mappingObject.id });
        mappings.splice(index, 1);
        if (mappingObject.id === 'new') {
            mappingObject.id = mappings.length;
        }
        mappings.push(mappingObject);
    }
    removeVariable(mappingObject) {
        const { mappings } = this.model;
        const index = findIndex(mappings, { id: mappingObject.id });
        mappings.splice(index, 1);
    }
    actionHandler(actionObject) {
        switch (actionObject.action) {
            case 'save':
                this.updateVariable(actionObject.data);
                break;
            case 'delete':
                this.removeVariable(actionObject.data);
                break;
        }
    }
    extractOverridSubscriptionType(_mapping) {
        const overriddenSubscriptions = [];
        const variableMapping = [];
        _mapping.forEach(element => {
            if (element.id !== 'new') {
                if (element.subscriptionType) {
                    overriddenSubscriptions.push(assign({ browsePath: element.browsePath }, { subscriptionType: element.subscriptionType }));
                }
                variableMapping.push(omit(element, ['subscriptionType']));
            }
        });
        return [variableMapping, overriddenSubscriptions];
    }
    prepareRequestJson(_model) {
        let requestJson = {};
        const [mappings, overriddenSubscriptions] = this.extractOverridSubscriptionType(_model.mappings);
        requestJson = assign(requestJson, pick(_model, Object.keys(this.initialModel)), {
            mappings,
            overriddenSubscriptions
        });
        return requestJson;
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const res = yield this.opcuaService.updateDeviceProtocol(this.prepareRequestJson(this.model));
                const data = yield res.json();
                if (res && res.status === 200) {
                    this.router.navigate(['deviceprotocols']);
                    this.alertService.success(gettext('Device protocol saved.'));
                }
                else {
                    const { details } = data;
                    this.alertService.addServerFailure({ res, data: details });
                }
            }
            catch (ex) {
                this.alertService.danger(gettext('Failed to save. Try again.'));
            }
        });
    }
    canSave(deviceTypeForm) {
        if (this.instanceList) {
            const activeInstances = this.instanceList.filter(item => item.isActive());
            if (activeInstances.length > 0) {
                return true;
            }
        }
        return !deviceTypeForm.form.valid;
    }
}
OpcuaDeviceProtocolDetailComponent.ɵfac = function OpcuaDeviceProtocolDetailComponent_Factory(t) { return new (t || OpcuaDeviceProtocolDetailComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.OpcuaService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.Router)); };
OpcuaDeviceProtocolDetailComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaDeviceProtocolDetailComponent, selectors: [["opcua-device-protocol-detail"]], viewQuery: function OpcuaDeviceProtocolDetailComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(OpcuaDeviceProtocolMapping, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.instanceList = _t);
    } }, decls: 3, vars: 2, consts: [[4, "ngIf"], ["name", "detailForm", 4, "ngIf"], ["name", "detailForm"], ["deviceTypeForm", "ngForm"], [3, "model"], [1, "card", "m-b-4"], [1, "card-header", "separator"], ["translate", ""], ["class", "list-group", "ngModelGroup", "variable", 4, "ngIf"], [1, "card-block"], ["class", "c8y-empty-state text-left", 4, "ngIf"], [1, "btn-add-block", "addVariableBtn", 3, "title", "click"], ["c8yIcon", "plus-circle"], ["ngModelGroup", "subscription", 1, "card-block"], [3, "groupName", "model"], [1, "card"], ["ngModelGroup", "autoApply", 1, "card-block", "overflow-visible"], [1, "text-center", "page-footer", "m-t-16"], [1, "btn-save-wrapper", "animated"], ["id", "deviceTypeSave", "translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["ngModelGroup", "variable", 1, "list-group"], [3, "index", "referencedServerId", "referencedRootNodeId", "resource", "getParentAttr", "onAction", 4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "index", "referencedServerId", "referencedRootNodeId", "resource", "getParentAttr", "onAction"], [1, "c8y-empty-state", "text-left"], ["c8yIcon", "sliders"]], template: function OpcuaDeviceProtocolDetailComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, OpcuaDeviceProtocolDetailComponent_c8y_title_0_Template, 2, 1, "c8y-title", 0);
        ɵngcc0.ɵɵelementStart(1, "div");
        ɵngcc0.ɵɵtemplate(2, OpcuaDeviceProtocolDetailComponent_form_2_Template, 33, 16, "form", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoaded);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoaded);
    } }, directives: [ɵngcc4.NgIf, ɵngcc2.TitleComponent, ɵngcc5.ɵNgNoValidate, ɵngcc5.NgControlStatusGroup, ɵngcc5.NgForm, ɵngcc6.OpcuaDeviceProtocolDescription, ɵngcc2.C8yTranslateDirective, ɵngcc2.IconDirective, ɵngcc5.NgModelGroup, ɵngcc7.OpcuaDeviceProtocolDataReportingComponent, ɵngcc8.OpcuaAutoApplySettingsComponent, ɵngcc4.NgForOf, ɵngcc9.OpcuaDeviceProtocolMapping], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
OpcuaDeviceProtocolDetailComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: OpcuaService },
    { type: AlertService },
    { type: Router }
];
OpcuaDeviceProtocolDetailComponent.propDecorators = {
    instanceList: [{ type: ViewChildren, args: [OpcuaDeviceProtocolMapping,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolDetailComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-device-protocol-detail',
                template: "<c8y-title *ngIf=\"!isLoaded\">{{ model.name }}</c8y-title>\n<div>\n  <form #deviceTypeForm=\"ngForm\" name=\"detailForm\" *ngIf=\"!isLoaded\">\n    <opcua-device-protocol-description [model]=\"model\"></opcua-device-protocol-description>\n    <div class=\"card m-b-4\">\n      <div class=\"card-header separator\">\n        <h4 translate>Variables</h4>\n      </div>\n      <div class=\"list-group\" *ngIf=\"model.mappings.length > 0\" ngModelGroup=\"variable\">\n        <opcua-device-protocol-mapping\n          *ngFor=\"let resource of getMapping(); trackBy: trackByIndex; let i = index\"\n          [index]=\"i\"\n          [referencedServerId]=\"model.referencedServerId\"\n          [referencedRootNodeId]=\"model.referencedRootNodeId\"\n          [resource]=\"getStructuredResource(resource)\"\n          [getParentAttr]=\"getParentAttr\"\n          (onAction)=\"actionHandler($event)\"\n        >\n        </opcua-device-protocol-mapping>\n      </div>\n      <div class=\"card-block\">\n        <div class=\"c8y-empty-state text-left\" *ngIf=\"model.mappings.length === 0\">\n          <h1 c8yIcon=\"sliders\"></h1>\n          <p translate>No variables to display. Click below to add.</p>\n        </div>\n        <button\n          title=\"{{ 'Add variable' | translate }}\"\n          class=\"btn-add-block addVariableBtn\"\n          (click)=\"addVariable()\"\n        >\n          <i c8yIcon=\"plus-circle\"></i> {{ 'Add variable' | translate }}\n        </button>\n      </div>\n    </div>\n    <div class=\"card m-b-4\">\n      <div class=\"card-header separator\">\n        <h4 translate>Data reporting</h4>\n      </div>\n      <div class=\"card-block\" ngModelGroup=\"subscription\">\n        <opcua-device-protocol-data-reporting\n          [groupName]=\"'subscription'\"\n          [model]=\"model\"\n        ></opcua-device-protocol-data-reporting>\n      </div>\n    </div>\n    <div class=\"card\">\n      <div class=\"card-header separator\">\n        <h4 translate>Auto apply constraints</h4>\n      </div>\n      <div class=\"card-block overflow-visible\" ngModelGroup=\"autoApply\">\n        <opcua-auto-apply [model]=\"model\"></opcua-auto-apply>\n      </div>\n    </div>\n\n    <span>\n      <div class=\"text-center page-footer m-t-16\">\n        <div class=\"btn-save-wrapper animated\">\n          <button\n            title=\"{{ 'Save' | translate }}\"\n            id=\"deviceTypeSave\"\n            class=\"btn btn-primary\"\n            (click)=\"save()\"\n            [disabled]=\"canSave(deviceTypeForm)\"\n            translate\n          >\n            Save\n          </button>\n        </div>\n      </div>\n    </span>\n  </form>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.OpcuaService }, { type: ɵngcc2.AlertService }, { type: ɵngcc3.Router }]; }, { instanceList: [{
            type: ViewChildren,
            args: [OpcuaDeviceProtocolMapping]
        }] }); })();
