// @c8y/ngx-components/esm2015/protocol-opcua/opcua-server-config.component.js
import { __awaiter } from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { gettext, DropAreaComponent } from '@c8y/ngx-components';
import { cloneDeep, has } from 'lodash-es';
import { OpcuaService } from './opcuaService';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './opcuaService';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from 'ngx-bootstrap/buttons';

const _c0 = ["opcuaConfigForm"];
function OpcuaServerConfigComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelement(1, "h1", 3);
    ɵngcc0.ɵɵelementStart(2, "div");
    ɵngcc0.ɵɵelementStart(3, "h3");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "p");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 2, "No server to display."));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 4, "Add or select a server."));
} }
function OpcuaServerConfigComponent_form_1_option_70_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 47);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const mode_r9 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngValue", mode_r9);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(mode_r9);
} }
function OpcuaServerConfigComponent_form_1_div_72_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 33);
    ɵngcc0.ɵɵelementStart(1, "div", 34);
    ɵngcc0.ɵɵelementStart(2, "label", 48);
    ɵngcc0.ɵɵtext(3, "Security policy");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "input", 49);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_72_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.model.config.securityMode = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("readonly", true)("ngModel", ctx_r4.model.config.securityMode);
} }
function OpcuaServerConfigComponent_form_1_div_73_select_5_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 47);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const policy_r15 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngValue", policy_r15);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(policy_r15);
} }
function OpcuaServerConfigComponent_form_1_div_73_select_5_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "select", 51);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_73_select_5_Template_select_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(3); return ctx_r16.model.config.securityMode = $event; });
    ɵngcc0.ɵɵtemplate(1, OpcuaServerConfigComponent_form_1_div_73_select_5_option_1_Template, 2, 2, "option", 38);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r12.model.config.securityMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r12.securityPolicies.sign);
} }
function OpcuaServerConfigComponent_form_1_div_73_select_6_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 47);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const policy_r19 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngValue", policy_r19);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(policy_r19);
} }
function OpcuaServerConfigComponent_form_1_div_73_select_6_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "select", 51);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_73_select_6_Template_select_ngModelChange_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); const ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.model.config.securityMode = $event; });
    ɵngcc0.ɵɵtemplate(1, OpcuaServerConfigComponent_form_1_div_73_select_6_option_1_Template, 2, 2, "option", 38);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r13.model.config.securityMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r13.securityPolicies.sign_enc);
} }
function OpcuaServerConfigComponent_form_1_div_73_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 33);
    ɵngcc0.ɵɵelementStart(1, "div", 34);
    ɵngcc0.ɵɵelementStart(2, "label", 48);
    ɵngcc0.ɵɵtext(3, "Security policy");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 36);
    ɵngcc0.ɵɵtemplate(5, OpcuaServerConfigComponent_form_1_div_73_select_5_Template, 2, 2, "select", 50);
    ɵngcc0.ɵɵtemplate(6, OpcuaServerConfigComponent_form_1_div_73_select_6_Template, 2, 2, "select", 50);
    ɵngcc0.ɵɵelement(7, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.currentSecMode === ctx_r5.SIGN);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.currentSecMode === ctx_r5.SIGN_ENC);
} }
function OpcuaServerConfigComponent_form_1_option_79_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 47);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const auth_r22 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngValue", auth_r22);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 2, auth_r22.value));
} }
function OpcuaServerConfigComponent_form_1_div_81_div_8_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Change password"), " ");
} }
function OpcuaServerConfigComponent_form_1_div_81_div_8_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Cancel password change"), " ");
} }
function OpcuaServerConfigComponent_form_1_div_81_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r28 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 34);
    ɵngcc0.ɵɵelementStart(1, "button", 57);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerConfigComponent_form_1_div_81_div_8_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r28); const ctx_r27 = ɵngcc0.ɵɵnextContext(3); return ctx_r27.toggleChangePassword(); });
    ɵngcc0.ɵɵtemplate(2, OpcuaServerConfigComponent_form_1_div_81_div_8_ng_container_2_Template, 3, 3, "ng-container", 56);
    ɵngcc0.ɵɵtemplate(3, OpcuaServerConfigComponent_form_1_div_81_div_8_ng_container_3_Template, 3, 3, "ng-container", 56);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r23 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r23.changePassword);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.changePassword);
} }
function OpcuaServerConfigComponent_form_1_div_81_div_10_Template(rf, ctx) { if (rf & 1) {
    const _r30 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "label", 58);
    ɵngcc0.ɵɵtext(2, "Password");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "input", 59);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_81_div_10_Template_input_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r30); const ctx_r29 = ɵngcc0.ɵɵnextContext(3); return ctx_r29.model.config.userPassword = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r24.model.config.userPassword);
} }
function OpcuaServerConfigComponent_form_1_div_81_Template(rf, ctx) { if (rf & 1) {
    const _r32 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "div", 52);
    ɵngcc0.ɵɵelementStart(2, "div", 34);
    ɵngcc0.ɵɵelementStart(3, "label", 53);
    ɵngcc0.ɵɵtext(4, "Username");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "input", 54);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_81_Template_input_ngModelChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r32); const ctx_r31 = ɵngcc0.ɵɵnextContext(2); return ctx_r31.model.config.userName = $event; });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 33);
    ɵngcc0.ɵɵtemplate(8, OpcuaServerConfigComponent_form_1_div_81_div_8_Template, 4, 2, "div", 55);
    ɵngcc0.ɵɵelementStart(9, "div", 34);
    ɵngcc0.ɵɵtemplate(10, OpcuaServerConfigComponent_form_1_div_81_div_10_Template, 4, 1, "div", 56);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(6, 4, "e.g. joe.doe`LOCALIZE`"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r7.model.config.userName);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r7.initialPasswordRequired);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r7.changePassword);
} }
function OpcuaServerConfigComponent_form_1_div_82_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "div", 33);
    ɵngcc0.ɵɵelementStart(2, "div", 34);
    ɵngcc0.ɵɵelementStart(3, "label", 60);
    ɵngcc0.ɵɵtext(4, "Keystore password");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "input", 61);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_82_Template_input_ngModelChange_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r34); const ctx_r33 = ɵngcc0.ɵɵnextContext(2); return ctx_r33.model.config.keystorePass = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 33);
    ɵngcc0.ɵɵelementStart(7, "div", 34);
    ɵngcc0.ɵɵelementStart(8, "label", 62);
    ɵngcc0.ɵɵtext(9, "Certificate password");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "input", 63);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_div_82_Template_input_ngModelChange_10_listener($event) { ɵngcc0.ɵɵrestoreView(_r34); const ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.model.config.certificatePass = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 52);
    ɵngcc0.ɵɵelementStart(12, "div", 34);
    ɵngcc0.ɵɵelementStart(13, "label", 64);
    ɵngcc0.ɵɵtext(14, "Upload keystore");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(15, "input", 65);
    ɵngcc0.ɵɵpipe(16, "translate");
    ɵngcc0.ɵɵelementStart(17, "c8y-drop-area", 66);
    ɵngcc0.ɵɵlistener("dropped", function OpcuaServerConfigComponent_form_1_div_82_Template_c8y_drop_area_dropped_17_listener($event) { ɵngcc0.ɵɵrestoreView(_r34); const ctx_r36 = ɵngcc0.ɵɵnextContext(2); return ctx_r36.uploadFile($event); });
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵpipe(19, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r8.model.config.keystorePass);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r8.model.config.certificatePass);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(16, 7, "e.g."), " yourKeystore.jks");
    ɵngcc0.ɵɵproperty("readonly", true)("ngModel", ctx_r8.fileName);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("loadingMessage", ɵngcc0.ɵɵpipeBind1(18, 9, "Importing, please wait."))("title", ɵngcc0.ɵɵpipeBind1(19, 11, "Import keystore with jks file extension"));
} }
function OpcuaServerConfigComponent_form_1_Template(rf, ctx) { if (rf & 1) {
    const _r38 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "form", 4, 5);
    ɵngcc0.ɵɵelementStart(2, "div", 6);
    ɵngcc0.ɵɵelementStart(3, "button", 7);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerConfigComponent_form_1_Template_button_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r37 = ɵngcc0.ɵɵnextContext(); return ctx_r37.cancel(); });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelement(5, "i", 8);
    ɵngcc0.ɵɵelementStart(6, "span", 9);
    ɵngcc0.ɵɵtext(7, "Back");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 10);
    ɵngcc0.ɵɵelementStart(9, "div", 11);
    ɵngcc0.ɵɵelementStart(10, "c8y-form-group", 12);
    ɵngcc0.ɵɵelementStart(11, "label");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "input", 13);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_input_ngModelChange_14_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r39 = ɵngcc0.ɵɵnextContext(); return ctx_r39.model.name = $event; });
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "c8y-messages");
    ɵngcc0.ɵɵelement(17, "c8y-message", 14);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "div", 15);
    ɵngcc0.ɵɵelementStart(20, "div", 16);
    ɵngcc0.ɵɵelementStart(21, "div", 17);
    ɵngcc0.ɵɵelementStart(22, "label", 18);
    ɵngcc0.ɵɵtext(23, "Server connection");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "button", 19);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_button_ngModelChange_24_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r40 = ɵngcc0.ɵɵnextContext(); return ctx_r40.targetConnectionState = $event; })("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_button_ngModelChange_24_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r41 = ɵngcc0.ɵɵnextContext(); return ctx_r41.setServerConnection($event); });
    ɵngcc0.ɵɵelementStart(25, "span", 20);
    ɵngcc0.ɵɵpipe(26, "translate");
    ɵngcc0.ɵɵtext(27, "Enabled");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "span", 20);
    ɵngcc0.ɵɵpipe(29, "translate");
    ɵngcc0.ɵɵtext(30, "Disabled");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(31, "div", 17);
    ɵngcc0.ɵɵelementStart(32, "label", 9);
    ɵngcc0.ɵɵtext(33, "Connection status");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(34, "div", 21);
    ɵngcc0.ɵɵelement(35, "device-status", 22);
    ɵngcc0.ɵɵelementStart(36, "span");
    ɵngcc0.ɵɵtext(37);
    ɵngcc0.ɵɵpipe(38, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(39, "div", 11);
    ɵngcc0.ɵɵelementStart(40, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(41, "label", 23);
    ɵngcc0.ɵɵtext(42, "Server URL");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(43, "input", 24);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_input_ngModelChange_43_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r42 = ɵngcc0.ɵɵnextContext(); return ctx_r42.model.config.serverUrl = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(44, "div", 25);
    ɵngcc0.ɵɵelementStart(45, "div", 26);
    ɵngcc0.ɵɵelementStart(46, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(47, "label", 27);
    ɵngcc0.ɵɵtext(48, "Timeout");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(49, "div", 28);
    ɵngcc0.ɵɵelementStart(50, "input", 29);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_input_ngModelChange_50_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r43 = ɵngcc0.ɵɵnextContext(); return ctx_r43.model.config.timeout = $event; });
    ɵngcc0.ɵɵpipe(51, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(52, "span", 30);
    ɵngcc0.ɵɵtext(53, " seconds ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(54, "div", 26);
    ɵngcc0.ɵɵelementStart(55, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(56, "label", 31);
    ɵngcc0.ɵɵtext(57, "Status check interval");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(58, "div", 28);
    ɵngcc0.ɵɵelementStart(59, "input", 32);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_input_ngModelChange_59_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r44 = ɵngcc0.ɵɵnextContext(); return ctx_r44.model.config.statusCheckInterval = $event; });
    ɵngcc0.ɵɵpipe(60, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(61, "span", 30);
    ɵngcc0.ɵɵtext(62, " seconds ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(63, "div", 16);
    ɵngcc0.ɵɵelementStart(64, "div", 33);
    ɵngcc0.ɵɵelementStart(65, "div", 34);
    ɵngcc0.ɵɵelementStart(66, "label", 35);
    ɵngcc0.ɵɵtext(67, "Security mode");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(68, "div", 36);
    ɵngcc0.ɵɵelementStart(69, "select", 37);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_select_ngModelChange_69_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r45 = ɵngcc0.ɵɵnextContext(); return ctx_r45.currentSecMode = $event; })("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_select_ngModelChange_69_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r46 = ɵngcc0.ɵɵnextContext(); return ctx_r46.setPolicy($event); });
    ɵngcc0.ɵɵtemplate(70, OpcuaServerConfigComponent_form_1_option_70_Template, 2, 2, "option", 38);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(71, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(72, OpcuaServerConfigComponent_form_1_div_72_Template, 5, 2, "div", 39);
    ɵngcc0.ɵɵtemplate(73, OpcuaServerConfigComponent_form_1_div_73_Template, 8, 2, "div", 39);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(74, "div", 34);
    ɵngcc0.ɵɵelementStart(75, "label", 40);
    ɵngcc0.ɵɵtext(76, "Authentication");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(77, "div", 36);
    ɵngcc0.ɵɵelementStart(78, "select", 41);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_select_ngModelChange_78_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r47 = ɵngcc0.ɵɵnextContext(); return ctx_r47.authenticationMode = $event; })("ngModelChange", function OpcuaServerConfigComponent_form_1_Template_select_ngModelChange_78_listener($event) { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r48 = ɵngcc0.ɵɵnextContext(); return ctx_r48.updateAuthentication($event); });
    ɵngcc0.ɵɵtemplate(79, OpcuaServerConfigComponent_form_1_option_79_Template, 3, 4, "option", 38);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(80, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(81, OpcuaServerConfigComponent_form_1_div_81_Template, 11, 6, "div", 42);
    ɵngcc0.ɵɵtemplate(82, OpcuaServerConfigComponent_form_1_div_82_Template, 20, 13, "div", 42);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(83, "div", 43);
    ɵngcc0.ɵɵelementStart(84, "button", 44);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerConfigComponent_form_1_Template_button_click_84_listener() { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r49 = ɵngcc0.ɵɵnextContext(); return ctx_r49.cancel(); });
    ɵngcc0.ɵɵpipe(85, "translate");
    ɵngcc0.ɵɵtext(86, " Cancel ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(87, "button", 45);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerConfigComponent_form_1_Template_button_click_87_listener() { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r50 = ɵngcc0.ɵɵnextContext(); return ctx_r50.remove(); });
    ɵngcc0.ɵɵpipe(88, "translate");
    ɵngcc0.ɵɵtext(89, " Remove ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(90, "button", 46);
    ɵngcc0.ɵɵlistener("click", function OpcuaServerConfigComponent_form_1_Template_button_click_90_listener() { ɵngcc0.ɵɵrestoreView(_r38); const ctx_r51 = ɵngcc0.ɵɵnextContext(); return ctx_r51.save(); });
    ɵngcc0.ɵɵpipe(91, "translate");
    ɵngcc0.ɵɵtext(92, " Save ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r2 = ɵngcc0.ɵɵreference(1);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 31, "Back"));
    ɵngcc0.ɵɵadvance(9);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 33, "Server name"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(15, 35, "e.g. My server"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.model.name);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(18, 37, "Server name is required"));
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.targetConnectionState);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(26, 39, "Enabled"));
    ɵngcc0.ɵɵproperty("hidden", ctx_r1.targetConnectionState !== "1");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(29, 41, "Disabled"));
    ɵngcc0.ɵɵproperty("hidden", ctx_r1.targetConnectionState !== "0");
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵproperty("mo", ctx_r1.server);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(38, 43, ctx_r1.connectionStatusLabel));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.model.config.serverUrl);
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(51, 45, "e.g."), " 30");
    ɵngcc0.ɵɵproperty("min", ctx_r1.minIntervalNumber)("ngModel", ctx_r1.model.config.timeout);
    ɵngcc0.ɵɵadvance(9);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(60, 47, "e.g."), " 40");
    ɵngcc0.ɵɵproperty("min", ctx_r1.minIntervalNumber)("ngModel", ctx_r1.model.config.statusCheckInterval);
    ɵngcc0.ɵɵadvance(10);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.currentSecMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.securityModes);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.currentSecMode === ctx_r1.NONE);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.currentSecMode !== ctx_r1.NONE);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.authenticationMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.authenticationModes);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.authenticationMode.id === 2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.authenticationMode.id === 3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(85, 49, "Cancel"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(88, 51, "Remove"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(91, 53, "Save"));
    ɵngcc0.ɵɵproperty("disabled", !_r2.valid);
} }
export class OpcuaServerConfigComponent {
    constructor(opcuaService) {
        this.fileName = '';
        this.targetConnectionState = '1';
        this.minIntervalNumber = 1;
        this.connectionStatusLabel = '';
        this.canceled = new EventEmitter();
        this.removed = new EventEmitter();
        this.saved = new EventEmitter();
        this.changePassword = false;
        this.initialPasswordRequired = true;
        this.NONE = 'NONE';
        this.SIGN = 'SIGN';
        this.SIGN_ENC = 'SIGN_ENCRYPT';
        this.securityPolicies = {
            sign: [
                `BASIC256_${this.SIGN}`,
                `BASIC128RSA15_${this.SIGN}`,
                `BASIC256SHA256_${this.SIGN}`
            ],
            sign_enc: [
                `BASIC256_${this.SIGN_ENC}`,
                `BASIC128RSA15_${this.SIGN_ENC}`,
                `BASIC256SHA256_${this.SIGN_ENC}`,
            ]
        };
        this.ANONYM = {
            id: 1,
            value: gettext('Anonymous')
        };
        this.USER_PASSWORD = {
            id: 2,
            value: gettext('Username/Password')
        };
        this.KEY_BASED = {
            id: 3,
            value: gettext('Key-based Authentication')
        };
        this.initialKeystore = {
            lastModified: 0,
            name: '',
            type: '',
            slice: null,
            size: 0
        };
        this.keystore = this.initialKeystore;
        this.authSwitch = false;
        this.opcuaService = opcuaService;
    }
    set server(server) {
        if (server) {
            this._server = cloneDeep(server);
            this.model = cloneDeep(server);
            this.fileName = this.model.config.keystoreFilename;
            if (server.id && server.id === 'new') {
                // enabled connection state
                this.targetConnectionState = '1';
                this.model.config.targetConnectionState = 'enabled';
            }
            else {
                this.targetConnectionState = (this.model.config.targetConnectionState === 'enabled') ? '1' : '0';
            }
            this.updateConnectionStatusLabel(this._server);
            this.setNewPassword();
        }
    }
    get server() {
        return this._server;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.authSwitch = false;
            this.securityModes = [
                this.NONE,
                this.SIGN,
                this.SIGN_ENC
            ];
            this.authenticationModes = [
                this.ANONYM,
                this.USER_PASSWORD,
                this.KEY_BASED
            ];
            this.setCurrentAuthenticationMode();
            this.setCurrentSecurityMode();
        });
    }
    ngOnChanges() {
        this.setCurrentSecurityMode();
        this.setCurrentAuthenticationMode();
    }
    cancel() {
        this.canceled.emit(this.model);
        this._server = null;
    }
    remove() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.removeKeystore(this.model);
            this.removed.emit(this.model);
            this._server = null;
        });
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.keystore && this.keystore.size > 0 && this.keystore.name && this.keystore.name.length > 0) {
                const response = yield this.uploadKeystore(this.model.config.keystoreBinaryId);
                if (response && response.data && response.data.id) {
                    this.model.config.keystoreBinaryId = response.data.id;
                }
                // if the keystore was uploaded successful we can remove
                // the local keystore. This will prevent another request to binary api
                // when the user will edit other inputs in the form and hit save again.
                this.keystore = this.initialKeystore;
            }
            // will remove keystore (binary) when the user switched
            // authentication settings from key-based to anonymous or username/password
            if (this.authSwitch) {
                this.removeKeystore(this.server);
            }
            // when the user sets a new password, make sure to mark it as
            // "not encrypted" by setting passwordEncrypted to false
            const userPassword = this.getModelConfig('userPassword');
            if (userPassword && userPassword.length > 0) {
                this.model.config.passwordEncrypted = false;
            }
            this.saved.emit(this.model);
        });
    }
    uploadFile(droppedFiles) {
        if (droppedFiles.length === 1) {
            this.keystore = droppedFiles[0].file;
            this.fileName = this.keystore.name;
        }
        else {
            // dropped more than one file
            console.warn('Tried to import... Import aborted.');
        }
    }
    setPolicy(data) {
        if (data === this.NONE) {
            this.model.config.securityMode = this.NONE;
        }
        else if (data === this.SIGN) {
            this.model.config.securityMode = this.securityPolicies.sign[0];
        }
        else if (data === this.SIGN_ENC) {
            this.model.config.securityMode = this.securityPolicies.sign_enc[0];
        }
    }
    setServerConnection(data) {
        this.model.config.targetConnectionState = (data !== '0') ? 'enabled' : 'disabled';
    }
    updateAuthentication(data) {
        if (data && data.id) {
            switch (data.id) {
                // Anonymous
                case 1:
                    this.resetUserAuthentication();
                    this.resetKeyBasedAuthentication();
                    break;
                // User/Password
                case 2:
                    this.resetKeyBasedAuthentication();
                    this.restoreUserData();
                    this.setNewPassword();
                    break;
                // Key-based
                case 3:
                    this.resetUserAuthentication();
                    this.restoreKeyBasedData();
                    break;
                default:
                    console.warn('Invalid authentication id', data.id);
                    break;
            }
        }
    }
    updateConnectionStatusLabel(server) {
        const connected = server.c8y_Connection && server.c8y_Connection.status === 'CONNECTED';
        const label = connected ? gettext('Connected') : gettext('Disconnected');
        this.connectionStatusLabel = label;
    }
    setNewPassword() {
        const username = this.getModelConfig('userName');
        if (username && username.length > 0) {
            // userName is given, NO need to change the password because it is already set
            this.changePassword = false;
            this.initialPasswordRequired = false;
        }
        else {
            // no userName in response, so require the user to set the initial pw
            this.changePassword = true;
            this.initialPasswordRequired = true;
        }
    }
    toggleChangePassword() {
        this.changePassword = !this.changePassword;
        // When the user hides the pw-input field but has entered a
        // string to it before, we need to discard the changes reflected in the model
        // otherwise we PUT it with the model when user hits the save button
        if (!this.changePassword) {
            if (this.getModelConfig('userPassword')) {
                delete this.model.config.userPassword;
            }
        }
    }
    uploadKeystore(binaryId) {
        if (!binaryId) {
            return this.opcuaService.uploadKeystore(this.keystore);
        }
        else if (binaryId && binaryId.length > 0) {
            // update existing binary
            return this.opcuaService.updateKeystore(binaryId, this.keystore);
        }
    }
    removeKeystore(server) {
        if (server &&
            server.config &&
            server.config.keystoreBinaryId &&
            server.config.keystoreBinaryId.length > 0) {
            this.authSwitch = false;
            return this.opcuaService.removeKeystore(this.server.config.keystoreBinaryId);
        }
    }
    resetUserAuthentication() {
        this.model.config.userName = null;
        this.model.config.userPassword = null;
        this.model.config.userIdentityMode = 'none';
    }
    resetKeyBasedAuthentication() {
        this.authSwitch = true;
        this.model.config.keystorePass = null;
        this.model.config.certificatePass = null;
        this.model.config.keystoreBinaryId = '';
        this.model.config.keystoreFilename = '';
        this.model.config.userIdentityMode = 'none';
    }
    restoreUserData() {
        this.model.config.userName = this._server.config.userName;
        this.model.config.userIdentityMode = 'userAndPassword';
    }
    restoreKeyBasedData() {
        this.authSwitch = false;
        this.model.config.keystorePass = this._server.config.keystorePass;
        this.model.config.certificatePass = this._server.config.certificatePass;
        this.model.config.keystoreBinaryId = this._server.config.keystoreBinaryId;
        this.model.config.keystoreFilename = this._server.config.keystoreFilename;
        this.model.config.userIdentityMode = 'certificate';
    }
    getServerConfig() {
        let cfg = {
            securityMode: this.NONE,
            userIdentityMode: 'none'
        };
        if (this.server && this.server.config) {
            cfg = this.server.config;
        }
        return cfg;
    }
    setCurrentSecurityMode() {
        const { securityMode } = this.getServerConfig();
        if (securityMode) {
            const foundInSign = this.securityPolicies.sign.find((el) => el === securityMode);
            if (foundInSign) {
                this.currentSecMode = this.SIGN;
            }
            else {
                const foundInSignEncrypt = this.securityPolicies.sign_enc.find((el) => el === securityMode);
                foundInSignEncrypt ? this.currentSecMode = this.SIGN_ENC : this.currentSecMode = this.NONE;
            }
        }
    }
    setCurrentAuthenticationMode() {
        const { userIdentityMode } = this.getServerConfig();
        switch (userIdentityMode) {
            case 'certificate':
                this.authenticationMode = this.KEY_BASED;
                break;
            case 'userAndPassword':
                this.authenticationMode = this.USER_PASSWORD;
                break;
            case 'Anonymous':
                this.authenticationMode = this.ANONYM;
                break;
            case 'UserName':
                this.authenticationMode = this.USER_PASSWORD;
                break;
            case 'Certificate':
                this.authenticationMode = this.KEY_BASED;
                break;
            default:
                this.authenticationMode = this.ANONYM;
                break;
        }
    }
    getModelConfig(fragment) {
        if (this.model && this.model.config) {
            if (fragment && fragment.length > 0) {
                return has(this.model.config, fragment) ? this.model.config[`${fragment}`] : undefined;
            }
        }
        return undefined;
    }
}
OpcuaServerConfigComponent.ɵfac = function OpcuaServerConfigComponent_Factory(t) { return new (t || OpcuaServerConfigComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.OpcuaService)); };
OpcuaServerConfigComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaServerConfigComponent, selectors: [["opcua-server-config"]], viewQuery: function OpcuaServerConfigComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
        ɵngcc0.ɵɵviewQuery(DropAreaComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.opcuaConfigForm = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropArea = _t.first);
    } }, inputs: { server: "server" }, outputs: { canceled: "canceled", removed: "removed", saved: "saved" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 2, vars: 2, consts: [["class", "c8y-empty-state m-t-24", "ng-if", "!vm.editInstruction.type", 4, "ngIf"], ["class", "d-contents", 4, "ngIf"], ["ng-if", "!vm.editInstruction.type", 1, "c8y-empty-state", "m-t-24"], ["c8yIcon", "server", 1, "c8y-icon", "c8y-icon-duocolor"], [1, "d-contents"], ["opcuaConfigForm", "ngForm"], [1, "card-header", "large-padding", "separator", "sticky-top", "visible-sm", "visible-xs"], [1, "btn", "btn-clean", "text-primary", "visible-sm", "visible-xs", 3, "title", "click"], ["c8yIcon", "chevron-left"], ["translate", ""], [1, "flex-grow"], [1, "card-block", "large-padding"], [1, "m-b-8"], ["type", "text", "id", "name", "name", "name", "required", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["name", "required", 3, "text"], [1, "card-block", "large-padding", "bg-gray-lighter"], [1, "tight-grid"], [1, "col-sm-6"], ["translate", "", 2, "width", "100%"], ["type", "button", "name", "serverConnection", "btnCheckbox", "", "btnCheckboxTrue", "1", "btnCheckboxFalse", "0", 1, "btn", "m-t-4", 3, "ngModel", "ngModelChange"], ["translate", "", 3, "title", "hidden"], [1, "form-control-static"], [1, "p-r-8", 3, "mo"], ["for", "configServerUrl", "translate", ""], ["type", "text", "id", "configServerUrl", "name", "serverUrl", "c8yDefaultValidation", "opcuaBrowsePath", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], [1, "content-flex-32"], [1, "col-6"], ["for", "config.timeout", "translate", ""], [1, "input-group"], ["type", "number", "id", "config.timeout", "name", "timeout", "required", "", 1, "form-control", 3, "min", "placeholder", "ngModel", "ngModelChange"], ["translate", "", 1, "input-group-addon", "units"], ["for", "config.statusCheckInterval", "translate", ""], ["type", "number", "id", "config.statusCheckInterval", "name", "statusCheckInterval", "required", "", 1, "form-control", 3, "min", "placeholder", "ngModel", "ngModelChange"], [1, "col-md-6"], [1, "form-group"], ["for", "config.securityMode", "translate", ""], [1, "c8y-select-wrapper"], ["id", "config.securityMode", "name", "securityMode", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], [3, "ngValue", 4, "ngFor", "ngForOf"], ["class", "col-md-6", 4, "ngIf"], ["for", "config.authenticationMode", "translate", ""], ["id", "config.authenticationMode", "name", "authenticationMode", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["class", "tight-grid", 4, "ngIf"], [1, "card-footer", "large-padding", "separator", "sticky-bottom"], ["translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", 1, "btn", "btn-danger", 3, "title", "click"], ["translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"], [3, "ngValue"], ["for", "config.securityPolicy", "translate", ""], ["type", "text", "id", "config.securityPolicy", "name", "securityPolicy", "required", "", 1, "form-control", 3, "readonly", "ngModel", "ngModelChange"], ["class", "form-control", "id", "config.securityPolicy", "name", "securityPolicy", "required", "", 3, "ngModel", "ngModelChange", 4, "ngIf"], ["id", "config.securityPolicy", "name", "securityPolicy", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], [1, "col-md-12"], ["for", "config.userName", "translate", ""], ["type", "text", "id", "config.userName", "name", "userName", "autocomplete", "new-password", "required", "", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["class", "form-group", 4, "ngIf"], [4, "ngIf"], ["type", "button", 1, "btn", "btn-default", 3, "click"], ["for", "config.password", "translate", ""], ["type", "password", "id", "config.userPassword", "name", "password", "autocomplete", "new-password", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["for", "config.keystorePass", "translate", ""], ["type", "password", "id", "config.keystorePass", "name", "keystorePass", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["for", "config.certificatePass", "translate", ""], ["type", "password", "id", "config.certificatePass", "name", "keystorePass", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["for", "certificateUpload", "translate", ""], ["type", "text", "name", "certificateUpload", "required", "", 1, "form-control", "m-b-8", 3, "readonly", "ngModel", "placeholder"], [3, "loadingMessage", "title", "dropped"]], template: function OpcuaServerConfigComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, OpcuaServerConfigComponent_div_0_Template, 9, 6, "div", 0);
        ɵngcc0.ɵɵtemplate(1, OpcuaServerConfigComponent_form_1_Template, 93, 55, "form", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.server);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.server);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.IconDirective, ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc3.C8yTranslateDirective, ɵngcc3.FormGroupComponent, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc3.MessagesComponent, ɵngcc3.MessageDirective, ɵngcc5.ButtonCheckboxDirective, ɵngcc3.DeviceStatusComponent, ɵngcc3.DefaultValidationDirective, ɵngcc4.MinValidator, ɵngcc4.NumberValueAccessor, ɵngcc3.MinValidationDirective, ɵngcc4.SelectControlValueAccessor, ɵngcc2.NgForOf, ɵngcc4.NgSelectOption, ɵngcc4.ɵNgSelectMultipleOption, ɵngcc3.DropAreaComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
OpcuaServerConfigComponent.ctorParameters = () => [
    { type: OpcuaService }
];
OpcuaServerConfigComponent.propDecorators = {
    opcuaConfigForm: [{ type: ViewChild, args: ['opcuaConfigForm', { static: false },] }],
    dropArea: [{ type: ViewChild, args: [DropAreaComponent, { static: false },] }],
    canceled: [{ type: Output }],
    removed: [{ type: Output }],
    saved: [{ type: Output }],
    server: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaServerConfigComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-server-config',
                template: "\n  <div class=\"c8y-empty-state m-t-24\"  ng-if=\"!vm.editInstruction.type\" *ngIf=\"!server\">\n    <h1 class=\"c8y-icon c8y-icon-duocolor\" c8yIcon=\"server\"></h1>\n    <div>\n      <h3>{{ 'No server to display.' | translate }}</h3>\n      <p>{{ 'Add or select a server.' | translate}}</p>\n    </div>\n  </div>\n\n\n<form #opcuaConfigForm=\"ngForm\" class=\"d-contents\" *ngIf=\"server\">\n  <div class=\"card-header large-padding separator sticky-top visible-sm visible-xs\" >\n    <button\n    class=\"btn btn-clean text-primary visible-sm visible-xs\"\n    title=\"{{ 'Back' | translate }}\"\n    (click)=\"cancel()\"\n    >\n      <i c8yIcon=\"chevron-left\"></i>\n      <span translate>Back</span>\n    </button>\n  </div>\n  <div class=\"flex-grow\">\n  <div class=\"card-block large-padding\">\n    <!-- SERVER NAME -->\n    <c8y-form-group class=\"m-b-8\">\n      <label>\n        {{ 'Server name' | translate }}\n      </label>\n      <input\n        type=\"text\"\n        class=\"form-control\"\n        placeholder=\"{{ 'e.g. My server' | translate }}\"\n        id=\"name\"\n        name=\"name\"\n        [(ngModel)]=\"model.name\"\n        required\n      />\n      <c8y-messages>\n        <c8y-message name=\"required\" text=\"{{ 'Server name is required' | translate }}\"></c8y-message>\n      </c8y-messages>\n    </c8y-form-group>\n  </div>\n\n    <div class=\"card-block large-padding bg-gray-lighter\">\n      <div class=\"tight-grid\">\n        <div class=\"col-sm-6\">\n          <label style=\"width: 100%;\" translate>Server connection</label>\n          <button\n            type=\"button\"\n            class=\"btn m-t-4\"\n            name=\"serverConnection\"\n            [(ngModel)]=\"targetConnectionState\"\n            (ngModelChange)=\"setServerConnection($event)\"\n            btnCheckbox\n            btnCheckboxTrue=\"1\"\n            btnCheckboxFalse=\"0\"\n          >\n            <span\n              title=\"{{ 'Enabled' | translate }}\"\n              [hidden]=\"targetConnectionState !== '1'\"\n              translate\n              >Enabled</span\n            >\n            <span\n              title=\"{{ 'Disabled' | translate }}\"\n              [hidden]=\"targetConnectionState !== '0'\"\n              translate\n              >Disabled</span\n            >\n          </button>\n        </div>\n        <div class=\"col-sm-6\">\n          <label translate>Connection status</label>\n          <div class=\"form-control-static\">\n            <device-status class=\"p-r-8\" [mo]=\"server\"></device-status>\n            <span>{{ connectionStatusLabel | translate }}</span>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"card-block large-padding\">\n      <!-- SERVER URL-->\n      <c8y-form-group>\n        <label for=\"configServerUrl\" translate>Server URL</label>\n        <input\n          type=\"text\"\n          class=\"form-control\"\n          id=\"configServerUrl\"\n          name=\"serverUrl\"\n          [(ngModel)]=\"model.config.serverUrl\"\n          c8yDefaultValidation=\"opcuaBrowsePath\"\n          required\n        />\n      </c8y-form-group>\n\n      <!-- TIMEOUT & STATUS-CHECK-INTERVAL-->\n      <div class=\"content-flex-32\">\n        <div class=\"col-6\">\n          <c8y-form-group>\n            <label for=\"config.timeout\" translate>Timeout</label>\n            <div class=\"input-group\">\n              <input\n                type=\"number\"\n                class=\"form-control\"\n                id=\"config.timeout\"\n                name=\"timeout\"\n                [min]=\"minIntervalNumber\"\n                placeholder=\"{{ 'e.g.' | translate }} 30\"\n                [(ngModel)]=\"model.config.timeout\"\n                required\n              />\n              <span class=\"input-group-addon units\" translate>\n                seconds\n              </span>\n            </div>\n          </c8y-form-group>\n        </div>\n        <div class=\"col-6\">\n          <c8y-form-group>\n            <label for=\"config.statusCheckInterval\" translate>Status check interval</label>\n            <div class=\"input-group\">\n              <input\n                type=\"number\"\n                class=\"form-control\"\n                id=\"config.statusCheckInterval\"\n                name=\"statusCheckInterval\"\n                [min]=\"minIntervalNumber\"\n                placeholder=\"{{ 'e.g.' | translate }} 40\"\n                [(ngModel)]=\"model.config.statusCheckInterval\"\n                required\n              />\n              <span class=\"input-group-addon units\" translate>\n                seconds\n              </span>\n            </div>\n          </c8y-form-group>\n        </div>\n      </div>\n\n      <!-- SECURITY MODE -->\n      <div class=\"tight-grid\">\n        <div class=\"col-md-6\">\n          <div class=\"form-group\">\n            <!-- NONE, SIGN, SIGN & ENCRYPT-->\n            <label for=\"config.securityMode\" translate>Security mode</label>\n            <div class=\"c8y-select-wrapper\">\n              <select\n                class=\"form-control\"\n                id=\"config.securityMode\"\n                [(ngModel)]=\"currentSecMode\"\n                (ngModelChange)=\"setPolicy($event)\"\n                name=\"securityMode\"\n                required\n              >\n                <option *ngFor=\"let mode of securityModes\" [ngValue]=\"mode\">{{ mode }}</option>\n              </select>\n              <span></span>\n            </div>\n          </div>\n        </div>\n        <div *ngIf=\"currentSecMode === NONE\" class=\"col-md-6\">\n          <div class=\"form-group\">\n            <label for=\"config.securityPolicy\" translate>Security policy</label>\n            <input\n              type=\"text\"\n              class=\"form-control\"\n              id=\"config.securityPolicy\"\n              name=\"securityPolicy\"\n              [readonly]=\"true\"\n              [(ngModel)]=\"model.config.securityMode\"\n              required\n            />\n          </div>\n        </div>\n        <div *ngIf=\"currentSecMode !== NONE\" class=\"col-md-6\">\n          <div class=\"form-group\">\n            <label for=\"config.securityPolicy\" translate>Security policy</label>\n            <div class=\"c8y-select-wrapper\">\n              <select\n                *ngIf=\"currentSecMode === SIGN\"\n                class=\"form-control\"\n                id=\"config.securityPolicy\"\n                [(ngModel)]=\"model.config.securityMode\"\n                name=\"securityPolicy\"\n                required\n              >\n                <option *ngFor=\"let policy of securityPolicies.sign\" [ngValue]=\"policy\">{{\n                  policy\n                }}</option>\n              </select>\n              <select\n                *ngIf=\"currentSecMode === SIGN_ENC\"\n                class=\"form-control\"\n                id=\"config.securityPolicy\"\n                [(ngModel)]=\"model.config.securityMode\"\n                name=\"securityPolicy\"\n                required\n              >\n                <option *ngFor=\"let policy of securityPolicies.sign_enc\" [ngValue]=\"policy\">{{\n                  policy\n                }}</option>\n              </select>\n              <span></span>\n            </div>\n          </div>\n        </div>\n      </div>\n\n      <!-- AUTHENTICATION -->\n      <div class=\"form-group\">\n        <label for=\"config.authenticationMode\" translate>Authentication</label>\n        <div class=\"c8y-select-wrapper\">\n          <select\n            class=\"form-control\"\n            id=\"config.authenticationMode\"\n            [(ngModel)]=\"authenticationMode\"\n            name=\"authenticationMode\"\n            (ngModelChange)=\"updateAuthentication($event)\"\n            required\n          >\n            <option *ngFor=\"let auth of authenticationModes\" [ngValue]=\"auth\">{{\n              auth.value | translate\n            }}</option>\n          </select>\n          <span></span>\n        </div>\n      </div>\n\n      <!-- User/Pw-->\n      <div *ngIf=\"authenticationMode.id === 2\" class=\"tight-grid\">\n        <div class=\"col-md-12\">\n          <div class=\"form-group\">\n            <label for=\"config.userName\" translate>Username</label>\n            <input\n              type=\"text\"\n              class=\"form-control\"\n              id=\"config.userName\"\n              name=\"userName\"\n              placeholder=\"{{ 'e.g. joe.doe`LOCALIZE`' | translate }}\"\n              [(ngModel)]=\"model.config.userName\"\n              autocomplete=\"new-password\"\n              required\n            />\n          </div>\n        </div>\n        <!-- change password section BEGINS-->\n        <div class=\"col-md-6\">\n          <div class=\"form-group\" *ngIf=\"!initialPasswordRequired\">\n            <button type=\"button\" class=\"btn btn-default\" (click)=\"toggleChangePassword()\">\n              <ng-container *ngIf=\"!changePassword\">\n                {{ 'Change password' | translate }}\n              </ng-container>\n              <ng-container *ngIf=\"changePassword\">\n                {{ 'Cancel password change' | translate }}\n              </ng-container>\n            </button>\n          </div>\n\n          <div class=\"form-group\">\n            <div *ngIf=\"changePassword\">\n              <label for=\"config.password\" translate>Password</label>\n              <input\n                type=\"password\"\n                class=\"form-control\"\n                id=\"config.userPassword\"\n                name=\"password\"\n                [(ngModel)]=\"model.config.userPassword\"\n                autocomplete=\"new-password\"\n                required\n              />\n            </div>\n          </div>\n        </div>\n        <!-- change password section ENDS-->\n      </div>\n      <!-- Key-based -->\n      <div *ngIf=\"authenticationMode.id === 3\" class=\"tight-grid\">\n        <!-- KEYSTORE PASSWORD -->\n        <div class=\"col-md-6\">\n          <div class=\"form-group\">\n            <label for=\"config.keystorePass\" translate>Keystore password</label>\n            <input\n              type=\"password\"\n              class=\"form-control\"\n              id=\"config.keystorePass\"\n              name=\"keystorePass\"\n              [(ngModel)]=\"model.config.keystorePass\"\n              required\n            />\n          </div>\n        </div>\n        <div class=\"col-md-6\">\n          <div class=\"form-group\">\n            <label for=\"config.certificatePass\" translate>Certificate password</label>\n            <input\n              type=\"password\"\n              class=\"form-control\"\n              id=\"config.certificatePass\"\n              name=\"keystorePass\"\n              [(ngModel)]=\"model.config.certificatePass\"\n              required\n            />\n          </div>\n        </div>\n        <!-- UPLOAD KEYSTORE -->\n        <div class=\"col-md-12\">\n          <div class=\"form-group\">\n            <label for=\"certificateUpload\" translate>Upload keystore</label>\n            <input\n              type=\"text\"\n              [readonly]=\"true\"\n              name=\"certificateUpload\"\n              class=\"form-control m-b-8\"\n              [ngModel]=\"fileName\"\n              placeholder=\"{{ 'e.g.' | translate }} yourKeystore.jks\"\n              required\n            />\n            <c8y-drop-area\n              (dropped)=\"uploadFile($event)\"\n              [loadingMessage]=\"'Importing, please wait.' | translate\"\n              [title]=\"'Import keystore with jks file extension' | translate\"\n            >\n            </c8y-drop-area>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"card-footer large-padding separator sticky-bottom\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"cancel()\" translate>\n      Cancel\n    </button>\n    <button title=\"{{ 'Remove' | translate }}\" class=\"btn btn-danger\" (click)=\"remove()\" translate>\n      Remove\n    </button>\n    <!-- Add [disabled]=\"method()\" when form is invalid-->\n    <button\n      title=\"{{ 'Save' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      [disabled]=\"!opcuaConfigForm.valid\"\n      translate\n    >\n      Save\n    </button>\n  </div> \n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.OpcuaService }]; }, { canceled: [{
            type: Output
        }], removed: [{
            type: Output
        }], saved: [{
            type: Output
        }], server: [{
            type: Input
        }], opcuaConfigForm: [{
            type: ViewChild,
            args: ['opcuaConfigForm', { static: false }]
        }], dropArea: [{
            type: ViewChild,
            args: [DropAreaComponent, { static: false }]
        }] }); })();
