// @c8y/ngx-components/esm2015/protocol-opcua/opcua-address-space.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter } from '@angular/core';
import { AddressSpaceService } from './address-space.service';
import { gettext } from '@c8y/ngx-components';
import { OpcuaService } from './opcuaService';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './address-space.service';
import * as ɵngcc2 from './opcuaService';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from './opcua-address-space-detail.component';
import * as ɵngcc6 from '@c8y/ngx-components';
import * as ɵngcc7 from './opcua-address-space-tree.component';

function OpcuaAddressSpaceComponent_button_8_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 13);
    ɵngcc0.ɵɵlistener("click", function OpcuaAddressSpaceComponent_button_8_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.searchNodes(); });
    ɵngcc0.ɵɵelement(1, "i", 14);
    ɵngcc0.ɵɵelementEnd();
} }
function OpcuaAddressSpaceComponent_button_9_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 15);
    ɵngcc0.ɵɵlistener("click", function OpcuaAddressSpaceComponent_button_9_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.clearSearch(); });
    ɵngcc0.ɵɵelement(1, "i", 16);
    ɵngcc0.ɵɵelementEnd();
} }
function OpcuaAddressSpaceComponent_div_10_p_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4, "\u00A0 ");
    ɵngcc0.ɵɵelementStart(5, "span", 19);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, ctx_r10.nodeList.resultLabel));
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r10.nodeList == null ? null : ctx_r10.nodeList.length);
} }
function OpcuaAddressSpaceComponent_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 17);
    ɵngcc0.ɵɵtemplate(1, OpcuaAddressSpaceComponent_div_10_p_1_Template, 7, 4, "p", 18);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.searchInProgress);
} }
function OpcuaAddressSpaceComponent_div_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelementStart(1, "div", 21);
    ɵngcc0.ɵɵelement(2, "div", 22);
    ɵngcc0.ɵɵelement(3, "div", 23);
    ɵngcc0.ɵɵelement(4, "div", 24);
    ɵngcc0.ɵɵelement(5, "div", 25);
    ɵngcc0.ɵɵelement(6, "div", 26);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function () { return { "list-group-item flex-row": true }; };
function OpcuaAddressSpaceComponent_div_12_div_1_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 31);
    ɵngcc0.ɵɵlistener("click", function OpcuaAddressSpaceComponent_div_12_div_1_button_1_Template_button_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r15); const nodeItem_r13 = restoredCtx.$implicit; const ctx_r14 = ɵngcc0.ɵɵnextContext(3); return ctx_r14.selectNode(nodeItem_r13); });
    ɵngcc0.ɵɵelementStart(1, "div", 32);
    ɵngcc0.ɵɵelement(2, "i", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 34);
    ɵngcc0.ɵɵelementStart(4, "span", 35);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const nodeItem_r13 = ctx.$implicit;
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction0(4, _c0));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r12.getIcon(nodeItem_r13.nodeClassName));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(nodeItem_r13.nodeId);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", nodeItem_r13.displayName, " ");
} }
function OpcuaAddressSpaceComponent_div_12_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 29);
    ɵngcc0.ɵɵtemplate(1, OpcuaAddressSpaceComponent_div_12_div_1_button_1_Template, 7, 5, "button", 30);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r11.nodeList);
} }
function OpcuaAddressSpaceComponent_div_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵtemplate(1, OpcuaAddressSpaceComponent_div_12_div_1_Template, 2, 1, "div", 28);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.isSearch && !ctx_r4.loading);
} }
function OpcuaAddressSpaceComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 27);
    ɵngcc0.ɵɵelementStart(1, "opcua-address-space-tree", 36);
    ɵngcc0.ɵɵlistener("selectedNode", function OpcuaAddressSpaceComponent_div_13_Template_opcua_address_space_tree_selectedNode_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.toggleCurrentNode($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("focusEmitter", ctx_r5.focusStatus);
} }
export class OpcuaAddressSpaceComponent {
    constructor(addressSpaceService, opcuaService) {
        this.addressSpaceService = addressSpaceService;
        this.opcuaService = opcuaService;
        this.selectednode = false;
        this.loading = false;
        this.searchInProgress = false;
        this.focusStatus = new EventEmitter();
        this.moId = '';
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.filterLabel = gettext('Filter…');
            this.moId = this.opcuaService.getMoId();
        });
    }
    ngOnDestroy() {
        // The BehaviourSubject will store the last array of ancestorNodes from the previous search
        // this would cause the component while subscribing in the init-phase to the subject to travers
        // to the last searched node again. From user perspective it does not make sense, because the user
        // left the Address space (tab) and should loose the context and just request a new search or
        // browse the tree manually.
        this.addressSpaceService.resetTreeToRootNode();
    }
    searchNodes() {
        return __awaiter(this, void 0, void 0, function* () {
            this.searchInProgress = true;
            this.clearNodeListAndCheckSearchString();
            if (this.isSearch) {
                this.currentNode = undefined;
                this.nodeList = yield this.addressSpaceService.getSearchedNodes(this.searchKey, this.moId);
                this.searchInProgress = false;
                this.nodeList.resultLabel = gettext('Results found');
            }
        });
    }
    clearNodeListAndCheckSearchString() {
        this.isSearch = this.searchKey !== undefined && this.searchKey !== '' ? true : false;
        if (!this.isSearch) {
            this.searchInProgress = false;
        }
    }
    clearSearch() {
        this.isSearch = false;
        this.searchKey = '';
        this.currentNode = undefined;
    }
    getIcon(nodeClassName) {
        return this.addressSpaceService.getIcon(nodeClassName);
    }
    selectNode(node) {
        return __awaiter(this, void 0, void 0, function* () {
            if (node && node.nodeId && node.nodeId.length > 0) {
                const res = yield this.addressSpaceService.getNodeById(this.moId, node.nodeId);
                this.toggleCurrentNode((yield res.json()));
            }
        });
    }
    toggleCurrentNode(node) {
        this.currentNode = this.isNodeSet(node) ? undefined : node;
    }
    backHandler(node) {
        this.isSearch = false;
        this.focusStatus.emit(node);
        this.toggleCurrentNode(node);
    }
    isNodeSet(node) {
        if (this.currentNode !== undefined && this.currentNode.nodeId === node.nodeId) {
            return true;
        }
        return false;
    }
}
OpcuaAddressSpaceComponent.ɵfac = function OpcuaAddressSpaceComponent_Factory(t) { return new (t || OpcuaAddressSpaceComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AddressSpaceService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OpcuaService)); };
OpcuaAddressSpaceComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaAddressSpaceComponent, selectors: [["opcua-address-space"]], outputs: { focusStatus: "focusStatus" }, decls: 15, vars: 11, consts: [[1, "row", "split-scroll"], [1, "col-md-5", "col-xs-12", "scroll-column", "no-gutter-r"], [1, "card", "bg-gray-lighter", "split-scroll", "flex-scroll"], [1, "card-block", "separator-bottom"], [1, "input-group", "input-group-search", 2, "margin", "-4px 0"], ["type", "search", 1, "form-control", 3, "placeholder", "ngModel", "keydown.enter", "ngModelChange"], [1, "input-group-btn"], ["type", "submit", "class", "btn btn-clean", 3, "click", 4, "ngIf"], ["class", "btn btn-clean", 3, "click", 4, "ngIf"], ["class", "p-t-16", 4, "ngIf"], ["class", "p-8", 4, "ngIf"], ["class", "flex-content-scroll", 4, "ngIf"], [1, "col-md-7", "col-xs-12", "scroll-column", "no-gutter-l", 2, "pointer-events", "none", 3, "node", "toggleAttrDetail"], ["type", "submit", 1, "btn", "btn-clean", 3, "click"], ["c8yIcon", "search"], [1, "btn", "btn-clean", 3, "click"], ["c8yIcon", "times"], [1, "p-t-16"], [4, "ngIf"], [1, "badge", "badge-info"], [1, "p-8"], [1, "spinner", "p-relative"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [1, "flex-content-scroll"], ["class", "list-group list-group-links", 4, "ngIf"], [1, "list-group", "list-group-links"], [3, "ngClass", "click", 4, "ngFor", "ngForOf"], [3, "ngClass", "click"], [1, "list-group-icon", "m-r-4"], [1, "m-r-4", 3, "c8yIcon"], [1, "list-item-body", "text-truncate"], ["tile", "nodeId"], [3, "focusEmitter", "selectedNode"]], template: function OpcuaAddressSpaceComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵelementStart(5, "input", 5);
        ɵngcc0.ɵɵlistener("keydown.enter", function OpcuaAddressSpaceComponent_Template_input_keydown_enter_5_listener() { return ctx.searchNodes(); })("ngModelChange", function OpcuaAddressSpaceComponent_Template_input_ngModelChange_5_listener($event) { return ctx.searchKey = $event; });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "span", 6);
        ɵngcc0.ɵɵtemplate(8, OpcuaAddressSpaceComponent_button_8_Template, 2, 0, "button", 7);
        ɵngcc0.ɵɵtemplate(9, OpcuaAddressSpaceComponent_button_9_Template, 2, 0, "button", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, OpcuaAddressSpaceComponent_div_10_Template, 2, 1, "div", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, OpcuaAddressSpaceComponent_div_11_Template, 7, 0, "div", 10);
        ɵngcc0.ɵɵtemplate(12, OpcuaAddressSpaceComponent_div_12_Template, 2, 1, "div", 11);
        ɵngcc0.ɵɵtemplate(13, OpcuaAddressSpaceComponent_div_13_Template, 2, 1, "div", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "opcua-address-space-detail", 12);
        ɵngcc0.ɵɵlistener("toggleAttrDetail", function OpcuaAddressSpaceComponent_Template_opcua_address_space_detail_toggleAttrDetail_14_listener($event) { return ctx.backHandler($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(6, 9, ctx.filterLabel));
        ɵngcc0.ɵɵproperty("ngModel", ctx.searchKey);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSearch);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSearch);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSearch && !ctx.loading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSearch && ctx.loading || ctx.searchInProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSearch && !ctx.loading && !ctx.searchInProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSearch);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("node", ctx.currentNode);
    } }, directives: [ɵngcc3.DefaultValueAccessor, ɵngcc3.NgControlStatus, ɵngcc3.NgModel, ɵngcc4.NgIf, ɵngcc5.OpcuaAddressSpaceDetailComponent, ɵngcc6.IconDirective, ɵngcc4.NgForOf, ɵngcc4.NgClass, ɵngcc7.OpcuaAddressSpaceTreeComponent], pipes: [ɵngcc6.C8yTranslatePipe], encapsulation: 2 });
OpcuaAddressSpaceComponent.ctorParameters = () => [
    { type: AddressSpaceService },
    { type: OpcuaService }
];
OpcuaAddressSpaceComponent.propDecorators = {
    focusStatus: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaAddressSpaceComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-address-space',
                template: "<div class=\"row split-scroll\">\n  <div class=\"col-md-5 col-xs-12 scroll-column no-gutter-r\">\n    <div class=\"card bg-gray-lighter split-scroll flex-scroll\">\n      <div class=\"card-block separator-bottom\">\n        <div class=\"input-group input-group-search\" style=\"margin: -4px 0;\">\n          <input\n            class=\"form-control\"\n            type=\"search\"\n            placeholder=\"{{ filterLabel | translate }}\"\n            (keydown.enter)=\"searchNodes()\"\n            [(ngModel)]=\"searchKey\"\n          />\n          <span class=\"input-group-btn\">\n            <button *ngIf=\"!isSearch\" type=\"submit\" class=\"btn btn-clean\" (click)=\"searchNodes()\">\n              <i c8yIcon=\"search\"></i>\n            </button>\n            <button *ngIf=\"isSearch\" class=\"btn btn-clean\" (click)=\"clearSearch()\">\n              <i c8yIcon=\"times\"></i>\n            </button>\n          </span>\n        </div>\n        <div *ngIf=\"isSearch && !loading\" class=\"p-t-16\">\n          <p *ngIf=\"!searchInProgress\">\n            <em>{{ nodeList.resultLabel | translate }}</em\n            >&nbsp;\n            <span class=\"badge badge-info\">{{ nodeList?.length }}</span>\n          </p>\n        </div>\n      </div>\n\n      <div class=\"p-8\" *ngIf=\"(isSearch && loading) || searchInProgress\">\n        <div class=\"spinner p-relative\">\n          <div class=\"rect1\"></div>\n          <div class=\"rect2\"></div>\n          <div class=\"rect3\"></div>\n          <div class=\"rect4\"></div>\n          <div class=\"rect5\"></div>\n        </div>\n      </div>\n\n      <div class=\"flex-content-scroll\" *ngIf=\"isSearch && !loading && !searchInProgress\">\n        <div\n          class=\"list-group list-group-links\"\n          *ngIf=\"isSearch && !loading\"\n          >\n          <button\n            *ngFor=\"let nodeItem of nodeList\"\n            (click)=\"selectNode(nodeItem)\"\n            [ngClass]=\"{ 'list-group-item flex-row': true }\"\n          >\n            <div class=\"list-group-icon m-r-4\">\n              <i class=\"m-r-4\" [c8yIcon]=\"getIcon(nodeItem.nodeClassName)\"></i>\n            </div>\n            <div class=\"list-item-body text-truncate\">\n              <span tile=\"nodeId\">{{ nodeItem.nodeId }}</span>\n              {{ nodeItem.displayName }}\n            </div>\n          </button>\n        </div>\n      </div>\n      <div class=\"flex-content-scroll\" *ngIf=\"!isSearch\">\n        <opcua-address-space-tree\n          (selectedNode)=\"toggleCurrentNode($event)\"\n          [focusEmitter]=\"focusStatus\"\n        ></opcua-address-space-tree>\n      </div>\n    </div>\n  </div>\n  <opcua-address-space-detail\n    [node]=\"currentNode\"\n    class=\"col-md-7 col-xs-12 scroll-column no-gutter-l\"\n    style=\"pointer-events: none;\"\n    (toggleAttrDetail)=\"backHandler($event)\"\n  ></opcua-address-space-detail>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AddressSpaceService }, { type: ɵngcc2.OpcuaService }]; }, { focusStatus: [{
            type: Output
        }] }); })();
