// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-object-mapping.component.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
import { UpgradeComponent } from '@angular/upgrade/static';
import * as ɵngcc0 from '@angular/core';
export class OpcuaDeviceProtocolObjectMapping extends UpgradeComponent {
    constructor(elementRef, injector) {
        super('c8yObjectMapping', elementRef, injector);
    }
}
OpcuaDeviceProtocolObjectMapping.ɵfac = function OpcuaDeviceProtocolObjectMapping_Factory(t) { return new (t || OpcuaDeviceProtocolObjectMapping)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
OpcuaDeviceProtocolObjectMapping.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: OpcuaDeviceProtocolObjectMapping, selectors: [["c8y-object-mapping"]], inputs: { mapping: "mapping", customActions: "customActions", autoObserveParams: "autoObserveParams", hideAutoObserve: "hideAutoObserve" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
OpcuaDeviceProtocolObjectMapping.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector }
];
OpcuaDeviceProtocolObjectMapping.propDecorators = {
    mapping: [{ type: Input }],
    customActions: [{ type: Input }],
    autoObserveParams: [{ type: Input }],
    hideAutoObserve: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolObjectMapping, [{
        type: Directive,
        args: [{
                selector: 'c8y-object-mapping'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Injector }]; }, { mapping: [{
            type: Input
        }], customActions: [{
            type: Input
        }], autoObserveParams: [{
            type: Input
        }], hideAutoObserve: [{
            type: Input
        }] }); })();
