// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-browse-path-validation.directive.js
import { Directive, forwardRef, ElementRef, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { some, isEqual, isEmpty } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
export class OpcuaDeviceProtocolBrowsePathValidation {
    constructor(el) {
        this.el = el;
    }
    validate(control) {
        if (control.value) {
            if (!this.isValidJson(control.value)) {
                return { invalidBrowsePathNotation: true };
            }
            else {
                if (this.isBrowsePathUnique(control.value)) {
                    return { browsePathNotUnique: true };
                }
            }
        }
        return null;
    }
    isValidJson(value) {
        try {
            const browsePath = JSON.parse(value);
            return !isEmpty(browsePath);
        }
        catch (error) {
            return false;
        }
    }
    toArray(str) {
        return JSON.parse(str);
    }
    isBrowsePathUnique(value) {
        const mappings = this.getMappings();
        const found = some(mappings, item => {
            if (isEqual(item.browsePath, this.toArray(value)) && item.id !== this.model.id) {
                return item;
            }
        });
        return found ? true : false;
    }
}
OpcuaDeviceProtocolBrowsePathValidation.ɵfac = function OpcuaDeviceProtocolBrowsePathValidation_Factory(t) { return new (t || OpcuaDeviceProtocolBrowsePathValidation)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
OpcuaDeviceProtocolBrowsePathValidation.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: OpcuaDeviceProtocolBrowsePathValidation, selectors: [["", "c8yBrowsePathValidator", "", "ngModel", ""]], inputs: { getMappings: "getMappings", model: "model" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: forwardRef(() => OpcuaDeviceProtocolBrowsePathValidation),
                multi: true
            }
        ])] });
OpcuaDeviceProtocolBrowsePathValidation.ctorParameters = () => [
    { type: ElementRef }
];
OpcuaDeviceProtocolBrowsePathValidation.propDecorators = {
    getMappings: [{ type: Input }],
    model: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolBrowsePathValidation, [{
        type: Directive,
        args: [{
                selector: '[c8yBrowsePathValidator][ngModel]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef(() => OpcuaDeviceProtocolBrowsePathValidation),
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { getMappings: [{
            type: Input
        }], model: [{
            type: Input
        }] }); })();
