// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-mapping.component.js
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ControlContainer, NgModelGroup } from '@angular/forms';
import { AlertService } from '@c8y/ngx-components';
import { isEmpty, assign, unset, get, set, cloneDeep, isEqual } from 'lodash-es';
import { OpcuaDeviceProtocolObjectMapping } from './opcua-device-protocol-object-mapping.component';
import { AddressSpaceService } from './address-space.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './address-space.service';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from './opcua-device-protocol-object-mapping-status-icon.component';
import * as ɵngcc5 from 'ngx-bootstrap/collapse';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from 'ngx-bootstrap/dropdown';
import * as ɵngcc8 from './opcua-device-protocol-browse-path-validation.directive';
import * as ɵngcc9 from './opcua-device-protocol-object-mapping.component';
import * as ɵngcc10 from './opcua-device-protocol-mapping-customaction.component';
import * as ɵngcc11 from './opcua-address-space-tree.component';
import * as ɵngcc12 from './opcua-device-protocol-data-reporting.component';

const _c0 = ["variableForm"];
const _c1 = ["browsePathModel"];
function OpcuaDeviceProtocolMapping_small_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 19);
    ɵngcc0.ɵɵpipe(1, "json");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "json");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r0.mapping.browsePath));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r0.mapping.browsePath));
} }
function OpcuaDeviceProtocolMapping_div_26_div_12_opcua_address_space_tree_1_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "opcua-address-space-tree", 48);
    ɵngcc0.ɵɵlistener("selectedNode", function OpcuaDeviceProtocolMapping_div_26_div_12_opcua_address_space_tree_1_Template_opcua_address_space_tree_selectedNode_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r9); ɵngcc0.ɵɵnextContext(2); const _r3 = ɵngcc0.ɵɵreference(7); const ctx_r8 = ɵngcc0.ɵɵnextContext(); ctx_r8.updateBrowsePath($event); return _r3.hide(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("node", ctx_r7.referencedNode)("moId", ctx_r7.referencedServerId);
} }
function OpcuaDeviceProtocolMapping_div_26_div_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 46);
    ɵngcc0.ɵɵtemplate(1, OpcuaDeviceProtocolMapping_div_26_div_12_opcua_address_space_tree_1_Template, 1, 2, "opcua-address-space-tree", 47);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.showAddressSpaceTree());
} }
function OpcuaDeviceProtocolMapping_div_26_div_44_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 49);
    ɵngcc0.ɵɵelement(1, "opcua-device-protocol-data-reporting", 50);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("model", ctx_r6.mapping);
} }
const _c2 = function () { return ["2:Node1", "2:SubNode1"]; };
function OpcuaDeviceProtocolMapping_div_26_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 20, 21);
    ɵngcc0.ɵɵelementStart(2, "div", 22);
    ɵngcc0.ɵɵelementStart(3, "c8y-form-group", 23);
    ɵngcc0.ɵɵelementStart(4, "label", 24);
    ɵngcc0.ɵɵtext(5, "Path");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 25, 26);
    ɵngcc0.ɵɵelementStart(8, "input", 27, 28);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMapping_div_26_Template_input_ngModelChange_8_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.browsePath = $event; })("change", function OpcuaDeviceProtocolMapping_div_26_Template_input_change_8_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.updateBrowsePathInput(); })("focus", function OpcuaDeviceProtocolMapping_div_26_Template_input_focus_8_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.setTreeFromRefNode(); });
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵpipe(11, "json");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(12, OpcuaDeviceProtocolMapping_div_26_div_12_Template, 2, 1, "div", 29);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "c8y-messages");
    ɵngcc0.ɵɵelement(14, "c8y-message", 30);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelement(16, "c8y-message", 31);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(18, "c8y-form-group", 32);
    ɵngcc0.ɵɵelementStart(19, "label", 24);
    ɵngcc0.ɵɵtext(20, "Name");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 33);
    ɵngcc0.ɵɵelementStart(22, "input", 34);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMapping_div_26_Template_input_ngModelChange_22_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.nodeDisplayName = $event; })("change", function OpcuaDeviceProtocolMapping_div_26_Template_input_change_22_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.updateDisplayname(); });
    ɵngcc0.ɵɵpipe(23, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "div", 35);
    ɵngcc0.ɵɵelementStart(25, "c8y-form-group", 36);
    ɵngcc0.ɵɵelementStart(26, "label");
    ɵngcc0.ɵɵelementStart(27, "span", 24);
    ɵngcc0.ɵɵtext(28, "Data reporting");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(29, "div", 33);
    ɵngcc0.ɵɵelementStart(30, "label", 37);
    ɵngcc0.ɵɵpipe(31, "translate");
    ɵngcc0.ɵɵelementStart(32, "input", 38);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMapping_div_26_Template_input_ngModelChange_32_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.dataReporting = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(33, "span");
    ɵngcc0.ɵɵelementStart(34, "span");
    ɵngcc0.ɵɵtext(35);
    ɵngcc0.ɵɵpipe(36, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(37, "label", 37);
    ɵngcc0.ɵɵpipe(38, "translate");
    ɵngcc0.ɵɵelementStart(39, "input", 39);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolMapping_div_26_Template_input_ngModelChange_39_listener($event) { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r17 = ɵngcc0.ɵɵnextContext(); return ctx_r17.dataReporting = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(40, "span");
    ɵngcc0.ɵɵelementStart(41, "span");
    ɵngcc0.ɵɵtext(42);
    ɵngcc0.ɵɵpipe(43, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(44, OpcuaDeviceProtocolMapping_div_26_div_44_Template, 2, 1, "div", 40);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(45, "c8y-object-mapping", 41);
    ɵngcc0.ɵɵelementStart(46, "div", 42);
    ɵngcc0.ɵɵelement(47, "opcua-device-protocol-mapping-customaction", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(48, "button", 44);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMapping_div_26_Template_button_click_48_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.cancel(); });
    ɵngcc0.ɵɵpipe(49, "translate");
    ɵngcc0.ɵɵtext(50);
    ɵngcc0.ɵɵpipe(51, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(52, "button", 45);
    ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMapping_div_26_Template_button_click_52_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.save(); });
    ɵngcc0.ɵɵpipe(53, "translate");
    ɵngcc0.ɵɵtext(54);
    ɵngcc0.ɵɵpipe(55, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r2 = ɵngcc0.ɵɵreference(1);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngModelGroup", ctx_r1.index);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("status", !ctx_r1.isBrowsePathUniq ? "error" : "");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("insideClick", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate2("placeholder", "", ɵngcc0.ɵɵpipeBind1(10, 29, "e.g."), " ", ɵngcc0.ɵɵpipeBind1(11, 31, ɵngcc0.ɵɵpureFunction0(55, _c2)), "");
    ɵngcc0.ɵɵproperty("getMappings", ctx_r1.getMappings)("model", ctx_r1.mapping)("ngModel", ctx_r1.browsePath);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(15, 33, "Must be a valid array of strings."));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(17, 35, "Variable with this path is already added."));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(23, 37, "e.g. childDevice2"), " ");
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.nodeDisplayName);
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(31, 39, "Default"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("name", ctx_r1.dataReportingName);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.dataReporting);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(36, 41, "Default"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(38, 43, "Custom"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("name", ctx_r1.dataReportingName);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.dataReporting);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(43, 45, "Custom"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.dataReporting === "custom");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("mapping", ctx_r1.mapping)("hideAutoObserve", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("customAction", ctx_r1.customAction);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(49, 47, "Cancel"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(51, 49, "Cancel"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(53, 51, "Save"));
    ɵngcc0.ɵɵproperty("disabled", !ctx_r1.canSave(_r2));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(55, 53, "Save"), " ");
} }
const _c3 = function (a0) { return { expanded: a0 }; };
const _c4 = function (a0) { return { active: a0 }; };
export class OpcuaDeviceProtocolMapping {
    constructor(alertService, addressSpaceService) {
        this.alertService = alertService;
        this.addressSpaceService = addressSpaceService;
        this.onAction = new EventEmitter();
        this.isPathFocused = false;
        this.isBrowsePathUniq = true;
        this.dataReporting = 'default';
        this.isTreeOpen = false;
        this.isNew = false;
        this.resetModel = false;
        this.moId = '';
        this.getMappings = () => this.getParentAttr('mappings');
    }
    toggleDetail() {
        this.isDetailOpen = !this.isDetailOpen;
        if (this.resetModel) {
            this.initialFormSetup();
        }
    }
    ngOnInit() {
        this.dataReportingName = 'ReportingMode' + this.index;
        this.initialFormSetup();
    }
    ngOnChanges(changes) {
        if (changes._model.previousValue &&
            !isEqual(this._model, changes._model.previousValue.SimpleChange)) {
            if (this.mapping && this.mapping.name === this._model.name) {
                this.mapping.id = this._model.id;
            }
        }
    }
    initialFormSetup() {
        const mapping = {
            id: '',
            browsePath: [],
            name: '',
            subscriptionType: {
                type: 'None'
            }
        };
        const customAction = {
            headers: [{ key: 'Authorization', value: '' }, { key: 'Content-Type', value: '' }],
            bodyTemplate: '',
            type: 'HttpPost',
            endpoint: ''
        };
        this.mapping = assign({}, mapping, cloneDeep(this._model));
        if (isEmpty(this.mapping.browsePath)) {
            this.isNew = true;
            this.isDetailOpen = true;
        }
        else {
            this.browsePath = this.stringfyBrowsePath(this.mapping.browsePath);
            this.nodeDisplayName = this.mapping.name;
        }
        if (this.referencedRootNodeId) {
            this.referencedNode = { nodeId: this.referencedRootNodeId };
            this.addressSpaceService.triggerNodeToOpen({
                node: {
                    nodeId: this.referencedRootNodeId,
                    children: [],
                    expanded: false,
                    absolutePaths: [[]]
                },
                selectedAncestorIds: []
            });
        }
        else {
            this.referencedNode = { nodeId: '' };
        }
        if (get(this.mapping, 'customAction')) {
            this.customAction = assign(customAction, get(this.mapping, 'customAction'));
            this.customAction.headers = this.mapHeadersObjectToList(get(this.customAction, 'headers'));
        }
        else {
            this.customAction = assign({}, customAction);
        }
        unset(this.mapping, 'customAction');
        if (get(this._model, 'subscriptionType')) {
            this.dataReporting = 'custom';
        }
        else {
            this.dataReporting = 'default';
        }
        this.resetModel = false;
    }
    showAddressSpaceTree() {
        return !isEmpty(this.referencedServerId);
    }
    ngAfterViewInit() {
        if (get(this.mapping, 'subscriptionType') &&
            get(this.mapping, 'subscriptionType.type') !== 'None') {
            this.dataReporting = 'custom';
        }
    }
    mapHeadersObjectToList(headers) {
        if (Object.keys(headers).length > 0) {
            return Object.keys(headers).map(item => {
                return { key: item, value: headers[item] };
            });
        }
    }
    stringfyBrowsePath(path) {
        return JSON.stringify(path);
    }
    updateBrowsePath(node) {
        this.mapping.browsePath = node.relativePath;
        this.nodeDisplayName = node.displayName;
        this.mapping.name = this.nodeDisplayName;
        this.browsePath = this.stringfyBrowsePath(this.mapping.browsePath);
        this.browsePathModel.control.markAsDirty();
    }
    updateDisplayname() {
        this.mapping.name = this.nodeDisplayName;
    }
    updateBrowsePathInput() {
        if (this.browsePath) {
            try {
                this.mapping.browsePath = JSON.parse(this.browsePath);
            }
            catch (error) {
                return;
            }
        }
    }
    save() {
        if (this.dataReporting === 'default') {
            unset(this.mapping, 'subscriptionType');
        }
        if (get(this.mapping, 'measurementCreation')) {
            const { measurementCreation } = this.mapping;
            set(measurementCreation, 'fragmentName', get(measurementCreation, 'type'));
        }
        const { customAction } = this.subFormRef.value;
        let modifiedCustomAction;
        if (customAction.hasCustomAction) {
            const reducedHeaders = this.customAction.headers.reduce((result, item) => {
                result[item.key] = item.value;
                return result;
            }, {});
            modifiedCustomAction = assign({}, this.customAction, { headers: reducedHeaders });
        }
        this.onAction.emit({
            action: 'save',
            data: assign({}, this.mapping, { customAction: modifiedCustomAction })
        });
        this.isDetailOpen = false;
    }
    cancel() {
        this.isDetailOpen = false;
        this.resetModel = true;
        if (this.mapping.id === 'new') {
            this.onAction.emit({ action: 'delete', data: assign({}, this.mapping) });
        }
    }
    onDelete() {
        this.onAction.emit({ action: 'delete', data: this.mapping });
    }
    canSave(variableForm) {
        const areValid = () => variableForm.valid && this.objectMappingForm.$componentScope.mappingForm.$valid;
        const areDirty = () => variableForm.dirty || this.objectMappingForm.$componentScope.mappingForm.$dirty;
        return areValid() && areDirty();
    }
    isActive() {
        return this.isDetailOpen;
    }
    setTreeFromRefNode() {
        if (this.referencedRootNodeId) {
            this.addressSpaceService.triggerNodeToOpen({
                node: {
                    nodeId: this.referencedRootNodeId,
                    children: [],
                    expanded: false,
                    absolutePaths: [[]]
                },
                selectedAncestorIds: []
            });
        }
    }
}
OpcuaDeviceProtocolMapping.ɵfac = function OpcuaDeviceProtocolMapping_Factory(t) { return new (t || OpcuaDeviceProtocolMapping)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AddressSpaceService)); };
OpcuaDeviceProtocolMapping.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaDeviceProtocolMapping, selectors: [["opcua-device-protocol-mapping"]], viewQuery: function OpcuaDeviceProtocolMapping_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(OpcuaDeviceProtocolObjectMapping, 5);
        ɵngcc0.ɵɵviewQuery(_c0, 5);
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.objectMappingForm = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.subFormRef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.browsePathModel = _t.first);
    } }, inputs: { _model: ["resource", "_model"], index: "index", getParentAttr: "getParentAttr", referencedServerId: "referencedServerId", referencedRootNodeId: "referencedRootNodeId" }, outputs: { onAction: "onAction" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgModelGroup }]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 27, vars: 18, consts: [[1, "list-group-item", "collapsible", 3, "ngClass"], [1, "flex-row", 3, "click"], [1, "list-item-actions"], [1, "btn", "btn-clean", "showOnHover", "flex-item-right", 3, "title"], ["c8yIcon", "minus-circle", 1, "text-danger", 3, "click"], ["type", "button", 1, "collapse-btn", 3, "title", "ngClass"], ["c8yIcon", "chevron-down"], [1, "list-item-icon"], ["c8yIcon", "sliders"], [1, "list-item-body"], [1, "row", "flex-row"], [1, "col-sm-7", "col-xs-12"], ["class", "text-muted text-truncate", 3, "title", 4, "ngIf"], [1, "col-sm-4", "col-xs-10"], [1, "list-functionalities"], ["translate", "", 1, "small", "m-r-8", "hidden-xs"], [3, "mapping"], [1, "detail", 3, "collapse", "isAnimated"], ["class", "form", 3, "ngModelGroup", 4, "ngIf"], [1, "text-muted", "text-truncate", 3, "title"], [1, "form", 3, "ngModelGroup"], ["variableForm", "ngModelGroup"], [1, "row", "p-t-8"], [1, "col-md-4", 3, "status"], ["translate", ""], ["dropdown", "", 1, "dropdown", 2, "width", "100%", 3, "insideClick"], ["dropdown", "bs-dropdown"], ["c8yBrowsePathValidator", "", "type", "text", "name", "browsePath", "dropdownToggle", "", "required", "", "autocomplete", "off", 1, "form-control", 3, "getMappings", "model", "placeholder", "ngModel", "ngModelChange", "change", "focus"], ["browsePathModel", "ngModel"], ["class", "dropdown-menu panel-inner-scroll", "style", "max-height:200px; width: 100%;", 4, "dropdownMenu"], ["name", "invalidBrowsePathNotation", 3, "text"], ["name", "browsePathNotUnique", 3, "text"], [1, "col-md-4"], [1, "input-group"], ["type", "test", "name", "displayName", "required", "", "autocomplete", "off", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange", "change"], ["ngModelGroup", "dataReportingSection", 1, "row"], [1, "col-sm-4", "col-md-3", "col-lg-2"], [1, "c8y-radio", "radio-inline", 3, "title"], ["type", "radio", "value", "default", 3, "ngModel", "name", "ngModelChange"], ["type", "radio", "value", "custom", 3, "ngModel", "name", "ngModelChange"], ["class", "col-sm-8 col-md-9 col-lg-10", "ngModelGroup", "overriddenSubscription", 4, "ngIf"], [3, "mapping", "hideAutoObserve"], ["ngModelGroup", "customAction"], [3, "customAction"], ["id", "cancelBtn", 1, "btn", "btn-default", "m-t-16", "m-b-16", 3, "title", "click"], ["id", "saveBtn", 1, "btn", "btn-primary", "m-t-16", "m-b-16", 3, "title", "disabled", "click"], [1, "dropdown-menu", "panel-inner-scroll", 2, "max-height", "200px", "width", "100%"], [3, "node", "moId", "selectedNode", 4, "ngIf"], [3, "node", "moId", "selectedNode"], ["ngModelGroup", "overriddenSubscription", 1, "col-sm-8", "col-md-9", "col-lg-10"], [3, "model"]], template: function OpcuaDeviceProtocolMapping_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMapping_Template_div_click_1_listener() { return ctx.toggleDetail(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "button", 3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementStart(5, "i", 4);
        ɵngcc0.ɵɵlistener("click", function OpcuaDeviceProtocolMapping_Template_i_click_5_listener() { return ctx.onDelete(); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "button", 5);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelement(8, "i", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelement(10, "i", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵelementStart(12, "div", 10);
        ɵngcc0.ɵɵelementStart(13, "div", 11);
        ɵngcc0.ɵɵelementStart(14, "p");
        ɵngcc0.ɵɵtext(15);
        ɵngcc0.ɵɵelement(16, "br");
        ɵngcc0.ɵɵtemplate(17, OpcuaDeviceProtocolMapping_small_17_Template, 4, 6, "small", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(18, "p");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "div", 13);
        ɵngcc0.ɵɵelementStart(20, "div", 14);
        ɵngcc0.ɵɵelementStart(21, "label", 15);
        ɵngcc0.ɵɵtext(22, "Functionalities");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(23, "\u00A0 ");
        ɵngcc0.ɵɵelement(24, "c8y-object-mapping-status-icons", 16);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "div", 17);
        ɵngcc0.ɵɵtemplate(26, OpcuaDeviceProtocolMapping_div_26_Template, 56, 56, "div", 18);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c3, ctx.isDetailOpen));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 10, "Delete"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 12, "Expand"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(16, _c4, ctx.isDetailOpen));
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.nodeDisplayName, "");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mapping.browsePath.length > 0);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("mapping", ctx.mapping);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("collapse", !ctx.isDetailOpen)("isAnimated", true);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDetailOpen);
    } }, directives: [ɵngcc3.NgClass, ɵngcc1.IconDirective, ɵngcc3.NgIf, ɵngcc1.C8yTranslateDirective, ɵngcc4.OpcuaDeviceProtocolObjectMappingStatus, ɵngcc5.CollapseDirective, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgModelGroup, ɵngcc1.FormGroupComponent, ɵngcc7.BsDropdownDirective, ɵngcc1.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc8.OpcuaDeviceProtocolBrowsePathValidation, ɵngcc7.BsDropdownToggleDirective, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.BsDropdownMenuDirective, ɵngcc1.MessagesComponent, ɵngcc1.MessageDirective, ɵngcc6.RadioControlValueAccessor, ɵngcc9.OpcuaDeviceProtocolObjectMapping, ɵngcc10.OpcuaDeviceProtocolMappingCustomAction, ɵngcc11.OpcuaAddressSpaceTreeComponent, ɵngcc12.OpcuaDeviceProtocolDataReportingComponent], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc3.JsonPipe], encapsulation: 2 });
OpcuaDeviceProtocolMapping.ctorParameters = () => [
    { type: AlertService },
    { type: AddressSpaceService }
];
OpcuaDeviceProtocolMapping.propDecorators = {
    objectMappingForm: [{ type: ViewChild, args: [OpcuaDeviceProtocolObjectMapping, { static: false },] }],
    subFormRef: [{ type: ViewChild, args: ['variableForm', { static: false },] }],
    browsePathModel: [{ type: ViewChild, args: ['browsePathModel', { static: false },] }],
    _model: [{ type: Input, args: ['resource',] }],
    index: [{ type: Input }],
    getParentAttr: [{ type: Input }],
    referencedServerId: [{ type: Input }],
    referencedRootNodeId: [{ type: Input }],
    onAction: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolMapping, [{
        type: Component,
        args: [{
                selector: 'opcua-device-protocol-mapping',
                template: "<div class=\"list-group-item collapsible\" [ngClass]=\"{ expanded: isDetailOpen }\">\n  <div class=\"flex-row\" (click)=\"toggleDetail()\">\n    <div class=\"list-item-actions\">\n      <button class=\"btn btn-clean showOnHover flex-item-right\" title=\"{{ 'Delete' | translate }}\">\n        <i c8yIcon=\"minus-circle\" class=\"text-danger\" (click)=\"onDelete()\"></i>\n      </button>\n      <button\n        type=\"button\"\n        title=\"{{ 'Expand' | translate }}\"\n        class=\"collapse-btn\"\n        [ngClass]=\"{ active: isDetailOpen }\"\n      >\n        <i c8yIcon=\"chevron-down\"></i>\n      </button>\n    </div>\n\n    <div class=\"list-item-icon\">\n      <i c8yIcon=\"sliders\"></i>\n    </div>\n\n    <div class=\"list-item-body\">\n      <div class=\"row flex-row\">\n        <div class=\"col-sm-7 col-xs-12\">\n          <p>\n            {{ nodeDisplayName }}<br />\n            <small\n              *ngIf=\"mapping.browsePath.length > 0\"\n              class=\"text-muted text-truncate\"\n              title=\"{{ mapping.browsePath | json }}\"\n              >{{ mapping.browsePath | json }}</small\n            >\n          </p>\n          <p></p>\n        </div>\n        <div class=\"col-sm-4 col-xs-10\">\n          <div class=\"list-functionalities\">\n            <label class=\"small m-r-8 hidden-xs\" translate>Functionalities</label>&nbsp;\n            <c8y-object-mapping-status-icons [mapping]=\"mapping\"></c8y-object-mapping-status-icons>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"detail\" [collapse]=\"!isDetailOpen\" [isAnimated]=\"true\">\n    <div class=\"form\" [ngModelGroup]=\"index\" #variableForm=\"ngModelGroup\" *ngIf=\"isDetailOpen\">\n      <div class=\"row p-t-8\">\n        <c8y-form-group class=\"col-md-4\" [status]=\"!isBrowsePathUniq ? 'error' : ''\">\n          <label translate>Path</label>\n          <div\n            class=\"dropdown\"\n            dropdown\n            #dropdown=\"bs-dropdown\"\n            [insideClick]=\"true\"\n            style=\"width:100%;\"\n          >\n            <input\n              class=\"form-control\"\n              c8yBrowsePathValidator\n              [getMappings]=\"getMappings\"\n              [model]=\"mapping\"\n              type=\"text\"\n              name=\"browsePath\"\n              dropdownToggle\n              placeholder=\"{{ 'e.g.' | translate }} {{ ['2:Node1', '2:SubNode1'] | json }}\"\n              [(ngModel)]=\"browsePath\"\n              (change)=\"updateBrowsePathInput()\"\n              (focus)=\"setTreeFromRefNode()\"\n              required\n              #browsePathModel=\"ngModel\"\n              autocomplete=\"off\"\n            />\n            <div\n              *dropdownMenu\n              class=\"dropdown-menu panel-inner-scroll\"\n              style=\"max-height:200px; width: 100%;\"\n            >\n              <opcua-address-space-tree\n                *ngIf=\"showAddressSpaceTree()\"\n                [node]=\"referencedNode\"\n                [moId]=\"referencedServerId\"\n                (selectedNode)=\"updateBrowsePath($event); dropdown.hide()\"\n              ></opcua-address-space-tree>\n            </div>\n          </div>\n          <c8y-messages>\n            <c8y-message\n              name=\"invalidBrowsePathNotation\"\n              text=\"{{ 'Must be a valid array of strings.' | translate }}\"\n            ></c8y-message>\n            <c8y-message\n              name=\"browsePathNotUnique\"\n              text=\"{{ 'Variable with this path is already added.' | translate }}\"\n            ></c8y-message>\n          </c8y-messages>\n        </c8y-form-group>\n\n        <c8y-form-group class=\"col-md-4\">\n          <label translate>Name</label>\n          <div class=\"input-group\">\n            <input\n              class=\"form-control\"\n              type=\"test\"\n              name=\"displayName\"\n              placeholder=\"{{ 'e.g. childDevice2' | translate }} \"\n              required\n              [(ngModel)]=\"nodeDisplayName\"\n              (change)=\"updateDisplayname()\"\n              autocomplete=\"off\"\n            />\n          </div>\n        </c8y-form-group>\n      </div>\n      <div class=\"row\" ngModelGroup=\"dataReportingSection\">\n        <c8y-form-group class=\"col-sm-4 col-md-3 col-lg-2\">\n          <label>\n            <span translate>Data reporting</span>\n          </label>\n          <div class=\"input-group\">\n            <label title=\"{{ 'Default' | translate }}\" class=\"c8y-radio radio-inline\">\n              <input\n                type=\"radio\"\n                [(ngModel)]=\"dataReporting\"\n                name=\"{{ dataReportingName }}\"\n                value=\"default\"\n              />\n              <span></span>\n              <span>{{ 'Default' | translate }}</span>\n            </label>\n            <label title=\"{{ 'Custom' | translate }}\" class=\"c8y-radio radio-inline\">\n              <input\n                type=\"radio\"\n                [(ngModel)]=\"dataReporting\"\n                name=\"{{ dataReportingName }}\"\n                value=\"custom\"\n              />\n              <span></span>\n              <span>{{ 'Custom' | translate }}</span>\n            </label>\n          </div>\n        </c8y-form-group>\n        <div\n          class=\"col-sm-8 col-md-9 col-lg-10\"\n          *ngIf=\"dataReporting === 'custom'\"\n          ngModelGroup=\"overriddenSubscription\"\n        >\n          <opcua-device-protocol-data-reporting\n            [model]=\"mapping\"\n          ></opcua-device-protocol-data-reporting>\n        </div>\n      </div>\n\n      <c8y-object-mapping [mapping]=\"mapping\" [hideAutoObserve]=\"true\"></c8y-object-mapping>\n      <div ngModelGroup=\"customAction\">\n        <opcua-device-protocol-mapping-customaction\n          [customAction]=\"customAction\"\n        ></opcua-device-protocol-mapping-customaction>\n      </div>\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        id=\"cancelBtn\"\n        class=\"btn btn-default m-t-16 m-b-16\"\n        (click)=\"cancel()\"\n      >\n        {{ 'Cancel' | translate }}\n      </button>\n      <button\n        title=\"{{ 'Save' | translate }}\"\n        id=\"saveBtn\"\n        class=\"btn btn-primary m-t-16 m-b-16\"\n        (click)=\"save()\"\n        [disabled]=\"!canSave(variableForm)\"\n      >\n        {{ 'Save' | translate }}\n      </button>\n    </div>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgModelGroup }]
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: ɵngcc2.AddressSpaceService }]; }, { onAction: [{
            type: Output
        }], objectMappingForm: [{
            type: ViewChild,
            args: [OpcuaDeviceProtocolObjectMapping, { static: false }]
        }], subFormRef: [{
            type: ViewChild,
            args: ['variableForm', { static: false }]
        }], browsePathModel: [{
            type: ViewChild,
            args: ['browsePathModel', { static: false }]
        }], _model: [{
            type: Input,
            args: ['resource']
        }], index: [{
            type: Input
        }], getParentAttr: [{
            type: Input
        }], referencedServerId: [{
            type: Input
        }], referencedRootNodeId: [{
            type: Input
        }] }); })();
