// @c8y/ngx-components/esm2015/protocol-opcua/opcua-device-protocol-data-reporting.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { set, unset, toInteger } from 'lodash';
import { gettext } from '@c8y/ngx-components';
import { ControlContainer, NgModelGroup } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-bootstrap/collapse';
import * as ɵngcc4 from '@angular/forms';

function OpcuaDeviceProtocolDataReportingComponent_label_3_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 15);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "input", 16);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_label_3_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.subscription.type = $event; })("change", function OpcuaDeviceProtocolDataReportingComponent_label_3_Template_input_change_2_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.updateModel(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const mechanism_r4 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 5, mechanism_r4.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("name", ctx_r0.subscriptionTypeName);
    ɵngcc0.ɵɵproperty("value", mechanism_r4.value)("ngModel", ctx_r0.subscription.type);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, mechanism_r4.label));
} }
function OpcuaDeviceProtocolDataReportingComponent_label_38_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 15);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "input", 17);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_label_38_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.subscriptionParameters.discardOldest = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const discardOption_r8 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 5, discardOption_r8.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("value", discardOption_r8.value)("ngModel", ctx_r1.subscriptionParameters.discardOldest)("required", ctx_r1.requireSubscription);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, discardOption_r8.label));
} }
function OpcuaDeviceProtocolDataReportingComponent_label_44_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 15);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "input", 18);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_label_44_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.subscriptionParameters.dataChangeTrigger = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const trigger_r11 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 5, trigger_r11.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("value", trigger_r11.value)("ngModel", ctx_r2.subscriptionParameters.dataChangeTrigger)("required", ctx_r2.requireSubscription);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, trigger_r11.label));
} }
function OpcuaDeviceProtocolDataReportingComponent_label_49_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 15);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementStart(2, "input", 19);
    ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_label_49_Template_input_ngModelChange_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.subscriptionParameters.deadbandType = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "span");
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const deadbandFilter_r14 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 5, deadbandFilter_r14.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("value", deadbandFilter_r14.value)("ngModel", ctx_r3.subscriptionParameters.deadbandType)("required", ctx_r3.requireSubscription);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, deadbandFilter_r14.label));
} }
export class OpcuaDeviceProtocolDataReportingComponent {
    constructor() {
        this.onSubscriptionChange = new EventEmitter();
        this.subscription = {
            type: 'None'
        };
        this.subscriptionParameters = {
            samplingRate: undefined,
            deadbandType: 'None',
            deadbandValue: undefined,
            ranges: '',
            queueSize: undefined,
            dataChangeTrigger: 'Status',
            discardOldest: true
        };
        this.cyclicReadParameters = {
            rate: undefined
        };
        this.types = [
            { value: 'None', label: gettext('None') },
            { value: 'CyclicRead', label: gettext('Cyclic read') },
            { value: 'Subscription', label: gettext('Subscription') }
        ];
        this.filters = [
            { value: 'None', label: gettext('None') },
            { value: 'absolute', label: gettext('Absolute') },
            { value: 'percent', label: gettext('Percent') }
        ];
        this.triggers = [
            { value: 'Status', label: gettext('Status') },
            { value: 'StatusValue', label: gettext('Status/Value') },
            { value: 'StatusValueTimestamp', label: gettext('Status/Value/Timestamp') }
        ];
        this.discard = [
            { value: true, label: gettext('oldest`data`') },
            { value: false, label: gettext('newest`data`') }
        ];
        this.requireCyclic = false;
        this.requireSubscription = false;
        this.subscriptionTypeName = 'subscriptionType';
        this.parseReadingInterval = $event => toInteger($event.target.value);
    }
    set model(_model) {
        if (_model.subscriptionType) {
            this.subscription = _model.subscriptionType;
            if (this.subscription.type === 'CyclicRead') {
                this.cyclicReadParameters = _model.subscriptionType.cyclicReadParameters;
                this.requireCyclic = true;
                this.requireSubscription = false;
            }
            if (this.subscription.type === 'Subscription') {
                this.subscriptionParameters = _model.subscriptionType.subscriptionParameters;
                this.requireCyclic = false;
                this.requireSubscription = true;
            }
            if (this.subscription.type === 'None') {
                this.requireCyclic = false;
                this.requireSubscription = false;
            }
        }
        this._model = _model;
        this.subscriptionTypeName = 'subscriptionType' + _model.id;
    }
    updateModel() {
        setTimeout(() => {
            unset(this.subscription, 'subscriptionParameters');
            unset(this.subscription, 'cyclicReadParameters');
            if (this.subscription.type === 'CyclicRead') {
                this.requireCyclic = true;
                this.requireSubscription = false;
                set(this.subscription, 'cyclicReadParameters', this.cyclicReadParameters);
            }
            else if (this.subscription.type === 'Subscription') {
                this.requireCyclic = false;
                this.requireSubscription = true;
                set(this.subscription, 'subscriptionParameters', this.subscriptionParameters);
            }
            else if (this.subscription.type === 'None') {
                this.requireCyclic = false;
                this.requireSubscription = false;
            }
        });
    }
}
OpcuaDeviceProtocolDataReportingComponent.ɵfac = function OpcuaDeviceProtocolDataReportingComponent_Factory(t) { return new (t || OpcuaDeviceProtocolDataReportingComponent)(); };
OpcuaDeviceProtocolDataReportingComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: OpcuaDeviceProtocolDataReportingComponent, selectors: [["opcua-device-protocol-data-reporting"]], inputs: { model: "model", groupName: "groupName" }, outputs: { onSubscriptionChange: "onSubscriptionChange" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgModelGroup }])], decls: 56, vars: 38, consts: [["translate", ""], ["class", "c8y-radio radio-inline", 3, "title", 4, "ngFor", "ngForOf"], [1, "row", "collapse", 3, "collapse", "isAnimated"], [1, "col-sm-6", "col-md-4", "col-lg-3"], [1, "input-group"], ["type", "number", "name", "rate", "min", "50", 1, "form-control", 3, "ngModel", "placeholder", "required", "ngModelChange"], [1, "input-group-addon", "units", 3, "title"], [1, "col-sm-6", "col-lg-4"], ["type", "number", "name", "samplingRate", "min", "0", 1, "form-control", 3, "ngModel", "placeholder", "required", "ngModelChange"], ["translate", "", 1, "input-group-addon", "units", 3, "title"], ["type", "number", "name", "queueSize", "min", "1", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], [1, "input-group-addon"], [1, "clearfix"], [3, "collapse", "isAnimated"], ["type", "number", "name", "deadbandValue", "min", "0", 1, "form-control", 3, "ngModel", "placeholder", "required", "ngModelChange"], [1, "c8y-radio", "radio-inline", 3, "title"], ["type", "radio", "required", "", 3, "name", "value", "ngModel", "ngModelChange", "change"], ["type", "radio", "name", "discard", 3, "value", "ngModel", "required", "ngModelChange"], ["type", "radio", "name", "dataChangeTrigger", 3, "value", "ngModel", "required", "ngModelChange"], ["type", "radio", "name", "deadbandType", 3, "value", "ngModel", "required", "ngModelChange"]], template: function OpcuaDeviceProtocolDataReportingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(1, "label", 0);
        ɵngcc0.ɵɵtext(2, "Mechanism");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, OpcuaDeviceProtocolDataReportingComponent_label_3_Template, 7, 9, "label", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 2);
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(7, "label", 0);
        ɵngcc0.ɵɵtext(8, "Reading interval");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 4);
        ɵngcc0.ɵɵelementStart(10, "input", 5);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_Template_input_ngModelChange_10_listener($event) { return ctx.cyclicReadParameters.rate = $event; });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "span", 6);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 2);
        ɵngcc0.ɵɵelementStart(17, "div", 7);
        ɵngcc0.ɵɵelementStart(18, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(19, "label", 0);
        ɵngcc0.ɵɵtext(20, "Sampling interval");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 4);
        ɵngcc0.ɵɵelementStart(22, "input", 8);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_Template_input_ngModelChange_22_listener($event) { return ctx.subscriptionParameters.samplingRate = $event; });
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "span", 9);
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵtext(26, " msec ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "div", 7);
        ɵngcc0.ɵɵelementStart(28, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(29, "label", 0);
        ɵngcc0.ɵɵtext(30, "Queue size");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "div", 4);
        ɵngcc0.ɵɵelementStart(32, "input", 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_Template_input_ngModelChange_32_listener($event) { return ctx.subscriptionParameters.queueSize = $event; });
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(34, "span", 11);
        ɵngcc0.ɵɵelementStart(35, "strong", 0);
        ɵngcc0.ɵɵtext(36, "Discard");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(37, "\u00A0 ");
        ɵngcc0.ɵɵtemplate(38, OpcuaDeviceProtocolDataReportingComponent_label_38_Template, 7, 9, "label", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(39, "div", 12);
        ɵngcc0.ɵɵelementStart(40, "div", 7);
        ɵngcc0.ɵɵelementStart(41, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(42, "label", 0);
        ɵngcc0.ɵɵtext(43, "Data change trigger");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(44, OpcuaDeviceProtocolDataReportingComponent_label_44_Template, 7, 9, "label", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "div", 7);
        ɵngcc0.ɵɵelementStart(46, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(47, "label", 0);
        ɵngcc0.ɵɵtext(48, "Deadband filter");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(49, OpcuaDeviceProtocolDataReportingComponent_label_49_Template, 7, 9, "label", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(50, "div", 13);
        ɵngcc0.ɵɵelementStart(51, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(52, "label", 0);
        ɵngcc0.ɵɵtext(53, "Deadband value");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(54, "input", 14);
        ɵngcc0.ɵɵlistener("ngModelChange", function OpcuaDeviceProtocolDataReportingComponent_Template_input_ngModelChange_54_listener($event) { return ctx.subscriptionParameters.deadbandValue = $event; });
        ɵngcc0.ɵɵpipe(55, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.types);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("collapse", ctx.subscription.type != "CyclicRead")("isAnimated", true);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(11, 24, "e.g."), " 50");
        ɵngcc0.ɵɵproperty("ngModel", ctx.cyclicReadParameters.rate)("required", ctx.requireCyclic);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(13, 26, "Milliseconds"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 28, "msec"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("collapse", ctx.subscription.type != "Subscription")("isAnimated", true);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(23, 30, "e.g."), " 500");
        ɵngcc0.ɵɵproperty("ngModel", ctx.subscriptionParameters.samplingRate)("required", ctx.requireSubscription);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(25, 32, "Milliseconds"));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(33, 34, "e.g."), " 10");
        ɵngcc0.ɵɵproperty("ngModel", ctx.subscriptionParameters.queueSize);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.discard);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.triggers);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.filters);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("collapse", ctx.subscriptionParameters.deadbandType == "None")("isAnimated", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(55, 36, "e.g."), " 10");
        ɵngcc0.ɵɵproperty("ngModel", ctx.subscriptionParameters.deadbandValue)("required", ctx.subscriptionParameters.deadbandType != "None");
    } }, directives: [ɵngcc1.FormGroupComponent, ɵngcc1.C8yTranslateDirective, ɵngcc2.NgForOf, ɵngcc3.CollapseDirective, ɵngcc4.MinValidator, ɵngcc4.NumberValueAccessor, ɵngcc4.DefaultValueAccessor, ɵngcc1.RequiredInputPlaceholderDirective, ɵngcc1.MinValidationDirective, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc4.RequiredValidator, ɵngcc4.RadioControlValueAccessor], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
OpcuaDeviceProtocolDataReportingComponent.propDecorators = {
    model: [{ type: Input }],
    groupName: [{ type: Input }],
    onSubscriptionChange: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaDeviceProtocolDataReportingComponent, [{
        type: Component,
        args: [{
                selector: 'opcua-device-protocol-data-reporting',
                template: "<c8y-form-group>\n  <label translate>Mechanism</label>\n  <label\n    title=\"{{ mechanism.label | translate }}\"\n    class=\"c8y-radio radio-inline\"\n    *ngFor=\"let mechanism of types\"\n  >\n    <input\n      type=\"radio\"\n      name=\"{{subscriptionTypeName}}\"\n      [value]=\"mechanism.value\"\n      [(ngModel)]=\"subscription.type\"\n      (change)=\"updateModel()\"\n      required\n    />\n    <span></span>\n    <span>{{ mechanism.label | translate }}</span>\n  </label>\n</c8y-form-group>\n\n<div class=\"row collapse\" [collapse]=\"subscription.type != 'CyclicRead'\" [isAnimated]=\"true\">\n  <div class=\"col-sm-6 col-md-4 col-lg-3\">\n    <c8y-form-group>\n      <label translate>Reading interval</label>\n      <div class=\"input-group\">\n        <input\n          class=\"form-control\"\n          type=\"number\"\n          name=\"rate\"\n          [(ngModel)]=\"cyclicReadParameters.rate\"\n          placeholder=\"{{ 'e.g.' | translate }} 50\"\n          [required]=\"requireCyclic\"\n          min=\"50\"\n        />\n        <span class=\"input-group-addon units\" title=\"{{ 'Milliseconds' | translate }}\">\n          {{ 'msec' | translate }}\n        </span>\n      </div>\n    </c8y-form-group>\n  </div>\n</div>\n\n<div class=\"row collapse\" [collapse]=\"subscription.type != 'Subscription'\" [isAnimated]=\"true\">\n  <div class=\"col-sm-6 col-lg-4\">\n    <c8y-form-group>\n      <label translate>Sampling interval</label>\n      <div class=\"input-group\">\n        <input\n          class=\"form-control\"\n          type=\"number\"\n          name=\"samplingRate\"\n          [(ngModel)]=\"subscriptionParameters.samplingRate\"\n          placeholder=\"{{ 'e.g.' | translate }} 500\"\n          [required]=\"requireSubscription\"\n          min=\"0\"\n        />\n        <span class=\"input-group-addon units\" title=\"{{ 'Milliseconds' | translate }}\" translate>\n          msec\n        </span>\n      </div>\n    </c8y-form-group>\n  </div>\n  <div class=\"col-sm-6 col-lg-4\">\n    <c8y-form-group>\n      <label translate>Queue size</label>\n      <div class=\"input-group\">\n        <input\n          class=\"form-control\"\n          type=\"number\"\n          name=\"queueSize\"\n          [(ngModel)]=\"subscriptionParameters.queueSize\"\n          placeholder=\"{{ 'e.g.' | translate }} 10\"\n          min=\"1\"\n        />\n        <span class=\"input-group-addon\">\n          <strong translate>Discard</strong>&nbsp;\n          <label\n            *ngFor=\"let discardOption of discard\"\n            title=\"{{ discardOption.label | translate }}\"\n            class=\"c8y-radio radio-inline\"\n          >\n            <input\n              type=\"radio\"\n              name=\"discard\"\n              [value]=\"discardOption.value\"\n              [(ngModel)]=\"subscriptionParameters.discardOldest\"\n              [required]=\"requireSubscription\"\n            />\n            <span></span>\n            <span>{{ discardOption.label | translate }}</span>\n          </label>\n        </span>\n      </div>\n    </c8y-form-group>\n  </div>\n  <div class=\"clearfix\"></div>\n  <div class=\"col-sm-6 col-lg-4\">\n    <c8y-form-group>\n      <label translate>Data change trigger</label>\n      <label\n        *ngFor=\"let trigger of triggers\"\n        title=\"{{ trigger.label | translate }}\"\n        class=\"c8y-radio radio-inline\"\n      >\n        <input\n          type=\"radio\"\n          name=\"dataChangeTrigger\"\n          [value]=\"trigger.value\"\n          [(ngModel)]=\"subscriptionParameters.dataChangeTrigger\"\n          [required]=\"requireSubscription\"\n        />\n        <span></span>\n        <span>{{ trigger.label | translate }}</span>\n      </label>\n    </c8y-form-group>\n  </div>\n  <div class=\"col-sm-6 col-lg-4\">\n    <c8y-form-group>\n      <label translate>Deadband filter</label>\n      <label\n        *ngFor=\"let deadbandFilter of filters\"\n        title=\"{{ deadbandFilter.label | translate }}\"\n        class=\"c8y-radio radio-inline\"\n      >\n        <input\n          type=\"radio\"\n          name=\"deadbandType\"\n          [value]=\"deadbandFilter.value\"\n          [(ngModel)]=\"subscriptionParameters.deadbandType\"\n          [required]=\"requireSubscription\"\n        />\n        <span></span>\n        <span>{{ deadbandFilter.label | translate }}</span>\n      </label>\n    </c8y-form-group>\n    <div [collapse]=\"subscriptionParameters.deadbandType == 'None'\" [isAnimated]=\"true\">\n      <c8y-form-group>\n        <label translate>Deadband value</label>\n        <input\n          class=\"form-control\"\n          type=\"number\"\n          name=\"deadbandValue\"\n          [(ngModel)]=\"subscriptionParameters.deadbandValue\"\n          placeholder=\"{{ 'e.g.' | translate }} 10\"\n          [required]=\"subscriptionParameters.deadbandType != 'None'\"\n          min=\"0\"\n        />\n      </c8y-form-group>\n    </div>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgModelGroup }]
            }]
    }], function () { return []; }, { onSubscriptionChange: [{
            type: Output
        }], model: [{
            type: Input
        }], groupName: [{
            type: Input
        }] }); })();
