// @c8y/ngx-components/esm2015/protocol-opcua/opcua-microservice.guard.js
import { Injectable } from '@angular/core';
import { AppStateService } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class OpcuaMicroserviceGuard {
    constructor(app) {
        this.app = app;
        this.microserviceName = 'opcua-mgmt-service';
    }
    canActivate() {
        if (!this.checkApp) {
            this.checkApp = this.app.isApplicationAvailable(this.microserviceName);
        }
        return this.checkApp;
    }
}
OpcuaMicroserviceGuard.ɵfac = function OpcuaMicroserviceGuard_Factory(t) { return new (t || OpcuaMicroserviceGuard)(ɵngcc0.ɵɵinject(ɵngcc1.AppStateService)); };
OpcuaMicroserviceGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: OpcuaMicroserviceGuard, factory: OpcuaMicroserviceGuard.ɵfac });
OpcuaMicroserviceGuard.ctorParameters = () => [
    { type: AppStateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OpcuaMicroserviceGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.AppStateService }]; }, null); })();
