// @c8y/ngx-components/esm2015/operations/operation-summary/operation-summary.module.js
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { OperationSummaryComponent } from './operation-summary.component';
/**
 * This module provides a component for displaying the summary of given bulk operation.
 */
import * as ɵngcc0 from '@angular/core';
export class OperationSummaryModule {
}
OperationSummaryModule.ɵfac = function OperationSummaryModule_Factory(t) { return new (t || OperationSummaryModule)(); };
OperationSummaryModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationSummaryModule });
OperationSummaryModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationSummaryModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule],
                declarations: [OperationSummaryComponent],
                entryComponents: [OperationSummaryComponent],
                exports: [OperationSummaryComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationSummaryModule, { declarations: function () { return [OperationSummaryComponent]; }, imports: function () { return [CoreModule]; }, exports: function () { return [OperationSummaryComponent]; } }); })();
