// @c8y/ngx-components/esm2015/core/modal/modal.module.js
import { NgModule } from '@angular/core';
import { ModalModule as BsModalModule } from 'ngx-bootstrap/modal';
import { ConfirmModalComponent } from './modal-confirm.component';
import { ModalComponent } from './modal.component';
import { ModalService } from './modal.service';
import { CommonModule } from '../common/common.module';
import { PopoverConfirmComponent } from './popover-confirm.component';
import { PopoverModule } from 'ngx-bootstrap/popover';
/**
 * The angular module definition for modal.
 * @exports ConfirmModalComponent
 * @exports ModalComponent
 * @exports PopoverConfirmComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from 'ngx-bootstrap/popover';
export class ModalModule {
}
ModalModule.ɵfac = function ModalModule_Factory(t) { return new (t || ModalModule)(); };
ModalModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ModalModule });
ModalModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ModalService
    ], imports: [[
            BsModalModule.forRoot(),
            CommonModule,
            PopoverModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ModalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BsModalModule.forRoot(),
                    CommonModule,
                    PopoverModule.forRoot()
                ],
                exports: [
                    ConfirmModalComponent,
                    ModalComponent,
                    PopoverConfirmComponent
                ],
                declarations: [
                    ConfirmModalComponent,
                    ModalComponent,
                    PopoverConfirmComponent
                ],
                providers: [
                    ModalService
                ],
                entryComponents: [
                    ConfirmModalComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ModalModule, { declarations: function () { return [ConfirmModalComponent,
        ModalComponent,
        PopoverConfirmComponent]; }, imports: function () { return [ɵngcc1.ModalModule, CommonModule, ɵngcc2.PopoverModule]; }, exports: function () { return [ConfirmModalComponent,
        ModalComponent,
        PopoverConfirmComponent]; } }); })();
