// @c8y/ngx-components/esm2015/core/docs/guide-href.directive.js
import { Directive, Input, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class GuideHrefDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.htmlAnchor = this.elementRef.nativeElement;
        if (!this.htmlAnchor.hasAttribute('target')) {
            this.htmlAnchor.setAttribute('target', '_blank');
        }
    }
    set baseHref(href) {
        this.htmlAnchor.href = `${href}/${this.c8yGuideHref}`;
    }
}
GuideHrefDirective.ɵfac = function GuideHrefDirective_Factory(t) { return new (t || GuideHrefDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
GuideHrefDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: GuideHrefDirective, selectors: [["", "c8y-guide-href", ""]], inputs: { c8yGuideHref: ["c8y-guide-href", "c8yGuideHref"] } });
GuideHrefDirective.ctorParameters = () => [
    { type: ElementRef }
];
GuideHrefDirective.propDecorators = {
    c8yGuideHref: [{ type: Input, args: ['c8y-guide-href',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GuideHrefDirective, [{
        type: Directive,
        args: [{
                selector: '[c8y-guide-href]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { c8yGuideHref: [{
            type: Input,
            args: ['c8y-guide-href']
        }] }); })();
