// @c8y/ngx-components/esm2015/connectivity/actility/actility-tab.factory.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { gettext } from '@c8y/ngx-components';
import { AbstractConnectivityTabFactory } from '../abstract-connectivity-tab.factory';
import { ActilityGuard } from './actility.guard';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './actility.guard';
export class ActilityTabFactory extends AbstractConnectivityTabFactory {
    constructor(router, guard) {
        super(router, guard);
        this.router = router;
        this.guard = guard;
        this.tabConfig = {
            label: gettext('Actility LoRa provider settings'),
            path: '/connectivitySettings/actility_lora_provider_settings',
            icon: 'c8y-device'
        };
    }
    config() {
        return this.tabConfig;
    }
}
ActilityTabFactory.ɵfac = function ActilityTabFactory_Factory(t) { return new (t || ActilityTabFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.ActilityGuard)); };
ActilityTabFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ActilityTabFactory, factory: ActilityTabFactory.ɵfac });
ActilityTabFactory.ctorParameters = () => [
    { type: Router },
    { type: ActilityGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityTabFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.ActilityGuard }]; }, null); })();
