// @ngx-formly/core/esm5/lib/templates/field-template.type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { FieldType } from './field.type';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/platform-browser';
var FormlyTemplateType = /** @class */ (function (_super) {
    tslib_1.__extends(FormlyTemplateType, _super);
    function FormlyTemplateType(sanitizer) {
        var _this = _super.call(this) || this;
        _this.sanitizer = sanitizer;
        _this.innerHtml = { content: null, template: null };
        return _this;
    }
    Object.defineProperty(FormlyTemplateType.prototype, "template", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.field && (this.field.template !== this.innerHtml.template)) {
                this.innerHtml = {
                    template: this.field.template,
                    content: this.to.safeHtml
                        ? this.sanitizer.bypassSecurityTrustHtml(this.field.template)
                        : this.field.template,
                };
            }
            return this.innerHtml.content;
        },
        enumerable: true,
        configurable: true
    });
    /** @nocollapse */
    FormlyTemplateType.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
FormlyTemplateType.ɵfac = function FormlyTemplateType_Factory(t) { return new (t || FormlyTemplateType)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DomSanitizer)); };
FormlyTemplateType.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FormlyTemplateType, selectors: [["formly-template"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 1, vars: 1, consts: [[3, "innerHtml"]], template: function FormlyTemplateType_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("innerHtml", ctx.template, ɵngcc0.ɵɵsanitizeHtml);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormlyTemplateType, [{
        type: Component,
        args: [{
                selector: 'formly-template',
                template: "<div [innerHtml]=\"template\"></div>"
            }]
    }], function () { return [{ type: ɵngcc1.DomSanitizer }]; }, null); })();
    return FormlyTemplateType;
}(FieldType));
export { FormlyTemplateType };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FormlyTemplateType.prototype.innerHtml;
    /**
     * @type {?}
     * @private
     */
    FormlyTemplateType.prototype.sanitizer;
}
