// @c8y/ngx-components/esm2015/operations/bulk-operation-scheduler/bulk-operation-scheduler.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule } from '@c8y/ngx-components';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { OperationSchedulerComponent } from './operation-scheduler.component';
/**
 * This module provides components for scheduling bulk operations.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/datepicker';
import * as ɵngcc2 from 'ngx-bootstrap/timepicker';
export class BulkOperationSchedulerModule {
}
BulkOperationSchedulerModule.ɵfac = function BulkOperationSchedulerModule_Factory(t) { return new (t || BulkOperationSchedulerModule)(); };
BulkOperationSchedulerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationSchedulerModule });
BulkOperationSchedulerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [], imports: [[
            CoreModule,
            ReactiveFormsModule,
            BsDatepickerModule.forRoot(),
            TimepickerModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationSchedulerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    ReactiveFormsModule,
                    BsDatepickerModule.forRoot(),
                    TimepickerModule.forRoot()
                ],
                providers: [],
                declarations: [OperationSchedulerComponent],
                entryComponents: [OperationSchedulerComponent],
                exports: [OperationSchedulerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationSchedulerModule, { declarations: function () { return [OperationSchedulerComponent]; }, imports: function () { return [CoreModule,
        ReactiveFormsModule, ɵngcc1.BsDatepickerModule, ɵngcc2.TimepickerModule]; }, exports: function () { return [OperationSchedulerComponent]; } }); })();
