// @c8y/ngx-components/esm2015/connectivity/sigfox/sigfox.module.js
import { NgModule } from '@angular/core';
import { HOOK_NAVIGATOR_NODES, HOOK_TABS } from '@c8y/ngx-components';
import { SigfoxNodeFactory } from './sigfox-node.factory';
import { SigfoxTabFactory } from './sigfox-tab.factory';
import { SigfoxGuard } from './sigfox.guard';
import * as ɵngcc0 from '@angular/core';
export class SigfoxModule {
}
SigfoxModule.ɵfac = function SigfoxModule_Factory(t) { return new (t || SigfoxModule)(); };
SigfoxModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SigfoxModule });
SigfoxModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        SigfoxGuard,
        SigfoxTabFactory,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: SigfoxNodeFactory,
            multi: true
        },
        {
            provide: HOOK_TABS,
            useClass: SigfoxTabFactory,
            multi: true
        }
    ] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxModule, [{
        type: NgModule,
        args: [{
                providers: [
                    SigfoxGuard,
                    SigfoxTabFactory,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: SigfoxNodeFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_TABS,
                        useClass: SigfoxTabFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
