// @c8y/ngx-components/esm2015/connectivity/sigfox/sigfox.guard.js
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class SigfoxGuard {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    canActivate(route, state) {
        if (this.showSigfoxProvider === undefined) {
            this.showSigfoxProvider = this.applicationService
                .isAvailable(SigfoxGuard.applicationName)
                .then(result => result && result.data);
        }
        return this.showSigfoxProvider;
    }
}
SigfoxGuard.ɵfac = function SigfoxGuard_Factory(t) { return new (t || SigfoxGuard)(ɵngcc0.ɵɵinject(ɵngcc1.ApplicationService)); };
SigfoxGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SigfoxGuard, factory: SigfoxGuard.ɵfac });
SigfoxGuard.applicationName = 'sigfox-agent';
SigfoxGuard.ctorParameters = () => [
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SigfoxGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ApplicationService }]; }, null); })();
