// @c8y/ngx-components/esm2015/protocol-opcua/opcua-server-list.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
export class OpcuaServerListComponent {
    constructor() {
        this.serverList = [];
        this.present = new EventEmitter();
    }
    presentConfig(server) {
        server.active = true;
        this.present.emit(server);
    }
}
OpcuaServerListComponent.decorators = [
    { type: Component, args: [{
                selector: 'opcua-server-list',
                template: "<!-- add active class to c8y-stacked-item when clicked -->\n<div\n  class=\"c8y-stacked-item\"\n  [class.active]=\"server.active\"\n  *ngFor=\"let server of serverList\"\n  (click)=\"presentConfig(server)\"\n>\n  <i [c8yIcon]=\"'server'\" class=\"icon-20 \"></i>\n  <div class=\"p-l-16 text-truncate\">\n    <p id=\"serverName\" class=\"text-truncate\">{{ server.name }}</p>\n    <small id=\"serverUrl\" class=\"text-truncate text-muted\">{{ server.config.serverUrl }}</small>\n  </div>\n  <div class=\"flex-item-right p-l-8 d-flex p-r-16\">\n    <device-status class=\"p-r-8\" [mo]=\"server\"></device-status>\n    <a\n      title=\"{{ server.quickInfo.padlockMsg | translate }}\"\n      placement=\"left\"\n      class=\"text-gray-darker\"\n    >\n      <i [c8yIcon]=\"server.quickInfo.padlock\"></i>\n    </a>\n  </div>\n</div>\n"
            },] }
];
OpcuaServerListComponent.propDecorators = {
    serverList: [{ type: Input }],
    present: [{ type: Output }]
};
