// @c8y/ngx-components/esm2015/core/help/help.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { HelpComponent } from './help.component';
import { ActionBarModule } from '../action-bar/action-bar.module';
/**
 * Provides components which allow to display help elements
 * inside the app.
 * @exports HelpComponent A component which shows a context help in
 *                        the action bar.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/collapse';
export class HelpModule {
}
HelpModule.ɵfac = function HelpModule_Factory(t) { return new (t || HelpModule)(); };
HelpModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: HelpModule });
HelpModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, CollapseModule.forRoot(), ActionBarModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HelpModule, [{
        type: NgModule,
        args: [{
                declarations: [HelpComponent],
                imports: [CommonModule, CollapseModule.forRoot(), ActionBarModule],
                exports: [HelpComponent],
                entryComponents: [HelpComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HelpModule, { declarations: function () { return [HelpComponent]; }, imports: function () { return [CommonModule, ɵngcc1.CollapseModule, ActionBarModule]; }, exports: function () { return [HelpComponent]; } }); })();
