// @c8y/ngx-components/esm2015/core/login/strength-validator-service.js
import { Injectable } from '@angular/core';
import { PasswordService } from '../authentication';
import { PasswordStrength } from '@c8y/client';
import * as i0 from "@angular/core";
import * as i1 from "../authentication/password.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../authentication';
export class StrengthValidatorService {
    constructor(passwordService) {
        this.passwordService = passwordService;
    }
    isStrong(password) {
        return this.isPasswordGreen(this.passwordService.getStrengthColor(password).passwordStrength);
    }
    isPasswordGreen(strength) {
        return strength === PasswordStrength.GREEN;
    }
}
StrengthValidatorService.ɵfac = function StrengthValidatorService_Factory(t) { return new (t || StrengthValidatorService)(ɵngcc0.ɵɵinject(ɵngcc1.PasswordService)); };
StrengthValidatorService.ɵprov = i0.ɵɵdefineInjectable({ factory: function StrengthValidatorService_Factory() { return new StrengthValidatorService(i0.ɵɵinject(i1.PasswordService)); }, token: StrengthValidatorService, providedIn: "root" });
StrengthValidatorService.ctorParameters = () => [
    { type: PasswordService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StrengthValidatorService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.PasswordService }]; }, null); })();
