// @c8y/ngx-components/esm2015/tenants/tenants.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE } from '@c8y/ngx-components';
import { TenantsNavigationFactory } from './tenants-navigation.factory';
import { TenantListComponent } from './tenant-list/tenant-list.component';
import { CreationTimeFilteringFormRendererComponent } from './tenant-list/creation-time.filtering-form-renderer.component';
import { StatusFilteringFormRendererComponent } from './tenant-list/status.filtering-form-renderer.component';
import { TENANTS_MODULE_CONFIG } from './tenants.model';
import { TenantListGuard } from './tenant-list/tenant-list.guard';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        path: 'tenants',
        component: TenantListComponent,
        canActivate: [TenantListGuard]
    }
];
export class TenantsModule {
    static config(config = {}) {
        return {
            ngModule: TenantsModule,
            providers: [
                {
                    provide: TENANTS_MODULE_CONFIG,
                    useValue: config
                }
            ]
        };
    }
}
TenantsModule.ɵfac = function TenantsModule_Factory(t) { return new (t || TenantsModule)(); };
TenantsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TenantsModule });
TenantsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        TenantListGuard,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: TenantsNavigationFactory,
            multi: true
        },
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[
            RouterModule,
            BsDatepickerModule,
            CoreModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    RouterModule,
                    BsDatepickerModule,
                    CoreModule
                ],
                exports: [],
                declarations: [
                    TenantListComponent,
                    CreationTimeFilteringFormRendererComponent,
                    StatusFilteringFormRendererComponent
                ],
                entryComponents: [
                    TenantListComponent,
                    CreationTimeFilteringFormRendererComponent,
                    StatusFilteringFormRendererComponent
                ],
                providers: [
                    TenantListGuard,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: TenantsNavigationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TenantsModule, { declarations: function () { return [TenantListComponent,
        CreationTimeFilteringFormRendererComponent,
        StatusFilteringFormRendererComponent]; }, imports: function () { return [RouterModule,
        BsDatepickerModule,
        CoreModule]; } }); })();
export { ɵ0 };
