// @c8y/ngx-components/esm2015/tenants/tenants-navigation.factory.js
import { __awaiter } from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { TenantListGuard } from './tenant-list/tenant-list.guard';
import { TENANTS_MODULE_CONFIG } from './tenants.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './tenant-list/tenant-list.guard';
export class TenantsNavigationFactory {
    constructor(tenantListGuard, config) {
        this.tenantListGuard = tenantListGuard;
        this.config = config;
        this.navs = [];
    }
    get() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const canActivateTenantList = yield this.tenantListGuard.canActivate();
            if (!this.navs.length) {
                const subtenantsNavigatorNode = (_b = (_a = this.config) === null || _a === void 0 ? void 0 : _a.subtenantsNavigatorNode) !== null && _b !== void 0 ? _b : true;
                if (subtenantsNavigatorNode !== false) {
                    this.navs.push(new NavigatorNode(Object.assign({ parent: gettext('Tenants'), label: gettext('Subtenants'), icon: 'c8y-sub-tenants', path: 'tenants', routerLinkExact: false, priority: 4000, hidden: !canActivateTenantList }, (subtenantsNavigatorNode === true ? {} : subtenantsNavigatorNode))));
                }
            }
            return this.navs;
        });
    }
}
TenantsNavigationFactory.ɵfac = function TenantsNavigationFactory_Factory(t) { return new (t || TenantsNavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.TenantListGuard), ɵngcc0.ɵɵinject(TENANTS_MODULE_CONFIG, 8)); };
TenantsNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TenantsNavigationFactory, factory: TenantsNavigationFactory.ɵfac });
TenantsNavigationFactory.ctorParameters = () => [
    { type: TenantListGuard },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [TENANTS_MODULE_CONFIG,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TenantsNavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.TenantListGuard }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [TENANTS_MODULE_CONFIG]
            }] }]; }, null); })();
