// @c8y/ngx-components/esm2015/upgrade/ng1/downgraded.components.js
import { downgradeComponent } from '@angular/upgrade/static';
import { BootstrapComponent, DataGridComponent, UserTotpRevokeComponent, LoadingComponent, HighlightComponent } from '@c8y/ngx-components';
import { AppLogsAutoRefreshComponent } from '@c8y/ngx-components/app-logs';
import { RangeDisplayComponent } from '@c8y/ngx-components';
import { HelpComponent } from '@c8y/ngx-components';
export const bootstrapComponent = downgradeComponent({ component: BootstrapComponent });
export const userTotpComponent = downgradeComponent({ component: UserTotpRevokeComponent });
export const appLogsAutoRefreshComponent = downgradeComponent({
    component: AppLogsAutoRefreshComponent
});
export const dataGridComponent = downgradeComponent({ component: DataGridComponent });
export const loadingComponent = downgradeComponent({ component: LoadingComponent });
export const rangeDisplayComponent = downgradeComponent({ component: RangeDisplayComponent });
export const helpComponent = downgradeComponent({ component: HelpComponent });
export const highlightComponent = downgradeComponent({
    component: HighlightComponent,
    inputs: ['pattern', 'text']
});
