// @c8y/ngx-components/esm2015/upgrade/ng1/serverMessages.service.js
import { Injectable, Inject } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { HOOK_PATTERN_MESSAGES } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-translate/core';
export class ServerMessagesService {
    constructor(translateService, patterns) {
        this.translateService = translateService;
        this.MESSAGE_PATTERNS = patterns;
    }
    translate(s) {
        return this.translateService.instant(s);
    }
}
ServerMessagesService.ɵfac = function ServerMessagesService_Factory(t) { return new (t || ServerMessagesService)(ɵngcc0.ɵɵinject(ɵngcc1.TranslateService), ɵngcc0.ɵɵinject(HOOK_PATTERN_MESSAGES)); };
ServerMessagesService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ServerMessagesService, factory: ServerMessagesService.ɵfac });
ServerMessagesService.ctorParameters = () => [
    { type: TranslateService },
    { type: undefined, decorators: [{ type: Inject, args: [HOOK_PATTERN_MESSAGES,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ServerMessagesService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.TranslateService }, { type: undefined, decorators: [{
                type: Inject,
                args: [HOOK_PATTERN_MESSAGES]
            }] }]; }, null); })();
