// @c8y/ngx-components/esm2015/core/list-display-switch/list-display-switch.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '../common/common.module';
import { ListDisplaySwitchComponent } from './list-display-switch.component';
/**
 * The angular module definition for the list display switcher.
 * It allows to switch between a list and a grid view.
 * @exports ListDisplaySwitchComponent
 */
import * as ɵngcc0 from '@angular/core';
export class ListDisplaySwitchModule {
}
ListDisplaySwitchModule.ɵfac = function ListDisplaySwitchModule_Factory(t) { return new (t || ListDisplaySwitchModule)(); };
ListDisplaySwitchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ListDisplaySwitchModule });
ListDisplaySwitchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, FormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListDisplaySwitchModule, [{
        type: NgModule,
        args: [{
                declarations: [ListDisplaySwitchComponent],
                imports: [CommonModule, FormsModule],
                exports: [ListDisplaySwitchComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ListDisplaySwitchModule, { declarations: function () { return [ListDisplaySwitchComponent]; }, imports: function () { return [CommonModule, FormsModule]; }, exports: function () { return [ListDisplaySwitchComponent]; } }); })();
