// @c8y/ngx-components/esm2015/operations/details/operation-details.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { BulkOperationSchedulerModule } from '@c8y/ngx-components/operations/bulk-operation-scheduler';
import { OperationDetailsComponent } from './operation-details.component';
/**
 * This module provides a component for entering operation details.
 */
import * as ɵngcc0 from '@angular/core';
export class OperationDetailsModule {
}
OperationDetailsModule.ɵfac = function OperationDetailsModule_Factory(t) { return new (t || OperationDetailsModule)(); };
OperationDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: OperationDetailsModule });
OperationDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(OperationDetailsModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule],
                declarations: [OperationDetailsComponent],
                exports: [OperationDetailsComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(OperationDetailsModule, { declarations: function () { return [OperationDetailsComponent]; }, imports: function () { return [CoreModule, FormsModule, ReactiveFormsModule, BulkOperationSchedulerModule]; }, exports: function () { return [OperationDetailsComponent]; } }); })();
