// @ngx-formly/core/esm5/select/select.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormlySelectOptionsPipe } from './select-options.pipe';
import * as ɵngcc0 from '@angular/core';
var FormlySelectModule = /** @class */ (function () {
    function FormlySelectModule() {
    }
FormlySelectModule.ɵfac = function FormlySelectModule_Factory(t) { return new (t || FormlySelectModule)(); };
FormlySelectModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: FormlySelectModule });
FormlySelectModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormlySelectModule, [{
        type: NgModule,
        args: [{
                declarations: [FormlySelectOptionsPipe],
                exports: [FormlySelectOptionsPipe]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormlySelectModule, { declarations: function () { return [FormlySelectOptionsPipe]; }, exports: function () { return [FormlySelectOptionsPipe]; } }); })();
    return FormlySelectModule;
}());
export { FormlySelectModule };
