// @c8y/ngx-components/esm2015/reports/reports.module.js
import { NgModule } from '@angular/core';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { ReportsService } from './reports.service';
import { CronService } from './cron.service';
import { ExportSchedulesComponent } from './export-schedules.component';
import { ScheduleModalComponent } from './schedule-modal.component';
import { CronComponent } from './cron.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
/**
 * The angular module definition for reports.
 * @exports ExportSchedulesComponent
 * @exports ScheduleModalComponent
 * @exports CronComponent
 */
export class ReportsModule {
}
ReportsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                imports: [CoreModule, FormsModule, BsDropdownModule, TooltipModule],
                entryComponents: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                providers: [ReportsService, CronService],
                exports: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent]
            },] }
];
