// @c8y/ngx-components/esm2015/core/list-group/list-item-footer.component.js
import { Component, Input } from '@angular/core';
/**
 * The footer can be used with either content projection
 * or a simple string as an input.
 *
 * @example
 * ```html
 * <c8y-li>
 *   Content
 *   <c8y-li-footer>I am a footer</c8y-li-footer>
 * </c8y-li>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function ListItemFooterComponent_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.footer);
} }
const _c0 = ["*"];
export class ListItemFooterComponent {
}
ListItemFooterComponent.ɵfac = function ListItemFooterComponent_Factory(t) { return new (t || ListItemFooterComponent)(); };
ListItemFooterComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemFooterComponent, selectors: [["c8y-list-item-footer"], ["c8y-li-footer"]], inputs: { footer: "footer" }, ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[4, "ngIf"]], template: function ListItemFooterComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ListItemFooterComponent_span_0_Template, 2, 1, "span", 0);
        ɵngcc0.ɵɵprojection(1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.footer);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2 });
ListItemFooterComponent.propDecorators = {
    footer: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemFooterComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-footer, c8y-li-footer',
                template: "<span *ngIf=\"footer\">{{footer}}</span>\n<ng-content></ng-content>\n"
            }]
    }], null, { footer: [{
            type: Input
        }] }); })();
