// @c8y/ngx-components/esm2015/core/list-group/list-item-action.component.js
import { Component, Output, Input, ViewChild, EventEmitter } from '@angular/core';
/**
 * Actions are functions that can be executed on the
 * current list item. They are displayed next to the
 * item and can be grouped in a dropdown.
 *
 * @example
 * ```html
 * <c8y-li>
 *   <c8y-li-action (click)="deleteDevice(device.id)" icon="times">
 *     Delete
 *   </c8y-li-action>
 *   <c8y-li-action (click)="(false)" icon="rocket">
 *     Launch to space
 *   </c8y-li-action>
 * </c8y-li>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';

const _c0 = ["template"];
function ListItemActionComponent_ng_template_0_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 4);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r2.icon);
} }
function ListItemActionComponent_ng_template_0_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r3.label));
} }
function ListItemActionComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "button", 1);
    ɵngcc0.ɵɵlistener("click", function ListItemActionComponent_ng_template_0_Template_button_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.click.emit($event); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtemplate(3, ListItemActionComponent_ng_template_0_i_3_Template, 1, 1, "i", 2);
    ɵngcc0.ɵɵtext(4, "\u00A0 ");
    ɵngcc0.ɵɵtemplate(5, ListItemActionComponent_ng_template_0_span_5_Template, 3, 3, "span", 3);
    ɵngcc0.ɵɵprojection(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r1.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.label);
} }
const _c1 = ["*"];
export class ListItemActionComponent {
    constructor() {
        /**
         * An EventEmitter to determine what to do if the
         * action is triggered.
         */
        this.click = new EventEmitter();
    }
}
ListItemActionComponent.ɵfac = function ListItemActionComponent_Factory(t) { return new (t || ListItemActionComponent)(); };
ListItemActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemActionComponent, selectors: [["c8y-list-item-action"], ["c8y-li-action"]], viewQuery: function ListItemActionComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { label: "label", icon: "icon" }, outputs: { click: "click" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""], [3, "title", "click"], [3, "c8yIcon", 4, "ngIf"], [4, "ngIf"], [3, "c8yIcon"]], template: function ListItemActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ListItemActionComponent_ng_template_0_Template, 7, 5, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.IconDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ListItemActionComponent.propDecorators = {
    template: [{ type: ViewChild, args: ['template', { static: false },] }],
    label: [{ type: Input }],
    icon: [{ type: Input }],
    click: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemActionComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-action, c8y-li-action',
                template: "<ng-template #template>\n  <li>\n    <button title=\"{{ label | translate }}\" (click)=\"click.emit($event)\">\n      <i [c8yIcon]=\"icon\" *ngIf=\"icon\"></i>&nbsp;\n      <span *ngIf=\"label\">{{ label | translate }}</span>\n      <ng-content></ng-content>\n    </button>\n  </li>\n</ng-template>\n"
            }]
    }], function () { return []; }, { click: [{
            type: Output
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }], label: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
