// @c8y/ngx-components/esm2015/core/list-group/list-item-collapse.component.js
import { Component, Input } from '@angular/core';
/**
 * Display collapsed content on a list item. Can be
 * used for more detailed information. Use it with
 * content projection.
 *
 * @example
 * ```html
 * <c8y-li>
 *   <c8y-collapsed>
 *     <strong>I am only visible if extended</strong>
 *   </c8y-collapsed>
 * </c8y-li>
 * ```
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ListItemCollapseComponent {
    constructor() {
        /**
         * The way how to toggle collapse state.
         * button (default) - additional button for toggle collapse state
         * row - toggling collapse state by clicking to entire row.
         */
        this.collapseWay = 'button';
    }
}
ListItemCollapseComponent.ɵfac = function ListItemCollapseComponent_Factory(t) { return new (t || ListItemCollapseComponent)(); };
ListItemCollapseComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemCollapseComponent, selectors: [["c8y-list-item-collapse"], ["c8y-li-collapse"]], inputs: { collapseWay: "collapseWay" }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ListItemCollapseComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ListItemCollapseComponent.propDecorators = {
    collapseWay: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemCollapseComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-collapse, c8y-li-collapse',
                template: "<ng-content></ng-content>\n"
            }]
    }], function () { return []; }, { collapseWay: [{
            type: Input
        }] }); })();
