// @c8y/ngx-components/esm2015/core/list-group/list-item-timeline.component.js
import { Component } from '@angular/core';
/**
 * Allows to display a timeline. Should be used with
 * content projection.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li-timeline>
 *     24.April 2020
 *     <c8y-li>
 *       Hello
 *     </c8y-li>
 *   </c8y-li-timeline>
 * </c8y-list-group>
 * ```
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*", [["c8y-li"], ["c8y-list-item"]]];
const _c1 = ["*", "c8y-li, c8y-list-item"];
export class ListItemTimelineComponent {
}
ListItemTimelineComponent.ɵfac = function ListItemTimelineComponent_Factory(t) { return new (t || ListItemTimelineComponent)(); };
ListItemTimelineComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemTimelineComponent, selectors: [["c8y-list-item-timeline"], ["c8y-li-timeline"]], hostAttrs: [1, "c8y-list--timeline__item"], ngContentSelectors: _c1, decls: 3, vars: 0, consts: [[1, "c8y-list--timeline__item__date"]], template: function ListItemTimelineComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "small", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵprojection(2, 1);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemTimelineComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-timeline, c8y-li-timeline',
                template: "<small class=\"c8y-list--timeline__item__date\">\n  <ng-content></ng-content>\n</small>\n<ng-content select=\"c8y-li, c8y-list-item\"></ng-content>\n",
                host: {
                    class: 'c8y-list--timeline__item'
                }
            }]
    }], null, null); })();
