// @c8y/ngx-components/esm2015/operations/bulk-operation-from-single/bulk-operation-from-single.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationStepperModule } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import { BulkOperationFromSingleComponent } from './bulk-operation-from-single.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export const bulkOperationCreateSingleRoutes = [
    {
        path: 'devicecontrol/single/create-bulk/:operationId',
        component: BulkOperationFromSingleComponent
    }
];
export class BulkOperationFromSingleModule {
}
BulkOperationFromSingleModule.ɵfac = function BulkOperationFromSingleModule_Factory(t) { return new (t || BulkOperationFromSingleModule)(); };
BulkOperationFromSingleModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationFromSingleModule });
BulkOperationFromSingleModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CoreModule,
            BulkOperationStepperModule,
            RouterModule.forChild(bulkOperationCreateSingleRoutes)
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationFromSingleModule, [{
        type: NgModule,
        args: [{
                declarations: [BulkOperationFromSingleComponent],
                imports: [
                    CoreModule,
                    BulkOperationStepperModule,
                    RouterModule.forChild(bulkOperationCreateSingleRoutes)
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationFromSingleModule, { declarations: function () { return [BulkOperationFromSingleComponent]; }, imports: function () { return [CoreModule,
        BulkOperationStepperModule, ɵngcc1.RouterModule]; } }); })();
