// @c8y/ngx-components/esm2015/core/date-picker/date-picker.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { DatePickerComponent } from './date-picker.component';
import { FormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { FormsModule as C8yFormsModule } from '../forms/index';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
/**
 * The angular module definition for a date picker - it allows you to pick either one date or a range of two.
 * @exports DatePickerComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
import * as ɵngcc2 from 'ngx-bootstrap/datepicker';
export class DatePickerModule {
}
DatePickerModule.ɵfac = function DatePickerModule_Factory(t) { return new (t || DatePickerModule)(); };
DatePickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DatePickerModule });
DatePickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            C8yFormsModule,
            CommonModule,
            FormsModule,
            ReactiveFormsModule,
            BsDropdownModule.forRoot(),
            BsDatepickerModule.forRoot(),
            C8yCommonModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatePickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    C8yFormsModule,
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    BsDropdownModule.forRoot(),
                    BsDatepickerModule.forRoot(),
                    C8yCommonModule
                ],
                exports: [DatePickerComponent],
                declarations: [DatePickerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatePickerModule, { declarations: function () { return [DatePickerComponent]; }, imports: function () { return [C8yFormsModule,
        CommonModule,
        FormsModule,
        ReactiveFormsModule, ɵngcc1.BsDropdownModule, ɵngcc2.BsDatepickerModule, C8yCommonModule]; }, exports: function () { return [DatePickerComponent]; } }); })();
