// @c8y/ngx-components/esm2015/context-dashboard/cockpit-dashboard/legacy-welcome/legacy-welcome.component.js
import { Component, Host } from '@angular/core';
import { ContextDashboardComponent } from '../../context-dashboard.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../context-dashboard.component';
import * as ɵngcc2 from '@c8y/ngx-components';
export class LegacyWelcomeComponent {
    constructor(dashboard) {
        this.dashboard = dashboard;
    }
    restore() {
        this.dashboard.restore();
    }
}
LegacyWelcomeComponent.ɵfac = function LegacyWelcomeComponent_Factory(t) { return new (t || LegacyWelcomeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ContextDashboardComponent, 1)); };
LegacyWelcomeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: LegacyWelcomeComponent, selectors: [["c8y-legacy-welcome-widget"]], decls: 17, vars: 9, consts: [[1, "card-block"], ["c8y-guide-docs", "", 1, "lead"], ["translate", "", 1, "cockpit-welcome-manual", 3, "title"], ["c8y-guide-href", "users-guide/cockpit/"], [1, "m-t-16"], ["translate", "", 1, "btn", "btn-primary", "btn-sm", "m-t-16", 3, "click"]], template: function LegacyWelcomeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "p", 1);
        ɵngcc0.ɵɵelementStart(2, "span");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "span", 2);
        ɵngcc0.ɵɵdisableBindings();
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵtext(7, "Browse the manual ");
        ɵngcc0.ɵɵelementStart(8, "a", 3);
        ɵngcc0.ɵɵtext(9, "here");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(10, ".");
        ɵngcc0.ɵɵenableBindings();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "p", 4);
        ɵngcc0.ɵɵelementStart(12, "em");
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "button", 5);
        ɵngcc0.ɵɵlistener("click", function LegacyWelcomeComponent_Template_button_click_15_listener() { ctx.restore(); return false; });
        ɵngcc0.ɵɵtext(16, " Get new dashboard ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1("", ɵngcc0.ɵɵpipeBind1(4, 3, "Welcome to Cockpit"), ".");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 5, "Browse the manual"));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵtextInterpolate1("", ɵngcc0.ɵɵpipeBind1(14, 7, "New default dashboard available."), " ");
    } }, directives: [ɵngcc2.GuideDocsComponent, ɵngcc2.C8yTranslateDirective, ɵngcc2.GuideHrefDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
LegacyWelcomeComponent.ctorParameters = () => [
    { type: ContextDashboardComponent, decorators: [{ type: Host }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LegacyWelcomeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-legacy-welcome-widget',
                template: "<div class=\"card-block\">\n  <p class=\"lead\" c8y-guide-docs>\n    <span>{{ 'Welcome to Cockpit' | translate }}.</span>\n    <span title=\"{{ 'Browse the manual' | translate }}\" class=\"cockpit-welcome-manual\" translate ngNonBindable\n      >Browse the manual <a c8y-guide-href=\"users-guide/cockpit/\">here</a>.</span\n    >\n  </p>\n  <p class=\"m-t-16\">\n    <em>{{ 'New default dashboard available.' | translate }} </em>\n  </p>\n  <button class=\"btn btn-primary btn-sm m-t-16\" (click)=\"restore(); (false)\" translate>\n    Get new dashboard\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ContextDashboardComponent, decorators: [{
                type: Host
            }] }]; }, null); })();
