// @c8y/cli/dist/webpack/config/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const path = require("path");
function config(env) {
    return {
        entry: {
            bootstrap: '@c8y/cli/dist/app-bootstrap'
        },
        mode: env.mode,
        devtool: 'eval-cheap-module-source-map',
        output: {
            filename: '[name].js',
            chunkFilename: '[name].js',
            assetModuleFilename: '[name]-[contenthash][ext]',
            path: path.resolve(env.output || 'dist', 'apps', env.app.contextPath)
        },
        resolve: {
            extensions: ['.ts', '.js', '.json'],
            // always look in the root node_modules first
            // trying to avoid issues with multiple versions of rxjs
            modules: [path.resolve(env.app.rootNodeModulesPath || 'node_modules'), 'node_modules']
        },
        // We ignore this message as Ivy includes everything into the bundle when
        // used with yarn workspace in our monorepo. As the build is not done via
        // mono-repo, the production build don't have this unused files.
        ignoreWarnings: [
            {
                message: /.*? part of the TypeScript compilation but it's unused\./
            }
        ],
        module: {
            // exprContextCritical: false -> Removes a warning caused by the dynamic imports here:
            // @c8y/ng1-modules/rotation3dWidget/webpack-specific.js
            // The issue is discussed here: https://github.com/webpack/webpack/issues/196
            exprContextCritical: false,
            rules: [
                {
                    test: /(jpe?g|gif|png|woff(2)?|eot|ttf|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
                    exclude: /(\/|\\)favicon/,
                    type: 'asset/resource'
                }
            ]
        }
    };
}
exports.config = config;
