// @c8y/cli/dist/webpack/config/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = void 0;
const CopyPlugin = require("copy-webpack-plugin");
const webpack = require("webpack");
const plugin_1 = require("../plugin/plugin");
const federation_1 = require("./federation");
function config(env) {
    const cumulocityPlugin = new plugin_1.CumulocityPlugin({
        app: env.app,
        mode: env.mode,
        indexTemplate: env.app.indexTemplate
    });
    const fileCopy = {
        patterns: [
            {
                from: 'node_modules/@angular/common/locales/*.js',
                to: `assets/locales/angular/[name][ext]`
            },
            ...(env.app.copy || []),
            {
                from: 'README.md',
                to: 'README.md',
                globOptions: {
                    caseSensitiveMatch: false
                },
                noErrorOnMissing: true
            }
        ]
    };
    return {
        plugins: [
            new webpack.DefinePlugin({
                __WEBPACK__: true,
                __BETA__: JSON.stringify(env.beta),
                __MODE__: JSON.stringify(env.mode),
                __ENTRY_APP__: JSON.stringify(env.entry),
                __ENTRY_BRANDING__: JSON.stringify(cumulocityPlugin.getBrandingEntry(env.branding)),
                __VERSION_NG1__: JSON.stringify(cumulocityPlugin.getVersion('@c8y/ng1-modules/package.json')),
                __VERSION_NGX__: JSON.stringify(cumulocityPlugin.getVersion('@c8y/ngx-components/package.json'))
            }),
            federation_1.getModuleFederationExports(env),
            cumulocityPlugin,
            new CopyPlugin(fileCopy)
        ].filter(Boolean)
    };
}
exports.config = config;
