// @c8y/ngx-components/esm2015/search/search-box.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { Router } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { TypeaheadComponent } from '@c8y/ngx-components';
import { DeviceGridService } from '@c8y/ngx-components/device-grid';
import { defer, empty, merge, of, pipe } from 'rxjs';
import { map, switchMap, tap } from 'rxjs/operators';
import { AssetSearchService } from './search.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components/device-grid';
import * as ɵngcc4 from './search.service';
import * as ɵngcc5 from 'ngx-bootstrap/dropdown';
import * as ɵngcc6 from '@c8y/ngx-components';
import * as ɵngcc7 from '@angular/forms';
import * as ɵngcc8 from '@angular/common';

const _c0 = ["dropdown"];
function SearchBoxComponent_div_4_c8y_li_5_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 22);
    ɵngcc0.ɵɵelementStart(1, "div", 18);
    ɵngcc0.ɵɵelementStart(2, "p", 23);
    ɵngcc0.ɵɵelementStart(3, "em", 24);
    ɵngcc0.ɵɵtext(4, "Searching by exact match. Click for other search options:");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 25);
    ɵngcc0.ɵɵelementStart(6, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.filter(ctx_r15.term + "*"); });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_10_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.filter("*" + ctx_r17.term + "*"); });
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_14_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.filter("*" + ctx_r18.term); });
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 6, "Starts with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 8, "Starts with"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 10, "Contains"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 12, "Contains"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 14, "Ends with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 16, "Ends with"), " ");
} }
function SearchBoxComponent_div_4_c8y_li_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Recent search results");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_7_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    const result_r19 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r19);
} }
function SearchBoxComponent_div_4_c8y_li_7_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_7_Template(rf, ctx) { if (rf & 1) {
    const _r24 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_7_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r24); const result_r19 = restoredCtx.$implicit; const ctx_r23 = ɵngcc0.ɵɵnextContext(2); return ctx_r23.open($event, result_r19); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_7_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_7_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r19 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r19.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r19.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r19.name || "--", " ");
} }
function SearchBoxComponent_div_4_c8y_li_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Recently registered devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_10_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    const result_r25 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r25);
} }
function SearchBoxComponent_div_4_c8y_li_10_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_10_Template(rf, ctx) { if (rf & 1) {
    const _r30 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_10_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r30); const result_r25 = restoredCtx.$implicit; const ctx_r29 = ɵngcc0.ɵɵnextContext(2); return ctx_r29.open($event, result_r25); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_10_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_10_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r25 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r25.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r25.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r25.name || "--", " ");
} }
function SearchBoxComponent_div_4_c8y_li_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Search results");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_12_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    const result_r31 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r31);
} }
function SearchBoxComponent_div_4_c8y_li_12_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_12_Template(rf, ctx) { if (rf & 1) {
    const _r36 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_12_Template_c8y_li_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r36); const result_r31 = restoredCtx.$implicit; const ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.open($event, result_r31); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_12_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_12_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const result_r31 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r31.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r31.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r31.name || "--", " ");
} }
function SearchBoxComponent_div_4_ng_template_13_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 35);
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 36);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 24);
    ɵngcc0.ɵɵtext(4, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "small", 24);
    ɵngcc0.ɵɵtext(6, " Use a filter or go to the asset data table to show all devices and groups. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "search");
} }
function SearchBoxComponent_div_4_ng_template_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SearchBoxComponent_div_4_ng_template_13_c8y_li_0_Template, 7, 2, "c8y-li", 34);
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.noMatch);
} }
function SearchBoxComponent_div_4_c8y_li_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 37);
    ɵngcc0.ɵɵelementStart(1, "div", 38);
    ɵngcc0.ɵɵelement(2, "div", 39);
    ɵngcc0.ɵɵelement(3, "div", 40);
    ɵngcc0.ɵɵelement(4, "div", 41);
    ɵngcc0.ɵɵelement(5, "div", 42);
    ɵngcc0.ɵɵelement(6, "div", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SearchBoxComponent_div_4_ng_template_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 44);
    ɵngcc0.ɵɵelementStart(1, "div", 38);
    ɵngcc0.ɵɵelement(2, "div", 39);
    ɵngcc0.ɵɵelement(3, "div", 40);
    ɵngcc0.ɵɵelement(4, "div", 41);
    ɵngcc0.ɵɵelement(5, "div", 42);
    ɵngcc0.ɵɵelement(6, "div", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function () { return []; };
const _c2 = function (a0) { return { data: a0 }; };
function SearchBoxComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    const _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelementStart(1, "form", 6, 7);
    ɵngcc0.ɵɵelementStart(3, "c8y-typeahead", 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function SearchBoxComponent_div_4_Template_c8y_typeahead_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r38 = ɵngcc0.ɵɵnextContext(); return ctx_r38.selected = $event; })("keydown", function SearchBoxComponent_div_4_Template_c8y_typeahead_keydown_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r40 = ɵngcc0.ɵɵnextContext(); return ctx_r40.keyDown($event); })("onIconClick", function SearchBoxComponent_div_4_Template_c8y_typeahead_onIconClick_3_listener() { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r41 = ɵngcc0.ɵɵnextContext(); return ctx_r41.reset(); });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtemplate(5, SearchBoxComponent_div_4_c8y_li_5_Template, 18, 18, "c8y-li", 9);
    ɵngcc0.ɵɵtemplate(6, SearchBoxComponent_div_4_c8y_li_6_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵtemplate(7, SearchBoxComponent_div_4_c8y_li_7_Template, 5, 3, "c8y-li", 11);
    ɵngcc0.ɵɵtemplate(8, SearchBoxComponent_div_4_c8y_li_8_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵtemplate(10, SearchBoxComponent_div_4_c8y_li_10_Template, 5, 3, "c8y-li", 12);
    ɵngcc0.ɵɵtemplate(11, SearchBoxComponent_div_4_c8y_li_11_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵtemplate(12, SearchBoxComponent_div_4_c8y_li_12_Template, 5, 3, "c8y-li", 13);
    ɵngcc0.ɵɵtemplate(13, SearchBoxComponent_div_4_ng_template_13_Template, 1, 1, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(15, SearchBoxComponent_div_4_c8y_li_15_Template, 7, 0, "c8y-li", 15);
    ɵngcc0.ɵɵtemplate(16, SearchBoxComponent_div_4_ng_template_16_Template, 7, 0, "ng-template", null, 16, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(18, "c8y-li", 17);
    ɵngcc0.ɵɵelementStart(19, "div", 18);
    ɵngcc0.ɵɵelement(20, "i", 19);
    ɵngcc0.ɵɵelementStart(21, "p", 20);
    ɵngcc0.ɵɵtext(22, "Need more filter possibilities?");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(23, "button", 21);
    ɵngcc0.ɵɵlistener("mousedown", function SearchBoxComponent_div_4_Template_button_mousedown_23_listener() { ɵngcc0.ɵɵrestoreView(_r39); const ctx_r42 = ɵngcc0.ɵɵnextContext(); return ctx_r42.openSearch(); });
    ɵngcc0.ɵɵtext(24, " Go to the asset data table ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r10 = ɵngcc0.ɵɵreference(14);
    const _r13 = ɵngcc0.ɵɵreference(17);
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    let tmp_7_0;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(4, 20, "Search for groups or assets\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.selected)("icon", ctx_r1.term ? "times" : "search")("allowFreeEntries", false);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length === 0 && ctx_r1.recentSearchResults.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.term.length === 0 ? ctx_r1.recentSearchResults : ɵngcc0.ɵɵpureFunction0(24, _c1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length === 0 && ((tmp_7_0 = ɵngcc0.ɵɵpipeBind1(9, 22, ctx_r1.recentlyRegisteredResults$)) == null ? null : tmp_7_0.data == null ? null : tmp_7_0.data.length) > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r1.term.length === 0 ? ctx_r1.recentlyRegisteredResults$ : ɵngcc0.ɵɵpureFunction1(26, _c2, ɵngcc0.ɵɵpureFunction0(25, _c1)))("c8yForLoadMore", "none")("c8yForPipe", ctx_r1.filterPipe);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r1.results$)("c8yForLoadMore", "auto")("c8yForPipe", ctx_r1.filterPipe)("c8yForNotFound", _r10)("c8yForLoadingTemplate", _r13)("c8yForLoadNextLabel", "Find more\u2026");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.isLoading);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("selectable", false);
} }
export class SearchBoxComponent {
    constructor(router, inventory, deviceGridService, searchService) {
        this.router = router;
        this.inventory = inventory;
        this.deviceGridService = deviceGridService;
        this.searchService = searchService;
        this.term = '';
        this.filterPipe = pipe(map((data) => {
            return this.searchService.filterOnlyAssets(data);
        }));
        this.recentSearchResults = [];
        this.isLoading = false;
        this.noMatch = false;
        this.RECENT_SEARCH_STORAGE_KEY = 'recent_search_view';
        this.MAX_RECENT_SEARCH_RESULTS = 5;
        this.DEFAULT_FILTER = {
            withTotalPages: true,
            pageSize: 5,
            withChildren: false
        };
        this.KEYCODE_ENTER = 13;
        this.KEYCODE_ESC = 27;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const recentSearchIds = JSON.parse(localStorage.getItem(this.RECENT_SEARCH_STORAGE_KEY));
            if (recentSearchIds && recentSearchIds.length > 0) {
                const { data } = yield this.inventory.list({ ids: recentSearchIds.join(',') });
                this.recentSearchResults = data;
                this.recentlyRegisteredResults$ = defer(() => this.inventory.list(Object.assign({ q: '$orderby=creationTime desc' }, this.DEFAULT_FILTER)));
            }
        });
    }
    onOpenChange(isOpen) {
        if (isOpen) {
            // needs to request an animation frame as
            // otherwise the typeahead is undefined
            requestAnimationFrame(() => {
                this.subscribeOnSearch();
                this.typeahead.dropdown.show();
                this.typeahead.searchControl.nativeElement.focus();
            });
        }
    }
    open(event, mo) {
        event.stopPropagation();
        const isAlreadyRecent = this.recentSearchResults.find(({ id }) => id === mo.id);
        if (!isAlreadyRecent) {
            this.recentSearchResults.unshift(mo);
            this.recentSearchResults = this.recentSearchResults.slice(0, this.MAX_RECENT_SEARCH_RESULTS);
        }
        const recentSearchIds = this.recentSearchResults.map(({ id }) => id);
        localStorage.setItem(this.RECENT_SEARCH_STORAGE_KEY, JSON.stringify(recentSearchIds));
        this.router.navigateByUrl(this.deviceGridService.getHref(mo, '/'));
        this.dropdown.hide();
    }
    reset() {
        this.typeahead.onSearch.emit('');
        this.selected = undefined;
        this.typeahead.searchControl.nativeElement.focus();
    }
    keyDown(event) {
        const keyCode = event.keyCode;
        if (keyCode === this.KEYCODE_ENTER) {
            // enter hit can be faster then typeahead debounce,
            // therefore we take the term from the DOM element
            // itself:
            const searchTerm = event.target.value;
            this.navigate(['/assetsearch'], { queryParams: { search: searchTerm } });
            this.dropdown.hide();
        }
        else if (keyCode === this.KEYCODE_ESC) {
            if (this.term === '') {
                this.dropdown.hide();
            }
            this.reset();
        }
    }
    filter(on) {
        this.navigate(['/assetsearch'], { queryParams: { filter: on } });
        this.dropdown.hide();
    }
    openSearch() {
        this.router.navigateByUrl('/assetsearch');
        this.dropdown.hide();
    }
    subscribeOnSearch() {
        if (!this.results$) {
            this.results$ = this.typeahead.onSearch.pipe(tap(term => this.onTypingStarted(term)), switchMap(term => this.mergeRequest(term)));
        }
    }
    navigate(commands, extras) {
        this.router
            .navigateByUrl('/', { skipLocationChange: true })
            .then(() => this.router.navigate(commands, extras));
    }
    mergeRequest(term) {
        return merge(of({ data: [] }), this.queryInventoryService(term).pipe(tap(({ data, paging }) => this.onLoadingDone(data, paging))));
    }
    queryInventoryService(term) {
        if (term) {
            return defer(() => this.searchService.search(term));
        }
        return empty();
    }
    onLoadingDone(data, paging) {
        this.isLoading = false;
        this.noMatch =
            paging && paging.nextPage === null && this.searchService.filterOnlyAssets(data).length === 0;
    }
    onTypingStarted(term) {
        this.noMatch = false;
        this.term = term;
        this.isLoading = term.length > 0;
    }
}
SearchBoxComponent.ɵfac = function SearchBoxComponent_Factory(t) { return new (t || SearchBoxComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AssetSearchService)); };
SearchBoxComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchBoxComponent, selectors: [["c8y-search-box"]], viewQuery: function SearchBoxComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(TypeaheadComponent, 5);
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.typeahead = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropdown = _t.first);
    } }, decls: 5, vars: 1, consts: [["dropdown", "", 1, "dropdown", 3, "insideClick", "isOpenChange"], ["dropdown", "bs-dropdown"], ["dropdownToggle", "", "type", "button", "title", "Search", "aria-controls", "searchDropdown", 1, "main-header-button", "dropdown-toggle", "c8y-dropdown"], ["c8yIcon", "search", 1, "icon-2x"], ["id", "searchDropdown", "class", "search-header-menu dropdown-menu dropdown-menu-center", 4, "dropdownMenu"], ["id", "searchDropdown", 1, "search-header-menu", "dropdown-menu", "dropdown-menu-center"], ["novalidate", "", 1, "c8y-search-form"], ["searchForm", "ngForm"], ["name", "selected", "container", "body", 3, "ngModel", "placeholder", "icon", "allowFreeEntries", "ngModelChange", "keydown", "onIconClick"], ["class", "p-l-16 p-r-16", 4, "ngIf"], ["class", "p-l-24 p-r-24", 3, "selectable", 4, "ngIf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "ngFor", "ngForOf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], ["class", "c8y-list__item--link  m-l-16 m-r-16", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe", "c8yForNotFound", "c8yForLoadingTemplate", "c8yForLoadNextLabel"], ["notFoundTemplate", ""], ["class", "p-t-32 p-b-0 p-relative", 4, "ngIf"], ["loading", ""], [1, "m-t-24", "bg-gray-lighter", "p-t-16", "p-b-16", "p-l-24", "p-r-24", 3, "selectable"], [1, "flex-row"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-4"], ["translate", "", 1, "m-r-8"], ["type", "button", "translate", "", 1, "m-l-auto", "btn", "btn-default", "btn-sm", 3, "mousedown"], [1, "p-l-16", "p-r-16"], [1, "m-r-4", "text-muted"], ["translate", ""], [1, "btn-group", "btn-group-sm"], [1, "btn", "btn-default", 3, "title", "click"], [1, "p-l-24", "p-r-24", 3, "selectable"], [1, "legend", "form-block"], [1, "c8y-list__item--link", "m-l-16", "m-r-16", 3, "click"], [3, "mo", 4, "ngIf"], ["class", "c8y-icon-duocolor", 3, "c8yIcon", 4, "ngIf"], [3, "mo"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], ["class", "p-16 c8y-empty-state", 3, "selectable", 4, "ngIf"], [1, "p-16", "c8y-empty-state", 3, "selectable"], [3, "icon"], [1, "p-t-32", "p-b-0", "p-relative"], [1, "spinner", 2, "right", "0"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [1, "text-center", "p-t-32", "p-b-0", "p-relative"]], template: function SearchBoxComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵlistener("isOpenChange", function SearchBoxComponent_Template_div_isOpenChange_0_listener($event) { return ctx.onOpenChange($event); });
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵelement(3, "i", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, SearchBoxComponent_div_4_Template, 25, 28, "div", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("insideClick", true);
    } }, directives: [ɵngcc5.BsDropdownDirective, ɵngcc5.BsDropdownToggleDirective, ɵngcc6.IconDirective, ɵngcc5.BsDropdownMenuDirective, ɵngcc7.ɵNgNoValidate, ɵngcc7.NgControlStatusGroup, ɵngcc7.NgForm, ɵngcc6.TypeaheadComponent, ɵngcc7.NgControlStatus, ɵngcc7.NgModel, ɵngcc8.NgIf, ɵngcc8.NgForOf, ɵngcc6.ForOfDirective, ɵngcc6.ListItemComponent, ɵngcc6.C8yTranslateDirective, ɵngcc6.ListItemIconComponent, ɵngcc6.DeviceStatusComponent], pipes: [ɵngcc6.C8yTranslatePipe, ɵngcc8.AsyncPipe], encapsulation: 2 });
SearchBoxComponent.ctorParameters = () => [
    { type: Router },
    { type: InventoryService },
    { type: DeviceGridService },
    { type: AssetSearchService }
];
SearchBoxComponent.propDecorators = {
    typeahead: [{ type: ViewChild, args: [TypeaheadComponent, { static: false },] }],
    dropdown: [{ type: ViewChild, args: ['dropdown', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchBoxComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search-box',
                template: "<div\n  class=\"dropdown\"\n  dropdown\n  #dropdown=\"bs-dropdown\"\n  [insideClick]=\"true\"\n  (isOpenChange)=\"onOpenChange($event)\"\n>\n  <button\n    class=\"main-header-button dropdown-toggle c8y-dropdown\"\n    dropdownToggle\n    type=\"button\"\n    title=\"Search\"\n    aria-controls=\"searchDropdown\"\n  >\n    <i c8yIcon=\"search\" class=\"icon-2x\"></i>\n  </button>\n\n  <div\n    id=\"searchDropdown\"\n    *dropdownMenu\n    class=\"search-header-menu dropdown-menu dropdown-menu-center\"\n  >\n    <form novalidate #searchForm=\"ngForm\" class=\"c8y-search-form\">\n      <c8y-typeahead\n        [(ngModel)]=\"selected\"\n        placeholder=\"{{ 'Search for groups or assets\u2026' | translate }}\"\n        (keydown)=\"keyDown($event)\"\n        (onIconClick)=\"reset()\"\n        [icon]=\"term ? 'times' : 'search'\"\n        [allowFreeEntries]=\"false\"\n        name=\"selected\"\n        container=\"body\"\n      >\n        <!-- filter buttons -->\n        <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-16 p-r-16\">\n          <div class=\"flex-row\">\n            <p class=\"m-r-4 text-muted\">\n              <em translate>Searching by exact match. Click for other search options:</em>\n            </p>\n            <div class=\"btn-group btn-group-sm\">\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Starts with' | translate }}\"\n                (click)=\"filter(term + '*')\"\n              >\n                {{ 'Starts with' | translate }}\n              </button>\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Contains' | translate }}\"\n                (click)=\"filter('*' + term + '*')\"\n              >\n                {{ 'Contains' | translate }}\n              </button>\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Ends with' | translate }}\"\n                (click)=\"filter('*' + term)\"\n              >\n                {{ 'Ends with' | translate }}\n              </button>\n            </div>\n          </div>\n        </c8y-li>\n\n        <!-- Recent search -->\n        <c8y-li\n          *ngIf=\"term.length === 0 && recentSearchResults.length > 0\"\n          [selectable]=\"false\"\n          class=\"p-l-24 p-r-24\"\n        >\n          <div class=\"legend form-block\">\n            <span translate>Recent search results</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *ngFor=\"let result of term.length === 0 ? recentSearchResults : []\"\n          class=\"c8y-list__item--link m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- Recently registered devices -->\n        <c8y-li\n          *ngIf=\"term.length === 0 && (recentlyRegisteredResults$ | async)?.data?.length > 0\"\n          class=\"p-l-24 p-r-24\"\n          [selectable]=\"false\"\n        >\n          <div class=\"legend form-block\">\n            <span translate>Recently registered devices</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *c8yFor=\"\n            let result of term.length === 0 ? recentlyRegisteredResults$ : { data: [] };\n            loadMore: 'none';\n            pipe: filterPipe\n          \"\n          class=\"c8y-list__item--link m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- Search results -->\n        <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-24 p-r-24\" [selectable]=\"false\">\n          <div class=\"legend form-block\">\n            <span translate>Search results</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *c8yFor=\"\n            let result of results$;\n            loadMore: 'auto';\n            pipe: filterPipe;\n            notFound: notFoundTemplate;\n            loadingTemplate: loading;\n            loadNextLabel: 'Find more\u2026'\n          \"\n          class=\"c8y-list__item--link  m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- No search results found entry -->\n        <ng-template #notFoundTemplate>\n          <c8y-li *ngIf=\"noMatch\" class=\"p-16 c8y-empty-state\" [selectable]=\"false\">\n            <c8y-li-icon [icon]=\"'search'\"></c8y-li-icon>\n            <p><strong translate>No match found.</strong></p>\n            <small translate>\n              Use a filter or go to the asset data table to show all devices and groups.\n            </small>\n          </c8y-li>\n        </ng-template>\n\n        <!-- loading bar first entries -->\n        <c8y-li *ngIf=\"isLoading\" class=\"p-t-32 p-b-0 p-relative\">\n          <div class=\"spinner\" style=\"right:0;\">\n            <div class=\"rect1\"></div>\n            <div class=\"rect2\"></div>\n            <div class=\"rect3\"></div>\n            <div class=\"rect4\"></div>\n            <div class=\"rect5\"></div>\n          </div>\n        </c8y-li>\n\n        <!-- loading bar for loading more entries (inventory roles) -->\n        <ng-template #loading>\n          <c8y-li class=\"text-center p-t-32 p-b-0 p-relative\">\n            <div class=\"spinner\" style=\"right:0;\">\n              <div class=\"rect1\"></div>\n              <div class=\"rect2\"></div>\n              <div class=\"rect3\"></div>\n              <div class=\"rect4\"></div>\n              <div class=\"rect5\"></div>\n            </div>\n          </c8y-li>\n        </ng-template>\n\n        <!-- more filter possibilities -->\n        <c8y-li class=\"m-t-24 bg-gray-lighter p-t-16 p-b-16 p-l-24 p-r-24\" [selectable]=\"false\">\n          <div class=\"flex-row\">\n            <i c8yIcon=\"info-circle\" class=\"text-info m-r-4\"></i>\n            <p translate class=\"m-r-8\">Need more filter possibilities?</p>\n            <button\n              type=\"button\"\n              class=\"m-l-auto btn btn-default btn-sm\"\n              translate\n              (mousedown)=\"openSearch()\"\n            >\n              Go to the asset data table\n            </button>\n          </div>\n        </c8y-li>\n      </c8y-typeahead>\n    </form>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.DeviceGridService }, { type: ɵngcc4.AssetSearchService }]; }, { typeahead: [{
            type: ViewChild,
            args: [TypeaheadComponent, { static: false }]
        }], dropdown: [{
            type: ViewChild,
            args: ['dropdown', { static: false }]
        }] }); })();
