// @c8y/ngx-components/esm2015/core/provider-configuration/provider-configuration-node.factory.js
import { Inject, Injectable, Injector, Optional } from '@angular/core';
import { flatten } from 'lodash-es';
import { forkJoin, of } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { toObservable } from '../common';
import { NavigatorNode } from '../navigator';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
import * as ɵngcc0 from '@angular/core';
export class ProviderConfigurationNodeFactory {
    constructor(config, injector) {
        this.injector = injector;
        this.config = flatten(config);
    }
    get() {
        if (!this.config || !this.config.length) {
            return;
        }
        if (!this.nodes) {
            const canActivate = this.config
                .map(this.getGuards.bind(this))
                .map(this.checkCanActivate.bind(this));
            return forkJoin(canActivate).pipe(map((canActivateResult) => this.config
                .map((c, index) => canActivateResult[index] ? new NavigatorNode(c.navigation) : undefined)
                .filter(el => !!el)), tap(nodes => (this.nodes = nodes)));
        }
        return this.nodes;
    }
    checkCanActivate(ca) {
        if (!!ca && ca.length) {
            const canActivateResult = ca
                .map((canActivate) => canActivate.canActivate(undefined, undefined))
                .map(toObservable);
            return forkJoin(canActivateResult).pipe(map((caResult) => caResult.reduce((acc, curr) => acc && curr)));
        }
        return of(true);
    }
    getGuards(c) {
        let result = [];
        if (c.tab && c.tab.canActivate && c.tab.canActivate.length) {
            result = [...result, ...c.tab.canActivate.map(ca => this.injector.get(ca))];
        }
        if (c.navigation.canActivate && c.navigation.canActivate.length) {
            result = [...result, ...c.navigation.canActivate.map(ca => this.injector.get(ca))];
        }
        return result;
    }
}
ProviderConfigurationNodeFactory.ɵfac = function ProviderConfigurationNodeFactory_Factory(t) { return new (t || ProviderConfigurationNodeFactory)(ɵngcc0.ɵɵinject(HOOK_DYNAMIC_PROVIDER_CONFIG, 8), ɵngcc0.ɵɵinject(ɵngcc0.Injector)); };
ProviderConfigurationNodeFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ProviderConfigurationNodeFactory, factory: ProviderConfigurationNodeFactory.ɵfac });
ProviderConfigurationNodeFactory.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [HOOK_DYNAMIC_PROVIDER_CONFIG,] }] },
    { type: Injector }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProviderConfigurationNodeFactory, [{
        type: Injectable
    }], function () { return [{ type: Array, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [HOOK_DYNAMIC_PROVIDER_CONFIG]
            }] }, { type: ɵngcc0.Injector }]; }, null); })();
