// @c8y/ngx-components/esm2015/core/provider-configuration/provider-configuration.module.js
import { forwardRef, NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { CoreModule } from '../core.module';
import { DynamicFormsModule } from '../dynamic-forms/dynamic-forms.module';
import { HOOK_NAVIGATOR_NODES } from '../navigator';
import { HOOK_ROUTE } from '../router';
import { HOOK_TABS } from '../tabs';
import { HOOK_DYNAMIC_PROVIDER_CONFIG } from './provider-configuration-hook';
import { ProviderConfigurationNodeFactory } from './provider-configuration-node.factory';
import { ProviderConfigurationRouteFactory } from './provider-configuration-route.factory';
import { ProviderConfigurationTabFactory } from './provider-configuration-tab.factory';
import { ProviderConfigurationComponent } from './provider-configuration.component';
import * as ɵngcc0 from '@angular/core';
export class ProviderConfigurationModule {
    static config(config) {
        return {
            ngModule: ProviderConfigurationModule,
            providers: [{ provide: HOOK_DYNAMIC_PROVIDER_CONFIG, useValue: config, multi: true }]
        };
    }
}
ProviderConfigurationModule.ɵfac = function ProviderConfigurationModule_Factory(t) { return new (t || ProviderConfigurationModule)(); };
ProviderConfigurationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ProviderConfigurationModule });
ProviderConfigurationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_ROUTE,
            useClass: forwardRef(() => ProviderConfigurationRouteFactory),
            multi: true
        },
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: forwardRef(() => ProviderConfigurationNodeFactory),
            multi: true
        },
        {
            provide: HOOK_TABS,
            useClass: forwardRef(() => ProviderConfigurationTabFactory),
            multi: true
        }
    ], imports: [[CoreModule, CommonModule, DynamicFormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProviderConfigurationModule, [{
        type: NgModule,
        args: [{
                declarations: [ProviderConfigurationComponent],
                exports: [ProviderConfigurationComponent],
                imports: [CoreModule, CommonModule, DynamicFormsModule],
                entryComponents: [ProviderConfigurationComponent],
                providers: [
                    {
                        provide: HOOK_ROUTE,
                        useClass: forwardRef(() => ProviderConfigurationRouteFactory),
                        multi: true
                    },
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: forwardRef(() => ProviderConfigurationNodeFactory),
                        multi: true
                    },
                    {
                        provide: HOOK_TABS,
                        useClass: forwardRef(() => ProviderConfigurationTabFactory),
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ProviderConfigurationModule, { declarations: function () { return [ProviderConfigurationComponent]; }, imports: function () { return [CoreModule, CommonModule, DynamicFormsModule]; }, exports: function () { return [ProviderConfigurationComponent]; } }); })();
