// @c8y/ngx-components/esm2015/sub-assets/sub-assets.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { DeviceGroupService } from '@c8y/ngx-components/assets-navigator';
import { SubAssetsService } from './sub-assets.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc3 from './sub-assets.service';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from './group-info.component';
import * as ɵngcc7 from './sub-assets-grid.component';
import * as ɵngcc8 from './add-group/add-group.component';
import * as ɵngcc9 from './assign-devices/assign-devices.component';
import * as ɵngcc10 from './assign-devices/assign-child-devices.component';

function SubAssetsComponent_c8y_action_bar_item_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 11);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.showAddGroup = !ctx_r5.showAddGroup; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵelementStart(4, "span", 14);
    ɵngcc0.ɵɵtext(5, "Add group");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Add group"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r0.shouldDisableAddGroup);
} }
function SubAssetsComponent_c8y_action_bar_item_3_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 11);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_3_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.showAssignDevices = !ctx_r7.showAssignDevices; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵelementStart(4, "span", 14);
    ɵngcc0.ɵɵtext(5, "Assign devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Assign devices"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r1.shouldDisableAssignDevices);
} }
function SubAssetsComponent_c8y_add_group_9_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-add-group", 15);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_add_group_9_Template_c8y_add_group_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.showAddGroup = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("currentGroupId", ctx_r2.currentGroupId)("refresh", ctx_r2.refresh);
} }
function SubAssetsComponent_c8y_assign_devices_13_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-devices", 16);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.showAssignDevices = false; })("onShowChildDevices", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_onShowChildDevices_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.showAssignChildDevices = $event; })("selectedDevice", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_selectedDevice_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.showChildrenForDevice = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r3.refresh)("currentGroupId", ctx_r3.currentGroupId);
} }
function SubAssetsComponent_c8y_assign_child_devices_17_Template(rf, ctx) { if (rf & 1) {
    const _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-child-devices", 17);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_child_devices_17_Template_c8y_assign_child_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); const ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.showAssignChildDevices = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r4.refresh)("currentGroupId", ctx_r4.currentGroupId)("parentDevice", ctx_r4.showChildrenForDevice);
} }
const _c0 = function (a0) { return { drawerOpen: a0 }; };
export class SubAssetsComponent {
    constructor(route, deviceGroupService, subAssetsService) {
        this.route = route;
        this.deviceGroupService = deviceGroupService;
        this.subAssetsService = subAssetsService;
        this.toggle = false;
        this.showAddGroup = false;
        this.showAssignDevices = false;
        this.showAssignChildDevices = false;
        this.refresh = new EventEmitter();
        this.filterable = true;
        this.shouldDisableAddGroup = false;
        this.shouldDisableAssignDevices = false;
        this.dataSub = this.route.parent.data.subscribe(({ contextData }) => {
            this.title = contextData.name;
            this.currentGroupId = contextData.id;
            this.group = contextData;
            this.filterable = !this.deviceGroupService.isDynamicGroup(contextData);
        });
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.shouldDisableAddGroup = !this.subAssetsService.canCreateGroup();
            this.shouldDisableAssignDevices = yield !this.subAssetsService.canAssignDevice(this.group);
            this.displayOptions = {
                striped: true,
                bordered: false,
                gridHeader: true,
                filter: !this.isSmartGroup()
            };
        });
    }
    groupChange(group) {
        /**
         * Group is cached in the context-route guard. To avoid loading old data (when switching tabs) we need to update
         * this object in the cache.
         */
        Object.assign(this.group, group);
        this.title = group.name;
        this.refresh.emit();
    }
    isSmartGroup() {
        return this.subAssetsService.isSmartGroup(this.group);
    }
    ngOnDestroy() {
        if (this.dataSub) {
            this.dataSub.unsubscribe();
        }
    }
}
SubAssetsComponent.ɵfac = function SubAssetsComponent_Factory(t) { return new (t || SubAssetsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGroupService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.SubAssetsService)); };
SubAssetsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SubAssetsComponent, selectors: [["c8y-sub-assets"]], decls: 18, vars: 20, consts: [[3, "placement", 4, "ngIf"], [1, "card", "content-fullpage", "d-flex", "d-col"], [3, "group", "onGroupChange"], [1, "d-contents", 3, "refresh", "parent-group", "filterable", "displayOptions"], [3, "ngClass"], [1, "bottom-drawer"], [3, "currentGroupId", "refresh", "onCancel", 4, "ngIf"], [1, "d-flex", "d-col", "no-align-items", "fit-h"], ["class", "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice", 4, "ngIf"], [1, "bottom-drawer", "m-t-40"], ["class", "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel", 4, "ngIf"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle", 1, "m-r-4"], ["translate", ""], [3, "currentGroupId", "refresh", "onCancel"], [1, "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice"], [1, "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel"]], template: function SubAssetsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, SubAssetsComponent_c8y_action_bar_item_2_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵtemplate(3, SubAssetsComponent_c8y_action_bar_item_3_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "div", 1);
        ɵngcc0.ɵɵelementStart(5, "c8y-group-info", 2);
        ɵngcc0.ɵɵlistener("onGroupChange", function SubAssetsComponent_Template_c8y_group_info_onGroupChange_5_listener($event) { return ctx.groupChange($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "c8y-sub-assets-grid", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵtemplate(9, SubAssetsComponent_c8y_add_group_9_Template, 1, 2, "c8y-add-group", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 4);
        ɵngcc0.ɵɵelementStart(11, "div", 5);
        ɵngcc0.ɵɵelementStart(12, "div", 7);
        ɵngcc0.ɵɵtemplate(13, SubAssetsComponent_c8y_assign_devices_13_Template, 1, 2, "c8y-assign-devices", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 4);
        ɵngcc0.ɵɵelementStart(15, "div", 9);
        ɵngcc0.ɵɵelementStart(16, "div", 7);
        ɵngcc0.ɵɵtemplate(17, SubAssetsComponent_c8y_assign_child_devices_17_Template, 1, 3, "c8y-assign-child-devices", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.title, "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("group", ctx.group);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("refresh", ctx.refresh)("parent-group", ctx.group)("filterable", ctx.filterable)("displayOptions", ctx.displayOptions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c0, ctx.showAddGroup));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(16, _c0, ctx.showAssignDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignDevices);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(18, _c0, ctx.showAssignChildDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignChildDevices);
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc5.NgIf, ɵngcc6.GroupInfoComponent, ɵngcc7.SubAssetsGridComponent, ɵngcc5.NgClass, ɵngcc4.ActionBarItemComponent, ɵngcc4.IconDirective, ɵngcc4.C8yTranslateDirective, ɵngcc8.AddGroupComponent, ɵngcc9.AssignDevicesComponent, ɵngcc10.AssignChildDevicesComponent], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
SubAssetsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: DeviceGroupService },
    { type: SubAssetsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-sub-assets',
                template: "<c8y-title>\n  {{ title }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n    [disabled]=\"shouldDisableAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Assign devices' | translate }}\"\n    (click)=\"showAssignDevices = !showAssignDevices\"\n    [disabled]=\"shouldDisableAssignDevices\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Assign devices</span>\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"card content-fullpage d-flex d-col\">\n  <c8y-group-info [group]=\"group\" (onGroupChange)=\"groupChange($event)\"></c8y-group-info>\n  <c8y-sub-assets-grid\n    [refresh]=\"refresh\"\n    [parent-group]=\"group\"\n    [filterable]=\"filterable\"\n    [displayOptions]=\"displayOptions\"\n    class=\"d-contents\"\n  ></c8y-sub-assets-grid>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [currentGroupId]=\"currentGroupId\"\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignDevices }\">\n  <div class=\"bottom-drawer\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-devices\n        class=\"d-contents\"\n        (onCancel)=\"showAssignDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        (onShowChildDevices)=\"showAssignChildDevices = $event\"\n        (selectedDevice)=\"showChildrenForDevice = $event\"\n        *ngIf=\"showAssignDevices\"\n      ></c8y-assign-devices>\n    </div>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignChildDevices }\">\n  <div class=\"bottom-drawer m-t-40\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-child-devices\n        *ngIf=\"showAssignChildDevices\"\n        class=\"d-contents\"\n        (onCancel)=\"showAssignChildDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        [parentDevice]=\"showChildrenForDevice\"\n      ></c8y-assign-child-devices>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.DeviceGroupService }, { type: ɵngcc3.SubAssetsService }]; }, null); })();
