// @c8y/ngx-components/esm2015/sub-assets/add-group/add-group.module.js
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_ROUTE } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { AddGroupComponent } from './add-group.component';
import { AddGroupService } from './add-group.service';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        path: 'add-group',
        component: AddGroupComponent
    }
];
export class AddGroupModule {
}
AddGroupModule.ɵfac = function AddGroupModule_Factory(t) { return new (t || AddGroupModule)(); };
AddGroupModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AddGroupModule });
AddGroupModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        AddGroupService
    ], imports: [[CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddGroupModule, [{
        type: NgModule,
        args: [{
                declarations: [AddGroupComponent],
                imports: [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule],
                entryComponents: [AddGroupComponent],
                exports: [AddGroupComponent],
                providers: [
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    AddGroupService
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AddGroupModule, { declarations: function () { return [AddGroupComponent]; }, imports: function () { return [CoreModule, DeviceGridModule, FormsModule, ReactiveFormsModule]; }, exports: function () { return [AddGroupComponent]; } }); })();
export { ɵ0 };
