// @c8y/ngx-components/esm2015/sub-assets/add-group/add-group.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, ViewChild, Input, HostListener } from '@angular/core';
import { Validators, FormBuilder } from '@angular/forms';
import { DeviceGridService } from '@c8y/ngx-components/device-grid';
import { C8yStepper, AlertService, gettext } from '@c8y/ngx-components';
import { AddGroupService } from './add-group.service';
import { SubAssetsService } from '../sub-assets.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/device-grid';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from './add-group.service';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '../sub-assets.service';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from '@angular/cdk/stepper';

const _c0 = ["nameRef"];
function AddGroupComponent_c8y_title_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-title");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Add group"), "\n");
} }
function AddGroupComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementContainer(1, 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    const _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function AddGroupComponent_ng_template_2_c8y_message_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 15);
    ɵngcc0.ɵɵtext(1, "This field is required.");
    ɵngcc0.ɵɵelementEnd();
} }
function AddGroupComponent_ng_template_2_div_31_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵtext(1, " You don't have permission to assign devices. ");
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = function () { return { edit: true, done: false }; };
const _c2 = function () { return ["col-md-6", "col-md-offset-3", "m-t-24", "m-b-40", "p-0", "flex-no-shrink"]; };
const _c3 = function (a0, a1) { return { next: a0, cancel: a1 }; };
const _c4 = function (a0, a1) { return { custom: a0, cancel: a1 }; };
const _c5 = function () { return { next: true, cancel: true }; };
const _c6 = function () { return { custom: true, cancel: true }; };
const _c7 = function () { return []; };
const _c8 = function (a0) { return { custom: a0 }; };
function AddGroupComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 5);
    ɵngcc0.ɵɵelementStart(1, "cdk-step", 6);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "div", 7);
    ɵngcc0.ɵɵelementStart(4, "div", 8);
    ɵngcc0.ɵɵelementStart(5, "div", 9);
    ɵngcc0.ɵɵelementStart(6, "h4", 10);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 11);
    ɵngcc0.ɵɵelementStart(10, "div", 12);
    ɵngcc0.ɵɵelementStart(11, "div", 13);
    ɵngcc0.ɵɵelementStart(12, "div", 8);
    ɵngcc0.ɵɵelementStart(13, "div", 9);
    ɵngcc0.ɵɵelementStart(14, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(15, "div", 14);
    ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(17, "label", 15);
    ɵngcc0.ɵɵtext(18, "Name");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(19, "input", 16, 17);
    ɵngcc0.ɵɵpipe(21, "translate");
    ɵngcc0.ɵɵelementStart(22, "c8y-messages");
    ɵngcc0.ɵɵtemplate(23, AddGroupComponent_ng_template_2_c8y_message_23_Template, 2, 0, "c8y-message", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(25, "label", 15);
    ɵngcc0.ɵɵtext(26, "Description");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(27, "input", 19);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(29, "c8y-form-group");
    ɵngcc0.ɵɵelement(30, "div", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(31, AddGroupComponent_ng_template_2_div_31_Template, 2, 0, "div", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(32, "c8y-stepper-buttons", 21);
    ɵngcc0.ɵɵlistener("onCancel", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCancel_32_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onCancel.emit(); })("onCustom", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCustom_32_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.createGroup(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(33, "cdk-step", 22);
    ɵngcc0.ɵɵpipe(34, "translate");
    ɵngcc0.ɵɵelementStart(35, "div", 7);
    ɵngcc0.ɵɵelementStart(36, "div", 8);
    ɵngcc0.ɵɵelementStart(37, "div", 9);
    ɵngcc0.ɵɵelementStart(38, "h4", 10);
    ɵngcc0.ɵɵtext(39);
    ɵngcc0.ɵɵpipe(40, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(41, "div", 23);
    ɵngcc0.ɵɵelementStart(42, "c8y-device-grid", 24);
    ɵngcc0.ɵɵlistener("itemsSelect", function AddGroupComponent_ng_template_2_Template_c8y_device_grid_itemsSelect_42_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.onSelected($event); });
    ɵngcc0.ɵɵpipe(43, "translate");
    ɵngcc0.ɵɵelementStart(44, "div", 25);
    ɵngcc0.ɵɵelement(45, "h1", 26);
    ɵngcc0.ɵɵelementStart(46, "div");
    ɵngcc0.ɵɵelementStart(47, "p");
    ɵngcc0.ɵɵelementStart(48, "strong");
    ɵngcc0.ɵɵtext(49);
    ɵngcc0.ɵɵpipe(50, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(51, "small");
    ɵngcc0.ɵɵtext(52);
    ɵngcc0.ɵɵpipe(53, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(54, "c8y-stepper-buttons", 27);
    ɵngcc0.ɵɵlistener("onCancel", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCancel_54_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onCancel.emit(); })("onCustom", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCustom_54_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.createGroup(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r4 = ɵngcc0.ɵɵreference(20);
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(45, _c1))("customClasses", ɵngcc0.ɵɵpureFunction0(46, _c2));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("stepControl", ctx_r3.formGroupStepOne)("label", ɵngcc0.ɵɵpipeBind1(2, 27, "New group"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 29, "New group"), " ");
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r3.formGroupStepOne);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(21, 31, "e.g. First floor"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.formGroupStepOne.untouched && !_r4.value);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(28, 33, "e.g. first floor devices"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r3.formGroupStepOne);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.canAssignDevice);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r3.canCreateGroup)("labels", ctx_r3.canAssignDevice ? ɵngcc0.ɵɵpureFunction2(47, _c3, ctx_r3.btnLabels.next, ctx_r3.btnLabels.cancel) : ɵngcc0.ɵɵpureFunction2(50, _c4, ctx_r3.btnLabels.create, ctx_r3.btnLabels.cancel))("showButtons", ctx_r3.canAssignDevice ? ɵngcc0.ɵɵpureFunction0(53, _c5) : ɵngcc0.ɵɵpureFunction0(54, _c6));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(34, 35, "Assign devices"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(40, 37, "Assign devices"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(43, 39, "Select target devices"))("actionControls", ɵngcc0.ɵɵpureFunction0(55, _c7))("infiniteScroll", "auto")("selectable", "true")("pagination", ctx_r3.pagination)("refresh", ctx_r3.refresh);
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(50, 41, "No matching devices."));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(53, 43, "Refine your search terms"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("labels", ɵngcc0.ɵɵpureFunction1(56, _c8, ctx_r3.btnLabels.create))("disabled", !ctx_r3.canAssignDevice)("pending", ctx_r3.pendingStatus);
} }
export class AddGroupComponent {
    constructor(deviceGridService, fb, addGroupService, alert, subAssetsService) {
        this.deviceGridService = deviceGridService;
        this.fb = fb;
        this.addGroupService = addGroupService;
        this.alert = alert;
        this.subAssetsService = subAssetsService;
        this.refresh = new EventEmitter();
        this.onDeviceQueryStringChange = new EventEmitter();
        this.onCancel = new EventEmitter();
        this.pendingStatus = false;
        this.pagination = { pageSize: 20, currentPage: 1 };
        this.selected = [];
        this.canCreateGroup = false;
        this.canAssignDevice = false;
        this.ITEMS_SELECT_LIMIT = 15;
        this.btnLabels = {
            next: gettext('Next'),
            cancel: gettext('Cancel'),
            create: gettext('Create')
        };
    }
    onEnterKeyDown(event) {
        // Order matters! Needs to be placed before this.stepper.next
        if ((this.isGroupDetailsStep() && !this.canAssignDevice) || this.isAssignDeviceStep()) {
            this.createGroup();
            return;
        }
        this.stepper.next();
    }
    onEscapeKeyDown(event) {
        this.onCancel.emit();
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.formGroupStepOne = this.fb.group({
                name: ['', Validators.required],
                description: ['']
            });
            this.subscription = this.onCancel.subscribe(() => this.resetStepper());
            this.canCreateGroup = this.subAssetsService.canCreateGroup();
            this.canAssignDevice = yield this.subAssetsService.canAssignDevice({
                id: this.currentGroupId
            });
        });
    }
    ngAfterViewInit() {
        this.nameInput = this.nameInputRef.nativeElement;
        this.setFocusOnNameInput();
    }
    createGroup() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.canCreateGroup === false) {
                return;
            }
            this.pendingStatus = true;
            yield this.addGroupService.createGroupAndAssignDevices(this.formGroupStepOne.value, this.currentGroupId, this.selected);
            this.pendingStatus = false;
            this.resetStepper();
            this.alert.success(gettext('Group created.'));
            this.refresh.emit();
            this.onCancel.emit();
        });
    }
    onSelected(selectedDevicesIDs) {
        this.selected = selectedDevicesIDs;
    }
    resetStepper() {
        this.stepper.reset();
        this.stepper.selectedIndex = 1;
        this.selected = [];
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    isGroupDetailsStep() {
        return this.stepper.selectedIndex === 0;
    }
    isAssignDeviceStep() {
        return this.stepper.selectedIndex === 1;
    }
    setFocusOnNameInput() {
        if (this.nameInput) {
            this.nameInput.focus();
            this.nameInput.select();
        }
    }
}
AddGroupComponent.ɵfac = function AddGroupComponent_Factory(t) { return new (t || AddGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AddGroupService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.SubAssetsService)); };
AddGroupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddGroupComponent, selectors: [["c8y-add-group"]], viewQuery: function AddGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(C8yStepper, 5);
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.nameInputRef = _t.first);
    } }, hostBindings: function AddGroupComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.enter", function AddGroupComponent_keydown_enter_HostBindingHandler($event) { return ctx.onEnterKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument)("keydown.escape", function AddGroupComponent_keydown_escape_HostBindingHandler($event) { return ctx.onEscapeKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { refresh: "refresh", currentGroupId: "currentGroupId" }, outputs: { onDeviceQueryStringChange: "onDeviceQueryStringChange", onCancel: "onCancel" }, decls: 4, vars: 3, consts: [[4, "ngIf"], ["class", "d-contents", 4, "ngIf", "ngIfElse"], ["stepper", ""], [1, "d-contents"], [3, "ngTemplateOutlet"], ["linear", "", 1, "flex-col", "flex-nowrap", "no-align-items", "fit-h", "c8y-stepper--no-btns", 3, "disableDefaultIcons", "customClasses"], [3, "stepControl", "label"], [1, "p-16", "p-t-0", "flex-no-shrink", "separator-bottom", "col-xs-12"], [1, "row"], [1, "col-md-6", "col-md-offset-3", "col-lg-4", "col-lg-offset-4"], [1, "text-center", "text-medium"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-b-0"], [3, "formGroup"], ["translate", ""], ["type", "text", "formControlName", "name", "maxlength", "254", "required", "", 1, "form-control", 3, "placeholder"], ["nameRef", ""], ["translate", "", 4, "ngIf"], ["type", "text", "formControlName", "description", 1, "form-control", 3, "placeholder"], ["class", "alert alert-info max-width-100", "translate", "", 4, "ngIf"], [1, "d-block", "card-footer", "p-24", "separator", 3, "disabled", "labels", "showButtons", "onCancel", "onCustom"], [3, "label"], [1, "col-xs-12", "no-gutter", "flex-grow"], [1, "d-contents", 3, "title", "actionControls", "infiniteScroll", "selectable", "pagination", "refresh", "itemsSelect"], [1, "c8y-empty-state"], ["c8yIcon", "search"], [1, "d-block", "card-footer", "p-24", "separator", 3, "labels", "disabled", "pending", "onCancel", "onCustom"], ["translate", "", 1, "alert", "alert-info", "max-width-100"]], template: function AddGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AddGroupComponent_c8y_title_0_Template, 3, 3, "c8y-title", 0);
        ɵngcc0.ɵɵtemplate(1, AddGroupComponent_div_1_Template, 2, 1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, AddGroupComponent_ng_template_2_Template, 55, 58, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.currentGroupId);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.currentGroupId)("ngIfElse", _r2);
    } }, directives: [ɵngcc6.NgIf, ɵngcc4.TitleComponent, ɵngcc6.NgTemplateOutlet, ɵngcc4.C8yStepper, ɵngcc7.CdkStep, ɵngcc4.FormGroupComponent, ɵngcc2.NgControlStatusGroup, ɵngcc2.FormGroupDirective, ɵngcc4.C8yTranslateDirective, ɵngcc2.DefaultValueAccessor, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc2.NgControlStatus, ɵngcc2.FormControlName, ɵngcc2.MaxLengthValidator, ɵngcc2.RequiredValidator, ɵngcc4.MessagesComponent, ɵngcc4.C8yStepperButtons, ɵngcc1.DeviceGridComponent, ɵngcc4.IconDirective, ɵngcc4.MessageDirective], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
AddGroupComponent.ctorParameters = () => [
    { type: DeviceGridService },
    { type: FormBuilder },
    { type: AddGroupService },
    { type: AlertService },
    { type: SubAssetsService }
];
AddGroupComponent.propDecorators = {
    currentGroupId: [{ type: Input }],
    refresh: [{ type: Input }],
    onDeviceQueryStringChange: [{ type: Output }],
    onCancel: [{ type: Output }],
    stepper: [{ type: ViewChild, args: [C8yStepper, { static: false },] }],
    nameInputRef: [{ type: ViewChild, args: ['nameRef', { static: false },] }],
    onEnterKeyDown: [{ type: HostListener, args: ['document:keydown.enter', ['$event'],] }],
    onEscapeKeyDown: [{ type: HostListener, args: ['document:keydown.escape', ['$event'],] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddGroupComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-group',
                template: "<c8y-title *ngIf=\"!currentGroupId\">\n  {{ 'Add group' | translate }}\n</c8y-title>\n\n<div class=\"d-contents\" *ngIf=\"!currentGroupId; else stepper\">\n  <ng-container [ngTemplateOutlet]=\"stepper\"></ng-container>\n</div>\n\n<ng-template #stepper>\n  <c8y-stepper\n    class=\"flex-col flex-nowrap no-align-items fit-h c8y-stepper--no-btns\"\n    [disableDefaultIcons]=\"{ edit: true, done: false }\"\n    [customClasses]=\"['col-md-6', 'col-md-offset-3', 'm-t-24', 'm-b-40', 'p-0', 'flex-no-shrink']\"\n    linear\n  >\n    <cdk-step [stepControl]=\"formGroupStepOne\" [label]=\"'New group' | translate\">\n      <div class=\"p-16 p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center text-medium\">\n              {{ 'New group' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-xs-12 flex-grow no-gutter\">\n        <div class=\"card-inner-scroll fit-h\">\n          <div class=\"card-block p-b-0\">\n            <div class=\"row\">\n              <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n                <c8y-form-group>\n                  <div [formGroup]=\"formGroupStepOne\">\n                    <c8y-form-group>\n                      <label translate>Name</label>\n                      <input\n                        class=\"form-control\"\n                        type=\"text\"\n                        formControlName=\"name\"\n                        placeholder=\"{{ 'e.g. First floor' | translate }} \"\n                        maxlength=\"254\"\n                        #nameRef\n                        required\n                      />\n                      <c8y-messages>\n                        <c8y-message *ngIf=\"!formGroupStepOne.untouched && !nameRef.value\" translate\n                          >This field is required.</c8y-message\n                        >\n                      </c8y-messages>\n                    </c8y-form-group>\n\n                    <c8y-form-group>\n                      <label translate>Description</label>\n                      <input\n                        class=\"form-control\"\n                        type=\"text\"\n                        formControlName=\"description\"\n                        placeholder=\"{{ 'e.g. first floor devices' | translate }}\"\n                      />\n                    </c8y-form-group>\n                  </div>\n                </c8y-form-group>\n                <c8y-form-group>\n                  <div [formGroup]=\"formGroupStepOne\"></div>\n                </c8y-form-group>\n                <div class=\"alert alert-info max-width-100\" translate *ngIf=\"!canAssignDevice\">\n                  You don't have permission to assign devices.\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        (onCancel)=\"onCancel.emit()\"\n        (onCustom)=\"createGroup()\"\n        [disabled]=\"!canCreateGroup\"\n        [labels]=\"\n          canAssignDevice\n            ? { next: btnLabels.next, cancel: btnLabels.cancel }\n            : { custom: btnLabels.create, cancel: btnLabels.cancel }\n        \"\n        [showButtons]=\"\n          canAssignDevice ? { next: true, cancel: true } : { custom: true, cancel: true }\n        \"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n    <cdk-step [label]=\"'Assign devices' | translate\">\n      <div class=\"p-16 p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center text-medium\">\n              {{ 'Assign devices' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-xs-12 no-gutter flex-grow\">\n        <c8y-device-grid\n          [title]=\"'Select target devices' | translate\"\n          [actionControls]=\"[]\"\n          [infiniteScroll]=\"'auto'\"\n          [selectable]=\"'true'\"\n          [pagination]=\"pagination\"\n          (itemsSelect)=\"onSelected($event)\"\n          [refresh]=\"refresh\"\n          class=\"d-contents\"\n        >\n          <div class=\"c8y-empty-state\">\n            <h1 c8yIcon=\"search\"></h1>\n            <div>\n              <p>\n                <strong>{{ 'No matching devices.' | translate }}</strong>\n              </p>\n              <small>{{ 'Refine your search terms' | translate }}</small>\n            </div>\n          </div>\n        </c8y-device-grid>\n      </div>\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        (onCancel)=\"onCancel.emit()\"\n        (onCustom)=\"createGroup()\"\n        [labels]=\"{ custom: btnLabels.create }\"\n        [disabled]=\"!canAssignDevice\"\n        [pending]=\"pendingStatus\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.DeviceGridService }, { type: ɵngcc2.FormBuilder }, { type: ɵngcc3.AddGroupService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.SubAssetsService }]; }, { refresh: [{
            type: Input
        }], onDeviceQueryStringChange: [{
            type: Output
        }], onCancel: [{
            type: Output
        }], onEnterKeyDown: [{
            type: HostListener,
            args: ['document:keydown.enter', ['$event']]
        }], onEscapeKeyDown: [{
            type: HostListener,
            args: ['document:keydown.escape', ['$event']]
        }], currentGroupId: [{
            type: Input
        }], stepper: [{
            type: ViewChild,
            args: [C8yStepper, { static: false }]
        }], nameInputRef: [{
            type: ViewChild,
            args: ['nameRef', { static: false }]
        }] }); })();
