// @c8y/ngx-components/esm2015/repository/configuration-device-tab/device-configuration.service.js
import { EventEmitter, Injectable } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class DeviceConfigurationService {
    constructor() {
        this.configurationsUpdated = new EventEmitter();
    }
    updateConfigurations(repositorySnapsOnly) {
        this.configurationsUpdated.emit(repositorySnapsOnly);
    }
    hasAnySupportedOperation(mo, operation) {
        const supported = mo.c8y_SupportedOperations;
        if (!supported) {
            return false;
        }
        if (!Array.isArray(operation)) {
            operation = [operation];
        }
        return supported.some(supportedOperation => operation.includes(supportedOperation));
    }
}
DeviceConfigurationService.ɵfac = function DeviceConfigurationService_Factory(t) { return new (t || DeviceConfigurationService)(); };
DeviceConfigurationService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceConfigurationService, factory: DeviceConfigurationService.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceConfigurationService, [{
        type: Injectable
    }], function () { return []; }, null); })();
