// @c8y/ngx-components/esm2015/repository/configuration-device-tab/save-to-repository.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { RepositoryService } from '../repository.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '../repository.service';
import * as ɵngcc4 from '@angular/forms';
export class SaveToRepositoryComponent {
    constructor(modal, alertService, repositoryService) {
        this.modal = modal;
        this.alertService = alertService;
        this.repositoryService = repositoryService;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            {
                try {
                    const configSnapshotData = {
                        selected: {
                            configurationType: this.configSnapshot.configurationType
                        },
                        version: this.configSnapshot.name,
                        deviceType: this.configSnapshot.deviceType,
                        description: this.configSnapshot.description,
                        binary: {
                            file: new File([this.configSnapshot.binary], this.configSnapshot.name)
                        }
                    };
                    yield this.repositoryService.save(configSnapshotData, RepositoryType.CONFIGURATION);
                    this.alertService.success(gettext('Configuration saved.'));
                    this._save();
                }
                catch (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
}
SaveToRepositoryComponent.ɵfac = function SaveToRepositoryComponent_Factory(t) { return new (t || SaveToRepositoryComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.RepositoryService)); };
SaveToRepositoryComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SaveToRepositoryComponent, selectors: [["c8y-save-config-to-configuration-repository"]], decls: 31, vars: 11, consts: [[1, "modal-header", "bg-primary", "text-center", "text-white"], [2, "font-size", "62px"], ["c8yIcon", "gears"], ["translate", "", 1, "text-uppercase"], [1, "modal-body"], [1, "p-t-24"], ["saveConfigurationSnapshot", "ngForm"], ["translate", "", "for", "name"], ["id", "name", "type", "text", "autocomplete", "off", "name", "name", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], ["translate", "", "for", "deviceType"], ["id", "deviceType", "rows", "6", "name", "deviceType", 1, "form-control", 3, "ngModel", "ngModelChange"], ["translate", "", "for", "description"], ["type", "text", "id", "description", "maxlength", "254", "autocomplete", "off", "name", "description", 1, "form-control", 3, "ngModel", "ngModelChange"], ["translate", "", "for", "configurationType"], ["id", "configurationType", "rows", "6", "name", "configurationType", 1, "form-control", 3, "ngModel", "ngModelChange"], [1, "modal-footer"], ["translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"]], template: function SaveToRepositoryComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, " Save configuration ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "form", 5, 6);
        ɵngcc0.ɵɵelementStart(8, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(9, "label", 7);
        ɵngcc0.ɵɵtext(10, "Name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "input", 8);
        ɵngcc0.ɵɵlistener("ngModelChange", function SaveToRepositoryComponent_Template_input_ngModelChange_11_listener($event) { return ctx.configSnapshot.name = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(13, "label", 9);
        ɵngcc0.ɵɵtext(14, "Device type");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "input", 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function SaveToRepositoryComponent_Template_input_ngModelChange_15_listener($event) { return ctx.configSnapshot.deviceType = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(17, "label", 11);
        ɵngcc0.ɵɵtext(18, "Description");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "input", 12);
        ɵngcc0.ɵɵlistener("ngModelChange", function SaveToRepositoryComponent_Template_input_ngModelChange_19_listener($event) { return ctx.configSnapshot.description = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(21, "label", 13);
        ɵngcc0.ɵɵtext(22, "Configuration type");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "input", 14);
        ɵngcc0.ɵɵlistener("ngModelChange", function SaveToRepositoryComponent_Template_input_ngModelChange_23_listener($event) { return ctx.configSnapshot.configurationType = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "div", 15);
        ɵngcc0.ɵɵelementStart(25, "button", 16);
        ɵngcc0.ɵɵlistener("click", function SaveToRepositoryComponent_Template_button_click_25_listener() { return ctx.close(); });
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵtext(27, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "button", 17);
        ɵngcc0.ɵɵlistener("click", function SaveToRepositoryComponent_Template_button_click_28_listener() { return ctx.save(); });
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵtext(30, " Save ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(7);
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵproperty("ngModel", ctx.configSnapshot.name);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngModel", ctx.configSnapshot.deviceType);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngModel", ctx.configSnapshot.description);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngModel", ctx.configSnapshot.configurationType);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(26, 7, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(29, 9, "Save configuration to repository"));
        ɵngcc0.ɵɵproperty("disabled", _r0.form.invalid);
    } }, directives: [ɵngcc2.IconDirective, ɵngcc2.C8yTranslateDirective, ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc2.FormGroupComponent, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc4.MaxLengthValidator], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
SaveToRepositoryComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: AlertService },
    { type: RepositoryService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SaveToRepositoryComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-save-config-to-configuration-repository',
                template: "<div class=\"modal-header bg-primary text-center text-white\">\n  <div style=\"font-size: 62px;\">\n    <span c8yIcon=\"gears\"></span>\n  </div>\n  <h4 class=\"text-uppercase\" translate>\n    Save configuration\n  </h4>\n</div>\n<div class=\"modal-body\">\n  <form #saveConfigurationSnapshot=\"ngForm\" class=\"p-t-24\">\n    <c8y-form-group>\n      <label translate for=\"name\">Name</label>\n      <input\n        id=\"name\"\n        type=\"text\"\n        class=\"form-control\"\n        autocomplete=\"off\"\n        name=\"name\"\n        [(ngModel)]=\"configSnapshot.name\"\n        required\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"deviceType\">Device type</label>\n      <input\n        id=\"deviceType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"deviceType\"\n        [(ngModel)]=\"configSnapshot.deviceType\"\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"description\">Description</label>\n      <input\n        type=\"text\"\n        id=\"description\"\n        class=\"form-control\"\n        maxlength=\"254\"\n        autocomplete=\"off\"\n        name=\"description\"\n        [(ngModel)]=\"configSnapshot.description\"\n      />\n    </c8y-form-group>\n    <c8y-form-group>\n      <label translate for=\"configurationType\">Configuration type</label>\n      <input\n        id=\"configurationType\"\n        class=\"form-control\"\n        rows=\"6\"\n        name=\"configurationType\"\n        [(ngModel)]=\"configSnapshot.configurationType\"\n      />\n    </c8y-form-group>\n  </form>\n</div>\n<div class=\"modal-footer\">\n  <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\" translate>\n    Cancel\n  </button>\n\n  <button\n    title=\"{{ 'Save configuration to repository' | translate }}\"\n    class=\"btn btn-primary\"\n    (click)=\"save()\"\n    [disabled]=\"saveConfigurationSnapshot.form.invalid\"\n    translate\n  >\n    Save\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.AlertService }, { type: ɵngcc3.RepositoryService }]; }, null); })();
