// @c8y/ngx-components/esm2015/repository/configuration-device-tab/source-code-preview.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class SourceCodePreviewComponent {
}
SourceCodePreviewComponent.ɵfac = function SourceCodePreviewComponent_Factory(t) { return new (t || SourceCodePreviewComponent)(); };
SourceCodePreviewComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SourceCodePreviewComponent, selectors: [["c8y-source-code-preview"]], inputs: { isDisabled: "isDisabled", text: "text" }, decls: 2, vars: 2, consts: [["rows", "4", 1, "text-monospace", "form-control", "no-resize", "flex-grow", 3, "disabled"]], template: function SourceCodePreviewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "textarea", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("disabled", ctx.isDisabled);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.text);
    } }, encapsulation: 2 });
SourceCodePreviewComponent.propDecorators = {
    isDisabled: [{ type: Input }],
    text: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SourceCodePreviewComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-source-code-preview',
                template: "<textarea\n  [disabled]=\"isDisabled\"\n  class=\"text-monospace form-control no-resize flex-grow\"\n  rows=\"4\"\n  >{{ text }}</textarea\n>\n"
            }]
    }], null, { isDisabled: [{
            type: Input
        }], text: [{
            type: Input
        }] }); })();
