// @c8y/ngx-components/esm2015/ecosystem/applications/add-web-application/add-web-application.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ApplicationType } from '@c8y/client';
import { gettext } from '@c8y/ngx-components';
import { EcosystemService } from '../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../ecosystem.service';
import * as ɵngcc2 from '../../shared/add-application.component';
export class AddWebApplicationComponent {
    constructor(ecosystemService) {
        this.ecosystemService = ecosystemService;
        this.headerText = gettext('Upload web application');
        this.successText = gettext('Application created');
        this.createApplicationHandler = (f) => this.createApplication(f);
        this.uploadApplicationHandler = (f, app) => this.uploadApplication(f, app);
    }
    createApplication(file) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.isValidAppType(file, ApplicationType.HOSTED);
            return this.ecosystemService.createAppForArchive(file);
        });
    }
    uploadApplication(file, app) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.ecosystemService.uploadArchiveToApp(file, app);
        });
    }
}
AddWebApplicationComponent.ɵfac = function AddWebApplicationComponent_Factory(t) { return new (t || AddWebApplicationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService)); };
AddWebApplicationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddWebApplicationComponent, selectors: [["c8y-add-web-application"]], decls: 1, vars: 6, consts: [[3, "headerText", "headerIcon", "successText", "createApplicationHandler", "uploadApplicationHandler", "canGoBack"]], template: function AddWebApplicationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-add-application", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("headerText", ctx.headerText)("headerIcon", "upload")("successText", ctx.successText)("createApplicationHandler", ctx.createApplicationHandler)("uploadApplicationHandler", ctx.uploadApplicationHandler)("canGoBack", true);
    } }, directives: [ɵngcc2.AddApplicationComponent], encapsulation: 2 });
AddWebApplicationComponent.ctorParameters = () => [
    { type: EcosystemService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddWebApplicationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-web-application',
                template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'upload'"
    [successText]="successText"
    [createApplicationHandler]="createApplicationHandler"
    [uploadApplicationHandler]="uploadApplicationHandler"
    [canGoBack]="true"
  ></c8y-add-application>`
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }]; }, null); })();
