// @c8y/ngx-components/esm2015/repository/software/software-details.component.js
import { __awaiter, __decorate } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { AlertService, gettext, memoize, ModalService, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BehaviorSubject, from, merge, Subject } from 'rxjs';
import { distinctUntilKeyChanged, map, shareReplay, switchMap, take, takeUntil, tap, withLatestFrom } from 'rxjs/operators';
import { RepositoryService } from '../repository.service';
import { AddSoftwareModalComponent } from './add-software-modal.component';
import { SoftwareTypeComponent } from './software-type.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../repository.service';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from 'ngx-bootstrap/modal';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '@angular/forms';
import * as ɵngcc9 from 'ngx-bootstrap/popover';
import * as ɵngcc10 from './software-type.component';
import * as ɵngcc11 from '../file-download/file-download.component';

function SoftwareDetailsComponent_button_7_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_button_7_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.addBaseVersion(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 30);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Add software"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Add software"), " ");
} }
function SoftwareDetailsComponent_div_101_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 31);
    ɵngcc0.ɵɵelement(2, "h1", 32);
    ɵngcc0.ɵɵelementStart(3, "h3", 33);
    ɵngcc0.ɵɵtext(4, "No versions to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 33);
    ɵngcc0.ɵɵtext(6, "Add a new version by clicking below.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "p");
    ɵngcc0.ɵɵelementStart(8, "button", 20);
    ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_div_101_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.addBaseVersion(); });
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 2, "Add software"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(11, 4, "Add software"), " ");
} }
function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_span_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 49);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Legacy"), " ");
} }
function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_17_Template(rf, ctx) { if (rf & 1) {
    const _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 50);
    ɵngcc0.ɵɵelementStart(1, "button", 51);
    ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_17_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r21); const baseVersion_r14 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.deleteBaseVersion(baseVersion_r14); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 52);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Delete"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Delete"), " ");
} }
function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_19_Template(rf, ctx) { if (rf & 1) {
    const _r24 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 53);
    ɵngcc0.ɵɵelementStart(1, "button", 54);
    ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_19_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r24); const baseVersion_r14 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r22 = ɵngcc0.ɵɵnextContext(2); return ctx_r22.deleteBaseVersion(baseVersion_r14); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 52);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Delete"));
} }
function SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 36);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 37);
    ɵngcc0.ɵɵelementStart(4, "div", 38);
    ɵngcc0.ɵɵelementStart(5, "p", 39);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 40);
    ɵngcc0.ɵɵelementStart(8, "p", 41);
    ɵngcc0.ɵɵelementStart(9, "span", 42);
    ɵngcc0.ɵɵtext(10, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "span", 43);
    ɵngcc0.ɵɵpipe(12, "async");
    ɵngcc0.ɵɵelement(13, "c8y-file-download", 44);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "div", 45);
    ɵngcc0.ɵɵtemplate(15, SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_span_15_Template, 3, 3, "span", 46);
    ɵngcc0.ɵɵpipe(16, "async");
    ɵngcc0.ɵɵtemplate(17, SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_17_Template, 6, 6, "div", 47);
    ɵngcc0.ɵɵpipe(18, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(19, SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_div_19_Template, 4, 3, "div", 48);
    ɵngcc0.ɵɵpipe(20, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r14 = ctx.$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", baseVersion_r14.c8y_Software.version);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", baseVersion_r14.c8y_Software.version, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(12, 7, ctx_r13.getBinaryName$(baseVersion_r14.c8y_Software.url)), "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("url", baseVersion_r14.c8y_Software.url);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(16, 9, ctx_r13.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(18, 11, ctx_r13.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(20, 13, ctx_r13.isLegacy$));
} }
const _c0 = function (a0) { return { "dd-low": a0 }; };
function SoftwareDetailsComponent_c8y_list_group_103_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 34);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵtemplate(2, SoftwareDetailsComponent_c8y_list_group_103_c8y_li_2_Template, 21, 15, "c8y-li", 35);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(7, _c0, ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r8.baseVersions$)) == null ? null : tmp_0_0.data.length) < 10));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(3, 5, ctx_r8.baseVersions$))("c8yForLoadMore", "auto");
} }
const _c1 = function (a0) { return { "icon-spin": a0 }; };
const _c2 = function (a0) { return { "width.ch": a0 }; };
export class SoftwareDetailsComponent {
    constructor(activatedRoute, inventoryService, repositoryService, alertService, translateService, modalService, bsModalService) {
        this.activatedRoute = activatedRoute;
        this.inventoryService = inventoryService;
        this.repositoryService = repositoryService;
        this.alertService = alertService;
        this.translateService = translateService;
        this.modalService = modalService;
        this.bsModalService = bsModalService;
        this.reload$ = new Subject();
        this.reloading$ = new BehaviorSubject(false);
        this.isSoftwareTypeChanged = false;
        this.updateSoftware$ = new Subject();
        this.softwareUpdated$ = new Subject();
        this.baseVersionsUpdated$ = new Subject();
        this.software$ = merge(this.activatedRoute.params.pipe(map(params => params.id), switchMap(id => from(this.inventoryService.detail(id).then(result => result.data)))), this.reload$.pipe(tap(() => this.reloading$.next(true)), switchMap(() => this.activatedRoute.params), map(params => params.id), switchMap(id => from(this.inventoryService.detail(id).then(result => result.data))), tap(() => this.reloading$.next(false))), this.softwareUpdated$).pipe(shareReplay(1));
        this.baseVersions$ = merge(this.software$.pipe(distinctUntilKeyChanged('id')), this.baseVersionsUpdated$, this.reload$).pipe(switchMap(() => this.software$), switchMap(software => this.repositoryService.listBaseVersions(software)), shareReplay(1));
        this.isLegacy$ = this.software$.pipe(map(software => this.repositoryService.isLegacyEntry(software)), shareReplay(1));
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        this.updateSoftware$
            .pipe(withLatestFrom(this.software$), switchMap(([softwarePartial, software]) => this.inventoryService.update(Object.assign({ id: software.id }, softwarePartial))), map(({ data }) => data), tap(software => this.softwareUpdated$.next(software)), tap(() => this.alertService.success(gettext('Saved.'))), takeUntil(this.destroy$))
            .subscribe();
        this.software$.subscribe(software => {
            this.softwareTypeObject = software;
        });
    }
    getBinaryName$(binaryUrl) {
        return this.repositoryService.getBinaryName$(binaryUrl);
    }
    addBaseVersion() {
        this.software$
            .pipe(take(1), switchMap(software => {
            const initialState = {
                model: {
                    selected: software,
                    description: software.description
                }
            };
            const config = {
                class: 'modal-sm',
                ignoreBackdropClick: true,
                initialState
            };
            const modalRef = this.bsModalService.show(AddSoftwareModalComponent, config);
            return modalRef.content.saved;
        }))
            .subscribe(() => this.baseVersionsUpdated$.next());
    }
    deleteBaseVersion(baseVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const title = gettext('Delete software');
                const body = `
        ${this.translateService.instant(gettext('You are about to delete software {{ version }}.'), { version: baseVersion.c8y_Software.version })}
        ${this.translateService.instant(gettext('This operation is irreversible.'))}
        ${this.translateService.instant(gettext('Do you want to proceed?'))}
      `;
                const labels = {
                    ok: gettext('Delete')
                };
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.repositoryService.delete(baseVersion);
                this.alertService.success(gettext('Software deleted.'));
                this.baseVersionsUpdated$.next();
            }
            catch (ex) {
                // only if not cancel from modal
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    onSelectSoftwareType(software) {
        var _a;
        this.isSoftwareTypeChanged = !(((_a = this.softwareTypeObject) === null || _a === void 0 ? void 0 : _a.softwareType) === (software === null || software === void 0 ? void 0 : software.softwareType));
        this.softwareTypeObject = software;
    }
    ngOnDestroy() {
        this.destroy$.next(true);
        this.destroy$.unsubscribe();
    }
}
SoftwareDetailsComponent.ɵfac = function SoftwareDetailsComponent_Factory(t) { return new (t || SoftwareDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.BsModalService)); };
SoftwareDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SoftwareDetailsComponent, selectors: [["c8y-software-details"]], viewQuery: function SoftwareDetailsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(SoftwareTypeComponent, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.softwareType = _t.first);
    } }, decls: 105, vars: 109, consts: [["path", "#/software", "icon", "c8y-tools", 3, "label"], [3, "placement"], ["class", "btn btn-link", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], [1, "row"], [1, "col-lg-12", "col-lg-max"], [1, "card", "m-b-4"], [1, "card-header", "separator"], ["translate", "", 1, "card-title"], [1, "card-block", "overflow-visible"], [1, "col-sm-6"], [1, "control-label"], [1, "input-group", "input-group-editable"], ["type", "text", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngStyle"], ["nameInput", "", "nameModel", "ngModel"], [1, "input-group-btn"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["type", "text", 1, "form-control", 3, "ngModel", "placeholder", "ngStyle"], ["descriptionInput", "", "descriptionModel", "ngModel"], [1, "btn", "btn-primary", 3, "title", "click"], ["container", "body", 1, "btn-clean", 3, "popover", "outsideClick"], ["c8yIcon", "question-circle-o", 1, "text-info"], ["deviceTypeInput", "", "deviceTypeModel", "ngModel"], [3, "softwareTypeMO", "onSelectSoftware"], [1, "btn", "btn-primary", 3, "ngClass", "title", "disabled", "click"], [1, "card"], [1, "card-block", "p-t-0", "p-b-24"], [4, "ngIf"], [3, "ngClass", 4, "ngIf"], ["c8yIcon", "plus-circle"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "c8y-tools", 1, "c8y-icon-duocolor"], ["translate", ""], [3, "ngClass"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["c8yIcon", "c8y-tools"], [1, "content-flex-50"], [1, "col-4"], [1, "text-truncate-wrap", 3, "title"], [1, "col-5"], [1, "text-truncate-wrap"], ["translate", "", 1, "text-label-small", "m-r-8"], [3, "title"], [3, "url"], [1, "col-2", "flex-row", "a-i-start"], ["class", "label label-warning flex-item-right-sm", 4, "ngIf"], ["class", "fit-h-20", 4, "ngIf"], ["class", "flex-item-right fit-h-20 p-r-8 hidden-xs", 4, "ngIf"], [1, "label", "label-warning", "flex-item-right-sm"], [1, "fit-h-20"], [1, "btn", "btn-danger", "btn-xs", "visible-xs", "flex-item-right", "m-r-8", "m-t-8", 3, "title", "click"], ["c8yIcon", "minus-circle"], [1, "flex-item-right", "fit-h-20", "p-r-8", "hidden-xs"], [1, "btn", "btn-dot", "text-danger", "showOnHover", 3, "title", "click"]], template: function SoftwareDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        const _r25 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(4, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(7, SoftwareDetailsComponent_button_7_Template, 5, 6, "button", 2);
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵelementStart(10, "button", 3);
        ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_Template_button_click_10_listener() { return ctx.reload$.next(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelement(12, "i", 4);
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 5);
        ɵngcc0.ɵɵelementStart(17, "div", 6);
        ɵngcc0.ɵɵelementStart(18, "div", 7);
        ɵngcc0.ɵɵelementStart(19, "div", 8);
        ɵngcc0.ɵɵelementStart(20, "h4", 9);
        ɵngcc0.ɵɵtext(21, "Software info");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "div", 10);
        ɵngcc0.ɵɵelementStart(23, "div", 5);
        ɵngcc0.ɵɵelementStart(24, "div", 11);
        ɵngcc0.ɵɵelementStart(25, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(26, "label", 12);
        ɵngcc0.ɵɵtext(27);
        ɵngcc0.ɵɵpipe(28, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(29, "div", 13);
        ɵngcc0.ɵɵelement(30, "input", 14, 15);
        ɵngcc0.ɵɵpipe(33, "async");
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵpipe(35, "async");
        ɵngcc0.ɵɵelement(36, "span");
        ɵngcc0.ɵɵelementStart(37, "div", 16);
        ɵngcc0.ɵɵelementStart(38, "button", 17);
        ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_Template_button_click_38_listener() { ɵngcc0.ɵɵrestoreView(_r25); const _r1 = ɵngcc0.ɵɵreference(31); const _r2 = ɵngcc0.ɵɵreference(32); ctx.updateSoftware$.next({ name: _r1.value }); return _r2.reset(); });
        ɵngcc0.ɵɵpipe(39, "translate");
        ɵngcc0.ɵɵtext(40);
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 11);
        ɵngcc0.ɵɵelementStart(43, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(44, "label", 12);
        ɵngcc0.ɵɵtext(45);
        ɵngcc0.ɵɵpipe(46, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(47, "div", 13);
        ɵngcc0.ɵɵelement(48, "input", 18, 19);
        ɵngcc0.ɵɵpipe(51, "async");
        ɵngcc0.ɵɵpipe(52, "translate");
        ɵngcc0.ɵɵpipe(53, "async");
        ɵngcc0.ɵɵelement(54, "span");
        ɵngcc0.ɵɵelementStart(55, "div", 16);
        ɵngcc0.ɵɵelementStart(56, "button", 20);
        ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_Template_button_click_56_listener() { ɵngcc0.ɵɵrestoreView(_r25); const _r3 = ɵngcc0.ɵɵreference(49); const _r4 = ɵngcc0.ɵɵreference(50); ctx.updateSoftware$.next({ description: _r3.value }); return _r4.reset(); });
        ɵngcc0.ɵɵpipe(57, "translate");
        ɵngcc0.ɵɵtext(58);
        ɵngcc0.ɵɵpipe(59, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(60, "div", 5);
        ɵngcc0.ɵɵelementStart(61, "div", 11);
        ɵngcc0.ɵɵelementStart(62, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(63, "label", 12);
        ɵngcc0.ɵɵtext(64);
        ɵngcc0.ɵɵpipe(65, "translate");
        ɵngcc0.ɵɵelementStart(66, "button", 21);
        ɵngcc0.ɵɵpipe(67, "translate");
        ɵngcc0.ɵɵelement(68, "i", 22);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(69, "div", 13);
        ɵngcc0.ɵɵelement(70, "input", 18, 23);
        ɵngcc0.ɵɵpipe(73, "async");
        ɵngcc0.ɵɵpipe(74, "translate");
        ɵngcc0.ɵɵpipe(75, "async");
        ɵngcc0.ɵɵelement(76, "span");
        ɵngcc0.ɵɵelementStart(77, "div", 16);
        ɵngcc0.ɵɵelementStart(78, "button", 20);
        ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_Template_button_click_78_listener() { ɵngcc0.ɵɵrestoreView(_r25); const _r5 = ɵngcc0.ɵɵreference(71); const _r6 = ɵngcc0.ɵɵreference(72); ctx.updateSoftware$.next({ c8y_Filter: { type: _r5.value } }); return _r6.reset(); });
        ɵngcc0.ɵɵpipe(79, "translate");
        ɵngcc0.ɵɵtext(80);
        ɵngcc0.ɵɵpipe(81, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(82, "div", 11);
        ɵngcc0.ɵɵelementStart(83, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(84, "label", 12);
        ɵngcc0.ɵɵtext(85);
        ɵngcc0.ɵɵpipe(86, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(87, "div", 13);
        ɵngcc0.ɵɵelementStart(88, "c8y-software-type", 24);
        ɵngcc0.ɵɵlistener("onSelectSoftware", function SoftwareDetailsComponent_Template_c8y_software_type_onSelectSoftware_88_listener($event) { return ctx.onSelectSoftwareType($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(89, "div", 16);
        ɵngcc0.ɵɵelementStart(90, "button", 25);
        ɵngcc0.ɵɵlistener("click", function SoftwareDetailsComponent_Template_button_click_90_listener() { ctx.updateSoftware$.next({ softwareType: ctx.softwareTypeObject.softwareType }); ctx.softwareType.resetInput(); return ctx.isSoftwareTypeChanged = false; });
        ɵngcc0.ɵɵpipe(91, "translate");
        ɵngcc0.ɵɵtext(92);
        ɵngcc0.ɵɵpipe(93, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(94, "div", 5);
        ɵngcc0.ɵɵelementStart(95, "div", 6);
        ɵngcc0.ɵɵelementStart(96, "div", 26);
        ɵngcc0.ɵɵelementStart(97, "div", 8);
        ɵngcc0.ɵɵelementStart(98, "h4", 9);
        ɵngcc0.ɵɵtext(99, "Versions");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(100, "div", 27);
        ɵngcc0.ɵɵtemplate(101, SoftwareDetailsComponent_div_101_Template, 12, 6, "div", 28);
        ɵngcc0.ɵɵpipe(102, "async");
        ɵngcc0.ɵɵtemplate(103, SoftwareDetailsComponent_c8y_list_group_103_Template, 4, 9, "c8y-list-group", 29);
        ɵngcc0.ɵɵpipe(104, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(31);
        let tmp_0_0;
        let tmp_10_0;
        let tmp_11_0;
        let tmp_17_0;
        let tmp_18_0;
        let tmp_25_0;
        let tmp_26_0;
        let tmp_36_0;
        let tmp_37_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 39, ctx.software$)) == null ? null : tmp_0_0.name, "\n");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(5, 41, "Software repository"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(8, 43, ctx.isLegacy$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 45, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(99, _c1, ɵngcc0.ɵɵpipeBind1(13, 47, ctx.reloading$)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 49, "Reload"), " ");
        ɵngcc0.ɵɵadvance(13);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(28, 51, "Name"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(34, 55, "e.g. My software"));
        ɵngcc0.ɵɵproperty("ngModel", (tmp_10_0 = ɵngcc0.ɵɵpipeBind1(33, 53, ctx.software$)) == null ? null : tmp_10_0.name)("ngStyle", ɵngcc0.ɵɵpureFunction1(101, _c2, ((tmp_11_0 = ɵngcc0.ɵɵpipeBind1(35, 57, ctx.software$)) == null ? null : tmp_11_0.name == null ? null : tmp_11_0.name.length) + 2 || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(39, 59, "Save"));
        ɵngcc0.ɵɵproperty("disabled", _r1.value.length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(41, 61, "Save"), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(46, 63, "Description"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(52, 67, "e.g. Cloud connectivity software"));
        ɵngcc0.ɵɵproperty("ngModel", (tmp_17_0 = ɵngcc0.ɵɵpipeBind1(51, 65, ctx.software$)) == null ? null : tmp_17_0.description)("ngStyle", ɵngcc0.ɵɵpureFunction1(103, _c2, ((tmp_18_0 = ɵngcc0.ɵɵpipeBind1(53, 69, ctx.software$)) == null ? null : tmp_18_0.description == null ? null : tmp_18_0.description.length) + 2 || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(57, 71, "Save"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(59, 73, "Save"), " ");
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(65, 75, "Device type"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(67, 77, "If the filter is set, the software will show up for installation only for devices of that type. If no filter is set, it will be available for all devices."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(74, 81, "e.g."), " c8y_Linux");
        ɵngcc0.ɵɵproperty("ngModel", (tmp_25_0 = ɵngcc0.ɵɵpipeBind1(73, 79, ctx.software$)) == null ? null : tmp_25_0.c8y_Filter == null ? null : tmp_25_0.c8y_Filter.type)("ngStyle", ɵngcc0.ɵɵpureFunction1(105, _c2, ((tmp_26_0 = ɵngcc0.ɵɵpipeBind1(75, 83, ctx.software$)) == null ? null : tmp_26_0.type == null ? null : tmp_26_0.type.length) + 2 || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(79, 85, "Save"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(81, 87, "Save"), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(86, 89, "Software type"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵstyleMap(ɵngcc0.ɵɵpureFunction1(107, _c2, (ctx.softwareTypeObject == null ? null : ctx.softwareTypeObject.softwareType == null ? null : ctx.softwareTypeObject.softwareType.length) + 2 || 31));
        ɵngcc0.ɵɵproperty("softwareTypeMO", ctx.softwareTypeObject);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(91, 91, "Save"));
        ɵngcc0.ɵɵproperty("ngClass", ctx.isSoftwareTypeChanged ? "" : "hidden")("disabled", (ctx.softwareTypeObject == null ? null : ctx.softwareTypeObject.softwareType) === "");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(93, 93, "Save"), " ");
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_36_0 = ɵngcc0.ɵɵpipeBind1(102, 95, ctx.baseVersions$)) == null ? null : tmp_36_0.data.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_37_0 = ɵngcc0.ɵɵpipeBind1(104, 97, ctx.baseVersions$)) == null ? null : tmp_37_0.data.length) > 0);
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc4.BreadcrumbComponent, ɵngcc4.BreadcrumbItemComponent, ɵngcc4.ActionBarItemComponent, ɵngcc7.NgIf, ɵngcc4.IconDirective, ɵngcc7.NgClass, ɵngcc4.C8yTranslateDirective, ɵngcc4.FormGroupComponent, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc8.DefaultValueAccessor, ɵngcc8.RequiredValidator, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc7.NgStyle, ɵngcc9.PopoverDirective, ɵngcc10.SoftwareTypeComponent, ɵngcc4.ListGroupComponent, ɵngcc4.ForOfDirective, ɵngcc4.ListItemComponent, ɵngcc4.ListItemIconComponent, ɵngcc4.ListItemBodyComponent, ɵngcc11.FileDownloadComponent], pipes: [ɵngcc7.AsyncPipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
SoftwareDetailsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: InventoryService },
    { type: RepositoryService },
    { type: AlertService },
    { type: TranslateService },
    { type: ModalService },
    { type: BsModalService }
];
SoftwareDetailsComponent.propDecorators = {
    softwareType: [{ type: ViewChild, args: [SoftwareTypeComponent, { static: true },] }]
};
__decorate([
    memoize()
], SoftwareDetailsComponent.prototype, "getBinaryName$", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SoftwareDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-software-details',
                template: "<c8y-title>\n  {{ (software$ | async)?.name }}\n</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item\n    path=\"#/software\"\n    label=\"{{ 'Software repository' | translate }}\"\n    icon=\"c8y-tools\"\n  >\n  </c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    *ngIf=\"!(isLegacy$ | async)\"\n    class=\"btn btn-link\"\n    title=\"{{ 'Add software' | translate }}\"\n    (click)=\"addBaseVersion()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add software' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"reload$.next()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': reloading$ | async }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"row\">\n  <div class=\"col-lg-12 col-lg-max\">\n    <div class=\"card m-b-4\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Software info</h4>\n      </div>\n      <div class=\"card-block overflow-visible\">\n        <div class=\"row\">\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Name' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #nameInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(software$ | async)?.name\"\n                  #nameModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g. My software' | translate }}\"\n                  [ngStyle]=\"{ 'width.ch': (software$ | async)?.name?.length + 2 || 31 }\"\n                  required\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\"updateSoftware$.next({ name: nameInput.value }); nameModel.reset()\"\n                    [disabled]=\"nameInput.value.length === 0\"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Description' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #descriptionInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(software$ | async)?.description\"\n                  #descriptionModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g. Cloud connectivity software' | translate }}\"\n                  [ngStyle]=\"{ 'width.ch': (software$ | async)?.description?.length + 2 || 31 }\"\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\"\n                      updateSoftware$.next({ description: descriptionInput.value });\n                      descriptionModel.reset()\n                    \"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n        </div>\n        <div class=\"row\">\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Device type' | translate }}\n                <button\n                  class=\"btn-clean\"\n                  popover=\"{{\n                    'If the filter is set, the software will show up for installation only for devices of that type. If no filter is set, it will be available for all devices.'\n                      | translate\n                  }}\"\n                  [outsideClick]=\"true\"\n                  container=\"body\"\n                >\n                  <i c8yIcon=\"question-circle-o\" class=\"text-info\"></i>\n                </button>\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #deviceTypeInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(software$ | async)?.c8y_Filter?.type\"\n                  #deviceTypeModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g.' | translate }} c8y_Linux\"\n                  [ngStyle]=\"{ 'width.ch': (software$ | async)?.type?.length + 2 || 31 }\"\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\"\n                      updateSoftware$.next({ c8y_Filter: { type: deviceTypeInput.value } });\n                      deviceTypeModel.reset()\n                    \"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n          <div class=\"col-sm-6\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Software type' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <c8y-software-type\n                  [softwareTypeMO]=\"softwareTypeObject\"\n                  [style]=\"{ 'width.ch': softwareTypeObject?.softwareType?.length + 2 || 31 }\"\n                  (onSelectSoftware)=\"onSelectSoftwareType($event)\"\n                >\n                </c8y-software-type>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    [ngClass]=\"isSoftwareTypeChanged ? '' : 'hidden'\"\n                    title=\"{{ 'Save' | translate }}\"\n                    [disabled]=\"softwareTypeObject?.softwareType === ''\"\n                    (click)=\"\n                      updateSoftware$.next({ softwareType: softwareTypeObject.softwareType });\n                      softwareType.resetInput();\n                      isSoftwareTypeChanged = false\n                    \"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n\n<div class=\"row\">\n  <div class=\"col-lg-12 col-lg-max\">\n    <div class=\"card\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Versions</h4>\n      </div>\n\n      <div class=\"card-block p-t-0 p-b-24\">\n        <div *ngIf=\"(baseVersions$ | async)?.data.length === 0\">\n          <div class=\"c8y-empty-state text-center\">\n            <h1 c8yIcon=\"c8y-tools\" class=\"c8y-icon-duocolor\"></h1>\n            <h3 translate>No versions to display.</h3>\n            <p translate>Add a new version by clicking below.</p>\n            <p>\n              <button\n                class=\"btn btn-primary\"\n                title=\"{{ 'Add software' | translate }}\"\n                (click)=\"addBaseVersion()\"\n              >\n                {{ 'Add software' | translate }}\n              </button>\n            </p>\n          </div>\n        </div>\n\n        <c8y-list-group\n          [ngClass]=\"{ 'dd-low': (baseVersions$ | async)?.data.length < 10 }\"\n          *ngIf=\"(baseVersions$ | async)?.data.length > 0\"\n        >\n          <c8y-li\n            *c8yFor=\"let baseVersion of baseVersions$ | async; let i = index; loadMore: 'auto'\"\n          >\n            <c8y-li-icon>\n              <i c8yIcon=\"c8y-tools\"></i>\n            </c8y-li-icon>\n\n            <c8y-li-body class=\"content-flex-50\">\n              <div class=\"col-4\">\n                <p class=\"text-truncate-wrap\" title=\"{{ baseVersion.c8y_Software.version }}\">\n                  {{ baseVersion.c8y_Software.version }}\n                </p>\n              </div>\n              <div class=\"col-5\">\n                <p class=\"text-truncate-wrap\">\n                  <span class=\"text-label-small m-r-8\" translate> File </span>\n                  <span title=\" {{ getBinaryName$(baseVersion.c8y_Software.url) | async }}\">\n                    <c8y-file-download url=\"{{ baseVersion.c8y_Software.url }}\"></c8y-file-download>\n                  </span>\n                </p>\n              </div>\n              <div class=\"col-2 flex-row a-i-start\">\n                <span *ngIf=\"isLegacy$ | async\" class=\"label label-warning flex-item-right-sm\">\n                  {{ 'Legacy' | translate }}\n                </span>\n\n                <div class=\"fit-h-20\" *ngIf=\"!(isLegacy$ | async)\">\n                  <button\n                    class=\"btn btn-danger btn-xs visible-xs flex-item-right m-r-8 m-t-8\"\n                    (click)=\"deleteBaseVersion(baseVersion)\"\n                    title=\"{{ 'Delete' | translate }}\"\n                  >\n                    <i c8yIcon=\"minus-circle\"></i>\n                    {{ 'Delete' | translate }}\n                  </button>\n                </div>\n              </div>\n              <div *ngIf=\"!(isLegacy$ | async)\" class=\"flex-item-right fit-h-20 p-r-8 hidden-xs\">\n                <button\n                  class=\"btn btn-dot text-danger showOnHover\"\n                  (click)=\"deleteBaseVersion(baseVersion)\"\n                  title=\"{{ 'Delete' | translate }}\"\n                >\n                  <i c8yIcon=\"minus-circle\"></i>\n                </button>\n              </div>\n            </c8y-li-body>\n          </c8y-li>\n        </c8y-list-group>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.RepositoryService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.TranslateService }, { type: ɵngcc4.ModalService }, { type: ɵngcc6.BsModalService }]; }, { softwareType: [{
            type: ViewChild,
            args: [SoftwareTypeComponent, { static: true }]
        }] }); })();
