// @c8y/ngx-components/esm2015/context-dashboard/widget-preview.component.js
import { Input, Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';

function WidgetPreviewComponent_label_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 8);
    ɵngcc0.ɵɵtext(1, "Tab and layout");
    ɵngcc0.ɵɵelementEnd();
} }
function WidgetPreviewComponent_label_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "label", 8);
    ɵngcc0.ɵɵtext(1, "Layout");
    ɵngcc0.ɵɵelementEnd();
} }
function WidgetPreviewComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelementStart(1, "div", 15);
    ɵngcc0.ɵɵelementStart(2, "ul", 16);
    ɵngcc0.ɵɵelementStart(3, "li", 17);
    ɵngcc0.ɵɵelementStart(4, "a", 18);
    ɵngcc0.ɵɵelement(5, "i", 19);
    ɵngcc0.ɵɵelementStart(6, "span", 20);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(8, "div", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r2.tab.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r2.tab.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.tab.name);
} }
const _c0 = function (a0) { return { "m-t-64": a0 }; };
export class WidgetPreviewComponent {
    constructor() {
        this.previewClasses = {
            'dashboard-theme-light': true,
            'panel-title-regular': true
        };
    }
}
WidgetPreviewComponent.ɵfac = function WidgetPreviewComponent_Factory(t) { return new (t || WidgetPreviewComponent)(); };
WidgetPreviewComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WidgetPreviewComponent, selectors: [["c8y-widget-preview"]], inputs: { previewClasses: "previewClasses", tab: "tab" }, decls: 19, vars: 7, consts: [["translate", "", 1, "legend", "form-block"], ["translate", "", 4, "ngIf"], [1, "dashboard-preview-slot", 3, "ngClass"], ["class", "page-tabs page-tabs-horizontal navigator-open p-absolute overflow-hidden", "style", "left: 0; margin: 0 15px;", 4, "ngIf"], [3, "ngClass"], [1, "card", "card-dashboard", 2, "height", "360px"], [1, "card-header-actions", 2, "z-index", "8"], [1, "card-title"], ["translate", ""], [1, "header-actions"], [1, "optionsBtn", "dropdown"], ["href", "", 1, "dropdown-toggle", "c8y-dropdown", 3, "click"], ["c8yIcon", "cog"], [1, "card-inner-scroll", "p-t-16"], [1, "page-tabs", "page-tabs-horizontal", "navigator-open", "p-absolute", "overflow-hidden", 2, "left", "0", "margin", "0 15px"], [1, "tabContainer", "hidden-xs"], [1, "nav", "nav-tabs", "nav-tabsc8y"], [1, "active", 2, "left", "16px"], ["routerlinkactive", "", 3, "title"], [1, "text-primary", 3, "c8yIcon"], [1, "txt"], [1, "visible-xs", "mobile-tabs"]], template: function WidgetPreviewComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtext(1, "Preview");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, WidgetPreviewComponent_label_2_Template, 2, 0, "label", 1);
        ɵngcc0.ɵɵtemplate(3, WidgetPreviewComponent_label_3_Template, 2, 0, "label", 1);
        ɵngcc0.ɵɵelementStart(4, "div", 2);
        ɵngcc0.ɵɵtemplate(5, WidgetPreviewComponent_div_5_Template, 9, 3, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵelementStart(7, "div", 5);
        ɵngcc0.ɵɵelementStart(8, "div", 6);
        ɵngcc0.ɵɵelementStart(9, "div", 7);
        ɵngcc0.ɵɵelementStart(10, "span", 8);
        ɵngcc0.ɵɵtext(11, "Widget title");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "div", 9);
        ɵngcc0.ɵɵelementStart(13, "div", 10);
        ɵngcc0.ɵɵelementStart(14, "a", 11);
        ɵngcc0.ɵɵlistener("click", function WidgetPreviewComponent_Template_a_click_14_listener() { return false; });
        ɵngcc0.ɵɵelement(15, "i", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 13);
        ɵngcc0.ɵɵelementStart(17, "p", 8);
        ɵngcc0.ɵɵtext(18, "Widget example content.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.tab);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.tab);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ctx.previewClasses);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.tab);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c0, !!ctx.tab));
    } }, directives: [ɵngcc1.C8yTranslateDirective, ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc1.IconDirective], encapsulation: 2 });
WidgetPreviewComponent.propDecorators = {
    previewClasses: [{ type: Input }],
    tab: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WidgetPreviewComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-widget-preview',
                template: "<div class=\"legend form-block\" translate>Preview</div>\n<label *ngIf=\"tab\" translate>Tab and layout</label>\n<label *ngIf=\"!tab\" translate>Layout</label>\n<div class=\"dashboard-preview-slot\" [ngClass]=\"previewClasses\">\n  <div\n    class=\"page-tabs page-tabs-horizontal navigator-open p-absolute overflow-hidden\"\n    style=\"left: 0; margin: 0 15px;\"\n    *ngIf=\"tab\"\n  >\n    <div class=\"tabContainer hidden-xs\">\n      <ul class=\"nav nav-tabs nav-tabsc8y\">\n        <li class=\"active\" style=\"left: 16px;\">\n          <a routerlinkactive=\"\" title=\"{{ tab.name }}\">\n            <i class=\"text-primary\" [c8yIcon]=\"tab.icon\"></i>\n            <span class=\"txt\">{{ tab.name }}</span></a\n          >\n        </li>\n      </ul>\n    </div>\n    <div class=\"visible-xs mobile-tabs\"></div>\n  </div>\n  <div [ngClass]=\"{ 'm-t-64': !!tab }\">\n    <div class=\"card card-dashboard\" style=\"height: 360px;\">\n      <div class=\"card-header-actions\" style=\"z-index: 8;\">\n        <div class=\"card-title\">\n          <span translate>Widget title</span>\n        </div>\n        <div class=\"header-actions\">\n          <div class=\"optionsBtn dropdown\">\n            <a href=\"\" (click)=\"(false)\" class=\"dropdown-toggle c8y-dropdown\">\n              <i c8yIcon=\"cog\"></i>\n            </a>\n          </div>\n        </div>\n      </div>\n      <div class=\"card-inner-scroll p-t-16\">\n        <p translate>Widget example content.</p>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { previewClasses: [{
            type: Input
        }], tab: [{
            type: Input
        }] }); })();
