// @c8y/ngx-components/esm2015/device-profile/device-tab-profile/device-tab-profile-detail.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from './device-profile-item-list.component';

function DeviceTabProfileDetailComponent_div_13_small_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r5.emptyStateDetails), " ");
} }
function DeviceTabProfileDetailComponent_div_13_ng_template_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 18);
    ɵngcc0.ɵɵtext(1, "No device profile selected");
    ɵngcc0.ɵɵelementEnd();
} }
function DeviceTabProfileDetailComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "div", 13);
    ɵngcc0.ɵɵelementStart(2, "div", 14);
    ɵngcc0.ɵɵelement(3, "h1", 15);
    ɵngcc0.ɵɵelementStart(4, "p");
    ɵngcc0.ɵɵelementStart(5, "span");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(8, "br");
    ɵngcc0.ɵɵtemplate(9, DeviceTabProfileDetailComponent_div_13_small_9_Template, 3, 3, "small", 16);
    ɵngcc0.ɵɵtemplate(10, DeviceTabProfileDetailComponent_div_13_ng_template_10_Template, 2, 0, "ng-template", null, 17, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const _r6 = ɵngcc0.ɵɵreference(11);
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r0.sectionIcon);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 4, ctx_r0.emptyStateText));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isProfileSelected)("ngIfElse", _r6);
} }
function DeviceTabProfileDetailComponent_div_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 19);
} }
function DeviceTabProfileDetailComponent_c8y_device_profile_item_list_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-device-profile-item-list", 20);
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("icon", ctx_r2.sectionIcon)("comparisonResults", ctx_r2.items)("showProfileItems", ctx_r2.isProfileSelected && !ctx_r2.isEmpty)("showTextLabel", ctx_r2.showTextLabel);
} }
function DeviceTabProfileDetailComponent_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 21);
} }
function DeviceTabProfileDetailComponent_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 22);
} }
export class DeviceTabProfileDetailComponent {
    constructor() {
        this.emptyStateText = '';
        this.emptyStateDetails = '';
        this.showTextLabel = true;
    }
}
DeviceTabProfileDetailComponent.ɵfac = function DeviceTabProfileDetailComponent_Factory(t) { return new (t || DeviceTabProfileDetailComponent)(); };
DeviceTabProfileDetailComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceTabProfileDetailComponent, selectors: [["c8y-device-tab-profile-detail"]], inputs: { emptyStateText: "emptyStateText", emptyStateDetails: "emptyStateDetails", showTextLabel: "showTextLabel", sectionTitle: "sectionTitle", sectionIcon: "sectionIcon", isProfileSelected: "isProfileSelected", isEmpty: "isEmpty", items: "items" }, decls: 18, vars: 11, consts: [[1, "card--grid", "grid__col--6-6--md"], [1, "bg-white", "card-block", "sticky-top"], [1, "legend", "form-block"], [1, "bg-gray-white", "card-block", "sticky-top", "hidden-xs", "hidden-sm"], [1, "bg-white", "p-l-16", "p-r-16"], [1, "m-0"], [1, "bg-gray-white", "p-l-16", "p-r-16"], ["class", "d-contents", 4, "ngIf"], ["class", "bg-gray-white", 4, "ngIf"], ["class", "d-contents", 3, "icon", "comparisonResults", "showProfileItems", "showTextLabel", 4, "ngIf"], ["class", "bg-white p-t-24", 4, "ngIf"], ["class", "bg-gray-white p-t-24", 4, "ngIf"], [1, "d-contents"], [1, "p-l-16", "p-r-16"], [1, "c8y-empty-state", "text-left"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], [4, "ngIf", "ngIfElse"], ["noItems", ""], ["translate", ""], [1, "bg-gray-white"], [1, "d-contents", 3, "icon", "comparisonResults", "showProfileItems", "showTextLabel"], [1, "bg-white", "p-t-24"], [1, "bg-gray-white", "p-t-24"]], template: function DeviceTabProfileDetailComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "h5", 2);
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 3);
        ɵngcc0.ɵɵelementStart(6, "h5", 2);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 4);
        ɵngcc0.ɵɵelement(10, "hr", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 6);
        ɵngcc0.ɵɵelement(12, "hr", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(13, DeviceTabProfileDetailComponent_div_13_Template, 12, 6, "div", 7);
        ɵngcc0.ɵɵtemplate(14, DeviceTabProfileDetailComponent_div_14_Template, 1, 0, "div", 8);
        ɵngcc0.ɵɵtemplate(15, DeviceTabProfileDetailComponent_c8y_device_profile_item_list_15_Template, 1, 4, "c8y-device-profile-item-list", 9);
        ɵngcc0.ɵɵtemplate(16, DeviceTabProfileDetailComponent_div_16_Template, 1, 0, "div", 10);
        ɵngcc0.ɵɵtemplate(17, DeviceTabProfileDetailComponent_div_17_Template, 1, 0, "div", 11);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 7, ctx.sectionTitle));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 9, ctx.sectionTitle));
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isProfileSelected || ctx.isEmpty);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.items.length === 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.items.length > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isProfileSelected && !ctx.isEmpty);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isProfileSelected && !ctx.isEmpty);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.IconDirective, ɵngcc2.C8yTranslateDirective, ɵngcc3.DeviceProfileItemListComponent], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
DeviceTabProfileDetailComponent.propDecorators = {
    sectionTitle: [{ type: Input }],
    sectionIcon: [{ type: Input }],
    emptyStateText: [{ type: Input }],
    emptyStateDetails: [{ type: Input }],
    isProfileSelected: [{ type: Input }],
    isEmpty: [{ type: Input }],
    items: [{ type: Input }],
    showTextLabel: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceTabProfileDetailComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-tab-profile-detail',
                template: "<div class=\"card--grid grid__col--6-6--md\">\n  <div class=\"bg-white card-block sticky-top\">\n    <h5 class=\"legend form-block\">{{ sectionTitle | translate }}</h5>\n  </div>\n  <div class=\"bg-gray-white  card-block sticky-top hidden-xs hidden-sm\">\n    <h5 class=\"legend form-block\">{{ sectionTitle | translate }}</h5>\n  </div>\n  <div class=\"bg-white p-l-16 p-r-16\">\n    <hr class=\"m-0\" />\n  </div>\n  <div class=\"bg-gray-white p-l-16 p-r-16\">\n    <hr class=\"m-0\" />\n  </div>\n  <div class=\"d-contents\" *ngIf=\"!isProfileSelected || isEmpty\">\n    <div class=\"p-l-16 p-r-16\">\n      <div class=\"c8y-empty-state text-left\">\n        <h1 [c8yIcon]=\"sectionIcon\" class=\"c8y-icon-duocolor\"></h1>\n        <p>\n          <span>{{ emptyStateText | translate }}</span\n          ><br />\n          <small *ngIf=\"isProfileSelected; else noItems\">\n            {{ emptyStateDetails | translate }}\n          </small>\n          <ng-template #noItems>\n            <small translate>No device profile selected</small>\n          </ng-template>\n        </p>\n      </div>\n    </div>\n  </div>\n  <div class=\"bg-gray-white\" *ngIf=\"items.length === 0\"></div>\n  <c8y-device-profile-item-list\n    *ngIf=\"items.length > 0\"\n    [icon]=\"sectionIcon\"\n    [comparisonResults]=\"items\"\n    [showProfileItems]=\"isProfileSelected && !isEmpty\"\n    [showTextLabel]=\"showTextLabel\"\n    class=\"d-contents\"\n  ></c8y-device-profile-item-list>\n  <div class=\"bg-white p-t-24\" *ngIf=\"isProfileSelected && !isEmpty\"></div>\n  <div class=\"bg-gray-white p-t-24\" *ngIf=\"isProfileSelected && !isEmpty\"></div>\n</div>\n"
            }]
    }], function () { return []; }, { emptyStateText: [{
            type: Input
        }], emptyStateDetails: [{
            type: Input
        }], showTextLabel: [{
            type: Input
        }], sectionTitle: [{
            type: Input
        }], sectionIcon: [{
            type: Input
        }], isProfileSelected: [{
            type: Input
        }], isEmpty: [{
            type: Input
        }], items: [{
            type: Input
        }] }); })();
