// ngx-bootstrap/rating/esm5/rating.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Default values provider for rating
 */
import * as ɵngcc0 from '@angular/core';
var RatingConfig = /** @class */ (function () {
    function RatingConfig() {
        /**
         * aria label for rating
         */
        this.ariaLabel = 'rating';
    }
    /** @nocollapse */ RatingConfig.ɵprov = i0.ɵɵdefineInjectable({ factory: function RatingConfig_Factory() { return new RatingConfig(); }, token: RatingConfig, providedIn: "root" });
RatingConfig.ɵfac = function RatingConfig_Factory(t) { return new (t || RatingConfig)(); };
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RatingConfig, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return RatingConfig;
}());
export { RatingConfig };
if (false) {
    /**
     * aria label for rating
     * @type {?}
     */
    RatingConfig.prototype.ariaLabel;
}
