// @c8y/ngx-components/esm2015/ecosystem/applications/application-plugins/application-plugins.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { BehaviorSubject, combineLatest } from 'rxjs';
import { ActivatedRoute } from '@angular/router';
import { EcosystemService } from '../../ecosystem.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { InstallPluginComponent } from './install-plugin.component';
import { AlertService, gettext, ModalService, PluginsService, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { map, shareReplay } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '../../ecosystem.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from './plugin-list.component';
import * as ɵngcc7 from '@angular/common';

const _c0 = function (a0) { return { "btn-pending": a0 }; };
export class ApplicationPluginsComponent {
    constructor(activatedRoute, ecosystemService, bsModalService, modal, translateService, pluginsService, alertService) {
        this.activatedRoute = activatedRoute;
        this.ecosystemService = ecosystemService;
        this.bsModalService = bsModalService;
        this.modal = modal;
        this.translateService = translateService;
        this.pluginsService = pluginsService;
        this.alertService = alertService;
        this.exportedPlugins$ = new BehaviorSubject([]);
        this.remotePlugins$ = new BehaviorSubject({});
        this.allAvailablePlugins$ = new BehaviorSubject([]);
        this.installedPlugins$ = combineLatest([
            this.remotePlugins$.pipe(map((remotes) => this.getRemotePluginsList(remotes))),
            this.allAvailablePlugins$
        ]).pipe(map(([remotePlugins, allPlugins]) => allPlugins.filter((p) => remotePlugins.includes(p.id))), shareReplay(1));
        this.pluginsToDelete = [];
        this.archives = [];
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            const { id } = this.activatedRoute.snapshot.parent.data.contextData;
            this.app = yield this.ecosystemService.getApplication(id);
            this.archives = yield this.ecosystemService.listArchives(this.app.id);
            this.archives.sort((a, b) => {
                return new Date(b.created) - new Date(a.created);
            });
            const manifest = this.app.manifest;
            if (manifest) {
                const exports = manifest.exports || [];
                this.exportedPlugins$.next(this.extendPluginList(exports, this.app.contextPath));
                const remotes = manifest.remotes || {};
                this.remotePlugins$.next(remotes);
            }
            this.allAvailablePlugins$.next(yield this.getAllPlugins());
        });
    }
    installPlugins() {
        return __awaiter(this, void 0, void 0, function* () {
            const initialState = {
                plugins$: combineLatest([
                    this.allAvailablePlugins$,
                    this.installedPlugins$.pipe(map((plugins) => plugins.map((p) => p.id))),
                    this.exportedPlugins$.pipe(map((plugins) => plugins.map((p) => p.id)))
                ]).pipe(map(([allPlugins, installedPlugins, exportedPluginNames]) => {
                    const plugins = [];
                    allPlugins
                        .filter((p) => !exportedPluginNames.includes(p.id))
                        .forEach((p) => plugins.push(Object.assign(Object.assign({}, p), { installed: !!installedPlugins.includes(p.id) })));
                    return plugins;
                }), shareReplay(1))
            };
            try {
                const pluginsToAdd = (yield this.bsModalService.show(InstallPluginComponent, {
                    class: 'modal-md',
                    initialState,
                    ignoreBackdropClick: true
                }).content.result);
                const currentRemotes = Object.assign({}, this.remotePlugins$.value);
                pluginsToAdd.forEach((pluginId) => {
                    const { contextPath, name } = this.parsePluginId(pluginId);
                    (currentRemotes[contextPath] = currentRemotes[contextPath] || []).push(name);
                });
                if (this.archives.length === 6) {
                    yield this.ecosystemService.removeOldestArchive(this.app, this.archives);
                }
                this.updateApplicationRemotes(currentRemotes);
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    removePlugins() {
        return __awaiter(this, void 0, void 0, function* () {
            const humanizedAppName = yield this.ecosystemService.getHumanizedAppName(this.app);
            try {
                yield this.modal.confirm(gettext('Remove plugins'), this.translateService.instant(gettext(`You are about to remove plugins from application "{{ humanizedAppName }}". The operation may take several minutes. During this time the application may be unavailable. Do you want to proceed?`), { humanizedAppName }), Status.DANGER, {
                    ok: gettext('Remove'),
                    cancel: gettext('Cancel')
                });
                const remotes = Object.assign({}, this.remotePlugins$.value);
                this.pluginsToDelete.forEach((pluginId) => {
                    const { contextPath, name } = this.parsePluginId(pluginId);
                    remotes[contextPath] = remotes[contextPath].filter((p) => p !== name);
                });
                this.updateApplicationRemotes(remotes);
            }
            catch (ex) {
                // do nothing
            }
        });
    }
    getAllPlugins() {
        return __awaiter(this, void 0, void 0, function* () {
            const plugins = [];
            const packages = yield this.ecosystemService.getPackageApplications();
            const apps = [...packages, this.app];
            apps.forEach((app) => {
                const exports = (app.manifest && app.manifest.exports) || [];
                plugins.push(...this.extendPluginList(exports, app.contextPath));
            });
            return plugins;
        });
    }
    updateApplicationRemotes(remotes) {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            try {
                yield this.pluginsService.updateRemotesInCumulocityJson(this.app, remotes);
                this.remotePlugins$.next(remotes);
            }
            catch (ex) {
                this.alertService.danger(ex);
            }
            finally {
                this.isLoading = false;
            }
        });
    }
    getRemotePluginsList(remotes) {
        const importContextPaths = Object.keys(remotes);
        const plugins = [];
        importContextPaths.forEach((contextPath) => {
            const moduleNames = remotes[contextPath];
            plugins.push(...moduleNames.map((module) => this.createPluginId(contextPath, module)));
        });
        return plugins;
    }
    extendPluginList(plugins, contextPath) {
        const extendedPlugins = [];
        plugins.map((p) => {
            extendedPlugins.push(Object.assign(Object.assign({}, p), { id: this.createPluginId(contextPath, p.name, p.module), contextPath }));
        });
        return extendedPlugins;
    }
    createPluginId(contextPath, pluginName, pluginModuleName) {
        return `${contextPath}/${pluginModuleName || pluginName}`;
    }
    parsePluginId(id) {
        const [contextPath, name] = id.split('/');
        return { contextPath, name };
    }
}
ApplicationPluginsComponent.ɵfac = function ApplicationPluginsComponent_Factory(t) { return new (t || ApplicationPluginsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.PluginsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService)); };
ApplicationPluginsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ApplicationPluginsComponent, selectors: [["c8y-app-plugins"]], decls: 47, vars: 46, consts: [[3, "icon", "label"], [3, "icon", "label", "path"], [3, "label"], [1, "card", "content-fullpage", "d-grid", "grid__col--6-6--md"], [1, "inner-scroll", "bg-gray-white"], [1, "card-header", "large-padding", "separator", "sticky-top"], ["translate", "", 1, "card-title"], [1, "card-block", "large-padding"], [1, "d-flex", "a-i-start", "p-b-16"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-8", "icon-20"], ["translate", ""], [1, "separator-top", "d-block", 3, "emptyListText", "plugins$"], [1, "content-fullpage", "d-flex", "d-col"], [1, "card-header", "large-padding", "separator"], [1, "inner-scroll", "flex-grow"], [1, "separator-top", "d-block", 3, "emptyListText", "plugins$", "selectable", "selectedItems"], [1, "card-footer", "large-padding", "separator"], ["translate", "", 1, "btn", "btn-default", 3, "disabled", "title", "click"], ["translate", "", 1, "btn", "btn-default", 3, "ngClass", "title", "click"]], template: function ApplicationPluginsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵpipe(3, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(5, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelement(7, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelement(9, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelement(11, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵpipe(13, "humanizeAppName");
        ɵngcc0.ɵɵelement(14, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 3);
        ɵngcc0.ɵɵelementStart(17, "div", 4);
        ɵngcc0.ɵɵelementStart(18, "div", 5);
        ɵngcc0.ɵɵelementStart(19, "h4", 6);
        ɵngcc0.ɵɵtext(20, "Included plugins");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 7);
        ɵngcc0.ɵɵelementStart(22, "p", 8);
        ɵngcc0.ɵɵelement(23, "i", 9);
        ɵngcc0.ɵɵelementStart(24, "span", 10);
        ɵngcc0.ɵɵtext(25, " Plugins included with the application. These plugins are required for running the application and cannot be removed. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(26, "c8y-plugin-list", 11);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "div", 12);
        ɵngcc0.ɵɵelementStart(29, "div", 13);
        ɵngcc0.ɵɵelementStart(30, "h4", 6);
        ɵngcc0.ɵɵtext(31, "Installed plugins");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "div", 14);
        ɵngcc0.ɵɵelementStart(33, "div", 7);
        ɵngcc0.ɵɵelementStart(34, "p", 8);
        ɵngcc0.ɵɵelement(35, "i", 9);
        ɵngcc0.ɵɵelementStart(36, "span", 10);
        ɵngcc0.ɵɵtext(37, " Plugins currently installed for this application. Upgrading the application won't change any of these plugins. ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(38, "c8y-plugin-list", 15);
        ɵngcc0.ɵɵlistener("selectedItems", function ApplicationPluginsComponent_Template_c8y_plugin_list_selectedItems_38_listener($event) { return ctx.pluginsToDelete = $event; });
        ɵngcc0.ɵɵpipe(39, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(40, "div", 16);
        ɵngcc0.ɵɵelementStart(41, "button", 17);
        ɵngcc0.ɵɵlistener("click", function ApplicationPluginsComponent_Template_button_click_41_listener() { return ctx.removePlugins(); });
        ɵngcc0.ɵɵpipe(42, "translate");
        ɵngcc0.ɵɵtext(43, " Remove plugins ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(44, "button", 18);
        ɵngcc0.ɵɵlistener("click", function ApplicationPluginsComponent_Template_button_click_44_listener() { return ctx.installPlugins(); });
        ɵngcc0.ɵɵpipe(45, "translate");
        ɵngcc0.ɵɵtext(46, " Install plugins ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 20, ɵngcc0.ɵɵpipeBind1(3, 22, ctx.app)));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(6, 24, "Ecosystem"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(8, 26, "Applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(10, 28, "All applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(12, 30, ɵngcc0.ɵɵpipeBind1(13, 32, ctx.app)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(15, 34, "Plugins"));
        ɵngcc0.ɵɵadvance(12);
        ɵngcc0.ɵɵproperty("emptyListText", ɵngcc0.ɵɵpipeBind1(27, 36, "This application doesn't contain plugins."))("plugins$", ctx.exportedPlugins$);
        ɵngcc0.ɵɵadvance(12);
        ɵngcc0.ɵɵproperty("emptyListText", ɵngcc0.ɵɵpipeBind1(39, 38, "No plugins installed"))("plugins$", ctx.installedPlugins$)("selectable", true);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(42, 40, "Remove plugins"));
        ɵngcc0.ɵɵproperty("disabled", ctx.pluginsToDelete.length === 0 || ctx.isLoading);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(45, 42, "Install plugins"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(44, _c0, ctx.isLoading));
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc4.BreadcrumbComponent, ɵngcc4.BreadcrumbItemComponent, ɵngcc4.C8yTranslateDirective, ɵngcc4.IconDirective, ɵngcc6.PluginListComponent, ɵngcc7.NgClass], pipes: [ɵngcc7.AsyncPipe, ɵngcc4.HumanizeAppNamePipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
ApplicationPluginsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: EcosystemService },
    { type: BsModalService },
    { type: ModalService },
    { type: TranslateService },
    { type: PluginsService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApplicationPluginsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-app-plugins',
                template: "<c8y-title>{{ app | humanizeAppName | async }}</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'All applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"app | humanizeAppName | async\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"'Plugins' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<div class=\"card content-fullpage d-grid grid__col--6-6--md\">\n  <div class=\"inner-scroll bg-gray-white\">\n    <div class=\"card-header large-padding separator sticky-top\">\n      <h4 class=\"card-title\" translate>Included plugins</h4>\n    </div>\n    <div class=\"card-block large-padding\">\n      <p class=\"d-flex a-i-start p-b-16\">\n        <i c8yIcon=\"info-circle\" class=\"text-info m-r-8 icon-20\"></i>\n        <span translate>\n          Plugins included with the application. These plugins are required for running the\n          application and cannot be removed.\n        </span>\n      </p>\n      <c8y-plugin-list\n        [emptyListText]=\"'This application doesn\\'t contain plugins.' | translate\"\n        [plugins$]=\"exportedPlugins$\"\n        class=\"separator-top d-block\"\n      ></c8y-plugin-list>\n    </div>\n  </div>\n\n  <div class=\"content-fullpage d-flex d-col\">\n    <div class=\"card-header large-padding separator\">\n      <h4 class=\"card-title\" translate>Installed plugins</h4>\n    </div>\n    <div class=\"inner-scroll flex-grow\">\n      <div class=\"card-block large-padding\">\n        <p class=\"d-flex a-i-start p-b-16\">\n          <i c8yIcon=\"info-circle\" class=\"text-info m-r-8 icon-20\"></i>\n          <span translate>\n            Plugins currently installed for this application. Upgrading the application won't change\n            any of these plugins.\n          </span>\n        </p>\n        <c8y-plugin-list\n          (selectedItems)=\"pluginsToDelete = $event\"\n          [emptyListText]=\"'No plugins installed' | translate\"\n          [plugins$]=\"installedPlugins$\"\n          [selectable]=\"true\"\n          class=\"separator-top d-block\"\n        ></c8y-plugin-list>\n      </div>\n    </div>\n\n    <div class=\"card-footer large-padding separator\">\n      <button\n        (click)=\"removePlugins()\"\n        [disabled]=\"pluginsToDelete.length === 0 || isLoading\"\n        class=\"btn btn-default\"\n        title=\"{{ 'Remove plugins' | translate }}\"\n        translate\n      >\n        Remove plugins\n      </button>\n      <button\n        (click)=\"installPlugins()\"\n        [ngClass]=\"{ 'btn-pending': isLoading }\"\n        class=\"btn btn-default\"\n        title=\"{{ 'Install plugins' | translate }}\"\n        translate\n      >\n        Install plugins\n      </button>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.EcosystemService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc4.ModalService }, { type: ɵngcc5.TranslateService }, { type: ɵngcc4.PluginsService }, { type: ɵngcc4.AlertService }]; }, null); })();
