// @c8y/ngx-components/esm2015/core/dynamic-forms/validation/validation.module.js
import { NgModule } from '@angular/core';
import { FormlyModule, FORMLY_CONFIG } from '@ngx-formly/core';
import { TranslateService } from '@ngx-translate/core';
import { C8yTranslateModule } from '../../i18n/c8y-translate.module';
import { c8yValidationConfigFactory } from './validation-config.factory';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = c8yValidationConfigFactory;
export class ValidationModule {
}
ValidationModule.ɵfac = function ValidationModule_Factory(t) { return new (t || ValidationModule)(); };
ValidationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ValidationModule });
ValidationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: FORMLY_CONFIG,
            multi: true,
            useFactory: ɵ0,
            deps: [TranslateService]
        }
    ], imports: [[C8yTranslateModule, FormlyModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ValidationModule, [{
        type: NgModule,
        args: [{
                imports: [C8yTranslateModule, FormlyModule],
                providers: [
                    {
                        provide: FORMLY_CONFIG,
                        multi: true,
                        useFactory: ɵ0,
                        deps: [TranslateService]
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ValidationModule, { imports: function () { return [C8yTranslateModule, FormlyModule]; } }); })();
export { ɵ0 };
