// @c8y/ngx-components/esm2015/core/common/virtual-scroll/virtual-scroller-wrapper.component.js
import { Component, Input } from '@angular/core';
import { of, pipe } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@angular/cdk/scrolling';
import * as ɵngcc3 from './virtual-scroll-window.directive';

function VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_div_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c0 = function (a0, a1) { return { $implicit: a0, index: a1 }; };
function VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_div_2_ng_container_1_Template, 1, 0, "ng-container", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    const i_r6 = ctx.index;
    const ctx_r4 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r4.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c0, item_r5, i_r6));
} }
function VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "cdk-virtual-scroll-viewport", 3);
    ɵngcc0.ɵɵtemplate(2, VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_div_2_Template, 2, 5, "div", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const items_r1 = ɵngcc0.ɵɵnextContext().ngIf;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("itemSizePx", ctx_r2.itemHeight);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkVirtualForOf", items_r1)("cdkVirtualForTrackBy", ctx_r2.trackByFn);
} }
function VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_div_2_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_div_2_ng_container_1_Template, 1, 0, "ng-container", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r10 = ctx.$implicit;
    const i_r11 = ctx.index;
    const ctx_r9 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r9.template)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c0, item_r10, i_r11));
} }
function VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "cdk-virtual-scroll-viewport", 6);
    ɵngcc0.ɵɵtemplate(2, VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_div_2_Template, 2, 5, "div", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const items_r1 = ɵngcc0.ɵɵnextContext().ngIf;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("height", ctx_r3.containerHeight + "px");
    ɵngcc0.ɵɵproperty("itemSize", ctx_r3.itemHeight);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("cdkVirtualForOf", items_r1)("cdkVirtualForTrackBy", ctx_r3.trackByFn);
} }
function VirtualScrollerWrapperComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0, 1);
    ɵngcc0.ɵɵtemplate(1, VirtualScrollerWrapperComponent_ng_container_0_ng_container_1_Template, 3, 3, "ng-container", 2);
    ɵngcc0.ɵɵtemplate(2, VirtualScrollerWrapperComponent_ng_container_0_ng_container_2_Template, 3, 5, "ng-container", 2);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r0.strategy);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "window");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "fixed");
} }
export class VirtualScrollerWrapperComponent {
    constructor() {
        this.filterPipe = pipe(tap());
        this.trackByFn = (i, item) => {
            return item.id;
        };
    }
    apply() {
        this.items$ = of(this.items).pipe(src => this.filterPipe(src), map(item => item));
    }
}
VirtualScrollerWrapperComponent.ɵfac = function VirtualScrollerWrapperComponent_Factory(t) { return new (t || VirtualScrollerWrapperComponent)(); };
VirtualScrollerWrapperComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: VirtualScrollerWrapperComponent, selectors: [["c8y-virtual-scroller-wrapper"]], inputs: { filterPipe: "filterPipe", trackByFn: "trackByFn", items: "items", itemHeight: "itemHeight", containerHeight: "containerHeight", template: "template", strategy: "strategy" }, decls: 2, vars: 3, consts: [[3, "ngSwitch", 4, "ngIf"], [3, "ngSwitch"], [4, "ngSwitchCase"], ["windowVirtualScrollStrategy", "", 3, "itemSizePx"], [4, "cdkVirtualFor", "cdkVirtualForOf", "cdkVirtualForTrackBy"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [3, "itemSize"]], template: function VirtualScrollerWrapperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, VirtualScrollerWrapperComponent_ng_container_0_Template, 3, 3, "ng-container", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(1, 1, ctx.items$));
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgSwitch, ɵngcc1.NgSwitchCase, ɵngcc2.CdkVirtualScrollViewport, ɵngcc3.VirtualScrollWindowDirective, ɵngcc2.CdkVirtualForOf, ɵngcc1.NgTemplateOutlet, ɵngcc2.CdkFixedSizeVirtualScroll], pipes: [ɵngcc1.AsyncPipe], encapsulation: 2 });
VirtualScrollerWrapperComponent.propDecorators = {
    items: [{ type: Input }],
    itemHeight: [{ type: Input }],
    containerHeight: [{ type: Input }],
    template: [{ type: Input }],
    filterPipe: [{ type: Input }],
    strategy: [{ type: Input }],
    trackByFn: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(VirtualScrollerWrapperComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-virtual-scroller-wrapper',
                template: "<ng-container *ngIf=\"items$ |async as items\" [ngSwitch]=\"strategy\">\n    <ng-container *ngSwitchCase=\"'window'\">\n      <cdk-virtual-scroll-viewport windowVirtualScrollStrategy [itemSizePx]=\"itemHeight\">\n        <div *cdkVirtualFor=\"let item of items; index as i; trackBy: trackByFn\">\n          <ng-container *ngTemplateOutlet=\"template; context: {$implicit: item, index: i}\"></ng-container>\n        </div>\n      </cdk-virtual-scroll-viewport>\n    </ng-container>\n\n    <ng-container *ngSwitchCase=\"'fixed'\">\n      <cdk-virtual-scroll-viewport [itemSize]=\"itemHeight\" [style.height]=\"containerHeight + 'px'\">\n        <div *cdkVirtualFor=\"let item of items; index as i; trackBy: trackByFn\" >\n          <ng-container *ngTemplateOutlet=\"template; context: {$implicit: item, index: i}\"></ng-container>\n        </div>\n      </cdk-virtual-scroll-viewport>\n    </ng-container>\n</ng-container>\n"
            }]
    }], function () { return []; }, { filterPipe: [{
            type: Input
        }], trackByFn: [{
            type: Input
        }], items: [{
            type: Input
        }], itemHeight: [{
            type: Input
        }], containerHeight: [{
            type: Input
        }], template: [{
            type: Input
        }], strategy: [{
            type: Input
        }] }); })();
