// @c8y/ngx-components/esm2015/sms-gateway/sms-gateway.guard.js
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { Permissions } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/client';
export class SmsGatewayGuard {
    constructor(permissions, applicationService) {
        this.permissions = permissions;
        this.applicationService = applicationService;
    }
    canActivate(route, state) {
        if (this.showSMSProvider === undefined) {
            const hasRole = Promise.resolve(this.permissions.hasRole('ROLE_SMS_READ'));
            this.showSMSProvider = hasRole.then((hasPermissions) => hasPermissions &&
                this.applicationService
                    .isAvailable(SmsGatewayGuard.applicationName)
                    .then(result => result && result.data));
        }
        return this.showSMSProvider;
    }
}
SmsGatewayGuard.ɵfac = function SmsGatewayGuard_Factory(t) { return new (t || SmsGatewayGuard)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions), ɵngcc0.ɵɵinject(ɵngcc2.ApplicationService)); };
SmsGatewayGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SmsGatewayGuard, factory: SmsGatewayGuard.ɵfac });
SmsGatewayGuard.applicationName = 'sms-gateway';
SmsGatewayGuard.ctorParameters = () => [
    { type: Permissions },
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SmsGatewayGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: ɵngcc2.ApplicationService }]; }, null); })();
