// @c8y/ngx-components/esm2015/connectivity/abstract-connectivity-node.factory.js
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, toObservable } from '@c8y/ngx-components';
import { map } from 'rxjs/operators';
import { AbstractConnectivityTabFactory } from './abstract-connectivity-tab.factory';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './abstract-connectivity-tab.factory';
export class AbstractConnectivityNodeFactory {
    constructor(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    get() {
        return toObservable(this.tabFactory.tab()).pipe(map(tab => {
            return !!tab ? [this.connectivityNode] : [];
        }));
    }
}
AbstractConnectivityNodeFactory.ɵfac = function AbstractConnectivityNodeFactory_Factory(t) { return new (t || AbstractConnectivityNodeFactory)(ɵngcc0.ɵɵinject(ɵngcc1.AbstractConnectivityTabFactory)); };
AbstractConnectivityNodeFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: AbstractConnectivityNodeFactory, factory: AbstractConnectivityNodeFactory.ɵfac });
AbstractConnectivityNodeFactory.ctorParameters = () => [
    { type: AbstractConnectivityTabFactory }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AbstractConnectivityNodeFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.AbstractConnectivityTabFactory }]; }, null); })();
