// @c8y/ngx-components/esm2015/core/header/header.module.js
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { ActionModule } from '../action/action.module';
import { BreadcrumbModule } from '../breadcrumb/breadcrumb.module';
import { CommonModule } from '../common/common.module';
import { NavigatorModule } from '../navigator/navigator.module';
import { ProductExperienceModule } from '../product-experience/product-experience.module';
import { SearchModule } from '../search/search.module';
import { UserModule } from '../user/user.module';
import { AppIconComponent } from './application/app-icon.component';
import { AppSwitcherComponent } from './application/app-switcher.component';
import { HeaderBarComponent } from './header-bar/header-bar.component';
import { RightDrawerComponent } from './right-drawer/right-drawer.component';
import { TitleOutletComponent } from './title/title-outlet.component';
import { TitleComponent } from './title/title.component';
/**
 * The angular module definition for the header of the application.
 * @exports HeaderBarComponent
 * @exports TitleComponent
 * @exports TitleOutletComponent
 * @exports AppIconComponent
 * @exports AppSwitcherComponent
 */
import * as ɵngcc0 from '@angular/core';
export class HeaderModule {
}
HeaderModule.ɵfac = function HeaderModule_Factory(t) { return new (t || HeaderModule)(); };
HeaderModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: HeaderModule });
HeaderModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            BsDropdownModule,
            UserModule,
            SearchModule,
            CommonModule,
            ActionModule,
            BreadcrumbModule,
            NavigatorModule,
            ProductExperienceModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HeaderModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TitleComponent,
                    HeaderBarComponent,
                    TitleOutletComponent,
                    AppIconComponent,
                    AppSwitcherComponent,
                    RightDrawerComponent
                ],
                imports: [
                    BsDropdownModule,
                    UserModule,
                    SearchModule,
                    CommonModule,
                    ActionModule,
                    BreadcrumbModule,
                    NavigatorModule,
                    ProductExperienceModule
                ],
                exports: [
                    HeaderBarComponent,
                    TitleComponent,
                    TitleOutletComponent,
                    AppIconComponent,
                    AppSwitcherComponent,
                    RightDrawerComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(HeaderModule, { declarations: function () { return [TitleComponent,
        HeaderBarComponent,
        TitleOutletComponent,
        AppIconComponent,
        AppSwitcherComponent,
        RightDrawerComponent]; }, imports: function () { return [BsDropdownModule,
        UserModule,
        SearchModule,
        CommonModule,
        ActionModule,
        BreadcrumbModule,
        NavigatorModule,
        ProductExperienceModule]; }, exports: function () { return [HeaderBarComponent,
        TitleComponent,
        TitleOutletComponent,
        AppIconComponent,
        AppSwitcherComponent,
        RightDrawerComponent]; } }); })();
