// @c8y/ngx-components/esm2015/core/product-experience/product-experience.directive.js
import { Directive, HostListener, Input } from '@angular/core';
import { GainsightService } from './gainsight.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './gainsight.service';
export class ProductExperienceDirective {
    constructor(gainsightService) {
        this.gainsightService = gainsightService;
    }
    onClick() {
        if (this.actionName) {
            this.gainsightService.triggerEvent(this.actionName, this.actionData);
        }
    }
}
ProductExperienceDirective.ɵfac = function ProductExperienceDirective_Factory(t) { return new (t || ProductExperienceDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.GainsightService)); };
ProductExperienceDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: ProductExperienceDirective, selectors: [["", "c8yProductExperience", ""]], hostBindings: function ProductExperienceDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ProductExperienceDirective_click_HostBindingHandler() { return ctx.onClick(); });
    } }, inputs: { actionName: "actionName", actionData: "actionData" } });
ProductExperienceDirective.ctorParameters = () => [
    { type: GainsightService }
];
ProductExperienceDirective.propDecorators = {
    actionName: [{ type: Input }],
    actionData: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProductExperienceDirective, [{
        type: Directive,
        args: [{
                selector: '[c8yProductExperience]'
            }]
    }], function () { return [{ type: ɵngcc1.GainsightService }]; }, { onClick: [{
            type: HostListener,
            args: ['click']
        }], actionName: [{
            type: Input
        }], actionData: [{
            type: Input
        }] }); })();
