// @c8y/ngx-components/esm2015/core/product-experience/product-experience.module.js
import { __awaiter } from "tslib";
import { NgModule } from '@angular/core';
import { combineLatest } from 'rxjs';
import { filter } from 'rxjs/operators';
import { CookieBannerService } from '../bootstrap/cookie-banner/cookie-banner.service';
import { AppStateService } from '../common/ui-state.service';
import { GainsightService } from './gainsight.service';
import { ProductExperienceDirective } from './product-experience.directive';
/**
 * This module enables an tenant to activate the product experience
 * software [Gainsight](https://www.gainsight.com/product-experience/) to help
 * and track user actions. Gainsight is only activated, if the tenant custom
 * property `gainsightEnabled` is set to true.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/ui-state.service';
import * as ɵngcc2 from './gainsight.service';
import * as ɵngcc3 from '../bootstrap/cookie-banner/cookie-banner.service';
export class ProductExperienceModule {
    constructor(appState, gainsightService, cookieBannerService) {
        this.appState = appState;
        this.gainsightService = gainsightService;
        this.cookieBannerService = cookieBannerService;
        combineLatest([
            this.appState.currentTenant.pipe(filter(Boolean)),
            this.cookieBannerService.isCookieBannerShowed$
        ]).subscribe(([currentTenant, isCookieBannerShowed]) => __awaiter(this, void 0, void 0, function* () {
            if (isCookieBannerShowed) {
                return;
            }
            const { customProperties, name } = currentTenant;
            if (this.gainsightService.shouldLoadGainsightTag(customProperties) &&
                !(yield this.gainsightService.isGainsightDisabledInUserPreferences())) {
                this.gainsightService.loadTag(currentTenant);
            }
        }));
    }
}
ProductExperienceModule.ɵfac = function ProductExperienceModule_Factory(t) { return new (t || ProductExperienceModule)(ɵngcc0.ɵɵinject(ɵngcc1.AppStateService), ɵngcc0.ɵɵinject(ɵngcc2.GainsightService), ɵngcc0.ɵɵinject(ɵngcc3.CookieBannerService)); };
ProductExperienceModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ProductExperienceModule });
ProductExperienceModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [GainsightService] });
ProductExperienceModule.ctorParameters = () => [
    { type: AppStateService },
    { type: GainsightService },
    { type: CookieBannerService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProductExperienceModule, [{
        type: NgModule,
        args: [{
                declarations: [ProductExperienceDirective],
                exports: [ProductExperienceDirective],
                providers: [GainsightService]
            }]
    }], function () { return [{ type: ɵngcc1.AppStateService }, { type: ɵngcc2.GainsightService }, { type: ɵngcc3.CookieBannerService }]; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ProductExperienceModule, { declarations: function () { return [ProductExperienceDirective]; }, exports: function () { return [ProductExperienceDirective]; } }); })();
