// @c8y/ngx-components/esm2015/repository/firmware/add-firmware-patch-modal.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild, Output, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryService } from '../repository.service';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { isUndefined } from 'lodash-es';
import { BehaviorSubject, of, from, pipe, merge } from 'rxjs';
import { debounceTime, tap, switchMap, distinctUntilChanged, shareReplay, map } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../repository.service';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '@angular/common';

const _c0 = ["dropdown"];
const _c1 = ["firmwarePatchForm"];
const _c2 = ["class", "component"];
function AddFirmwarePatchModalComponent_c8y_li_18_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 25);
    ɵngcc0.ɵɵlistener("click", function AddFirmwarePatchModalComponent_c8y_li_18_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r5); const firmware_r3 = restoredCtx.$implicit; const ctx_r4 = ɵngcc0.ɵɵnextContext(); ctx_r4.model.selected = firmware_r3; return ctx_r4.firmwareSelected$.next(firmware_r3); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 26);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const firmware_r3 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("active", ctx_r1.model.selected === firmware_r3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", firmware_r3.name || "--")("pattern", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r1.firmwareInput$));
} }
function AddFirmwarePatchModalComponent_c8y_li_30_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 25);
    ɵngcc0.ɵɵlistener("click", function AddFirmwarePatchModalComponent_c8y_li_30_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r8); const baseVersion_r6 = restoredCtx.$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.model.dependency = baseVersion_r6; });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 26);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r6 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("active", ctx_r2.model.dependency === baseVersion_r6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", baseVersion_r6.c8y_Firmware.version || "--")("pattern", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r2.patchDependencyInput$));
} }
const _c3 = function (a0) { return { "btn-pending": a0 }; };
export class AddFirmwarePatchModalComponent {
    constructor(modal, repository, alert) {
        this.modal = modal;
        this.repository = repository;
        this.alert = alert;
        this.saved = new EventEmitter();
        this.textForFirmwareUrlPopover = gettext(`Path for binaries can vary depending on device agent implementation, e.g.:
    /firmware/binaries/firmware1.bin
    https://firmware/binary/123
    ftp://firmware/binary/123.tar.gz
  `);
        this.model = {
            selected: undefined,
            dependency: null,
            patchVersion: undefined,
            binary: {
                file: undefined,
                url: undefined
            }
        };
        this.firmwareInput$ = new BehaviorSubject('');
        this.firmwares$ = this.firmwareInput$.pipe(debounceTime(300), distinctUntilChanged(), switchMap(searchStr => from(this.repository.listRepositoryEntries(RepositoryType.FIRMWARE, {
            partialName: searchStr,
            skipLegacy: true
        }))), shareReplay(1));
        this.firmwareSelected$ = new BehaviorSubject(null);
        this.patchDependencyInput$ = new BehaviorSubject('');
        this.saving = false;
        this.firmwarePreselected = false;
        this.baseVersions$ = merge(this.firmwareInput$.pipe(tap(() => {
            this.model.dependency = null;
            if (this.form) {
                this.form.form.get('patchDependency').reset();
            }
        }), switchMap(() => of(null))), this.firmwareSelected$).pipe(switchMap(selectedFirmware => selectedFirmware ? this.repository.listBaseVersions(selectedFirmware) : of(null)), shareReplay(1));
        this.baseVersionsFilterPipe = pipe(switchMap((data) => this.patchDependencyInput$.pipe(map(partialVersion => data.filter((mo) => {
            const version = mo.c8y_Firmware.version.toLowerCase();
            return (partialVersion.length === 0 || version.indexOf(partialVersion.toLowerCase()) > -1);
        })))));
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.setInitialState();
        });
    }
    setInitialState() {
        if (this.model.selected) {
            this.firmwarePreselected = true;
            this.firmwareSelected$.next(this.model.selected);
        }
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            this.saving = true;
            this.repository
                .create(this.model, RepositoryType.FIRMWARE)
                .then(savedFirmware => {
                this.successMsg();
                this.saving = false;
                this.saved.next(savedFirmware);
                this.cancel();
            })
                .catch(e => {
                this.saving = false;
                this.saved.error(e);
                this.cancel();
            });
        });
    }
    successMsg() {
        const msg = gettext('Firmware patch added.');
        this.alert.success(msg);
    }
    cancel() {
        this.modal.hide();
        this.saved.complete();
    }
    onFile(dropped) {
        if (!isUndefined(dropped.url)) {
            this.model.binary = {
                url: dropped.url
            };
            return;
        }
        else if (!isUndefined(dropped.droppedFiles)) {
            this.model.binary = {
                file: dropped.droppedFiles[0].file
            };
            return;
        }
        else {
            this.model.binary = {
                file: undefined,
                url: undefined
            };
        }
    }
}
AddFirmwarePatchModalComponent.ɵfac = function AddFirmwarePatchModalComponent_Factory(t) { return new (t || AddFirmwarePatchModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
AddFirmwarePatchModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddFirmwarePatchModalComponent, selectors: [["c8y-add-firmware-patch-modal", 8, "component"]], viewQuery: function AddFirmwarePatchModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropdown = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.form = _t.first);
    } }, outputs: { saved: "saved" }, attrs: _c2, decls: 50, vars: 50, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], ["translate", ""], [1, "p-16", "text-center", "separator-bottom"], ["translate", "", 1, "lead", "m-0"], ["autocomplete", "off", 1, "d-contents", 3, "ngSubmit"], ["firmwarePatchForm", "ngForm"], [1, "modal-inner-scroll"], [1, "modal-body"], [3, "hidden"], ["for", "firmwareName", "translate", ""], ["name", "firmwareName", 3, "ngModel", "placeholder", "allowFreeEntries", "required", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore"], ["name", "notExisting", 3, "text"], ["for", "patchDependency", "translate", "", 1, "m-r-8"], ["name", "patchDependency", 3, "ngModel", "placeholder", "displayProperty", "allowFreeEntries", "disabled", "required", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], ["for", "patchVersion", "translate", ""], ["id", "patchVersion", "autocomplete", "off", "name", "patchVersion", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["translate", "", 1, "legend", "form-block", "m-t-40"], [3, "maxAllowedFiles", "fileUrlPopover", "onFilesPicked"], [1, "modal-footer"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "disabled", "click"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", 3, "title", "ngClass", "disabled"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"]], template: function AddFirmwarePatchModalComponent_Template(rf, ctx) { if (rf & 1) {
        const _r9 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, " Add firmware patch ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "p", 5);
        ɵngcc0.ɵɵtext(7, " Select a firmware version ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "form", 6, 7);
        ɵngcc0.ɵɵlistener("ngSubmit", function AddFirmwarePatchModalComponent_Template_form_ngSubmit_8_listener() { ɵngcc0.ɵɵrestoreView(_r9); const _r0 = ɵngcc0.ɵɵreference(9); return _r0.form.valid && ctx.save(); });
        ɵngcc0.ɵɵelementStart(10, "div", 8);
        ɵngcc0.ɵɵelementStart(11, "div", 9);
        ɵngcc0.ɵɵelementStart(12, "div", 10);
        ɵngcc0.ɵɵelementStart(13, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(14, "label", 11);
        ɵngcc0.ɵɵtext(15, "Firmware");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-typeahead", 12);
        ɵngcc0.ɵɵlistener("onSearch", function AddFirmwarePatchModalComponent_Template_c8y_typeahead_onSearch_16_listener($event) { return ctx.firmwareInput$.next($event); });
        ɵngcc0.ɵɵpipe(17, "translate");
        ɵngcc0.ɵɵtemplate(18, AddFirmwarePatchModalComponent_c8y_li_18_Template, 3, 5, "c8y-li", 13);
        ɵngcc0.ɵɵpipe(19, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "c8y-messages");
        ɵngcc0.ɵɵelement(21, "c8y-message", 14);
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(24, "label", 15);
        ɵngcc0.ɵɵtext(25, "Version");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(26, "c8y-typeahead", 16);
        ɵngcc0.ɵɵlistener("onSearch", function AddFirmwarePatchModalComponent_Template_c8y_typeahead_onSearch_26_listener($event) { return ctx.patchDependencyInput$.next($event); });
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵpipe(28, "async");
        ɵngcc0.ɵɵpipe(29, "async");
        ɵngcc0.ɵɵtemplate(30, AddFirmwarePatchModalComponent_c8y_li_30_Template, 3, 5, "c8y-li", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "c8y-messages");
        ɵngcc0.ɵɵelement(32, "c8y-message", 14);
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(34, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(35, "label", 18);
        ɵngcc0.ɵɵtext(36, "Patch");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(37, "input", 19);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddFirmwarePatchModalComponent_Template_input_ngModelChange_37_listener($event) { return ctx.model.patchVersion = $event; });
        ɵngcc0.ɵɵpipe(38, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(40, "div", 20);
        ɵngcc0.ɵɵtext(41, "Patch file");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "c8y-file-picker", 21);
        ɵngcc0.ɵɵlistener("onFilesPicked", function AddFirmwarePatchModalComponent_Template_c8y_file_picker_onFilesPicked_42_listener($event) { return ctx.onFile($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(43, "div", 22);
        ɵngcc0.ɵɵelementStart(44, "button", 23);
        ɵngcc0.ɵɵlistener("click", function AddFirmwarePatchModalComponent_Template_button_click_44_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(45, "translate");
        ɵngcc0.ɵɵtext(46, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(47, "button", 24);
        ɵngcc0.ɵɵpipe(48, "translate");
        ɵngcc0.ɵɵtext(49, " Add firmware patch ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(9);
        let tmp_13_0;
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-firmware");
        ɵngcc0.ɵɵadvance(10);
        ɵngcc0.ɵɵproperty("hidden", ctx.firmwarePreselected);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(17, 28, "Select or enter"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.selected)("allowFreeEntries", false)("required", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(19, 30, ctx.firmwares$))("c8yForLoadMore", "auto");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(22, 32, "Select one of the existing firmwares."));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(27, 34, "Select or enter"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.dependency)("displayProperty", "c8y_Firmware.version")("allowFreeEntries", false)("disabled", ɵngcc0.ɵɵpipeBind1(28, 36, ctx.baseVersions$) === null || ((tmp_13_0 = ɵngcc0.ɵɵpipeBind1(29, 38, ctx.baseVersions$)) == null ? null : tmp_13_0.data.length) === 0)("required", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.baseVersions$)("c8yForLoadMore", "auto")("c8yForPipe", ctx.baseVersionsFilterPipe);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("text", ɵngcc0.ɵɵpipeBind1(33, 40, "Select one of the existing versions."));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(38, 42, "e.g."), " 1.0.0");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.patchVersion);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("maxAllowedFiles", 1)("fileUrlPopover", ctx.textForFirmwareUrlPopover);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(45, 44, "Cancel"));
        ɵngcc0.ɵɵproperty("disabled", ctx.saving);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(48, 46, "Add firmware patch"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(48, _c3, ctx.saving))("disabled", !_r0.form.valid || _r0.form.pristine || !(ctx.model.binary == null ? null : ctx.model.binary.url) && !(ctx.model.binary == null ? null : ctx.model.binary.file) || ctx.saving);
    } }, directives: [ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc3.FormGroupComponent, ɵngcc3.TypeaheadComponent, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc4.RequiredValidator, ɵngcc3.ForOfDirective, ɵngcc3.MessagesComponent, ɵngcc3.MessageDirective, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc3.FilePickerComponent, ɵngcc5.NgClass, ɵngcc3.ListItemComponent, ɵngcc3.HighlightComponent], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc5.AsyncPipe], encapsulation: 2 });
AddFirmwarePatchModalComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: RepositoryService },
    { type: AlertService }
];
AddFirmwarePatchModalComponent.propDecorators = {
    saved: [{ type: Output }],
    dropdown: [{ type: ViewChild, args: ['dropdown', { static: false },] }],
    form: [{ type: ViewChild, args: ['firmwarePatchForm', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddFirmwarePatchModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-firmware-patch-modal.component',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'c8y-firmware'\"></i>\n    <h4 translate>\n      Add firmware patch\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\">\n    <p class=\"lead m-0\" translate>\n      Select a firmware version\n    </p>\n  </div>\n\n  <form\n    class=\"d-contents\"\n    autocomplete=\"off\"\n    #firmwarePatchForm=\"ngForm\"\n    (ngSubmit)=\"firmwarePatchForm.form.valid && save()\">\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <div [hidden]=\"firmwarePreselected\">\n          <c8y-form-group>\n            <label for=\"firmwareName\" translate>Firmware</label>\n            <c8y-typeahead\n              [ngModel]=\"model.selected\"\n              name=\"firmwareName\"\n              placeholder=\"{{ 'Select or enter' | translate }}\"\n              (onSearch)=\"firmwareInput$.next($event)\"\n              [allowFreeEntries]=\"false\"\n              [required]=\"true\"\n            >\n              <c8y-li\n                *c8yFor=\"let firmware of firmwares$ | async; loadMore: 'auto'\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"model.selected = firmware; firmwareSelected$.next(firmware)\"\n                [active]=\"model.selected === firmware\"\n              >\n                <c8y-highlight\n                  [text]=\"firmware.name || '--'\"\n                  [pattern]=\"firmwareInput$ | async\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <c8y-messages\n              ><c8y-message\n                name=\"notExisting\"\n                [text]=\"'Select one of the existing firmwares.' | translate\"\n              ></c8y-message>\n            </c8y-messages>\n          </c8y-form-group>\n        </div>\n\n        <c8y-form-group>\n          <label for=\"patchDependency\" class=\"m-r-8\" translate>Version</label>\n          <c8y-typeahead\n            [ngModel]=\"model.dependency\"\n            name=\"patchDependency\"\n            placeholder=\"{{ 'Select or enter' | translate }}\"\n            (onSearch)=\"patchDependencyInput$.next($event)\"\n            [displayProperty]=\"'c8y_Firmware.version'\"\n            [allowFreeEntries]=\"false\"\n            [disabled]=\"\n              (baseVersions$ | async) === null || (baseVersions$ | async)?.data.length === 0\n            \"\n            [required]=\"true\"\n          >\n            <c8y-li\n              *c8yFor=\"\n                let baseVersion of baseVersions$;\n                loadMore: 'auto';\n                pipe: baseVersionsFilterPipe\n              \"\n              class=\"p-l-8 p-r-8 c8y-list__item--link\"\n              (click)=\"model.dependency = baseVersion\"\n              [active]=\"model.dependency === baseVersion\"\n            >\n              <c8y-highlight\n                [text]=\"baseVersion.c8y_Firmware.version || '--'\"\n                [pattern]=\"patchDependencyInput$ | async\"\n              ></c8y-highlight>\n            </c8y-li>\n          </c8y-typeahead>\n          <c8y-messages\n            ><c8y-message\n              name=\"notExisting\"\n              [text]=\"'Select one of the existing versions.' | translate\"\n            ></c8y-message>\n          </c8y-messages>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label for=\"patchVersion\" translate>Patch</label>\n          <input\n            id=\"patchVersion\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"patchVersion\"\n            [(ngModel)]=\"model.patchVersion\"\n            placeholder=\"{{ 'e.g.' | translate }} 1.0.0\"\n            required\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <div class=\"legend form-block m-t-40\" translate>Patch file</div>\n          <c8y-file-picker [maxAllowedFiles]=\"1\" (onFilesPicked)=\"onFile($event)\" [fileUrlPopover]=\"textForFirmwareUrlPopover\">\n          </c8y-file-picker>\n        </c8y-form-group>\n      </div>\n    </div>\n    <div class=\"modal-footer\">\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"cancel()\"\n        [disabled]=\"saving\"\n        translate\n      >\n        Cancel\n      </button>\n      <button\n        title=\"{{ 'Add firmware patch' | translate }}\"\n        class=\"btn btn-primary\"\n        type=\"submit\"\n        [ngClass]=\"{ 'btn-pending': saving }\"\n        [disabled]=\"\n          !firmwarePatchForm.form.valid ||\n          firmwarePatchForm.form.pristine ||\n          (!model.binary?.url && !model.binary?.file) ||\n          saving\n        \"\n        translate\n      >\n        Add firmware patch\n      </button>\n    </div>\n  </form>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.RepositoryService }, { type: ɵngcc3.AlertService }]; }, { saved: [{
            type: Output
        }], dropdown: [{
            type: ViewChild,
            args: ['dropdown', { static: false }]
        }], form: [{
            type: ViewChild,
            args: ['firmwarePatchForm', { static: false }]
        }] }); })();
