// @c8y/ngx-components/esm2015/repository/firmware/add-firmware-modal.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, ViewChild } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { RepositoryService } from '../repository.service';
import { RepositoryType } from '../repository.model';
import { AlertService, gettext } from '@c8y/ngx-components';
import { isUndefined, get, assign } from 'lodash-es';
import { BehaviorSubject, from } from 'rxjs';
import { debounceTime, distinctUntilChanged, tap, switchMap } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../repository.service';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from 'ngx-bootstrap/popover';

const _c0 = ["firmwareForm"];
function AddFirmwareModalComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelementStart(1, "p", 27);
    ɵngcc0.ɵɵtext(2, " Select or create new firmware ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function AddFirmwareModalComponent_c8y_li_16_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 28);
    ɵngcc0.ɵɵlistener("click", function AddFirmwareModalComponent_c8y_li_16_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r7); const firmware_r5 = restoredCtx.$implicit; const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.onFirmwareSelect(firmware_r5); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 29);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const firmware_r5 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("active", ctx_r2.model.selected === firmware_r5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", firmware_r5.name || "--")("pattern", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r2.onInput));
} }
function AddFirmwareModalComponent_ng_template_17_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 31);
    ɵngcc0.ɵɵelementStart(1, "span", 3);
    ɵngcc0.ɵɵtext(2, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 32);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5, " Add new`firmware` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 1, "Add new`firmware`"));
} }
function AddFirmwareModalComponent_ng_template_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, AddFirmwareModalComponent_ng_template_17_c8y_li_0_Template, 6, 3, "c8y-li", 30);
    ɵngcc0.ɵɵpipe(1, "async");
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    ɵngcc0.ɵɵproperty("ngIf", ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 1, ctx_r4.onInput)) == null ? null : tmp_0_0.length) > 0);
} }
const _c1 = function (a0) { return { "btn-pending": a0 }; };
export class AddFirmwareModalComponent {
    constructor(modal, repositoryService, alert) {
        this.modal = modal;
        this.repositoryService = repositoryService;
        this.alert = alert;
        this.saved = new EventEmitter();
        this.onInput = new BehaviorSubject('');
        this.model = {
            selected: undefined,
            version: undefined,
            description: undefined,
            deviceType: undefined,
            binary: {
                file: undefined,
                url: undefined
            }
        };
        this.saving = false;
        this.firmwarePreselected = false;
        this.textForFirmwareUrlPopover = gettext(`Path for binaries can vary depending on device agent implementation, e.g.:
    /firmware/binaries/firmware1.bin
    https://firmware/binary/123
    ftp://firmware/binary/123.tar.gz
  `);
    }
    ngOnInit() {
        this.setInitialState();
        this.loadFirmwares();
    }
    setInitialState() {
        if (this.model.selected) {
            this.firmwarePreselected = true;
        }
    }
    loadFirmwares() {
        this.inputSubscription$ = this.onInput
            .pipe(tap(() => {
            if (!this.firmwarePreselected) {
                this.model.description = null;
                if (this.form) {
                    this.form.form.get('description').reset();
                }
            }
        }), debounceTime(300), distinctUntilChanged(), switchMap(searchStr => this.getFirmwareResult(searchStr)))
            .subscribe(result => {
            this.firmwaresResult = result;
        });
    }
    getFirmwareResult(searchStr) {
        return from(this.repositoryService.listRepositoryEntries(RepositoryType.FIRMWARE, {
            partialName: searchStr,
            skipLegacy: true
        }));
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            this.saving = true;
            this.repositoryService
                .create(this.model, RepositoryType.FIRMWARE)
                .then(savedFirmware => {
                this.successMsg();
                this.saving = false;
                this.saved.next(savedFirmware);
                this.cancel();
            })
                .catch(e => {
                this.saving = false;
                this.saved.error(e);
                this.cancel();
            });
        });
    }
    successMsg() {
        const msg = gettext('Firmware added.');
        this.alert.success(msg);
    }
    cancel() {
        this.modal.hide();
        this.saved.complete();
    }
    ngOnDestroy() {
        this.inputSubscription$.unsubscribe();
    }
    onFile(dropped) {
        if (!isUndefined(dropped.url)) {
            this.model.binary = {
                url: dropped.url
            };
            return;
        }
        else if (!isUndefined(dropped.droppedFiles)) {
            this.model.binary = {
                file: dropped.droppedFiles[0].file
            };
            return;
        }
        else {
            this.model.binary = {
                file: undefined,
                url: undefined
            };
        }
    }
    onFirmwareSelect(firmware) {
        assign(this.model, {
            selected: firmware,
            description: firmware.description,
            deviceType: get(firmware, 'c8y_Filter.type')
        });
    }
}
AddFirmwareModalComponent.ɵfac = function AddFirmwareModalComponent_Factory(t) { return new (t || AddFirmwareModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
AddFirmwareModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddFirmwareModalComponent, selectors: [["c8y-add-firmware-software-modal"]], viewQuery: function AddFirmwareModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.form = _t.first);
    } }, outputs: { saved: "saved" }, decls: 49, vars: 47, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], ["translate", ""], ["class", "p-16 text-center separator-bottom", 4, "ngIf"], ["autocomplete", "off", 1, "d-contents", 3, "ngSubmit"], ["firmwareForm", "ngForm"], [1, "modal-inner-scroll"], [1, "modal-body"], [3, "hidden"], ["for", "firmwareName", "translate", ""], ["name", "firmwareName", 3, "ngModel", "placeholder", "required", "ngModelChange", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForNotFound"], ["notFoundTemplate", ""], ["for", "firmwareDescription", "translate", ""], ["id", "firmwareDescription", "autocomplete", "off", "name", "description", 1, "form-control", 3, "ngModel", "placeholder", "disabled", "required", "ngModelChange"], ["for", "firmwareDeviceTypeFilter", 1, "control-label"], ["triggers", "focus", "container", "body", "placement", "right", 1, "btn", "btn-clean", "text-primary", 3, "popover"], ["id", "firmwareDeviceTypeFilter", "name", "firmwareDeviceTypeFilter", 1, "form-control", 3, "ngModel", "placeholder", "disabled", "ngModelChange"], ["for", "firmwareVersion", "translate", ""], ["id", "firmwareVersion", "autocomplete", "off", "name", "version", 1, "form-control", 3, "ngModel", "placeholder", "required", "ngModelChange"], ["translate", "", 1, "legend", "form-block", "m-t-40"], [3, "maxAllowedFiles", "fileUrlPopover", "onFilesPicked"], [1, "modal-footer"], ["type", "button", "translate", "", 1, "btn", "btn-default", 3, "title", "disabled", "click"], ["type", "submit", "translate", "", 1, "btn", "btn-primary", 3, "title", "ngClass", "disabled"], [1, "p-16", "text-center", "separator-bottom"], ["translate", "", 1, "lead", "m-0"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"], ["class", "bg-gray-lighter p-8", 4, "ngIf"], [1, "bg-gray-lighter", "p-8"], ["type", "button", "translate", "", 1, "btn", "btn-primary", "btn-xs", "m-l-16", 3, "title"]], template: function AddFirmwareModalComponent_Template(rf, ctx) { if (rf & 1) {
        const _r9 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, " Add firmware ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, AddFirmwareModalComponent_div_5_Template, 3, 0, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "form", 5, 6);
        ɵngcc0.ɵɵlistener("ngSubmit", function AddFirmwareModalComponent_Template_form_ngSubmit_6_listener() { ɵngcc0.ɵɵrestoreView(_r9); const _r1 = ɵngcc0.ɵɵreference(7); return _r1.form.valid && ctx.save(); });
        ɵngcc0.ɵɵelementStart(8, "div", 7);
        ɵngcc0.ɵɵelementStart(9, "div", 8);
        ɵngcc0.ɵɵelementStart(10, "div", 9);
        ɵngcc0.ɵɵelementStart(11, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(12, "label", 10);
        ɵngcc0.ɵɵtext(13, "Firmware");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-typeahead", 11);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddFirmwareModalComponent_Template_c8y_typeahead_ngModelChange_14_listener($event) { return ctx.model.selected = $event; })("onSearch", function AddFirmwareModalComponent_Template_c8y_typeahead_onSearch_14_listener($event) { return ctx.onInput.next($event); });
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵtemplate(16, AddFirmwareModalComponent_c8y_li_16_Template, 3, 5, "c8y-li", 12);
        ɵngcc0.ɵɵtemplate(17, AddFirmwareModalComponent_ng_template_17_Template, 2, 3, "ng-template", null, 13, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(20, "label", 14);
        ɵngcc0.ɵɵtext(21, "Description");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "input", 15);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddFirmwareModalComponent_Template_input_ngModelChange_22_listener($event) { return ctx.model.description = $event; });
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(25, "label", 16);
        ɵngcc0.ɵɵtext(26);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelementStart(28, "button", 17);
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵelement(30, "i", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "input", 18);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddFirmwareModalComponent_Template_input_ngModelChange_31_listener($event) { return ctx.model.deviceType = $event; });
        ɵngcc0.ɵɵpipe(32, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(33, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(34, "label", 19);
        ɵngcc0.ɵɵtext(35, "Version");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "input", 20);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddFirmwareModalComponent_Template_input_ngModelChange_36_listener($event) { return ctx.model.version = $event; });
        ɵngcc0.ɵɵpipe(37, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(38, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(39, "div", 21);
        ɵngcc0.ɵɵtext(40, "Firmware file");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(41, "c8y-file-picker", 22);
        ɵngcc0.ɵɵlistener("onFilesPicked", function AddFirmwareModalComponent_Template_c8y_file_picker_onFilesPicked_41_listener($event) { return ctx.onFile($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 23);
        ɵngcc0.ɵɵelementStart(43, "button", 24);
        ɵngcc0.ɵɵlistener("click", function AddFirmwareModalComponent_Template_button_click_43_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(44, "translate");
        ɵngcc0.ɵɵtext(45, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(46, "button", 25);
        ɵngcc0.ɵɵpipe(47, "translate");
        ɵngcc0.ɵɵtext(48, " Add firmware ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r1 = ɵngcc0.ɵɵreference(7);
        const _r3 = ɵngcc0.ɵɵreference(18);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "c8y-firmware");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.firmwarePreselected);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("hidden", ctx.firmwarePreselected);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(15, 29, "Select or enter"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.selected)("required", true);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.firmwaresResult)("c8yForLoadMore", "auto")("c8yForNotFound", _r3);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(23, 31, "e.g. Firmware for hardware revision b"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.description)("disabled", ctx.model.selected == null ? null : ctx.model.selected.id)("required", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(27, 33, "Device type filter"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(29, 35, "If the filter is set, the firmware will show up for installation only for devices of that type. If no filter is set, it will be available for all devices."));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(32, 37, "e.g."), " c8y_Linux");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.deviceType)("disabled", ctx.model.selected == null ? null : ctx.model.selected.id);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(37, 39, "e.g."), " 1.0.0");
        ɵngcc0.ɵɵproperty("ngModel", ctx.model.version)("required", true);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("maxAllowedFiles", 1)("fileUrlPopover", ctx.textForFirmwareUrlPopover);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(44, 41, "Cancel"));
        ɵngcc0.ɵɵproperty("disabled", ctx.saving);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(47, 43, "Add firmware"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(45, _c1, ctx.saving))("disabled", !_r1.form.valid || _r1.form.pristine || ctx.saving || !(ctx.model.binary == null ? null : ctx.model.binary.url) && !(ctx.model.binary == null ? null : ctx.model.binary.file));
    } }, directives: [ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc4.NgIf, ɵngcc5.ɵNgNoValidate, ɵngcc5.NgControlStatusGroup, ɵngcc5.NgForm, ɵngcc3.FormGroupComponent, ɵngcc3.TypeaheadComponent, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc5.RequiredValidator, ɵngcc3.ForOfDirective, ɵngcc5.DefaultValueAccessor, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc6.PopoverDirective, ɵngcc3.FilePickerComponent, ɵngcc4.NgClass, ɵngcc3.ListItemComponent, ɵngcc3.HighlightComponent], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc4.AsyncPipe], encapsulation: 2 });
AddFirmwareModalComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: RepositoryService },
    { type: AlertService }
];
AddFirmwareModalComponent.propDecorators = {
    form: [{ type: ViewChild, args: ['firmwareForm', { static: false },] }],
    saved: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddFirmwareModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-firmware-software-modal',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'c8y-firmware'\"></i>\n    <h4 translate>\n      Add firmware\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\" *ngIf=\"!firmwarePreselected\">\n    <p class=\"lead m-0\" translate>\n      Select or create new firmware\n    </p>\n  </div>\n  <form\n    class=\"d-contents\"\n    autocomplete=\"off\"\n    #firmwareForm=\"ngForm\"\n    (ngSubmit)=\"firmwareForm.form.valid && save()\"\n  >\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <div [hidden]=\"firmwarePreselected\">\n          <c8y-form-group>\n            <label for=\"firmwareName\" translate>Firmware</label>\n            <c8y-typeahead\n              [(ngModel)]=\"model.selected\"\n              name=\"firmwareName\"\n              placeholder=\"{{ 'Select or enter' | translate }}\"\n              (onSearch)=\"onInput.next($event)\"\n              [required]=\"true\"\n            >\n              <c8y-li\n                *c8yFor=\"\n                  let firmware of firmwaresResult;\n                  loadMore: 'auto';\n                  notFound: notFoundTemplate\n                \"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"onFirmwareSelect(firmware)\"\n                [active]=\"model.selected === firmware\"\n              >\n                <c8y-highlight\n                  [text]=\"firmware.name || '--'\"\n                  [pattern]=\"onInput | async\"\n                ></c8y-highlight>\n              </c8y-li>\n              <ng-template #notFoundTemplate>\n                <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"(onInput | async)?.length > 0\">\n                  <span translate>No match found.</span>\n                  <button\n                    title=\"{{ 'Add new`firmware`' | translate }}\"\n                    type=\"button\"\n                    class=\"btn btn-primary btn-xs m-l-16\"\n                    translate\n                  >\n                    Add new`firmware`\n                  </button>\n                </c8y-li>\n              </ng-template>\n            </c8y-typeahead>\n          </c8y-form-group>\n\n          <c8y-form-group>\n            <label for=\"firmwareDescription\" translate>Description</label>\n            <input\n              id=\"firmwareDescription\"\n              class=\"form-control\"\n              autocomplete=\"off\"\n              name=\"description\"\n              [(ngModel)]=\"model.description\"\n              placeholder=\"{{ 'e.g. Firmware for hardware revision b' | translate }}\"\n              [disabled]=\"model.selected?.id\"\n              [required]=\"true\"\n            />\n          </c8y-form-group>\n\n          <c8y-form-group>\n            <label class=\"control-label\" for=\"firmwareDeviceTypeFilter\">\n              {{ 'Device type filter' | translate }}\n              <button\n                class=\"btn btn-clean text-primary\"\n                popover=\"{{\n                  'If the filter is set, the firmware will show up for installation only for devices of that type. If no filter is set, it will be available for all devices.'\n                    | translate\n                }}\"\n                triggers=\"focus\"\n                container=\"body\"\n                placement=\"right\"\n              >\n                <i [c8yIcon]=\"'question-circle-o'\"></i>\n              </button>\n            </label>\n            <input\n              id=\"firmwareDeviceTypeFilter\"\n              class=\"form-control\"\n              name=\"firmwareDeviceTypeFilter\"\n              [(ngModel)]=\"model.deviceType\"\n              placeholder=\"{{ 'e.g.' | translate }} c8y_Linux\"\n              [disabled]=\"model.selected?.id\"\n            />\n          </c8y-form-group>\n        </div>\n\n        <c8y-form-group>\n          <label for=\"firmwareVersion\" translate>Version</label>\n          <input\n            id=\"firmwareVersion\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"version\"\n            [(ngModel)]=\"model.version\"\n            placeholder=\"{{ 'e.g.' | translate }} 1.0.0\"\n            [required]=\"true\"\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <div class=\"legend form-block m-t-40\" translate>Firmware file</div>\n          <c8y-file-picker [maxAllowedFiles]=\"1\" (onFilesPicked)=\"onFile($event)\" [fileUrlPopover]=\"textForFirmwareUrlPopover\">\n          </c8y-file-picker>\n        </c8y-form-group>\n      </div>\n    </div>\n    <div class=\"modal-footer\">\n      <button\n        title=\"{{ 'Cancel' | translate }}\"\n        class=\"btn btn-default\"\n        type=\"button\"\n        (click)=\"cancel()\"\n        [disabled]=\"saving\"\n        translate\n      >\n        Cancel\n      </button>\n\n      <button\n        title=\"{{ 'Add firmware' | translate }}\"\n        class=\"btn btn-primary\"\n        type=\"submit\"\n        [ngClass]=\"{ 'btn-pending': saving }\"\n        [disabled]=\"\n          !firmwareForm.form.valid ||\n          firmwareForm.form.pristine ||\n          saving ||\n          (!model.binary?.url && !model.binary?.file)\n        \"\n        translate\n      >\n        Add firmware\n      </button>\n    </div>\n  </form>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.RepositoryService }, { type: ɵngcc3.AlertService }]; }, { saved: [{
            type: Output
        }], form: [{
            type: ViewChild,
            args: ['firmwareForm', { static: false }]
        }] }); })();
