// @c8y/ngx-components/esm2015/repository/firmware/firmware-details.component.js
import { __awaiter, __decorate } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { InventoryService } from '@c8y/client';
import { AlertService, gettext, memoize, ModalService, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { property } from 'lodash-es';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BehaviorSubject, combineLatest, defer, merge, Subject } from 'rxjs';
import { distinctUntilKeyChanged, map, shareReplay, switchMap, take, takeUntil, tap, withLatestFrom } from 'rxjs/operators';
import { RepositoryService } from './../repository.service';
import { AddFirmwareModalComponent } from './add-firmware-modal.component';
import { AddFirmwarePatchModalComponent } from './add-firmware-patch-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from './../repository.service';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@ngx-translate/core';
import * as ɵngcc6 from 'ngx-bootstrap/modal';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from '@angular/forms';
import * as ɵngcc9 from 'ngx-bootstrap/popover';
import * as ɵngcc10 from '../file-download/file-download.component';

function FirmwareDetailsComponent_button_7_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_button_7_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r11); const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.addBaseVersion(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 28);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Add firmware"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Add firmware"), " ");
} }
function FirmwareDetailsComponent_button_10_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 3);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_button_10_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.addPatchVersion(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 28);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Add firmware patch"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 4, "Add firmware patch"), " ");
} }
function FirmwareDetailsComponent_div_91_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 29);
    ɵngcc0.ɵɵelement(2, "h1", 30);
    ɵngcc0.ɵɵelementStart(3, "h3", 31);
    ɵngcc0.ɵɵtext(4, "No versions to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 31);
    ɵngcc0.ɵɵtext(6, "Add a new version by clicking below.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "p");
    ɵngcc0.ɵɵelementStart(8, "button", 20);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_div_91_Template_button_click_8_listener() { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.addBaseVersion(); });
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 2, "Add firmware"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(11, 4, "Add firmware"), " ");
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 48);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Legacy"), " ");
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 49);
    ɵngcc0.ɵɵelement(2, "i", 50);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 55);
    ɵngcc0.ɵɵelementStart(1, "span", 31);
    ɵngcc0.ɵɵtext(2, "No patches");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 49);
    ɵngcc0.ɵɵelementStart(1, "span", 31);
    ɵngcc0.ɵɵtext(2, "1 patch");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { count: a0 }; };
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 49);
    ɵngcc0.ɵɵelementStart(1, "span", 56);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtext(3, " {{ count }} patches ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r17 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const ctx_r28 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(3, _c0, ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r28.getPatchVersionsCount$(baseVersion_r17))));
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelementStart(1, "span", 51);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_3_Template, 3, 0, "ng-template", 52);
    ɵngcc0.ɵɵtemplate(4, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_4_Template, 3, 0, "ng-template", 53);
    ɵngcc0.ɵɵtemplate(5, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_ng_template_5_Template, 4, 5, "ng-template", 54);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r17 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const ctx_r25 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngPlural", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r25.getPatchVersionsCount$(baseVersion_r17)));
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_1_Template, 3, 0, "span", 26);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵtemplate(3, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_span_3_Template, 6, 3, "span", 26);
    ɵngcc0.ɵɵpipe(4, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r17 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r20 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(2, 2, ctx_r20.getPatchVersionsCount$(baseVersion_r17)) === null);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(4, 4, ctx_r20.getPatchVersionsCount$(baseVersion_r17)) !== null);
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_20_Template(rf, ctx) { if (rf & 1) {
    const _r34 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 57);
    ɵngcc0.ɵɵelementStart(1, "button", 58);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_20_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r34); const baseVersion_r17 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r32 = ɵngcc0.ɵɵnextContext(2); return ctx_r32.deleteBaseVersion(baseVersion_r17); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 59);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Delete"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Delete"), " ");
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_22_Template(rf, ctx) { if (rf & 1) {
    const _r37 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 60);
    ɵngcc0.ɵɵelementStart(1, "button", 61);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_22_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r37); const baseVersion_r17 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.deleteBaseVersion(baseVersion_r17); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 59);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 1, "Delete"));
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r42 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 35);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 36);
    ɵngcc0.ɵɵelementStart(4, "div", 37);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 39);
    ɵngcc0.ɵɵelementStart(7, "div", 64);
    ɵngcc0.ɵɵelementStart(8, "span", 41);
    ɵngcc0.ɵɵtext(9, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(10, "c8y-file-download", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 65);
    ɵngcc0.ɵɵelementStart(12, "button", 66);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_c8y_li_2_Template_button_click_12_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r42); const patchVersion_r39 = restoredCtx.$implicit; const ctx_r41 = ɵngcc0.ɵɵnextContext(4); return ctx_r41.deletePatchVersion(patchVersion_r39); });
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelement(14, "i", 59);
    ɵngcc0.ɵɵtext(15);
    ɵngcc0.ɵɵpipe(16, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "div", 67);
    ɵngcc0.ɵɵelementStart(18, "button", 61);
    ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_c8y_li_2_Template_button_click_18_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r42); const patchVersion_r39 = restoredCtx.$implicit; const ctx_r43 = ɵngcc0.ɵɵnextContext(4); return ctx_r43.deletePatchVersion(patchVersion_r39); });
    ɵngcc0.ɵɵpipe(19, "translate");
    ɵngcc0.ɵɵelement(20, "i", 59);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const patchVersion_r39 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", patchVersion_r39.c8y_Firmware.version, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("url", patchVersion_r39.c8y_Firmware.url);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(13, 5, "Delete"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(16, 7, "Delete"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(19, 9, "Delete"));
} }
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(1, "c8y-list-group", 62);
    ɵngcc0.ɵɵtemplate(2, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_c8y_li_2_Template, 21, 11, "c8y-li", 63);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r17 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r23 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(3, 2, ctx_r23.getPatchVersions$(baseVersion_r17)))("c8yForLoadMore", "auto");
} }
const _c1 = function (a0) { return { "c8y-list__item--empty-actions": a0 }; };
function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r46 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 34);
    ɵngcc0.ɵɵlistener("collapsedChange", function FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_Template_c8y_li_collapsedChange_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r46); const baseVersion_r17 = restoredCtx.$implicit; const ctx_r45 = ɵngcc0.ɵɵnextContext(2); return (ctx_r45.expanded[baseVersion_r17.id] = !$event); });
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
    ɵngcc0.ɵɵelement(3, "i", 35);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 36);
    ɵngcc0.ɵɵelementStart(5, "div", 37);
    ɵngcc0.ɵɵelementStart(6, "p", 38);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 39);
    ɵngcc0.ɵɵelementStart(9, "p", 40);
    ɵngcc0.ɵɵelementStart(10, "span", 41);
    ɵngcc0.ɵɵtext(11, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "span", 42);
    ɵngcc0.ɵɵpipe(13, "async");
    ɵngcc0.ɵɵelement(14, "c8y-file-download", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 44);
    ɵngcc0.ɵɵtemplate(16, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_16_Template, 3, 3, "span", 45);
    ɵngcc0.ɵɵpipe(17, "async");
    ɵngcc0.ɵɵtemplate(18, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_span_18_Template, 5, 6, "span", 26);
    ɵngcc0.ɵɵpipe(19, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(20, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_20_Template, 6, 6, "div", 46);
    ɵngcc0.ɵɵpipe(21, "async");
    ɵngcc0.ɵɵtemplate(22, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_div_22_Template, 4, 3, "div", 47);
    ɵngcc0.ɵɵpipe(23, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(24, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_c8y_li_collapse_24_Template, 4, 4, "c8y-li-collapse", 26);
    ɵngcc0.ɵɵpipe(25, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const baseVersion_r17 = ctx.$implicit;
    const ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    let tmp_0_0;
    let tmp_10_0;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(25, _c1, !((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 11, ctx_r16.getPatchVersions$(baseVersion_r17))) == null ? null : tmp_0_0.data.length)))("collapsed", !ctx_r16.expanded[baseVersion_r17.id]);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", baseVersion_r17.c8y_Firmware.version);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", baseVersion_r17.c8y_Firmware.version, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(13, 13, ctx_r16.getBinaryName$(baseVersion_r17.c8y_Firmware.url)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("url", baseVersion_r17.c8y_Firmware.url);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(17, 15, ctx_r16.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(19, 17, ctx_r16.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(21, 19, ctx_r16.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(23, 21, ctx_r16.isLegacy$));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", (tmp_10_0 = ɵngcc0.ɵɵpipeBind1(25, 23, ctx_r16.getPatchVersions$(baseVersion_r17))) == null ? null : tmp_10_0.data.length);
} }
const _c2 = function (a0) { return { "dd-low": a0 }; };
function FirmwareDetailsComponent_c8y_list_group_93_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 32);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵtemplate(2, FirmwareDetailsComponent_c8y_list_group_93_c8y_li_2_Template, 26, 27, "c8y-li", 33);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r9 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(7, _c2, ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r9.baseVersions$)) == null ? null : tmp_0_0.data.length) < 10));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(3, 5, ctx_r9.baseVersions$))("c8yForLoadMore", "auto");
} }
const _c3 = function (a0) { return { "icon-spin": a0 }; };
const _c4 = function (a0) { return { "width.ch": a0 }; };
export class FirmwareDetailsComponent {
    constructor(activatedRoute, inventoryService, repositoryService, alertService, translateService, modalService, bsModalService) {
        this.activatedRoute = activatedRoute;
        this.inventoryService = inventoryService;
        this.repositoryService = repositoryService;
        this.alertService = alertService;
        this.translateService = translateService;
        this.modalService = modalService;
        this.bsModalService = bsModalService;
        this.reload$ = new Subject();
        this.reloading$ = new BehaviorSubject(false);
        this.updateFirmware$ = new Subject();
        this.firmwareUpdated$ = new Subject();
        this.baseVersionsUpdated$ = new Subject();
        this.patchVersionsUpdated$ = new Subject();
        this.firmware$ = merge(this.activatedRoute.params.pipe(map(params => params.id), switchMap(id => defer(() => this.inventoryService.detail(id).then(result => result.data)))), this.reload$.pipe(tap(() => this.reloading$.next(true)), switchMap(() => this.activatedRoute.params), map(params => params.id), switchMap(id => defer(() => this.inventoryService.detail(id).then(result => result.data))), tap(() => this.reloading$.next(false))), this.firmwareUpdated$).pipe(shareReplay(1));
        this.baseVersions$ = merge(this.firmware$.pipe(distinctUntilKeyChanged('id')), this.baseVersionsUpdated$, this.patchVersionsUpdated$, this.reload$).pipe(switchMap(() => this.firmware$), switchMap(firmware => this.repositoryService.listBaseVersions(firmware)), shareReplay(1));
        this.isLegacy$ = this.firmware$.pipe(map(firmware => this.repositoryService.isLegacyEntry(firmware)), shareReplay(1));
        this.canAddPatchVersions$ = combineLatest(this.isLegacy$, this.baseVersions$.pipe(map(({ data }) => data.length > 0))).pipe(map(([isLegacy, hasBaseVersions]) => !isLegacy && hasBaseVersions));
        this.expanded = {};
        this.destroy$ = new Subject();
    }
    ngOnInit() {
        this.updateFirmware$
            .pipe(withLatestFrom(this.firmware$), switchMap(([firmwarePartial, firmware]) => this.inventoryService.update(Object.assign({ id: firmware.id }, firmwarePartial))), map(({ data }) => data), tap(firmware => this.firmwareUpdated$.next(firmware)), tap(() => this.alertService.success(gettext('Saved.'))), takeUntil(this.destroy$))
            .subscribe();
    }
    getPatchVersionsCount$(baseVersion) {
        return merge(this.firmware$.pipe(distinctUntilKeyChanged('id')), this.baseVersionsUpdated$, this.patchVersionsUpdated$, this.reload$).pipe(switchMap(() => this.firmware$), switchMap(firmware => this.repositoryService.getPatchVersionsCount$(firmware, baseVersion)), shareReplay(1));
    }
    getBinaryName$(binaryUrl) {
        return this.repositoryService.getBinaryName$(binaryUrl);
    }
    getPatchVersions$(baseVersion) {
        return merge(this.firmware$.pipe(distinctUntilKeyChanged('id')), this.patchVersionsUpdated$, this.reload$).pipe(switchMap(() => this.firmware$), switchMap(firmware => this.repositoryService.listPatchVersions(firmware, baseVersion)), shareReplay(1));
    }
    addBaseVersion() {
        this.firmware$
            .pipe(take(1), switchMap(firmware => {
            const initialState = {
                model: {
                    selected: firmware,
                    description: firmware.description
                }
            };
            const config = {
                class: 'modal-sm',
                ignoreBackdropClick: true,
                initialState
            };
            const modalRef = this.bsModalService.show(AddFirmwareModalComponent, config);
            return modalRef.content.saved;
        }))
            .subscribe(() => this.baseVersionsUpdated$.next());
    }
    addPatchVersion() {
        this.firmware$
            .pipe(take(1), switchMap(firmware => {
            const initialState = {
                model: {
                    selected: firmware
                }
            };
            const config = {
                class: 'modal-sm',
                ignoreBackdropClick: true,
                initialState
            };
            const modalRef = this.bsModalService.show(AddFirmwarePatchModalComponent, config);
            return modalRef.content.saved;
        }))
            .subscribe(() => this.patchVersionsUpdated$.next());
    }
    deleteBaseVersion(baseVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const title = gettext('Delete firmware');
                const body = `
        ${this.translateService.instant(gettext('You are about to delete firmware {{ version }} with all its patches.'), { version: baseVersion.c8y_Firmware.version })}
        ${this.translateService.instant(gettext('This operation is irreversible.'))}
        ${this.translateService.instant(gettext('Do you want to proceed?'))}
      `;
                const labels = {
                    ok: gettext('Delete')
                };
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.repositoryService.delete(baseVersion);
                this.alertService.success(gettext('Firmware deleted.'));
                this.baseVersionsUpdated$.next();
            }
            catch (ex) {
                // only if not cancel from modal
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    deletePatchVersion(patchVersion) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const title = gettext('Delete firmware patch');
                const body = `
        ${this.translateService.instant(gettext('You are about to delete firmware patch {{ version }}.'), { version: patchVersion.c8y_Firmware.version })}
        ${this.translateService.instant(gettext('This operation is irreversible.'))}
        ${this.translateService.instant(gettext('Do you want to proceed?'))}
      `;
                const labels = {
                    ok: gettext('Delete')
                };
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.repositoryService.delete(patchVersion);
                this.alertService.success(gettext('Firmware patch deleted.'));
                this.patchVersionsUpdated$.next();
            }
            catch (ex) {
                // only if not cancel from modal
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    ngOnDestroy() {
        this.destroy$.next(true);
        this.destroy$.unsubscribe();
    }
}
FirmwareDetailsComponent.ɵfac = function FirmwareDetailsComponent_Factory(t) { return new (t || FirmwareDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.BsModalService)); };
FirmwareDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FirmwareDetailsComponent, selectors: [["c8y-firmware-details"]], decls: 95, vars: 97, consts: [["path", "#/firmware", "icon", "c8y-firmware", 3, "label"], [3, "placement"], ["class", "btn btn-link", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], [1, "row"], [1, "col-lg-12", "col-lg-max"], [1, "card", "m-b-4"], [1, "card-header", "separator"], ["translate", "", 1, "card-title"], [1, "card-block"], [1, "col-md-4"], [1, "control-label"], [1, "input-group", "input-group-editable"], ["type", "text", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngStyle"], ["nameInput", "", "nameModel", "ngModel"], [1, "input-group-btn"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["type", "text", 1, "form-control", 3, "ngModel", "placeholder", "ngStyle"], ["descriptionInput", "", "descriptionModel", "ngModel"], [1, "btn", "btn-primary", 3, "title", "click"], ["container", "body", 1, "btn-clean", 3, "popover", "outsideClick"], ["c8yIcon", "question-circle-o", 1, "text-info"], ["deviceTypeInput", "", "deviceTypeModel", "ngModel"], [1, "card"], [1, "card-block", "p-t-0", "p-b-24"], [4, "ngIf"], [3, "ngClass", 4, "ngIf"], ["c8yIcon", "plus-circle"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "c8y-firmware", 1, "c8y-icon-duocolor"], ["translate", ""], [3, "ngClass"], [3, "ngClass", "collapsed", "collapsedChange", 4, "c8yFor", "c8yForOf", "c8yForLoadMore"], [3, "ngClass", "collapsed", "collapsedChange"], ["c8yIcon", "c8y-firmware"], [1, "content-flex-50"], [1, "col-4"], [1, "text-truncate-wrap", 3, "title"], [1, "col-5"], [1, "text-truncate-wrap"], ["translate", "", 1, "text-label-small", "m-r-8"], [3, "title"], [3, "url"], [1, "col-2", "flex-row", "a-i-start"], ["class", "label label-warning flex-item-right-sm", 4, "ngIf"], ["class", "fit-h-20 visible-xs", 4, "ngIf"], ["class", "flex-item-right fit-h-20 p-r-8 hidden-xs", 4, "ngIf"], [1, "label", "label-warning", "flex-item-right-sm"], [1, "label", "label-info"], ["c8yIcon", "circle-o-notch", 1, "icon-spin"], [3, "ngPlural"], ["ngPluralCase", "=0"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], [1, "label", "label-default", "flex-item-right-sm"], ["translate", "", 3, "translateParams"], [1, "fit-h-20", "visible-xs"], [1, "btn", "btn-danger", "btn-xs", "m-t-8", 3, "title", "click"], ["c8yIcon", "minus-circle"], [1, "flex-item-right", "fit-h-20", "p-r-8", "hidden-xs"], [1, "btn", "btn-dot", "text-danger", "showOnHover", 3, "title", "click"], [1, "separator-top"], [4, "c8yFor", "c8yForOf", "c8yForLoadMore"], [1, "text-truncate"], [1, "visible-xs", "m-t-8"], [1, "btn", "btn-danger", "btn-xs", 3, "title", "click"], [1, "flex-item-right", "p-r-8", "hidden-xs", "fit-h-20"]], template: function FirmwareDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        const _r47 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(4, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(7, FirmwareDetailsComponent_button_7_Template, 5, 6, "button", 2);
        ɵngcc0.ɵɵpipe(8, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵtemplate(10, FirmwareDetailsComponent_button_10_Template, 5, 6, "button", 2);
        ɵngcc0.ɵɵpipe(11, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "c8y-action-bar-item", 1);
        ɵngcc0.ɵɵelementStart(13, "button", 3);
        ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_Template_button_click_13_listener() { return ctx.reload$.next(); });
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelement(15, "i", 4);
        ɵngcc0.ɵɵpipe(16, "async");
        ɵngcc0.ɵɵtext(17);
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "div", 5);
        ɵngcc0.ɵɵelementStart(20, "div", 6);
        ɵngcc0.ɵɵelementStart(21, "div", 7);
        ɵngcc0.ɵɵelementStart(22, "div", 8);
        ɵngcc0.ɵɵelementStart(23, "h4", 9);
        ɵngcc0.ɵɵtext(24, " Name, description and device type filter ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "div", 10);
        ɵngcc0.ɵɵelementStart(26, "div", 5);
        ɵngcc0.ɵɵelementStart(27, "div", 11);
        ɵngcc0.ɵɵelementStart(28, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(29, "label", 12);
        ɵngcc0.ɵɵtext(30);
        ɵngcc0.ɵɵpipe(31, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "div", 13);
        ɵngcc0.ɵɵelement(33, "input", 14, 15);
        ɵngcc0.ɵɵpipe(36, "async");
        ɵngcc0.ɵɵpipe(37, "translate");
        ɵngcc0.ɵɵpipe(38, "async");
        ɵngcc0.ɵɵelement(39, "span");
        ɵngcc0.ɵɵelementStart(40, "div", 16);
        ɵngcc0.ɵɵelementStart(41, "button", 17);
        ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_Template_button_click_41_listener() { ɵngcc0.ɵɵrestoreView(_r47); const _r2 = ɵngcc0.ɵɵreference(34); const _r3 = ɵngcc0.ɵɵreference(35); ctx.updateFirmware$.next({ name: _r2.value }); return _r3.reset(); });
        ɵngcc0.ɵɵpipe(42, "translate");
        ɵngcc0.ɵɵtext(43);
        ɵngcc0.ɵɵpipe(44, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "div", 11);
        ɵngcc0.ɵɵelementStart(46, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(47, "label", 12);
        ɵngcc0.ɵɵtext(48);
        ɵngcc0.ɵɵpipe(49, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(50, "div", 13);
        ɵngcc0.ɵɵelement(51, "input", 18, 19);
        ɵngcc0.ɵɵpipe(54, "async");
        ɵngcc0.ɵɵpipe(55, "translate");
        ɵngcc0.ɵɵpipe(56, "async");
        ɵngcc0.ɵɵelement(57, "span");
        ɵngcc0.ɵɵelementStart(58, "div", 16);
        ɵngcc0.ɵɵelementStart(59, "button", 20);
        ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_Template_button_click_59_listener() { ɵngcc0.ɵɵrestoreView(_r47); const _r4 = ɵngcc0.ɵɵreference(52); const _r5 = ɵngcc0.ɵɵreference(53); ctx.updateFirmware$.next({ description: _r4.value }); return _r5.reset(); });
        ɵngcc0.ɵɵpipe(60, "translate");
        ɵngcc0.ɵɵtext(61);
        ɵngcc0.ɵɵpipe(62, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(63, "div", 11);
        ɵngcc0.ɵɵelementStart(64, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(65, "label", 12);
        ɵngcc0.ɵɵtext(66);
        ɵngcc0.ɵɵpipe(67, "translate");
        ɵngcc0.ɵɵelementStart(68, "button", 21);
        ɵngcc0.ɵɵpipe(69, "translate");
        ɵngcc0.ɵɵelement(70, "i", 22);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(71, "div", 13);
        ɵngcc0.ɵɵelement(72, "input", 18, 23);
        ɵngcc0.ɵɵpipe(75, "async");
        ɵngcc0.ɵɵpipe(76, "translate");
        ɵngcc0.ɵɵpipe(77, "async");
        ɵngcc0.ɵɵelement(78, "span");
        ɵngcc0.ɵɵelementStart(79, "div", 16);
        ɵngcc0.ɵɵelementStart(80, "button", 20);
        ɵngcc0.ɵɵlistener("click", function FirmwareDetailsComponent_Template_button_click_80_listener() { ɵngcc0.ɵɵrestoreView(_r47); const _r6 = ɵngcc0.ɵɵreference(73); const _r7 = ɵngcc0.ɵɵreference(74); ctx.updateFirmware$.next({ c8y_Filter: { type: _r6.value } }); return _r7.reset(); });
        ɵngcc0.ɵɵpipe(81, "translate");
        ɵngcc0.ɵɵtext(82);
        ɵngcc0.ɵɵpipe(83, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(84, "div", 5);
        ɵngcc0.ɵɵelementStart(85, "div", 6);
        ɵngcc0.ɵɵelementStart(86, "div", 24);
        ɵngcc0.ɵɵelementStart(87, "div", 8);
        ɵngcc0.ɵɵelementStart(88, "h4", 9);
        ɵngcc0.ɵɵtext(89, " Versions and patches ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(90, "div", 25);
        ɵngcc0.ɵɵtemplate(91, FirmwareDetailsComponent_div_91_Template, 12, 6, "div", 26);
        ɵngcc0.ɵɵpipe(92, "async");
        ɵngcc0.ɵɵtemplate(93, FirmwareDetailsComponent_c8y_list_group_93_Template, 4, 9, "c8y-list-group", 27);
        ɵngcc0.ɵɵpipe(94, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(34);
        let tmp_0_0;
        let tmp_12_0;
        let tmp_13_0;
        let tmp_19_0;
        let tmp_20_0;
        let tmp_27_0;
        let tmp_28_0;
        let tmp_31_0;
        let tmp_32_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", (tmp_0_0 = ɵngcc0.ɵɵpipeBind1(2, 33, ctx.firmware$)) == null ? null : tmp_0_0.name, "\n");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(5, 35, "Firmware repository"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ɵngcc0.ɵɵpipeBind1(8, 37, ctx.isLegacy$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(11, 39, ctx.canAddPatchVersions$));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(14, 41, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(89, _c3, ɵngcc0.ɵɵpipeBind1(16, 43, ctx.reloading$)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(18, 45, "Reload"), " ");
        ɵngcc0.ɵɵadvance(13);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(31, 47, "Name"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(37, 51, "e.g. My firmware"));
        ɵngcc0.ɵɵproperty("ngModel", (tmp_12_0 = ɵngcc0.ɵɵpipeBind1(36, 49, ctx.firmware$)) == null ? null : tmp_12_0.name)("ngStyle", ɵngcc0.ɵɵpureFunction1(91, _c4, ((tmp_13_0 = ɵngcc0.ɵɵpipeBind1(38, 53, ctx.firmware$)) == null ? null : tmp_13_0.name == null ? null : tmp_13_0.name.length) || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(42, 55, "Save"));
        ɵngcc0.ɵɵproperty("disabled", _r2.value.length == 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(44, 57, "Save"), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(49, 59, "Description"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(55, 63, "e.g. Firmware for hardware revision b"));
        ɵngcc0.ɵɵproperty("ngModel", (tmp_19_0 = ɵngcc0.ɵɵpipeBind1(54, 61, ctx.firmware$)) == null ? null : tmp_19_0.description)("ngStyle", ɵngcc0.ɵɵpureFunction1(93, _c4, ((tmp_20_0 = ɵngcc0.ɵɵpipeBind1(56, 65, ctx.firmware$)) == null ? null : tmp_20_0.description == null ? null : tmp_20_0.description.length) || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(60, 67, "Save"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(62, 69, "Save"), " ");
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(67, 71, "Device type filter"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(69, 73, "If the filter is set, the firmware will show up for installation only for devices of that type. If no filter is set, it will be available for all devices."));
        ɵngcc0.ɵɵproperty("outsideClick", true);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(76, 77, "e.g."), " c8y_Linux");
        ɵngcc0.ɵɵproperty("ngModel", (tmp_27_0 = ɵngcc0.ɵɵpipeBind1(75, 75, ctx.firmware$)) == null ? null : tmp_27_0.c8y_Filter == null ? null : tmp_27_0.c8y_Filter.type)("ngStyle", ɵngcc0.ɵɵpureFunction1(95, _c4, ((tmp_28_0 = ɵngcc0.ɵɵpipeBind1(77, 79, ctx.firmware$)) == null ? null : tmp_28_0.type == null ? null : tmp_28_0.type.length) || 31));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(81, 81, "Save"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(83, 83, "Save"), " ");
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_31_0 = ɵngcc0.ɵɵpipeBind1(92, 85, ctx.baseVersions$)) == null ? null : tmp_31_0.data.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_32_0 = ɵngcc0.ɵɵpipeBind1(94, 87, ctx.baseVersions$)) == null ? null : tmp_32_0.data.length) > 0);
    } }, directives: [ɵngcc4.TitleComponent, ɵngcc4.BreadcrumbComponent, ɵngcc4.BreadcrumbItemComponent, ɵngcc4.ActionBarItemComponent, ɵngcc7.NgIf, ɵngcc4.IconDirective, ɵngcc7.NgClass, ɵngcc4.C8yTranslateDirective, ɵngcc4.FormGroupComponent, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc8.DefaultValueAccessor, ɵngcc8.RequiredValidator, ɵngcc8.NgControlStatus, ɵngcc8.NgModel, ɵngcc7.NgStyle, ɵngcc9.PopoverDirective, ɵngcc4.ListGroupComponent, ɵngcc4.ForOfDirective, ɵngcc4.ListItemComponent, ɵngcc4.ListItemIconComponent, ɵngcc4.ListItemBodyComponent, ɵngcc10.FileDownloadComponent, ɵngcc7.NgPlural, ɵngcc7.NgPluralCase, ɵngcc4.ListItemCollapseComponent], pipes: [ɵngcc7.AsyncPipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
FirmwareDetailsComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: InventoryService },
    { type: RepositoryService },
    { type: AlertService },
    { type: TranslateService },
    { type: ModalService },
    { type: BsModalService }
];
__decorate([
    memoize(property('id'))
], FirmwareDetailsComponent.prototype, "getPatchVersionsCount$", null);
__decorate([
    memoize()
], FirmwareDetailsComponent.prototype, "getBinaryName$", null);
__decorate([
    memoize(property('id'))
], FirmwareDetailsComponent.prototype, "getPatchVersions$", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FirmwareDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-firmware-details',
                template: "<c8y-title>\n  {{ (firmware$ | async)?.name }}\n</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item\n    path=\"#/firmware\"\n    label=\"{{ 'Firmware repository' | translate }}\"\n    icon=\"c8y-firmware\"\n  >\n  </c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    *ngIf=\"!(isLegacy$ | async)\"\n    class=\"btn btn-link\"\n    title=\"{{ 'Add firmware' | translate }}\"\n    (click)=\"addBaseVersion()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add firmware' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    *ngIf=\"canAddPatchVersions$ | async\"\n    class=\"btn btn-link\"\n    title=\"{{ 'Add firmware patch' | translate }}\"\n    (click)=\"addPatchVersion()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add firmware patch' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"reload$.next()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': reloading$ | async }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"row\">\n  <div class=\"col-lg-12 col-lg-max\">\n    <div class=\"card m-b-4\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>\n          Name, description and device type filter\n        </h4>\n      </div>\n      <div class=\"card-block\">\n        <div class=\"row\">\n          <div class=\"col-md-4\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Name' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #nameInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(firmware$ | async)?.name\"\n                  #nameModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g. My firmware' | translate }}\"\n                  [ngStyle]=\"{ 'width.ch': (firmware$ | async)?.name?.length || 31 }\"\n                  required\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\"updateFirmware$.next({ name: nameInput.value }); nameModel.reset()\"\n                    [disabled]=\"nameInput.value.length == 0\"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n          <div class=\"col-md-4\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Description' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #descriptionInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(firmware$ | async)?.description\"\n                  #descriptionModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g. Firmware for hardware revision b' | translate }}\"\n                  [ngStyle]=\"{ 'width.ch': (firmware$ | async)?.description?.length || 31 }\"\n\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\" updateFirmware$.next({ description: descriptionInput.value }); descriptionModel.reset() \"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n          <div class=\"col-md-4\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Device type filter' | translate }}\n\n                <button\n                 class=\"btn-clean\"\n                  popover=\"{{ 'If the filter is set, the firmware will show up for installation only for devices of that type. If no filter is set, it will be available for all devices.' | translate }}\"\n                  [outsideClick]=\"true\"\n                  container=\"body\"\n                >\n                  <i c8yIcon=\"question-circle-o\" class=\"text-info\"></i>\n                </button>\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  #deviceTypeInput\n                  type=\"text\"\n                  class=\"form-control\"\n                  [ngModel]=\"(firmware$ | async)?.c8y_Filter?.type\"\n                  #deviceTypeModel=\"ngModel\"\n                  placeholder=\"{{ 'e.g.' | translate }} c8y_Linux\"\n                  [ngStyle]=\"{ 'width.ch': (firmware$ | async)?.type?.length || 31 }\"\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    (click)=\" updateFirmware$.next({ c8y_Filter: { type: deviceTypeInput.value } }); deviceTypeModel.reset()\"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n<div class=\"row\">\n  <div class=\"col-lg-12 col-lg-max\">\n    <div class=\"card\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>\n          Versions and patches\n        </h4>\n      </div>\n\n      <div class=\"card-block p-t-0 p-b-24\">\n        <div *ngIf=\"(baseVersions$ | async)?.data.length === 0\">\n          <div class=\"c8y-empty-state text-center\">\n            <h1 c8yIcon=\"c8y-firmware\" class=\"c8y-icon-duocolor\"></h1>\n            <h3 translate>No versions to display.</h3>\n            <p translate>Add a new version by clicking below.</p>\n            <p>\n              <button\n                class=\"btn btn-primary\"\n                title=\"{{ 'Add firmware' | translate }}\"\n                (click)=\"addBaseVersion()\"\n              >\n                {{ 'Add firmware' | translate }}\n              </button>\n            </p>\n          </div>\n        </div>\n\n        <c8y-list-group\n          [ngClass]=\"{ 'dd-low': (baseVersions$ | async)?.data.length < 10 }\"\n          *ngIf=\"(baseVersions$ | async)?.data.length > 0\"\n        >\n          <c8y-li\n            *c8yFor=\"let baseVersion of baseVersions$ | async; let i = index; loadMore: 'auto'\"\n            [ngClass]=\"{ 'c8y-list__item--empty-actions': !(getPatchVersions$(baseVersion) | async)?.data.length }\"\n            [collapsed]=\"!expanded[baseVersion.id]\"\n            (collapsedChange)=\"expanded[baseVersion.id] = !$event\"\n          >\n            <c8y-li-icon>\n              <i c8yIcon=\"c8y-firmware\"></i>\n            </c8y-li-icon>\n\n            <c8y-li-body class=\"content-flex-50\">\n              <div class=\"col-4\">\n                <p class=\"text-truncate-wrap\" title=\"{{ baseVersion.c8y_Firmware.version }}\">\n                  {{ baseVersion.c8y_Firmware.version }}\n                </p>\n              </div>\n              <div class=\"col-5\">\n                <p class=\"text-truncate-wrap\">\n                  <span class=\"text-label-small m-r-8\" translate>\n                    File\n                  </span>\n                  <span title=\"{{ getBinaryName$(baseVersion.c8y_Firmware.url) | async }}\">\n                    <c8y-file-download url=\"{{ baseVersion.c8y_Firmware.url }}\"></c8y-file-download>\n                  </span>\n                </p>\n              </div>\n              <div class=\"col-2 flex-row a-i-start\">\n                <span *ngIf=\"isLegacy$ | async\" class=\"label label-warning flex-item-right-sm\">\n                  {{ 'Legacy' | translate }}\n                </span>\n\n                <span *ngIf=\"!(isLegacy$ | async)\">\n                  <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) === null\">\n                    <span class=\"label label-info\">\n                      <i c8yIcon=\"circle-o-notch\" class=\"icon-spin\"></i>\n                    </span>\n                  </span>\n                  <span *ngIf=\"(getPatchVersionsCount$(baseVersion) | async) !== null\">\n                    <span [ngPlural]=\"getPatchVersionsCount$(baseVersion) | async\">\n                      <ng-template ngPluralCase=\"=0\">\n                        <span class=\"label label-default flex-item-right-sm\">\n                          <span translate>No patches</span>\n                        </span>\n                      </ng-template>\n                      <ng-template ngPluralCase=\"=1\">\n                        <span class=\"label label-info\">\n                          <span translate>1 patch</span>\n                        </span>\n                      </ng-template>\n                      <ng-template ngPluralCase=\"other\">\n                        <span class=\"label label-info\">\n                          <span\n                            ngNonBindable\n                            translate\n                            [translateParams]=\"{ count: getPatchVersionsCount$(baseVersion) | async }\"\n                          >\n                            {{ count }} patches\n                          </span>\n                        </span>\n                      </ng-template>\n                    </span>\n                  </span>\n                </span>\n              </div>\n              <div class=\"fit-h-20 visible-xs\" *ngIf=\"!(isLegacy$ | async)\">\n                <button\n                  class=\"btn btn-danger btn-xs m-t-8 \"\n                  (click)=\"deleteBaseVersion(baseVersion)\"\n                  title=\"{{ 'Delete' | translate }}\"\n                >\n                  <i c8yIcon=\"minus-circle\"></i>\n                  {{ 'Delete' | translate }}\n                </button>\n              </div>\n              <div *ngIf=\"!(isLegacy$ | async)\" class=\"flex-item-right fit-h-20 p-r-8 hidden-xs\">\n                <button\n                  class=\"btn btn-dot text-danger showOnHover\"\n                  (click)=\"deleteBaseVersion(baseVersion)\"\n                  title=\"{{ 'Delete' | translate }}\"\n                >\n                  <i c8yIcon=\"minus-circle\"></i>\n                </button>\n              </div>\n            </c8y-li-body>\n            <c8y-li-collapse *ngIf=\"(getPatchVersions$(baseVersion) | async)?.data.length\">\n              <c8y-list-group class=\"separator-top\">\n                <c8y-li\n                  *c8yFor=\" let patchVersion of getPatchVersions$(baseVersion) | async; let i = index; loadMore: 'auto' \"\n                >\n                  <c8y-li-icon>\n                    <i c8yIcon=\"c8y-firmware\"></i>\n                  </c8y-li-icon>\n                  <c8y-li-body class=\"content-flex-50\">\n                    <div class=\"col-4\">\n                      {{ patchVersion.c8y_Firmware.version }}\n                    </div>\n                    <div class=\"col-5\">\n                      <div class=\"text-truncate\">\n                        <span class=\"text-label-small m-r-8\" translate>\n                          File\n                        </span>\n                        <c8y-file-download\n                          url=\"{{ patchVersion.c8y_Firmware.url }}\"\n                        ></c8y-file-download>\n                      </div>\n                    </div>\n                    <div class=\"visible-xs m-t-8\">\n                      <button\n                        class=\"btn btn-danger btn-xs\"\n                        (click)=\"deletePatchVersion(patchVersion)\"\n                        title=\"{{ 'Delete' | translate }}\"\n                      >\n                        <i c8yIcon=\"minus-circle\"></i>\n                        {{ 'Delete' | translate }}\n                      </button>\n                    </div>\n                    <div class=\"flex-item-right p-r-8 hidden-xs fit-h-20\">\n                      <button\n                        class=\"btn btn-dot text-danger showOnHover\"\n                        (click)=\"deletePatchVersion(patchVersion)\"\n                        title=\"{{ 'Delete' | translate }}\"\n                      >\n                        <i c8yIcon=\"minus-circle\"></i>\n                      </button>\n                    </div>\n                  </c8y-li-body>\n                </c8y-li>\n              </c8y-list-group>\n            </c8y-li-collapse>\n          </c8y-li>\n        </c8y-list-group>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.RepositoryService }, { type: ɵngcc4.AlertService }, { type: ɵngcc5.TranslateService }, { type: ɵngcc4.ModalService }, { type: ɵngcc6.BsModalService }]; }, null); })();
