// @c8y/ngx-components/esm2015/connectivity/connectivity.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ConditionalTabsOutletComponent, CoreModule } from '@c8y/ngx-components';
import { ActilityModule } from './actility/actility.module';
import { ImpactModule } from './impact/impact.module';
import { SigfoxModule } from './sigfox/sigfox.module';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const connectivityRoutes = [
    {
        path: 'connectivitySettings',
        component: ConditionalTabsOutletComponent
    }
];
export class ConnectivityModule {
}
ConnectivityModule.ɵfac = function ConnectivityModule_Factory(t) { return new (t || ConnectivityModule)(); };
ConnectivityModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ConnectivityModule });
ConnectivityModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CoreModule,
            ActilityModule,
            ImpactModule,
            SigfoxModule,
            RouterModule.forChild(connectivityRoutes)
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConnectivityModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    ActilityModule,
                    ImpactModule,
                    SigfoxModule,
                    RouterModule.forChild(connectivityRoutes)
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ConnectivityModule, { imports: function () { return [CoreModule,
        ActilityModule,
        ImpactModule,
        SigfoxModule, ɵngcc1.RouterModule]; } }); })();
