// @c8y/ngx-components/esm2015/context-dashboard/appearance-settings.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { DASHBOARD_THEME_CLASSES, WIDGET_HEADER_CLASSES } from './context-dashboard.model';
import { ControlContainer, NgForm } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from 'ngx-bootstrap/popover';

function AppearanceSettingsComponent_li_8_small_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 16);
    ɵngcc0.ɵɵtext(1, "Dashboard default");
    ɵngcc0.ɵɵelementEnd();
} }
function AppearanceSettingsComponent_li_8_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 8);
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵelementStart(2, "label", 10);
    ɵngcc0.ɵɵelementStart(3, "input", 11);
    ɵngcc0.ɵɵlistener("click", function AppearanceSettingsComponent_li_8_Template_input_click_3_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); const themeClassItem_r2 = restoredCtx.$implicit; const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.themeClassClick(themeClassItem_r2.class); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "label", 12);
    ɵngcc0.ɵɵelementStart(6, "span");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(9, "br");
    ɵngcc0.ɵɵtemplate(10, AppearanceSettingsComponent_li_8_small_10_Template, 2, 0, "small", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "button", 14);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelement(13, "i", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const themeClassItem_r2 = ctx.$implicit;
    const i_r3 = ctx.index;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("id", "groupradiocontentclass" + i_r3)("value", themeClassItem_r2.class)("ngModel", ctx_r0.themeClass);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("for", "groupradiocontentclass" + i_r3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 8, themeClassItem_r2.label));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", themeClassItem_r2.class === ctx_r0.defaultThemeClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(12, 10, themeClassItem_r2.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
} }
function AppearanceSettingsComponent_li_13_small_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 16);
    ɵngcc0.ɵɵtext(1, "Dashboard default");
    ɵngcc0.ɵɵelementEnd();
} }
function AppearanceSettingsComponent_li_13_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 17);
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵelementStart(2, "label", 10);
    ɵngcc0.ɵɵelementStart(3, "input", 18);
    ɵngcc0.ɵɵlistener("click", function AppearanceSettingsComponent_li_13_Template_input_click_3_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r11); const headerClassItem_r7 = restoredCtx.$implicit; const ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.headerClassClick(headerClassItem_r7.class); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "label", 12);
    ɵngcc0.ɵɵelementStart(6, "span");
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(9, "br");
    ɵngcc0.ɵɵtemplate(10, AppearanceSettingsComponent_li_13_small_10_Template, 2, 0, "small", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "button", 14);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelement(13, "i", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const headerClassItem_r7 = ctx.$implicit;
    const i_r8 = ctx.index;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("id", "groupradioheaderclass" + i_r8)("value", headerClassItem_r7.class)("ngModel", ctx_r1.headerClass);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("for", "groupradioheaderclass" + i_r8);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 8, headerClassItem_r7.label));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", headerClassItem_r7.class === ctx_r1.defaultHeaderClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(12, 10, headerClassItem_r7.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
} }
export class AppearanceSettingsComponent {
    constructor() {
        this.themeClass = 'dashboard-theme-light';
        this.headerClass = 'panel-title-regular';
        this.themeClassChange = new EventEmitter();
        this.headerClassChange = new EventEmitter();
        this.possibleStylingTheme = DASHBOARD_THEME_CLASSES;
        this.possibleStylingHeader = WIDGET_HEADER_CLASSES;
    }
    themeClassClick(value) {
        value = this.resetToDefault(value, this.themeClass, this.defaultThemeClass);
        this.themeClass = value;
        this.themeClassChange.emit(value);
    }
    headerClassClick(value) {
        value = this.resetToDefault(value, this.headerClass, this.defaultHeaderClass);
        this.headerClass = value;
        this.headerClassChange.emit(value);
    }
    resetToDefault(value, compareTo, defaultValue) {
        if (defaultValue && value === compareTo) {
            value = defaultValue;
        }
        return value;
    }
}
AppearanceSettingsComponent.ɵfac = function AppearanceSettingsComponent_Factory(t) { return new (t || AppearanceSettingsComponent)(); };
AppearanceSettingsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AppearanceSettingsComponent, selectors: [["c8y-appearance-settings"]], inputs: { themeClass: "themeClass", headerClass: "headerClass", possibleStylingTheme: "possibleStylingTheme", possibleStylingHeader: "possibleStylingHeader", defaultThemeClass: "defaultThemeClass", defaultHeaderClass: "defaultHeaderClass" }, outputs: { themeClassChange: "themeClassChange", headerClassChange: "headerClassChange" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgForm }])], decls: 14, vars: 2, consts: [[1, "legend", "form-block"], ["translate", ""], [1, "row"], [1, "col-sm-6", "m-b-32"], [1, "list-group", 2, "box-shadow", "0 0 0 1px rgba(0, 0, 0, 0.05)"], ["class", "list-group-item flex-row", "style", "align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);", 4, "ngFor", "ngForOf"], [1, "col-sm-6"], ["class", "list-group-item flex-row a-i-center", "style", "box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);", 4, "ngFor", "ngForOf"], [1, "list-group-item", "flex-row", 2, "align-items", "center", "box-shadow", "0 1px 0 0 rgba(0, 0, 0, 0.05)"], [1, "list-item-checkbox"], [1, "c8y-radio"], ["type", "radio", "name", "content", 3, "id", "value", "ngModel", "click"], [3, "for"], ["class", "text-muted", "translate", "", 4, "ngIf"], ["container", "body", "placement", "top", "triggers", "focus", 1, "btn", "btn-clean", "flex-item-top", 3, "popover"], [1, "text-primary", 3, "c8yIcon"], ["translate", "", 1, "text-muted"], [1, "list-group-item", "flex-row", "a-i-center", 2, "box-shadow", "0 1px 0 0 rgba(0, 0, 0, 0.05)"], ["type", "radio", "name", "header", 3, "id", "value", "ngModel", "click"]], template: function AppearanceSettingsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "h6", 0);
        ɵngcc0.ɵɵelementStart(1, "span", 1);
        ɵngcc0.ɵɵtext(2, "Layout");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "label", 1);
        ɵngcc0.ɵɵtext(6, "Theme");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "ul", 4);
        ɵngcc0.ɵɵtemplate(8, AppearanceSettingsComponent_li_8_Template, 14, 12, "li", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "label", 1);
        ɵngcc0.ɵɵtext(11, "Widget header style");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "ul", 4);
        ɵngcc0.ɵɵtemplate(13, AppearanceSettingsComponent_li_13_Template, 14, 12, "li", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.possibleStylingTheme);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.possibleStylingHeader);
    } }, directives: [ɵngcc1.C8yTranslateDirective, ɵngcc2.NgForOf, ɵngcc3.RadioControlValueAccessor, ɵngcc3.DefaultValueAccessor, ɵngcc3.NgControlStatus, ɵngcc3.NgModel, ɵngcc2.NgIf, ɵngcc4.PopoverDirective, ɵngcc1.IconDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
AppearanceSettingsComponent.propDecorators = {
    themeClass: [{ type: Input }],
    headerClass: [{ type: Input }],
    defaultThemeClass: [{ type: Input }],
    defaultHeaderClass: [{ type: Input }],
    themeClassChange: [{ type: Output }],
    headerClassChange: [{ type: Output }],
    possibleStylingTheme: [{ type: Input }],
    possibleStylingHeader: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AppearanceSettingsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-appearance-settings',
                template: "<h6 class=\"legend form-block\"><span translate>Layout</span></h6>\n<div class=\"row \">\n  <div class=\"col-sm-6 m-b-32\">\n    <label translate>Theme</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row\"\n        style=\"align-items: center; box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let themeClassItem of possibleStylingTheme; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradiocontentclass' + i\"\n              name=\"content\"\n              [value]=\"themeClassItem.class\"\n              [ngModel]=\"themeClass\"\n              (click)=\"themeClassClick(themeClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradiocontentclass' + i\">\n          <span>{{ themeClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"themeClassItem.class === defaultThemeClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button\n          class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ themeClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\"\n        >\n          <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-6\">\n    <label translate>Widget header style</label>\n    <ul class=\"list-group\" style=\"box-shadow: 0 0 0 1px rgba(0, 0, 0, 0.05);\">\n      <li\n        class=\"list-group-item flex-row a-i-center\"\n        style=\"box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.05);\"\n        *ngFor=\"let headerClassItem of possibleStylingHeader; let i = index\"\n      >\n        <div class=\"list-item-checkbox\">\n          <label class=\"c8y-radio\">\n            <input\n              type=\"radio\"\n              [id]=\"'groupradioheaderclass' + i\"\n              name=\"header\"\n              [value]=\"headerClassItem.class\"\n              [ngModel]=\"headerClass\"\n              (click)=\"headerClassClick(headerClassItem.class)\"\n            />\n            <span></span>\n          </label>\n        </div>\n        <label [for]=\"'groupradioheaderclass' + i\">\n          <span>{{ headerClassItem.label | translate }}</span>\n          <br />\n          <small *ngIf=\"headerClassItem.class === defaultHeaderClass\" class=\"text-muted\" translate\n            >Dashboard default</small\n          >\n        </label>\n        <button\n          class=\"btn btn-clean flex-item-top\"\n          popover=\"{{ headerClassItem.description | translate }}\"\n          container=\"body\"\n          placement=\"top\"\n          triggers=\"focus\"\n        >\n          <i [c8yIcon]=\"'question-circle-o'\" class=\"text-primary\"></i>\n        </button>\n      </li>\n    </ul>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
            }]
    }], function () { return []; }, { themeClass: [{
            type: Input
        }], headerClass: [{
            type: Input
        }], themeClassChange: [{
            type: Output
        }], headerClassChange: [{
            type: Output
        }], possibleStylingTheme: [{
            type: Input
        }], possibleStylingHeader: [{
            type: Input
        }], defaultThemeClass: [{
            type: Input
        }], defaultHeaderClass: [{
            type: Input
        }] }); })();
