// @c8y/ngx-components/esm2015/context-dashboard/dashboard-actions.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, ViewContext, Permissions } from '@c8y/ngx-components';
import { Router } from '@angular/router';
import { ContextDashboardService } from './context-dashboard.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { DashboardDetailComponent } from './dashboard-detail.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './context-dashboard.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@c8y/ngx-components';
export class DashboardActionsFactory {
    constructor(router, contextDashboardService, bsModal, permissions) {
        this.router = router;
        this.contextDashboardService = contextDashboardService;
        this.bsModal = bsModal;
        this.permissions = permissions;
    }
    get(activatedRoute) {
        return __awaiter(this, void 0, void 0, function* () {
            const context = !activatedRoute.parent || activatedRoute.snapshot.data.context
                ? activatedRoute.snapshot.data
                : activatedRoute.parent.snapshot.data;
            const canEditDashboard = yield this.isEnabled(context);
            return [
                {
                    label: gettext('Add dashboard'),
                    priority: 2500,
                    action: () => this.addDashboard(context),
                    disabled: !canEditDashboard
                }
            ];
        });
    }
    addDashboard(context) {
        return __awaiter(this, void 0, void 0, function* () {
            let initialState;
            if (context.context === ViewContext.Device) {
                initialState = {
                    deviceType: context.contextData.type
                };
            }
            const modal = this.bsModal.show(DashboardDetailComponent, {
                class: 'modal-lg',
                initialState,
                ignoreBackdropClick: true
            }).content;
            try {
                const dashboardCfg = yield modal.result;
                const dashboardMO = yield this.contextDashboardService.create(dashboardCfg, context);
                yield this.contextDashboardService.navigateToDashboard(dashboardMO);
                modal.close();
            }
            catch (ex) {
                // intended emptys
            }
        });
    }
    isEnabled(context) {
        return __awaiter(this, void 0, void 0, function* () {
            if (context.contextData && context.contextData.id) {
                return ((yield this.permissions.canEdit(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE'], context.contextData)) && [ViewContext.Device, ViewContext.Group].includes(context.context));
            }
            return (this.permissions.hasAnyRole(['ROLE_INVENTORY_ADMIN', 'ROLE_INVENTORY_CREATE']) &&
                [ViewContext.Device, ViewContext.Group].includes(context.context));
        });
    }
}
DashboardActionsFactory.ɵfac = function DashboardActionsFactory_Factory(t) { return new (t || DashboardActionsFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(ɵngcc2.ContextDashboardService), ɵngcc0.ɵɵinject(ɵngcc3.BsModalService), ɵngcc0.ɵɵinject(ɵngcc4.Permissions)); };
DashboardActionsFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DashboardActionsFactory, factory: DashboardActionsFactory.ɵfac });
DashboardActionsFactory.ctorParameters = () => [
    { type: Router },
    { type: ContextDashboardService },
    { type: BsModalService },
    { type: Permissions }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DashboardActionsFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.ContextDashboardService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc4.Permissions }]; }, null); })();
