// @c8y/ngx-components/esm2015/context-dashboard/widget.service.js
import { __awaiter } from "tslib";
import { Injectable, Inject, Optional } from '@angular/core';
import { sortBy, cloneDeep, get } from 'lodash-es';
import { DynamicComponentService } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { CONTEXT_DASHBOARD_CONFIG } from './context-dashboard.model';
import * as i0 from "@angular/core";
import * as i1 from "@c8y/ngx-components";
import * as i2 from "@ngx-translate/core";
import * as i3 from "./context-dashboard.model";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@ngx-translate/core';
export class WidgetService {
    constructor(dynamicComponentService, translateService, moduleConfig) {
        this.dynamicComponentService = dynamicComponentService;
        this.translateService = translateService;
        this.moduleConfig = moduleConfig;
        this.dynamicComponentService.items$.subscribe(widgets => {
            this.widgets = widgets;
        });
    }
    getWidgetDefinition(componentId) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.dynamicComponentService.getById(componentId);
        });
    }
    getWidgetDefinitions() {
        const translatedComponents = this.widgets.map(cmp => (Object.assign(Object.assign({}, cmp), { label: this.translateService.instant(cmp.label) })));
        return cloneDeep(sortBy(translatedComponents, 'label').filter(this.moduleConfig.widgetFilter));
    }
    mapLegacy(widget) {
        return __awaiter(this, void 0, void 0, function* () {
            const cmp = yield this.getWidgetDefinition(widget.componentId || widget.name);
            if (get(cmp, 'data.settings.upgrade')) {
                widget.widgetComponent = cmp.data.settings.widgetComponent;
                widget.configComponent = cmp.data.settings.configComponent;
                widget.templateUrl = cmp.data.settings.templateUrl;
                widget.configTemplateUrl = cmp.data.settings.configTemplateUrl;
                widget.transformConfigWithContext =
                    cmp.data.settings.componentTransformConfigWithContext ||
                        cmp.data.settings.transformConfigWithContext ||
                        widget.transformConfigWithContext;
            }
            else {
                delete widget.templateUrl;
                delete widget.configTemplateUrl;
            }
            return widget;
        });
    }
}
WidgetService.ɵfac = function WidgetService_Factory(t) { return new (t || WidgetService)(ɵngcc0.ɵɵinject(ɵngcc1.DynamicComponentService), ɵngcc0.ɵɵinject(ɵngcc2.TranslateService), ɵngcc0.ɵɵinject(CONTEXT_DASHBOARD_CONFIG, 8)); };
WidgetService.ɵprov = i0.ɵɵdefineInjectable({ factory: function WidgetService_Factory() { return new WidgetService(i0.ɵɵinject(i1.DynamicComponentService), i0.ɵɵinject(i2.TranslateService), i0.ɵɵinject(i3.CONTEXT_DASHBOARD_CONFIG, 8)); }, token: WidgetService, providedIn: "root" });
WidgetService.ctorParameters = () => [
    { type: DynamicComponentService },
    { type: TranslateService },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [CONTEXT_DASHBOARD_CONFIG,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WidgetService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.DynamicComponentService }, { type: ɵngcc2.TranslateService }, { type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [CONTEXT_DASHBOARD_CONFIG]
            }] }]; }, null); })();
