// @c8y/ngx-components/esm2015/operations/stepper-bulk-type-device-profile/stepper-bulk-type-device-profile.component.js
import { Component } from '@angular/core';
import { gettext } from '@c8y/ngx-components';
import { get } from 'lodash-es';
import { BaseStepperComponent } from '@c8y/ngx-components/operations/bulk-operation-stepper';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/operations/bulk-operation-stepper';
import * as ɵngcc2 from './select-device-profile-step.component';
import * as ɵngcc3 from './confirm-device-profile-selection-step.component';
import * as ɵngcc4 from '@c8y/ngx-components';

function StepperBulkTypeDeviceProfileComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "c8y-select-device-profile-step", 2);
    ɵngcc0.ɵɵlistener("deviceProfile", function StepperBulkTypeDeviceProfileComponent_ng_container_1_Template_c8y_select_device_profile_step_deviceProfile_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.onDeviceProfileSelected($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function StepperBulkTypeDeviceProfileComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-confirm-device-profile-selection-step", 3);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("selectedDeviceProfile", ctx_r1.selectedDeviceProfile);
} }
export class StepperBulkTypeDeviceProfileComponent extends BaseStepperComponent {
    goToConfirmSelection($event) {
        this.deviceTypes = get(this.selectedDeviceProfile, 'c8y_Filter.type');
        $event.stepper.next();
    }
    onDeviceProfileSelected(selectedItem) {
        this.selectedDeviceProfile = selectedItem;
    }
    retrieveOperationPrototype() {
        return {
            name: gettext('Apply device profile'),
            description: get(this.selectedDeviceProfile, 'name'),
            prototype: {
                description: `Apply device profile: ${this.selectedDeviceProfile.name}`,
                profileId: this.selectedDeviceProfile.id,
                profileName: this.selectedDeviceProfile.name,
                c8y_DeviceProfile: this.selectedDeviceProfile.c8y_DeviceProfile
            }
        };
    }
}
StepperBulkTypeDeviceProfileComponent.ɵfac = /*@__PURE__*/ function () { let ɵStepperBulkTypeDeviceProfileComponent_BaseFactory; return function StepperBulkTypeDeviceProfileComponent_Factory(t) { return (ɵStepperBulkTypeDeviceProfileComponent_BaseFactory || (ɵStepperBulkTypeDeviceProfileComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(StepperBulkTypeDeviceProfileComponent)))(t || StepperBulkTypeDeviceProfileComponent); }; }();
StepperBulkTypeDeviceProfileComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StepperBulkTypeDeviceProfileComponent, selectors: [["c8y-stepper-bulk-type-device-profile"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 9, consts: [[4, "customStep", "customStepCompleted", "customStepButtonsDisabled", "customStepOnNext"], [4, "customStep"], [1, "d-contents", 3, "deviceProfile"], [1, "d-contents", 3, "selectedDeviceProfile"]], template: function StepperBulkTypeDeviceProfileComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-bulk-operation-stepper");
        ɵngcc0.ɵɵtemplate(1, StepperBulkTypeDeviceProfileComponent_ng_container_1_Template, 2, 0, "ng-container", 0);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, StepperBulkTypeDeviceProfileComponent_ng_container_3_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(2, 5, "Select device profile"))("customStepCompleted", !!ctx.selectedDeviceProfile)("customStepButtonsDisabled", !ctx.selectedDeviceProfile)("customStepOnNext", ctx.goToConfirmSelection.bind(ctx));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(4, 7, "Confirm selected device profile"));
    } }, directives: [ɵngcc1.BulkOperationStepper, ɵngcc1.CustomStep, ɵngcc2.SelectDeviceProfileStepComponent, ɵngcc3.ConfirmDeviceProfileSelectionStepComponent], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeDeviceProfileComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-stepper-bulk-type-device-profile',
                template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select device profile' | translate; \n      completed: !!selectedDeviceProfile;\n      buttonsDisabled: !selectedDeviceProfile; \n      onNext: goToConfirmSelection.bind(this)\"\n  >\n    <c8y-select-device-profile-step\n      (deviceProfile)=\"onDeviceProfileSelected($event)\"\n      class=\"d-contents\"\n    ></c8y-select-device-profile-step>\n  </ng-container>\n  <ng-container *customStep=\"'Confirm selected device profile' | translate\">\n    <c8y-confirm-device-profile-selection-step\n      class=\"d-contents\"\n      [selectedDeviceProfile]=\"selectedDeviceProfile\"\n    ></c8y-confirm-device-profile-selection-step>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
            }]
    }], null, null); })();
