// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation-details.module.js
import { NgModule } from '@angular/core';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsListModule } from '@c8y/ngx-components/operations/bulk-operations-list';
import { BulkOperationsServiceModule } from '@c8y/ngx-components/operations/bulk-operations-service';
import { SingleOperationDetailsService } from './single-operation-details.service';
import { SingleOperationComponent } from './single-operation.component';
import { SingleOperationModalComponent } from './single-operation-modal.component';
import { SingleOperationListItemComponent } from './single-operation-list-item.component';
import { SingleOperationDetailsComponent } from './single-operation-details.component';
/**
 * This module allows to display details of a single operation in a modal.
 */
import * as ɵngcc0 from '@angular/core';
export class SingleOperationDetailsModule {
}
SingleOperationDetailsModule.ɵfac = function SingleOperationDetailsModule_Factory(t) { return new (t || SingleOperationDetailsModule)(); };
SingleOperationDetailsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SingleOperationDetailsModule });
SingleOperationDetailsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [SingleOperationDetailsService], imports: [[
            CoreModule,
            PopoverModule,
            TooltipModule,
            CollapseModule,
            BulkOperationsListModule,
            BulkOperationsServiceModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SingleOperationDetailsModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    PopoverModule,
                    TooltipModule,
                    CollapseModule,
                    BulkOperationsListModule,
                    BulkOperationsServiceModule
                ],
                providers: [SingleOperationDetailsService],
                declarations: [
                    SingleOperationComponent,
                    SingleOperationModalComponent,
                    SingleOperationListItemComponent,
                    SingleOperationDetailsComponent
                ],
                entryComponents: [
                    SingleOperationComponent,
                    SingleOperationModalComponent,
                    SingleOperationListItemComponent,
                    SingleOperationDetailsComponent
                ],
                exports: [SingleOperationComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SingleOperationDetailsModule, { declarations: function () { return [SingleOperationComponent,
        SingleOperationModalComponent,
        SingleOperationListItemComponent,
        SingleOperationDetailsComponent]; }, imports: function () { return [CoreModule,
        PopoverModule,
        TooltipModule,
        CollapseModule,
        BulkOperationsListModule,
        BulkOperationsServiceModule]; }, exports: function () { return [SingleOperationComponent]; } }); })();
