// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation-details.component.js
import { Component, Input } from '@angular/core';
import { BulkOperationDetailsService } from '@c8y/ngx-components/operations/bulk-operations-list';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components/operations/bulk-operations-list';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';

function SingleOperationDetailsComponent_li_5_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵtext(2, "Bulk operation");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 7);
    ɵngcc0.ɵɵelementStart(4, "a", 10);
    ɵngcc0.ɵɵlistener("click", function SingleOperationDetailsComponent_li_5_Template_a_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.openBulkOperationDetails(); });
    ɵngcc0.ɵɵtext(5, "See details");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SingleOperationDetailsComponent_li_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵtext(2, "Failure reason");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 7);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 1, ctx_r1.operation.failureReason));
} }
function SingleOperationDetailsComponent_li_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 11);
    ɵngcc0.ɵɵtext(2, "Result");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 7);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.operation.c8y_Command.result);
} }
export class SingleOperationDetailsComponent {
    constructor(bulkOperationDetailsService) {
        this.bulkOperationDetailsService = bulkOperationDetailsService;
        this.readOnly = false;
    }
    openBulkOperationDetails() {
        this.bulkOperationDetailsService.openDetails(this.operation.bulkOperationId);
    }
}
SingleOperationDetailsComponent.ɵfac = function SingleOperationDetailsComponent_Factory(t) { return new (t || SingleOperationDetailsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationDetailsService)); };
SingleOperationDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SingleOperationDetailsComponent, selectors: [["c8y-single-operation-details"]], inputs: { readOnly: "readOnly", operation: "operation" }, decls: 20, vars: 9, consts: [[1, "row"], [1, "col-sm-3"], ["translate", "", 1, "legend", "form-block"], [1, "list-unstyled", "small"], ["class", "p-t-4 p-b-4 flex-row separator-bottom wrap", 4, "ngIf"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom", "wrap"], ["translate", "", 1, "small", "m-0", "m-r-4"], [1, "flex-item-right"], [1, "col-sm-9"], [1, "p-16"], ["role", "button", "translate", "", 1, "pointer", 3, "click"], ["translate", "", 1, "small", "flex-item-top", "m-r-4"]], template: function SingleOperationDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Details ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "ul", 3);
        ɵngcc0.ɵɵtemplate(5, SingleOperationDetailsComponent_li_5_Template, 6, 0, "li", 4);
        ɵngcc0.ɵɵelementStart(6, "li", 5);
        ɵngcc0.ɵɵelementStart(7, "label", 6);
        ɵngcc0.ɵɵtext(8, "Status");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "span", 7);
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(12, SingleOperationDetailsComponent_li_12_Template, 6, 3, "li", 4);
        ɵngcc0.ɵɵtemplate(13, SingleOperationDetailsComponent_li_13_Template, 5, 1, "li", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 8);
        ɵngcc0.ɵɵelementStart(15, "div", 2);
        ɵngcc0.ɵɵtext(16, " Operation ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "pre", 9);
        ɵngcc0.ɵɵtext(18);
        ɵngcc0.ɵɵpipe(19, "json");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.bulkOperationId);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 5, ctx.operation.status.toString()));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.failureReason);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation.c8y_Command);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(19, 7, ctx.operation));
    } }, directives: [ɵngcc2.C8yTranslateDirective, ɵngcc3.NgIf], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.JsonPipe], encapsulation: 2 });
SingleOperationDetailsComponent.ctorParameters = () => [
    { type: BulkOperationDetailsService }
];
SingleOperationDetailsComponent.propDecorators = {
    operation: [{ type: Input }],
    readOnly: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SingleOperationDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-single-operation-details',
                template: "<div class=\"row\">\n  <div class=\"col-sm-3\">\n    <div class=\"legend form-block\" translate>\n      Details\n    </div>\n    <ul class=\"list-unstyled small\">\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\" *ngIf=\"operation.bulkOperationId\">\n        <label class=\"small m-0 m-r-4\" translate>Bulk operation</label>\n        <span class=\"flex-item-right\">\n          <a class=\"pointer\" role=\"button\" (click)=\"openBulkOperationDetails()\" translate\n            >See details</a\n          >\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small m-0 m-r-4\" translate>Status</label>\n        <span class=\"flex-item-right\">{{ operation.status.toString() | translate }}</span>\n      </li>\n      <li *ngIf=\"operation.failureReason\" class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small m-0 m-r-4\" translate>Failure reason</label>\n        <span class=\"flex-item-right\">{{ operation.failureReason | translate }}</span>\n      </li>\n      <li *ngIf=\"operation.c8y_Command\" class=\"p-t-4 p-b-4 flex-row separator-bottom wrap\">\n        <label class=\"small flex-item-top m-r-4\" translate>Result</label>\n        <span class=\"flex-item-right\">{{ operation.c8y_Command.result }}</span>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-sm-9\">\n    <div class=\"legend form-block\" translate>\n      Operation\n    </div>\n    <pre class=\"p-16\">{{ operation | json }}</pre>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkOperationDetailsService }]; }, { readOnly: [{
            type: Input
        }], operation: [{
            type: Input
        }] }); })();
