// @c8y/ngx-components/esm2015/reports/export-schedules.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { ActionType } from './export-schedules.interface';
import { ReportsService } from './reports.service';
import { BsModalService } from 'ngx-bootstrap/modal';
import { ScheduleModalComponent } from './schedule-modal.component';
import { gettext, OptionsService } from '@c8y/ngx-components';
import { cloneDeep } from 'lodash-es';
import { CronService } from './cron.service';
import { TranslateService } from '@ngx-translate/core';
import { UserService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './reports.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from './cron.service';
import * as ɵngcc4 from '@ngx-translate/core';
import * as ɵngcc5 from '@c8y/client';
import * as ɵngcc6 from '@c8y/ngx-components';
import * as ɵngcc7 from '@angular/common';
import * as ɵngcc8 from 'ngx-bootstrap/dropdown';

function ExportSchedulesComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementStart(2, "span", 6);
    ɵngcc0.ɵɵtext(3, "Retrieving schedules\u2026");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ExportSchedulesComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 7);
    ɵngcc0.ɵɵtext(2, " Could not load schedules list. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ExportSchedulesComponent_div_3_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelement(1, "h1", 11);
    ɵngcc0.ɵɵelementStart(2, "h3", 6);
    ɵngcc0.ɵɵtext(3, "No export schedules defined.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0) { return { minutes: a0 }; };
function ExportSchedulesComponent_div_3_div_2_div_16_span_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 34);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(1, "number");
    ɵngcc0.ɵɵtext(2, " Hourly: {{ minutes }} minute(s) past the hour. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(4, _c0, ɵngcc0.ɵɵpipeBind2(1, 1, schedule_r7.cronConfig.minute, "2.0-0")));
} }
const _c1 = function (a0, a1) { return { hour: a0, minutes: a1 }; };
function ExportSchedulesComponent_div_3_div_2_div_16_span_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 34);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(1, "number");
    ɵngcc0.ɵɵpipe(2, "number");
    ɵngcc0.ɵɵtext(3, " Daily: at {{ hour }}:{{ minutes }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction2(7, _c1, ɵngcc0.ɵɵpipeBind2(1, 1, schedule_r7.cronConfig.hour, "2.0-0"), ɵngcc0.ɵɵpipeBind2(2, 4, schedule_r7.cronConfig.minute, "2.0-0")));
} }
const _c2 = function (a0, a1, a2) { return { weekDay: a0, hour: a1, minutes: a2 }; };
function ExportSchedulesComponent_div_3_div_2_div_16_span_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 34);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(1, "number");
    ɵngcc0.ɵɵpipe(2, "number");
    ɵngcc0.ɵɵtext(3, " Weekly: {{ weekDay }}, at {{ hour }}:{{ minutes }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction3(7, _c2, ctx_r11.cronService.getWeekDayName(schedule_r7.cronConfig), ɵngcc0.ɵɵpipeBind2(1, 1, schedule_r7.cronConfig.hour, "2.0-0"), ɵngcc0.ɵɵpipeBind2(2, 4, schedule_r7.cronConfig.minute, "2.0-0")));
} }
const _c3 = function (a0, a1, a2) { return { monthDay: a0, hour: a1, minutes: a2 }; };
function ExportSchedulesComponent_div_3_div_2_div_16_span_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 34);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(1, "number");
    ɵngcc0.ɵɵpipe(2, "number");
    ɵngcc0.ɵɵtext(3, " Monthly: {{ monthDay }} day of the month, at {{ hour }}:{{ minutes }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction3(7, _c3, ctx_r12.cronService.getMonthDayName(schedule_r7.cronConfig), ɵngcc0.ɵɵpipeBind2(1, 1, schedule_r7.cronConfig.hour, "2.0-0"), ɵngcc0.ɵɵpipeBind2(2, 4, schedule_r7.cronConfig.minute, "2.0-0")));
} }
const _c4 = function (a0, a1, a2, a3) { return { month: a0, monthDay: a1, hour: a2, minutes: a3 }; };
function ExportSchedulesComponent_div_3_div_2_div_16_span_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 34);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵpipe(1, "number");
    ɵngcc0.ɵɵpipe(2, "number");
    ɵngcc0.ɵɵtext(3, " Yearly: {{ month }}, {{ monthDay }} day of the month, at {{ hour }}:{{ minutes }}. ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction4(7, _c4, ctx_r13.cronService.getMonthName(schedule_r7.cronConfig), ctx_r13.cronService.getMonthDayName(schedule_r7.cronConfig), ɵngcc0.ɵɵpipeBind2(1, 1, schedule_r7.cronConfig.hour, "2.0-0"), ɵngcc0.ɵɵpipeBind2(2, 4, schedule_r7.cronConfig.minute, "2.0-0")));
} }
function ExportSchedulesComponent_div_3_div_2_div_16_ul_22_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "ul", 35);
    ɵngcc0.ɵɵelementStart(1, "li", 36);
    ɵngcc0.ɵɵelementStart(2, "button", 37);
    ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_div_3_div_2_div_16_ul_22_Template_button_click_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); const schedule_r7 = ctx_r21.$implicit; const i_r8 = ctx_r21.index; const ctx_r20 = ɵngcc0.ɵɵnextContext(3); return ctx_r20.editSchedule(schedule_r7, i_r8, $event); });
    ɵngcc0.ɵɵelement(3, "i", 4);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "li", 36);
    ɵngcc0.ɵɵelementStart(7, "button", 37);
    ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_div_3_div_2_div_16_ul_22_Template_button_click_7_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const schedule_r7 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r23 = ɵngcc0.ɵɵnextContext(3); return ctx_r23.duplicateSchedule(schedule_r7, $event); });
    ɵngcc0.ɵɵelement(8, "i", 4);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "li", 36);
    ɵngcc0.ɵɵelementStart(12, "button", 37);
    ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_div_3_div_2_div_16_ul_22_Template_button_click_12_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r26 = ɵngcc0.ɵɵnextContext(); const schedule_r7 = ctx_r26.$implicit; const i_r8 = ctx_r26.index; const ctx_r25 = ɵngcc0.ɵɵnextContext(3); return ctx_r25.removeSchedule(schedule_r7, i_r8, $event); });
    ɵngcc0.ɵɵelement(13, "i", 4);
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("title", ctx_r14.hasRequiredRole ? ctx_r14.buttonLabels.edit : ctx_r14.buttonLabels.editNoPermission)("disabled", !ctx_r14.hasRequiredRole);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "pencil");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 12, "Edit"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", ctx_r14.hasRequiredRole ? ctx_r14.buttonLabels.duplicate : ctx_r14.buttonLabels.duplicateNoPermission)("disabled", !ctx_r14.hasRequiredRole);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "copy");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 14, "Duplicate"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", ctx_r14.hasRequiredRole ? ctx_r14.buttonLabels.delete : ctx_r14.buttonLabels.deleteNoPermission)("disabled", !ctx_r14.hasRequiredRole);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "trash");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 16, "Delete"), " ");
} }
function ExportSchedulesComponent_div_3_div_2_div_16_Template(rf, ctx) { if (rf & 1) {
    const _r28 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_div_3_div_2_div_16_Template_div_click_0_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r28); const schedule_r7 = restoredCtx.$implicit; const i_r8 = restoredCtx.index; const ctx_r27 = ɵngcc0.ɵɵnextContext(3); return ctx_r27.editSchedule(schedule_r7, i_r8, $event); });
    ɵngcc0.ɵɵelementStart(1, "div", 14);
    ɵngcc0.ɵɵelementStart(2, "div", 15);
    ɵngcc0.ɵɵelement(3, "i", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 23);
    ɵngcc0.ɵɵelementStart(5, "div", 24);
    ɵngcc0.ɵɵelementStart(6, "div", 25);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 24);
    ɵngcc0.ɵɵelementStart(9, "div", 26);
    ɵngcc0.ɵɵelement(10, "i", 27);
    ɵngcc0.ɵɵelementStart(11, "span", 28);
    ɵngcc0.ɵɵtemplate(12, ExportSchedulesComponent_div_3_div_2_div_16_span_12_Template, 3, 6, "span", 29);
    ɵngcc0.ɵɵtemplate(13, ExportSchedulesComponent_div_3_div_2_div_16_span_13_Template, 4, 10, "span", 29);
    ɵngcc0.ɵɵtemplate(14, ExportSchedulesComponent_div_3_div_2_div_16_span_14_Template, 4, 11, "span", 29);
    ɵngcc0.ɵɵtemplate(15, ExportSchedulesComponent_div_3_div_2_div_16_span_15_Template, 4, 11, "span", 29);
    ɵngcc0.ɵɵtemplate(16, ExportSchedulesComponent_div_3_div_2_div_16_span_16_Template, 4, 12, "span", 29);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "div", 30);
    ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_div_3_div_2_div_16_Template_div_click_17_listener($event) { return $event.stopPropagation(); });
    ɵngcc0.ɵɵelementStart(18, "div", 31);
    ɵngcc0.ɵɵelementStart(19, "button", 32);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelement(21, "i", 4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(22, ExportSchedulesComponent_div_3_div_2_div_16_ul_22_Template, 16, 18, "ul", 33);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const schedule_r7 = ctx.$implicit;
    const ctx_r6 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", schedule_r7.emailConfig.subject);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", schedule_r7.emailConfig.subject, " ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.cronService.getBase(schedule_r7.cronConfig) === 2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.cronService.getBase(schedule_r7.cronConfig) === 3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.cronService.getBase(schedule_r7.cronConfig) === 4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.cronService.getBase(schedule_r7.cronConfig) === 5);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r6.cronService.getBase(schedule_r7.cronConfig) === 6);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 9, "Actions"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "ellipsis-v");
} }
function ExportSchedulesComponent_div_3_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "div", 13);
    ɵngcc0.ɵɵelementStart(2, "div", 14);
    ɵngcc0.ɵɵelementStart(3, "div", 15);
    ɵngcc0.ɵɵelement(4, "i", 16);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 17);
    ɵngcc0.ɵɵelementStart(6, "div", 5);
    ɵngcc0.ɵɵelementStart(7, "div", 18);
    ɵngcc0.ɵɵelementStart(8, "label", 19);
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 20);
    ɵngcc0.ɵɵelementStart(12, "label", 19);
    ɵngcc0.ɵɵtext(13);
    ɵngcc0.ɵɵpipe(14, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(15, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(16, ExportSchedulesComponent_div_3_div_2_div_16_Template, 23, 11, "div", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(9);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 3, "Description"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(14, 5, "Frequency"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r5.scheduleList);
} }
function ExportSchedulesComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, ExportSchedulesComponent_div_3_div_1_Template, 4, 0, "div", 8);
    ɵngcc0.ɵɵtemplate(2, ExportSchedulesComponent_div_3_div_2_Template, 17, 7, "div", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.scheduleList.length);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.scheduleList.length);
} }
function ExportSchedulesComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 38);
    ɵngcc0.ɵɵtext(1, " You don't have the permission required to schedule exports. ");
    ɵngcc0.ɵɵelementEnd();
} }
export class ExportSchedulesComponent {
    constructor(reportsService, bsModalService, cronService, translateService, userService, optionsService) {
        this.reportsService = reportsService;
        this.bsModalService = bsModalService;
        this.cronService = cronService;
        this.translateService = translateService;
        this.userService = userService;
        this.optionsService = optionsService;
        this.onSchedulesUpdate = new EventEmitter();
        this.scheduleList = [];
        this.initialSchedule = {
            timestamp: null,
            emailConfig: {
                to: [],
                cc: [],
                bcc: [],
                replyTo: '',
                text: '',
                subject: ''
            },
            cronConfig: {
                minute: '0',
                hour: '0',
                day: '1',
                month: '1',
                weekday: '?'
            }
        };
        this.listClass = 'interact-list';
        this.sortReverse = false;
        this.isOpen = {};
        this.isEditMenuOpen = false;
        this.currentUserEmail = '';
        this.hasRequiredRole = false;
        this.defaultExportEmailTemplate = this.translateService.instant(gettext('File with exported data can be downloaded from {tenant-domain}/inventory/binaries/{binaryId}.'));
        this.loadingStatus = {
            inProgress: false,
            done: false,
            error: false
        };
    }
    set exportId(exportId) {
        this._exportId = exportId;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.hasRequiredRole = yield this.checkRole();
            this.getScheduleList(true);
            const currentUserEmail = yield this.getCurrentUserEmail();
            this.initialSchedule.emailConfig.text = yield this.optionsService.getTenantOption('configuration', 'export.data.mail.text', this.defaultExportEmailTemplate);
            this.initialSchedule.emailConfig.to = currentUserEmail;
            this.exp = yield this.reportsService.getExport(this._exportId);
            this.initialSchedule.emailConfig.subject = this.translateService.instant(gettext('Export of "{{expName}}"'), { expName: this.exp.name });
        });
    }
    ngOnChanges() {
        this.translateButtonTitles();
    }
    translateButtonTitles() {
        this.buttonLabels = {
            edit: this.translateService.instant(gettext('Edit schedule')),
            editNoPermission: this.translateService.instant(gettext('Edit schedule - no permissions')),
            duplicate: this.translateService.instant(gettext('Duplicate schedule')),
            duplicateNoPermission: this.translateService.instant(gettext('Duplicate schedule - no permissions')),
            delete: this.translateService.instant(gettext('Delete schedule')),
            deleteNoPermission: this.translateService.instant(gettext('Delete schedule - no permissions'))
        };
    }
    getCurrentUserEmail() {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.userService.current();
            return data && data.email ? [data.email] : [];
        });
    }
    checkRole() {
        return __awaiter(this, void 0, void 0, function* () {
            const { data } = yield this.userService.current();
            const role = 'ROLE_SCHEDULE_REPORT_ADMIN';
            const hasRole = this.userService.hasRole(data, role);
            return hasRole;
        });
    }
    getScheduleList(withProgress) {
        return __awaiter(this, void 0, void 0, function* () {
            if (withProgress) {
                this.loadingStatus.inProgress = true;
            }
            this.scheduleList = yield this.reportsService.getScheduleList(this._exportId);
            if (withProgress) {
                this.loadingStatus.inProgress = false;
            }
        });
    }
    addSchedule() {
        this.openAddEditModal(this._exportId, this.initialSchedule, ActionType.CREATE);
    }
    editSchedule(schedule, index, event) {
        if (this.hasRequiredRole) {
            event.preventDefault();
            this.openAddEditModal(this._exportId, schedule, ActionType.EDIT, index);
        }
    }
    duplicateSchedule(schedule, event) {
        event.preventDefault();
        this.openAddEditModal(this._exportId, schedule, ActionType.DUPLICATE);
    }
    openAddEditModal(exportId, schedule, actionType, index) {
        const payload = { actionType, exportId, schedule: cloneDeep(schedule) };
        const modalOptions = { class: 'modal-sm', initialState: payload };
        this.modalRef = this.bsModalService.show(ScheduleModalComponent, modalOptions);
        this.modalRef.content.emitter.subscribe((load) => this.getMessageFromModal(load, index));
    }
    getMessageFromModal(payload, index) {
        if (payload.success) {
            if (index !== undefined) {
                this.scheduleList[index] = payload.schedule;
            }
            else {
                this.scheduleList.push(payload.schedule);
            }
            this.onSchedulesUpdate.emit(this.scheduleList);
        }
    }
    removeSchedule(schedule, index, event) {
        event.preventDefault();
        this.scheduleList.splice(index, 1);
        this.onSchedulesUpdate.emit(this.scheduleList);
    }
}
ExportSchedulesComponent.ɵfac = function ExportSchedulesComponent_Factory(t) { return new (t || ExportSchedulesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ReportsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.CronService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.OptionsService)); };
ExportSchedulesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ExportSchedulesComponent, selectors: [["export-schedules"]], inputs: { exportId: "exportId" }, outputs: { onSchedulesUpdate: "onSchedulesUpdate" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 10, vars: 12, consts: [["class", "flex-row", 4, "ngIf"], [4, "ngIf"], ["class", "alert alert-warning max-width-100", "role", "alert", "translate", "", 4, "ngIf"], ["type", "button", 1, "btn-add-block", "m-t-16", 3, "title", "disabled", "click"], [3, "c8yIcon"], [1, "flex-row"], ["translate", ""], ["translate", "", 1, "alert", "alert-warning", "max-width-100"], ["class", "c8y-empty-state text-center max-width-100", 4, "ngIf"], ["class", "c8y-list__group", 4, "ngIf"], [1, "c8y-empty-state", "text-center", "max-width-100"], ["c8yIcon", "c8y-report", 1, "c8y-icon-duocolor"], [1, "c8y-list__group"], [1, "c8y-list__item", "hidden-xs"], [1, "c8y-list__item__block"], [1, "c8y-list__item__icon"], [1, "p-l-24"], [1, "c8y-list__item__body"], [1, "col-sm-6"], [1, "m-0"], [1, "col-sm-6", "m-r-40"], ["class", "c8y-list__item flex-row pointer", 3, "click", 4, "ngFor", "ngForOf"], [1, "c8y-list__item", "flex-row", "pointer", 3, "click"], [1, "c8y-list__item__body", "flex-row"], [1, "col-sm-6", "col-xs-6"], [1, "text-truncate", 3, "title"], [1, "flex-row", "a-i-baseline"], ["c8yIcon", "calendar", 1, "text-muted", "m-r-4"], [1, "smart-rule-information"], ["ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], [1, "c8y-list__item__actions", 3, "click"], ["dropdown", "", 1, "settings", "dropdown"], ["dropdownToggle", "", "aria-haspopup", "true", "aria-expanded", "false", 1, "dropdown-toggle", "c8y-dropdown", 3, "title"], ["class", "dropdown-menu dropdown-menu-right", 4, "dropdownMenu"], ["translate", "", 3, "translateParams"], [1, "dropdown-menu", "dropdown-menu-right"], ["role", "menuitem"], [3, "title", "disabled", "click"], ["role", "alert", "translate", "", 1, "alert", "alert-warning", "max-width-100"]], template: function ExportSchedulesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtemplate(1, ExportSchedulesComponent_div_1_Template, 4, 0, "div", 0);
        ɵngcc0.ɵɵtemplate(2, ExportSchedulesComponent_div_2_Template, 3, 0, "div", 1);
        ɵngcc0.ɵɵtemplate(3, ExportSchedulesComponent_div_3_Template, 3, 2, "div", 1);
        ɵngcc0.ɵɵtemplate(4, ExportSchedulesComponent_div_4_Template, 2, 0, "div", 2);
        ɵngcc0.ɵɵelementStart(5, "button", 3);
        ɵngcc0.ɵɵlistener("click", function ExportSchedulesComponent_Template_button_click_5_listener() { return ctx.addSchedule(); });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelement(7, "i", 4);
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.loadingStatus.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.loadingStatus.inProgress && ctx.loadingStatus.done && ctx.loadingStatus.error);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.loadingStatus.inProgress && !ctx.loadingStatus.done && !ctx.loadingStatus.error);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.hasRequiredRole);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 8, "Add schedule"));
        ɵngcc0.ɵɵproperty("disabled", !ctx.hasRequiredRole);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "plus-circle");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 10, "Add schedule"), " ");
    } }, directives: [ɵngcc7.NgIf, ɵngcc6.IconDirective, ɵngcc6.LoadingComponent, ɵngcc6.C8yTranslateDirective, ɵngcc7.NgForOf, ɵngcc8.BsDropdownDirective, ɵngcc8.BsDropdownToggleDirective, ɵngcc8.BsDropdownMenuDirective], pipes: [ɵngcc6.C8yTranslatePipe, ɵngcc7.DecimalPipe], encapsulation: 2 });
ExportSchedulesComponent.ctorParameters = () => [
    { type: ReportsService },
    { type: BsModalService },
    { type: CronService },
    { type: TranslateService },
    { type: UserService },
    { type: OptionsService }
];
ExportSchedulesComponent.propDecorators = {
    exportId: [{ type: Input }],
    onSchedulesUpdate: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ExportSchedulesComponent, [{
        type: Component,
        args: [{
                selector: 'export-schedules',
                template: "<div>\n  <div *ngIf=\"loadingStatus.inProgress\" class=\"flex-row\">\n    <c8y-loading></c8y-loading>\n    <span translate>Retrieving schedules\u2026</span>\n  </div>\n\n  <div *ngIf=\"!loadingStatus.inProgress && loadingStatus.done && loadingStatus.error\">\n    <div class=\"alert alert-warning max-width-100\" translate>\n      Could not load schedules list.\n    </div>\n  </div>\n\n  <div *ngIf=\"!loadingStatus.inProgress && !loadingStatus.done && !loadingStatus.error\">\n    <div class=\"c8y-empty-state text-center max-width-100\" *ngIf=\"!scheduleList.length\">\n      <h1 c8yIcon=\"c8y-report\" class=\"c8y-icon-duocolor\"></h1>\n      <h3 translate>No export schedules defined.</h3>\n    </div>\n\n    <div class=\"c8y-list__group\" *ngIf=\"scheduleList.length\">\n      <div class=\"c8y-list__item hidden-xs\">\n        <div class=\"c8y-list__item__block\">\n          <div class=\"c8y-list__item__icon\">\n            <i class=\"p-l-24\"></i>\n          </div>\n          <div class=\"c8y-list__item__body\">\n            <div class=\"flex-row\">\n              <div class=\"col-sm-6\">\n                <label class=\"m-0\">\n                  {{ 'Description' | translate }}\n                </label>\n              </div>\n              <div class=\"col-sm-6 m-r-40\">\n                <label class=\"m-0\">\n                  {{ 'Frequency' | translate }}\n                </label>\n              </div>\n            </div>\n          </div>\n          <span></span>\n        </div>\n      </div>\n\n      <div\n        class=\"c8y-list__item flex-row pointer\"\n        *ngFor=\"let schedule of scheduleList; index as i\"\n        (click)=\"editSchedule(schedule, i, $event)\"\n      >\n        <div class=\"c8y-list__item__block\">\n          <div class=\"c8y-list__item__icon\">\n            <i c8yIcon=\"c8y-report\" class=\"c8y-icon-duocolor\"></i>\n          </div>\n          <div class=\"c8y-list__item__body flex-row\">\n            <div class=\"col-sm-6 col-xs-6\">\n              <div class=\"text-truncate\" title=\"{{ schedule.emailConfig.subject }}\">\n                {{ schedule.emailConfig.subject }}\n              </div>\n            </div>\n            <div class=\"col-sm-6 col-xs-6\">\n              <div class=\"flex-row a-i-baseline\">\n                <i c8yIcon=\"calendar\" class=\"text-muted m-r-4\"></i>\n                <span class=\"smart-rule-information\">\n                  <span\n                    *ngIf=\"cronService.getBase(schedule.cronConfig) === 2\"\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{ minutes: schedule.cronConfig.minute | number: '2.0-0' }\"\n                  >\n                    Hourly: {{ minutes }} minute(s) past the hour.\n                  </span>\n                  <span\n                    *ngIf=\"cronService.getBase(schedule.cronConfig) === 3\"\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{\n                      hour: schedule.cronConfig.hour | number: '2.0-0',\n                      minutes: schedule.cronConfig.minute | number: '2.0-0'\n                    }\"\n                  >\n                    Daily: at {{ hour }}:{{ minutes }}.\n                  </span>\n                  <span\n                    *ngIf=\"cronService.getBase(schedule.cronConfig) === 4\"\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{\n                      weekDay: cronService.getWeekDayName(schedule.cronConfig),\n                      hour: schedule.cronConfig.hour | number: '2.0-0',\n                      minutes: schedule.cronConfig.minute | number: '2.0-0'\n                    }\"\n                  >\n                    Weekly: {{ weekDay }}, at {{ hour }}:{{ minutes }}.\n                  </span>\n                  <span\n                    *ngIf=\"cronService.getBase(schedule.cronConfig) === 5\"\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{\n                      monthDay: cronService.getMonthDayName(schedule.cronConfig),\n                      hour: schedule.cronConfig.hour | number: '2.0-0',\n                      minutes: schedule.cronConfig.minute | number: '2.0-0'\n                    }\"\n                  >\n                    Monthly: {{ monthDay }} day of the month, at {{ hour }}:{{ minutes }}.\n                  </span>\n                  <span\n                    *ngIf=\"cronService.getBase(schedule.cronConfig) === 6\"\n                    ngNonBindable\n                    translate\n                    [translateParams]=\"{\n                      month: cronService.getMonthName(schedule.cronConfig),\n                      monthDay: cronService.getMonthDayName(schedule.cronConfig),\n                      hour: schedule.cronConfig.hour | number: '2.0-0',\n                      minutes: schedule.cronConfig.minute | number: '2.0-0'\n                    }\"\n                  >\n                    Yearly: {{ month }}, {{ monthDay }} day of the month, at {{ hour }}:{{\n                      minutes\n                    }}.\n                  </span>\n                </span>\n              </div>\n            </div>\n          </div>\n          <div class=\"c8y-list__item__actions\" (click)=\"$event.stopPropagation()\">\n            <div class=\"settings dropdown\" dropdown>\n              <button\n                class=\"dropdown-toggle c8y-dropdown\"\n                dropdownToggle\n                aria-haspopup=\"true\"\n                aria-expanded=\"false\"\n                title=\"{{ 'Actions' | translate }}\"\n              >\n                <i [c8yIcon]=\"'ellipsis-v'\"></i>\n              </button>\n              <ul class=\"dropdown-menu dropdown-menu-right\" *dropdownMenu>\n                <li role=\"menuitem\">\n                  <button\n                    [title]=\"hasRequiredRole ? buttonLabels.edit : buttonLabels.editNoPermission\"\n                    (click)=\"editSchedule(schedule, i, $event)\"\n                    [disabled]=\"!hasRequiredRole\"\n                  >\n                    <i [c8yIcon]=\"'pencil'\"></i> {{ 'Edit' | translate }}\n                  </button>\n                </li>\n                <li role=\"menuitem\">\n                  <button\n                    [title]=\"\n                      hasRequiredRole ? buttonLabels.duplicate : buttonLabels.duplicateNoPermission\n                    \"\n                    (click)=\"duplicateSchedule(schedule, $event)\"\n                    [disabled]=\"!hasRequiredRole\"\n                  >\n                    <i [c8yIcon]=\"'copy'\"></i> {{ 'Duplicate' | translate }}\n                  </button>\n                </li>\n                <li role=\"menuitem\">\n                  <button\n                    [title]=\"\n                      hasRequiredRole ? buttonLabels.delete : buttonLabels.deleteNoPermission\n                    \"\n                    (click)=\"removeSchedule(schedule, i, $event)\"\n                    [disabled]=\"!hasRequiredRole\"\n                  >\n                    <i [c8yIcon]=\"'trash'\"></i> {{ 'Delete' | translate }}\n                  </button>\n                </li>\n              </ul>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n  <div class=\"alert alert-warning max-width-100\" *ngIf=\"!hasRequiredRole\" role=\"alert\" translate>\n    You don't have the permission required to schedule exports.\n  </div>\n  <button\n    type=\"button\"\n    class=\"btn-add-block m-t-16\"\n    title=\"{{ 'Add schedule' | translate }}\"\n    (click)=\"addSchedule()\"\n    [disabled]=\"!hasRequiredRole\"\n  >\n    <i [c8yIcon]=\"'plus-circle'\"></i>\n    {{ 'Add schedule' | translate }}\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ReportsService }, { type: ɵngcc2.BsModalService }, { type: ɵngcc3.CronService }, { type: ɵngcc4.TranslateService }, { type: ɵngcc5.UserService }, { type: ɵngcc6.OptionsService }]; }, { onSchedulesUpdate: [{
            type: Output
        }], exportId: [{
            type: Input
        }] }); })();
