// @c8y/ngx-components/esm2015/reports/reports.module.js
import { NgModule } from '@angular/core';
import { CoreModule, FormsModule } from '@c8y/ngx-components';
import { ReportsService } from './reports.service';
import { CronService } from './cron.service';
import { ExportSchedulesComponent } from './export-schedules.component';
import { ScheduleModalComponent } from './schedule-modal.component';
import { CronComponent } from './cron.component';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
/**
 * The angular module definition for reports.
 * @exports ExportSchedulesComponent
 * @exports ScheduleModalComponent
 * @exports CronComponent
 */
import * as ɵngcc0 from '@angular/core';
export class ReportsModule {
}
ReportsModule.ɵfac = function ReportsModule_Factory(t) { return new (t || ReportsModule)(); };
ReportsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ReportsModule });
ReportsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [ReportsService, CronService], imports: [[CoreModule, FormsModule, BsDropdownModule, TooltipModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ReportsModule, [{
        type: NgModule,
        args: [{
                declarations: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                imports: [CoreModule, FormsModule, BsDropdownModule, TooltipModule],
                entryComponents: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent],
                providers: [ReportsService, CronService],
                exports: [ExportSchedulesComponent, ScheduleModalComponent, CronComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ReportsModule, { declarations: function () { return [ExportSchedulesComponent, ScheduleModalComponent, CronComponent]; }, imports: function () { return [CoreModule, FormsModule, BsDropdownModule, TooltipModule]; }, exports: function () { return [ExportSchedulesComponent, ScheduleModalComponent, CronComponent]; } }); })();
