// @c8y/ngx-components/esm2015/reports/schedule-modal.component.js
import { Component, Output, EventEmitter } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { ActionType } from './export-schedules.interface';
import { CronService } from './cron.service';
import { cloneDeep } from 'lodash-es';
import { gettext } from '@c8y/ngx-components';
import { ReportsService } from './reports.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './reports.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from './cron.service';
import * as ɵngcc4 from '@c8y/ngx-components';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '@angular/forms';
import * as ɵngcc7 from './cron.component';

function ScheduleModalComponent_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 15);
    ɵngcc0.ɵɵtext(1, "New export schedule");
    ɵngcc0.ɵɵelementEnd();
} }
function ScheduleModalComponent_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 15);
    ɵngcc0.ɵɵtext(1, "Edit export schedule");
    ɵngcc0.ɵɵelementEnd();
} }
function ScheduleModalComponent_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 15);
    ɵngcc0.ɵɵtext(1, "Duplicate export schedule");
    ɵngcc0.ɵɵelementEnd();
} }
function ScheduleModalComponent_span_82_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 36);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Create"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Create"), " ");
} }
function ScheduleModalComponent_span_83_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 36);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Save"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Save"), " ");
} }
function ScheduleModalComponent_span_84_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 36);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Duplicate"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Duplicate"), " ");
} }
export class ScheduleModalComponent {
    constructor(reportsService, modalRef, cronService) {
        this.reportsService = reportsService;
        this.modalRef = modalRef;
        this.cronService = cronService;
        this.emitter = new EventEmitter();
        this.ActionType = ActionType;
        this.cronExpression = '* * * * *';
        this.validCron = false;
        this.emitterPayload = {
            success: false,
            message: '',
            schedule: { timestamp: undefined, emailConfig: undefined, cronConfig: undefined }
        };
        this.placeholdersInfo = gettext('Available placeholders: {tenant-domain}, {host}, {binaryId}. Whole link to downloadable file is: {tenant-domain}/inventory/binaries/{binaryId}.');
    }
    ngOnInit() {
        this.oldSchedule = cloneDeep(this.schedule);
        this.populateEmailFieldsFromSchedule(this.schedule);
        this.cronExpression = this.cronService.generateCron(this.schedule.cronConfig);
        this.validCron = this.cronService.validateModels(this.cronService.getBase(this.schedule.cronConfig), this.schedule.cronConfig);
    }
    populateEmailFieldsFromSchedule(schedule) {
        if (schedule.emailConfig.to && schedule.emailConfig.to.length) {
            this.emailTo = schedule.emailConfig.to.toString();
        }
        if (schedule.emailConfig.cc && schedule.emailConfig.cc.length) {
            this.emailCc = schedule.emailConfig.cc.toString();
        }
        if (schedule.emailConfig.bcc && schedule.emailConfig.bcc.length) {
            this.emailBcc = schedule.emailConfig.bcc.toString();
        }
        if (schedule.emailConfig.replyTo) {
            this.emailReplyTo = schedule.emailConfig.replyTo;
        }
        if (schedule.emailConfig.subject) {
            this.emailSubject = schedule.emailConfig.subject;
        }
        if (schedule.emailConfig.text) {
            this.emailText = schedule.emailConfig.text;
        }
    }
    save() {
        this.populateScheduleFromCronExpression();
        this.populateScheduleFromEmailFields();
        if (this.actionType === ActionType.CREATE || this.actionType === ActionType.DUPLICATE) {
            const date = new Date();
            const timestamp = date.getTime();
            this.schedule.timestamp = timestamp;
        }
        this.emitterPayload.success = true;
        this.emitterPayload.schedule = this.schedule;
        this.modalRef.hide();
        // signal to the parent component to update list
        this.emitter.emit(this.emitterPayload);
    }
    cancel() {
        this.modalRef.hide();
    }
    getCron(cron) {
        this.validCron = cron.valid;
        if (cron.valid) {
            this.cronExpression = cron.cron;
        }
    }
    populateScheduleFromCronExpression() {
        this.schedule.cronConfig = this.cronService.generateCronConfig(this.cronExpression);
    }
    convertStringOfEmailsToArray(stringOfEmails) {
        const arr = [];
        if (stringOfEmails) {
            const parts = stringOfEmails.split(',');
            parts.forEach(item => {
                if (item) {
                    arr.push(item);
                }
            });
        }
        return arr;
    }
    populateScheduleFromEmailFields() {
        this.schedule.emailConfig.to = this.emailTo
            ? this.convertStringOfEmailsToArray(this.emailTo)
            : null;
        this.schedule.emailConfig.cc = this.emailCc
            ? this.convertStringOfEmailsToArray(this.emailCc)
            : null;
        this.schedule.emailConfig.bcc = this.emailBcc
            ? this.convertStringOfEmailsToArray(this.emailBcc)
            : null;
        this.schedule.emailConfig.replyTo = this.emailReplyTo;
        this.schedule.emailConfig.subject = this.emailSubject;
        this.schedule.emailConfig.text = this.emailText;
    }
}
ScheduleModalComponent.ɵfac = function ScheduleModalComponent_Factory(t) { return new (t || ScheduleModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ReportsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.CronService)); };
ScheduleModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ScheduleModalComponent, selectors: [["schedule-modal"]], outputs: { emitter: "emitter" }, decls: 85, vars: 41, consts: [[1, "modal-header", "text-center", "bg-primary"], [1, "text-white"], [2, "font-size", "62px"], ["c8yIcon", "c8y-report"], [1, "text-uppercase"], ["translate", "", 4, "ngIf"], [1, "modal-body"], ["translate", "", 1, "lead", "text-center", "p-t-24", "m-b-0"], [1, "modal-inner-scroll", "smart-rule-control"], [1, "edit-smart-rule-details"], ["scheduleForm", "ngForm"], [1, "list-group"], [1, "list-group-item", "bg-gray-white"], [1, "smart-list-icon-label"], [1, "dot", "bg-primary-light", "m-r-8"], ["translate", ""], [1, "p-t-16"], [1, "form-group"], ["name", "cron", 3, "cronIn", "emitter"], [1, "list-group-item"], [1, "d-inline-block"], [1, "help-block", "text-muted", "small", "p-absolute"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-4", "text-14"], [1, "p-t-24"], ["translate", "", 1, "control-label"], ["emails", "", "type", "text", "name", "to", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["emails", "", "type", "text", "name", "cc", 1, "form-control", "span", 3, "placeholder", "ngModel", "ngModelChange"], ["emails", "", "type", "text", "name", "bcc", 1, "form-control", "span", 3, "placeholder", "ngModel", "ngModelChange"], ["email", "", "type", "text", "name", "replyTo", 1, "form-control", "span", 3, "placeholder", "ngModel", "ngModelChange"], ["type", "text", "name", "subject", "required", "", 1, "form-control", "span", 3, "ngModel", "placeholder", "ngModelChange"], ["name", "text", "rows", "4", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], [1, "help-block", "text-muted"], [1, "modal-footer"], [1, "btn", "btn-default", 3, "title", "click"], [1, "btn", "btn-primary", 3, "disabled", "click"], [3, "title", 4, "ngIf"], [3, "title"]], template: function ScheduleModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "header", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "span", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "h4", 4);
        ɵngcc0.ɵɵtemplate(5, ScheduleModalComponent_span_5_Template, 2, 0, "span", 5);
        ɵngcc0.ɵɵtemplate(6, ScheduleModalComponent_span_6_Template, 2, 0, "span", 5);
        ɵngcc0.ɵɵtemplate(7, ScheduleModalComponent_span_7_Template, 2, 0, "span", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "div", 6);
        ɵngcc0.ɵɵelementStart(9, "p", 7);
        ɵngcc0.ɵɵtext(10, "On schedule send export via email");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 8);
        ɵngcc0.ɵɵelementStart(12, "form", 9, 10);
        ɵngcc0.ɵɵelementStart(14, "div", 11);
        ɵngcc0.ɵɵelementStart(15, "div", 12);
        ɵngcc0.ɵɵelementStart(16, "div", 13);
        ɵngcc0.ɵɵelementStart(17, "span", 14);
        ɵngcc0.ɵɵtext(18, "1");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "strong", 15);
        ɵngcc0.ɵɵtext(20, "Frequency");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 16);
        ɵngcc0.ɵɵelementStart(22, "div", 17);
        ɵngcc0.ɵɵelementStart(23, "cron", 18);
        ɵngcc0.ɵɵlistener("emitter", function ScheduleModalComponent_Template_cron_emitter_23_listener($event) { return ctx.getCron($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "div", 19);
        ɵngcc0.ɵɵelementStart(25, "div", 13);
        ɵngcc0.ɵɵelementStart(26, "span", 14);
        ɵngcc0.ɵɵtext(27, "2");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "div", 20);
        ɵngcc0.ɵɵelementStart(29, "strong", 15);
        ɵngcc0.ɵɵtext(30, "Send email");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "p", 21);
        ɵngcc0.ɵɵelement(32, "i", 22);
        ɵngcc0.ɵɵelementStart(33, "span", 15);
        ɵngcc0.ɵɵtext(34, "Enter one or more valid email addresses, separated with a comma.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "div", 23);
        ɵngcc0.ɵɵelementStart(36, "div", 17);
        ɵngcc0.ɵɵelementStart(37, "label", 24);
        ɵngcc0.ɵɵtext(38, "Send to");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(39, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(40, "input", 25);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_input_ngModelChange_40_listener($event) { return ctx.emailTo = $event; });
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 17);
        ɵngcc0.ɵɵelementStart(43, "label", 24);
        ɵngcc0.ɵɵtext(44, "CC");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(46, "input", 26);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_input_ngModelChange_46_listener($event) { return ctx.emailCc = $event; });
        ɵngcc0.ɵɵpipe(47, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(48, "div", 17);
        ɵngcc0.ɵɵelementStart(49, "label", 24);
        ɵngcc0.ɵɵtext(50, "BCC");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(51, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(52, "input", 27);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_input_ngModelChange_52_listener($event) { return ctx.emailBcc = $event; });
        ɵngcc0.ɵɵpipe(53, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(54, "div", 17);
        ɵngcc0.ɵɵelementStart(55, "label", 24);
        ɵngcc0.ɵɵtext(56, "Reply to (single email address)");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(57, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(58, "input", 28);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_input_ngModelChange_58_listener($event) { return ctx.emailReplyTo = $event; });
        ɵngcc0.ɵɵpipe(59, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(60, "div", 17);
        ɵngcc0.ɵɵelementStart(61, "label", 24);
        ɵngcc0.ɵɵtext(62, "Subject");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(63, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(64, "input", 29);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_input_ngModelChange_64_listener($event) { return ctx.emailSubject = $event; });
        ɵngcc0.ɵɵpipe(65, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(66, "div", 17);
        ɵngcc0.ɵɵelementStart(67, "label", 24);
        ɵngcc0.ɵɵtext(68, "Message");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(69, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(70, "textarea", 30);
        ɵngcc0.ɵɵlistener("ngModelChange", function ScheduleModalComponent_Template_textarea_ngModelChange_70_listener($event) { return ctx.emailText = $event; });
        ɵngcc0.ɵɵpipe(71, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(72, "p", 31);
        ɵngcc0.ɵɵtext(73);
        ɵngcc0.ɵɵpipe(74, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(75, "div", 32);
        ɵngcc0.ɵɵelementStart(76, "button", 33);
        ɵngcc0.ɵɵlistener("click", function ScheduleModalComponent_Template_button_click_76_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(77, "translate");
        ɵngcc0.ɵɵtext(78);
        ɵngcc0.ɵɵpipe(79, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(80, "button", 34);
        ɵngcc0.ɵɵlistener("click", function ScheduleModalComponent_Template_button_click_80_listener() { return ctx.save(); });
        ɵngcc0.ɵɵelementStart(81, "span");
        ɵngcc0.ɵɵtemplate(82, ScheduleModalComponent_span_82_Template, 4, 6, "span", 35);
        ɵngcc0.ɵɵtemplate(83, ScheduleModalComponent_span_83_Template, 4, 6, "span", 35);
        ɵngcc0.ɵɵtemplate(84, ScheduleModalComponent_span_84_Template, 4, 6, "span", 35);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r3 = ɵngcc0.ɵɵreference(13);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.CREATE);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.EDIT);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.DUPLICATE);
        ɵngcc0.ɵɵadvance(16);
        ɵngcc0.ɵɵproperty("cronIn", ctx.cronExpression);
        ɵngcc0.ɵɵadvance(17);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(41, 23, "e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailTo);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(47, 25, "e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailCc);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(53, 27, "e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailBcc);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(59, 29, "e.g. joe.doe@example.com`LOCALIZE`"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailReplyTo);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(65, 31, "e.g. Daily report"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailSubject);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(71, 33, "Message"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.emailText);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(74, 35, ctx.placeholdersInfo), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(77, 37, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(79, 39, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("disabled", !ctx.validCron || !_r3.form.valid);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.CREATE);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.EDIT);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.actionType === ctx.ActionType.DUPLICATE);
    } }, directives: [ɵngcc4.IconDirective, ɵngcc5.NgIf, ɵngcc4.C8yTranslateDirective, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc7.CronComponent, ɵngcc4.FormGroupComponent, ɵngcc4.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc4.EmailsValidatorDirective, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc6.EmailValidator], pipes: [ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
ScheduleModalComponent.ctorParameters = () => [
    { type: ReportsService },
    { type: BsModalRef },
    { type: CronService }
];
ScheduleModalComponent.propDecorators = {
    emitter: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ScheduleModalComponent, [{
        type: Component,
        args: [{
                selector: 'schedule-modal',
                template: "<div class=\"modal-header text-center bg-primary\">\n  <header class=\"text-white\">\n    <div style=\"font-size: 62px;\">\n      <span c8yIcon=\"c8y-report\"></span>\n    </div>\n    <h4 class=\"text-uppercase\">\n      <span *ngIf=\"actionType === ActionType.CREATE\" translate>New export schedule</span>\n      <span *ngIf=\"actionType === ActionType.EDIT\" translate>Edit export schedule</span>\n      <span *ngIf=\"actionType === ActionType.DUPLICATE\" translate>Duplicate export schedule</span>\n    </h4>\n  </header>\n</div>\n\n<div class=\"modal-body\">\n  <p class=\"lead text-center p-t-24 m-b-0\" translate>On schedule send export via email</p>\n</div>\n<div class=\"modal-inner-scroll smart-rule-control\">\n  <form #scheduleForm=\"ngForm\" class=\"edit-smart-rule-details\">\n    <div class=\"list-group\">\n      <div class=\"list-group-item bg-gray-white\">\n        <div class=\"smart-list-icon-label\">\n          <span class=\"dot bg-primary-light m-r-8\">1</span>\n          <strong translate>Frequency</strong>\n        </div>\n        <div class=\"p-t-16\">\n          <div class=\"form-group\">\n            <cron [cronIn]=\"cronExpression\" (emitter)=\"getCron($event)\" name=\"cron\"></cron>\n          </div>\n        </div>\n      </div>\n      <div class=\"list-group-item\">\n        <div class=\"smart-list-icon-label\">\n          <span class=\"dot bg-primary-light m-r-8\">2</span>\n          <div class=\"d-inline-block\">\n            <strong translate>Send email</strong>\n            <p class=\"help-block text-muted small p-absolute\">\n              <i class=\"text-info m-r-4 text-14\" c8yIcon=\"info-circle\"></i>\n              <span translate\n                >Enter one or more valid email addresses, separated with a comma.</span\n              >\n            </p>\n          </div>\n        </div>\n        <div class=\"p-t-24\">\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>Send to</label>\n            <c8y-form-group>\n              <input\n                emails\n                type=\"text\"\n                class=\"form-control\"\n                name=\"to\"\n                [(ngModel)]=\"emailTo\"\n                placeholder=\"{{\n                  'e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`' | translate\n                }}\"\n                required\n              />\n            </c8y-form-group>\n          </div>\n\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>CC</label>\n            <c8y-form-group>\n              <input\n                emails\n                type=\"text\"\n                class=\"form-control span\"\n                name=\"cc\"\n                placeholder=\"{{\n                  'e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`' | translate\n                }}\"\n                [(ngModel)]=\"emailCc\"\n              />\n            </c8y-form-group>\n          </div>\n\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>BCC</label>\n            <c8y-form-group>\n              <input\n                emails\n                type=\"text\"\n                class=\"form-control span\"\n                name=\"bcc\"\n                placeholder=\"{{\n                  'e.g. joe.doe@example.com,john.smith@example.com`LOCALIZE`' | translate\n                }}\"\n                [(ngModel)]=\"emailBcc\"\n              />\n            </c8y-form-group>\n          </div>\n\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>Reply to (single email address)</label>\n            <c8y-form-group>\n              <input\n                email\n                type=\"text\"\n                class=\"form-control span\"\n                name=\"replyTo\"\n                placeholder=\"{{ 'e.g. joe.doe@example.com`LOCALIZE`' | translate }}\"\n                [(ngModel)]=\"emailReplyTo\"\n              />\n            </c8y-form-group>\n          </div>\n\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>Subject</label>\n            <c8y-form-group>\n              <input\n                type=\"text\"\n                class=\"form-control span\"\n                name=\"subject\"\n                [(ngModel)]=\"emailSubject\"\n                placeholder=\"{{ 'e.g. Daily report' | translate }}\"\n                required\n              />\n            </c8y-form-group>\n          </div>\n\n          <div class=\"form-group\">\n            <label class=\"control-label\" translate>Message</label>\n            <c8y-form-group>\n              <textarea\n                class=\"form-control\"\n                name=\"text\"\n                [(ngModel)]=\"emailText\"\n                placeholder=\"{{ 'Message' | translate }}\"\n                rows=\"4\"\n                required\n              ></textarea>\n              <p class=\"help-block text-muted\">\n                {{ placeholdersInfo | translate }}\n              </p>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </div>\n  </form>\n</div>\n\n<div class=\"modal-footer\">\n  <button class=\"btn btn-default\" (click)=\"cancel()\" title=\"{{ 'Cancel' | translate }}\">\n    {{ 'Cancel' | translate }}\n  </button>\n  <button\n    class=\"btn btn-primary\"\n    (click)=\"save()\"\n    [disabled]=\"!validCron || !scheduleForm.form.valid\"\n  >\n    <span>\n      <span *ngIf=\"actionType === ActionType.CREATE\" title=\"{{ 'Create' | translate }}\">\n        {{ 'Create' | translate }}\n      </span>\n      <span *ngIf=\"actionType === ActionType.EDIT\" title=\"{{ 'Save' | translate }}\">\n        {{ 'Save' | translate }}\n      </span>\n      <span *ngIf=\"actionType === ActionType.DUPLICATE\" title=\"{{ 'Duplicate' | translate }}\">\n        {{ 'Duplicate' | translate }}\n      </span>\n    </span>\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ReportsService }, { type: ɵngcc2.BsModalRef }, { type: ɵngcc3.CronService }]; }, { emitter: [{
            type: Output
        }] }); })();
