// @c8y/ngx-components/esm2015/operations/bulk-operations-list/bulk-operation-details.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from 'ngx-bootstrap/tooltip';
import * as ɵngcc3 from '@angular/common';

function BulkOperationDetailsComponent_ng_template_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
    ɵngcc0.ɵɵelementStart(1, "span", 21);
    ɵngcc0.ɵɵtext(2, "second");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r0.bulkOperation.creationRamp, " ");
} }
function BulkOperationDetailsComponent_ng_template_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
    ɵngcc0.ɵɵelementStart(1, "span", 21);
    ɵngcc0.ɵɵtext(2, "seconds");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.bulkOperation.creationRamp, " ");
} }
const _c0 = function (a0) { return { bulkOperationsCountSuccessful: a0 }; };
function BulkOperationDetailsComponent_div_32_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelement(1, "i", 23);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountSuccessful }} successful ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c0, ctx_r2.bulkOperation.progress.successful));
} }
const _c1 = function (a0) { return { bulkOperationsCountFailed: a0 }; };
function BulkOperationDetailsComponent_div_33_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵelement(1, "i", 25);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountFailed }} failed ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c1, ctx_r3.bulkOperation.progress.failed));
} }
const _c2 = function (a0) { return { bulkOperationsCountPending: a0 }; };
function BulkOperationDetailsComponent_div_34_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "i", 27);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(3, " {{ bulkOperationsCountPending }} pending ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c2, ctx_r4.bulkOperation.progress.pending));
} }
function BulkOperationDetailsComponent_li_35_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 4);
    ɵngcc0.ɵɵelementStart(1, "label", 5);
    ɵngcc0.ɵɵtext(2, "Retried from");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 6);
    ɵngcc0.ɵɵelementStart(4, "button", 28);
    ɵngcc0.ɵɵlistener("click", function BulkOperationDetailsComponent_li_35_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r8); const ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.scrollToRetriedFromOperation(ctx_r7.bulkOperation.failedParentId); });
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r5.bulkOperation.failedParentId, " ");
} }
function BulkOperationDetailsComponent_li_36_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 4);
    ɵngcc0.ɵɵelementStart(1, "label", 14);
    ɵngcc0.ɵɵtext(2, "Description");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 29);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r6.bulkOperation.note);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r6.bulkOperation.note, " ");
} }
export class BulkOperationDetailsComponent {
    constructor() {
        this.allFilterIsSelected = false;
        this.readOnly = false;
        this.showFailedOperation = new EventEmitter();
    }
    scrollToRetriedFromOperation(failedOperationId) {
        this.showFailedOperation.emit(failedOperationId);
    }
}
BulkOperationDetailsComponent.ɵfac = function BulkOperationDetailsComponent_Factory(t) { return new (t || BulkOperationDetailsComponent)(); };
BulkOperationDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationDetailsComponent, selectors: [["c8y-bulk-operation-details"]], inputs: { readOnly: "readOnly", bulkOperation: "bulkOperation" }, outputs: { showFailedOperation: "showFailedOperation" }, decls: 43, vars: 22, consts: [[1, "row", "d-flex-md", "fit-h"], [1, "col-md-4", "flex-item-top"], ["translate", "", 1, "legend", "form-block"], [1, "list-unstyled", "small"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom"], ["translate", "", 1, "small", "m-0"], [1, "flex-item-right"], [1, "small", "m-0"], ["container", "body", 1, "text-primary", 3, "tooltip"], [3, "c8yIcon"], [1, "flex-item-right", 3, "ngPlural"], ["ngPluralCase", "=1"], ["ngPluralCase", "other"], [1, "flex-item-right", "text-uppercase"], ["translate", "", 1, "small", "flex-item-top"], [1, "flex-row", "flex-grow", "flex-item-right", "flex-end"], ["class", "icon-flex flex-item-left m-r-8", 4, "ngIf"], ["class", "icon-flex flex-item-left", 4, "ngIf"], ["class", "p-t-4 p-b-4 flex-row separator-bottom", 4, "ngIf"], [1, "col-md-8", "fit-h", "d-flex", "d-col"], [1, "inner-scroll", "p-16", "m-b-0", "text-pre-wrap"], ["translate", ""], [1, "icon-flex", "flex-item-left", "m-r-8"], ["c8yIcon", "check-circle", 1, "text-success", "m-r-4"], ["translate", "", 3, "translateParams"], ["c8yIcon", "warning", 1, "text-danger", "m-r-4"], [1, "icon-flex", "flex-item-left"], ["c8yIcon", "clock-o", 1, "m-r-4", "text-info"], [1, "btn", "btn-default", "btn-xs", 3, "click"], [1, "flex-item-right", "text-right", "m-l-8", "text-truncate", 3, "title"]], template: function BulkOperationDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtext(3, " Details ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "ul", 3);
        ɵngcc0.ɵɵelementStart(5, "li", 4);
        ɵngcc0.ɵɵelementStart(6, "label", 5);
        ɵngcc0.ɵɵtext(7, "Start date");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "span", 6);
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "c8yDate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "li", 4);
        ɵngcc0.ɵɵelementStart(12, "label", 7);
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelementStart(15, "a", 8);
        ɵngcc0.ɵɵpipe(16, "translate");
        ɵngcc0.ɵɵelement(17, "i", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "span", 10);
        ɵngcc0.ɵɵtemplate(19, BulkOperationDetailsComponent_ng_template_19_Template, 3, 1, "ng-template", 11);
        ɵngcc0.ɵɵtemplate(20, BulkOperationDetailsComponent_ng_template_20_Template, 3, 1, "ng-template", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "li", 4);
        ɵngcc0.ɵɵelementStart(22, "label", 5);
        ɵngcc0.ɵɵtext(23, "Status");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "span", 13);
        ɵngcc0.ɵɵtext(25);
        ɵngcc0.ɵɵpipe(26, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "li", 4);
        ɵngcc0.ɵɵelementStart(28, "label", 14);
        ɵngcc0.ɵɵtext(29, "Result");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(30, "div", 15);
        ɵngcc0.ɵɵelementStart(31, "div");
        ɵngcc0.ɵɵtemplate(32, BulkOperationDetailsComponent_div_32_Template, 4, 3, "div", 16);
        ɵngcc0.ɵɵtemplate(33, BulkOperationDetailsComponent_div_33_Template, 4, 3, "div", 16);
        ɵngcc0.ɵɵtemplate(34, BulkOperationDetailsComponent_div_34_Template, 4, 3, "div", 17);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(35, BulkOperationDetailsComponent_li_35_Template, 6, 1, "li", 18);
        ɵngcc0.ɵɵtemplate(36, BulkOperationDetailsComponent_li_36_Template, 5, 2, "li", 18);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(37, "div", 19);
        ɵngcc0.ɵɵelementStart(38, "div", 2);
        ɵngcc0.ɵɵtext(39, " Operation ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(40, "pre", 20);
        ɵngcc0.ɵɵtext(41);
        ɵngcc0.ɵɵpipe(42, "json");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(10, 12, ctx.bulkOperation.startDate));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(14, 14, "Delay"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("tooltip", ɵngcc0.ɵɵpipeBind1(16, 16, "Delay before sending the operation to the next device."));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "info-circle");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngPlural", ctx.bulkOperation.creationRamp);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(26, 18, ctx.bulkOperation.generalStatus.toString()), " ");
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵproperty("ngIf", ctx.bulkOperation.progress.successful > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.bulkOperation.progress.failed > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.bulkOperation.progress.pending > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.bulkOperation.failedParentId);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.bulkOperation.note);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(42, 20, ctx.bulkOperation.operationPrototype));
    } }, directives: [ɵngcc1.C8yTranslateDirective, ɵngcc2.TooltipDirective, ɵngcc1.IconDirective, ɵngcc3.NgPlural, ɵngcc3.NgPluralCase, ɵngcc3.NgIf], pipes: [ɵngcc1.DatePipe, ɵngcc1.C8yTranslatePipe, ɵngcc3.JsonPipe], encapsulation: 2 });
BulkOperationDetailsComponent.propDecorators = {
    bulkOperation: [{ type: Input }],
    readOnly: [{ type: Input }],
    showFailedOperation: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-bulk-operation-details',
                template: "<div class=\"row d-flex-md fit-h\">\n  <div class=\"col-md-4 flex-item-top\">\n    <div class=\"legend form-block\" translate>\n      Details\n    </div>\n    <ul class=\"list-unstyled small\">\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n        <label class=\"small m-0\" translate>Start date</label>\n        <span class=\"flex-item-right\">{{ bulkOperation.startDate | c8yDate }}</span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n        <label class=\"small m-0\">\n          {{ 'Delay' | translate }}\n          <a\n            class=\"text-primary\"\n            tooltip=\"{{ 'Delay before sending the operation to the next device.' | translate }}\"\n            container=\"body\"\n          >\n            <i [c8yIcon]=\"'info-circle'\"></i>\n          </a>\n        </label>\n        <span [ngPlural]=\"bulkOperation.creationRamp\" class=\"flex-item-right\">\n          <ng-template ngPluralCase=\"=1\">\n            {{ bulkOperation.creationRamp }} <span translate>second</span>\n          </ng-template>\n          <ng-template ngPluralCase=\"other\">\n            {{ bulkOperation.creationRamp }} <span translate>seconds</span>\n          </ng-template>\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n        <label class=\"small m-0\" translate>Status</label>\n        <span class=\"flex-item-right text-uppercase\">\n          {{ bulkOperation.generalStatus.toString() | translate }}\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n        <label class=\"small flex-item-top\" translate>Result</label>\n        <div class=\"flex-row flex-grow flex-item-right flex-end\">\n          <div>\n            <div *ngIf=\"bulkOperation.progress.successful > 0\" class=\"icon-flex flex-item-left m-r-8\">\n              <i c8yIcon=\"check-circle\" class=\"text-success m-r-4\"></i>\n              <span\n                ngNonBindable\n                translate\n                [translateParams]=\"{\n                  bulkOperationsCountSuccessful: bulkOperation.progress.successful\n                }\"\n              >\n                {{ bulkOperationsCountSuccessful }} successful\n              </span>\n            </div>\n            <div class=\"icon-flex flex-item-left m-r-8\" *ngIf=\"bulkOperation.progress.failed > 0\">\n              <i c8yIcon=\"warning\" class=\"text-danger m-r-4\"></i>\n              <span\n                ngNonBindable\n                translate\n                [translateParams]=\"{ bulkOperationsCountFailed: bulkOperation.progress.failed }\"\n              >\n                {{ bulkOperationsCountFailed }} failed\n              </span>\n            </div>\n            <div class=\"icon-flex flex-item-left\" *ngIf=\"bulkOperation.progress.pending > 0\">\n              <i c8yIcon=\"clock-o\" class=\"m-r-4 text-info\"></i>\n              <span\n                ngNonBindable\n                translate\n                [translateParams]=\"{ bulkOperationsCountPending: bulkOperation.progress.pending }\"\n              >\n                {{ bulkOperationsCountPending }} pending\n              </span>\n            </div>\n          </div>\n        </div>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"bulkOperation.failedParentId\">\n        <label class=\"small m-0\" translate>Retried from</label>\n        <span class=\"flex-item-right\">\n          <button\n            class=\"btn btn-default btn-xs\"\n            (click)=\"scrollToRetriedFromOperation(bulkOperation.failedParentId)\"\n          >\n            {{ bulkOperation.failedParentId }}\n          </button>\n        </span>\n      </li>\n      <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" *ngIf=\"bulkOperation.note\">\n        <label class=\"small flex-item-top\" translate>Description</label>\n        <div\n          title=\"{{ bulkOperation.note }}\"\n          class=\"flex-item-right text-right m-l-8 text-truncate\"\n        >\n          {{ bulkOperation.note }}\n        </div>\n      </li>\n    </ul>\n  </div>\n  <div class=\"col-md-8 fit-h d-flex d-col\">\n    <div class=\"legend form-block\" translate>\n      Operation\n    </div>\n    <pre class=\"inner-scroll p-16 m-b-0 text-pre-wrap\">{{\n      bulkOperation.operationPrototype | json\n    }}</pre>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { readOnly: [{
            type: Input
        }], showFailedOperation: [{
            type: Output
        }], bulkOperation: [{
            type: Input
        }] }); })();
