// @c8y/ngx-components/esm2015/operations/bulk-operations-list/type-list/bulk-operation-type-list.component.js
import { Component, EventEmitter, Output } from '@angular/core';
import { Router } from '@angular/router';
import { BulkOperationsService } from '@c8y/ngx-components/operations/bulk-operations-service';
import { BulkOperationModalsService } from '../modals/bulk-operation-modals.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components/operations/bulk-operations-service';
import * as ɵngcc3 from '../modals/bulk-operation-modals.service';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@c8y/ngx-components';

function BulkOperationTypeListComponent_button_1_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelement(1, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r1 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", item_r1.c8yIcon);
} }
function BulkOperationTypeListComponent_button_1_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelement(1, "i", 8);
    ɵngcc0.ɵɵelementEnd();
} }
function BulkOperationTypeListComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 2);
    ɵngcc0.ɵɵlistener("click", function BulkOperationTypeListComponent_button_1_Template_button_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r7); const item_r1 = restoredCtx.$implicit; const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.handleClick(item_r1); });
    ɵngcc0.ɵɵtemplate(1, BulkOperationTypeListComponent_button_1_div_1_Template, 2, 1, "div", 3);
    ɵngcc0.ɵɵtemplate(2, BulkOperationTypeListComponent_button_1_ng_template_2_Template, 2, 0, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(4, "div", 5);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r1 = ctx.$implicit;
    const _r3 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r1.c8yIcon)("ngIfElse", _r3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 4, item_r1.name));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 6, item_r1.name), " ");
} }
export class BulkOperationTypeListComponent {
    constructor(router, bulkOperationsService, bulkOperationModalsService) {
        this.router = router;
        this.bulkOperationsService = bulkOperationsService;
        this.bulkOperationModalsService = bulkOperationModalsService;
        this.onCustomType = new EventEmitter();
        this.bulkList = [];
    }
    ngOnInit() {
        this.bulkList = this.bulkOperationsService.getBulkTypes();
    }
    handleClick(item) {
        if (item.path) {
            this.bulkOperationModalsService.hideNewBulkOperationModal();
            this.router.navigateByUrl(item.path);
        }
        else if (item.component) {
            this.onCustomType.emit(item);
        }
    }
}
BulkOperationTypeListComponent.ɵfac = function BulkOperationTypeListComponent_Factory(t) { return new (t || BulkOperationTypeListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BulkOperationModalsService)); };
BulkOperationTypeListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationTypeListComponent, selectors: [["c8y-bulk-operation-type-list"]], outputs: { onCustomType: "onCustomType" }, decls: 2, vars: 1, consts: [[1, "list-group", "list-group-links"], ["type", "button", "class", "list-group-item flex-row", 3, "click", 4, "ngFor", "ngForOf"], ["type", "button", 1, "list-group-item", "flex-row", 3, "click"], ["class", "list-item-icon", 4, "ngIf", "ngIfElse"], ["icon", ""], [1, "list-item-body", "text-truncate", 3, "title"], [1, "list-item-icon"], [3, "c8yIcon"], ["c8yIcon", "cogs"]], template: function BulkOperationTypeListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, BulkOperationTypeListComponent_button_1_Template, 8, 8, "button", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.bulkList);
    } }, directives: [ɵngcc4.NgForOf, ɵngcc4.NgIf, ɵngcc5.IconDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
BulkOperationTypeListComponent.ctorParameters = () => [
    { type: Router },
    { type: BulkOperationsService },
    { type: BulkOperationModalsService }
];
BulkOperationTypeListComponent.propDecorators = {
    onCustomType: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationTypeListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-bulk-operation-type-list',
                template: "<div class=\"list-group list-group-links\">\n  <button\n    *ngFor=\"let item of bulkList\"\n    type=\"button\"\n    class=\"list-group-item flex-row\"\n    (click)=\"handleClick(item)\"\n  >\n    <div *ngIf=\"item.c8yIcon; else icon\" class=\"list-item-icon\">\n      <i [c8yIcon]=\"item.c8yIcon\"></i>\n    </div>\n    <ng-template #icon>\n      <div class=\"list-item-icon\">\n        <i c8yIcon=\"cogs\"></i>\n      </div>\n    </ng-template>\n    <div title=\"{{ item.name | translate }}\" class=\"list-item-body text-truncate\">\n      {{ item.name | translate }}\n    </div>\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.BulkOperationsService }, { type: ɵngcc3.BulkOperationModalsService }]; }, { onCustomType: [{
            type: Output
        }] }); })();
