// @c8y/ngx-components/esm2015/core/navigator/navigator-entry.component.js
import { Component, ViewChild, Input } from '@angular/core';
import { NavigatorNode } from './navigator-node';
import { NavigatorService } from './navigator.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './navigator.service';

const _c0 = ["iconTpl"];
function NavigatorItemComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c1 = ["*"];
export class NavigatorItemComponent {
    constructor(service) {
        this.service = service;
        this.priority = 0;
        this.nodeData = new NavigatorNode();
    }
    ngOnInit() {
        this.ngOnChanges();
        setTimeout(() => this.service.add(this.nodeData));
    }
    ngOnChanges() {
        this.nodeData.icon = this.icon;
        this.nodeData.label = this.label;
        this.nodeData.path = this.path;
        this.nodeData.priority = this.priority;
        this.nodeData.iconTemplate = !this.icon ? this.template : undefined;
        this.nodeData.icon = this.icon;
    }
    ngOnDestroy() {
        this.service.remove(this.nodeData);
    }
}
NavigatorItemComponent.ɵfac = function NavigatorItemComponent_Factory(t) { return new (t || NavigatorItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NavigatorService)); };
NavigatorItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NavigatorItemComponent, selectors: [["c8y-navigator-item"]], viewQuery: function NavigatorItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { priority: "priority", label: "label", icon: "icon", path: "path" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["iconTpl", ""]], template: function NavigatorItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, NavigatorItemComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
NavigatorItemComponent.ctorParameters = () => [
    { type: NavigatorService }
];
NavigatorItemComponent.propDecorators = {
    template: [{ type: ViewChild, args: ['iconTpl', { static: true },] }],
    label: [{ type: Input }],
    icon: [{ type: Input }],
    path: [{ type: Input }],
    priority: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavigatorItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-navigator-item',
                template: '<ng-template #iconTpl><ng-content></ng-content></ng-template>'
            }]
    }], function () { return [{ type: ɵngcc1.NavigatorService }]; }, { priority: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['iconTpl', { static: true }]
        }], label: [{
            type: Input
        }], icon: [{
            type: Input
        }], path: [{
            type: Input
        }] }); })();
