// @c8y/ngx-components/esm2015/core/navigator/navigator-outlet.component.js
import { Component, Input } from '@angular/core';
import { HeaderService } from '../header/header.service';
import { BehaviorSubject } from 'rxjs';
/**
 * The component shows all nodes that have been added to the navigator.
 * In addition, it allows you to add a header and a footer to the navigator.
 *
 * ## Example:
 * ```html
 * <c8y-navigator-outlet [nodes]="navigatorService.items$ | async" [open]="isOpen">
 *   <header class="navigator-slot-top"></header>
 *   <footer class="navigator-slot-bottom"></footer>
 * </c8y-navigator-outlet>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../header/header.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './navigator-node.component';

function NavigatorOutletComponent_nav_0_c8y_navigator_node_display_3_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-navigator-node-display", 4);
    ɵngcc0.ɵɵlistener("nodeClick", function NavigatorOutletComponent_nav_0_c8y_navigator_node_display_3_Template_c8y_navigator_node_display_nodeClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(2); return ctx_r3.collapseNavigator($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const node_r2 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("node", node_r2)("isRoot", true);
} }
const _c0 = function (a0) { return { open: a0 }; };
function NavigatorOutletComponent_nav_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "nav", 1);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementStart(2, "ul", 2);
    ɵngcc0.ɵɵtemplate(3, NavigatorOutletComponent_nav_0_c8y_navigator_node_display_3_Template, 1, 2, "c8y-navigator-node-display", 3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵprojection(4, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx_r0.open));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.nodes)("ngForTrackBy", ctx_r0.trackByPathOrLabel);
} }
const _c1 = [[["", 8, "navigator-slot-top"]], [["", 8, "navigator-slot-bottom"]]];
const _c2 = [".navigator-slot-top", ".navigator-slot-bottom"];
export class NavigatorOutletComponent {
    constructor(headerService) {
        this.headerService = headerService;
        /**
         * Determines whether the navigator menu should be open/closed.
         */
        this.open = true;
        /**
         * Determines whether the navigator pane should be rendered.
         */
        this.isShown$ = new BehaviorSubject(false);
    }
    /**
     * @ignore
     */
    ngOnChanges() {
        this.isShown$.next(this.nodes && this.nodes.length > 0);
    }
    /**
     * Closes the navigator menu after clicking an item in the menu, in case of insufficient screen width.
     */
    collapseNavigator(from) {
        if (from === 'link' && !this.headerService.largeWidth) {
            this.headerService.closeNavigator();
        }
    }
    /**
     * @ignore
     */
    trackByPathOrLabel(index, node) {
        return node.path ? node.path : node.label;
    }
}
NavigatorOutletComponent.ɵfac = function NavigatorOutletComponent_Factory(t) { return new (t || NavigatorOutletComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.HeaderService)); };
NavigatorOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NavigatorOutletComponent, selectors: [["c8y-navigator-outlet"], ["c8y-ui-navigator"]], inputs: { open: "open", nodes: "nodes" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], ngContentSelectors: _c2, decls: 2, vars: 3, consts: [["class", "navigator", 3, "ngClass", 4, "ngIf"], [1, "navigator", 3, "ngClass"], [1, "navigatorContent"], [3, "node", "isRoot", "nodeClick", 4, "ngFor", "ngForOf", "ngForTrackBy"], [3, "node", "isRoot", "nodeClick"]], template: function NavigatorOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵtemplate(0, NavigatorOutletComponent_nav_0_Template, 5, 5, "nav", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(1, 1, ctx.isShown$));
    } }, directives: [ɵngcc2.NgIf, ɵngcc2.NgClass, ɵngcc2.NgForOf, ɵngcc3.NavigatorNodeComponent], pipes: [ɵngcc2.AsyncPipe], encapsulation: 2 });
NavigatorOutletComponent.ctorParameters = () => [
    { type: HeaderService }
];
NavigatorOutletComponent.propDecorators = {
    nodes: [{ type: Input }],
    open: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavigatorOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-navigator-outlet, c8y-ui-navigator',
                template: "<nav class=\"navigator\" *ngIf=\"isShown$ | async\" [ngClass]=\"{ open: open }\">\n  <ng-content select=\".navigator-slot-top\"></ng-content>\n  <ul class=\"navigatorContent\">\n    <c8y-navigator-node-display (nodeClick)=\"collapseNavigator($event)\" [node]=\"node\" *ngFor=\"let node of nodes trackBy trackByPathOrLabel\" [isRoot]=\"true\"></c8y-navigator-node-display>\n  </ul>\n  <ng-content select=\".navigator-slot-bottom\"></ng-content>\n</nav>\n"
            }]
    }], function () { return [{ type: ɵngcc1.HeaderService }]; }, { open: [{
            type: Input
        }], nodes: [{
            type: Input
        }] }); })();
