// @c8y/ngx-components/esm2015/operations/status-filter/status-filter.module.js
import { NgModule } from '@angular/core';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CoreModule } from '@c8y/ngx-components';
import { StatusFilterComponent } from './status-filter.component';
/**
 * This module provides a component for selecting status filter.
 */
import * as ɵngcc0 from '@angular/core';
export class StatusFilterModule {
}
StatusFilterModule.ɵfac = function StatusFilterModule_Factory(t) { return new (t || StatusFilterModule)(); };
StatusFilterModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StatusFilterModule });
StatusFilterModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, TooltipModule, BsDropdownModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StatusFilterModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, TooltipModule, BsDropdownModule],
                declarations: [StatusFilterComponent],
                entryComponents: [StatusFilterComponent],
                exports: [StatusFilterComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StatusFilterModule, { declarations: function () { return [StatusFilterComponent]; }, imports: function () { return [CoreModule, TooltipModule, BsDropdownModule]; }, exports: function () { return [StatusFilterComponent]; } }); })();
