// @c8y/ngx-components/esm2015/ecosystem/applications/duplicate-application/duplicate-application-list/duplicate-application-list.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { WizardComponent } from '@c8y/ngx-components';
import { EcosystemService } from '../../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';

function DuplicateApplicationListComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 4);
    ɵngcc0.ɵɵlistener("click", function DuplicateApplicationListComponent_button_1_Template_button_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r3); const app_r1 = restoredCtx.$implicit; const ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.select(app_r1); });
    ɵngcc0.ɵɵelement(1, "c8y-app-icon", 5);
    ɵngcc0.ɵɵelement(2, "span", 6);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵpipe(4, "humanizeAppName");
    ɵngcc0.ɵɵelementStart(5, "span", 7);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const app_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("app", app_r1)("contextPath", app_r1.contextPath)("name", app_r1.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerText", ɵngcc0.ɵɵpipeBind1(3, 6, ɵngcc0.ɵɵpipeBind1(4, 8, app_r1)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r0.ecosystemService.getAppState(app_r1).class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 10, ctx_r0.ecosystemService.getAppState(app_r1).label));
} }
export class DuplicateApplicationListComponent {
    constructor(ecosystemService, wizardComponent) {
        this.ecosystemService = ecosystemService;
        this.wizardComponent = wizardComponent;
        this.onSelectedApp = new EventEmitter();
    }
    select(selectedApp) {
        this.onSelectedApp.emit(selectedApp);
        this.stepper.next();
    }
    cancel() {
        this.wizardComponent.close();
    }
    back() {
        this.wizardComponent.reset();
    }
}
DuplicateApplicationListComponent.ɵfac = function DuplicateApplicationListComponent_Factory(t) { return new (t || DuplicateApplicationListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.WizardComponent)); };
DuplicateApplicationListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DuplicateApplicationListComponent, selectors: [["c8y-duplicate-application-list"]], inputs: { stepper: "stepper", existingApps: "existingApps" }, outputs: { onSelectedApp: "onSelectedApp" }, decls: 9, vars: 7, consts: [[1, "c8y-wizard-list-nav"], ["class", "list-group-item text-truncate p-t-8 p-b-8", "type", "button", 3, "click", 4, "ngFor", "ngForOf"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "list-group-item", "text-truncate", "p-t-8", "p-b-8", 3, "click"], [1, "list-group-icon", 3, "app", "contextPath", "name"], [1, "pointer", 3, "innerText"], [1, "label", "m-l-16", 3, "ngClass"]], template: function DuplicateApplicationListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, DuplicateApplicationListComponent_button_1_Template, 8, 12, "button", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(2, "c8y-wizard-footer");
        ɵngcc0.ɵɵelementStart(3, "button", 2);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationListComponent_Template_button_click_3_listener() { return ctx.back(); });
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵtext(5, " Back ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "button", 3);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationListComponent_Template_button_click_6_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵtext(8, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.existingApps);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 3, "Back"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 5, "Cancel"));
    } }, directives: [ɵngcc3.NgForOf, ɵngcc2.WizardFooterComponent, ɵngcc2.C8yTranslateDirective, ɵngcc2.AppIconComponent, ɵngcc3.NgClass], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.AsyncPipe, ɵngcc2.HumanizeAppNamePipe], encapsulation: 2 });
DuplicateApplicationListComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: WizardComponent }
];
DuplicateApplicationListComponent.propDecorators = {
    stepper: [{ type: Input }],
    existingApps: [{ type: Input }],
    onSelectedApp: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DuplicateApplicationListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-duplicate-application-list',
                template: "<div class=\"c8y-wizard-list-nav\">\n  <button\n    (click)=\"select(app)\"\n    *ngFor=\"let app of existingApps\"\n    class=\"list-group-item text-truncate p-t-8 p-b-8\"\n    type=\"button\"\n  >\n    <c8y-app-icon\n      [app]=\"app\"\n      [contextPath]=\"app.contextPath\"\n      [name]=\"app.name\"\n      class=\"list-group-icon\"\n    ></c8y-app-icon>\n    <span [innerText]=\"app | humanizeAppName | async\" class=\"pointer\"></span>\n    <span [ngClass]=\"ecosystemService.getAppState(app).class\" class=\"label m-l-16\">{{\n      ecosystemService.getAppState(app).label | translate\n    }}</span>\n  </button>\n</div>\n<c8y-wizard-footer>\n  <button\n    (click)=\"back()\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Back' | translate }}\"\n    translate\n    type=\"button\"\n  >\n    Back\n  </button>\n  <button (click)=\"cancel()\" class=\"btn btn-default\" title=\"{{ 'Cancel' | translate }}\" translate>\n    Cancel\n  </button>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.WizardComponent }]; }, { onSelectedApp: [{
            type: Output
        }], stepper: [{
            type: Input
        }], existingApps: [{
            type: Input
        }] }); })();
