// @c8y/ngx-components/esm2015/default-subscriptions/default-subscriptions.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE } from '@c8y/ngx-components';
import { DefaultSubscriptionsComponent } from './default-subscriptions.component';
import { DefaultSubscriptionsService } from './default-subscriptions.service';
import { DefaultSubscriptionsGuard } from './default-subscriptions.guard';
import { DefaultSubscriptionsNavigationFactory } from './default-subscriptions-navigation.factory';
import * as ɵngcc0 from '@angular/core';
const ɵ0 = [
    {
        path: 'default-subscriptions',
        component: DefaultSubscriptionsComponent,
        canActivate: [DefaultSubscriptionsGuard]
    }
];
export class DefaultSubscriptionsModule {
}
DefaultSubscriptionsModule.ɵfac = function DefaultSubscriptionsModule_Factory(t) { return new (t || DefaultSubscriptionsModule)(); };
DefaultSubscriptionsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DefaultSubscriptionsModule });
DefaultSubscriptionsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        DefaultSubscriptionsService,
        DefaultSubscriptionsGuard,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: DefaultSubscriptionsNavigationFactory,
            multi: true
        },
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CoreModule, ReactiveFormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DefaultSubscriptionsModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, ReactiveFormsModule],
                declarations: [DefaultSubscriptionsComponent],
                entryComponents: [DefaultSubscriptionsComponent],
                exports: [DefaultSubscriptionsComponent],
                providers: [
                    DefaultSubscriptionsService,
                    DefaultSubscriptionsGuard,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: DefaultSubscriptionsNavigationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DefaultSubscriptionsModule, { declarations: function () { return [DefaultSubscriptionsComponent]; }, imports: function () { return [CoreModule, ReactiveFormsModule]; }, exports: function () { return [DefaultSubscriptionsComponent]; } }); })();
export { ɵ0 };
