// @c8y/ngx-components/esm2015/default-subscriptions/default-subscriptions-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { DefaultSubscriptionsGuard } from './default-subscriptions.guard';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './default-subscriptions.guard';
export class DefaultSubscriptionsNavigationFactory {
    constructor(defaultSubscriptionsGuard) {
        this.defaultSubscriptionsGuard = defaultSubscriptionsGuard;
        this.navNode = new NavigatorNode({
            label: gettext('Default subscriptions'),
            path: '/default-subscriptions',
            icon: 'c8y-c8y-data',
            parent: gettext('Ecosystem'),
            priority: 100
        });
    }
    /** Returns the navigation node if the feature is active. */
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (yield this.defaultSubscriptionsGuard.canActivate()) {
                return this.navNode;
            }
        });
    }
}
DefaultSubscriptionsNavigationFactory.ɵfac = function DefaultSubscriptionsNavigationFactory_Factory(t) { return new (t || DefaultSubscriptionsNavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.DefaultSubscriptionsGuard)); };
DefaultSubscriptionsNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DefaultSubscriptionsNavigationFactory, factory: DefaultSubscriptionsNavigationFactory.ɵfac });
DefaultSubscriptionsNavigationFactory.ctorParameters = () => [
    { type: DefaultSubscriptionsGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DefaultSubscriptionsNavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DefaultSubscriptionsGuard }]; }, null); })();
