// @c8y/ngx-components/fesm2015/c8y-ngx-components-child-devices.js
import { __awaiter } from 'tslib';
import { Component, Injectable, NgModule } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { ManagedObjectRealtimeService, ViewContext, gettext, CoreModule, HOOK_ONCE_ROUTE } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { InventoryService } from '@c8y/client';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/ngx-components/device-grid';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@c8y/client';
class ChildDevicesComponent {
    constructor(route) {
        this.route = route;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.getAllDevices();
        });
    }
    getAllDevices() {
        return __awaiter(this, void 0, void 0, function* () {
            this.id = this.route.snapshot.parent.data.contextData;
        });
    }
}
ChildDevicesComponent.ɵfac = function ChildDevicesComponent_Factory(t) { return new (t || ChildDevicesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute)); };
ChildDevicesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ChildDevicesComponent, selectors: [["c8y-child-devices"]], features: [ɵngcc0.ɵɵProvidersFeature([ManagedObjectRealtimeService])], decls: 12, vars: 11, consts: [[1, "flex-grow", "col-xs-12", "no-gutter", 3, "title", "childDeviceGrid", "parentDeviceId"], [1, "c8y-empty-state"], ["c8yIcon", "search"]], template: function ChildDevicesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-device-grid", 0);
        ɵngcc0.ɵɵpipe(1, "translate");
        ɵngcc0.ɵɵelementStart(2, "div", 1);
        ɵngcc0.ɵɵelement(3, "h1", 2);
        ɵngcc0.ɵɵelementStart(4, "div");
        ɵngcc0.ɵɵelementStart(5, "p");
        ɵngcc0.ɵɵelementStart(6, "strong");
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "small");
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(1, 5, "Child devices"))("childDeviceGrid", true)("parentDeviceId", ctx.id);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 7, "No matching devices."));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 9, "Refine your search terms"));
    } }, directives: [ɵngcc2.DeviceGridComponent, ɵngcc3.IconDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ChildDevicesComponent.ctorParameters = () => [
    { type: ActivatedRoute }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChildDevicesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-child-devices',
                template: "<c8y-device-grid\n  [title]=\"'Child devices' | translate\"\n  class=\"flex-grow col-xs-12 no-gutter\"\n  [childDeviceGrid]=\"true\"\n  [parentDeviceId]=\"id\"\n>\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div>\n      <p>\n        <strong>{{ 'No matching devices.' | translate }}</strong>\n      </p>\n      <small>{{ 'Refine your search terms' | translate }}</small>\n    </div>\n  </div>\n</c8y-device-grid>\n",
                providers: [ManagedObjectRealtimeService]
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }]; }, null); })();

class ChildDevicesTabGuard {
    constructor(inventory) {
        this.inventory = inventory;
    }
    canActivate(route) {
        return __awaiter(this, void 0, void 0, function* () {
            const contextData = route.data.contextData;
            const filter = {
                pageSize: 1
            };
            if (!contextData) {
                return true;
            }
            const device = yield this.inventory.childDevicesList(contextData.id, filter);
            const hasChildren = device.data.length;
            const isVendme = route.data.contextData.type === 'com_nsn_startups_vendme_VendingMachine';
            return hasChildren && !isVendme;
        });
    }
}
ChildDevicesTabGuard.ɵfac = function ChildDevicesTabGuard_Factory(t) { return new (t || ChildDevicesTabGuard)(ɵngcc0.ɵɵinject(ɵngcc4.InventoryService)); };
ChildDevicesTabGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ChildDevicesTabGuard, factory: ChildDevicesTabGuard.ɵfac });
ChildDevicesTabGuard.ctorParameters = () => [
    { type: InventoryService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChildDevicesTabGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc4.InventoryService }]; }, null); })();

const ɵ0 = [
    {
        context: ViewContext.Device,
        path: 'child-devices',
        priority: 1000,
        icon: 'code-fork',
        label: gettext('Child devices'),
        component: ChildDevicesComponent,
        canActivate: [ChildDevicesTabGuard]
    }
];
class ChildDevicesModule {
}
ChildDevicesModule.ɵfac = function ChildDevicesModule_Factory(t) { return new (t || ChildDevicesModule)(); };
ChildDevicesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ChildDevicesModule });
ChildDevicesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ChildDevicesTabGuard,
        {
            provide: HOOK_ONCE_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[CoreModule, DeviceGridModule, PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChildDevicesModule, [{
        type: NgModule,
        args: [{
                declarations: [ChildDevicesComponent],
                imports: [CoreModule, DeviceGridModule, PopoverModule],
                entryComponents: [ChildDevicesComponent],
                providers: [
                    ChildDevicesTabGuard,
                    {
                        provide: HOOK_ONCE_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ChildDevicesModule, { declarations: function () { return [ChildDevicesComponent]; }, imports: function () { return [CoreModule, DeviceGridModule, PopoverModule]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { ChildDevicesComponent, ChildDevicesModule, ɵ0, ChildDevicesTabGuard as ɵa };
