// @c8y/ngx-components/fesm2015/c8y-ngx-components-trusted-certificates.js
import { Component, Injectable, NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { gettext, AlertService, Status, ModalService, NavigatorNode, CoreModule, CommonModule, HOOK_NAVIGATOR_NODES } from '@c8y/ngx-components';
import { __awaiter } from 'tslib';
import { BsModalRef, BsModalService } from 'ngx-bootstrap/modal';
import { TrustedCertificateService } from '@c8y/client';
import { TranslateService } from '@ngx-translate/core';
import { tap, switchMap } from 'rxjs/operators';
import { BehaviorSubject, pipe } from 'rxjs';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { ReactiveFormsModule } from '@angular/forms';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from 'ngx-bootstrap/popover';
import * as ɵngcc6 from 'ngx-bootstrap/buttons';
import * as ɵngcc7 from '@ngx-translate/core';
import * as ɵngcc8 from '@angular/common';
import * as ɵngcc9 from '@angular/router';
import * as ɵngcc10 from 'ngx-bootstrap/dropdown';

function TrustedCertificatesComponent_div_16_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelement(1, "h1", 9);
    ɵngcc0.ɵɵelementStart(2, "h3", 10);
    ɵngcc0.ɵɵtext(3, "No trusted certificates to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 10);
    ɵngcc0.ɵɵtext(5, "Add your first certificate by clicking below.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "p");
    ɵngcc0.ɵɵelementStart(7, "button", 11);
    ɵngcc0.ɵɵlistener("click", function TrustedCertificatesComponent_div_16_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r4); const ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.addTrustedCertificate(); });
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(8, 2, "Add trusted certificate"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 4, "Add trusted certificate"), " ");
} }
function TrustedCertificatesComponent_div_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "div", 13);
    ɵngcc0.ɵɵelementStart(2, "div", 14);
    ɵngcc0.ɵɵelement(3, "i", 15);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 16);
    ɵngcc0.ɵɵelementStart(5, "div", 17);
    ɵngcc0.ɵɵelementStart(6, "div", 18);
    ɵngcc0.ɵɵelementStart(7, "span", 19);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵtext(9);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 20);
    ɵngcc0.ɵɵelementStart(12, "span", 19);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 20);
    ɵngcc0.ɵɵelementStart(17, "span", 19);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵtext(19);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(21, "div", 18);
    ɵngcc0.ɵɵelementStart(22, "span", 19);
    ɵngcc0.ɵɵpipe(23, "translate");
    ɵngcc0.ɵɵtext(24);
    ɵngcc0.ɵɵpipe(25, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(26, "div", 20);
    ɵngcc0.ɵɵelementStart(27, "span", 19);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵtext(29);
    ɵngcc0.ɵɵpipe(30, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(31, "div", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(8, 10, "Certificate"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(10, 12, "Certificate"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(13, 14, "Status"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 16, "Status"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 18, "Algorithm"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(20, 20, "Algorithm"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(23, 22, "Expiration date"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(25, 24, "Expiration date"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate1("title", " ", ɵngcc0.ɵɵpipeBind1(28, 26, "Auto registration"), "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(30, 28, "Auto registration"), " ");
} }
function TrustedCertificatesComponent_c8y_li_22_i_32_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 52);
} }
function TrustedCertificatesComponent_c8y_li_22_i_33_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 53);
} }
function TrustedCertificatesComponent_c8y_li_22_span_42_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 54);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 3, "Auto registration"), ":  ", ɵngcc0.ɵɵpipeBind1(2, 5, "Enabled`auto registration`"), "");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, "Enabled`auto registration`"), " ");
} }
function TrustedCertificatesComponent_c8y_li_22_span_43_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 54);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 3, "Auto registration"), ":  ", ɵngcc0.ɵɵpipeBind1(2, 5, "Disabled`auto registration`"), "");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, "Disabled`auto registration`"), " ");
} }
function TrustedCertificatesComponent_c8y_li_22_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 17);
    ɵngcc0.ɵɵelementStart(4, "div", 18);
    ɵngcc0.ɵɵelementStart(5, "span", 19);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 20);
    ɵngcc0.ɵɵelement(8, "div", 22);
    ɵngcc0.ɵɵelementStart(9, "button", 23);
    ɵngcc0.ɵɵlistener("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_button_ngModelChange_9_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; return trustedCertificate_r5.status = $event; })("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_button_ngModelChange_9_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; const ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.updateCertificate(trustedCertificate_r5, { status: $event }); });
    ɵngcc0.ɵɵelementStart(10, "small", 24);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "small", 24);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(18, "div", 22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(19, "div", 20);
    ɵngcc0.ɵɵelementStart(20, "div", 19);
    ɵngcc0.ɵɵpipe(21, "translate");
    ɵngcc0.ɵɵelementStart(22, "span", 25);
    ɵngcc0.ɵɵtext(23, "Algorithm");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(24);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 18);
    ɵngcc0.ɵɵelementStart(26, "div", 19);
    ɵngcc0.ɵɵpipe(27, "translate");
    ɵngcc0.ɵɵpipe(28, "c8yDate");
    ɵngcc0.ɵɵelementStart(29, "span", 25);
    ɵngcc0.ɵɵtext(30, "Expiration date");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(31, "small", 26);
    ɵngcc0.ɵɵtemplate(32, TrustedCertificatesComponent_c8y_li_22_i_32_Template, 1, 0, "i", 27);
    ɵngcc0.ɵɵtemplate(33, TrustedCertificatesComponent_c8y_li_22_i_33_Template, 1, 0, "i", 28);
    ɵngcc0.ɵɵelementStart(34, "span");
    ɵngcc0.ɵɵtext(35);
    ɵngcc0.ɵɵpipe(36, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(37, "div", 20);
    ɵngcc0.ɵɵelementStart(38, "div", 29);
    ɵngcc0.ɵɵelementStart(39, "span", 30);
    ɵngcc0.ɵɵtext(40);
    ɵngcc0.ɵɵpipe(41, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(42, TrustedCertificatesComponent_c8y_li_22_span_42_Template, 5, 9, "span", 31);
    ɵngcc0.ɵɵtemplate(43, TrustedCertificatesComponent_c8y_li_22_span_43_Template, 5, 9, "span", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(44, "c8y-li-action", 32);
    ɵngcc0.ɵɵlistener("click", function TrustedCertificatesComponent_c8y_li_22_Template_c8y_li_action_click_44_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; const i_r6 = restoredCtx.index; const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.deleteTrustedCertificate(trustedCertificate_r5, i_r6); });
    ɵngcc0.ɵɵpipe(45, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(46, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(47, "div", 33);
    ɵngcc0.ɵɵelementStart(48, "div", 34);
    ɵngcc0.ɵɵelementStart(49, "div", 35);
    ɵngcc0.ɵɵelementStart(50, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(51, "label", 36);
    ɵngcc0.ɵɵtext(52);
    ɵngcc0.ɵɵpipe(53, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(54, "div", 37);
    ɵngcc0.ɵɵelementStart(55, "input", 38);
    ɵngcc0.ɵɵlistener("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_input_ngModelChange_55_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; return trustedCertificate_r5.name = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(56, "span");
    ɵngcc0.ɵɵelementStart(57, "div", 39);
    ɵngcc0.ɵɵelementStart(58, "button", 40);
    ɵngcc0.ɵɵlistener("click", function TrustedCertificatesComponent_c8y_li_22_Template_button_click_58_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.updateCertificate(trustedCertificate_r5, { name: trustedCertificate_r5.name }); });
    ɵngcc0.ɵɵpipe(59, "translate");
    ɵngcc0.ɵɵtext(60);
    ɵngcc0.ɵɵpipe(61, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(62, "div", 41);
    ɵngcc0.ɵɵtext(63, "Additional properties");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(64, "ul", 42);
    ɵngcc0.ɵɵelementStart(65, "li", 43);
    ɵngcc0.ɵɵelementStart(66, "label", 44);
    ɵngcc0.ɵɵtext(67);
    ɵngcc0.ɵɵpipe(68, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(69, "span", 45);
    ɵngcc0.ɵɵtext(70);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(71, "li", 43);
    ɵngcc0.ɵɵelementStart(72, "label", 44);
    ɵngcc0.ɵɵtext(73);
    ɵngcc0.ɵɵpipe(74, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(75, "span", 45);
    ɵngcc0.ɵɵtext(76);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(77, "li", 43);
    ɵngcc0.ɵɵelementStart(78, "label", 44);
    ɵngcc0.ɵɵtext(79);
    ɵngcc0.ɵɵpipe(80, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(81, "span", 45);
    ɵngcc0.ɵɵtext(82);
    ɵngcc0.ɵɵpipe(83, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(84, "li", 43);
    ɵngcc0.ɵɵelementStart(85, "label", 44);
    ɵngcc0.ɵɵtext(86);
    ɵngcc0.ɵɵpipe(87, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(88, "span", 45);
    ɵngcc0.ɵɵtext(89);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(90, "li", 43);
    ɵngcc0.ɵɵelementStart(91, "label", 44);
    ɵngcc0.ɵɵtext(92);
    ɵngcc0.ɵɵpipe(93, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(94, "span", 45);
    ɵngcc0.ɵɵtext(95);
    ɵngcc0.ɵɵpipe(96, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(97, "li", 43);
    ɵngcc0.ɵɵelementStart(98, "label", 44);
    ɵngcc0.ɵɵtext(99);
    ɵngcc0.ɵɵpipe(100, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(101, "span", 45);
    ɵngcc0.ɵɵtext(102);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(103, "li", 43);
    ɵngcc0.ɵɵelementStart(104, "label", 44);
    ɵngcc0.ɵɵtext(105);
    ɵngcc0.ɵɵpipe(106, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(107, "span", 45);
    ɵngcc0.ɵɵtext(108);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(109, "div", 46);
    ɵngcc0.ɵɵelementStart(110, "div", 47);
    ɵngcc0.ɵɵelementStart(111, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(112, "label", 48);
    ɵngcc0.ɵɵtext(113, " Certificate ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(114, "textarea", 49);
    ɵngcc0.ɵɵlistener("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_textarea_ngModelChange_114_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; return trustedCertificate_r5.certInPemFormat = $event; });
    ɵngcc0.ɵɵtext(115, "              ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(116, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(117, "label", 50);
    ɵngcc0.ɵɵpipe(118, "translate");
    ɵngcc0.ɵɵelementStart(119, "input", 51);
    ɵngcc0.ɵɵlistener("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_input_ngModelChange_119_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; return trustedCertificate_r5.autoRegistrationEnabled = $event; })("ngModelChange", function TrustedCertificatesComponent_c8y_li_22_Template_input_ngModelChange_119_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const trustedCertificate_r5 = restoredCtx.$implicit; const ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.updateCertificate(trustedCertificate_r5, { autoRegistrationEnabled: $event }); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(120, "span");
    ɵngcc0.ɵɵelementStart(121, "span");
    ɵngcc0.ɵɵtext(122);
    ɵngcc0.ɵɵpipe(123, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const trustedCertificate_r5 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", trustedCertificate_r5.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.name, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngModel", trustedCertificate_r5.status);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 45, "Disabled`trusted certificate status`"));
    ɵngcc0.ɵɵproperty("hidden", trustedCertificate_r5.status !== "DISABLED");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 47, "Disabled`trusted certificate status`"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 49, "Enabled`trusted certificate status`"));
    ɵngcc0.ɵɵproperty("hidden", trustedCertificate_r5.status !== "ENABLED");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 51, "Enabled`trusted certificate status`"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(21, 53, "Algorithm"), ": ", trustedCertificate_r5.algorithmName, "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.algorithmName, " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(27, 55, "Expiration date"), ": ", ɵngcc0.ɵɵpipeBind1(28, 57, trustedCertificate_r5.notAfter), "");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.highlightDependingOnExpirationStatus(trustedCertificate_r5));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.highlightDependingOnExpirationStatus(trustedCertificate_r5));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.highlightDependingOnExpirationStatus(trustedCertificate_r5));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(36, 59, trustedCertificate_r5.notAfter));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(41, 61, "Auto registration"), "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", trustedCertificate_r5.autoRegistrationEnabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !trustedCertificate_r5.autoRegistrationEnabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(45, 63, "Delete"));
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(53, 65, "Certificate name"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngModel", trustedCertificate_r5.name);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(59, 67, "Update certificate name"));
    ɵngcc0.ɵɵproperty("disabled", !trustedCertificate_r5.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(61, 69, "Save"), " ");
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(68, 71, "Algorithm"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.algorithmName, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(74, 73, "Version"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.version, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(80, 75, "Valid from"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(83, 77, trustedCertificate_r5.notBefore), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(87, 79, "Issuer"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.issuer, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(93, 81, "Expiration date"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(96, 83, trustedCertificate_r5.notAfter), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(100, 85, "Serial number"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.serialNumber, " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(106, 87, "Subject`of a certificate`"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", trustedCertificate_r5.subject, " ");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("ngModel", trustedCertificate_r5.certInPemFormat);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(118, 89, "Auto registration"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", trustedCertificate_r5.autoRegistrationEnabled);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(123, 91, "Auto registration"), "");
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
const _c1 = function (a0) { return { "dd-low": a0 }; };
class AddTrustedCertificateComponent {
    constructor(modal, trustedCertificateService, alertService) {
        this.modal = modal;
        this.trustedCertificateService = trustedCertificateService;
        this.alertService = alertService;
        this.maxAllowedFiles = 1;
        this.AUTO_REGISTRATION_POPOVER = gettext('All devices with credentials signed by this certificate will be able to communicate with the platform without a prior registration.');
        this.trustedCertificate = {
            status: 'DISABLED'
        };
        this.fileIsUploaded = false;
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    uploadFile(dropped) {
        this.fileIsUploaded = false;
        if (dropped) {
            const reader = new FileReader();
            reader.onload = () => {
                this.fileIsUploaded = true;
                this.trustedCertificate.certInPemFormat = reader.result;
            };
            reader.readAsText(dropped[0].file);
        }
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.trustedCertificateService.create(this.trustedCertificate);
                this.alertService.success(gettext('Certificate saved.'));
                this._save();
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    close() {
        this._cancel();
        this.modal.hide();
    }
}
AddTrustedCertificateComponent.ɵfac = function AddTrustedCertificateComponent_Factory(t) { return new (t || AddTrustedCertificateComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TrustedCertificateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
AddTrustedCertificateComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddTrustedCertificateComponent, selectors: [["c8y-add-trusted-certificate"]], decls: 49, vars: 50, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], ["c8yIcon", "certificate"], ["translate", ""], [1, "modal-inner-scroll"], [1, "modal-body"], ["addTrustedCertificateForm", "ngForm"], ["translate", "", "for", "certificateName"], ["id", "certificateName", "type", "text", "autocomplete", "off", "name", "certificateName", "required", "", 1, "form-control", 3, "ngModel", "placeholder", "ngModelChange"], ["translate", "", "for", "certificate"], [1, "drop-area-sm", 3, "loadingMessage", "title", "maxAllowedFiles", "dropped"], [1, "c8y-checkbox", 3, "title"], ["id", "autoRegistration", "type", "checkbox", "name", "autoRegistration", 3, "ngModel", "ngModelChange"], ["triggers", "focus", 1, "btn", "btn-clean", 3, "popover"], [3, "c8yIcon"], ["type", "button", "name", "certificateStatus", "btnCheckbox", "", "btnCheckboxTrue", "ENABLED", "btnCheckboxFalse", "DISABLED", 1, "btn", 3, "ngModel", "ngModelChange"], [3, "title", "hidden"], [1, "modal-footer"], [1, "btn", "btn-default", 3, "title", "click"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"]], template: function AddTrustedCertificateComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, "Add trusted certificate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "form", null, 6);
        ɵngcc0.ɵɵelementStart(9, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(10, "label", 7);
        ɵngcc0.ɵɵtext(11, "Certificate name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "input", 8);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddTrustedCertificateComponent_Template_input_ngModelChange_12_listener($event) { return ctx.trustedCertificate.name = $event; });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(15, "label", 9);
        ɵngcc0.ɵɵtext(16, "Certificate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "c8y-drop-area", 10);
        ɵngcc0.ɵɵlistener("dropped", function AddTrustedCertificateComponent_Template_c8y_drop_area_dropped_17_listener($event) { return ctx.uploadFile($event); });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(20, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(21, "label", 11);
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelementStart(23, "input", 12);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddTrustedCertificateComponent_Template_input_ngModelChange_23_listener($event) { return ctx.trustedCertificate.autoRegistrationEnabled = $event; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(24, "span");
        ɵngcc0.ɵɵelementStart(25, "span");
        ɵngcc0.ɵɵtext(26);
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(28, "button", 13);
        ɵngcc0.ɵɵpipe(29, "translate");
        ɵngcc0.ɵɵelement(30, "i", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "button", 15);
        ɵngcc0.ɵɵlistener("ngModelChange", function AddTrustedCertificateComponent_Template_button_ngModelChange_31_listener($event) { return ctx.trustedCertificate.status = $event; });
        ɵngcc0.ɵɵelementStart(32, "span", 16);
        ɵngcc0.ɵɵpipe(33, "translate");
        ɵngcc0.ɵɵtext(34);
        ɵngcc0.ɵɵpipe(35, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "span", 16);
        ɵngcc0.ɵɵpipe(37, "translate");
        ɵngcc0.ɵɵtext(38);
        ɵngcc0.ɵɵpipe(39, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(40, "div", 17);
        ɵngcc0.ɵɵelementStart(41, "button", 18);
        ɵngcc0.ɵɵlistener("click", function AddTrustedCertificateComponent_Template_button_click_41_listener() { return ctx.close(); });
        ɵngcc0.ɵɵpipe(42, "translate");
        ɵngcc0.ɵɵtext(43);
        ɵngcc0.ɵɵpipe(44, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(45, "button", 19);
        ɵngcc0.ɵɵlistener("click", function AddTrustedCertificateComponent_Template_button_click_45_listener() { return ctx.save(); });
        ɵngcc0.ɵɵpipe(46, "translate");
        ɵngcc0.ɵɵtext(47);
        ɵngcc0.ɵɵpipe(48, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r0 = ɵngcc0.ɵɵreference(8);
        ɵngcc0.ɵɵadvance(12);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(13, 22, "e.g. My certificate"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.trustedCertificate.name);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("loadingMessage", ɵngcc0.ɵɵpipeBind1(18, 24, "Importing, please wait."))("title", ɵngcc0.ɵɵpipeBind1(19, 26, "Paste the certificate in PEM format."))("maxAllowedFiles", ctx.maxAllowedFiles);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(22, 28, "Auto registration"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngModel", ctx.trustedCertificate.autoRegistrationEnabled);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(27, 30, "Auto registration"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(29, 32, ctx.AUTO_REGISTRATION_POPOVER));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngModel", ctx.trustedCertificate.status);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(33, 34, "Disabled`trusted certificate status`"));
        ɵngcc0.ɵɵproperty("hidden", ctx.trustedCertificate.status !== "DISABLED");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(35, 36, "Disabled`trusted certificate status`"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(37, 38, "Enabled`trusted certificate status`"));
        ɵngcc0.ɵɵproperty("hidden", ctx.trustedCertificate.status !== "ENABLED");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(39, 40, "Enabled`trusted certificate status`"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(42, 42, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(44, 44, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(46, 46, "Add certificate"));
        ɵngcc0.ɵɵproperty("disabled", _r0.form.invalid || _r0.form.pristine || !ctx.fileIsUploaded);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(48, 48, "Add certificate"), " ");
    } }, directives: [ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc4.ɵNgNoValidate, ɵngcc4.NgControlStatusGroup, ɵngcc4.NgForm, ɵngcc3.FormGroupComponent, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc3.DropAreaComponent, ɵngcc4.CheckboxControlValueAccessor, ɵngcc5.PopoverDirective, ɵngcc6.ButtonCheckboxDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
AddTrustedCertificateComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: TrustedCertificateService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddTrustedCertificateComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-trusted-certificate',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i c8yIcon=\"certificate\"></i>\n    <h4 translate>Add trusted certificate</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body\">\n      <form #addTrustedCertificateForm=\"ngForm\">\n        <c8y-form-group>\n          <label translate for=\"certificateName\">Certificate name</label>\n          <input\n            id=\"certificateName\"\n            type=\"text\"\n            class=\"form-control\"\n            autocomplete=\"off\"\n            name=\"certificateName\"\n            [(ngModel)]=\"trustedCertificate.name\"\n            placeholder=\"{{ 'e.g. My certificate' | translate }}\"\n            required\n          />\n        </c8y-form-group>\n        <c8y-form-group>\n          <label translate for=\"certificate\">Certificate</label>\n          <c8y-drop-area\n            class=\"drop-area-sm\"\n            (dropped)=\"uploadFile($event)\"\n            [loadingMessage]=\"'Importing, please wait.' | translate\"\n            [title]=\"'Paste the certificate in PEM format.' | translate\"\n            [maxAllowedFiles]=\"maxAllowedFiles\"\n          >\n          </c8y-drop-area>\n        </c8y-form-group>\n        <c8y-form-group>\n          <label title=\"{{ 'Auto registration' | translate }}\" class=\"c8y-checkbox\">\n            <input\n              id=\"autoRegistration\"\n              type=\"checkbox\"\n              name=\"autoRegistration\"\n              [(ngModel)]=\"trustedCertificate.autoRegistrationEnabled\"\n            /><span></span>\n            <span>{{ 'Auto registration' | translate }}</span>\n            <button\n              class=\"btn btn-clean\"\n              popover=\"{{ AUTO_REGISTRATION_POPOVER | translate }}\"\n              triggers=\"focus\"\n            >\n              <i [c8yIcon]=\"'question-circle-o'\"></i>\n            </button>\n          </label>\n        </c8y-form-group>\n        <button\n          type=\"button\"\n          class=\"btn\"\n          name=\"certificateStatus\"\n          [(ngModel)]=\"trustedCertificate.status\"\n          btnCheckbox\n          btnCheckboxTrue=\"ENABLED\"\n          btnCheckboxFalse=\"DISABLED\"\n        >\n          <span\n            title=\"{{ 'Disabled`trusted certificate status`' | translate }}\"\n            [hidden]=\"trustedCertificate.status !== 'DISABLED'\"\n          >\n            {{ 'Disabled`trusted certificate status`' | translate }}\n          </span>\n          <span\n            title=\"{{ 'Enabled`trusted certificate status`' | translate }}\"\n            [hidden]=\"trustedCertificate.status !== 'ENABLED'\"\n          >\n            {{ 'Enabled`trusted certificate status`' | translate }}\n          </span>\n        </button>\n      </form>\n    </div>\n  </div>\n  <div class=\"modal-footer\">\n    <button title=\"{{ 'Cancel' | translate }}\" class=\"btn btn-default\" (click)=\"close()\">\n      {{ 'Cancel' | translate }}\n    </button>\n\n    <button\n      title=\"{{ 'Add certificate' | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"save()\"\n      [disabled]=\"\n        addTrustedCertificateForm.form.invalid ||\n        addTrustedCertificateForm.form.pristine ||\n        !fileIsUploaded\n      \"\n    >\n      {{ 'Add certificate' | translate }}\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.TrustedCertificateService }, { type: ɵngcc3.AlertService }]; }, null); })();

class TrustedCertificatesComponent {
    constructor(bsModal, alertService, trustedCertificateService, modalService, translateService) {
        this.bsModal = bsModal;
        this.alertService = alertService;
        this.trustedCertificateService = trustedCertificateService;
        this.modalService = modalService;
        this.translateService = translateService;
        this.reloading = new BehaviorSubject(false);
        this.reload = new BehaviorSubject(null);
        this.trustedCertificates = this.reload.pipe(tap(() => this.reloading.next(true)), switchMap(() => this.getTrustedCertificates()), tap(() => this.reloading.next(false)));
        this.sortByExpirationDateAsc = pipe(tap((results) => {
            return results.sort((n1, n2) => n1.notAfter.localeCompare(n2.notAfter));
        }));
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.loadTrustedCertificates();
        });
    }
    loadTrustedCertificates() {
        return __awaiter(this, void 0, void 0, function* () {
            this.reload.next();
        });
    }
    getTrustedCertificates() {
        const filter = {
            pageSize: 1000,
            withTotalPages: true,
        };
        return this.trustedCertificateService.list(filter);
    }
    addTrustedCertificate() {
        return __awaiter(this, void 0, void 0, function* () {
            const modal = this.bsModal.show(AddTrustedCertificateComponent, {
                class: 'modal-sm',
                ignoreBackdropClick: true,
            }).content;
            try {
                yield modal.result;
                modal.close();
                this.reload.next();
            }
            catch (ex) {
                // do nothing
            }
        });
    }
    deleteTrustedCertificate(certificate, index) {
        return __awaiter(this, void 0, void 0, function* () {
            const title = gettext('Delete trusted certificate');
            const confirmationText = gettext('You are about to delete a trusted certificate "{{ certificateName }}".');
            const finalQuestion = gettext('Do you want to proceed?');
            const certificateName = certificate.name;
            const body = [
                this.translateService.instant(confirmationText, {
                    certificateName,
                }),
                this.translateService.instant(finalQuestion),
            ].join(' ');
            try {
                yield this.modalService.confirm(title, body, Status.DANGER, { ok: gettext('Delete') });
                yield this.trustedCertificateService.delete(certificate.fingerprint);
                this.alertService.success(gettext('Certificate deleted.'));
                this.reload.next();
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    updateCertificate(trustedCertificate, updatingPart) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.trustedCertificateService.update(Object.assign({ fingerprint: trustedCertificate.fingerprint }, updatingPart));
                this.alertService.success(gettext('Certificate saved.'));
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
    highlightDependingOnExpirationStatus(item) {
        const warningOffset = 24 * 60 * 60 * 1000 * 90; // 90 days
        const todayTimestamp = new Date().getTime();
        const warningTimestamp = new Date().getTime() + warningOffset;
        const notAfterTimestamp = new Date(item.notAfter).getTime();
        const expired = notAfterTimestamp < todayTimestamp;
        const expiresInLessThan90Days = notAfterTimestamp < warningTimestamp;
        if (expired) {
            return 'text-danger';
        }
        if (expiresInLessThan90Days) {
            return 'text-warning';
        }
        return '';
    }
}
TrustedCertificatesComponent.ɵfac = function TrustedCertificatesComponent_Factory(t) { return new (t || TrustedCertificatesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TrustedCertificateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.TranslateService)); };
TrustedCertificatesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TrustedCertificatesComponent, selectors: [["c8y-trusted-certificates"]], decls: 24, vars: 38, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "plus-circle"], ["c8yIcon", "refresh", 3, "ngClass"], ["class", "c8y-empty-state text-center", 4, "ngIf"], [1, "m-b-24", 3, "ngClass"], ["class", "page-sticky-header hidden-xs c8y-list__item c8y-list__item--double-actions", 4, "ngIf"], [4, "c8yFor", "c8yForOf", "c8yForPipe", "c8yForLoadMore"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "certificate"], ["translate", ""], [1, "btn", "btn-primary", 3, "title", "click"], [1, "page-sticky-header", "hidden-xs", "c8y-list__item", "c8y-list__item--double-actions"], [1, "c8y-list__item__block"], [1, "c8y-list__item__icon"], [1, "p-l-24"], [1, "c8y-list__item__body"], [1, "content-flex-59"], [1, "col-3"], [1, "text-truncate", 3, "title"], [1, "col-2"], [1, "c8y-list__item__actions"], [1, "visible-xs", "p-8"], ["type", "button", "name", "certificateStatus", "btnCheckbox", "", "btnCheckboxTrue", "ENABLED", "btnCheckboxFalse", "DISABLED", 1, "btn", "c8y-btn-checkbox--inline", 3, "ngModel", "ngModelChange"], [3, "title", "hidden"], ["translate", "", 1, "text-label-small", "m-t-8", "m-r-8", "visible-xs-inline"], [3, "ngClass"], ["c8yIcon", "calendar", "class", "m-r-4", 4, "ngIf"], ["c8yIcon", "warning", "class", "m-r-4", 4, "ngIf"], [1, "text-truncate"], [1, "text-label-small", "m-t-8", "m-r-4", "visible-xs-inline"], [3, "title", 4, "ngIf"], ["icon", "trash", 3, "label", "click"], [1, "p-t-16", "p-b-16"], [1, "row"], [1, "col-md-4"], [1, "control-label"], [1, "input-group", "input-group-editable"], ["type", "text", "required", "", 1, "form-control", 3, "ngModel", "ngModelChange"], [1, "input-group-btn"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["translate", "", 1, "legend", "form-block"], [1, "list-unstyled"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom"], [1, "small", "m-b-0", "m-r-8", "flex-item-v-start", "flex-grow"], [1, "flex-item-right", "text-break-word"], [1, "col-md-1"], [1, "col-md-7", "col-lg-5"], ["translate", "", 1, "control-label"], ["name", "name", "type", "text", "rows", "15", "readonly", "", 1, "form-control", "no-resize", 3, "ngModel", "ngModelChange"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", 3, "ngModel", "ngModelChange"], ["c8yIcon", "calendar", 1, "m-r-4"], ["c8yIcon", "warning", 1, "m-r-4"], [3, "title"]], template: function TrustedCertificatesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TrustedCertificatesComponent_Template_button_click_4_listener() { return ctx.addTrustedCertificate(); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 2);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(10, "button", 1);
        ɵngcc0.ɵɵlistener("click", function TrustedCertificatesComponent_Template_button_click_10_listener() { return ctx.loadTrustedCertificates(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelement(12, "i", 3);
        ɵngcc0.ɵɵpipe(13, "async");
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(16, TrustedCertificatesComponent_div_16_Template, 11, 6, "div", 4);
        ɵngcc0.ɵɵpipe(17, "async");
        ɵngcc0.ɵɵelementStart(18, "c8y-list-group", 5);
        ɵngcc0.ɵɵpipe(19, "async");
        ɵngcc0.ɵɵtemplate(20, TrustedCertificatesComponent_div_20_Template, 32, 30, "div", 6);
        ɵngcc0.ɵɵpipe(21, "async");
        ɵngcc0.ɵɵtemplate(22, TrustedCertificatesComponent_c8y_li_22_Template, 124, 93, "c8y-li", 7);
        ɵngcc0.ɵɵpipe(23, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_8_0;
        let tmp_9_0;
        let tmp_10_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 14, "Trusted certificates"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 16, "Add trusted certificate"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 18, "Add trusted certificate"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 20, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(34, _c0, ɵngcc0.ɵɵpipeBind1(13, 22, ctx.reloading)));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 24, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_8_0 = ɵngcc0.ɵɵpipeBind1(17, 26, ctx.trustedCertificates)) == null ? null : tmp_8_0.data.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(36, _c1, ((tmp_9_0 = ɵngcc0.ɵɵpipeBind1(19, 28, ctx.trustedCertificates)) == null ? null : tmp_9_0.data.length) < 10));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_10_0 = ɵngcc0.ɵɵpipeBind1(21, 30, ctx.trustedCertificates)) == null ? null : tmp_10_0.data.length) > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(23, 32, ctx.trustedCertificates))("c8yForPipe", ctx.sortByExpirationDateAsc)("c8yForLoadMore", "none");
    } }, directives: [ɵngcc3.TitleComponent, ɵngcc3.ActionBarItemComponent, ɵngcc3.IconDirective, ɵngcc8.NgClass, ɵngcc8.NgIf, ɵngcc3.ListGroupComponent, ɵngcc3.ForOfDirective, ɵngcc3.C8yTranslateDirective, ɵngcc3.ListItemComponent, ɵngcc3.ListItemIconComponent, ɵngcc3.ListItemBodyComponent, ɵngcc6.ButtonCheckboxDirective, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc3.ListItemActionComponent, ɵngcc3.ListItemCollapseComponent, ɵngcc3.FormGroupComponent, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc4.DefaultValueAccessor, ɵngcc4.RequiredValidator, ɵngcc4.CheckboxControlValueAccessor], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc8.AsyncPipe, ɵngcc3.DatePipe], encapsulation: 2 });
TrustedCertificatesComponent.ctorParameters = () => [
    { type: BsModalService },
    { type: AlertService },
    { type: TrustedCertificateService },
    { type: ModalService },
    { type: TranslateService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TrustedCertificatesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-trusted-certificates',
                template: "<c8y-title>{{ 'Trusted certificates' | translate }}</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    title=\"{{ 'Add trusted certificate' | translate }}\"\n    class=\"btn btn-link\"\n    (click)=\"addTrustedCertificate()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add trusted certificate' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    title=\"{{ 'Reload' | translate }}\"\n    class=\"btn btn-link\"\n    (click)=\"loadTrustedCertificates()\"\n  >\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': reloading | async }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"c8y-empty-state text-center\" *ngIf=\"(trustedCertificates | async)?.data.length === 0\">\n  <h1 c8yIcon=\"certificate\"></h1>\n  <h3 translate>No trusted certificates to display.</h3>\n  <p translate>Add your first certificate by clicking below.</p>\n  <p>\n    <button\n      title=\"{{ 'Add trusted certificate' | translate }}\"\n      (click)=\"addTrustedCertificate()\"\n      class=\"btn btn-primary\"\n    >\n      {{ 'Add trusted certificate' | translate }}\n    </button>\n  </p>\n</div>\n\n<c8y-list-group \n  [ngClass]=\"{'dd-low': (trustedCertificates | async)?.data.length < 10 }\"\n  class=\"m-b-24\"\n>\n  <div *ngIf=\"(trustedCertificates | async)?.data.length > 0 \"\n    class=\"page-sticky-header hidden-xs c8y-list__item c8y-list__item--double-actions\">\n    <div class=\"c8y-list__item__block\">\n      <div class=\"c8y-list__item__icon\">\n        <i class=\"p-l-24\"></i>\n      </div>\n      <div class=\"c8y-list__item__body\">\n        <div class=\"content-flex-59\">\n          <div class=\"col-3\">\n            <span class=\"text-truncate\" title=\" {{'Certificate' | translate}} \">\n              {{'Certificate' | translate}}\n            </span>\n          </div>\n          <div class=\"col-2\">\n            <span class=\"text-truncate\" title=\"{{'Status' | translate}}\">\n              {{'Status' | translate}}\n            </span>\n          </div>\n          <div class=\"col-2\">\n            <span class=\"text-truncate\" title=\"{{'Algorithm' | translate}}\">\n              {{'Algorithm' | translate}}\n            </span>\n          </div>\n          <div class=\"col-3\">\n            <span class=\"text-truncate\" title=\"{{'Expiration date' | translate}}\">\n              {{'Expiration date' | translate}}\n            </span>\n          </div>\n          <div class=\"col-2\">\n            <span class=\"text-truncate\" title=\" {{'Auto registration' | translate}}\">\n              {{'Auto registration' | translate}}\n            </span>\n          </div>\n        </div>\n      </div>\n      <div class=\"c8y-list__item__actions\"></div>\n    </div>\n  </div>\n  \n  <c8y-li\n    *c8yFor=\"\n      let trustedCertificate of trustedCertificates | async;\n      let i = index;\n      pipe: sortByExpirationDateAsc;\n      loadMore: 'none'\n    \"\n  >\n    <c8y-li-icon>\n      <i c8yIcon=\"certificate\"></i>\n    </c8y-li-icon>\n\n    <c8y-li-body class=\"content-flex-59\">\n      <div class=\"col-3\">\n        <span class=\"text-truncate\" title=\"{{ trustedCertificate.name }}\">\n          {{ trustedCertificate.name }}\n        </span>\n      </div>\n      <div class=\"col-2\">\n        <div class=\"visible-xs p-8\"></div>\n        <button\n          type=\"button\"\n          class=\"btn c8y-btn-checkbox--inline\"\n          name=\"certificateStatus\"\n          [(ngModel)]=\"trustedCertificate.status\"\n          btnCheckbox\n          btnCheckboxTrue=\"ENABLED\"\n          btnCheckboxFalse=\"DISABLED\"\n          (ngModelChange)=\"updateCertificate(trustedCertificate, { status: $event })\"\n        >\n          <small\n            title=\"{{ 'Disabled`trusted certificate status`' | translate }}\"\n            [hidden]=\"trustedCertificate.status !== 'DISABLED'\"\n          >\n            {{ 'Disabled`trusted certificate status`' | translate }}\n          </small>\n          <small\n            title=\"{{ 'Enabled`trusted certificate status`' | translate }}\"\n            [hidden]=\"trustedCertificate.status !== 'ENABLED'\"\n          >\n            {{ 'Enabled`trusted certificate status`' | translate }}\n          </small>\n        </button>\n        <div class=\"visible-xs p-8\"></div>\n      </div>\n      <div class=\"col-2\">\n        <div class=\"text-truncate\" title=\"{{'Algorithm' | translate }}: {{ trustedCertificate.algorithmName }}\">\n          <span class=\"text-label-small m-t-8 m-r-8 visible-xs-inline \" translate>Algorithm</span>\n          {{ trustedCertificate.algorithmName }}\n        </div>\n      </div>\n\n      <div class=\"col-3\">\n        <div class=\"text-truncate\" \n          title=\"{{'Expiration date' | translate}}: {{ trustedCertificate.notAfter | c8yDate }}\">\n          <span class=\"text-label-small m-t-8 m-r-8 visible-xs-inline\" translate>Expiration date</span>\n          <small [ngClass]=\"highlightDependingOnExpirationStatus(trustedCertificate)\">\n            <i c8yIcon=\"calendar\" class=\"m-r-4\" *ngIf=\"!highlightDependingOnExpirationStatus(trustedCertificate)\"></i>\n            <i c8yIcon=\"warning\" class=\"m-r-4\" *ngIf=\"highlightDependingOnExpirationStatus(trustedCertificate)\"></i>\n            <span>{{ trustedCertificate.notAfter | c8yDate }}</span>\n          </small>\n        </div>\n      </div>\n      <div class=\"col-2\">\n        <div class=\"text-truncate\">\n          <span class=\"text-label-small m-t-8 m-r-4 visible-xs-inline\">\n            {{ 'Auto registration' | translate }}</span>\n          <span *ngIf=\"trustedCertificate.autoRegistrationEnabled\"\n          title=\"{{ 'Auto registration' | translate }}:  {{ 'Enabled`auto registration`' | translate }}\">\n            {{ 'Enabled`auto registration`' | translate }}\n          </span>\n          <span *ngIf=\"!trustedCertificate.autoRegistrationEnabled\"\n          title=\"{{ 'Auto registration' | translate }}:  {{ 'Disabled`auto registration`' | translate }}\">\n            {{ 'Disabled`auto registration`' | translate }}\n          </span>\n        </div>\n      </div>\n    </c8y-li-body>\n\n    <c8y-li-action\n      (click)=\"deleteTrustedCertificate(trustedCertificate, i)\"\n      icon=\"trash\"\n      label=\"{{ 'Delete' | translate }}\"\n    >\n    </c8y-li-action>\n\n    <c8y-li-collapse>\n      <div class=\"p-t-16 p-b-16\">\n        <div class=\"row\">\n          <div class=\"col-md-4\">\n            <c8y-form-group>\n              <label class=\"control-label\">\n                {{ 'Certificate name' | translate }}\n              </label>\n              <div class=\"input-group input-group-editable\">\n                <input\n                  type=\"text\"\n                  class=\"form-control\"\n                  [(ngModel)]=\"trustedCertificate.name\"\n                  required\n                />\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    title=\"{{ 'Update certificate name' | translate }}\"\n                    (click)=\"\n                      updateCertificate(trustedCertificate, { name: trustedCertificate.name })\n                    \"\n                    class=\"btn btn-primary\"\n                    [disabled]=\"!trustedCertificate.name\"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </c8y-form-group>\n            <div class=\"legend form-block\" translate>Additional properties</div>\n            <ul class=\"list-unstyled\">\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\" >\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Algorithm' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.algorithmName }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Version' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.version }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Valid from' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.notBefore | c8yDate }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Issuer' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.issuer }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Expiration date' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.notAfter | c8yDate }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Serial number' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.serialNumber }}\n                </span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom\">\n                <label class=\"small m-b-0 m-r-8 flex-item-v-start flex-grow\">\n                  {{ 'Subject`of a certificate`' | translate }}\n                </label>\n                <span class=\"flex-item-right text-break-word\">\n                  {{ trustedCertificate.subject }}\n                </span>\n              </li>\n            </ul>\n          </div>\n          <div class=\"col-md-1\"></div>\n          <div class=\"col-md-7 col-lg-5\">\n            <c8y-form-group>\n              <label class=\"control-label\" translate>\n                Certificate\n              </label>\n              <textarea\n                class=\"form-control no-resize\"\n                name=\"name\"\n                type=\"text\"\n                rows=\"15\"\n                readonly\n                [(ngModel)]=\"trustedCertificate.certInPemFormat\"\n              >\n              </textarea>\n            </c8y-form-group>\n            <c8y-form-group>\n              <label title=\"{{ 'Auto registration' | translate }}\" class=\"c8y-checkbox\">\n                <input\n                  type=\"checkbox\"\n                  [(ngModel)]=\"trustedCertificate.autoRegistrationEnabled\"\n                  (ngModelChange)=\"\n                    updateCertificate(trustedCertificate, { autoRegistrationEnabled: $event })\n                  \"\n                />\n                <span></span>\n                <span> {{ 'Auto registration' | translate }}</span>\n              </label>\n            </c8y-form-group>\n          </div>\n        </div>\n      </div>\n    </c8y-li-collapse>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalService }, { type: ɵngcc3.AlertService }, { type: ɵngcc2.TrustedCertificateService }, { type: ɵngcc3.ModalService }, { type: ɵngcc7.TranslateService }]; }, null); })();

class TrustedCertificatesNavigationFactory {
    constructor() {
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Trusted certificates'),
                    icon: 'certificate',
                    path: '/trusted-certificates',
                    parent: gettext('Management'),
                    priority: 100
                }));
            }
            return this.navs;
        });
    }
}
TrustedCertificatesNavigationFactory.ɵfac = function TrustedCertificatesNavigationFactory_Factory(t) { return new (t || TrustedCertificatesNavigationFactory)(); };
TrustedCertificatesNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TrustedCertificatesNavigationFactory, factory: TrustedCertificatesNavigationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TrustedCertificatesNavigationFactory, [{
        type: Injectable
    }], function () { return []; }, null); })();

const trustedCertificatesRoutes = [
    {
        path: 'trusted-certificates',
        component: TrustedCertificatesComponent
    }
];
class TrustedCertificatesModule {
}
TrustedCertificatesModule.ɵfac = function TrustedCertificatesModule_Factory(t) { return new (t || TrustedCertificatesModule)(); };
TrustedCertificatesModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TrustedCertificatesModule });
TrustedCertificatesModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: TrustedCertificatesNavigationFactory,
            multi: true
        }
    ], imports: [[
            CoreModule,
            CommonModule,
            RouterModule.forRoot(trustedCertificatesRoutes, { useHash: true }),
            BsDropdownModule.forRoot(),
            TooltipModule,
            ReactiveFormsModule,
            ButtonsModule,
            PopoverModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TrustedCertificatesModule, [{
        type: NgModule,
        args: [{
                declarations: [TrustedCertificatesComponent, AddTrustedCertificateComponent],
                exports: [],
                imports: [
                    CoreModule,
                    CommonModule,
                    RouterModule.forRoot(trustedCertificatesRoutes, { useHash: true }),
                    BsDropdownModule.forRoot(),
                    TooltipModule,
                    ReactiveFormsModule,
                    ButtonsModule,
                    PopoverModule
                ],
                entryComponents: [TrustedCertificatesComponent, AddTrustedCertificateComponent],
                providers: [
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: TrustedCertificatesNavigationFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TrustedCertificatesModule, { declarations: function () { return [TrustedCertificatesComponent, AddTrustedCertificateComponent]; }, imports: function () { return [CoreModule,
        CommonModule, ɵngcc9.RouterModule, ɵngcc10.BsDropdownModule, TooltipModule,
        ReactiveFormsModule,
        ButtonsModule,
        PopoverModule]; } }); })();

/**
 * Generated bundle index. Do not edit.
 */

export { AddTrustedCertificateComponent, TrustedCertificatesComponent, TrustedCertificatesModule, TrustedCertificatesNavigationFactory };
