// @c8y/ngx-components/esm2015/device-grid-example/device-grid-example-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import { DeviceGridExampleGuard } from './device-grid-example-guard';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './device-grid-example-guard';
export class DeviceGridExampleNavigationFactory {
    constructor(deviceGridExampleGuard) {
        this.deviceGridExampleGuard = deviceGridExampleGuard;
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.nav && this.deviceGridExampleGuard.canActivate(null, null)) {
                this.nav = new NavigatorNode({
                    label: gettext('Grids'),
                    icon: 'table'
                });
                this.nav.add(new NavigatorNode({
                    label: gettext('Device grid'),
                    path: '/grids/device-grid',
                    icon: 'table'
                }));
            }
            return this.nav;
        });
    }
}
DeviceGridExampleNavigationFactory.ɵfac = function DeviceGridExampleNavigationFactory_Factory(t) { return new (t || DeviceGridExampleNavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.DeviceGridExampleGuard)); };
DeviceGridExampleNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceGridExampleNavigationFactory, factory: DeviceGridExampleNavigationFactory.ɵfac });
DeviceGridExampleNavigationFactory.ctorParameters = () => [
    { type: DeviceGridExampleGuard }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceGridExampleNavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.DeviceGridExampleGuard }]; }, null); })();
