// @c8y/ngx-components/esm2015/operations/device-selector/device-selector.module.js
import { NgModule } from '@angular/core';
import { CoreModule } from '@c8y/ngx-components';
import { DeviceGridModule } from '@c8y/ngx-components/device-grid';
import { DeviceSelectorComponent } from './device-selector.component';
/**
 * This module provides a component for selecting devices based on filters.
 */
import * as ɵngcc0 from '@angular/core';
export class DeviceSelectorModule {
}
DeviceSelectorModule.ɵfac = function DeviceSelectorModule_Factory(t) { return new (t || DeviceSelectorModule)(); };
DeviceSelectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceSelectorModule });
DeviceSelectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, DeviceGridModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [CoreModule, DeviceGridModule],
                declarations: [DeviceSelectorComponent],
                exports: [DeviceSelectorComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceSelectorModule, { declarations: function () { return [DeviceSelectorComponent]; }, imports: function () { return [CoreModule, DeviceGridModule]; }, exports: function () { return [DeviceSelectorComponent]; } }); })();
