// @c8y/ngx-components/esm2015/repository/configuration/configuration-detail.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { uniqBy, isUndefined } from 'lodash-es';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { pipe } from 'rxjs';
import { map } from 'rxjs/operators';
import { RepositoryType } from '../repository.model';
import { RepositoryService } from '../repository.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../repository.service';
import * as ɵngcc2 from 'ngx-bootstrap/modal';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';

const _c0 = ["configurationForm"];
function ConfigurationDetailComponent_h4_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h4", 8);
    ɵngcc0.ɵɵtext(1, "Update configuration");
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationDetailComponent_h4_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h4", 8);
    ɵngcc0.ɵɵtext(1, "Add configuration");
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationDetailComponent_c8y_li_29_Template(rf, ctx) { if (rf & 1) {
    const _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 20);
    ɵngcc0.ɵɵlistener("click", function ConfigurationDetailComponent_c8y_li_29_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r8); const config_r6 = restoredCtx.$implicit; const ctx_r7 = ɵngcc0.ɵɵnextContext(); ctx_r7.selected = config_r6; return ctx_r7.setPipe(""); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const config_r6 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("active", ctx_r3.selected === config_r6);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", config_r6.configurationType || "--")("pattern", ctx_r3.pattern);
} }
function ConfigurationDetailComponent_ng_template_30_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 23);
    ɵngcc0.ɵɵelementStart(1, "span", 8);
    ɵngcc0.ɵɵtext(2, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "button", 24);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtext(5, " Add new`configuration type` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(4, 1, "Add new`configuration type`"));
} }
function ConfigurationDetailComponent_ng_template_30_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, ConfigurationDetailComponent_ng_template_30_c8y_li_0_Template, 6, 3, "c8y-li", 22);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r5.pattern.length > 0);
} }
const _c1 = function (a0) { return { "btn-pending": a0 }; };
export class ConfigurationDetailComponent {
    constructor(repositoryService, bsModalRef, alert) {
        this.repositoryService = repositoryService;
        this.bsModalRef = bsModalRef;
        this.alert = alert;
        this.binary = {
            file: undefined,
            url: undefined
        };
        this.pattern = '';
        this.mo = {};
        this.saving = false;
        this.uploadChoice = 'uploadBinary';
        this.textForConfigurationUrlPopover = gettext(`Path for binaries can vary depending on device agent implementation, e.g.:
    /configuration/binaries/configuration1.bin
    https://configuration/binary/123
    ftp://configuration/binary/123.tar.gz
  `);
        this.result = new Promise((resolve, reject) => {
            this._save = resolve;
            this._cancel = reject;
        });
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.configs = yield this.repositoryService.listRepositoryEntries(RepositoryType.CONFIGURATION);
            if (this.selected) {
                this.uploadChoice = this.binary.file ? 'uploadBinary' : 'uploadUrl';
                this.existingBinary = this.binary.file;
            }
            this.setPipe('');
            this.submitButtonTitle = this.mo.id
                ? gettext('Update configuration')
                : gettext('Add configuration');
        });
    }
    cancel() {
        this.bsModalRef.hide();
        this._cancel();
    }
    setPipe(filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map((data) => uniqBy(data, 'configurationType')), map((data) => {
            return data.filter((mo) => mo.configurationType &&
                mo.configurationType.toLowerCase().indexOf(filterStr.toLowerCase()) > -1);
        }));
    }
    onFile(dropped) {
        this.configurationForm.form.markAsDirty();
        if (!isUndefined(dropped.url)) {
            this.binary = {
                url: dropped.url
            };
            return;
        }
        else if (!isUndefined(dropped.droppedFiles)) {
            this.binary = {
                file: dropped.droppedFiles[0].file
            };
            return;
        }
        else {
            this.binary = {
                file: undefined,
                url: undefined
            };
        }
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.saving = true;
                const { selected, version, description, binary, deviceType } = this;
                if (this.existingBinary === this.binary.file) {
                    binary.file = undefined;
                }
                yield this.repositoryService.save({ selected, version, description, binary, deviceType }, RepositoryType.CONFIGURATION, this.mo);
                this.alert.success(this.mo.id ? gettext('Configuration updated.') : gettext('Configuration created.'));
                this.bsModalRef.hide();
                this._save();
            }
            catch (ex) {
                this.alert.addServerFailure(ex);
                this._cancel();
            }
            finally {
                this.saving = false;
            }
        });
    }
}
ConfigurationDetailComponent.ɵfac = function ConfigurationDetailComponent_Factory(t) { return new (t || ConfigurationDetailComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
ConfigurationDetailComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ConfigurationDetailComponent, selectors: [["c8y-configuration-detail"]], viewQuery: function ConfigurationDetailComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.configurationForm = _t.first);
    } }, decls: 45, vars: 41, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], ["translate", "", 4, "ngIf"], [1, "d-contents", 3, "ngSubmit"], ["configurationForm", "ngForm"], [1, "modal-inner-scroll"], [1, "modal-body"], ["translate", ""], ["type", "text", "autocomplete", "off", "required", "", "maxlength", "254", "name", "version", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["type", "text", "maxlength", "254", "autocomplete", "off", "name", "deviceType", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["type", "text", "maxlength", "254", "autocomplete", "off", "name", "description", 1, "form-control", 3, "placeholder", "ngModel", "ngModelChange"], ["name", "confType", "maxlength", "254", "displayProperty", "configurationType", 3, "ngModel", "placeholder", "ngModelChange", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "c8yFor", "c8yForOf", "c8yForPipe", "c8yForNotFound"], ["notFoundTemplate", ""], ["translate", "", 1, "legend", "form-block", "m-t-40"], [3, "maxAllowedFiles", "uploadChoice", "fileUrl", "fileBinary", "fileUrlPopover", "onFilesPicked"], [1, "modal-footer"], ["type", "button", 1, "btn", "btn-default", 3, "title", "disabled", "click"], ["type", "submit", 1, "btn", "btn-primary", 3, "title", "ngClass", "disabled"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"], ["class", "bg-gray-lighter p-8", 4, "ngIf"], [1, "bg-gray-lighter", "p-8"], ["type", "button", "translate", "", 1, "btn", "btn-primary", "btn-xs", "m-l-16", 3, "title"]], template: function ConfigurationDetailComponent_Template(rf, ctx) { if (rf & 1) {
        const _r10 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "i", 2);
        ɵngcc0.ɵɵtemplate(3, ConfigurationDetailComponent_h4_3_Template, 2, 0, "h4", 3);
        ɵngcc0.ɵɵtemplate(4, ConfigurationDetailComponent_h4_4_Template, 2, 0, "h4", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "form", 4, 5);
        ɵngcc0.ɵɵlistener("ngSubmit", function ConfigurationDetailComponent_Template_form_ngSubmit_5_listener() { ɵngcc0.ɵɵrestoreView(_r10); const _r2 = ɵngcc0.ɵɵreference(6); return _r2.form.valid && ctx.save(); });
        ɵngcc0.ɵɵelementStart(7, "div", 6);
        ɵngcc0.ɵɵelementStart(8, "div", 7);
        ɵngcc0.ɵɵelementStart(9, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(10, "label", 8);
        ɵngcc0.ɵɵtext(11, "Name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "input", 9);
        ɵngcc0.ɵɵlistener("ngModelChange", function ConfigurationDetailComponent_Template_input_ngModelChange_12_listener($event) { return ctx.version = $event; });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(15, "label", 8);
        ɵngcc0.ɵɵtext(16, "Device type");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "input", 10);
        ɵngcc0.ɵɵlistener("ngModelChange", function ConfigurationDetailComponent_Template_input_ngModelChange_17_listener($event) { return ctx.deviceType = $event; });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(20, "label", 8);
        ɵngcc0.ɵɵtext(21, "Description");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "input", 11);
        ɵngcc0.ɵɵlistener("ngModelChange", function ConfigurationDetailComponent_Template_input_ngModelChange_22_listener($event) { return ctx.description = $event; });
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(25, "label", 8);
        ɵngcc0.ɵɵtext(26, "Configuration type");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(27, "c8y-typeahead", 12);
        ɵngcc0.ɵɵlistener("ngModelChange", function ConfigurationDetailComponent_Template_c8y_typeahead_ngModelChange_27_listener($event) { return ctx.selected = $event; })("onSearch", function ConfigurationDetailComponent_Template_c8y_typeahead_onSearch_27_listener($event) { return ctx.setPipe($event); });
        ɵngcc0.ɵɵpipe(28, "translate");
        ɵngcc0.ɵɵtemplate(29, ConfigurationDetailComponent_c8y_li_29_Template, 2, 3, "c8y-li", 13);
        ɵngcc0.ɵɵtemplate(30, ConfigurationDetailComponent_ng_template_30_Template, 1, 1, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(33, "div", 15);
        ɵngcc0.ɵɵtext(34, "Configuration file");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(35, "c8y-file-picker", 16);
        ɵngcc0.ɵɵlistener("onFilesPicked", function ConfigurationDetailComponent_Template_c8y_file_picker_onFilesPicked_35_listener($event) { return ctx.onFile($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(36, "div", 17);
        ɵngcc0.ɵɵelementStart(37, "button", 18);
        ɵngcc0.ɵɵlistener("click", function ConfigurationDetailComponent_Template_button_click_37_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(38, "translate");
        ɵngcc0.ɵɵelementStart(39, "span", 8);
        ɵngcc0.ɵɵtext(40, "Cancel");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(41, "button", 19);
        ɵngcc0.ɵɵpipe(42, "translate");
        ɵngcc0.ɵɵtext(43);
        ɵngcc0.ɵɵpipe(44, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(6);
        const _r4 = ɵngcc0.ɵɵreference(31);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yIcon", "cogs");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mo.id);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.mo.id);
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(13, 25, "e.g."), " hosts");
        ɵngcc0.ɵɵproperty("ngModel", ctx.version);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(18, 27, "e.g."), " c8y_Linux");
        ɵngcc0.ɵɵproperty("ngModel", ctx.deviceType);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(23, 29, "e.g. Host configuration"), " c8y_Linux");
        ɵngcc0.ɵɵproperty("ngModel", ctx.description);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(28, 31, "e.g."), " ssh");
        ɵngcc0.ɵɵproperty("ngModel", ctx.selected);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.configs)("c8yForPipe", ctx.filterPipe)("c8yForNotFound", _r4);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("maxAllowedFiles", 1)("uploadChoice", ctx.uploadChoice)("fileUrl", ctx.binary.url)("fileBinary", ctx.binary.file)("fileUrlPopover", ctx.textForConfigurationUrlPopover);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(38, 33, "Cancel"));
        ɵngcc0.ɵɵproperty("disabled", ctx.saving);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(42, 35, ctx.submitButtonTitle));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(39, _c1, ctx.saving))("disabled", !_r2.valid || _r2.pristine || !(ctx.binary == null ? null : ctx.binary.url) && !(ctx.binary == null ? null : ctx.binary.file) || ctx.saving);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(44, 37, ctx.submitButtonTitle), " ");
    } }, directives: [ɵngcc3.IconDirective, ɵngcc4.NgIf, ɵngcc5.ɵNgNoValidate, ɵngcc5.NgControlStatusGroup, ɵngcc5.NgForm, ɵngcc3.FormGroupComponent, ɵngcc3.C8yTranslateDirective, ɵngcc3.RequiredInputPlaceholderDirective, ɵngcc5.DefaultValueAccessor, ɵngcc5.RequiredValidator, ɵngcc5.MaxLengthValidator, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc3.TypeaheadComponent, ɵngcc3.ForOfDirective, ɵngcc3.FilePickerComponent, ɵngcc4.NgClass, ɵngcc3.ListItemComponent, ɵngcc3.HighlightComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ConfigurationDetailComponent.ctorParameters = () => [
    { type: RepositoryService },
    { type: BsModalRef },
    { type: AlertService }
];
ConfigurationDetailComponent.propDecorators = {
    configurationForm: [{ type: ViewChild, args: ['configurationForm', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfigurationDetailComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-configuration-detail',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <i [c8yIcon]=\"'cogs'\"></i>\n    <h4 translate *ngIf=\"mo.id\">Update configuration</h4>\n    <h4 translate *ngIf=\"!mo.id\">Add configuration</h4>\n  </div>\n\n  <form\n    class=\"d-contents\"\n    #configurationForm=\"ngForm\"\n    (ngSubmit)=\"configurationForm.form.valid && save()\"\n  >\n    <div class=\"modal-inner-scroll\">\n      <div class=\"modal-body\">\n        <c8y-form-group>\n          <label translate>Name</label>\n          <input\n            type=\"text\"\n            class=\"form-control\"\n            placeholder=\"{{ 'e.g.' | translate }} hosts\"\n            autocomplete=\"off\"\n            required\n            maxlength=\"254\"\n            [(ngModel)]=\"version\"\n            name=\"version\"\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label translate>Device type</label>\n          <input\n            type=\"text\"\n            class=\"form-control\"\n            placeholder=\"{{ 'e.g.' | translate }} c8y_Linux\"\n            maxlength=\"254\"\n            autocomplete=\"off\"\n            [(ngModel)]=\"deviceType\"\n            name=\"deviceType\"\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label translate>Description</label>\n          <input\n            type=\"text\"\n            class=\"form-control\"\n            placeholder=\"{{ 'e.g. Host configuration' | translate }} c8y_Linux\"\n            maxlength=\"254\"\n            autocomplete=\"off\"\n            [(ngModel)]=\"description\"\n            name=\"description\"\n          />\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <label translate>Configuration type</label>\n          <c8y-typeahead\n            [(ngModel)]=\"selected\"\n            name=\"confType\"\n            placeholder=\"{{ 'e.g.' | translate }} ssh\"\n            maxlength=\"254\"\n            (onSearch)=\"setPipe($event)\"\n            displayProperty=\"configurationType\"\n          >\n            <c8y-li\n              *c8yFor=\"let config of configs; pipe: filterPipe; notFound: notFoundTemplate\"\n              class=\"p-l-8 p-r-8 c8y-list__item--link\"\n              (click)=\"selected = config; setPipe('')\"\n              [active]=\"selected === config\"\n            >\n              <c8y-highlight\n                [text]=\"config.configurationType || '--'\"\n                [pattern]=\"pattern\"\n              ></c8y-highlight>\n            </c8y-li>\n            <ng-template #notFoundTemplate>\n              <c8y-li class=\"bg-gray-lighter p-8\" *ngIf=\"pattern.length > 0\">\n                <span translate>No match found.</span>\n                <button\n                  title=\"{{ 'Add new`configuration type`' | translate }}\"\n                  type=\"button\"\n                  class=\"btn btn-primary btn-xs m-l-16\"\n                  translate\n                >\n                  Add new`configuration type`\n                </button>\n              </c8y-li>\n            </ng-template>\n          </c8y-typeahead>\n        </c8y-form-group>\n\n        <c8y-form-group>\n          <div class=\"legend form-block m-t-40\" translate>Configuration file</div>\n          <c8y-file-picker\n            [maxAllowedFiles]=\"1\"\n            (onFilesPicked)=\"onFile($event)\"\n            [uploadChoice]=\"uploadChoice\"\n            [fileUrl]=\"binary.url\"\n            [fileBinary]=\"binary.file\"\n            [fileUrlPopover]=\"textForConfigurationUrlPopover\"\n          >\n          </c8y-file-picker>\n        </c8y-form-group>\n      </div>\n    </div>\n\n    <div class=\"modal-footer\">\n      <button\n        (click)=\"cancel()\"\n        type=\"button\"\n        class=\"btn btn-default\"\n        title=\"{{ 'Cancel' | translate }}\"\n        [disabled]=\"saving\"\n      >\n        <span translate>Cancel</span>\n      </button>\n      <button\n        class=\"btn btn-primary\"\n        type=\"submit\"\n        title=\"{{ submitButtonTitle | translate }}\"\n        [ngClass]=\"{ 'btn-pending': saving }\"\n        [disabled]=\"\n          !configurationForm.valid ||\n          configurationForm.pristine ||\n          (!binary?.url && !binary?.file) ||\n          saving\n        \"\n      >\n        {{ submitButtonTitle | translate }}\n      </button>\n    </div>\n  </form>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.RepositoryService }, { type: ɵngcc2.BsModalRef }, { type: ɵngcc3.AlertService }]; }, { configurationForm: [{
            type: ViewChild,
            args: ['configurationForm', { static: true }]
        }] }); })();
