// @c8y/ngx-components/esm2015/core/authentication/password-strength.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { PasswordService } from './password.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './password.service';
import * as ɵngcc2 from '../i18n/c8y-translate.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';
export class PasswordStrengthComponent {
    constructor(passwordStrengthCheckerService) {
        this.passwordStrengthCheckerService = passwordStrengthCheckerService;
        this.strength = new EventEmitter();
        this.password = '';
    }
    set password(password) {
        if (typeof password === 'string') {
            this.model = this.passwordStrengthCheckerService.getStrengthColor(password);
            this.strength.emit(this.model);
        }
    }
}
PasswordStrengthComponent.ɵfac = function PasswordStrengthComponent_Factory(t) { return new (t || PasswordStrengthComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.PasswordService)); };
PasswordStrengthComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PasswordStrengthComponent, selectors: [["c8y-password-strength"]], inputs: { password: "password" }, outputs: { strength: "strength" }, decls: 8, vars: 6, consts: [["translate", "", 1, "password-strength-label", "small"], [1, "password-bar"], [1, "small"]], template: function PasswordStrengthComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵelementStart(1, "label", 0);
        ɵngcc0.ɵɵtext(2, " Password strength ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div");
        ɵngcc0.ɵɵelement(4, "div", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "span", 2);
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMapInterpolate1("password-strength password-", ctx.model.colorName, "");
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 4, ctx.model.description), " ");
    } }, directives: [ɵngcc2.C8yTranslateDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
PasswordStrengthComponent.ctorParameters = () => [
    { type: PasswordService }
];
PasswordStrengthComponent.propDecorators = {
    password: [{ type: Input, args: ['password',] }],
    strength: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PasswordStrengthComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-password-strength',
                template: "<div class=\"password-strength password-{{ model.colorName }}\">\r\n  <label class=\"password-strength-label small\" translate>\r\n    Password strength\r\n  </label>\r\n  <div>\r\n    <div class=\"password-bar\"></div>\r\n  </div>\r\n  <span class=\"small\">\r\n    {{ model.description | translate }}\r\n  </span>\r\n</div>\r\n"
            }]
    }], function () { return [{ type: ɵngcc1.PasswordService }]; }, { strength: [{
            type: Output
        }], password: [{
            type: Input,
            args: ['password']
        }] }); })();
