// @c8y/ngx-components/esm2015/core/authentication/password-confirm.directive.js
import { Directive, Attribute } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
export class PasswordConfirm {
    constructor(passwordConfirm) {
        this.passwordConfirm = passwordConfirm;
    }
    validate(abControl) {
        const value = abControl.value;
        const controlToCompareWith = abControl.root.get(this.passwordConfirm);
        if (controlToCompareWith && value !== controlToCompareWith.value) {
            return { passwordConfirm: true };
        }
        return null;
    }
}
PasswordConfirm.ɵfac = function PasswordConfirm_Factory(t) { return new (t || PasswordConfirm)(ɵngcc0.ɵɵinjectAttribute('passwordConfirm')); };
PasswordConfirm.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PasswordConfirm, selectors: [["", "passwordConfirm", ""]], features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALIDATORS,
                useExisting: PasswordConfirm,
                multi: true
            }
        ])] });
PasswordConfirm.ctorParameters = () => [
    { type: String, decorators: [{ type: Attribute, args: ['passwordConfirm',] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PasswordConfirm, [{
        type: Directive,
        args: [{
                selector: '[passwordConfirm]',
                providers: [
                    {
                        provide: NG_VALIDATORS,
                        useExisting: PasswordConfirm,
                        multi: true
                    }
                ]
            }]
    }], function () { return [{ type: String, decorators: [{
                type: Attribute,
                args: ['passwordConfirm']
            }] }]; }, null); })();
