// @c8y/ngx-components/esm2015/core/alert/alert.component.js
import { Component, Input, ViewChild, ContentChild } from '@angular/core';
import { AlertService } from './alert.service';
import { AlertDetailsComponent } from './alert-details.component';
/**
 * Alert component allows creating alert with complex view.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './alert.service';

const _c0 = ["detailedData"];
const _c1 = ["text"];
function AlertComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
    ɵngcc0.ɵɵprojection(1, 1);
} }
function AlertComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0, 2);
} }
const _c2 = [[["c8y-alert-text"]], "*", [["c8y-alert-details"]]];
const _c3 = ["c8y-alert-text", "*", "c8y-alert-details"];
export class AlertComponent {
    /**
     * @ignore
     */
    constructor(alertService) {
        this.alertService = alertService;
    }
    /**
     * AlertComponent adds new alert to collection stored in alertService after view init.
     */
    ngAfterViewInit() {
        if (!this.containDetailsComponent) {
            this.detailedData = null;
        }
        this.viewInitTimeout = setTimeout(() => this.alertService.add(this));
    }
    /**
     * AlertComponent removes alert from collection stored in alertService on destroy component.
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.alertService.remove(this);
    }
}
AlertComponent.ɵfac = function AlertComponent_Factory(t) { return new (t || AlertComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
AlertComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AlertComponent, selectors: [["c8y-alert"]], contentQueries: function AlertComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, AlertDetailsComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.containDetailsComponent = _t.first);
    } }, viewQuery: function AlertComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.detailedData = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.text = _t.first);
    } }, inputs: { type: "type", onDetail: "onDetail", onClose: "onClose" }, ngContentSelectors: _c3, decls: 4, vars: 0, consts: [["text", ""], ["detailedData", ""]], template: function AlertComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c2);
        ɵngcc0.ɵɵtemplate(0, AlertComponent_ng_template_0_Template, 2, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, AlertComponent_ng_template_2_Template, 1, 0, "ng-template", null, 1, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
AlertComponent.ctorParameters = () => [
    { type: AlertService }
];
AlertComponent.propDecorators = {
    type: [{ type: Input }],
    onDetail: [{ type: Input }],
    onClose: [{ type: Input }],
    detailedData: [{ type: ViewChild, args: ['detailedData', { static: false },] }],
    text: [{ type: ViewChild, args: ['text', { static: false },] }],
    containDetailsComponent: [{ type: ContentChild, args: [AlertDetailsComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlertComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-alert',
                template: "<ng-template #text>\r\n  <ng-content select=\"c8y-alert-text\"></ng-content>\r\n  <ng-content></ng-content>\r\n</ng-template>\r\n<ng-template #detailedData>\r\n  <ng-content select=\"c8y-alert-details\"></ng-content>\r\n</ng-template>"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }]; }, { detailedData: [{
            type: ViewChild,
            args: ['detailedData', { static: false }]
        }], type: [{
            type: Input
        }], onDetail: [{
            type: Input
        }], onClose: [{
            type: Input
        }], text: [{
            type: ViewChild,
            args: ['text', { static: false }]
        }], containDetailsComponent: [{
            type: ContentChild,
            args: [AlertDetailsComponent, { static: false }]
        }] }); })();
