// @c8y/ngx-components/esm2015/core/alert/alert-details.component.js
import { Component } from '@angular/core';
/**
 * The content of `c8y-alert-details` component will be displayed as alert's details.
 * This component must be placed inside `c8y-alert` component.
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class AlertDetailsComponent {
}
AlertDetailsComponent.ɵfac = function AlertDetailsComponent_Factory(t) { return new (t || AlertDetailsComponent)(); };
AlertDetailsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AlertDetailsComponent, selectors: [["c8y-alert-details"]], ngContentSelectors: _c0, decls: 1, vars: 0, template: function AlertDetailsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlertDetailsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-alert-details',
                template: `<ng-content></ng-content>`
            }]
    }], null, null); })();
