// @c8y/ngx-components/esm2015/core/alert/alert-outlet.component.js
import { Component, TemplateRef, Attribute, HostBinding } from '@angular/core';
import { AlertService } from './alert.service';
/**
 * Alert outlet component shows all added alerts. It is used in the application's `c8y-bootstrap` component.
 *
 * For applications which don’t use `c8y-bootstrap` component but want to display alerts via AlertService, a c8y-alert-outlet component needs to be added explicitly in the template.
 *
 * ## Example:
 * ```html
 *  <div class="alerts">
 *    <c8y-alert-outlet></c8y-alert-outlet>
 *  </div>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './alert.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../common/icon.directive';
import * as ɵngcc4 from '../i18n/c8y-translate.pipe';

function AlertOutletComponent_div_0_div_8_strong_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "strong", 10);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const alert_r1 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("textContent", ɵngcc0.ɵɵpipeBind1(1, 1, alert_r1.text));
} }
function AlertOutletComponent_div_0_div_8_strong_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "strong", 11);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const alert_r1 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(1, 1, alert_r1.text), ɵngcc0.ɵɵsanitizeHtml);
} }
function AlertOutletComponent_div_0_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, AlertOutletComponent_div_0_div_8_strong_1_Template, 2, 3, "strong", 8);
    ɵngcc0.ɵɵtemplate(2, AlertOutletComponent_div_0_div_8_strong_2_Template, 2, 3, "strong", 9);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const alert_r1 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !alert_r1.allowHtml);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", alert_r1.allowHtml);
} }
function AlertOutletComponent_div_0_p_9_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "p", 12);
    ɵngcc0.ɵɵelementStart(1, "button", 13);
    ɵngcc0.ɵɵlistener("click", function AlertOutletComponent_div_0_p_9_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r13); const alert_r1 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.showDetails(alert_r1); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 14);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Show details"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Show details"), " ");
} }
function AlertOutletComponent_div_0_p_10_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "p", 12);
    ɵngcc0.ɵɵelementStart(1, "button", 13);
    ɵngcc0.ɵɵlistener("click", function AlertOutletComponent_div_0_p_10_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(2); return ctx_r14.closeDetails(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 15);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Hide details"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Hide details"), " ");
} }
function AlertOutletComponent_div_0_div_11_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "pre");
    ɵngcc0.ɵɵelement(2, "code", 16);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("textContent", ɵngcc0.ɵɵpipeBind1(3, 1, ctx_r16.detailDisplayedAlert.contents));
} }
function AlertOutletComponent_div_0_div_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, AlertOutletComponent_div_0_div_11_div_1_Template, 4, 3, "div", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const alert_r1 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r5.isTemplateRef(ctx_r5.detailDisplayedAlert.contents))("ngIfElse", ctx_r5.template(alert_r1.detailedData));
} }
const _c0 = function (a0, a1) { return [a0, a1]; };
function AlertOutletComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "button", 2);
    ɵngcc0.ɵɵlistener("click", function AlertOutletComponent_div_0_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r19); const alert_r1 = restoredCtx.$implicit; const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.close(alert_r1); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "span", 3);
    ɵngcc0.ɵɵtext(4, "\u00D7");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span", 4);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, AlertOutletComponent_div_0_div_8_Template, 3, 2, "div", 5);
    ɵngcc0.ɵɵtemplate(9, AlertOutletComponent_div_0_p_9_Template, 6, 6, "p", 6);
    ɵngcc0.ɵɵtemplate(10, AlertOutletComponent_div_0_p_10_Template, 6, 6, "p", 6);
    ɵngcc0.ɵɵtemplate(11, AlertOutletComponent_div_0_div_11_Template, 2, 2, "div", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const alert_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(12, _c0, ctx_r0.getClass(alert_r1), ctx_r0.isDetailsShow(alert_r1) ? "expanded" : ""));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 8, "Close"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 10, "Close"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r0.isTemplateRef(alert_r1.text))("ngIfElse", ctx_r0.template(alert_r1.text));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.hasDetails(alert_r1) && !ctx_r0.isDetailsShow(alert_r1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.hasDetails(alert_r1) && ctx_r0.isDetailsShow(alert_r1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.isDetailsShow(alert_r1));
} }
export class AlertOutletComponent {
    /**
     * @ignore
     */
    constructor(position, alertService) {
        this.position = position;
        this.alertService = alertService;
        /**
         * Currently displayed details of an alert. Only one alert’s details can be displayed at a time.”
         */
        this.detailDisplayedAlert = {};
        /**
         * @ignore
         */
        this.isStatic = false;
        this.isStatic = this.position === 'static';
    }
    /**
     * The array of current alerts.
     */
    get alerts() {
        return this.alertService.state$;
    }
    /**
     * Returns a CSS class for an alert depending on its type to style the corresponding alert component.
     */
    getClass(alert) {
        const classes = [];
        if (alert.type !== 'danger') {
            classes.push('interact');
        }
        classes.push(`alert-${alert.type}`);
        return classes.join(' ');
    }
    /**
     * Executes alert’s onClose method, if provided, and closes the alert.
     * @param alert The alert which will be closed.
     */
    close(alert) {
        if (alert.onClose) {
            alert.onClose();
        }
        this.alertService.remove(alert);
    }
    /**
     * Expands details part of particular alert, and collapses details for another alerts.
     * @param alert The alert which details should be expanded.
     */
    showDetails(alert) {
        if (alert.onDetail) {
            return alert.onDetail();
        }
        this.detailDisplayedAlert = Object.assign({}, alert);
        this.detailDisplayedAlert.mode = 'raw';
        this.setDetailsContent(alert);
    }
    /**
     * Returns true if alert has extra detailedData or defines custom onDetail method.
     */
    hasDetails(alert) {
        return (alert.detailedData || alert.onDetail);
    }
    /**
     * Collapses alert's details.
     */
    closeDetails() {
        this.detailDisplayedAlert = {};
    }
    /**
     * Returns true if details part of alert is expanded.
     */
    isDetailsShow(alert) {
        return this.alertService.areSame(alert, this.detailDisplayedAlert);
    }
    /**
     * @ignore
     */
    isTemplateRef(value) {
        return value instanceof TemplateRef;
    }
    /**
     * @ignore
     */
    template(content) {
        return this.isTemplateRef(content) ? content : undefined;
    }
    setDetailsContent(alert) {
        if (this.isTemplateRef(alert.detailedData) || typeof alert.detailedData === 'string') {
            this.detailDisplayedAlert.contents = alert.detailedData;
        }
        else {
            this.detailDisplayedAlert.contents = JSON.stringify(alert.detailedData, null, 2);
        }
    }
}
AlertOutletComponent.ɵfac = function AlertOutletComponent_Factory(t) { return new (t || AlertOutletComponent)(ɵngcc0.ɵɵinjectAttribute('position'), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
AlertOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AlertOutletComponent, selectors: [["c8y-alert-outlet"]], hostVars: 2, hostBindings: function AlertOutletComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("static", ctx.isStatic);
    } }, decls: 2, vars: 3, consts: [["class", "alert animated fadeInRightBig m-t-16", 3, "ngClass", 4, "ngFor", "ngForOf"], [1, "alert", "animated", "fadeInRightBig", "m-t-16", 3, "ngClass"], ["type", "button", 1, "close", 3, "title", "click"], ["aria-hidden", "true"], [1, "sr-only"], [4, "ngIf", "ngIfElse"], ["class", "text-muted m-t-8", 4, "ngIf"], [4, "ngIf"], ["class", "message", 3, "textContent", 4, "ngIf"], ["class", "message", 3, "innerHTML", 4, "ngIf"], [1, "message", 3, "textContent"], [1, "message", 3, "innerHTML"], [1, "text-muted", "m-t-8"], [1, "btn", "btn-clean", 3, "title", "click"], ["c8yIcon", "chevron-down"], ["c8yIcon", "chevron-up"], [3, "textContent"]], template: function AlertOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AlertOutletComponent_div_0_Template, 12, 15, "div", 0);
        ɵngcc0.ɵɵpipe(1, "async");
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(1, 1, ctx.alerts));
    } }, directives: [ɵngcc2.NgForOf, ɵngcc2.NgClass, ɵngcc2.NgIf, ɵngcc3.IconDirective], pipes: [ɵngcc2.AsyncPipe, ɵngcc4.C8yTranslatePipe], encapsulation: 2 });
AlertOutletComponent.ctorParameters = () => [
    { type: String, decorators: [{ type: Attribute, args: ['position',] }] },
    { type: AlertService }
];
AlertOutletComponent.propDecorators = {
    isStatic: [{ type: HostBinding, args: ['class.static',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AlertOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-alert-outlet',
                template: "<div\n  class=\"alert animated fadeInRightBig m-t-16\"\n  *ngFor=\"let alert of alerts | async\"\n  [ngClass]=\"[getClass(alert), isDetailsShow(alert) ? 'expanded' : '']\"\n>\n  <button title=\"{{ 'Close' | translate }}\" type=\"button\" class=\"close\" (click)=\"close(alert)\">\n    <span aria-hidden=\"true\">&times;</span>\n    <span class=\"sr-only\">{{ 'Close' | translate }}</span>\n  </button>\n\n  <div *ngIf=\"!isTemplateRef(alert.text); else template(alert.text)\">\n    <strong *ngIf=\"!alert.allowHtml\" [textContent]=\"alert.text | translate\" class=\"message\"></strong>\n    <strong *ngIf=\"alert.allowHtml\" [innerHTML]=\"alert.text | translate\" class=\"message\"></strong>\n  </div>\n  <p *ngIf=\"hasDetails(alert) && !isDetailsShow(alert)\" class=\"text-muted m-t-8\">\n    <button\n      title=\"{{ 'Show details' | translate }}\"\n      class=\"btn btn-clean\"\n      (click)=\"showDetails(alert)\"\n    >\n      <i c8yIcon=\"chevron-down\"></i>\n      {{ 'Show details' | translate }}\n    </button>\n  </p>\n  <p *ngIf=\"hasDetails(alert) && isDetailsShow(alert)\" class=\"text-muted m-t-8\">\n    <button title=\"{{ 'Hide details' | translate }}\" class=\"btn btn-clean\" (click)=\"closeDetails()\">\n      <i c8yIcon=\"chevron-up\"></i>\n      {{ 'Hide details' | translate }}\n    </button>\n  </p>\n  <div *ngIf=\"isDetailsShow(alert)\">\n    <div *ngIf=\"!isTemplateRef(detailDisplayedAlert.contents); else template(alert.detailedData)\">\n      <pre><code [textContent]=\"detailDisplayedAlert.contents | translate\"></code></pre>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: String, decorators: [{
                type: Attribute,
                args: ['position']
            }] }, { type: ɵngcc1.AlertService }]; }, { isStatic: [{
            type: HostBinding,
            args: ['class.static']
        }] }); })();
