// @c8y/ngx-components/bundles/c8y-ngx-components-operations-stepper-bulk-type-configuration.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@c8y/ngx-components/operations/bulk-operations-service'),require('@c8y/ngx-components/repository'),require('@ngx-translate/core'),require('@c8y/ngx-components/operations/bulk-operation-stepper'),require('@angular/forms'),require('@angular/common'),require('@c8y/ngx-components'),exports, require('@angular/core'), require('@angular/forms'), require('@c8y/ngx-components'), require('@c8y/ngx-components/repository'), require('@c8y/ngx-components/operations/bulk-operation-stepper'), require('@c8y/ngx-components/operations/bulk-operations-service'), require('@ngx-translate/core'), require('lodash-es'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/stepper-bulk-type-configuration', ['@angular/core','@c8y/ngx-components/operations/bulk-operations-service','@c8y/ngx-components/repository','@ngx-translate/core','@c8y/ngx-components/operations/bulk-operation-stepper','@angular/forms','@angular/common','@c8y/ngx-components','exports', '@angular/core', '@angular/forms', '@c8y/ngx-components', '@c8y/ngx-components/repository', '@c8y/ngx-components/operations/bulk-operation-stepper', '@c8y/ngx-components/operations/bulk-operations-service', '@ngx-translate/core', 'lodash-es', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.c8y.ngxComponents.operations.bulkOperationsService,global.c8y.ngxComponents.repository,global.ngxTranslate.core,global.c8y.ngxComponents.operations.bulkOperationStepper,global.ng.forms,global.ng.common,global.c8y.ngxComponents,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["stepper-bulk-type-configuration"] = {}), global.ng.core, global.ng.forms, global.c8y["ngx-components"], global.c8y["ngx-components"].repository, global.c8y["ngx-components"].operations["bulk-operation-stepper"], global.c8y["ngx-components"].operations["bulk-operations-service"], global.core$1, global.lodashEs, global.rxjs, global.rxjs.operators));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,exports, core, forms, ngxComponents, repository, bulkOperationStepper, bulkOperationsService, core$1, lodashEs, rxjs, operators) { 
function StepperBulkTypeConfigurationComponent_ng_container_1_i_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 23);
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_i_17_Template(rf, ctx) { if (rf & 1) {
    var _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 24);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_i_17_Template_i_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.textFilter$.next(""); });
    ɵngcc0.ɵɵelementEnd();
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_28_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 25);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_28_Template_c8y_li_click_0_listener() { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r10); var configType_r8 = restoredCtx.$implicit; var ctx_r9 = ɵngcc0.ɵɵnextContext(2); ctx_r9.selectedConfigType = configType_r8; return ctx_r9.configType$.next(configType_r8.name); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 26);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var configType_r8 = ctx.$implicit;
    var ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("active", ctx_r4.selectedConfigType === configType_r8);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configType_r8.name)("pattern", ɵngcc0.ɵɵpipeBind1(2, 3, ctx_r4.configType$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_em_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 39);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Undefined`device type`"), " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_span_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 40);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "c8y-highlight", 41);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var configuration_r11 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r14 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 4, "Configuration type"), ": ", configuration_r11.configurationType, "");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("text", configuration_r11.configurationType)("pattern", ɵngcc0.ɵɵpipeBind1(3, 6, ctx_r14.configType$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_span_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 42);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(1, 3, "Configuration type"), ": ", ɵngcc0.ɵɵpipeBind1(2, 5, "Undefined"), "");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 7, "Undefined"), " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_Template(rf, ctx) { if (rf & 1) {
    var _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", null, 27);
    ɵngcc0.ɵɵelementStart(2, "c8y-li-radio", 28);
    ɵngcc0.ɵɵlistener("onSelect", function StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_Template_c8y_li_radio_onSelect_2_listener() { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r18); var configuration_r11 = restoredCtx.$implicit; var ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.selectConfiguration(configuration_r11); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "c8y-li-icon", 29);
    ɵngcc0.ɵɵelementStart(4, "c8y-li-body", 30);
    ɵngcc0.ɵɵelementStart(5, "div", 31);
    ɵngcc0.ɵɵelementStart(6, "div", 32);
    ɵngcc0.ɵɵelement(7, "c8y-highlight", 26);
    ɵngcc0.ɵɵpipe(8, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 33);
    ɵngcc0.ɵɵelementStart(10, "div", 32);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementStart(12, "span", 34);
    ɵngcc0.ɵɵtext(13, "Device type");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵtemplate(15, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_em_15_Template, 3, 3, "em", 35);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 36);
    ɵngcc0.ɵɵtemplate(17, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_span_17_Template, 4, 8, "span", 37);
    ɵngcc0.ɵɵtemplate(18, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_span_18_Template, 5, 9, "span", 38);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var configuration_r11 = ctx.$implicit;
    var ctx_r5 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", configuration_r11.name || "-");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r11.name || "-")("pattern", ɵngcc0.ɵɵpipeBind1(8, 9, ctx_r5.textFilter$));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(11, 11, "Device type"), ": ", ctx_r5.getDeviceTypeTitle(configuration_r11), "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", configuration_r11.deviceType, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !configuration_r11.deviceType);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", configuration_r11.configurationType);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !configuration_r11.configurationType);
} }
var _c0 = function (a0) { return { "dd-low": a0 }; };
function StepperBulkTypeConfigurationComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    var _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵelementStart(2, "div", 3);
    ɵngcc0.ɵɵelementStart(3, "div", 4);
    ɵngcc0.ɵɵelementStart(4, "h4", 5);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵelementStart(8, "div", 7);
    ɵngcc0.ɵɵelementStart(9, "div", 8);
    ɵngcc0.ɵɵelementStart(10, "input", 9);
    ɵngcc0.ɵɵlistener("ngModelChange", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_input_ngModelChange_10_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); var ctx_r19 = ɵngcc0.ɵɵnextContext(); return ctx_r19.textFilter$.next($event); });
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵpipe(13, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "span", 10);
    ɵngcc0.ɵɵtemplate(15, StepperBulkTypeConfigurationComponent_ng_container_1_i_15_Template, 1, 0, "i", 11);
    ɵngcc0.ɵɵpipe(16, "async");
    ɵngcc0.ɵɵtemplate(17, StepperBulkTypeConfigurationComponent_ng_container_1_i_17_Template, 1, 0, "i", 12);
    ɵngcc0.ɵɵpipe(18, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(19, "div", 13);
    ɵngcc0.ɵɵelementStart(20, "div", 7);
    ɵngcc0.ɵɵelementStart(21, "c8y-form-group", 14);
    ɵngcc0.ɵɵelementStart(22, "c8y-typeahead", 15);
    ɵngcc0.ɵɵlistener("ngModelChange", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_typeahead_ngModelChange_22_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); var ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.selectedConfigType = $event; })("onSearch", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_typeahead_onSearch_22_listener($event) { ɵngcc0.ɵɵrestoreView(_r20); var ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.configType$.next($event); });
    ɵngcc0.ɵɵpipe(23, "translate");
    ɵngcc0.ɵɵelementStart(24, "c8y-li", 16);
    ɵngcc0.ɵɵlistener("click", function StepperBulkTypeConfigurationComponent_ng_container_1_Template_c8y_li_click_24_listener() { ɵngcc0.ɵɵrestoreView(_r20); var ctx_r23 = ɵngcc0.ɵɵnextContext(); ctx_r23.selectedConfigType = { name: "" }; return ctx_r23.configType$.next(""); });
    ɵngcc0.ɵɵelementStart(25, "span");
    ɵngcc0.ɵɵtext(26);
    ɵngcc0.ɵɵpipe(27, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(28, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_28_Template, 3, 5, "c8y-li", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(29, "div", 18);
    ɵngcc0.ɵɵelementStart(30, "div", 19);
    ɵngcc0.ɵɵelementStart(31, "div", 20);
    ɵngcc0.ɵɵelementStart(32, "c8y-list-group", 21);
    ɵngcc0.ɵɵtemplate(33, StepperBulkTypeConfigurationComponent_ng_container_1_c8y_li_33_Template, 19, 13, "c8y-li", 22);
    ɵngcc0.ɵɵpipe(34, "async");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 12, "Select configuration"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 14, "Filter\u2026"));
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(12, 16, "Filter\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ɵngcc0.ɵɵpipeBind1(13, 18, ctx_r0.textFilter$));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(16, 20, ctx_r0.textFilter$).length === 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ɵngcc0.ɵɵpipeBind1(18, 22, ctx_r0.textFilter$).length > 0);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(23, 24, "Type to filter configuration types\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.selectedConfigType);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(27, 26, "All configuration types"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.configTypes);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(30, _c0, ctx_r0.elementCount < ctx_r0.DD_LOW_COUNT));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ɵngcc0.ɵɵpipeBind1(34, 28, ctx_r0.configurations$));
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_span_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 57);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r24 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r24.selectedConfiguration.configurationType, " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_span_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 58);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Undefined"), " ");
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_em_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "em", 59);
    ɵngcc0.ɵɵtext(1, " Undefined`device type` ");
    ɵngcc0.ɵɵelementEnd();
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_div_35_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 60);
    ɵngcc0.ɵɵelement(1, "h1", 61);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 62);
    ɵngcc0.ɵɵtext(4, "No preview available.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "br");
    ɵngcc0.ɵɵelementStart(6, "small", 62);
    ɵngcc0.ɵɵtext(7, "Could not fetch the file.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "file-code-o");
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_ng_template_36_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "pre", 63);
    ɵngcc0.ɵɵelementStart(1, "code");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r29 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r29.selectedConfiguration == null ? null : ctx_r29.selectedConfiguration.binary);
} }
function StepperBulkTypeConfigurationComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 43);
    ɵngcc0.ɵɵelementStart(2, "div", 44);
    ɵngcc0.ɵɵelementStart(3, "div", 6);
    ɵngcc0.ɵɵelementStart(4, "div", 4);
    ɵngcc0.ɵɵelementStart(5, "h4", 5);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(8, "div", 45);
    ɵngcc0.ɵɵelementStart(9, "div", 46);
    ɵngcc0.ɵɵelementStart(10, "div", 47);
    ɵngcc0.ɵɵelementStart(11, "strong");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(13, StepperBulkTypeConfigurationComponent_ng_container_3_span_13_Template, 2, 1, "span", 48);
    ɵngcc0.ɵɵtemplate(14, StepperBulkTypeConfigurationComponent_ng_container_3_span_14_Template, 3, 3, "span", 49);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 50);
    ɵngcc0.ɵɵelementStart(16, "div", 32);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementStart(18, "span", 34);
    ɵngcc0.ɵɵtext(19, " Device type ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(20);
    ɵngcc0.ɵɵtemplate(21, StepperBulkTypeConfigurationComponent_ng_container_3_em_21_Template, 2, 0, "em", 51);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(22, "p");
    ɵngcc0.ɵɵelementStart(23, "span", 34);
    ɵngcc0.ɵɵtext(24, "Updated");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "small");
    ɵngcc0.ɵɵtext(26);
    ɵngcc0.ɵɵpipe(27, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(28, "div", 18);
    ɵngcc0.ɵɵelementStart(29, "div", 19);
    ɵngcc0.ɵɵelementStart(30, "div", 52);
    ɵngcc0.ɵɵelementStart(31, "div", 6);
    ɵngcc0.ɵɵelementStart(32, "div", 53);
    ɵngcc0.ɵɵelementStart(33, "div", 54);
    ɵngcc0.ɵɵtext(34, "Preview");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(35, StepperBulkTypeConfigurationComponent_ng_container_3_div_35_Template, 8, 1, "div", 55);
    ɵngcc0.ɵɵtemplate(36, StepperBulkTypeConfigurationComponent_ng_container_3_ng_template_36_Template, 3, 1, "ng-template", null, 56, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var _r28 = ɵngcc0.ɵɵreference(37);
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 12, "Preview configuration"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.configurationType);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.configurationType));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(17, 14, "Device type"), ": ", ctx_r1.getDeviceTypeTitle(ctx_r1.selectedConfiguration), "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.deviceType, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.deviceType));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(27, 16, ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.lastUpdated));
    ɵngcc0.ɵɵadvance(9);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r1.selectedConfiguration == null ? null : ctx_r1.selectedConfiguration.binary))("ngIfElse", _r28);
} }
'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var StepperBulkTypeConfigurationComponent = /** @class */ (function (_super) {
        __extends(StepperBulkTypeConfigurationComponent, _super);
        function StepperBulkTypeConfigurationComponent(bulkOperationService, repositoryService, translate) {
            var _this = _super.call(this) || this;
            _this.bulkOperationService = bulkOperationService;
            _this.repositoryService = repositoryService;
            _this.translate = translate;
            _this.NO_DEVICE_TYPE_AVAILABLE = ngxComponents.gettext('Undefined`device type`');
            _this.elementCount = 0;
            _this.DD_LOW_COUNT = 10;
            _this.textFilter$ = new rxjs.BehaviorSubject('');
            _this.configType$ = new rxjs.BehaviorSubject('');
            _this.configTypes = [];
            _this.selectedConfigType = { name: '' };
            _this.configurations$ = rxjs.combineLatest(_this.textFilter$, _this.configType$).pipe(operators.switchMap(function (_a) {
                var _b = __read(_a, 2), name = _b[0], configType = _b[1];
                return _this.getConfiguration(name, configType);
            }), operators.shareReplay(1));
            _this.DD_LOW_COUNT = _this.bulkOperationService.DD_LOW_COUNT;
            _this.loadConfigurationTypes();
            return _this;
        }
        StepperBulkTypeConfigurationComponent.prototype.loadConfigurationTypes = function () {
            var _this = this;
            this.configTypeSubscription = this.configType$
                .pipe(operators.debounceTime(300), operators.distinctUntilChanged(), operators.switchMap(function (searchStr) {
                var query = { configurationType: "*" + searchStr + "*" };
                return rxjs.from(_this.repositoryService.listRepositoryEntries(repository.RepositoryType.CONFIGURATION, { query: query }));
            }))
                .subscribe(function (result) {
                var data = result.data;
                _this.configTypes = lodashEs.uniqWith(data.map(function (val) { return ({ name: val.configurationType }); }), lodashEs.isEqual);
            });
        };
        StepperBulkTypeConfigurationComponent.prototype.ngOnDestroy = function () {
            this.configTypeSubscription.unsubscribe();
        };
        StepperBulkTypeConfigurationComponent.prototype.selectConfiguration = function (configuration) {
            this.selectedConfiguration = configuration;
        };
        StepperBulkTypeConfigurationComponent.prototype.goToSecondStep = function ($event) {
            $event.stepper.next();
            this.getConfigBinary();
            this.deviceTypes = this.selectedConfiguration.deviceType;
        };
        StepperBulkTypeConfigurationComponent.prototype.getConfigBinary = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            if (!this.selectedConfiguration.url) return [3 /*break*/, 2];
                            _a = this.selectedConfiguration;
                            return [4 /*yield*/, this.repositoryService.getBinaryText(this.selectedConfiguration.url, { allowExternal: true, noAlerts: true })];
                        case 1:
                            _a.binary = _b.sent();
                            _b.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        StepperBulkTypeConfigurationComponent.prototype.getDeviceTypeTitle = function (configuration) {
            return lodashEs.get(configuration, 'deviceType', this.translate.instant(this.NO_DEVICE_TYPE_AVAILABLE));
        };
        StepperBulkTypeConfigurationComponent.prototype.retrieveOperationPrototype = function () {
            var configuration = {
                type: this.selectedConfiguration.configurationType,
                url: this.selectedConfiguration.url
            };
            return {
                name: ngxComponents.gettext('Configuration update'),
                description: lodashEs.get(this.selectedConfiguration, 'name'),
                prototype: {
                    description: "Update configuration to: " + this.selectedConfiguration.name + ".",
                    c8y_DownloadConfigFile: configuration
                }
            };
        };
        StepperBulkTypeConfigurationComponent.prototype.getConfiguration = function (name, configurationType) {
            var query = name ? { name: "*" + name + "*" } : {};
            if (configurationType) {
                query.__or = [{ configurationType: configurationType }, { __not: { __has: "configurationType" } }];
            }
            return this.repositoryService.listRepositoryEntries(repository.RepositoryType.CONFIGURATION, { query: query });
        };
StepperBulkTypeConfigurationComponent.ɵfac = function StepperBulkTypeConfigurationComponent_Factory(t) { return new (t || StepperBulkTypeConfigurationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.TranslateService)); };
StepperBulkTypeConfigurationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StepperBulkTypeConfigurationComponent, selectors: [["c8y-stepper-bulk-type-configuration"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 9, consts: [[4, "customStep", "customStepCompleted", "customStepButtonsDisabled", "customStepOnNext"], [4, "customStep"], [1, "card-block", "p-t-0", "overflow-visible", "flex-no-shrink", "separator-bottom", "col-xs-12"], [1, "row", "p-b-16"], [1, "col-md-8", "col-md-offset-2", "col-lg-6", "col-lg-offset-3"], [1, "text-center", "m-b-16"], [1, "row"], [1, "col-md-6"], [1, "input-group", "input-group-search"], ["type", "search", 1, "form-control", 3, "title", "placeholder", "ngModel", "ngModelChange"], [1, "input-group-addon"], ["c8yIcon", "filter", 4, "ngIf"], ["c8yIcon", "times", "class", "text-muted", 3, "click", 4, "ngIf"], [1, "col-xs-12", "p-b-8", "visible-xs", "visible-sm"], [1, "m-0"], ["name", "configType", 3, "ngModel", "placeholder", "ngModelChange", "onSearch"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "click"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "active", "click", 4, "ngFor", "ngForOf"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-t-0", "p-b-0"], [3, "ngClass"], [4, "c8yFor", "c8yForOf"], ["c8yIcon", "filter"], ["c8yIcon", "times", 1, "text-muted", 3, "click"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "active", "click"], [3, "text", "pattern"], ["listItem", ""], [3, "onSelect"], ["icon", "gears"], [1, "content-flex-60"], [1, "col-5"], [1, "text-truncate", 3, "title"], [1, "col-4"], ["translate", "", 1, "text-label-small", "m-r-4"], ["class", "text-muted", 4, "ngIf"], [1, "col-3"], ["class", "label label-info", 3, "title", 4, "ngIf"], ["class", "label label-default", 3, "title", 4, "ngIf"], [1, "text-muted"], [1, "label", "label-info", 3, "title"], ["elementClass", "text-gray-lighter", 3, "text", "pattern"], [1, "label", "label-default", 3, "title"], [1, "d-contents"], [1, "card-block", "p-t-0", "flex-no-shrink", "separator-bottom"], [1, "row", "no-gutter"], [1, "col-xs-6"], [1, "text-right", "text-truncate", 3, "title"], ["class", "label label-primary", 4, "ngIf"], ["class", "label label-default", 4, "ngIf"], [1, "col-xs-6", "text-right"], ["class", "text-muted", "translate", "", 4, "ngIf"], [1, "card-block", "p-t-0"], [1, "col-md-8", "col-md-offset-2"], ["translate", "", 1, "legend", "form-block"], ["class", "c8y-empty-state text-left", 4, "ngIf", "ngIfElse"], ["binaryPreview", ""], [1, "label", "label-primary"], [1, "label", "label-default"], ["translate", "", 1, "text-muted"], [1, "c8y-empty-state", "text-left"], [3, "c8yIcon"], ["translate", ""], [2, "min-height", "98px"]], template: function StepperBulkTypeConfigurationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-bulk-operation-stepper");
        ɵngcc0.ɵɵtemplate(1, StepperBulkTypeConfigurationComponent_ng_container_1_Template, 35, 32, "ng-container", 0);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, StepperBulkTypeConfigurationComponent_ng_container_3_Template, 38, 18, "ng-container", 1);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(2, 5, "Select configuration"))("customStepCompleted", !!ctx.selectedConfiguration)("customStepButtonsDisabled", !ctx.selectedConfiguration)("customStepOnNext", ctx.goToSecondStep.bind(ctx));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("customStep", ɵngcc0.ɵɵpipeBind1(4, 7, "Preview configuration"));
    } }, directives: [ɵngcc4.BulkOperationStepper, ɵngcc4.CustomStep, ɵngcc5.DefaultValueAccessor, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc6.NgIf, ɵngcc7.FormGroupComponent, ɵngcc7.TypeaheadComponent, ɵngcc7.ListItemComponent, ɵngcc6.NgForOf, ɵngcc7.ListGroupComponent, ɵngcc6.NgClass, ɵngcc7.ForOfDirective, ɵngcc7.IconDirective, ɵngcc7.HighlightComponent, ɵngcc7.ListItemRadioComponent, ɵngcc7.ListItemIconComponent, ɵngcc7.ListItemBodyComponent, ɵngcc7.C8yTranslateDirective], pipes: [ɵngcc7.C8yTranslatePipe, ɵngcc6.AsyncPipe, ɵngcc7.DatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeConfigurationComponent, [{
        type: core.Component,
        args: [{
                selector: 'c8y-stepper-bulk-type-configuration',
                template: "<c8y-bulk-operation-stepper>\n  <ng-container\n    *customStep=\"\n      'Select configuration' | translate;\n      completed: !!selectedConfiguration;\n      buttonsDisabled: !selectedConfiguration;\n      onNext: goToSecondStep.bind(this)\"\n  >\n    <div class=\"card-block p-t-0 overflow-visible flex-no-shrink separator-bottom col-xs-12\">\n      <div class=\"row p-b-16\">\n        <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3\">\n          <h4 class=\"text-center m-b-16\">{{ 'Select configuration' | translate }}</h4>\n          <div class=\"row\">\n            <div class=\"col-md-6\">\n              <div class=\"input-group input-group-search\">\n                <input\n                  type=\"search\"\n                  class=\"form-control\"\n                  title=\"{{ 'Filter\u2026' | translate }}\"\n                  placeholder=\"{{ 'Filter\u2026' | translate }}\"\n                  [ngModel]=\"textFilter$ | async\"\n                  (ngModelChange)=\"textFilter$.next($event)\"\n                />\n                <span class=\"input-group-addon\">\n                  <i c8yIcon=\"filter\" *ngIf=\"(textFilter$ | async).length === 0\"></i>\n                  <i\n                    c8yIcon=\"times\"\n                    class=\"text-muted\"\n                    *ngIf=\"(textFilter$ | async).length > 0\"\n                    (click)=\"textFilter$.next('')\"\n                  ></i>\n                </span>\n              </div>\n            </div>\n            <div class=\"col-xs-12 p-b-8 visible-xs visible-sm\"></div>\n            <div class=\"col-md-6\">\n              <c8y-form-group class=\"m-0\">\n                <c8y-typeahead\n                  name=\"configType\"\n                  [(ngModel)]=\"selectedConfigType\"\n                  placeholder=\"{{ 'Type to filter configuration types\u2026' | translate }}\"\n                  (onSearch)=\"configType$.next($event)\"\n                >\n                  <c8y-li\n                    class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                    (click)=\"selectedConfigType = { name: '' }; configType$.next('')\"\n                  >\n                    <span>{{ 'All configuration types' | translate }}</span>\n                  </c8y-li>\n                  <c8y-li\n                    *ngFor=\"let configType of configTypes\"\n                    class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                    (click)=\"selectedConfigType = configType; configType$.next(configType.name)\"\n                    [active]=\"selectedConfigType === configType\"\n                  >\n                    <c8y-highlight\n                      [text]=\"configType.name\"\n                      [pattern]=\"configType$ | async\"\n                    ></c8y-highlight>\n                  </c8y-li>\n                </c8y-typeahead>\n              </c8y-form-group>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n    <div class=\"col-xs-12 flex-grow no-gutter\">\n      <div class=\"card-inner-scroll fit-h\">\n        <div class=\"card-block p-t-0 p-b-0\">\n          <c8y-list-group [ngClass]=\"{ 'dd-low': elementCount < DD_LOW_COUNT }\">\n            <c8y-li #listItem *c8yFor=\"let configuration of configurations$ | async\">\n              <c8y-li-radio (onSelect)=\"selectConfiguration(configuration)\"></c8y-li-radio>\n              <c8y-li-icon icon=\"gears\"></c8y-li-icon>\n              <c8y-li-body class=\"content-flex-60\">\n                <div class=\"col-5\">\n                  <div class=\"text-truncate\" title=\"{{ configuration.name || '-' }}\">\n                    <c8y-highlight\n                      [text]=\"configuration.name || '-'\"\n                      [pattern]=\"textFilter$ | async\"\n                    ></c8y-highlight>\n                  </div>\n                </div>\n                <div class=\"col-4\">\n                  <div\n                    class=\"text-truncate\"\n                    title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(configuration) }}\"\n                  >\n                    <span translate class=\"text-label-small m-r-4\">Device type</span>\n                    {{ configuration.deviceType }}\n                    <em class=\"text-muted\" *ngIf=\"!configuration.deviceType\">\n                      {{ 'Undefined`device type`' | translate }}\n                    </em>\n                  </div>\n                </div>\n                <div class=\"col-3\">\n                  <span\n                    class=\"label label-info\"\n                    *ngIf=\"configuration.configurationType\"\n                    title=\"{{ 'Configuration type' | translate }}: {{\n                      configuration.configurationType\n                    }}\"\n                  >\n                    <c8y-highlight\n                      [text]=\"configuration.configurationType\"\n                      elementClass=\"text-gray-lighter\"\n                      [pattern]=\"configType$ | async\"\n                    ></c8y-highlight>\n                  </span>\n                  <span\n                    class=\"label label-default\"\n                    *ngIf=\"!configuration.configurationType\"\n                    title=\"{{ 'Configuration type' | translate }}: {{ 'Undefined' | translate }}\"\n                  >\n                    {{ 'Undefined' | translate }}\n                  </span>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </div>\n      </div>\n    </div>\n  </ng-container>\n  <ng-container *customStep=\"'Preview configuration' | translate\">\n    <div class=\"d-contents\">\n      <div class=\"card-block p-t-0 flex-no-shrink separator-bottom\">\n        <div class=\"row\">\n          <div class=\"col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3\">\n            <h4 class=\"text-center m-b-16\">\n              {{ 'Preview configuration' | translate }}\n            </h4>\n            <div class=\"row no-gutter\">\n              <div class=\"col-xs-6\">\n                <div class=\"text-right text-truncate\" title=\"{{ selectedConfiguration?.name }}\">\n                  <strong>{{ selectedConfiguration?.name }}</strong>\n                </div>\n                <span class=\"label label-primary\" *ngIf=\"selectedConfiguration?.configurationType\">\n                  {{ selectedConfiguration.configurationType }}\n                </span>\n                <span class=\"label label-default\" *ngIf=\"!selectedConfiguration?.configurationType\">\n                  {{ 'Undefined' | translate }}\n                </span>\n              </div>\n              <div class=\"col-xs-6 text-right\">\n                <div class=\"text-truncate\" title=\"{{ 'Device type' | translate }}: {{ getDeviceTypeTitle(selectedConfiguration) }}\">\n                  <span translate class=\"text-label-small m-r-4\">\n                    Device type\n                  </span>\n                  {{ selectedConfiguration?.deviceType }}\n                  <em class=\"text-muted\" *ngIf=\"!selectedConfiguration?.deviceType\" translate>\n                    Undefined`device type`\n                  </em>\n                </div>\n                <p>\n                  <span translate class=\"text-label-small m-r-4\">Updated</span>\n                  <small>{{ selectedConfiguration?.lastUpdated | c8yDate }}</small>\n                </p>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-xs-12 flex-grow no-gutter\">\n        <div class=\"card-inner-scroll fit-h\">\n          <div class=\"card-block p-t-0\">\n            <div class=\"row\">\n              <div class=\"col-md-8 col-md-offset-2\">\n                <div class=\"legend form-block\" translate>Preview</div>\n                <div\n                  class=\"c8y-empty-state text-left\"\n                  *ngIf=\"!this.selectedConfiguration?.binary; else binaryPreview\"\n                >\n                  <h1 [c8yIcon]=\"'file-code-o'\"></h1>\n                  <p>\n                    <strong translate>No preview available.</strong><br />\n                    <small translate>Could not fetch the file.</small>\n                  </p>\n                </div>\n                <ng-template #binaryPreview>\n                  <pre\n                    style=\"min-height: 98px;\"\n                  ><code>{{ this.selectedConfiguration?.binary }}</code></pre>\n                </ng-template>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n    </div>\n  </ng-container>\n</c8y-bulk-operation-stepper>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BulkOperationsService }, { type: ɵngcc2.RepositoryService }, { type: ɵngcc3.TranslateService }]; }, null); })();
        return StepperBulkTypeConfigurationComponent;
    }(bulkOperationStepper.BaseStepperComponent));
    StepperBulkTypeConfigurationComponent.ctorParameters = function () { return [
        { type: bulkOperationsService.BulkOperationsService },
        { type: repository.RepositoryService },
        { type: core$1.TranslateService }
    ]; };

    var ɵ0 = {
        type: bulkOperationsService.BulkOperationType.CONFIGURATION,
        c8yIcon: 'cogs',
        name: ngxComponents.gettext('Configuration update'),
        path: bulkOperationsService.baseUrl + "configuration",
        component: StepperBulkTypeConfigurationComponent,
        fragments: ['c8y_DownloadConfigFile', 'c8y_Configuration'],
        selected: false
    };
    /** Module for the 'Configuration update' operation type stepper */
    var StepperBulkTypeConfigurationModule = /** @class */ (function () {
        function StepperBulkTypeConfigurationModule() {
        }
StepperBulkTypeConfigurationModule.ɵfac = function StepperBulkTypeConfigurationModule_Factory(t) { return new (t || StepperBulkTypeConfigurationModule)(); };
StepperBulkTypeConfigurationModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: StepperBulkTypeConfigurationModule });
StepperBulkTypeConfigurationModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: bulkOperationsService.HOOK_LIST_BULK_TYPE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[
            ngxComponents.CoreModule,
            ngxComponents.FormsModule,
            forms.ReactiveFormsModule,
            bulkOperationStepper.BulkOperationStepperModule,
            repository.RepositoryModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StepperBulkTypeConfigurationModule, [{
        type: core.NgModule,
        args: [{
                declarations: [StepperBulkTypeConfigurationComponent],
                imports: [
                    ngxComponents.CoreModule,
                    ngxComponents.FormsModule,
                    forms.ReactiveFormsModule,
                    bulkOperationStepper.BulkOperationStepperModule,
                    repository.RepositoryModule
                ],
                providers: [
                    {
                        provide: bulkOperationsService.HOOK_LIST_BULK_TYPE,
                        useValue: ɵ0,
                        multi: true
                    }
                ],
                entryComponents: [StepperBulkTypeConfigurationComponent],
                exports: [StepperBulkTypeConfigurationComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(StepperBulkTypeConfigurationModule, { declarations: [StepperBulkTypeConfigurationComponent], imports: [ɵngcc7.CoreModule, ɵngcc7.FormsModule, ɵngcc5.ReactiveFormsModule, ɵngcc4.BulkOperationStepperModule, ɵngcc2.RepositoryModule], exports: [StepperBulkTypeConfigurationComponent] }); })();
        return StepperBulkTypeConfigurationModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.StepperBulkTypeConfigurationComponent = StepperBulkTypeConfigurationComponent;
    exports.StepperBulkTypeConfigurationModule = StepperBulkTypeConfigurationModule;
    exports["ɵ0"] = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
