// @c8y/ngx-components/bundles/c8y-ngx-components-search.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/router'),require('@c8y/client'),require('@c8y/ngx-components/device-grid'),require('ngx-bootstrap/dropdown'),require('@c8y/ngx-components'),require('@angular/forms'),require('@angular/common'),require('ngx-bootstrap/modal'),require('@c8y/ngx-components/sub-assets'),exports, require('@angular/core'), require('@c8y/ngx-components/sub-assets'), require('lodash-es'), require('@ngx-translate/core'), require('@c8y/client'), require('@c8y/ngx-components'), require('@c8y/client'), require('@c8y/ngx-components/assets-navigator'), require('@c8y/client'), require('@c8y/client'), require('@angular/router'), require('@c8y/client'), require('@c8y/ngx-components/device-grid'), require('rxjs'), require('rxjs/operators'), require('ngx-bootstrap/dropdown'), require('ngx-bootstrap/modal')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/search', ['@angular/core','@angular/router','@c8y/client','@c8y/ngx-components/device-grid','ngx-bootstrap/dropdown','@c8y/ngx-components','@angular/forms','@angular/common','ngx-bootstrap/modal','@c8y/ngx-components/sub-assets','exports', '@angular/core', '@c8y/ngx-components/sub-assets', 'lodash-es', '@ngx-translate/core', '@c8y/client', '@c8y/ngx-components', '@c8y/client', '@c8y/ngx-components/assets-navigator', '@c8y/client', '@c8y/client', '@angular/router', '@c8y/client', '@c8y/ngx-components/device-grid', 'rxjs', 'rxjs/operators', 'ngx-bootstrap/dropdown', 'ngx-bootstrap/modal'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.router,global.c8y.client,global.c8y.ngxComponents.deviceGrid,global.ngxBootstrap.dropdown,global.c8y.ngxComponents,global.ng.forms,global.ng.common,global.ngxBootstrap.modal,global.c8y.ngxComponents.subAssets,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].search = {}), global.ng.core, global.c8y["ngx-components"]["sub-assets"], global.lodashEs, global.i1, global.i2, global.c8y["ngx-components"], global.i4, global.i5, global.i6, global.i7, global.ng.router, global.client, global.c8y["ngx-components"]["device-grid"], global.rxjs, global.rxjs.operators, global.dropdown, global.modal));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,exports, i0, subAssets, lodashEs, i1, i2, i3, i4, i5, i6, i7, router, client, deviceGrid, rxjs, operators, dropdown, modal) { 
var _c0 = ["dropdown"];
function SearchBoxComponent_div_4_c8y_li_5_Template(rf, ctx) { if (rf & 1) {
    var _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 22);
    ɵngcc0.ɵɵelementStart(1, "div", 18);
    ɵngcc0.ɵɵelementStart(2, "p", 23);
    ɵngcc0.ɵɵelementStart(3, "em", 24);
    ɵngcc0.ɵɵtext(4, "Searching by exact match. Click for other search options:");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 25);
    ɵngcc0.ɵɵelementStart(6, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r16); var ctx_r15 = ɵngcc0.ɵɵnextContext(2); return ctx_r15.filter(ctx_r15.term + "*"); });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_10_listener() { ɵngcc0.ɵɵrestoreView(_r16); var ctx_r17 = ɵngcc0.ɵɵnextContext(2); return ctx_r17.filter("*" + ctx_r17.term + "*"); });
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(14, "button", 26);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_5_Template_button_click_14_listener() { ɵngcc0.ɵɵrestoreView(_r16); var ctx_r18 = ɵngcc0.ɵɵnextContext(2); return ctx_r18.filter("*" + ctx_r18.term); });
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵtext(16);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 6, "Starts with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 8, "Starts with"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(11, 10, "Contains"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 12, "Contains"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(15, 14, "Ends with"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 16, "Ends with"), " ");
} }
function SearchBoxComponent_div_4_c8y_li_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Recent search results");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_7_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    var result_r19 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r19);
} }
function SearchBoxComponent_div_4_c8y_li_7_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_7_Template(rf, ctx) { if (rf & 1) {
    var _r24 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_7_Template_c8y_li_click_0_listener($event) { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r24); var result_r19 = restoredCtx.$implicit; var ctx_r23 = ɵngcc0.ɵɵnextContext(2); return ctx_r23.open($event, result_r19); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_7_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_7_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var result_r19 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r19.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r19.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r19.name || "--", " ");
} }
function SearchBoxComponent_div_4_c8y_li_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Recently registered devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_10_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    var result_r25 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r25);
} }
function SearchBoxComponent_div_4_c8y_li_10_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_10_Template(rf, ctx) { if (rf & 1) {
    var _r30 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_10_Template_c8y_li_click_0_listener($event) { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r30); var result_r25 = restoredCtx.$implicit; var ctx_r29 = ɵngcc0.ɵɵnextContext(2); return ctx_r29.open($event, result_r25); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_10_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_10_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var result_r25 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r25.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r25.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r25.name || "--", " ");
} }
function SearchBoxComponent_div_4_c8y_li_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 27);
    ɵngcc0.ɵɵelementStart(1, "div", 28);
    ɵngcc0.ɵɵelementStart(2, "span", 24);
    ɵngcc0.ɵɵtext(3, "Search results");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchBoxComponent_div_4_c8y_li_12_device_status_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 32);
} if (rf & 2) {
    var result_r31 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("mo", result_r31);
} }
function SearchBoxComponent_div_4_c8y_li_12_i_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 33);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-group-open");
} }
function SearchBoxComponent_div_4_c8y_li_12_Template(rf, ctx) { if (rf & 1) {
    var _r36 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 29);
    ɵngcc0.ɵɵlistener("click", function SearchBoxComponent_div_4_c8y_li_12_Template_c8y_li_click_0_listener($event) { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r36); var result_r31 = restoredCtx.$implicit; var ctx_r35 = ɵngcc0.ɵɵnextContext(2); return ctx_r35.open($event, result_r31); });
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵtemplate(2, SearchBoxComponent_div_4_c8y_li_12_device_status_2_Template, 1, 1, "device-status", 30);
    ɵngcc0.ɵɵtemplate(3, SearchBoxComponent_div_4_c8y_li_12_i_3_Template, 1, 1, "i", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var result_r31 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !result_r31.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", result_r31.c8y_IsDeviceGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", result_r31.name || "--", " ");
} }
function SearchBoxComponent_div_4_ng_template_13_c8y_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 35);
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 36);
    ɵngcc0.ɵɵelementStart(2, "p");
    ɵngcc0.ɵɵelementStart(3, "strong", 24);
    ɵngcc0.ɵɵtext(4, "No match found.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "small", 24);
    ɵngcc0.ɵɵtext(6, " Use a filter or go to the asset data table to show all devices and groups. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("selectable", false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "search");
} }
function SearchBoxComponent_div_4_ng_template_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, SearchBoxComponent_div_4_ng_template_13_c8y_li_0_Template, 7, 2, "c8y-li", 34);
} if (rf & 2) {
    var ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.noMatch);
} }
function SearchBoxComponent_div_4_c8y_li_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 37);
    ɵngcc0.ɵɵelementStart(1, "div", 38);
    ɵngcc0.ɵɵelement(2, "div", 39);
    ɵngcc0.ɵɵelement(3, "div", 40);
    ɵngcc0.ɵɵelement(4, "div", 41);
    ɵngcc0.ɵɵelement(5, "div", 42);
    ɵngcc0.ɵɵelement(6, "div", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SearchBoxComponent_div_4_ng_template_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 44);
    ɵngcc0.ɵɵelementStart(1, "div", 38);
    ɵngcc0.ɵɵelement(2, "div", 39);
    ɵngcc0.ɵɵelement(3, "div", 40);
    ɵngcc0.ɵɵelement(4, "div", 41);
    ɵngcc0.ɵɵelement(5, "div", 42);
    ɵngcc0.ɵɵelement(6, "div", 43);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
var _c1 = function () { return []; };
var _c2 = function (a0) { return { data: a0 }; };
function SearchBoxComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    var _r39 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵelementStart(1, "form", 6, 7);
    ɵngcc0.ɵɵelementStart(3, "c8y-typeahead", 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function SearchBoxComponent_div_4_Template_c8y_typeahead_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r39); var ctx_r38 = ɵngcc0.ɵɵnextContext(); return ctx_r38.selected = $event; })("keydown", function SearchBoxComponent_div_4_Template_c8y_typeahead_keydown_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r39); var ctx_r40 = ɵngcc0.ɵɵnextContext(); return ctx_r40.keyDown($event); })("onIconClick", function SearchBoxComponent_div_4_Template_c8y_typeahead_onIconClick_3_listener() { ɵngcc0.ɵɵrestoreView(_r39); var ctx_r41 = ɵngcc0.ɵɵnextContext(); return ctx_r41.reset(); });
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵtemplate(5, SearchBoxComponent_div_4_c8y_li_5_Template, 18, 18, "c8y-li", 9);
    ɵngcc0.ɵɵtemplate(6, SearchBoxComponent_div_4_c8y_li_6_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵtemplate(7, SearchBoxComponent_div_4_c8y_li_7_Template, 5, 3, "c8y-li", 11);
    ɵngcc0.ɵɵtemplate(8, SearchBoxComponent_div_4_c8y_li_8_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵpipe(9, "async");
    ɵngcc0.ɵɵtemplate(10, SearchBoxComponent_div_4_c8y_li_10_Template, 5, 3, "c8y-li", 12);
    ɵngcc0.ɵɵtemplate(11, SearchBoxComponent_div_4_c8y_li_11_Template, 4, 1, "c8y-li", 10);
    ɵngcc0.ɵɵtemplate(12, SearchBoxComponent_div_4_c8y_li_12_Template, 5, 3, "c8y-li", 13);
    ɵngcc0.ɵɵtemplate(13, SearchBoxComponent_div_4_ng_template_13_Template, 1, 1, "ng-template", null, 14, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(15, SearchBoxComponent_div_4_c8y_li_15_Template, 7, 0, "c8y-li", 15);
    ɵngcc0.ɵɵtemplate(16, SearchBoxComponent_div_4_ng_template_16_Template, 7, 0, "ng-template", null, 16, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementStart(18, "c8y-li", 17);
    ɵngcc0.ɵɵelementStart(19, "div", 18);
    ɵngcc0.ɵɵelement(20, "i", 19);
    ɵngcc0.ɵɵelementStart(21, "p", 20);
    ɵngcc0.ɵɵtext(22, "Need more filter possibilities?");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(23, "button", 21);
    ɵngcc0.ɵɵlistener("mousedown", function SearchBoxComponent_div_4_Template_button_mousedown_23_listener() { ɵngcc0.ɵɵrestoreView(_r39); var ctx_r42 = ɵngcc0.ɵɵnextContext(); return ctx_r42.openSearch(); });
    ɵngcc0.ɵɵtext(24, " Go to the asset data table ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r10 = ɵngcc0.ɵɵreference(14);
    var _r13 = ɵngcc0.ɵɵreference(17);
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    var tmp_7_0;
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(4, 20, "Search for groups or assets\u2026"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.selected)("icon", ctx_r1.term ? "times" : "search")("allowFreeEntries", false);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length === 0 && ctx_r1.recentSearchResults.length > 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.term.length === 0 ? ctx_r1.recentSearchResults : ɵngcc0.ɵɵpureFunction0(24, _c1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length === 0 && ((tmp_7_0 = ɵngcc0.ɵɵpipeBind1(9, 22, ctx_r1.recentlyRegisteredResults$)) == null ? null : tmp_7_0.data == null ? null : tmp_7_0.data.length) > 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r1.term.length === 0 ? ctx_r1.recentlyRegisteredResults$ : ɵngcc0.ɵɵpureFunction1(26, _c2, ɵngcc0.ɵɵpureFunction0(25, _c1)))("c8yForLoadMore", "none")("c8yForPipe", ctx_r1.filterPipe);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.term.length !== 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r1.results$)("c8yForLoadMore", "auto")("c8yForPipe", ctx_r1.filterPipe)("c8yForNotFound", _r10)("c8yForLoadingTemplate", _r13)("c8yForLoadNextLabel", "Find more\u2026");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.isLoading);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("selectable", false);
} }
function SearchGridComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-column", 5);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", column_r1.name);
} }
var _c3 = function (a0) { return { searchHint: a0 }; };
function SearchResultsComponent_small_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 3);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, "searching \"{{ searchHint }}\"");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c3, ctx_r0.searchTerm));
} }
var _c4 = function (a0) { return { filterHint: a0 }; };
function SearchResultsComponent_small_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 3);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(1, "filtered by \"{{ filterHint }}\"");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("translateParams", ɵngcc0.ɵɵpureFunction1(1, _c4, ctx_r1.filter));
} }
'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var AssetSearchService = /** @class */ (function (_super) {
        __extends(AssetSearchService, _super);
        function AssetSearchService() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.GRID_CONFIG_STORAGE_KEY = 'search-grid-config';
            _this.DEFAULT_PAGE_SIZE = 50;
            _this.FRAGMENTS_FOR_NO_DEVICE = [
                'c8y_Dashboard',
                'c8y_Report',
                'c8y_Kpi',
                'c8y_ExportConfiguration',
                'c8y_IsBinary',
                'c8y_NoDevice',
                'c8y_IsDeviceGroup',
                'c8y_Group',
                'com_cumulocity_model_smartrest_SmartRestTemplate',
                'com_cumulocity_model_devicesimulator_SensorTemplate',
                '_attachments',
                'c8y_IsDeviceType',
                'c8y_objectmapping_ObjectMapping'
            ];
            _this.TYPES_FOR_NO_DEVICE = [
                'c8y_ConfigurationDump',
                'c8y_Firmware',
                'c8y_SmartRule',
                'c8y_Software',
                'impact_object_mapping',
                'c8y_UserPreference',
                'c8y_TenantPolicy',
                'c8y_PrivateSmartRule',
                'c8y_SmartRest2Template',
                'c8y_JsonSchema',
                'c8y_DeviceShellTemplate',
                'c8y_DemoStatus',
                'c8y_DataBroker',
                'c8y_Application_',
                'brandingVariables',
                'c8y_DeviceSimulator',
                'c8y_CertificateMetadata',
                'lwm2m_post_registration',
                'c8y_microservice_manifest_',
                'c8y_CepAgent'
            ];
            _this.DEVICE_FRAGMENT_TYPE = 'c8y_IsDevice';
            return _this;
        }
        /**
         * Will return only valid assets (groups and devices) and filter out
         * none useful inventories (e.g. c8y_JsonSchema).
         * @param data All managed objects that should be filtered.
         */
        AssetSearchService.prototype.filterOnlyAssets = function (data) {
            var _this = this;
            return data.filter(function (mo) { return mo.c8y_IsDeviceGroup || _this.isAnyDevice(mo); });
        };
        AssetSearchService.prototype.buildCombinedRootQueryFilter = function (columns, pagination) {
            var rootQuery = {
                __filter: {
                    __and: { __not: { __has: "c8y_IsBinary" } }
                }
            };
            var userQuery = this.getQueryObj(columns, pagination);
            var queryPart = this.queriesUtil.addOrderbys(rootQuery, userQuery.__orderby, 'append');
            var fullQuery = this.queriesUtil.addAndFilter(queryPart, userQuery.__filter);
            return this.queriesUtil.buildQuery(fullQuery);
        };
        /**
         * Returns the full-text search results.
         *
         * @param term The search term.
         * @param pagination The currently used pagination.
         */
        AssetSearchService.prototype.search = function (term, pagination) {
            if (pagination === void 0) { pagination = { currentPage: 1, pageSize: this.DEFAULT_PAGE_SIZE }; }
            return this.inventoryService.list({
                text: term,
                withTotalPages: true,
                pageSize: pagination.pageSize,
                withChildren: false,
                currentPage: pagination.currentPage || 1
            });
        };
        AssetSearchService.prototype.isRootDevice = function (mo) {
            return !!mo[this.DEVICE_FRAGMENT_TYPE];
        };
        AssetSearchService.prototype.isAnyDevice = function (mo) {
            var isDevice = this.isRootDevice(mo) || !this.hasFragmentOrTypeFromBlacklist(mo);
            return isDevice;
        };
        AssetSearchService.prototype.hasFragmentOrTypeFromBlacklist = function (mo) {
            return this.hasTypeFromBlacklist(mo) || this.hasFragmentFromBlacklist(mo);
        };
        AssetSearchService.prototype.hasTypeFromBlacklist = function (mo) {
            var moType = lodashEs.get(mo, 'type', '');
            return lodashEs.some(this.TYPES_FOR_NO_DEVICE, function (type) { return moType.includes(type); });
        };
        AssetSearchService.prototype.hasFragmentFromBlacklist = function (mo) {
            return lodashEs.find(this.FRAGMENTS_FOR_NO_DEVICE, function (f) { return !lodashEs.isUndefined(mo[f]); });
        };
AssetSearchService.ɵfac = /*@__PURE__*/ function () { var ɵAssetSearchService_BaseFactory; return function AssetSearchService_Factory(t) { return (ɵAssetSearchService_BaseFactory || (ɵAssetSearchService_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(AssetSearchService)))(t || AssetSearchService); }; }();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSearchService, [{
        type: i0.Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
        return AssetSearchService;
    }(subAssets.SubAssetsService));
    AssetSearchService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function AssetSearchService_Factory() { return new AssetSearchService(i0__namespace.ɵɵinject(i1__namespace.TranslateService), i0__namespace.ɵɵinject(i2__namespace.InventoryService), i0__namespace.ɵɵinject(i3__namespace.AppStateService), i0__namespace.ɵɵinject(i4__namespace.UserService), i0__namespace.ɵɵinject(i5__namespace.AssetNodeService), i0__namespace.ɵɵinject(i5__namespace.DeviceGroupService), i0__namespace.ɵɵinject(i6__namespace.SmartGroupsService), i0__namespace.ɵɵinject(i7__namespace.SmartRulesService), i0__namespace.ɵɵinject(i3__namespace.AlertService), i0__namespace.ɵɵinject(i3__namespace.Permissions), i0__namespace.ɵɵinject(i3__namespace.ModalService)); }, token: AssetSearchService, providedIn: "root" });

    var SearchBoxComponent = /** @class */ (function () {
        function SearchBoxComponent(router, inventory, deviceGridService, searchService) {
            var _this = this;
            this.router = router;
            this.inventory = inventory;
            this.deviceGridService = deviceGridService;
            this.searchService = searchService;
            this.term = '';
            this.filterPipe = rxjs.pipe(operators.map(function (data) {
                return _this.searchService.filterOnlyAssets(data);
            }));
            this.recentSearchResults = [];
            this.isLoading = false;
            this.noMatch = false;
            this.RECENT_SEARCH_STORAGE_KEY = 'recent_search_view';
            this.MAX_RECENT_SEARCH_RESULTS = 5;
            this.DEFAULT_FILTER = {
                withTotalPages: true,
                pageSize: 5,
                withChildren: false
            };
            this.KEYCODE_ENTER = 13;
            this.KEYCODE_ESC = 27;
        }
        SearchBoxComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var recentSearchIds, data;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            recentSearchIds = JSON.parse(localStorage.getItem(this.RECENT_SEARCH_STORAGE_KEY));
                            if (!(recentSearchIds && recentSearchIds.length > 0)) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.inventory.list({ ids: recentSearchIds.join(',') })];
                        case 1:
                            data = (_a.sent()).data;
                            this.recentSearchResults = data;
                            this.recentlyRegisteredResults$ = rxjs.defer(function () { return _this.inventory.list(Object.assign({ q: '$orderby=creationTime desc' }, _this.DEFAULT_FILTER)); });
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            });
        };
        SearchBoxComponent.prototype.onOpenChange = function (isOpen) {
            var _this = this;
            if (isOpen) {
                // needs to request an animation frame as
                // otherwise the typeahead is undefined
                requestAnimationFrame(function () {
                    _this.subscribeOnSearch();
                    _this.typeahead.dropdown.show();
                    _this.typeahead.searchControl.nativeElement.focus();
                });
            }
        };
        SearchBoxComponent.prototype.open = function (event, mo) {
            event.stopPropagation();
            var isAlreadyRecent = this.recentSearchResults.find(function (_a) {
                var id = _a.id;
                return id === mo.id;
            });
            if (!isAlreadyRecent) {
                this.recentSearchResults.unshift(mo);
                this.recentSearchResults = this.recentSearchResults.slice(0, this.MAX_RECENT_SEARCH_RESULTS);
            }
            var recentSearchIds = this.recentSearchResults.map(function (_a) {
                var id = _a.id;
                return id;
            });
            localStorage.setItem(this.RECENT_SEARCH_STORAGE_KEY, JSON.stringify(recentSearchIds));
            this.router.navigateByUrl(this.deviceGridService.getHref(mo, '/'));
            this.dropdown.hide();
        };
        SearchBoxComponent.prototype.reset = function () {
            this.typeahead.onSearch.emit('');
            this.selected = undefined;
            this.typeahead.searchControl.nativeElement.focus();
        };
        SearchBoxComponent.prototype.keyDown = function (event) {
            var keyCode = event.keyCode;
            if (keyCode === this.KEYCODE_ENTER) {
                // enter hit can be faster then typeahead debounce,
                // therefore we take the term from the DOM element
                // itself:
                var searchTerm = event.target.value;
                this.navigate(['/assetsearch'], { queryParams: { search: searchTerm } });
                this.dropdown.hide();
            }
            else if (keyCode === this.KEYCODE_ESC) {
                if (this.term === '') {
                    this.dropdown.hide();
                }
                this.reset();
            }
        };
        SearchBoxComponent.prototype.filter = function (on) {
            this.navigate(['/assetsearch'], { queryParams: { filter: on } });
            this.dropdown.hide();
        };
        SearchBoxComponent.prototype.openSearch = function () {
            this.router.navigateByUrl('/assetsearch');
            this.dropdown.hide();
        };
        SearchBoxComponent.prototype.subscribeOnSearch = function () {
            var _this = this;
            if (!this.results$) {
                this.results$ = this.typeahead.onSearch.pipe(operators.tap(function (term) { return _this.onTypingStarted(term); }), operators.switchMap(function (term) { return _this.mergeRequest(term); }));
            }
        };
        SearchBoxComponent.prototype.navigate = function (commands, extras) {
            var _this = this;
            this.router
                .navigateByUrl('/', { skipLocationChange: true })
                .then(function () { return _this.router.navigate(commands, extras); });
        };
        SearchBoxComponent.prototype.mergeRequest = function (term) {
            var _this = this;
            return rxjs.merge(rxjs.of({ data: [] }), this.queryInventoryService(term).pipe(operators.tap(function (_a) {
                var data = _a.data, paging = _a.paging;
                return _this.onLoadingDone(data, paging);
            })));
        };
        SearchBoxComponent.prototype.queryInventoryService = function (term) {
            var _this = this;
            if (term) {
                return rxjs.defer(function () { return _this.searchService.search(term); });
            }
            return rxjs.empty();
        };
        SearchBoxComponent.prototype.onLoadingDone = function (data, paging) {
            this.isLoading = false;
            this.noMatch =
                paging && paging.nextPage === null && this.searchService.filterOnlyAssets(data).length === 0;
        };
        SearchBoxComponent.prototype.onTypingStarted = function (term) {
            this.noMatch = false;
            this.term = term;
            this.isLoading = term.length > 0;
        };
SearchBoxComponent.ɵfac = function SearchBoxComponent_Factory(t) { return new (t || SearchBoxComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.InventoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(AssetSearchService)); };
SearchBoxComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchBoxComponent, selectors: [["c8y-search-box"]], viewQuery: function SearchBoxComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(i3.TypeaheadComponent, 5);
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.typeahead = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropdown = _t.first);
    } }, decls: 5, vars: 1, consts: [["dropdown", "", 1, "dropdown", 3, "insideClick", "isOpenChange"], ["dropdown", "bs-dropdown"], ["dropdownToggle", "", "type", "button", "title", "Search", "aria-controls", "searchDropdown", 1, "main-header-button", "dropdown-toggle", "c8y-dropdown"], ["c8yIcon", "search", 1, "icon-2x"], ["id", "searchDropdown", "class", "search-header-menu dropdown-menu dropdown-menu-center", 4, "dropdownMenu"], ["id", "searchDropdown", 1, "search-header-menu", "dropdown-menu", "dropdown-menu-center"], ["novalidate", "", 1, "c8y-search-form"], ["searchForm", "ngForm"], ["name", "selected", "container", "body", 3, "ngModel", "placeholder", "icon", "allowFreeEntries", "ngModelChange", "keydown", "onIconClick"], ["class", "p-l-16 p-r-16", 4, "ngIf"], ["class", "p-l-24 p-r-24", 3, "selectable", 4, "ngIf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "ngFor", "ngForOf"], ["class", "c8y-list__item--link m-l-16 m-r-16", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe"], ["class", "c8y-list__item--link  m-l-16 m-r-16", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForLoadMore", "c8yForPipe", "c8yForNotFound", "c8yForLoadingTemplate", "c8yForLoadNextLabel"], ["notFoundTemplate", ""], ["class", "p-t-32 p-b-0 p-relative", 4, "ngIf"], ["loading", ""], [1, "m-t-24", "bg-gray-lighter", "p-t-16", "p-b-16", "p-l-24", "p-r-24", 3, "selectable"], [1, "flex-row"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-4"], ["translate", "", 1, "m-r-8"], ["type", "button", "translate", "", 1, "m-l-auto", "btn", "btn-default", "btn-sm", 3, "mousedown"], [1, "p-l-16", "p-r-16"], [1, "m-r-4", "text-muted"], ["translate", ""], [1, "btn-group", "btn-group-sm"], [1, "btn", "btn-default", 3, "title", "click"], [1, "p-l-24", "p-r-24", 3, "selectable"], [1, "legend", "form-block"], [1, "c8y-list__item--link", "m-l-16", "m-r-16", 3, "click"], [3, "mo", 4, "ngIf"], ["class", "c8y-icon-duocolor", 3, "c8yIcon", 4, "ngIf"], [3, "mo"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], ["class", "p-16 c8y-empty-state", 3, "selectable", 4, "ngIf"], [1, "p-16", "c8y-empty-state", 3, "selectable"], [3, "icon"], [1, "p-t-32", "p-b-0", "p-relative"], [1, "spinner", 2, "right", "0"], [1, "rect1"], [1, "rect2"], [1, "rect3"], [1, "rect4"], [1, "rect5"], [1, "text-center", "p-t-32", "p-b-0", "p-relative"]], template: function SearchBoxComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵlistener("isOpenChange", function SearchBoxComponent_Template_div_isOpenChange_0_listener($event) { return ctx.onOpenChange($event); });
        ɵngcc0.ɵɵelementStart(2, "button", 2);
        ɵngcc0.ɵɵelement(3, "i", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, SearchBoxComponent_div_4_Template, 25, 28, "div", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("insideClick", true);
    } }, directives: [ɵngcc4.BsDropdownDirective, ɵngcc4.BsDropdownToggleDirective, ɵngcc5.IconDirective, ɵngcc4.BsDropdownMenuDirective, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc5.TypeaheadComponent, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.NgIf, ɵngcc7.NgForOf, ɵngcc5.ForOfDirective, ɵngcc5.ListItemComponent, ɵngcc5.C8yTranslateDirective, ɵngcc5.ListItemIconComponent, ɵngcc5.DeviceStatusComponent], pipes: [ɵngcc5.C8yTranslatePipe, ɵngcc7.AsyncPipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchBoxComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-box',
                template: "<div\n  class=\"dropdown\"\n  dropdown\n  #dropdown=\"bs-dropdown\"\n  [insideClick]=\"true\"\n  (isOpenChange)=\"onOpenChange($event)\"\n>\n  <button\n    class=\"main-header-button dropdown-toggle c8y-dropdown\"\n    dropdownToggle\n    type=\"button\"\n    title=\"Search\"\n    aria-controls=\"searchDropdown\"\n  >\n    <i c8yIcon=\"search\" class=\"icon-2x\"></i>\n  </button>\n\n  <div\n    id=\"searchDropdown\"\n    *dropdownMenu\n    class=\"search-header-menu dropdown-menu dropdown-menu-center\"\n  >\n    <form novalidate #searchForm=\"ngForm\" class=\"c8y-search-form\">\n      <c8y-typeahead\n        [(ngModel)]=\"selected\"\n        placeholder=\"{{ 'Search for groups or assets\u2026' | translate }}\"\n        (keydown)=\"keyDown($event)\"\n        (onIconClick)=\"reset()\"\n        [icon]=\"term ? 'times' : 'search'\"\n        [allowFreeEntries]=\"false\"\n        name=\"selected\"\n        container=\"body\"\n      >\n        <!-- filter buttons -->\n        <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-16 p-r-16\">\n          <div class=\"flex-row\">\n            <p class=\"m-r-4 text-muted\">\n              <em translate>Searching by exact match. Click for other search options:</em>\n            </p>\n            <div class=\"btn-group btn-group-sm\">\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Starts with' | translate }}\"\n                (click)=\"filter(term + '*')\"\n              >\n                {{ 'Starts with' | translate }}\n              </button>\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Contains' | translate }}\"\n                (click)=\"filter('*' + term + '*')\"\n              >\n                {{ 'Contains' | translate }}\n              </button>\n              <button\n                class=\"btn btn-default\"\n                title=\"{{ 'Ends with' | translate }}\"\n                (click)=\"filter('*' + term)\"\n              >\n                {{ 'Ends with' | translate }}\n              </button>\n            </div>\n          </div>\n        </c8y-li>\n\n        <!-- Recent search -->\n        <c8y-li\n          *ngIf=\"term.length === 0 && recentSearchResults.length > 0\"\n          [selectable]=\"false\"\n          class=\"p-l-24 p-r-24\"\n        >\n          <div class=\"legend form-block\">\n            <span translate>Recent search results</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *ngFor=\"let result of term.length === 0 ? recentSearchResults : []\"\n          class=\"c8y-list__item--link m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- Recently registered devices -->\n        <c8y-li\n          *ngIf=\"term.length === 0 && (recentlyRegisteredResults$ | async)?.data?.length > 0\"\n          class=\"p-l-24 p-r-24\"\n          [selectable]=\"false\"\n        >\n          <div class=\"legend form-block\">\n            <span translate>Recently registered devices</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *c8yFor=\"\n            let result of term.length === 0 ? recentlyRegisteredResults$ : { data: [] };\n            loadMore: 'none';\n            pipe: filterPipe\n          \"\n          class=\"c8y-list__item--link m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- Search results -->\n        <c8y-li *ngIf=\"term.length !== 0\" class=\"p-l-24 p-r-24\" [selectable]=\"false\">\n          <div class=\"legend form-block\">\n            <span translate>Search results</span>\n          </div>\n        </c8y-li>\n        <c8y-li\n          *c8yFor=\"\n            let result of results$;\n            loadMore: 'auto';\n            pipe: filterPipe;\n            notFound: notFoundTemplate;\n            loadingTemplate: loading;\n            loadNextLabel: 'Find more\u2026'\n          \"\n          class=\"c8y-list__item--link  m-l-16 m-r-16\"\n          (click)=\"open($event, result)\"\n        >\n          <c8y-li-icon>\n            <device-status [mo]=\"result\" *ngIf=\"!result.c8y_IsDeviceGroup\"></device-status>\n            <i\n              [c8yIcon]=\"'c8y-group-open'\"\n              class=\"c8y-icon-duocolor\"\n              *ngIf=\"result.c8y_IsDeviceGroup\"\n            ></i>\n          </c8y-li-icon>\n          {{ result.name || '--' }}\n        </c8y-li>\n\n        <!-- No search results found entry -->\n        <ng-template #notFoundTemplate>\n          <c8y-li *ngIf=\"noMatch\" class=\"p-16 c8y-empty-state\" [selectable]=\"false\">\n            <c8y-li-icon [icon]=\"'search'\"></c8y-li-icon>\n            <p><strong translate>No match found.</strong></p>\n            <small translate>\n              Use a filter or go to the asset data table to show all devices and groups.\n            </small>\n          </c8y-li>\n        </ng-template>\n\n        <!-- loading bar first entries -->\n        <c8y-li *ngIf=\"isLoading\" class=\"p-t-32 p-b-0 p-relative\">\n          <div class=\"spinner\" style=\"right:0;\">\n            <div class=\"rect1\"></div>\n            <div class=\"rect2\"></div>\n            <div class=\"rect3\"></div>\n            <div class=\"rect4\"></div>\n            <div class=\"rect5\"></div>\n          </div>\n        </c8y-li>\n\n        <!-- loading bar for loading more entries (inventory roles) -->\n        <ng-template #loading>\n          <c8y-li class=\"text-center p-t-32 p-b-0 p-relative\">\n            <div class=\"spinner\" style=\"right:0;\">\n              <div class=\"rect1\"></div>\n              <div class=\"rect2\"></div>\n              <div class=\"rect3\"></div>\n              <div class=\"rect4\"></div>\n              <div class=\"rect5\"></div>\n            </div>\n          </c8y-li>\n        </ng-template>\n\n        <!-- more filter possibilities -->\n        <c8y-li class=\"m-t-24 bg-gray-lighter p-t-16 p-b-16 p-l-24 p-r-24\" [selectable]=\"false\">\n          <div class=\"flex-row\">\n            <i c8yIcon=\"info-circle\" class=\"text-info m-r-4\"></i>\n            <p translate class=\"m-r-8\">Need more filter possibilities?</p>\n            <button\n              type=\"button\"\n              class=\"m-l-auto btn btn-default btn-sm\"\n              translate\n              (mousedown)=\"openSearch()\"\n            >\n              Go to the asset data table\n            </button>\n          </div>\n        </c8y-li>\n      </c8y-typeahead>\n    </form>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.InventoryService }, { type: ɵngcc3.DeviceGridService }, { type: AssetSearchService }]; }, { typeahead: [{
            type: i0.ViewChild,
            args: [i3.TypeaheadComponent, { static: false }]
        }], dropdown: [{
            type: i0.ViewChild,
            args: ['dropdown', { static: false }]
        }] }); })();
        return SearchBoxComponent;
    }());
    SearchBoxComponent.ctorParameters = function () { return [
        { type: router.Router },
        { type: client.InventoryService },
        { type: deviceGrid.DeviceGridService },
        { type: AssetSearchService }
    ]; };
    SearchBoxComponent.propDecorators = {
        typeahead: [{ type: i0.ViewChild, args: [i3.TypeaheadComponent, { static: false },] }],
        dropdown: [{ type: i0.ViewChild, args: ['dropdown', { static: false },] }]
    };

    var SearchGridComponent = /** @class */ (function () {
        function SearchGridComponent(searchService, bsModalService, smartGroupsService, subAssetsGridService) {
            this.searchService = searchService;
            this.bsModalService = bsModalService;
            this.smartGroupsService = smartGroupsService;
            this.subAssetsGridService = subAssetsGridService;
            this.title = '';
            this.loadingItemsLabel = i3.gettext('Loading results…');
            this.selectable = false;
            this.onColumnsChange = new i0.EventEmitter();
            this.searchText = '';
            this.pagination = this.searchService.getDefaultPagination();
            this.bulkActionControls = this.searchService.getDefaultBulkActionControls();
            this.refresh = new i0.EventEmitter();
            this.sizeCount = 0;
        }
        Object.defineProperty(SearchGridComponent.prototype, "_columns", {
            set: function (value) {
                if (value) {
                    this.columns = value;
                }
                else {
                    this.columns = this.searchService.getDefaultColumns();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_pagination", {
            set: function (value) {
                if (value) {
                    this.pagination = value;
                }
                else {
                    this.pagination = this.searchService.getDefaultPagination();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_actionControls", {
            set: function (value) {
                if (value) {
                    this.actionControls = value;
                }
                else {
                    this.actionControls = this.searchService.getDefaultActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SearchGridComponent.prototype, "_bulkActionControls", {
            set: function (value) {
                if (value) {
                    this.bulkActionControls = value;
                }
                else {
                    this.bulkActionControls = this.searchService.getDefaultBulkActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        SearchGridComponent.prototype.ngOnInit = function () {
            if (!this.filteringName) {
                this.columns = this.searchService.getDefaultColumns();
            }
            else {
                this.columns = [
                    new subAssets.AssetTypeGridColumn({ sortOrder: 'desc' }),
                    new deviceGrid.NameDeviceGridColumn({
                        sortOrder: 'asc',
                        filter: { externalFilterQuery: { names: [this.filteringName] } }
                    }),
                    new deviceGrid.ModelDeviceGridColumn(),
                    new deviceGrid.SerialNumberDeviceGridColumn({ visible: false }),
                    new deviceGrid.RegistrationDateDeviceGridColumn({ visible: false }),
                    new deviceGrid.SystemIdDeviceGridColumn({ visible: false }),
                    new deviceGrid.ImeiDeviceGridColumn({ visible: false }),
                    new deviceGrid.AlarmsDeviceGridColumn()
                ];
            }
            this.serverSideDataCallback = this.onDataSourceModifier.bind(this);
            this.setActionControls();
        };
        SearchGridComponent.prototype.trackByName = function (_index, column) {
            return column.name;
        };
        SearchGridComponent.prototype.onDataSourceModifier = function (dataSourceModifier) {
            return __awaiter(this, void 0, void 0, function () {
                var response, res, data, paging, filteredData;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!dataSourceModifier.searchText) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.searchService.search(dataSourceModifier.searchText, dataSourceModifier.pagination)];
                        case 1:
                            response = _a.sent();
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, this.searchService.getData(dataSourceModifier.columns, dataSourceModifier.pagination, undefined)];
                        case 3:
                            response = _a.sent();
                            _a.label = 4;
                        case 4:
                            res = response.res, data = response.data, paging = response.paging;
                            filteredData = this.searchService.filterOnlyAssets(data);
                            if (paging.currentPage === 1) {
                                this.sizeCount = 0;
                            }
                            this.sizeCount += filteredData.length;
                            this.onColumnsChange.emit(dataSourceModifier.columns);
                            return [2 /*return*/, {
                                    res: res,
                                    data: filteredData,
                                    paging: paging,
                                    filteredSize: this.sizeCount,
                                    size: undefined
                                }];
                    }
                });
            });
        };
        SearchGridComponent.prototype.setActionControls = function () {
            var _this = this;
            var actionControls = [];
            var deleteAction = {
                type: "DELETE" /* Delete */,
                callback: function (asset) { return _this.onDeleteAsset(asset, _this.parentGroup); }
            };
            actionControls.push(deleteAction);
            if (!this.actionControls) {
                this.actionControls = actionControls;
            }
        };
        SearchGridComponent.prototype.updateFiltering = function (columnNames, action) {
            var type = action.type;
            if (type === i3.FilteringActionType.ResetFilter) {
                this.dataGrid.clearFilters();
            }
            else {
                this.dataGrid.updateFiltering(columnNames, action, false);
            }
        };
        SearchGridComponent.prototype.configChange = function (config) {
            this.searchService.saveConfig(config);
        };
        SearchGridComponent.prototype.onDeleteAsset = function (asset, parentRef) {
            var _this = this;
            var initialState = {
                showWithDeviceUserCheckbox: this.subAssetsGridService.shouldShowWithDeviceUserCheckbox(asset),
                asset: asset,
                showWithCascadeCheckbox: !this.smartGroupsService.isSmartGroup(asset) &&
                    !this.smartGroupsService.isSmartGroupV2(asset)
            };
            var modalRef = this.bsModalService.show(subAssets.DeleteAssetsModalComponent, { initialState: initialState });
            modalRef.content.closeSubject.subscribe(function (result) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!result) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.subAssetsGridService.deleteAsset(asset, parentRef, result)];
                        case 1:
                            _a.sent();
                            this.refresh.emit();
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            }); });
        };
SearchGridComponent.ɵfac = function SearchGridComponent_Factory(t) { return new (t || SearchGridComponent)(ɵngcc0.ɵɵdirectiveInject(AssetSearchService), ɵngcc0.ɵɵdirectiveInject(ɵngcc8.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SmartGroupsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc9.SubAssetsService)); };
SearchGridComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchGridComponent, selectors: [["c8y-search-grid"]], viewQuery: function SearchGridComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(i3.DataGridComponent, 7);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dataGrid = _t.first);
    } }, inputs: { title: "title", loadingItemsLabel: "loadingItemsLabel", selectable: "selectable", searchText: "searchText", _columns: ["columns", "_columns"], _pagination: ["pagination", "_pagination"], _actionControls: ["actionControls", "_actionControls"], _bulkActionControls: ["bulkActionControls", "_bulkActionControls"], parentGroup: ["parent-group", "parentGroup"], filteringName: "filteringName" }, outputs: { onColumnsChange: "onColumnsChange" }, decls: 14, vars: 22, consts: [[1, "card--grid--fullpage"], [1, "d-contents", 3, "title", "loadingItemsLabel", "columns", "pagination", "actionControls", "selectable", "bulkActionControls", "serverSideDataCallback", "infiniteScroll", "showSearch", "searchText", "refresh", "onConfigChange"], [4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "c8y-empty-state"], ["c8yIcon", "search"], [3, "name"]], template: function SearchGridComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "c8y-data-grid", 1);
        ɵngcc0.ɵɵlistener("onConfigChange", function SearchGridComponent_Template_c8y_data_grid_onConfigChange_1_listener($event) { return ctx.configChange($event); });
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵtemplate(3, SearchGridComponent_ng_container_3_Template, 2, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelement(5, "h1", 4);
        ɵngcc0.ɵɵelementStart(6, "div");
        ɵngcc0.ɵɵelementStart(7, "p");
        ɵngcc0.ɵɵelementStart(8, "strong");
        ɵngcc0.ɵɵtext(9);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "small");
        ɵngcc0.ɵɵtext(12);
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 16, "Search results"))("loadingItemsLabel", ctx.loadingItemsLabel)("columns", ctx.columns)("pagination", ctx.pagination)("actionControls", ctx.actionControls)("selectable", ctx.selectable)("bulkActionControls", ctx.bulkActionControls)("serverSideDataCallback", ctx.serverSideDataCallback)("infiniteScroll", "auto")("showSearch", true)("searchText", ctx.searchText)("refresh", ctx.refresh);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns)("ngForTrackBy", ctx.trackByName);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(10, 18, "No items found."));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(13, 20, "Change your search or filter criteria."));
    } }, directives: [ɵngcc5.DataGridComponent, ɵngcc7.NgForOf, ɵngcc5.IconDirective, ɵngcc5.ColumnDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchGridComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-grid',
                template: "<div class=\"card--grid--fullpage\">\n  <c8y-data-grid\n    [title]=\"'Search results' | translate\"\n    [loadingItemsLabel]=\"loadingItemsLabel\"\n    [columns]=\"columns\"\n    [pagination]=\"pagination\"\n    [actionControls]=\"actionControls\"\n    [selectable]=\"selectable\"\n    [bulkActionControls]=\"bulkActionControls\"\n    [serverSideDataCallback]=\"serverSideDataCallback\"\n    [infiniteScroll]=\"'auto'\"\n    [showSearch]=\"true\"\n    [searchText]=\"searchText\"\n    [refresh]=\"refresh\"\n    (onConfigChange)=\"configChange($event)\"\n    class=\"d-contents\"\n  >\n    <ng-container *ngFor=\"let column of columns; trackBy: trackByName\">\n      <c8y-column [name]=\"column.name\"></c8y-column>\n    </ng-container>\n    <div class=\"c8y-empty-state\">\n      <h1 c8yIcon=\"search\"></h1>\n      <div>\n        <p>\n          <strong>{{ 'No items found.' | translate }}</strong>\n        </p>\n        <small>{{ 'Change your search or filter criteria.' | translate }}</small>\n      </div>\n    </div>\n  </c8y-data-grid>\n</div>\n"
            }]
    }], function () { return [{ type: AssetSearchService }, { type: ɵngcc8.BsModalService }, { type: ɵngcc2.SmartGroupsService }, { type: ɵngcc9.SubAssetsService }]; }, { title: [{
            type: i0.Input
        }], loadingItemsLabel: [{
            type: i0.Input
        }], selectable: [{
            type: i0.Input
        }], onColumnsChange: [{
            type: i0.Output
        }], searchText: [{
            type: i0.Input
        }], _columns: [{
            type: i0.Input,
            args: ['columns']
        }], _pagination: [{
            type: i0.Input,
            args: ['pagination']
        }], _actionControls: [{
            type: i0.Input,
            args: ['actionControls']
        }], _bulkActionControls: [{
            type: i0.Input,
            args: ['bulkActionControls']
        }], parentGroup: [{
            type: i0.Input,
            args: ['parent-group']
        }], filteringName: [{
            type: i0.Input
        }], dataGrid: [{
            type: i0.ViewChild,
            args: [i3.DataGridComponent, { static: true }]
        }] }); })();
        return SearchGridComponent;
    }());
    SearchGridComponent.ctorParameters = function () { return [
        { type: AssetSearchService },
        { type: modal.BsModalService },
        { type: client.SmartGroupsService },
        { type: subAssets.SubAssetsService }
    ]; };
    SearchGridComponent.propDecorators = {
        parentGroup: [{ type: i0.Input, args: ['parent-group',] }],
        title: [{ type: i0.Input }],
        loadingItemsLabel: [{ type: i0.Input }],
        _columns: [{ type: i0.Input, args: ['columns',] }],
        _pagination: [{ type: i0.Input, args: ['pagination',] }],
        _actionControls: [{ type: i0.Input, args: ['actionControls',] }],
        selectable: [{ type: i0.Input }],
        _bulkActionControls: [{ type: i0.Input, args: ['bulkActionControls',] }],
        onColumnsChange: [{ type: i0.Output }],
        searchText: [{ type: i0.Input }],
        filteringName: [{ type: i0.Input }],
        dataGrid: [{ type: i0.ViewChild, args: [i3.DataGridComponent, { static: true },] }]
    };

    var SearchResultsComponent = /** @class */ (function () {
        function SearchResultsComponent(route, alert) {
            this.route = route;
            this.alert = alert;
            this.filter = '';
            this.searchTerm = '';
            this.unsubscribe$ = new rxjs.Subject();
        }
        SearchResultsComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.route.queryParams.subscribe(function (params) {
                if (params.filter) {
                    _this.filteringName = params.filter;
                }
            });
        };
        SearchResultsComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            this.route.queryParams
                .pipe(operators.takeUntil(this.unsubscribe$))
                .subscribe(function (_a) {
                var filter = _a.filter, search = _a.search;
                return _this.onQueryParamsChange(filter, search);
            });
            this.searchGrid.dataGrid.searchText$.pipe(operators.takeUntil(this.unsubscribe$)).subscribe(function (text) {
                if (text) {
                    _this.resetFilter();
                }
                _this.searchTerm = text;
            });
            this.searchGrid.dataGrid.onFilter
                .pipe(operators.takeUntil(this.unsubscribe$))
                .subscribe(function () { return _this.resetSearch(); });
            // to prevent race condition (search empty):
            this.searchTerm = this.route.snapshot.queryParams.search || '';
        };
        SearchResultsComponent.prototype.resetSearch = function () {
            this.searchTerm = '';
            if (this.searchTerm) {
                this.alert.add({
                    text: i3.gettext('Search reset. Full text search does not support filtering.'),
                    type: i3.Status.WARNING,
                    timeout: 5000
                });
            }
        };
        SearchResultsComponent.prototype.resetFilter = function () {
            this.filter = '';
            if (this.searchGrid.dataGrid.filteringApplied) {
                this.alert.add({
                    text: i3.gettext('Filter reset. Full text search does not support filtering.'),
                    type: i3.Status.WARNING,
                    timeout: 5000
                });
                this.searchGrid.dataGrid.clearFilters();
            }
        };
        SearchResultsComponent.prototype.ngOnDestroy = function () {
            this.unsubscribe$.next();
            this.unsubscribe$.complete();
        };
        SearchResultsComponent.prototype.onQueryParamsChange = function (filter, searchTerm) {
            if (!this.shouldFilter(filter) && searchTerm) {
                this.search(searchTerm);
            }
            else if (this.shouldFilter(filter) && searchTerm) {
                this.search(searchTerm);
            }
        };
        SearchResultsComponent.prototype.shouldFilter = function (filter) {
            if (!filter) {
                return false;
            }
            this.resetSearch();
            this.filter = filter || '';
            this.searchGrid.updateFiltering(['name'], {
                type: i3.FilteringActionType.ApplyFilter,
                payload: {
                    filteringModifier: {
                        externalFilterQuery: {
                            names: [this.filter]
                        }
                    }
                }
            });
            return true;
        };
        SearchResultsComponent.prototype.search = function (searchTerm) {
            this.searchTerm = searchTerm || '';
            this.searchGrid.dataGrid.searchText$.next(this.searchTerm);
        };
SearchResultsComponent.ɵfac = function SearchResultsComponent_Factory(t) { return new (t || SearchResultsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.AlertService)); };
SearchResultsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchResultsComponent, selectors: [["c8y-search-results"]], viewQuery: function SearchResultsComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(SearchGridComponent, 7);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.searchGrid = _t.first);
    } }, decls: 6, vars: 4, consts: [["translate", "", 1, "p-r-4"], ["ngNonBindable", "", "translate", "", 3, "translateParams", 4, "ngIf"], [3, "searchText", "filteringName"], ["translate", "", 3, "translateParams"]], template: function SearchResultsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵtext(2, "Search");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, SearchResultsComponent_small_3_Template, 2, 3, "small", 1);
        ɵngcc0.ɵɵtemplate(4, SearchResultsComponent_small_4_Template, 2, 3, "small", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(5, "c8y-search-grid", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.searchTerm);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.filter);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("searchText", ctx.searchTerm)("filteringName", ctx.filteringName);
    } }, directives: [ɵngcc5.TitleComponent, ɵngcc5.C8yTranslateDirective, ɵngcc7.NgIf, SearchGridComponent], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchResultsComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-search-results',
                template: "<c8y-title>\n  <span translate class=\"p-r-4\">Search</span>\n  <small\n    ngNonBindable\n    translate\n    *ngIf=\"searchTerm\"\n    [translateParams]=\"{\n      searchHint: searchTerm\n    }\"\n    >searching \"{{ searchHint }}\"</small\n  >\n  <small\n    ngNonBindable\n    translate\n    *ngIf=\"filter\"\n    [translateParams]=\"{\n      filterHint: filter\n    }\"\n    >filtered by \"{{ filterHint }}\"</small\n  >\n</c8y-title>\n\n<c8y-search-grid [searchText]=\"searchTerm\" [filteringName]=\"filteringName\"></c8y-search-grid>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc5.AlertService }]; }, { searchGrid: [{
            type: i0.ViewChild,
            args: [SearchGridComponent, { static: true }]
        }] }); })();
        return SearchResultsComponent;
    }());
    SearchResultsComponent.ctorParameters = function () { return [
        { type: router.ActivatedRoute },
        { type: i3.AlertService }
    ]; };
    SearchResultsComponent.propDecorators = {
        searchGrid: [{ type: i0.ViewChild, args: [SearchGridComponent, { static: true },] }]
    };

    var ɵ0 = [
        {
            path: 'assetsearch',
            component: SearchResultsComponent
        }
    ], ɵ1 = [
        {
            template: SearchBoxComponent
        }
    ];
    var SearchModule = /** @class */ (function () {
        function SearchModule() {
        }
SearchModule.ɵfac = function SearchModule_Factory(t) { return new (t || SearchModule)(); };
SearchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SearchModule });
SearchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        AssetSearchService,
        {
            provide: i3.HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        {
            provide: i3.HOOK_SEARCH,
            useValue: ɵ1,
            multi: true
        }
    ], imports: [[i3.CoreModule, deviceGrid.DeviceGridModule, dropdown.BsDropdownModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchModule, [{
        type: i0.NgModule,
        args: [{
                declarations: [SearchResultsComponent, SearchBoxComponent, SearchGridComponent],
                imports: [i3.CoreModule, deviceGrid.DeviceGridModule, dropdown.BsDropdownModule],
                entryComponents: [SearchResultsComponent, SearchBoxComponent],
                exports: [SearchResultsComponent, SearchBoxComponent, SearchGridComponent],
                providers: [
                    AssetSearchService,
                    {
                        provide: i3.HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    {
                        provide: i3.HOOK_SEARCH,
                        useValue: ɵ1,
                        multi: true
                    }
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SearchModule, { declarations: [SearchResultsComponent, SearchBoxComponent, SearchGridComponent], imports: [ɵngcc5.CoreModule, ɵngcc3.DeviceGridModule, ɵngcc4.BsDropdownModule], exports: [SearchResultsComponent, SearchBoxComponent, SearchGridComponent] }); })();
        return SearchModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AssetSearchService = AssetSearchService;
    exports.SearchBoxComponent = SearchBoxComponent;
    exports.SearchGridComponent = SearchGridComponent;
    exports.SearchModule = SearchModule;
    exports.SearchResultsComponent = SearchResultsComponent;
    exports["ɵ0"] = ɵ0;
    exports["ɵ1"] = ɵ1;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
