// @c8y/ngx-components/bundles/c8y-ngx-components-sub-assets.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@c8y/ngx-components'),require('@c8y/ngx-components/assets-navigator'),require('@angular/common'),require('@ngx-translate/core'),require('@c8y/client'),require('@angular/forms'),require('ngx-bootstrap/popover'),require('@angular/router'),require('ngx-bootstrap/modal'),require('@c8y/ngx-components/device-grid'),require('@angular/cdk/stepper'),exports, require('@angular/core'), require('@c8y/client'), require('@c8y/ngx-components'), require('@c8y/ngx-components/assets-navigator'), require('@ngx-translate/core'), require('@c8y/ngx-components/device-grid'), require('@c8y/client'), require('@c8y/client'), require('@c8y/client'), require('@c8y/client'), require('@angular/router'), require('ngx-bootstrap/modal'), require('rxjs'), require('ngx-bootstrap/popover'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/sub-assets', ['@angular/core','@c8y/ngx-components','@c8y/ngx-components/assets-navigator','@angular/common','@ngx-translate/core','@c8y/client','@angular/forms','ngx-bootstrap/popover','@angular/router','ngx-bootstrap/modal','@c8y/ngx-components/device-grid','@angular/cdk/stepper','exports', '@angular/core', '@c8y/client', '@c8y/ngx-components', '@c8y/ngx-components/assets-navigator', '@ngx-translate/core', '@c8y/ngx-components/device-grid', '@c8y/client', '@c8y/client', '@c8y/client', '@c8y/client', '@angular/router', 'ngx-bootstrap/modal', 'rxjs', 'ngx-bootstrap/popover', '@angular/forms'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.c8y.ngxComponents,global.c8y.ngxComponents.assetsNavigator,global.ng.common,global.ngxTranslate.core,global.c8y.client,global.ng.forms,global.ngxBootstrap.popover,global.ng.router,global.ngxBootstrap.modal,global.c8y.ngxComponents.deviceGrid,global.ng.cdk.stepper,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"]["sub-assets"] = {}), global.ng.core, global.client, global.c8y["ngx-components"], global.c8y["ngx-components"]["assets-navigator"], global.i1, global.c8y["ngx-components"]["device-grid"], global.i2, global.i4, global.i6, global.i7, global.ng.router, global.modal, global.rxjs, global.popover, global.ng.forms));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,ɵngcc5,ɵngcc6,ɵngcc7,ɵngcc8,ɵngcc9,ɵngcc10,ɵngcc11,exports, i0, client, i3, i5, i1, deviceGrid, i2, i4, i6, i7, router, modal, rxjs, popover, forms) { 
function AssetTypeCellRendererComponent_device_status_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 2);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("mo", ctx_r0.context.item);
} }
function AssetTypeCellRendererComponent_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 3);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r1.iconTitle));
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r1.groupIcon);
} }
function GroupInfoComponent_small_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 26);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Smart group"), " ");
} }
function GroupInfoComponent_small_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 26);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Group"), " ");
} }
function GroupInfoComponent_small_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 26);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Remote group"), " ");
} }
function GroupInfoComponent_p_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.group.name);
} }
function GroupInfoComponent_div_18_Template(rf, ctx) { if (rf & 1) {
    var _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵelementStart(1, "input", 29);
    ɵngcc0.ɵɵlistener("ngModelChange", function GroupInfoComponent_div_18_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.group.name = $event; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementStart(5, "div", 30);
    ɵngcc0.ɵɵelementStart(6, "button", 31);
    ɵngcc0.ɵɵlistener("click", function GroupInfoComponent_div_18_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r13 = ɵngcc0.ɵɵnextContext(); var _r3 = ɵngcc0.ɵɵreference(13); ctx_r13.update({ name: ctx_r13.group.name }); return _r3.form.markAsPristine(); });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r5 = ɵngcc0.ɵɵnextContext();
    var _r3 = ɵngcc0.ɵɵreference(13);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 7, "Name"));
    ɵngcc0.ɵɵpropertyInterpolate("size", ctx_r5.group.name.length + 2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(3, 9, "e.g. My group"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r5.group.name);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 11, "Save"));
    ɵngcc0.ɵɵproperty("disabled", _r3.form.invalid);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 13, "Save"), " ");
} }
function GroupInfoComponent_p_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r7.group.c8y_Notes);
} }
function GroupInfoComponent_div_24_Template(rf, ctx) { if (rf & 1) {
    var _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵelementStart(1, "textarea", 32);
    ɵngcc0.ɵɵlistener("ngModelChange", function GroupInfoComponent_div_24_Template_textarea_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r15); var ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.group.c8y_Notes = $event; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementStart(5, "div", 30);
    ɵngcc0.ɵɵelementStart(6, "button", 31);
    ɵngcc0.ɵɵlistener("click", function GroupInfoComponent_div_24_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r15); var ctx_r16 = ɵngcc0.ɵɵnextContext(); var _r6 = ɵngcc0.ɵɵreference(20); ctx_r16.update({ c8y_Notes: ctx_r16.group.c8y_Notes }); return _r6.form.markAsPristine(); });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r8 = ɵngcc0.ɵɵnextContext();
    var _r6 = ɵngcc0.ɵɵreference(20);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 7, "Description"));
    ɵngcc0.ɵɵpropertyInterpolate("cols", ctx_r8.group.c8y_Notes ? ctx_r8.group.c8y_Notes.length : 25);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(3, 9, "e.g. My description"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r8.group.c8y_Notes);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 11, "Save"));
    ɵngcc0.ɵɵproperty("disabled", _r6.form.invalid);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 13, "Save"), " ");
} }
function GroupInfoComponent_form_25_p_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r18 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r18.smartGroupFilter, " ");
} }
function GroupInfoComponent_form_25_div_10_Template(rf, ctx) { if (rf & 1) {
    var _r21 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵelementStart(1, "input", 38);
    ɵngcc0.ɵɵlistener("ngModelChange", function GroupInfoComponent_form_25_div_10_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r21); var ctx_r20 = ɵngcc0.ɵɵnextContext(2); return ctx_r20.smartGroupFilter = $event; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementStart(5, "div", 30);
    ɵngcc0.ɵɵelementStart(6, "button", 31);
    ɵngcc0.ɵɵlistener("click", function GroupInfoComponent_form_25_div_10_Template_button_click_6_listener() { ɵngcc0.ɵɵrestoreView(_r21); ɵngcc0.ɵɵnextContext(); var _r17 = ɵngcc0.ɵɵreference(1); var ctx_r22 = ɵngcc0.ɵɵnextContext(); ctx_r22.update({ c8y_DeviceQueryString: ctx_r22.smartGroupFilter }); return _r17.form.markAsPristine(); });
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    var _r17 = ɵngcc0.ɵɵreference(1);
    var ctx_r19 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("size", ctx_r19.smartGroupFilter.length + 2);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(2, 7, "e.g."), " $filter=(id eq '12*')");
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 9, "Smart group filter"));
    ɵngcc0.ɵɵproperty("ngModel", ctx_r19.smartGroupFilter);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 11, "Save"));
    ɵngcc0.ɵɵproperty("disabled", _r17.form.invalid);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 13, "Save"), " ");
} }
function GroupInfoComponent_form_25_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "form", null, 33);
    ɵngcc0.ɵɵelementStart(2, "c8y-form-group", 34);
    ɵngcc0.ɵɵelementStart(3, "label", 35);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementStart(6, "button", 36);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelement(8, "i", 37);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(9, GroupInfoComponent_form_25_p_9_Template, 2, 1, "p", 12);
    ɵngcc0.ɵɵtemplate(10, GroupInfoComponent_form_25_div_10_Template, 10, 15, "div", 13);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r9 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Smart group filter"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("popover", ɵngcc0.ɵɵpipeBind1(7, 6, ctx_r9.filterHintMsg));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r9.canEdit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.canEdit);
} }
function GroupInfoComponent_li_47_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 23);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r23 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r23.group.c8y_BrokerSource.status, " ");
} }
function GroupInfoComponent_li_47_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 23);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Offline"), " ");
} }
function GroupInfoComponent_li_47_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 24);
    ɵngcc0.ɵɵelementStart(1, "label", 22);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, GroupInfoComponent_li_47_span_4_Template, 2, 1, "span", 39);
    ɵngcc0.ɵɵtemplate(5, GroupInfoComponent_li_47_span_5_Template, 3, 3, "span", 39);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r10 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 3, "Status"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r10.group.c8y_BrokerSource);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r10.group.c8y_BrokerSource);
} }
function SubAssetsComponent_c8y_action_bar_item_2_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 11);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.showAddGroup = !ctx_r5.showAddGroup; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵelementStart(4, "span", 14);
    ɵngcc0.ɵɵtext(5, "Add group");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Add group"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r0.shouldDisableAddGroup);
} }
function SubAssetsComponent_c8y_action_bar_item_3_Template(rf, ctx) { if (rf & 1) {
    var _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 11);
    ɵngcc0.ɵɵelementStart(1, "button", 12);
    ɵngcc0.ɵɵlistener("click", function SubAssetsComponent_c8y_action_bar_item_3_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.showAssignDevices = !ctx_r7.showAssignDevices; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 13);
    ɵngcc0.ɵɵelementStart(4, "span", 14);
    ɵngcc0.ɵɵtext(5, "Assign devices");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Assign devices"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r1.shouldDisableAssignDevices);
} }
function SubAssetsComponent_c8y_add_group_9_Template(rf, ctx) { if (rf & 1) {
    var _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-add-group", 15);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_add_group_9_Template_c8y_add_group_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.showAddGroup = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("currentGroupId", ctx_r2.currentGroupId)("refresh", ctx_r2.refresh);
} }
function SubAssetsComponent_c8y_assign_devices_13_Template(rf, ctx) { if (rf & 1) {
    var _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-devices", 16);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.showAssignDevices = false; })("onShowChildDevices", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_onShowChildDevices_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.showAssignChildDevices = $event; })("selectedDevice", function SubAssetsComponent_c8y_assign_devices_13_Template_c8y_assign_devices_selectedDevice_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); var ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.showChildrenForDevice = $event; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r3.refresh)("currentGroupId", ctx_r3.currentGroupId);
} }
function SubAssetsComponent_c8y_assign_child_devices_17_Template(rf, ctx) { if (rf & 1) {
    var _r16 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-assign-child-devices", 17);
    ɵngcc0.ɵɵlistener("onCancel", function SubAssetsComponent_c8y_assign_child_devices_17_Template_c8y_assign_child_devices_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r16); var ctx_r15 = ɵngcc0.ɵɵnextContext(); return ctx_r15.showAssignChildDevices = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r4.refresh)("currentGroupId", ctx_r4.currentGroupId)("parentDevice", ctx_r4.showChildrenForDevice);
} }
var _c0 = function (a0) { return { drawerOpen: a0 }; };
var _c1 = ["modalRef"];
function DeleteAssetsModalComponent_c8y_form_group_7_Template(rf, ctx) { if (rf & 1) {
    var _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "input", 7);
    ɵngcc0.ɵɵlistener("ngModelChange", function DeleteAssetsModalComponent_c8y_form_group_7_Template_input_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r5); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.config.cascade = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementStart(5, "span");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Delete devices"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r2.config.cascade);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 5, "Also delete all devices inside selected asset and its subassets."), " ");
} }
function DeleteAssetsModalComponent_c8y_form_group_8_Template(rf, ctx) { if (rf & 1) {
    var _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group", 5);
    ɵngcc0.ɵɵelementStart(1, "label", 6);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "input", 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function DeleteAssetsModalComponent_c8y_form_group_8_Template_input_ngModelChange_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.config.withDeviceUser = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(4, "span");
    ɵngcc0.ɵɵelementStart(5, "span");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Delete associated device owner"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r3.config.withDeviceUser);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 5, "Also delete associated device owner."), " ");
} }
function SubAssetsGridComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "c8y-column", 4);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var column_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", column_r1.name);
} }
var _c2 = ["nameRef"];
function AddGroupComponent_c8y_title_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-title");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Add group"), "\n");
} }
function AddGroupComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementContainer(1, 4);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵnextContext();
    var _r2 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
} }
function AddGroupComponent_ng_template_2_c8y_message_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 15);
    ɵngcc0.ɵɵtext(1, "This field is required.");
    ɵngcc0.ɵɵelementEnd();
} }
function AddGroupComponent_ng_template_2_div_31_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 28);
    ɵngcc0.ɵɵtext(1, " You don't have permission to assign devices. ");
    ɵngcc0.ɵɵelementEnd();
} }
var _c3 = function () { return { edit: true, done: false }; };
var _c4 = function () { return ["col-md-6", "col-md-offset-3", "m-t-24", "m-b-40", "p-0", "flex-no-shrink"]; };
var _c5 = function (a0, a1) { return { next: a0, cancel: a1 }; };
var _c6 = function (a0, a1) { return { custom: a0, cancel: a1 }; };
var _c7 = function () { return { next: true, cancel: true }; };
var _c8 = function () { return { custom: true, cancel: true }; };
var _c9 = function () { return []; };
var _c10 = function (a0) { return { custom: a0 }; };
function AddGroupComponent_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    var _r8 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-stepper", 5);
    ɵngcc0.ɵɵelementStart(1, "cdk-step", 6);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementStart(3, "div", 7);
    ɵngcc0.ɵɵelementStart(4, "div", 8);
    ɵngcc0.ɵɵelementStart(5, "div", 9);
    ɵngcc0.ɵɵelementStart(6, "h4", 10);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 11);
    ɵngcc0.ɵɵelementStart(10, "div", 12);
    ɵngcc0.ɵɵelementStart(11, "div", 13);
    ɵngcc0.ɵɵelementStart(12, "div", 8);
    ɵngcc0.ɵɵelementStart(13, "div", 9);
    ɵngcc0.ɵɵelementStart(14, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(15, "div", 14);
    ɵngcc0.ɵɵelementStart(16, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(17, "label", 15);
    ɵngcc0.ɵɵtext(18, "Name");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(19, "input", 16, 17);
    ɵngcc0.ɵɵpipe(21, "translate");
    ɵngcc0.ɵɵelementStart(22, "c8y-messages");
    ɵngcc0.ɵɵtemplate(23, AddGroupComponent_ng_template_2_c8y_message_23_Template, 2, 0, "c8y-message", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(25, "label", 15);
    ɵngcc0.ɵɵtext(26, "Description");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(27, "input", 19);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(29, "c8y-form-group");
    ɵngcc0.ɵɵelement(30, "div", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(31, AddGroupComponent_ng_template_2_div_31_Template, 2, 0, "div", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(32, "c8y-stepper-buttons", 21);
    ɵngcc0.ɵɵlistener("onCancel", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCancel_32_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onCancel.emit(); })("onCustom", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCustom_32_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.createGroup(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(33, "cdk-step", 22);
    ɵngcc0.ɵɵpipe(34, "translate");
    ɵngcc0.ɵɵelementStart(35, "div", 7);
    ɵngcc0.ɵɵelementStart(36, "div", 8);
    ɵngcc0.ɵɵelementStart(37, "div", 9);
    ɵngcc0.ɵɵelementStart(38, "h4", 10);
    ɵngcc0.ɵɵtext(39);
    ɵngcc0.ɵɵpipe(40, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(41, "div", 23);
    ɵngcc0.ɵɵelementStart(42, "c8y-device-grid", 24);
    ɵngcc0.ɵɵlistener("itemsSelect", function AddGroupComponent_ng_template_2_Template_c8y_device_grid_itemsSelect_42_listener($event) { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.onSelected($event); });
    ɵngcc0.ɵɵpipe(43, "translate");
    ɵngcc0.ɵɵelementStart(44, "div", 25);
    ɵngcc0.ɵɵelement(45, "h1", 26);
    ɵngcc0.ɵɵelementStart(46, "div");
    ɵngcc0.ɵɵelementStart(47, "p");
    ɵngcc0.ɵɵelementStart(48, "strong");
    ɵngcc0.ɵɵtext(49);
    ɵngcc0.ɵɵpipe(50, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(51, "small");
    ɵngcc0.ɵɵtext(52);
    ɵngcc0.ɵɵpipe(53, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(54, "c8y-stepper-buttons", 27);
    ɵngcc0.ɵɵlistener("onCancel", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCancel_54_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onCancel.emit(); })("onCustom", function AddGroupComponent_ng_template_2_Template_c8y_stepper_buttons_onCustom_54_listener() { ɵngcc0.ɵɵrestoreView(_r8); var ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.createGroup(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r4 = ɵngcc0.ɵɵreference(20);
    var ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disableDefaultIcons", ɵngcc0.ɵɵpureFunction0(45, _c3))("customClasses", ɵngcc0.ɵɵpureFunction0(46, _c4));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("stepControl", ctx_r3.formGroupStepOne)("label", ɵngcc0.ɵɵpipeBind1(2, 27, "New group"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(8, 29, "New group"), " ");
    ɵngcc0.ɵɵadvance(8);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r3.formGroupStepOne);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(21, 31, "e.g. First floor"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.formGroupStepOne.untouched && !_r4.value);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(28, 33, "e.g. first floor devices"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r3.formGroupStepOne);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.canAssignDevice);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("disabled", !ctx_r3.canCreateGroup)("labels", ctx_r3.canAssignDevice ? ɵngcc0.ɵɵpureFunction2(47, _c5, ctx_r3.btnLabels.next, ctx_r3.btnLabels.cancel) : ɵngcc0.ɵɵpureFunction2(50, _c6, ctx_r3.btnLabels.create, ctx_r3.btnLabels.cancel))("showButtons", ctx_r3.canAssignDevice ? ɵngcc0.ɵɵpureFunction0(53, _c7) : ɵngcc0.ɵɵpureFunction0(54, _c8));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(34, 35, "Assign devices"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(40, 37, "Assign devices"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(43, 39, "Select target devices"))("actionControls", ɵngcc0.ɵɵpureFunction0(55, _c9))("infiniteScroll", "auto")("selectable", "true")("pagination", ctx_r3.pagination)("refresh", ctx_r3.refresh);
    ɵngcc0.ɵɵadvance(7);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(50, 41, "No matching devices."));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(53, 43, "Refine your search terms"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("labels", ɵngcc0.ɵɵpureFunction1(56, _c10, ctx_r3.btnLabels.create))("disabled", !ctx_r3.canAssignDevice)("pending", ctx_r3.pendingStatus);
} }
function GroupsComponent_c8y_add_group_11_Template(rf, ctx) { if (rf & 1) {
    var _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-add-group", 8);
    ɵngcc0.ɵɵlistener("onCancel", function GroupsComponent_c8y_add_group_11_Template_c8y_add_group_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.showAddGroup = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r0.refresh);
} }
var _c11 = ["showDevicesToggle"];
function AssignDevicesComponent_ng_template_29_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 16);
    ɵngcc0.ɵɵtext(1, " Displays the button ");
    ɵngcc0.ɵɵelementStart(2, "button", 17);
    ɵngcc0.ɵɵelement(3, "i", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(4, " next to target devices with children. Clicking it displays a list with all child devices of the selected target device. ");
    ɵngcc0.ɵɵelementEnd();
} }
function AssignDevicesComponent_ng_template_29_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "label", 11);
    ɵngcc0.ɵɵelementStart(1, "input", 12);
    ɵngcc0.ɵɵlistener("ngModelChange", function AssignDevicesComponent_ng_template_29_Template_input_ngModelChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.showChildren = $event; })("click", function AssignDevicesComponent_ng_template_29_Template_input_click_1_listener() { var restoredCtx = ɵngcc0.ɵɵrestoreView(_r6); var control_r2 = restoredCtx.headerActionControl; return control_r2.callback(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(2, "span");
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "button", 13);
    ɵngcc0.ɵɵelement(7, "i", 14);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(8, AssignDevicesComponent_ng_template_29_ng_template_8_Template, 5, 0, "ng-template", null, 15, ɵngcc0.ɵɵtemplateRefExtractor);
} if (rf & 2) {
    var control_r2 = ctx.headerActionControl;
    var _r3 = ɵngcc0.ɵɵreference(9);
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r1.showChildren);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 4, control_r2.text));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("popover", _r3)("outsideClick", true);
} }
var _c12 = function (a0) { return { "btn-pending": a0 }; };
'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);
    var i6__namespace = /*#__PURE__*/_interopNamespace(i6);
    var i7__namespace = /*#__PURE__*/_interopNamespace(i7);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || Array.prototype.slice.call(from));
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var AssetTypeCellRendererComponent = /** @class */ (function () {
        function AssetTypeCellRendererComponent(context, deviceGroupService) {
            this.context = context;
            this.deviceGroupService = deviceGroupService;
        }
        AssetTypeCellRendererComponent.prototype.ngOnInit = function () {
            this.isAnyGroup = this.deviceGroupService.isAnyGroup(this.context.item);
            this.isDevice = this.deviceGroupService.isDevice(this.context.item);
            this.groupIcon = this.deviceGroupService.icon(this.context.item);
            this.iconTitle = this.getIconTitle(this.groupIcon);
        };
        AssetTypeCellRendererComponent.prototype.getIconTitle = function (icon) {
            switch (icon) {
                case 'c8y-group':
                    return i3.gettext('Group');
                case 'c8y-group-smart':
                    return i3.gettext('Smart group');
                case 'c8y-group-remote-inactive':
                    return 'Inactive remote group';
                case 'c8y-group-remote':
                    return i3.gettext('Remote group');
                default:
                    return i3.gettext('Child device');
            }
        };
AssetTypeCellRendererComponent.ɵfac = function AssetTypeCellRendererComponent_Factory(t) { return new (t || AssetTypeCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGroupService)); };
AssetTypeCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssetTypeCellRendererComponent, selectors: [["ng-component"]], decls: 2, vars: 2, consts: [["class", "m-l-8", 3, "mo", 4, "ngIf"], ["class", "m-l-8 c8y-icon-duocolor", 3, "c8yIcon", "title", 4, "ngIf"], [1, "m-l-8", 3, "mo"], [1, "m-l-8", "c8y-icon-duocolor", 3, "c8yIcon", "title"]], template: function AssetTypeCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AssetTypeCellRendererComponent_device_status_0_Template, 1, 1, "device-status", 0);
        ɵngcc0.ɵɵtemplate(1, AssetTypeCellRendererComponent_i_1_Template, 2, 4, "i", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDevice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isAnyGroup || !ctx.isDevice);
    } }, directives: [ɵngcc3.NgIf, ɵngcc1.DeviceStatusComponent, ɵngcc1.IconDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetTypeCellRendererComponent, [{
        type: i0.Component,
        args: [{
                template: "<device-status class=\"m-l-8\" [mo]=\"context.item\" *ngIf=\"isDevice\"></device-status>\n<i\n  [c8yIcon]=\"groupIcon\"\n  title=\"{{ iconTitle | translate }}\"\n  class=\"m-l-8 c8y-icon-duocolor\"\n  *ngIf=\"isAnyGroup || !isDevice\"\n></i>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }, { type: ɵngcc2.DeviceGroupService }]; }, null); })();
        return AssetTypeCellRendererComponent;
    }());
    AssetTypeCellRendererComponent.ctorParameters = function () { return [
        { type: i3.CellRendererContext },
        { type: i5.DeviceGroupService }
    ]; };

    var AssetTypeGridColumn = /** @class */ (function (_super) {
        __extends(AssetTypeGridColumn, _super);
        function AssetTypeGridColumn(initialColumnConfig) {
            var _this = _super.call(this, initialColumnConfig) || this;
            _this.name = 'type';
            _this.header = i3.gettext('Type');
            _this.dataType = "icon" /* Icon */;
            _this.cellRendererComponent = AssetTypeCellRendererComponent;
            _this.sortable = true;
            _this.sortingConfig = {
                pathSortingConfigs: [{ path: "c8y_IsDeviceGroup" }]
            };
            return _this;
        }
        return AssetTypeGridColumn;
    }(deviceGrid.BaseDeviceGridColumn));

    var SubAssetsService = /** @class */ (function (_super) {
        __extends(SubAssetsService, _super);
        function SubAssetsService(translateService, inventoryService, appState, user, assetNodeService, deviceGroupService, smartGroupsService, smartRulesService, alertService, permissionsService, modal) {
            var _this = _super.call(this, inventoryService, translateService, alertService, modal) || this;
            _this.translateService = translateService;
            _this.inventoryService = inventoryService;
            _this.appState = appState;
            _this.user = user;
            _this.assetNodeService = assetNodeService;
            _this.deviceGroupService = deviceGroupService;
            _this.smartGroupsService = smartGroupsService;
            _this.smartRulesService = smartRulesService;
            _this.alertService = alertService;
            _this.permissionsService = permissionsService;
            _this.modal = modal;
            _this.GRID_CONFIG_DEFAULT_STORAGE_KEY = 'sub-assets-grid-config';
            _this.IS_DEVICE_GROUP_FRAGMENT = 'c8y_IsDeviceGroup';
            _this.IS_DYNAMIC_GROUP_FRAGMENT = 'c8y_IsDynamicGroup';
            return _this;
        }
        SubAssetsService.prototype.getDefaultColumns = function (filterable, sortable) {
            if (filterable === void 0) { filterable = true; }
            if (sortable === void 0) { sortable = true; }
            var defaultColumns = [
                new AssetTypeGridColumn({ sortOrder: 'desc' }),
                new deviceGrid.NameDeviceGridColumn({ sortOrder: 'asc' }),
                new deviceGrid.ModelDeviceGridColumn(),
                new deviceGrid.SerialNumberDeviceGridColumn({ visible: false }),
                new deviceGrid.RegistrationDateDeviceGridColumn({ visible: false }),
                new deviceGrid.SystemIdDeviceGridColumn({ visible: false }),
                new deviceGrid.ImeiDeviceGridColumn({ visible: false }),
                new deviceGrid.AlarmsDeviceGridColumn()
            ];
            return defaultColumns;
        };
        SubAssetsService.prototype.getDefaultPagination = function () {
            var pagination = this.getConfig().pagination;
            return {
                pageSize: pagination.pageSize,
                currentPage: 1
            };
        };
        SubAssetsService.prototype.getDefaultActionControls = function () {
            return [];
        };
        SubAssetsService.prototype.unassignAsset = function (asset, parentRef) {
            return __awaiter(this, void 0, void 0, function () {
                var assetId, parentId, alertMessage, error_1, alertMessage;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            assetId = asset.id;
                            parentId = parentRef.id;
                            if (!this.isDevice(asset)) return [3 /*break*/, 6];
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.inventoryService.childAssetsRemove(assetId, parentId)];
                        case 2:
                            _a.sent();
                            alertMessage = this.translateService.instant(i3.gettext('Asset unassigned.'));
                            this.alertService.success(alertMessage);
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _a.sent();
                            alertMessage = this.translateService.instant(i3.gettext('Could not unassign asset.'));
                            this.alertService.danger(alertMessage);
                            return [3 /*break*/, 4];
                        case 4: return [4 /*yield*/, this.deactivateSmartrulesForAsset(asset, parentRef)];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.isDevice = function (asset) {
            return (!asset.hasOwnProperty(this.IS_DEVICE_GROUP_FRAGMENT) &&
                !asset.hasOwnProperty(this.IS_DYNAMIC_GROUP_FRAGMENT));
        };
        SubAssetsService.prototype.deleteAsset = function (asset, parentRef, params) {
            if (params === void 0) { params = {}; }
            return __awaiter(this, void 0, void 0, function () {
                var isGroup;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            isGroup = asset.hasOwnProperty(this.IS_DEVICE_GROUP_FRAGMENT) ||
                                this.smartGroupsService.isSmartGroup(asset);
                            if (!isGroup) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.deleteGroup(asset, params)];
                        case 1:
                            _a.sent();
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, this.deleteDevice(asset, params)];
                        case 3:
                            _a.sent();
                            _a.label = 4;
                        case 4:
                            if (!(parentRef &&
                                !this.smartGroupsService.isSmartGroup(asset) &&
                                !this.smartGroupsService.isSmartGroupV2(asset))) return [3 /*break*/, 6];
                            return [4 /*yield*/, this.deactivateSmartrulesForAsset(asset, parentRef)];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.shouldShowWithDeviceUserCheckbox = function (asset) {
            var owner = asset.owner, isRootDevice = asset.c8y_IsDevice;
            var hasDeviceUserAsOwner = asset.owner && this.isDeviceUser(owner);
            return Boolean(isRootDevice && hasDeviceUserAsOwner);
        };
        SubAssetsService.prototype.getDefaultBulkActionControls = function () {
            return [];
        };
        SubAssetsService.prototype.getData = function (columns, pagination, parentReference, baseQuery) {
            if (baseQuery === void 0) { baseQuery = {}; }
            return __awaiter(this, void 0, void 0, function () {
                var isRoot, query, filters;
                return __generator(this, function (_a) {
                    isRoot = !parentReference;
                    if (isRoot) {
                        query = this.buildCombinedRootQueryFilter(columns, pagination);
                        return [2 /*return*/, this.assetNodeService.getRootNodes(Object.assign(Object.assign({}, pagination), { query: query }))];
                    }
                    filters = Object.assign(Object.assign({}, this.getAssetsFilters(columns, pagination, baseQuery)), { withParents: false });
                    if (this.deviceGroupService.isGroup(parentReference)) {
                        return [2 /*return*/, this.assetNodeService.getGroupItems(parentReference.id, filters)];
                    }
                    if (this.deviceGroupService.isDynamicGroup(parentReference)) {
                        return [2 /*return*/, this.assetNodeService.getDynamicGroupItems(parentReference.c8y_DeviceQueryString, filters)];
                    }
                    if (this.deviceGroupService.isDevice(parentReference)) {
                        return [2 /*return*/, this.assetNodeService.getDeviceChildren(parentReference.id, filters)];
                    }
                    return [2 /*return*/];
                });
            });
        };
        SubAssetsService.prototype.getCount = function (columns, pagination, parentReference, baseQuery) {
            if (baseQuery === void 0) { baseQuery = {}; }
            return __awaiter(this, void 0, void 0, function () {
                var defaultFilters, filters;
                return __generator(this, function (_a) {
                    defaultFilters = {
                        pageSize: 1,
                        withChildren: false
                    };
                    filters = !parentReference
                        ? Object.assign({ query: this.buildCombinedRootQueryFilter(columns, pagination) }, defaultFilters) : Object.assign(Object.assign({}, this.getAssetsFilters(columns, pagination, baseQuery)), defaultFilters);
                    return [2 /*return*/, this.getAssetsStatistics(parentReference, filters)];
                });
            });
        };
        SubAssetsService.prototype.getTotal = function (parentReference, baseQuery) {
            if (baseQuery === void 0) { baseQuery = {}; }
            var queryFilter = this.assetNodeService.rootQueryFilter();
            var query = !parentReference
                ? this.queriesUtil.addAndFilter(queryFilter, baseQuery)
                : baseQuery;
            var filters = {
                query: this.queriesUtil.buildQuery(query),
                withChildren: false,
                withTotalPages: true,
                pageSize: 1
            };
            return this.getAssetsStatistics(parentReference, filters);
        };
        SubAssetsService.prototype.canEditGroup = function (group) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.permissionsService.canEdit(['ROLE_INVENTORY_ADMIN'], group)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        SubAssetsService.prototype.canCreateGroup = function () {
            var currentUser = this.appState.currentUser.value;
            var hasAdminRole = this.user.hasAnyRole(currentUser, [
                'ROLE_INVENTORY_ADMIN',
                'ROLE_INVENTORY_CREATE'
            ]);
            return hasAdminRole;
        };
        SubAssetsService.prototype.canAssignDevice = function (group) {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.permissionsService.canEdit(['ROLE_INVENTORY_ADMIN'], group)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
        SubAssetsService.prototype.canEditSmartGroup = function () {
            var SMART_GROUPS_ROLES_EDIT = ['ROLE_SMARTGROUP_UPDATE', 'ROLE_SMARTGROUP_ADMIN'];
            return this.permissionsService.hasAnyRole(SMART_GROUPS_ROLES_EDIT);
        };
        SubAssetsService.prototype.canDeleteSmartGroup = function () {
            var SMART_GROUPS_ROLES_DELETE = ['ROLE_SMARTGROUP_ADMIN', 'ROLE_INVENTORY_ADMIN'];
            return this.permissionsService.hasAnyRole(SMART_GROUPS_ROLES_DELETE);
        };
        SubAssetsService.prototype.isSmartGroup = function (group) {
            return (this.smartGroupsService.isSmartGroup(group) || this.smartGroupsService.isSmartGroupV2(group));
        };
        SubAssetsService.prototype.isUsingInventoryRoles = function () {
            var currentUser = this.appState.currentUser.value;
            var hasAnyInventoryRole = this.user.hasAnyRole(currentUser, [
                'ROLE_INVENTORY_ADMIN',
                'ROLE_INVENTORY_READ',
                'ROLE_INVENTORY_CREATE'
            ]);
            return !hasAnyInventoryRole;
        };
        SubAssetsService.prototype.getAssetsStatistics = function (parentReference, filters) {
            return __awaiter(this, void 0, void 0, function () {
                var isRoot;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            isRoot = !parentReference;
                            if (!isRoot) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.assetNodeService.getRootNodes(filters)];
                        case 1: return [2 /*return*/, (_a.sent()).paging.totalPages];
                        case 2:
                            if (!this.deviceGroupService.isGroup(parentReference)) return [3 /*break*/, 4];
                            return [4 /*yield*/, this.assetNodeService.getGroupItems(parentReference.id, filters)];
                        case 3: return [2 /*return*/, (_a.sent()).paging
                                .totalPages];
                        case 4:
                            if (!this.deviceGroupService.isDynamicGroup(parentReference)) return [3 /*break*/, 6];
                            return [4 /*yield*/, this.assetNodeService.getDynamicGroupItems(parentReference.c8y_DeviceQueryString, filters)];
                        case 5: return [2 /*return*/, (_a.sent()).paging.totalPages];
                        case 6:
                            if (!this.deviceGroupService.isDevice(parentReference)) return [3 /*break*/, 8];
                            return [4 /*yield*/, this.assetNodeService.getDeviceChildren(parentReference.id, filters)];
                        case 7: return [2 /*return*/, (_a.sent()).paging
                                .totalPages];
                        case 8: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.buildCombinedRootQueryFilter = function (columns, pagination) {
            var queryFilter = this.assetNodeService.rootQueryFilter();
            var userQuery = this.getQueryObj(columns, pagination);
            var queryPart = this.queriesUtil.addOrderbys(queryFilter, userQuery.__orderby, 'append');
            var fullQuery = this.queriesUtil.addAndFilter(queryPart, userQuery.__filter);
            return this.queriesUtil.buildQuery(fullQuery);
        };
        SubAssetsService.prototype.deleteGroup = function (group, params) {
            if (params === void 0) { params = {}; }
            return __awaiter(this, void 0, void 0, function () {
                var cascade, _a, alertMessage, error_2, alertMessage;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            cascade = params.cascade;
                            _b.label = 1;
                        case 1:
                            _b.trys.push([1, 6, , 7]);
                            if (!(this.smartGroupsService.isSmartGroup(group) || this.smartGroupsService.isSmartGroupV2(group))) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.smartGroupsService.delete(group, { cascade: cascade })];
                        case 2:
                            _a = _b.sent();
                            return [3 /*break*/, 5];
                        case 3: return [4 /*yield*/, this.inventoryService.delete(group, { cascade: cascade })];
                        case 4:
                            _a = _b.sent();
                            _b.label = 5;
                        case 5:
                            _a;
                            alertMessage = this.translateService.instant(i3.gettext('Asset deleted.'));
                            this.alertService.success(alertMessage);
                            return [3 /*break*/, 7];
                        case 6:
                            error_2 = _b.sent();
                            alertMessage = this.translateService.instant(i3.gettext('Could not delete asset.'));
                            this.alertService.danger(alertMessage);
                            return [3 /*break*/, 7];
                        case 7: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.deleteDevice = function (device, params) {
            if (params === void 0) { params = {}; }
            return __awaiter(this, void 0, void 0, function () {
                var cascade, withDeviceUser, owner, shouldRemoveOwner, _a, alertMessage, error_3, alertMessage;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            cascade = params.cascade, withDeviceUser = params.withDeviceUser;
                            _b.label = 1;
                        case 1:
                            _b.trys.push([1, 6, , 7]);
                            owner = device.owner;
                            shouldRemoveOwner = withDeviceUser && owner && this.isDeviceUser(owner);
                            if (!shouldRemoveOwner) return [3 /*break*/, 3];
                            return [4 /*yield*/, this.deleteDeviceWithUser(device, cascade)];
                        case 2:
                            _a = _b.sent();
                            return [3 /*break*/, 5];
                        case 3: return [4 /*yield*/, this.inventoryService.delete(device, { cascade: cascade })];
                        case 4:
                            _a = _b.sent();
                            _b.label = 5;
                        case 5:
                            _a;
                            alertMessage = this.translateService.instant(i3.gettext('Asset deleted.'));
                            this.alertService.success(alertMessage);
                            return [3 /*break*/, 7];
                        case 6:
                            error_3 = _b.sent();
                            alertMessage = this.translateService.instant(i3.gettext('Could not delete asset.'));
                            this.alertService.danger(alertMessage);
                            return [3 /*break*/, 7];
                        case 7: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.deactivateSmartrulesForAsset = function (asset, parentRef) {
            return __awaiter(this, void 0, void 0, function () {
                var assetId, parentId, rules, upateSmartrulesPromises, error_4, alertMessage;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            assetId = asset.id;
                            parentId = parentRef.id;
                            return [4 /*yield*/, this.smartRulesService.listByContext(parentId)];
                        case 1:
                            rules = (_a.sent()).data;
                            upateSmartrulesPromises = rules.map(function (rule) { return _this.smartRulesService.bulkDeactivateEnabledSources(rule, [assetId]); });
                            _a.label = 2;
                        case 2:
                            _a.trys.push([2, 4, , 5]);
                            return [4 /*yield*/, Promise.all(upateSmartrulesPromises)];
                        case 3:
                            _a.sent();
                            return [3 /*break*/, 5];
                        case 4:
                            error_4 = _a.sent();
                            alertMessage = this.translateService.instant(i3.gettext('Could not deactivate smart rules.'));
                            this.alertService.danger(alertMessage);
                            return [3 /*break*/, 5];
                        case 5: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.isDeviceUser = function (userId) {
            return userId.match(/^device_/);
        };
        SubAssetsService.prototype.deleteDeviceWithUser = function (device, cascade) {
            return __awaiter(this, void 0, void 0, function () {
                var params, error_5;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            params = { cascade: cascade, withDeviceUser: true };
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 5]);
                            return [4 /*yield*/, this.inventoryService.delete(device, params)];
                        case 2: return [2 /*return*/, _a.sent()];
                        case 3:
                            error_5 = _a.sent();
                            return [4 /*yield*/, this.inventoryService.delete(device, { cascade: cascade })];
                        case 4: return [2 /*return*/, _a.sent()];
                        case 5: return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsService.prototype.getAssetsFilters = function (columns, pagination, baseQuery) {
            var query = this.queriesUtil.addAndFilter(this.getQueryObj(columns), baseQuery);
            return {
                query: this.queriesUtil.buildQuery(query),
                pageSize: pagination.pageSize || this.DEFAULT_PAGE_SIZE,
                currentPage: pagination.currentPage,
                withTotalPages: true
            };
        };
SubAssetsService.ɵfac = function SubAssetsService_Factory(t) { return new (t || SubAssetsService)(ɵngcc0.ɵɵinject(ɵngcc4.TranslateService), ɵngcc0.ɵɵinject(ɵngcc5.InventoryService), ɵngcc0.ɵɵinject(ɵngcc1.AppStateService), ɵngcc0.ɵɵinject(ɵngcc5.UserService), ɵngcc0.ɵɵinject(ɵngcc2.AssetNodeService), ɵngcc0.ɵɵinject(ɵngcc2.DeviceGroupService), ɵngcc0.ɵɵinject(ɵngcc5.SmartGroupsService), ɵngcc0.ɵɵinject(ɵngcc5.SmartRulesService), ɵngcc0.ɵɵinject(ɵngcc1.AlertService), ɵngcc0.ɵɵinject(ɵngcc1.Permissions), ɵngcc0.ɵɵinject(ɵngcc1.ModalService)); };
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsService, [{
        type: i0.Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc4.TranslateService }, { type: ɵngcc5.InventoryService }, { type: ɵngcc1.AppStateService }, { type: ɵngcc5.UserService }, { type: ɵngcc2.AssetNodeService }, { type: ɵngcc2.DeviceGroupService }, { type: ɵngcc5.SmartGroupsService }, { type: ɵngcc5.SmartRulesService }, { type: ɵngcc1.AlertService }, { type: ɵngcc1.Permissions }, { type: ɵngcc1.ModalService }]; }, null); })();
        return SubAssetsService;
    }(deviceGrid.DeviceGridService));
    SubAssetsService.ɵprov = i0__namespace.ɵɵdefineInjectable({ factory: function SubAssetsService_Factory() { return new SubAssetsService(i0__namespace.ɵɵinject(i1__namespace.TranslateService), i0__namespace.ɵɵinject(i2__namespace.InventoryService), i0__namespace.ɵɵinject(i3__namespace.AppStateService), i0__namespace.ɵɵinject(i4__namespace.UserService), i0__namespace.ɵɵinject(i5__namespace.AssetNodeService), i0__namespace.ɵɵinject(i5__namespace.DeviceGroupService), i0__namespace.ɵɵinject(i6__namespace.SmartGroupsService), i0__namespace.ɵɵinject(i7__namespace.SmartRulesService), i0__namespace.ɵɵinject(i3__namespace.AlertService), i0__namespace.ɵɵinject(i3__namespace.Permissions), i0__namespace.ɵɵinject(i3__namespace.ModalService)); }, token: SubAssetsService, providedIn: "root" });
    SubAssetsService.ctorParameters = function () { return [
        { type: i1.TranslateService },
        { type: client.InventoryService },
        { type: i3.AppStateService },
        { type: client.UserService },
        { type: i5.AssetNodeService },
        { type: i5.DeviceGroupService },
        { type: client.SmartGroupsService },
        { type: client.SmartRulesService },
        { type: i3.AlertService },
        { type: i3.Permissions },
        { type: i3.ModalService }
    ]; };

    var GroupInfoComponent = /** @class */ (function () {
        function GroupInfoComponent(inventory, subAssetsService, deviceGroupService, smartGroupsService, alertService, translate, modalService) {
            this.inventory = inventory;
            this.subAssetsService = subAssetsService;
            this.deviceGroupService = deviceGroupService;
            this.smartGroupsService = smartGroupsService;
            this.alertService = alertService;
            this.translate = translate;
            this.modalService = modalService;
            this.onGroupChange = new i0.EventEmitter();
            this.filterMsg = i3.gettext('Smart groups are groups dynamically constructed based on filtering criteria.');
            this.canEditMsg = i3.gettext('You can edit the filter here.');
            this.GROUP_UPDATED_MSG = i3.gettext('Group updated.');
        }
        GroupInfoComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a, _b;
                return __generator(this, function (_c) {
                    switch (_c.label) {
                        case 0:
                            _a = this;
                            if (!this.smartGroupsService.isSmartGroupV2(this.group)) return [3 /*break*/, 1];
                            _b = this.subAssetsService.canEditSmartGroup();
                            return [3 /*break*/, 3];
                        case 1: return [4 /*yield*/, this.subAssetsService.canEditGroup(this.group)];
                        case 2:
                            _b = _c.sent();
                            _c.label = 3;
                        case 3:
                            _a.canEdit = _b;
                            this.groupIcon = this.deviceGroupService.icon(this.group);
                            this.smartGroupFilter = this.group.c8y_DeviceQueryString;
                            this.setHintMsg();
                            return [2 /*return*/];
                    }
                });
            });
        };
        GroupInfoComponent.prototype.isSmartGroup = function () {
            return this.subAssetsService.isSmartGroup(this.group);
        };
        GroupInfoComponent.prototype.update = function (partialGroup) {
            return __awaiter(this, void 0, void 0, function () {
                var isSmartGroup, _a, error_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _b.trys.push([0, 5, , 6]);
                            isSmartGroup = this.subAssetsService.isSmartGroup(this.group);
                            if (!isSmartGroup) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.updateSmartGroup(partialGroup)];
                        case 1:
                            _a = _b.sent();
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, this.updateGroup(partialGroup)];
                        case 3:
                            _a = _b.sent();
                            _b.label = 4;
                        case 4:
                            _a;
                            return [3 /*break*/, 6];
                        case 5:
                            error_1 = _b.sent();
                            this.alertService.addServerFailure(error_1);
                            return [3 /*break*/, 6];
                        case 6: return [2 /*return*/];
                    }
                });
            });
        };
        GroupInfoComponent.prototype.setHintMsg = function () {
            var filterMsgTranslated = this.translate.instant(this.filterMsg);
            var canEditMsgTranslated = this.translate.instant(this.canEditMsg);
            this.filterHintMsg = this.canEdit
                ? filterMsgTranslated + " " + canEditMsgTranslated
                : this.filterMsg;
        };
        GroupInfoComponent.prototype.updateGroup = function (partialGroup) {
            return __awaiter(this, void 0, void 0, function () {
                var id, group, updatedGroup;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            id = this.group.id;
                            group = Object.assign({ id: id }, partialGroup);
                            return [4 /*yield*/, this.inventory.update(group)];
                        case 1:
                            updatedGroup = (_a.sent()).data;
                            this.setGroup(updatedGroup);
                            return [2 /*return*/];
                    }
                });
            });
        };
        GroupInfoComponent.prototype.updateSmartGroup = function (partialGroup) {
            return __awaiter(this, void 0, void 0, function () {
                var id, c8y_DeviceQueryString, group, updatedGroup, modalBody, title, e_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            id = this.group.id;
                            c8y_DeviceQueryString = partialGroup.c8y_DeviceQueryString;
                            group = Object.assign({ id: id }, partialGroup);
                            if (!!c8y_DeviceQueryString) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.smartGroupsService.update(group)];
                        case 1:
                            updatedGroup = (_a.sent()).data;
                            this.setGroup(updatedGroup);
                            return [2 /*return*/];
                        case 2:
                            _a.trys.push([2, 5, , 6]);
                            modalBody = i3.gettext('You are about to change the smart group filter. Do you want to proceed?');
                            title = i3.gettext('Smart group filter');
                            return [4 /*yield*/, this.modalService.confirm(title, modalBody, i3.Status.WARNING, {
                                    ok: i3.gettext('Save'),
                                    cancel: i3.gettext('Cancel')
                                })];
                        case 3:
                            _a.sent();
                            return [4 /*yield*/, this.isQueryExecutable(c8y_DeviceQueryString)];
                        case 4:
                            if (!(_a.sent())) {
                                return [2 /*return*/];
                            }
                            return [3 /*break*/, 6];
                        case 5:
                            e_1 = _a.sent();
                            return [2 /*return*/];
                        case 6: return [4 /*yield*/, this.smartGroupsService.update(group)];
                        case 7:
                            updatedGroup = (_a.sent()).data;
                            this.setGroup(updatedGroup);
                            return [2 /*return*/];
                    }
                });
            });
        };
        GroupInfoComponent.prototype.setGroup = function (group) {
            this.onGroupChange.emit(group);
            this.alertService.success(this.GROUP_UPDATED_MSG);
        };
        GroupInfoComponent.prototype.isQueryExecutable = function (query) {
            return __awaiter(this, void 0, void 0, function () {
                var filter, error_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            filter = { q: query };
                            return [4 /*yield*/, this.inventory.list(filter)];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, true];
                        case 2:
                            error_2 = _a.sent();
                            this.alertService.addServerFailure(error_2);
                            return [2 /*return*/, false];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
GroupInfoComponent.ɵfac = function GroupInfoComponent_Factory(t) { return new (t || GroupInfoComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc5.InventoryService), ɵngcc0.ɵɵdirectiveInject(SubAssetsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGroupService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.SmartGroupsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ModalService)); };
GroupInfoComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: GroupInfoComponent, selectors: [["c8y-group-info"]], inputs: { group: "group" }, outputs: { onGroupChange: "onGroupChange" }, decls: 48, vars: 25, consts: [[1, "bg-gray-white"], [1, "card-block", "p-t-24", "p-b-24", "large-padding"], [1, "content-flex-70"], [1, "text-center"], [1, "c8y-icon-duocolor", "icon-48", 3, "c8yIcon"], ["class", "label label-info", 4, "ngIf"], [1, "flex-grow", "col-10"], [1, "content-flex-80"], [1, "col-9"], ["groupNameForm", "ngForm"], [1, "form-group-lg", "m-b-0"], ["translate", "", 1, "sr-only"], ["class", "form-control-static", 4, "ngIf"], ["class", "input-group input-group-editable", 4, "ngIf"], ["groupDescriptionForm", "ngForm"], [4, "ngIf"], [1, "flex-grow"], [1, "m-b-8"], ["c8yIcon", "info-circle", 1, "text-info", "m-r-8"], [1, "text-label-small"], [1, "list-unstyled", "small"], [1, "p-t-4", "p-b-4", "flex-row", "separator-top-bottom", "text-nowrap"], [1, "small", "m-b-0", "m-r-8"], [1, "flex-item-right"], [1, "p-t-4", "p-b-4", "flex-row", "separator-bottom", "text-nowrap"], ["class", "p-t-4 p-b-4 flex-row separator-bottom text-nowrap", 4, "ngIf"], [1, "label", "label-info"], [1, "form-control-static"], [1, "input-group", "input-group-editable"], ["type", "text", "name", "name", "maxlength", "254", "required", "", 1, "form-control", 3, "ngModel", "title", "size", "placeholder", "ngModelChange"], [1, "input-group-btn"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], ["name", "description", 1, "form-control", 3, "ngModel", "title", "cols", "placeholder", "ngModelChange"], ["smartGroupFilterForm", "ngForm"], [1, "m-b-0", "m-t-8"], [1, "m-b-0", "text-nowrap"], ["type", "button", "triggers", "focus", 1, "btn-clean", "text-primary", "m-r-4", 3, "popover"], ["c8yIcon", "question-circle-o"], ["type", "text", "name", "filter", "maxlength", "254", 1, "form-control", 3, "ngModel", "size", "placeholder", "title", "ngModelChange"], ["class", "flex-item-right", 4, "ngIf"]], template: function GroupInfoComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelement(4, "i", 4);
        ɵngcc0.ɵɵelementStart(5, "p");
        ɵngcc0.ɵɵtemplate(6, GroupInfoComponent_small_6_Template, 3, 3, "small", 5);
        ɵngcc0.ɵɵtemplate(7, GroupInfoComponent_small_7_Template, 3, 3, "small", 5);
        ɵngcc0.ɵɵtemplate(8, GroupInfoComponent_small_8_Template, 3, 3, "small", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 6);
        ɵngcc0.ɵɵelementStart(10, "div", 7);
        ɵngcc0.ɵɵelementStart(11, "div", 8);
        ɵngcc0.ɵɵelementStart(12, "form", null, 9);
        ɵngcc0.ɵɵelementStart(14, "c8y-form-group", 10);
        ɵngcc0.ɵɵelementStart(15, "label", 11);
        ɵngcc0.ɵɵtext(16, " Name ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(17, GroupInfoComponent_p_17_Template, 2, 1, "p", 12);
        ɵngcc0.ɵɵtemplate(18, GroupInfoComponent_div_18_Template, 10, 15, "div", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(19, "form", null, 14);
        ɵngcc0.ɵɵelementStart(21, "label", 11);
        ɵngcc0.ɵɵtext(22, " Description ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(23, GroupInfoComponent_p_23_Template, 2, 1, "p", 12);
        ɵngcc0.ɵɵtemplate(24, GroupInfoComponent_div_24_Template, 10, 15, "div", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(25, GroupInfoComponent_form_25_Template, 11, 8, "form", 15);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(26, "div", 16);
        ɵngcc0.ɵɵelementStart(27, "p", 17);
        ɵngcc0.ɵɵelement(28, "i", 18);
        ɵngcc0.ɵɵelementStart(29, "span", 19);
        ɵngcc0.ɵɵtext(30);
        ɵngcc0.ɵɵpipe(31, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(32, "ul", 20);
        ɵngcc0.ɵɵelementStart(33, "li", 21);
        ɵngcc0.ɵɵelementStart(34, "label", 22);
        ɵngcc0.ɵɵtext(35);
        ɵngcc0.ɵɵpipe(36, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(37, "span", 23);
        ɵngcc0.ɵɵtext(38);
        ɵngcc0.ɵɵpipe(39, "c8yDate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(40, "li", 24);
        ɵngcc0.ɵɵelementStart(41, "label", 22);
        ɵngcc0.ɵɵtext(42);
        ɵngcc0.ɵɵpipe(43, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(44, "span", 23);
        ɵngcc0.ɵɵtext(45);
        ɵngcc0.ɵɵpipe(46, "c8yDate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(47, GroupInfoComponent_li_47_Template, 6, 5, "li", 25);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.groupIcon + "-open");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.group.c8y_IsDynamicGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.group.c8y_IsDynamicGroup && !ctx.group.com_cumulocity_model_Agent);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.group.com_cumulocity_model_Agent);
        ɵngcc0.ɵɵadvance(9);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.canEdit);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.canEdit);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.canEdit);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.canEdit);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(31, 15, "Group info"));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(36, 17, "Created"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(39, 19, ctx.group.creationTime));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(43, 21, "Last updated"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(46, 23, ctx.group.lastUpdated));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.group.com_cumulocity_model_Agent);
    } }, directives: [ɵngcc1.IconDirective, ɵngcc3.NgIf, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc1.FormGroupComponent, ɵngcc1.C8yTranslateDirective, ɵngcc1.RequiredInputPlaceholderDirective, ɵngcc6.DefaultValueAccessor, ɵngcc6.MaxLengthValidator, ɵngcc6.RequiredValidator, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.PopoverDirective], pipes: [ɵngcc1.C8yTranslatePipe, ɵngcc1.DatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GroupInfoComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-group-info',
                template: "<div class=\"bg-gray-white\">\n  <div class=\"card-block p-t-24 p-b-24 large-padding\">\n    <div class=\"content-flex-70\">\n      <div class=\"text-center\">\n        <i class=\"c8y-icon-duocolor icon-48\" [c8yIcon]=\"groupIcon + '-open'\"></i>\n        <p>\n          <small class=\"label label-info\" *ngIf=\"group.c8y_IsDynamicGroup\">\n            {{ 'Smart group' | translate }}\n          </small>\n          <small\n            class=\"label label-info\"\n            *ngIf=\"!group.c8y_IsDynamicGroup && !group.com_cumulocity_model_Agent\"\n          >\n            {{ 'Group' | translate }}\n          </small>\n          <small class=\"label label-info\" *ngIf=\"group.com_cumulocity_model_Agent\">\n            {{ 'Remote group' | translate }}\n          </small>\n        </p>\n      </div>\n\n      <div class=\"flex-grow col-10\">\n        <div class=\"content-flex-80\">\n          <div class=\"col-9\">\n            <form #groupNameForm=\"ngForm\">\n              <c8y-form-group class=\"form-group-lg m-b-0\">\n                <label class=\"sr-only\" translate>\n                  Name\n                </label>\n\n                <p *ngIf=\"!canEdit\" class=\"form-control-static\">{{ group.name }}</p>\n                <div *ngIf=\"canEdit\" class=\"input-group input-group-editable\">\n                  <input\n                    type=\"text\"\n                    class=\"form-control\"\n                    [(ngModel)]=\"group.name\"\n                    name=\"name\"\n                    title=\"{{ 'Name' | translate }}\"\n                    size=\"{{ group.name.length + 2 }}\"\n                    placeholder=\"{{ 'e.g. My group' | translate }}\"\n                    maxlength=\"254\"\n                    required\n                  />\n                  <span></span>\n                  <div class=\"input-group-btn\">\n                    <button\n                      (click)=\"update({ name: group.name }); groupNameForm.form.markAsPristine()\"\n                      class=\"btn btn-primary\"\n                      title=\"{{ 'Save' | translate }}\"\n                      [disabled]=\"groupNameForm.form.invalid\"\n                    >\n                      {{ 'Save' | translate }}\n                    </button>\n                  </div>\n                </div>\n              </c8y-form-group>\n            </form>\n            <form #groupDescriptionForm=\"ngForm\">\n              <label class=\"sr-only\" translate>\n                Description\n              </label>\n              <p *ngIf=\"!canEdit\" class=\"form-control-static\">{{ group.c8y_Notes }}</p>\n              <div *ngIf=\"canEdit\" class=\"input-group input-group-editable\">\n                <textarea\n                  class=\"form-control\"\n                  [(ngModel)]=\"group.c8y_Notes\"\n                  name=\"description\"\n                  title=\"{{ 'Description' | translate }}\"\n                  cols=\"{{ group.c8y_Notes ? group.c8y_Notes.length : 25 }}\"\n                  placeholder=\"{{ 'e.g. My description' | translate }}\"\n                ></textarea>\n                <span></span>\n                <div class=\"input-group-btn\">\n                  <button\n                    (click)=\"\n                      update({ c8y_Notes: group.c8y_Notes });\n                      groupDescriptionForm.form.markAsPristine()\n                    \"\n                    class=\"btn btn-primary\"\n                    title=\"{{ 'Save' | translate }}\"\n                    [disabled]=\"groupDescriptionForm.form.invalid\"\n                  >\n                    {{ 'Save' | translate }}\n                  </button>\n                </div>\n              </div>\n            </form>\n\n            <form #smartGroupFilterForm=\"ngForm\" *ngIf=\"isSmartGroup()\">\n              <c8y-form-group class=\"m-b-0 m-t-8\">\n                <label class=\"m-b-0 text-nowrap\">\n                  {{ 'Smart group filter' | translate }}\n                  <button\n                    class=\"btn-clean text-primary m-r-4\"\n                    type=\"button\"\n                    popover=\"{{ filterHintMsg | translate }}\"\n                    triggers=\"focus\"\n                  >\n                    <i c8yIcon=\"question-circle-o\"></i>\n                  </button>\n                </label>\n\n                <p *ngIf=\"!canEdit\" class=\"form-control-static\">\n                  {{ smartGroupFilter }}\n                </p>\n                <div *ngIf=\"canEdit\" class=\"input-group input-group-editable\">\n                  <input\n                    type=\"text\"\n                    class=\"form-control\"\n                    [(ngModel)]=\"smartGroupFilter\"\n                    name=\"filter\"\n                    size=\"{{ smartGroupFilter.length + 2 }}\"\n                    placeholder=\"{{ 'e.g.' | translate }} $filter=(id eq '12*')\"\n                    maxlength=\"254\"\n                    title=\"{{ 'Smart group filter' | translate }}\"\n                  />\n                  <span></span>\n                  <div class=\"input-group-btn\">\n                    <button\n                      (click)=\"\n                        update({ c8y_DeviceQueryString: smartGroupFilter });\n                        smartGroupFilterForm.form.markAsPristine()\n                      \"\n                      class=\"btn btn-primary\"\n                      title=\"{{ 'Save' | translate }}\"\n                      [disabled]=\"smartGroupFilterForm.form.invalid\"\n                    >\n                      {{ 'Save' | translate }}\n                    </button>\n                  </div>\n                </div>\n              </c8y-form-group>\n            </form>\n          </div>\n          <div class=\"flex-grow\">\n            <p class=\"m-b-8\">\n              <i c8yIcon=\"info-circle\" class=\"text-info m-r-8\"></i>\n              <span class=\"text-label-small\">{{ 'Group info' | translate }}</span>\n            </p>\n            <ul class=\"list-unstyled small\">\n              <li class=\"p-t-4 p-b-4 flex-row separator-top-bottom text-nowrap\">\n                <label class=\"small m-b-0 m-r-8\">{{ 'Created' | translate }}</label>\n                <span class=\"flex-item-right\">{{ group.creationTime | c8yDate }}</span>\n              </li>\n              <li class=\"p-t-4 p-b-4 flex-row separator-bottom text-nowrap\">\n                <label class=\"small m-b-0 m-r-8\">{{ 'Last updated' | translate }}</label>\n                <span class=\"flex-item-right\">{{ group.lastUpdated | c8yDate }}</span>\n              </li>\n\n              <li\n                *ngIf=\"group.com_cumulocity_model_Agent\"\n                class=\"p-t-4 p-b-4 flex-row separator-bottom text-nowrap\"\n              >\n                <label class=\"small m-b-0 m-r-8\">{{ 'Status' | translate }}</label>\n                <span class=\"flex-item-right\" *ngIf=\"group.c8y_BrokerSource\">\n                  {{ group.c8y_BrokerSource.status }}\n                </span>\n                <span class=\"flex-item-right\" *ngIf=\"!group.c8y_BrokerSource\">\n                  {{ 'Offline' | translate }}\n                </span>\n              </li>\n            </ul>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc5.InventoryService }, { type: SubAssetsService }, { type: ɵngcc2.DeviceGroupService }, { type: ɵngcc5.SmartGroupsService }, { type: ɵngcc1.AlertService }, { type: ɵngcc4.TranslateService }, { type: ɵngcc1.ModalService }]; }, { onGroupChange: [{
            type: i0.Output
        }], group: [{
            type: i0.Input
        }] }); })();
        return GroupInfoComponent;
    }());
    GroupInfoComponent.ctorParameters = function () { return [
        { type: client.InventoryService },
        { type: SubAssetsService },
        { type: i5.DeviceGroupService },
        { type: client.SmartGroupsService },
        { type: i3.AlertService },
        { type: i1.TranslateService },
        { type: i3.ModalService }
    ]; };
    GroupInfoComponent.propDecorators = {
        group: [{ type: i0.Input }],
        onGroupChange: [{ type: i0.Output }]
    };

    var SubAssetsComponent = /** @class */ (function () {
        function SubAssetsComponent(route, deviceGroupService, subAssetsService) {
            var _this = this;
            this.route = route;
            this.deviceGroupService = deviceGroupService;
            this.subAssetsService = subAssetsService;
            this.toggle = false;
            this.showAddGroup = false;
            this.showAssignDevices = false;
            this.showAssignChildDevices = false;
            this.refresh = new i0.EventEmitter();
            this.filterable = true;
            this.shouldDisableAddGroup = false;
            this.shouldDisableAssignDevices = false;
            this.dataSub = this.route.parent.data.subscribe(function (_a) {
                var contextData = _a.contextData;
                _this.title = contextData.name;
                _this.currentGroupId = contextData.id;
                _this.group = contextData;
                _this.filterable = !_this.deviceGroupService.isDynamicGroup(contextData);
            });
        }
        SubAssetsComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            this.shouldDisableAddGroup = !this.subAssetsService.canCreateGroup();
                            _a = this;
                            return [4 /*yield*/, !this.subAssetsService.canAssignDevice(this.group)];
                        case 1:
                            _a.shouldDisableAssignDevices = _b.sent();
                            this.displayOptions = {
                                striped: true,
                                bordered: false,
                                gridHeader: true,
                                filter: !this.isSmartGroup()
                            };
                            return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsComponent.prototype.groupChange = function (group) {
            /**
             * Group is cached in the context-route guard. To avoid loading old data (when switching tabs) we need to update
             * this object in the cache.
             */
            Object.assign(this.group, group);
            this.title = group.name;
            this.refresh.emit();
        };
        SubAssetsComponent.prototype.isSmartGroup = function () {
            return this.subAssetsService.isSmartGroup(this.group);
        };
        SubAssetsComponent.prototype.ngOnDestroy = function () {
            if (this.dataSub) {
                this.dataSub.unsubscribe();
            }
        };
SubAssetsComponent.ɵfac = function SubAssetsComponent_Factory(t) { return new (t || SubAssetsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc8.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGroupService), ɵngcc0.ɵɵdirectiveInject(SubAssetsService)); };
SubAssetsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SubAssetsComponent, selectors: [["c8y-sub-assets"]], decls: 18, vars: 20, consts: [[3, "placement", 4, "ngIf"], [1, "card", "content-fullpage", "d-flex", "d-col"], [3, "group", "onGroupChange"], [1, "d-contents", 3, "refresh", "parent-group", "filterable", "displayOptions"], [3, "ngClass"], [1, "bottom-drawer"], [3, "currentGroupId", "refresh", "onCancel", 4, "ngIf"], [1, "d-flex", "d-col", "no-align-items", "fit-h"], ["class", "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice", 4, "ngIf"], [1, "bottom-drawer", "m-t-40"], ["class", "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel", 4, "ngIf"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle", 1, "m-r-4"], ["translate", ""], [3, "currentGroupId", "refresh", "onCancel"], [1, "d-contents", 3, "refresh", "currentGroupId", "onCancel", "onShowChildDevices", "selectedDevice"], [1, "d-contents", 3, "refresh", "currentGroupId", "parentDevice", "onCancel"]], template: function SubAssetsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, SubAssetsComponent_c8y_action_bar_item_2_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵtemplate(3, SubAssetsComponent_c8y_action_bar_item_3_Template, 6, 5, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "div", 1);
        ɵngcc0.ɵɵelementStart(5, "c8y-group-info", 2);
        ɵngcc0.ɵɵlistener("onGroupChange", function SubAssetsComponent_Template_c8y_group_info_onGroupChange_5_listener($event) { return ctx.groupChange($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(6, "c8y-sub-assets-grid", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "div", 4);
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵtemplate(9, SubAssetsComponent_c8y_add_group_9_Template, 1, 2, "c8y-add-group", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 4);
        ɵngcc0.ɵɵelementStart(11, "div", 5);
        ɵngcc0.ɵɵelementStart(12, "div", 7);
        ɵngcc0.ɵɵtemplate(13, SubAssetsComponent_c8y_assign_devices_13_Template, 1, 2, "c8y-assign-devices", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 4);
        ɵngcc0.ɵɵelementStart(15, "div", 9);
        ɵngcc0.ɵɵelementStart(16, "div", 7);
        ɵngcc0.ɵɵtemplate(17, SubAssetsComponent_c8y_assign_child_devices_17_Template, 1, 3, "c8y-assign-child-devices", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.title, "\n");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSmartGroup());
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("group", ctx.group);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("refresh", ctx.refresh)("parent-group", ctx.group)("filterable", ctx.filterable)("displayOptions", ctx.displayOptions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c0, ctx.showAddGroup));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(16, _c0, ctx.showAssignDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignDevices);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(18, _c0, ctx.showAssignChildDevices));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAssignChildDevices);
    } }, directives: function () { return [ɵngcc1.TitleComponent, ɵngcc3.NgIf, GroupInfoComponent, SubAssetsGridComponent, ɵngcc3.NgClass, ɵngcc1.ActionBarItemComponent, ɵngcc1.IconDirective, ɵngcc1.C8yTranslateDirective, AddGroupComponent, AssignDevicesComponent, AssignChildDevicesComponent]; }, pipes: function () { return [ɵngcc1.C8yTranslatePipe]; }, encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-sub-assets',
                template: "<c8y-title>\n  {{ title }}\n</c8y-title>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n    [disabled]=\"shouldDisableAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\" *ngIf=\"!isSmartGroup()\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Assign devices' | translate }}\"\n    (click)=\"showAssignDevices = !showAssignDevices\"\n    [disabled]=\"shouldDisableAssignDevices\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Assign devices</span>\n  </button>\n</c8y-action-bar-item>\n\n<div class=\"card content-fullpage d-flex d-col\">\n  <c8y-group-info [group]=\"group\" (onGroupChange)=\"groupChange($event)\"></c8y-group-info>\n  <c8y-sub-assets-grid\n    [refresh]=\"refresh\"\n    [parent-group]=\"group\"\n    [filterable]=\"filterable\"\n    [displayOptions]=\"displayOptions\"\n    class=\"d-contents\"\n  ></c8y-sub-assets-grid>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [currentGroupId]=\"currentGroupId\"\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignDevices }\">\n  <div class=\"bottom-drawer\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-devices\n        class=\"d-contents\"\n        (onCancel)=\"showAssignDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        (onShowChildDevices)=\"showAssignChildDevices = $event\"\n        (selectedDevice)=\"showChildrenForDevice = $event\"\n        *ngIf=\"showAssignDevices\"\n      ></c8y-assign-devices>\n    </div>\n  </div>\n</div>\n\n<div [ngClass]=\"{ drawerOpen: showAssignChildDevices }\">\n  <div class=\"bottom-drawer m-t-40\">\n    <div class=\"d-flex d-col no-align-items fit-h\">\n      <c8y-assign-child-devices\n        *ngIf=\"showAssignChildDevices\"\n        class=\"d-contents\"\n        (onCancel)=\"showAssignChildDevices = false\"\n        [refresh]=\"refresh\"\n        [currentGroupId]=\"currentGroupId\"\n        [parentDevice]=\"showChildrenForDevice\"\n      ></c8y-assign-child-devices>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc8.ActivatedRoute }, { type: ɵngcc2.DeviceGroupService }, { type: SubAssetsService }]; }, null); })();
        return SubAssetsComponent;
    }());
    SubAssetsComponent.ctorParameters = function () { return [
        { type: router.ActivatedRoute },
        { type: i5.DeviceGroupService },
        { type: SubAssetsService }
    ]; };

    var DeleteAssetsModalComponent = /** @class */ (function () {
        function DeleteAssetsModalComponent(translateService) {
            this.translateService = translateService;
            this.showWithCascadeCheckbox = true;
            this.showWithDeviceUserCheckbox = false;
            this.closeSubject = new rxjs.Subject();
            this.labels = { ok: i3.gettext('Delete'), cancel: i3.gettext('Cancel') };
            this.title = i3.gettext('Delete asset');
            this.status = i3.Status.DANGER;
            this.config = {
                cascade: false,
                withDeviceUser: false
            };
        }
        DeleteAssetsModalComponent.prototype.ngOnInit = function () {
            this.message = this.translateService.instant(i3.gettext('You are about to delete asset: "{{name}}". This operation is irreversible. Do you want to proceed?'), { name: this.asset.name });
        };
        DeleteAssetsModalComponent.prototype.ngAfterViewInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, this.modalRef.result];
                        case 1:
                            _a.sent();
                            this.onClose();
                            return [3 /*break*/, 3];
                        case 2:
                            error_1 = _a.sent();
                            this.onDismiss();
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        DeleteAssetsModalComponent.prototype.onClose = function () {
            this.closeSubject.next(this.config);
            this.closeSubject.complete();
        };
        DeleteAssetsModalComponent.prototype.onDismiss = function () {
            this.closeSubject.complete();
        };
DeleteAssetsModalComponent.ɵfac = function DeleteAssetsModalComponent_Factory(t) { return new (t || DeleteAssetsModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService)); };
DeleteAssetsModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeleteAssetsModalComponent, selectors: [["c8y-delete-assets-modal"]], viewQuery: function DeleteAssetsModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalRef = _t.first);
    } }, inputs: { showWithCascadeCheckbox: "showWithCascadeCheckbox", showWithDeviceUserCheckbox: "showWithDeviceUserCheckbox", asset: "asset" }, decls: 9, vars: 8, consts: [[3, "title", "status", "labels"], ["modalRef", ""], ["assetsForm", "ngForm"], [1, "text-wrap", "m-b-16"], ["class", "m-b-0", 4, "ngIf"], [1, "m-b-0"], [1, "c8y-checkbox", 3, "title"], ["type", "checkbox", "name", "cascade", 3, "ngModel", "ngModelChange"], ["type", "checkbox", "name", "withDeviceUser", 3, "ngModel", "ngModelChange"]], template: function DeleteAssetsModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-confirm-modal", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "form", null, 2);
        ɵngcc0.ɵɵelementStart(4, "p", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, DeleteAssetsModalComponent_c8y_form_group_7_Template, 8, 7, "c8y-form-group", 4);
        ɵngcc0.ɵɵtemplate(8, DeleteAssetsModalComponent_c8y_form_group_8_Template, 8, 7, "c8y-form-group", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ctx.title)("status", ctx.status)("labels", ctx.labels);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 6, ctx.message), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showWithCascadeCheckbox);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showWithDeviceUserCheckbox);
    } }, directives: [ɵngcc1.ConfirmModalComponent, ɵngcc6.ɵNgNoValidate, ɵngcc6.NgControlStatusGroup, ɵngcc6.NgForm, ɵngcc3.NgIf, ɵngcc1.FormGroupComponent, ɵngcc6.CheckboxControlValueAccessor, ɵngcc6.NgControlStatus, ɵngcc6.NgModel], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeleteAssetsModalComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-delete-assets-modal',
                template: "<c8y-confirm-modal [title]=\"title\" [status]=\"status\" [labels]=\"labels\" #modalRef>\n  <form #assetsForm=\"ngForm\">\n    <p class=\"text-wrap m-b-16\">\n      {{ message | translate }}\n    </p>\n    <c8y-form-group *ngIf=\"showWithCascadeCheckbox\" class=\"m-b-0\">\n      <label title=\"{{ 'Delete devices' | translate }}\" class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"cascade\" [(ngModel)]=\"config.cascade\" />\n        <span></span>\n        <span>\n          {{ 'Also delete all devices inside selected asset and its subassets.' | translate }}\n        </span>\n      </label>\n    </c8y-form-group>\n    <c8y-form-group *ngIf=\"showWithDeviceUserCheckbox\" class=\"m-b-0\">\n      <label title=\"{{ 'Delete associated device owner' | translate }}\" class=\"c8y-checkbox\">\n        <input type=\"checkbox\" name=\"withDeviceUser\" [(ngModel)]=\"config.withDeviceUser\" />\n        <span></span>\n        <span>\n          {{ 'Also delete associated device owner.' | translate }}\n        </span>\n      </label>\n    </c8y-form-group>\n  </form>\n</c8y-confirm-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc4.TranslateService }]; }, { showWithCascadeCheckbox: [{
            type: i0.Input
        }], showWithDeviceUserCheckbox: [{
            type: i0.Input
        }], asset: [{
            type: i0.Input
        }], modalRef: [{
            type: i0.ViewChild,
            args: ['modalRef', { static: false }]
        }] }); })();
        return DeleteAssetsModalComponent;
    }());
    DeleteAssetsModalComponent.ctorParameters = function () { return [
        { type: i1.TranslateService }
    ]; };
    DeleteAssetsModalComponent.propDecorators = {
        showWithCascadeCheckbox: [{ type: i0.Input }],
        showWithDeviceUserCheckbox: [{ type: i0.Input }],
        asset: [{ type: i0.Input }],
        modalRef: [{ type: i0.ViewChild, args: ['modalRef', { static: false },] }]
    };

    var UnassignModalComponent = /** @class */ (function () {
        function UnassignModalComponent(translateService) {
            this.translateService = translateService;
            this.closeSubject = new rxjs.Subject();
            this.labels = { ok: i3.gettext('Unassign'), cancel: i3.gettext('Cancel') };
            this.title = i3.gettext('Unassign asset');
            this.status = i3.Status.WARNING;
        }
        UnassignModalComponent.prototype.ngOnInit = function () {
            this.message = this.translateService.instant(i3.gettext('You are about to unassign asset "{{name}}". Do you want to proceed?'), { name: this.asset.name });
        };
        UnassignModalComponent.prototype.ngAfterViewInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            _a.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, this.modalRef.result];
                        case 1:
                            _a.sent();
                            this.onClose();
                            return [3 /*break*/, 3];
                        case 2:
                            error_1 = _a.sent();
                            this.onDismiss();
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        };
        UnassignModalComponent.prototype.onClose = function () {
            this.closeSubject.next(true);
            this.closeSubject.complete();
        };
        UnassignModalComponent.prototype.onDismiss = function () {
            this.closeSubject.complete();
        };
UnassignModalComponent.ɵfac = function UnassignModalComponent_Factory(t) { return new (t || UnassignModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService)); };
UnassignModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UnassignModalComponent, selectors: [["c8y-unassign-modal"]], viewQuery: function UnassignModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c1, 5);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalRef = _t.first);
    } }, inputs: { asset: "asset" }, decls: 5, vars: 6, consts: [[3, "title", "status", "labels"], ["modalRef", ""]], template: function UnassignModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-confirm-modal", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "span");
        ɵngcc0.ɵɵtext(3);
        ɵngcc0.ɵɵpipe(4, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ctx.title)("status", ctx.status)("labels", ctx.labels);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, ctx.message));
    } }, directives: [ɵngcc1.ConfirmModalComponent], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UnassignModalComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-unassign-modal',
                template: "<c8y-confirm-modal [title]=\"title\" [status]=\"status\" [labels]=\"labels\" #modalRef>\n  <span>{{ message | translate }}</span>\n</c8y-confirm-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc4.TranslateService }]; }, { asset: [{
            type: i0.Input
        }], modalRef: [{
            type: i0.ViewChild,
            args: ['modalRef', { static: false }]
        }] }); })();
        return UnassignModalComponent;
    }());
    UnassignModalComponent.ctorParameters = function () { return [
        { type: i1.TranslateService }
    ]; };
    UnassignModalComponent.propDecorators = {
        asset: [{ type: i0.Input }],
        modalRef: [{ type: i0.ViewChild, args: ['modalRef', { static: false },] }]
    };

    var SubAssetsGridComponent = /** @class */ (function () {
        function SubAssetsGridComponent(subAssetsGridService, bsModalService, smartGroupsService) {
            this.subAssetsGridService = subAssetsGridService;
            this.bsModalService = bsModalService;
            this.smartGroupsService = smartGroupsService;
            this.title = i3.gettext('Subassets');
            this.emptyStateText = i3.gettext('Add your first group or assign devices using the buttons on the action bar.');
            this.loadingItemsLabel = i3.gettext('Loading assets…');
            this.selectable = false;
            this.baseQuery = {};
            this.filterable = true;
            this.sortable = true;
            this.onColumnsChange = new i0.EventEmitter();
            this.itemsSelect = new i0.EventEmitter();
            this.pagination = this.subAssetsGridService.getDefaultPagination();
            this.showCounterWarning = false;
            this.bulkActionControls = this.subAssetsGridService.getDefaultBulkActionControls();
            this.displayOptions = {
                striped: true,
                bordered: false,
                gridHeader: true,
                filter: true
            };
            this.serverSideDataCallback = this.onDataSourceModifier.bind(this);
        }
        Object.defineProperty(SubAssetsGridComponent.prototype, "columns", {
            get: function () {
                return this._columns;
            },
            set: function (value) {
                if (value) {
                    this._columns = this.subAssetsGridService.getUserConfiguredColumns(value);
                }
                else {
                    this._columns = this.subAssetsGridService.getUserConfiguredColumns(this.subAssetsGridService.getDefaultColumns());
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "_pagination", {
            set: function (value) {
                if (value) {
                    this.pagination = value;
                }
                else {
                    this.pagination = this.subAssetsGridService.getDefaultPagination();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "_actionControls", {
            set: function (value) {
                if (value) {
                    this.actionControls = value;
                }
                else {
                    this.actionControls = this.subAssetsGridService.getDefaultActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "_bulkActionControls", {
            set: function (value) {
                if (value) {
                    this.bulkActionControls = value;
                }
                else {
                    this.bulkActionControls = this.subAssetsGridService.getDefaultBulkActionControls();
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "isRootGroup", {
            get: function () {
                return !this.parentGroup;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "getInfiniteScrollMode", {
            get: function () {
                return this.isRootGroup && this.subAssetsGridService.isUsingInventoryRoles()
                    ? 'auto'
                    : undefined;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(SubAssetsGridComponent.prototype, "_displayOptions", {
            set: function (displayOptions) {
                this.displayOptions = Object.assign(Object.assign({}, this.displayOptions), displayOptions);
            },
            enumerable: false,
            configurable: true
        });
        SubAssetsGridComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.columns = this.subAssetsGridService.getDefaultColumns(this.filterable, this.sortable);
            if (!this.filterable || !this.sortable) {
                this.columns.forEach(function (column) {
                    column.filterable = _this.filterable;
                    column.sortable = _this.sortable;
                });
            }
            this.setActionControls();
        };
        SubAssetsGridComponent.prototype.setActionControls = function () {
            return __awaiter(this, void 0, void 0, function () {
                var actionControls, isMicroserviceInstalled, deleteAction, unassignAction;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            actionControls = [];
                            return [4 /*yield*/, this.smartGroupsService.isSmartGroupsV2MicroserviceInstalled()];
                        case 1:
                            isMicroserviceInstalled = (_a.sent()).data;
                            deleteAction = {
                                type: "DELETE" /* Delete */,
                                callback: function (asset) { return _this.onDeleteAsset(asset, _this.parentGroup); },
                                showIf: function (asset) {
                                    if (_this.smartGroupsService.isSmartGroupV2(asset)) {
                                        return isMicroserviceInstalled ? _this.subAssetsGridService.canDeleteSmartGroup() : false;
                                    }
                                    if (_this.smartGroupsService.isSmartGroup(asset)) {
                                        return _this.subAssetsGridService.canDeleteSmartGroup();
                                    }
                                    return true;
                                }
                            };
                            actionControls.push(deleteAction);
                            unassignAction = {
                                type: 'UNASSIGN',
                                icon: 'unlink',
                                text: i3.gettext('Unassign'),
                                callback: function (asset) { return _this.onUnassignAsset(asset, _this.parentGroup); },
                                showIf: function (asset) { return _this.subAssetsGridService.isDevice(asset) &&
                                    !_this.subAssetsGridService.isSmartGroup(_this.parentGroup); }
                            };
                            actionControls.push(unassignAction);
                            if (!this.actionControls) {
                                this.actionControls = actionControls;
                            }
                            return [2 /*return*/];
                    }
                });
            });
        };
        SubAssetsGridComponent.prototype.onUnassignAsset = function (asset, parentRef) {
            var _this = this;
            var initialState = {
                asset: asset
            };
            var modalRef = this.bsModalService.show(UnassignModalComponent, { initialState: initialState });
            modalRef.content.closeSubject.subscribe(function (result) { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (!result) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.subAssetsGridService.unassignAsset(asset, parentRef)];
                        case 1:
                            _a.sent();
                            this.refresh.emit();
                            _a.label = 2;
                        case 2: return [2 /*return*/];
                    }
                });
            }); });
        };
        SubAssetsGridComponent.prototype.onDeleteAsset = function (asset, parentRef) {
            return __awaiter(this, void 0, void 0, function () {
                var initialState, modalRef;
                var _this = this;
                return __generator(this, function (_a) {
                    initialState = {
                        showWithDeviceUserCheckbox: this.subAssetsGridService.shouldShowWithDeviceUserCheckbox(asset),
                        asset: asset,
                        showWithCascadeCheckbox: !this.smartGroupsService.isSmartGroup(asset) &&
                            !this.smartGroupsService.isSmartGroupV2(asset)
                    };
                    modalRef = this.bsModalService.show(DeleteAssetsModalComponent, { initialState: initialState });
                    modalRef.content.closeSubject.subscribe(function (result) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    if (!result) return [3 /*break*/, 2];
                                    return [4 /*yield*/, this.subAssetsGridService.deleteAsset(asset, parentRef, result)];
                                case 1:
                                    _a.sent();
                                    this.showCounterWarning = true;
                                    this.refresh.emit();
                                    _a.label = 2;
                                case 2: return [2 /*return*/];
                            }
                        });
                    }); });
                    return [2 /*return*/];
                });
            });
        };
        SubAssetsGridComponent.prototype.ngOnChanges = function (changes) {
            if (changes.parentGroup && !changes.parentGroup.firstChange) {
                this.dataGrid.reload();
            }
        };
        SubAssetsGridComponent.prototype.trackByName = function (_index, column) {
            return column.name;
        };
        SubAssetsGridComponent.prototype.onDataSourceModifier = function (dataSourceModifier) {
            return __awaiter(this, void 0, void 0, function () {
                var promises, counters, _a, dataResponse, size, filteredSize;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            promises = [];
                            promises.push(this.subAssetsGridService.getData(dataSourceModifier.columns, dataSourceModifier.pagination, this.parentGroup, this.baseQuery));
                            promises.push(this.subAssetsGridService.getTotal(this.parentGroup, this.baseQuery));
                            promises.push(this.subAssetsGridService.getCount(dataSourceModifier.columns, dataSourceModifier.pagination, this.parentGroup, this.baseQuery));
                            return [4 /*yield*/, Promise.all(promises)];
                        case 1:
                            _a = __read.apply(void 0, [_b.sent(), 3]), dataResponse = _a[0], size = _a[1], filteredSize = _a[2];
                            if (!counters) {
                                counters = {
                                    size: size,
                                    filteredSize: filteredSize
                                };
                            }
                            this.onColumnsChange.emit(dataSourceModifier.columns);
                            return [2 /*return*/, Object.assign({ res: dataResponse.res, data: dataResponse.data, paging: dataResponse.paging }, counters)];
                    }
                });
            });
        };
        SubAssetsGridComponent.prototype.configChange = function (config) {
            this.subAssetsGridService.saveConfig(config);
        };
SubAssetsGridComponent.ɵfac = function SubAssetsGridComponent_Factory(t) { return new (t || SubAssetsGridComponent)(ɵngcc0.ɵɵdirectiveInject(SubAssetsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc9.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.SmartGroupsService)); };
SubAssetsGridComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SubAssetsGridComponent, selectors: [["c8y-sub-assets-grid"]], viewQuery: function SubAssetsGridComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(i3.DataGridComponent, 7);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dataGrid = _t.first);
    } }, inputs: { title: "title", emptyStateText: "emptyStateText", loadingItemsLabel: "loadingItemsLabel", selectable: "selectable", baseQuery: "baseQuery", filterable: "filterable", sortable: "sortable", columns: "columns", _pagination: ["pagination", "_pagination"], _actionControls: ["actionControls", "_actionControls"], _bulkActionControls: ["bulkActionControls", "_bulkActionControls"], _displayOptions: ["displayOptions", "_displayOptions"], parentGroup: ["parent-group", "parentGroup"], refresh: "refresh" }, outputs: { onColumnsChange: "onColumnsChange", itemsSelect: "itemsSelect" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 12, vars: 20, consts: [[1, "d-contents", 3, "title", "loadingItemsLabel", "columns", "pagination", "actionControls", "selectable", "bulkActionControls", "serverSideDataCallback", "infiniteScroll", "showCounterWarning", "refresh", "displayOptions", "onConfigChange", "itemsSelect"], [4, "ngFor", "ngForOf", "ngForTrackBy"], [1, "c8y-empty-state"], ["c8yIcon", "c8y-group-add", 1, "c8y-icon-duocolor"], [3, "name"]], template: function SubAssetsGridComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-data-grid", 0);
        ɵngcc0.ɵɵlistener("onConfigChange", function SubAssetsGridComponent_Template_c8y_data_grid_onConfigChange_0_listener($event) { return ctx.configChange($event); })("itemsSelect", function SubAssetsGridComponent_Template_c8y_data_grid_itemsSelect_0_listener($event) { return ctx.itemsSelect.emit($event); });
        ɵngcc0.ɵɵtemplate(1, SubAssetsGridComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "h1", 3);
        ɵngcc0.ɵɵelementStart(4, "div");
        ɵngcc0.ɵɵelementStart(5, "p");
        ɵngcc0.ɵɵelementStart(6, "strong");
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "small");
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ctx.title)("loadingItemsLabel", ctx.loadingItemsLabel)("columns", ctx.columns)("pagination", ctx.pagination)("actionControls", ctx.actionControls)("selectable", ctx.selectable)("bulkActionControls", ctx.bulkActionControls)("serverSideDataCallback", ctx.serverSideDataCallback)("infiniteScroll", ctx.getInfiniteScrollMode)("showCounterWarning", ctx.showCounterWarning)("refresh", ctx.refresh)("displayOptions", ctx.displayOptions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.columns)("ngForTrackBy", ctx.trackByName);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 16, "No items to display."));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 18, ctx.emptyStateText));
    } }, directives: [ɵngcc1.DataGridComponent, ɵngcc3.NgForOf, ɵngcc1.IconDirective, ɵngcc1.ColumnDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsGridComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-sub-assets-grid',
                template: "<c8y-data-grid\n  [title]=\"title\"\n  [loadingItemsLabel]=\"loadingItemsLabel\"\n  [columns]=\"columns\"\n  [pagination]=\"pagination\"\n  [actionControls]=\"actionControls\"\n  [selectable]=\"selectable\"\n  [bulkActionControls]=\"bulkActionControls\"\n  [serverSideDataCallback]=\"serverSideDataCallback\"\n  [infiniteScroll]=\"getInfiniteScrollMode\"\n  [showCounterWarning]=\"showCounterWarning\"\n  [refresh]=\"refresh\"\n  [displayOptions]=\"displayOptions\"\n  (onConfigChange)=\"configChange($event)\"\n  (itemsSelect)=\"itemsSelect.emit($event)\"\n  class=\"d-contents\"\n>\n  <ng-container *ngFor=\"let column of columns; trackBy: trackByName\">\n    <c8y-column [name]=\"column.name\"></c8y-column>\n  </ng-container>\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"c8y-group-add\" class=\"c8y-icon-duocolor\"></h1>\n    <div>\n      <p>\n        <strong>{{ 'No items to display.' | translate }}</strong>\n      </p>\n      <small>{{ emptyStateText | translate }}</small>\n    </div>\n  </div>\n</c8y-data-grid>\n"
            }]
    }], function () { return [{ type: SubAssetsService }, { type: ɵngcc9.BsModalService }, { type: ɵngcc5.SmartGroupsService }]; }, { title: [{
            type: i0.Input
        }], emptyStateText: [{
            type: i0.Input
        }], loadingItemsLabel: [{
            type: i0.Input
        }], selectable: [{
            type: i0.Input
        }], baseQuery: [{
            type: i0.Input
        }], filterable: [{
            type: i0.Input
        }], sortable: [{
            type: i0.Input
        }], onColumnsChange: [{
            type: i0.Output
        }], itemsSelect: [{
            type: i0.Output
        }], columns: [{
            type: i0.Input
        }], _pagination: [{
            type: i0.Input,
            args: ['pagination']
        }], _actionControls: [{
            type: i0.Input,
            args: ['actionControls']
        }], _bulkActionControls: [{
            type: i0.Input,
            args: ['bulkActionControls']
        }], _displayOptions: [{
            type: i0.Input,
            args: ['displayOptions']
        }], parentGroup: [{
            type: i0.Input,
            args: ['parent-group']
        }], refresh: [{
            type: i0.Input
        }], dataGrid: [{
            type: i0.ViewChild,
            args: [i3.DataGridComponent, { static: true }]
        }] }); })();
        return SubAssetsGridComponent;
    }());
    SubAssetsGridComponent.ctorParameters = function () { return [
        { type: SubAssetsService },
        { type: modal.BsModalService },
        { type: client.SmartGroupsService }
    ]; };
    SubAssetsGridComponent.propDecorators = {
        parentGroup: [{ type: i0.Input, args: ['parent-group',] }],
        refresh: [{ type: i0.Input }],
        title: [{ type: i0.Input }],
        emptyStateText: [{ type: i0.Input }],
        loadingItemsLabel: [{ type: i0.Input }],
        columns: [{ type: i0.Input }],
        _pagination: [{ type: i0.Input, args: ['pagination',] }],
        _actionControls: [{ type: i0.Input, args: ['actionControls',] }],
        selectable: [{ type: i0.Input }],
        baseQuery: [{ type: i0.Input }],
        _bulkActionControls: [{ type: i0.Input, args: ['bulkActionControls',] }],
        filterable: [{ type: i0.Input }],
        sortable: [{ type: i0.Input }],
        onColumnsChange: [{ type: i0.Output }],
        itemsSelect: [{ type: i0.Output }],
        dataGrid: [{ type: i0.ViewChild, args: [i3.DataGridComponent, { static: true },] }],
        _displayOptions: [{ type: i0.Input, args: ['displayOptions',] }]
    };

    var AddGroupService = /** @class */ (function () {
        function AddGroupService(inventoryService) {
            this.inventoryService = inventoryService;
            this.GROUP_FRAGMENT_TYPE = 'c8y_IsDeviceGroup';
        }
        AddGroupService.prototype.createGroupAndAssignDevices = function (groupForm, groupContextId, selectedDevices) {
            return __awaiter(this, void 0, void 0, function () {
                var group, name, description, newGroupMO;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            name = groupForm.name, description = groupForm.description;
                            newGroupMO = this.getGroupMO(name, description, groupContextId);
                            if (!groupContextId) return [3 /*break*/, 2];
                            return [4 /*yield*/, this.inventoryService.childAssetsCreate(newGroupMO, groupContextId)];
                        case 1:
                            group = (_a.sent()).data;
                            return [3 /*break*/, 4];
                        case 2: return [4 /*yield*/, this.inventoryService.create(newGroupMO)];
                        case 3:
                            group = (_a.sent()).data;
                            _a.label = 4;
                        case 4:
                            if (!(selectedDevices.length > 0)) return [3 /*break*/, 6];
                            return [4 /*yield*/, this.assignDevices(group.id, selectedDevices)];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6: return [2 /*return*/, group];
                    }
                });
            });
        };
        AddGroupService.prototype.getGroupMO = function (name, description, groupContextId) {
            var _a;
            if (description === void 0) { description = ''; }
            var group = (_a = {
                    type: this.getGroupType(groupContextId)
                },
                _a[this.GROUP_FRAGMENT_TYPE] = {},
                _a.name = name,
                _a.c8y_Notes = description,
                _a);
            return group;
        };
        AddGroupService.prototype.getGroupType = function (groupContextId) {
            return groupContextId ? i5.GroupFragment.subGroupType : i5.GroupFragment.groupType;
        };
        AddGroupService.prototype.assignDevices = function (id, selectedDevices) {
            return __awaiter(this, void 0, void 0, function () {
                var promises;
                var _this = this;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            promises = [];
                            selectedDevices.forEach(function (moId) {
                                promises.push(_this.inventoryService.childAssetsAdd(moId, id));
                            });
                            return [4 /*yield*/, Promise.all(promises)];
                        case 1: return [2 /*return*/, _a.sent()];
                    }
                });
            });
        };
AddGroupService.ɵfac = function AddGroupService_Factory(t) { return new (t || AddGroupService)(ɵngcc0.ɵɵinject(ɵngcc5.InventoryService)); };
AddGroupService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: AddGroupService, factory: function (t) { return AddGroupService.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddGroupService, [{
        type: i0.Injectable
    }], function () { return [{ type: ɵngcc5.InventoryService }]; }, null); })();
        return AddGroupService;
    }());
    AddGroupService.ctorParameters = function () { return [
        { type: client.InventoryService }
    ]; };

    var AddGroupComponent = /** @class */ (function () {
        function AddGroupComponent(deviceGridService, fb, addGroupService, alert, subAssetsService) {
            this.deviceGridService = deviceGridService;
            this.fb = fb;
            this.addGroupService = addGroupService;
            this.alert = alert;
            this.subAssetsService = subAssetsService;
            this.refresh = new i0.EventEmitter();
            this.onDeviceQueryStringChange = new i0.EventEmitter();
            this.onCancel = new i0.EventEmitter();
            this.pendingStatus = false;
            this.pagination = { pageSize: 20, currentPage: 1 };
            this.selected = [];
            this.canCreateGroup = false;
            this.canAssignDevice = false;
            this.ITEMS_SELECT_LIMIT = 15;
            this.btnLabels = {
                next: i3.gettext('Next'),
                cancel: i3.gettext('Cancel'),
                create: i3.gettext('Create')
            };
        }
        AddGroupComponent.prototype.onEnterKeyDown = function (event) {
            // Order matters! Needs to be placed before this.stepper.next
            if ((this.isGroupDetailsStep() && !this.canAssignDevice) || this.isAssignDeviceStep()) {
                this.createGroup();
                return;
            }
            this.stepper.next();
        };
        AddGroupComponent.prototype.onEscapeKeyDown = function (event) {
            this.onCancel.emit();
        };
        AddGroupComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                var _this = this;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            this.formGroupStepOne = this.fb.group({
                                name: ['', forms.Validators.required],
                                description: ['']
                            });
                            this.subscription = this.onCancel.subscribe(function () { return _this.resetStepper(); });
                            this.canCreateGroup = this.subAssetsService.canCreateGroup();
                            _a = this;
                            return [4 /*yield*/, this.subAssetsService.canAssignDevice({
                                    id: this.currentGroupId
                                })];
                        case 1:
                            _a.canAssignDevice = _b.sent();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AddGroupComponent.prototype.ngAfterViewInit = function () {
            this.nameInput = this.nameInputRef.nativeElement;
            this.setFocusOnNameInput();
        };
        AddGroupComponent.prototype.createGroup = function () {
            return __awaiter(this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (this.canCreateGroup === false) {
                                return [2 /*return*/];
                            }
                            this.pendingStatus = true;
                            return [4 /*yield*/, this.addGroupService.createGroupAndAssignDevices(this.formGroupStepOne.value, this.currentGroupId, this.selected)];
                        case 1:
                            _a.sent();
                            this.pendingStatus = false;
                            this.resetStepper();
                            this.alert.success(i3.gettext('Group created.'));
                            this.refresh.emit();
                            this.onCancel.emit();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AddGroupComponent.prototype.onSelected = function (selectedDevicesIDs) {
            this.selected = selectedDevicesIDs;
        };
        AddGroupComponent.prototype.resetStepper = function () {
            this.stepper.reset();
            this.stepper.selectedIndex = 1;
            this.selected = [];
        };
        AddGroupComponent.prototype.ngOnDestroy = function () {
            if (this.subscription) {
                this.subscription.unsubscribe();
            }
        };
        AddGroupComponent.prototype.isGroupDetailsStep = function () {
            return this.stepper.selectedIndex === 0;
        };
        AddGroupComponent.prototype.isAssignDeviceStep = function () {
            return this.stepper.selectedIndex === 1;
        };
        AddGroupComponent.prototype.setFocusOnNameInput = function () {
            if (this.nameInput) {
                this.nameInput.focus();
                this.nameInput.select();
            }
        };
AddGroupComponent.ɵfac = function AddGroupComponent_Factory(t) { return new (t || AddGroupComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc10.DeviceGridService), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.FormBuilder), ɵngcc0.ɵɵdirectiveInject(AddGroupService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(SubAssetsService)); };
AddGroupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddGroupComponent, selectors: [["c8y-add-group"]], viewQuery: function AddGroupComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(i3.C8yStepper, 5);
        ɵngcc0.ɵɵviewQuery(_c2, 5);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.stepper = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.nameInputRef = _t.first);
    } }, hostBindings: function AddGroupComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.enter", function AddGroupComponent_keydown_enter_HostBindingHandler($event) { return ctx.onEnterKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument)("keydown.escape", function AddGroupComponent_keydown_escape_HostBindingHandler($event) { return ctx.onEscapeKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { refresh: "refresh", currentGroupId: "currentGroupId" }, outputs: { onDeviceQueryStringChange: "onDeviceQueryStringChange", onCancel: "onCancel" }, decls: 4, vars: 3, consts: [[4, "ngIf"], ["class", "d-contents", 4, "ngIf", "ngIfElse"], ["stepper", ""], [1, "d-contents"], [3, "ngTemplateOutlet"], ["linear", "", 1, "flex-col", "flex-nowrap", "no-align-items", "fit-h", "c8y-stepper--no-btns", 3, "disableDefaultIcons", "customClasses"], [3, "stepControl", "label"], [1, "p-16", "p-t-0", "flex-no-shrink", "separator-bottom", "col-xs-12"], [1, "row"], [1, "col-md-6", "col-md-offset-3", "col-lg-4", "col-lg-offset-4"], [1, "text-center", "text-medium"], [1, "col-xs-12", "flex-grow", "no-gutter"], [1, "card-inner-scroll", "fit-h"], [1, "card-block", "p-b-0"], [3, "formGroup"], ["translate", ""], ["type", "text", "formControlName", "name", "maxlength", "254", "required", "", 1, "form-control", 3, "placeholder"], ["nameRef", ""], ["translate", "", 4, "ngIf"], ["type", "text", "formControlName", "description", 1, "form-control", 3, "placeholder"], ["class", "alert alert-info max-width-100", "translate", "", 4, "ngIf"], [1, "d-block", "card-footer", "p-24", "separator", 3, "disabled", "labels", "showButtons", "onCancel", "onCustom"], [3, "label"], [1, "col-xs-12", "no-gutter", "flex-grow"], [1, "d-contents", 3, "title", "actionControls", "infiniteScroll", "selectable", "pagination", "refresh", "itemsSelect"], [1, "c8y-empty-state"], ["c8yIcon", "search"], [1, "d-block", "card-footer", "p-24", "separator", 3, "labels", "disabled", "pending", "onCancel", "onCustom"], ["translate", "", 1, "alert", "alert-info", "max-width-100"]], template: function AddGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AddGroupComponent_c8y_title_0_Template, 3, 3, "c8y-title", 0);
        ɵngcc0.ɵɵtemplate(1, AddGroupComponent_div_1_Template, 2, 1, "div", 1);
        ɵngcc0.ɵɵtemplate(2, AddGroupComponent_ng_template_2_Template, 55, 58, "ng-template", null, 2, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        var _r2 = ɵngcc0.ɵɵreference(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.currentGroupId);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.currentGroupId)("ngIfElse", _r2);
    } }, directives: [ɵngcc3.NgIf, ɵngcc1.TitleComponent, ɵngcc3.NgTemplateOutlet, ɵngcc1.C8yStepper, ɵngcc11.CdkStep, ɵngcc1.FormGroupComponent, ɵngcc6.NgControlStatusGroup, ɵngcc6.FormGroupDirective, ɵngcc1.C8yTranslateDirective, ɵngcc6.DefaultValueAccessor, ɵngcc1.RequiredInputPlaceholderDirective, ɵngcc6.NgControlStatus, ɵngcc6.FormControlName, ɵngcc6.MaxLengthValidator, ɵngcc6.RequiredValidator, ɵngcc1.MessagesComponent, ɵngcc1.C8yStepperButtons, ɵngcc10.DeviceGridComponent, ɵngcc1.IconDirective, ɵngcc1.MessageDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddGroupComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-add-group',
                template: "<c8y-title *ngIf=\"!currentGroupId\">\n  {{ 'Add group' | translate }}\n</c8y-title>\n\n<div class=\"d-contents\" *ngIf=\"!currentGroupId; else stepper\">\n  <ng-container [ngTemplateOutlet]=\"stepper\"></ng-container>\n</div>\n\n<ng-template #stepper>\n  <c8y-stepper\n    class=\"flex-col flex-nowrap no-align-items fit-h c8y-stepper--no-btns\"\n    [disableDefaultIcons]=\"{ edit: true, done: false }\"\n    [customClasses]=\"['col-md-6', 'col-md-offset-3', 'm-t-24', 'm-b-40', 'p-0', 'flex-no-shrink']\"\n    linear\n  >\n    <cdk-step [stepControl]=\"formGroupStepOne\" [label]=\"'New group' | translate\">\n      <div class=\"p-16 p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center text-medium\">\n              {{ 'New group' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-xs-12 flex-grow no-gutter\">\n        <div class=\"card-inner-scroll fit-h\">\n          <div class=\"card-block p-b-0\">\n            <div class=\"row\">\n              <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n                <c8y-form-group>\n                  <div [formGroup]=\"formGroupStepOne\">\n                    <c8y-form-group>\n                      <label translate>Name</label>\n                      <input\n                        class=\"form-control\"\n                        type=\"text\"\n                        formControlName=\"name\"\n                        placeholder=\"{{ 'e.g. First floor' | translate }} \"\n                        maxlength=\"254\"\n                        #nameRef\n                        required\n                      />\n                      <c8y-messages>\n                        <c8y-message *ngIf=\"!formGroupStepOne.untouched && !nameRef.value\" translate\n                          >This field is required.</c8y-message\n                        >\n                      </c8y-messages>\n                    </c8y-form-group>\n\n                    <c8y-form-group>\n                      <label translate>Description</label>\n                      <input\n                        class=\"form-control\"\n                        type=\"text\"\n                        formControlName=\"description\"\n                        placeholder=\"{{ 'e.g. first floor devices' | translate }}\"\n                      />\n                    </c8y-form-group>\n                  </div>\n                </c8y-form-group>\n                <c8y-form-group>\n                  <div [formGroup]=\"formGroupStepOne\"></div>\n                </c8y-form-group>\n                <div class=\"alert alert-info max-width-100\" translate *ngIf=\"!canAssignDevice\">\n                  You don't have permission to assign devices.\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        (onCancel)=\"onCancel.emit()\"\n        (onCustom)=\"createGroup()\"\n        [disabled]=\"!canCreateGroup\"\n        [labels]=\"\n          canAssignDevice\n            ? { next: btnLabels.next, cancel: btnLabels.cancel }\n            : { custom: btnLabels.create, cancel: btnLabels.cancel }\n        \"\n        [showButtons]=\"\n          canAssignDevice ? { next: true, cancel: true } : { custom: true, cancel: true }\n        \"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n    <cdk-step [label]=\"'Assign devices' | translate\">\n      <div class=\"p-16 p-t-0 flex-no-shrink separator-bottom col-xs-12\">\n        <div class=\"row\">\n          <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n            <h4 class=\"text-center text-medium\">\n              {{ 'Assign devices' | translate }}\n            </h4>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-xs-12 no-gutter flex-grow\">\n        <c8y-device-grid\n          [title]=\"'Select target devices' | translate\"\n          [actionControls]=\"[]\"\n          [infiniteScroll]=\"'auto'\"\n          [selectable]=\"'true'\"\n          [pagination]=\"pagination\"\n          (itemsSelect)=\"onSelected($event)\"\n          [refresh]=\"refresh\"\n          class=\"d-contents\"\n        >\n          <div class=\"c8y-empty-state\">\n            <h1 c8yIcon=\"search\"></h1>\n            <div>\n              <p>\n                <strong>{{ 'No matching devices.' | translate }}</strong>\n              </p>\n              <small>{{ 'Refine your search terms' | translate }}</small>\n            </div>\n          </div>\n        </c8y-device-grid>\n      </div>\n      <c8y-stepper-buttons\n        class=\"d-block card-footer p-24 separator\"\n        (onCancel)=\"onCancel.emit()\"\n        (onCustom)=\"createGroup()\"\n        [labels]=\"{ custom: btnLabels.create }\"\n        [disabled]=\"!canAssignDevice\"\n        [pending]=\"pendingStatus\"\n      ></c8y-stepper-buttons>\n    </cdk-step>\n  </c8y-stepper>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc10.DeviceGridService }, { type: ɵngcc6.FormBuilder }, { type: AddGroupService }, { type: ɵngcc1.AlertService }, { type: SubAssetsService }]; }, { refresh: [{
            type: i0.Input
        }], onDeviceQueryStringChange: [{
            type: i0.Output
        }], onCancel: [{
            type: i0.Output
        }], onEnterKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.enter', ['$event']]
        }], onEscapeKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.escape', ['$event']]
        }], currentGroupId: [{
            type: i0.Input
        }], stepper: [{
            type: i0.ViewChild,
            args: [i3.C8yStepper, { static: false }]
        }], nameInputRef: [{
            type: i0.ViewChild,
            args: ['nameRef', { static: false }]
        }] }); })();
        return AddGroupComponent;
    }());
    AddGroupComponent.ctorParameters = function () { return [
        { type: deviceGrid.DeviceGridService },
        { type: forms.FormBuilder },
        { type: AddGroupService },
        { type: i3.AlertService },
        { type: SubAssetsService }
    ]; };
    AddGroupComponent.propDecorators = {
        currentGroupId: [{ type: i0.Input }],
        refresh: [{ type: i0.Input }],
        onDeviceQueryStringChange: [{ type: i0.Output }],
        onCancel: [{ type: i0.Output }],
        stepper: [{ type: i0.ViewChild, args: [i3.C8yStepper, { static: false },] }],
        nameInputRef: [{ type: i0.ViewChild, args: ['nameRef', { static: false },] }],
        onEnterKeyDown: [{ type: i0.HostListener, args: ['document:keydown.enter', ['$event'],] }],
        onEscapeKeyDown: [{ type: i0.HostListener, args: ['document:keydown.escape', ['$event'],] }]
    };

    var ɵ0$1 = [
        {
            path: 'add-group',
            component: AddGroupComponent
        }
    ];
    var AddGroupModule = /** @class */ (function () {
        function AddGroupModule() {
        }
AddGroupModule.ɵfac = function AddGroupModule_Factory(t) { return new (t || AddGroupModule)(); };
AddGroupModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AddGroupModule });
AddGroupModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: i3.HOOK_ROUTE,
            useValue: ɵ0$1,
            multi: true
        },
        AddGroupService
    ], imports: [[i3.CoreModule, deviceGrid.DeviceGridModule, forms.FormsModule, forms.ReactiveFormsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddGroupModule, [{
        type: i0.NgModule,
        args: [{
                declarations: [AddGroupComponent],
                imports: [i3.CoreModule, deviceGrid.DeviceGridModule, forms.FormsModule, forms.ReactiveFormsModule],
                entryComponents: [AddGroupComponent],
                exports: [AddGroupComponent],
                providers: [
                    {
                        provide: i3.HOOK_ROUTE,
                        useValue: ɵ0$1,
                        multi: true
                    },
                    AddGroupService
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AddGroupModule, { declarations: [AddGroupComponent], imports: [ɵngcc1.CoreModule, ɵngcc10.DeviceGridModule, ɵngcc6.FormsModule, ɵngcc6.ReactiveFormsModule], exports: [AddGroupComponent] }); })();
        return AddGroupModule;
    }());

    var GroupsComponent = /** @class */ (function () {
        function GroupsComponent(permissions, subAssetsService) {
            this.permissions = permissions;
            this.subAssetsService = subAssetsService;
            this.refresh = new i0.EventEmitter();
            this.showAddGroup = false;
            this.filterable = false;
            this.sortable = true;
            this.shouldDisableAddGroup = false;
        }
        GroupsComponent.prototype.ngOnInit = function () {
            if (!this.permissions.hasRole('ROLE_INVENTORY_READ')) {
                this.sortable = false;
            }
            this.shouldDisableAddGroup = !this.subAssetsService.canCreateGroup();
        };
GroupsComponent.ɵfac = function GroupsComponent_Factory(t) { return new (t || GroupsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Permissions), ɵngcc0.ɵɵdirectiveInject(SubAssetsService)); };
GroupsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: GroupsComponent, selectors: [["c8y-groups-name"]], decls: 13, vars: 15, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle", 1, "m-r-4"], ["translate", ""], [3, "ngClass"], [1, "bottom-drawer"], [3, "refresh", "onCancel", 4, "ngIf"], [1, "content-fullpage", "d-flex", "d-col", 3, "refresh", "filterable", "sortable"], [3, "refresh", "onCancel"]], template: function GroupsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "button", 1);
        ɵngcc0.ɵɵlistener("click", function GroupsComponent_Template_button_click_4_listener() { return ctx.showAddGroup = !ctx.showAddGroup; });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 2);
        ɵngcc0.ɵɵelementStart(7, "span", 3);
        ɵngcc0.ɵɵtext(8, "Add group");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 4);
        ɵngcc0.ɵɵelementStart(10, "div", 5);
        ɵngcc0.ɵɵtemplate(11, GroupsComponent_c8y_add_group_11_Template, 1, 1, "c8y-add-group", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(12, "c8y-sub-assets-grid", 7);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 9, "Groups"), "\n");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 11, "Add group"));
        ɵngcc0.ɵɵproperty("disabled", ctx.shouldDisableAddGroup);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(13, _c0, ctx.showAddGroup));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("refresh", ctx.refresh)("filterable", ctx.filterable)("sortable", ctx.sortable);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc1.ActionBarItemComponent, ɵngcc1.IconDirective, ɵngcc1.C8yTranslateDirective, ɵngcc3.NgClass, ɵngcc3.NgIf, SubAssetsGridComponent, AddGroupComponent], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GroupsComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-groups-name',
                template: "<c8y-title>\n  {{ 'Groups' | translate }}\n</c8y-title>\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n    [disabled]=\"shouldDisableAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n <!-- TODO: uncomment when help is available \n  <c8y-help src=\"/users-guide/device-management/#grouping-devices\"></c8y-help> \n-->\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n<c8y-sub-assets-grid\n  [refresh]=\"refresh\"\n  [filterable]=\"filterable\"\n  [sortable]=\"sortable\"\n  class=\"content-fullpage d-flex d-col\"\n></c8y-sub-assets-grid>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: SubAssetsService }]; }, null); })();
        return GroupsComponent;
    }());
    GroupsComponent.ctorParameters = function () { return [
        { type: i3.Permissions },
        { type: SubAssetsService }
    ]; };

    var AssignDevicesComponent = /** @class */ (function () {
        function AssignDevicesComponent(alert, subAssetsService, inventoryService) {
            this.alert = alert;
            this.subAssetsService = subAssetsService;
            this.inventoryService = inventoryService;
            this.refresh = new i0.EventEmitter();
            this.onCancel = new i0.EventEmitter();
            this.onShowChildDevices = new i0.EventEmitter();
            this.selectedDevice = new i0.EventEmitter();
            this.pendingStatus = false;
            this.pagination = { pageSize: 20, currentPage: 1 };
            this.selected = [];
            this.canAssignDevice = false;
            this.actionControls = [];
            this.headerActionControls = [];
            this.showChildren = false;
            this.isSelectable = true;
        }
        AssignDevicesComponent.prototype.onEnterKeyDown = function (event) {
            if (this.selected.length > 0) {
                this.assignDevices();
            }
        };
        AssignDevicesComponent.prototype.onEscapeKeyDown = function (event) {
            this.onCancel.emit();
        };
        AssignDevicesComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            this.setNotIncludedInGroupQuery();
                            _a = this;
                            return [4 /*yield*/, this.subAssetsService.canAssignDevice({
                                    id: this.currentGroupId
                                })];
                        case 1:
                            _a.canAssignDevice = _b.sent();
                            this.setHeaderActionControls();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AssignDevicesComponent.prototype.setNotIncludedInGroupQuery = function () {
            var notIncludedInGroupQuery = { __not: { __bygroupid: this.currentGroupId } };
            this.baseQuery = notIncludedInGroupQuery;
        };
        AssignDevicesComponent.prototype.setHeaderActionControls = function () {
            var _this = this;
            var headerActionControls = [];
            var showChildDevices = {
                type: 'DISPLAY_CHILD_DEVICES_BUTTON',
                text: i3.gettext('Display child devices button'),
                template: this.showDevicesToggle,
                callback: function () {
                    _this.showChildren = !_this.showChildren;
                    _this.setActionControls(_this.showChildren);
                }
            };
            headerActionControls.push(showChildDevices);
            this.headerActionControls = headerActionControls;
        };
        AssignDevicesComponent.prototype.setActionControls = function (showChildren) {
            var _this = this;
            var actionControls = [];
            var selectChildrenAction = {
                type: 'SHOW_TARGET_CHILD_DEVICES',
                icon: 'enter-bottom',
                text: i3.gettext('Show target child devices'),
                callback: function (asset) { return _this.selectChildren(asset); },
                showIf: function (asset) { return asset.childDevices.references.length > 0; }
            };
            if (showChildren) {
                actionControls.push(selectChildrenAction);
            }
            this.actionControls = actionControls;
            this.refresh.emit();
        };
        AssignDevicesComponent.prototype.assignDevices = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (this.canAssignDevice === false) {
                                return [2 /*return*/];
                            }
                            this.pendingStatus = true;
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.inventoryService.childAssetsBulkAdd(this.selected, this.currentGroupId)];
                        case 2:
                            _a.sent();
                            this.refresh.emit();
                            this.alert.success(i3.gettext('Devices assigned to the group.'));
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _a.sent();
                            this.alert.danger(i3.gettext('Could not assign devices to the group'), error_1);
                            return [3 /*break*/, 4];
                        case 4:
                            this.pendingStatus = false;
                            this.selected = [];
                            this.onCancel.emit();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AssignDevicesComponent.prototype.onSelected = function (selectedDevicesIDs) {
            this.selected = selectedDevicesIDs;
        };
        AssignDevicesComponent.prototype.selectChildren = function (asset) {
            this.onShowChildDevices.emit(true);
            this.selectedDevice.emit(asset);
        };
AssignDevicesComponent.ɵfac = function AssignDevicesComponent_Factory(t) { return new (t || AssignDevicesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(SubAssetsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.InventoryService)); };
AssignDevicesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssignDevicesComponent, selectors: [["c8y-assign-devices"]], viewQuery: function AssignDevicesComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c11, 5, i0.TemplateRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.showDevicesToggle = _t.first);
    } }, hostBindings: function AssignDevicesComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.enter", function AssignDevicesComponent_keydown_enter_HostBindingHandler($event) { return ctx.onEnterKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument)("keydown.escape", function AssignDevicesComponent_keydown_escape_HostBindingHandler($event) { return ctx.onEscapeKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { refresh: "refresh", currentGroupId: "currentGroupId" }, outputs: { onCancel: "onCancel", onShowChildDevices: "onShowChildDevices", selectedDevice: "selectedDevice" }, decls: 31, vars: 36, consts: [[1, "card-block", "flex-no-shrink", "separator-bottom", "col-xs-12", "large-padding", "p-t-24", "p-b-24"], [1, "row"], [1, "col-md-6", "col-md-offset-3", "col-lg-4", "col-lg-offset-4"], [1, "text-center", "text-medium"], [1, "flex-grow", "col-xs-12", "no-gutter", 3, "title", "actionControls", "infiniteScroll", "selectable", "pagination", "refresh", "baseQuery", "headerActionControls", "withChildren", "itemsSelect"], [1, "c8y-empty-state"], ["c8yIcon", "search"], [1, "text-center", "card-footer", "p-24", "separator"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "ngClass", "title", "disabled", "click"], ["showDevicesToggle", ""], [1, "c8y-switch"], ["type", "checkbox", 3, "ngModel", "ngModelChange", "click"], ["placement", "bottom", 1, "btn-clean", "m-r-8", 3, "popover", "outsideClick"], ["c8yIcon", "question-circle-o", 1, "text-primary"], ["childDevicesPop", ""], ["translate", ""], [1, "btn", "btn-xs", "btn-icon", "btn-default", "no-pointer"], [1, "text-primary", "dlt-c8y-icon-enter-bottom"]], template: function AssignDevicesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-device-grid", 4);
        ɵngcc0.ɵɵlistener("itemsSelect", function AssignDevicesComponent_Template_c8y_device_grid_itemsSelect_6_listener($event) { return ctx.onSelected($event); });
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementStart(8, "div", 5);
        ɵngcc0.ɵɵelement(9, "h1", 6);
        ɵngcc0.ɵɵelementStart(10, "div");
        ɵngcc0.ɵɵelementStart(11, "p");
        ɵngcc0.ɵɵelementStart(12, "strong");
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "small");
        ɵngcc0.ɵɵtext(16);
        ɵngcc0.ɵɵpipe(17, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 7);
        ɵngcc0.ɵɵelementStart(19, "button", 8);
        ɵngcc0.ɵɵlistener("click", function AssignDevicesComponent_Template_button_click_19_listener() { return ctx.onCancel.emit(); });
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementStart(21, "span");
        ɵngcc0.ɵɵtext(22);
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(24, "button", 9);
        ɵngcc0.ɵɵlistener("click", function AssignDevicesComponent_Template_button_click_24_listener() { return ctx.assignDevices(); });
        ɵngcc0.ɵɵpipe(25, "translate");
        ɵngcc0.ɵɵelementStart(26, "span");
        ɵngcc0.ɵɵtext(27);
        ɵngcc0.ɵɵpipe(28, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(29, AssignDevicesComponent_ng_template_29_Template, 10, 6, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 18, "Assign devices"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(7, 20, "Select devices"))("actionControls", ctx.actionControls)("infiniteScroll", "auto")("selectable", ctx.isSelectable)("pagination", ctx.pagination)("refresh", ctx.refresh)("baseQuery", ctx.baseQuery)("headerActionControls", ctx.headerActionControls)("withChildren", true);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(14, 22, "No matching devices."));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(17, 24, "Refine your search terms"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 26, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(23, 28, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(25, 30, "Assign"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(34, _c12, ctx.pendingStatus))("disabled", ctx.selected.length === 0 || !ctx.canAssignDevice);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(28, 32, "Assign"));
    } }, directives: [ɵngcc10.DeviceGridComponent, ɵngcc1.IconDirective, ɵngcc3.NgClass, ɵngcc6.CheckboxControlValueAccessor, ɵngcc6.NgControlStatus, ɵngcc6.NgModel, ɵngcc7.PopoverDirective, ɵngcc1.C8yTranslateDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssignDevicesComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-assign-devices',
                template: "<div class=\"card-block flex-no-shrink separator-bottom col-xs-12 large-padding p-t-24 p-b-24\">\n  <div class=\"row\">\n    <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n      <h4 class=\"text-center text-medium\">\n        {{ 'Assign devices' | translate }}\n      </h4>\n    </div>\n  </div>\n</div>\n<c8y-device-grid\n  [title]=\"'Select devices' | translate\"\n  [actionControls]=\"actionControls\"\n  [infiniteScroll]=\"'auto'\"\n  [selectable]=\"isSelectable\"\n  [pagination]=\"pagination\"\n  (itemsSelect)=\"onSelected($event)\"\n  [refresh]=\"refresh\"\n  [baseQuery]=\"baseQuery\"\n  [headerActionControls]=\"headerActionControls\"\n  [withChildren]=\"true\"\n  class=\"flex-grow col-xs-12 no-gutter\"\n>\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div>\n      <p>\n        <strong>{{ 'No matching devices.' | translate }}</strong>\n      </p>\n      <small>{{ 'Refine your search terms' | translate }}</small>\n    </div>\n  </div>\n</c8y-device-grid>\n\n<div class=\"text-center card-footer p-24 separator\">\n  <button\n    (click)=\"onCancel.emit()\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Cancel' | translate }}\"\n  >\n    <span>{{ 'Cancel' | translate }}</span>\n  </button>\n  <button\n    (click)=\"assignDevices()\"\n    type=\"button\"\n    class=\"btn btn-primary\"\n    [ngClass]=\"{ 'btn-pending': pendingStatus }\"\n    title=\"{{ 'Assign' | translate }}\"\n    [disabled]=\"selected.length === 0 || !canAssignDevice\"\n  >\n    <span>{{ 'Assign' | translate }}</span>\n  </button>\n</div>\n\n<ng-template #showDevicesToggle let-control=\"headerActionControl\">\n  <label class=\"c8y-switch\">\n    <input type=\"checkbox\" [(ngModel)]=\"showChildren\" (click)=\"control.callback()\" />\n    <span></span>\n    <span>{{ control.text | translate }}</span>\n  </label>\n  <button\n    class=\"btn-clean m-r-8\"\n    [popover]=\"childDevicesPop\"\n    placement=\"bottom\"\n    [outsideClick]=\"true\"\n  >\n    <i c8yIcon=\"question-circle-o\" class=\"text-primary\"></i>\n  </button>\n  <ng-template #childDevicesPop>\n    <span translate>\n      Displays the button\n      <button class=\"btn btn-xs btn-icon btn-default no-pointer\">\n        <i class=\"text-primary dlt-c8y-icon-enter-bottom\"></i>\n      </button>\n      next to target devices with children. Clicking it displays a list with all child devices of\n      the selected target device.\n    </span>\n  </ng-template>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: SubAssetsService }, { type: ɵngcc5.InventoryService }]; }, { refresh: [{
            type: i0.Input
        }], onCancel: [{
            type: i0.Output
        }], onShowChildDevices: [{
            type: i0.Output
        }], selectedDevice: [{
            type: i0.Output
        }], onEnterKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.enter', ['$event']]
        }], onEscapeKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.escape', ['$event']]
        }], currentGroupId: [{
            type: i0.Input
        }], showDevicesToggle: [{
            type: i0.ViewChild,
            args: ['showDevicesToggle', { read: i0.TemplateRef }]
        }] }); })();
        return AssignDevicesComponent;
    }());
    AssignDevicesComponent.ctorParameters = function () { return [
        { type: i3.AlertService },
        { type: SubAssetsService },
        { type: client.InventoryService }
    ]; };
    AssignDevicesComponent.propDecorators = {
        currentGroupId: [{ type: i0.Input }],
        refresh: [{ type: i0.Input }],
        onCancel: [{ type: i0.Output }],
        onShowChildDevices: [{ type: i0.Output }],
        selectedDevice: [{ type: i0.Output }],
        showDevicesToggle: [{ type: i0.ViewChild, args: ['showDevicesToggle', { read: i0.TemplateRef },] }],
        onEnterKeyDown: [{ type: i0.HostListener, args: ['document:keydown.enter', ['$event'],] }],
        onEscapeKeyDown: [{ type: i0.HostListener, args: ['document:keydown.escape', ['$event'],] }]
    };

    var AssignChildDevicesComponent = /** @class */ (function () {
        function AssignChildDevicesComponent(alert, subAssetsService, inventoryService) {
            this.alert = alert;
            this.subAssetsService = subAssetsService;
            this.inventoryService = inventoryService;
            this.onCancel = new i0.EventEmitter();
            this.refresh = new i0.EventEmitter();
            this.selected = [];
            this.canAssignDevice = false;
            this.pendingStatus = false;
        }
        AssignChildDevicesComponent.prototype.onEnterKeyDown = function (event) {
            if (this.selected.length > 0) {
                this.assignDevices();
            }
        };
        AssignChildDevicesComponent.prototype.onEscapeKeyDown = function (event) {
            this.onCancel.emit();
        };
        AssignChildDevicesComponent.prototype.ngOnInit = function () {
            return __awaiter(this, void 0, void 0, function () {
                var _a;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            this.setNotIncludedInGroupQuery();
                            _a = this;
                            return [4 /*yield*/, this.subAssetsService.canAssignDevice({
                                    id: this.currentGroupId
                                })];
                        case 1:
                            _a.canAssignDevice = _b.sent();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AssignChildDevicesComponent.prototype.setNotIncludedInGroupQuery = function () {
            var notIncludedInGroupQuery = { __not: { __bygroupid: this.currentGroupId } };
            this.baseQuery = notIncludedInGroupQuery;
        };
        AssignChildDevicesComponent.prototype.assignDevices = function () {
            return __awaiter(this, void 0, void 0, function () {
                var error_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            if (this.canAssignDevice === false) {
                                return [2 /*return*/];
                            }
                            this.pendingStatus = true;
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 3, , 4]);
                            return [4 /*yield*/, this.inventoryService.childAssetsBulkAdd(this.selected, this.currentGroupId)];
                        case 2:
                            _a.sent();
                            this.refresh.emit();
                            this.alert.success(i3.gettext('Child devices assigned to the group.'));
                            return [3 /*break*/, 4];
                        case 3:
                            error_1 = _a.sent();
                            this.alert.danger(i3.gettext('Could not assign child devices to the group'), error_1);
                            return [3 /*break*/, 4];
                        case 4:
                            this.pendingStatus = false;
                            this.selected = [];
                            this.onCancel.emit();
                            return [2 /*return*/];
                    }
                });
            });
        };
        AssignChildDevicesComponent.prototype.onSelected = function (selectedDevicesIDs) {
            this.selected = selectedDevicesIDs;
        };
AssignChildDevicesComponent.ɵfac = function AssignChildDevicesComponent_Factory(t) { return new (t || AssignChildDevicesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(SubAssetsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.InventoryService)); };
AssignChildDevicesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssignChildDevicesComponent, selectors: [["c8y-assign-child-devices"]], hostBindings: function AssignChildDevicesComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown.enter", function AssignChildDevicesComponent_keydown_enter_HostBindingHandler($event) { return ctx.onEnterKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument)("keydown.escape", function AssignChildDevicesComponent_keydown_escape_HostBindingHandler($event) { return ctx.onEscapeKeyDown($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, inputs: { refresh: "refresh", currentGroupId: "currentGroupId", parentDevice: "parentDevice" }, outputs: { onCancel: "onCancel" }, decls: 21, vars: 31, consts: [[1, "card-block", "flex-no-shrink", "separator-bottom", "col-xs-12", "large-padding", "p-t-24", "p-b-24"], [1, "row"], [1, "col-md-6", "col-md-offset-3", "col-lg-4", "col-lg-offset-4"], [1, "text-center", "text-medium"], [1, "d-flex", "d-col", "no-align-items", "fit-h"], [1, "d-contents", 3, "title", "emptyStateText", "refresh", "actionControls", "selectable", "parent-group", "baseQuery", "itemsSelect"], [1, "text-center", "card-footer", "p-24", "separator"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "ngClass", "title", "disabled", "click"]], template: function AssignChildDevicesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵelementStart(7, "c8y-sub-assets-grid", 5);
        ɵngcc0.ɵɵlistener("itemsSelect", function AssignChildDevicesComponent_Template_c8y_sub_assets_grid_itemsSelect_7_listener($event) { return ctx.onSelected($event); });
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "button", 7);
        ɵngcc0.ɵɵlistener("click", function AssignChildDevicesComponent_Template_button_click_11_listener() { return ctx.onCancel.emit(); });
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementStart(13, "span");
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "button", 8);
        ɵngcc0.ɵɵlistener("click", function AssignChildDevicesComponent_Template_button_click_16_listener() { return ctx.assignDevices(); });
        ɵngcc0.ɵɵpipe(17, "translate");
        ɵngcc0.ɵɵelementStart(18, "span");
        ɵngcc0.ɵɵtext(19);
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 14, "Assign child devices"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(8, 16, "Select child devices"))("emptyStateText", ɵngcc0.ɵɵpipeBind1(9, 18, "All child devices are already assigned to the group"))("refresh", ctx.refresh)("actionControls", ɵngcc0.ɵɵpureFunction0(28, _c9))("selectable", true)("parent-group", ctx.parentDevice)("baseQuery", ctx.baseQuery);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 20, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(15, 22, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(17, 24, "Assign"));
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(29, _c12, ctx.pendingStatus))("disabled", ctx.selected.length === 0 || !ctx.canAssignDevice);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(20, 26, "Assign"));
    } }, directives: [SubAssetsGridComponent, ɵngcc3.NgClass], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssignChildDevicesComponent, [{
        type: i0.Component,
        args: [{
                selector: 'c8y-assign-child-devices',
                template: "<div class=\"card-block flex-no-shrink separator-bottom col-xs-12 large-padding p-t-24 p-b-24\">\n  <div class=\"row\">\n    <div class=\"col-md-6 col-md-offset-3 col-lg-4 col-lg-offset-4\">\n      <h4 class=\"text-center text-medium\">\n        {{ 'Assign child devices' | translate }}\n      </h4>\n    </div>\n  </div>\n</div>\n\n<div class=\"d-flex d-col no-align-items fit-h\">\n  <c8y-sub-assets-grid\n    [title]=\"'Select child devices' | translate\"\n    [emptyStateText]=\"'All child devices are already assigned to the group' | translate\"\n    [refresh]=\"refresh\"\n    [actionControls]=\"[]\"\n    [selectable]=\"true\"\n    [parent-group]=\"parentDevice\"\n    [baseQuery]=\"baseQuery\"\n    (itemsSelect)=\"onSelected($event)\"\n    class=\"d-contents\"\n  >\n  </c8y-sub-assets-grid>\n</div>\n\n<div class=\"text-center card-footer p-24 separator\">\n  <button\n    (click)=\"onCancel.emit()\"\n    type=\"button\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Cancel' | translate }}\"\n  >\n    <span>{{ 'Cancel' | translate }}</span>\n  </button>\n  <button\n    (click)=\"assignDevices()\"\n    type=\"button\"\n    class=\"btn btn-primary\"\n    [ngClass]=\"{ 'btn-pending': pendingStatus }\"\n    title=\"{{ 'Assign' | translate }}\"\n    [disabled]=\"selected.length === 0 || !canAssignDevice\"\n  >\n    <span>{{ 'Assign' | translate }}</span>\n  </button>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: SubAssetsService }, { type: ɵngcc5.InventoryService }]; }, { onCancel: [{
            type: i0.Output
        }], refresh: [{
            type: i0.Input
        }], onEnterKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.enter', ['$event']]
        }], onEscapeKeyDown: [{
            type: i0.HostListener,
            args: ['document:keydown.escape', ['$event']]
        }], currentGroupId: [{
            type: i0.Input
        }], parentDevice: [{
            type: i0.Input
        }] }); })();
        return AssignChildDevicesComponent;
    }());
    AssignChildDevicesComponent.ctorParameters = function () { return [
        { type: i3.AlertService },
        { type: SubAssetsService },
        { type: client.InventoryService }
    ]; };
    AssignChildDevicesComponent.propDecorators = {
        currentGroupId: [{ type: i0.Input }],
        parentDevice: [{ type: i0.Input }],
        onCancel: [{ type: i0.Output }],
        refresh: [{ type: i0.Input }],
        onEnterKeyDown: [{ type: i0.HostListener, args: ['document:keydown.enter', ['$event'],] }],
        onEscapeKeyDown: [{ type: i0.HostListener, args: ['document:keydown.escape', ['$event'],] }]
    };

    var ɵ0 = [
        {
            context: i3.ViewContext.Group,
            path: 'subassets',
            priority: 1000,
            icon: 'c8y-group-open',
            label: i3.gettext('Subassets'),
            component: SubAssetsComponent
        },
        {
            path: 'group',
            component: GroupsComponent
        }
    ];
    var SubAssetsModule = /** @class */ (function () {
        function SubAssetsModule() {
        }
SubAssetsModule.ɵfac = function SubAssetsModule_Factory(t) { return new (t || SubAssetsModule)(); };
SubAssetsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SubAssetsModule });
SubAssetsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: i3.HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        SubAssetsService
    ], imports: [[i3.CoreModule, deviceGrid.DeviceGridModule, AddGroupModule, popover.PopoverModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubAssetsModule, [{
        type: i0.NgModule,
        args: [{
                declarations: [
                    SubAssetsComponent,
                    SubAssetsGridComponent,
                    AssetTypeCellRendererComponent,
                    GroupsComponent,
                    GroupInfoComponent,
                    DeleteAssetsModalComponent,
                    UnassignModalComponent,
                    AssignDevicesComponent,
                    AssignChildDevicesComponent
                ],
                imports: [i3.CoreModule, deviceGrid.DeviceGridModule, AddGroupModule, popover.PopoverModule],
                entryComponents: [
                    SubAssetsComponent,
                    AssetTypeCellRendererComponent,
                    GroupsComponent,
                    DeleteAssetsModalComponent,
                    UnassignModalComponent
                ],
                exports: [],
                providers: [
                    {
                        provide: i3.HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    SubAssetsService
                ]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SubAssetsModule, { declarations: [SubAssetsComponent, SubAssetsGridComponent, AssetTypeCellRendererComponent, GroupsComponent, GroupInfoComponent, DeleteAssetsModalComponent, UnassignModalComponent, AssignDevicesComponent, AssignChildDevicesComponent], imports: [ɵngcc1.CoreModule, ɵngcc10.DeviceGridModule, AddGroupModule, ɵngcc7.PopoverModule] }); })();
        return SubAssetsModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.AddGroupComponent = AddGroupComponent;
    exports.AddGroupModule = AddGroupModule;
    exports.AddGroupService = AddGroupService;
    exports.AssetTypeCellRendererComponent = AssetTypeCellRendererComponent;
    exports.AssetTypeGridColumn = AssetTypeGridColumn;
    exports.AssignChildDevicesComponent = AssignChildDevicesComponent;
    exports.AssignDevicesComponent = AssignDevicesComponent;
    exports.DeleteAssetsModalComponent = DeleteAssetsModalComponent;
    exports.GroupInfoComponent = GroupInfoComponent;
    exports.GroupsComponent = GroupsComponent;
    exports.SubAssetsComponent = SubAssetsComponent;
    exports.SubAssetsGridComponent = SubAssetsGridComponent;
    exports.SubAssetsModule = SubAssetsModule;
    exports.SubAssetsService = SubAssetsService;
    exports.UnassignModalComponent = UnassignModalComponent;
    exports["ɵa"] = AssetTypeCellRendererComponent;
    exports["ɵb"] = DeleteAssetsModalComponent;
    exports["ɵc"] = UnassignModalComponent;
    exports["ɵd"] = AddGroupModule;
    exports["ɵe"] = AddGroupComponent;
    exports["ɵf"] = AddGroupService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
