// @c8y/ngx-components/bundles/c8y-ngx-components-operations-bulk-operations-stepper-container.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/router'),require('@c8y/ngx-components/operations/bulk-operations-service'),require('@c8y/ngx-components'),exports, require('@angular/core'), require('@angular/router'), require('@c8y/ngx-components'), require('@c8y/ngx-components/operations/bulk-operations-service')) :
    typeof define === 'function' && define.amd ? define('@c8y/ngx-components/operations/bulk-operations-stepper-container', ['@angular/core','@angular/router','@c8y/ngx-components/operations/bulk-operations-service','@c8y/ngx-components','exports', '@angular/core', '@angular/router', '@c8y/ngx-components', '@c8y/ngx-components/operations/bulk-operations-service'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.router,global.c8y.ngxComponents.operations.bulkOperationsService,global.c8y.ngxComponents,(global.c8y = global.c8y || {}, global.c8y["ngx-components"] = global.c8y["ngx-components"] || {}, global.c8y["ngx-components"].operations = global.c8y["ngx-components"].operations || {}, global.c8y["ngx-components"].operations["bulk-operations-stepper-container"] = {}), global.ng.core, global.ng.router, global.c8y["ngx-components"], global.c8y["ngx-components"].operations["bulk-operations-service"]));
})(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, router, ngxComponents, bulkOperationsService) { 
var _c0 = ["container"];
'use strict';

    var BulkOperationsStepperContainerComponent = /** @class */ (function () {
        function BulkOperationsStepperContainerComponent(activatedRoute, bulkOpsService, resolver, changeDetector) {
            this.activatedRoute = activatedRoute;
            this.bulkOpsService = bulkOpsService;
            this.resolver = resolver;
            this.changeDetector = changeDetector;
        }
        BulkOperationsStepperContainerComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.activatedRoute.params.subscribe(function (params) {
                if (params && params.type) {
                    _this.bulkOperationType = params.type;
                    _this.component = _this.bulkOpsService
                        .getBulkTypes()
                        .find(function (element) { return element.path.includes(_this.bulkOperationType); }).component;
                    _this.changeDetector.detectChanges();
                    _this.createComponent(_this.component);
                }
            });
        };
        BulkOperationsStepperContainerComponent.prototype.createComponent = function (component) {
            if (component) {
                var factory = this.resolver.resolveComponentFactory(component);
                this.container.createComponent(factory);
            }
        };
BulkOperationsStepperContainerComponent.ɵfac = function BulkOperationsStepperContainerComponent_Factory(t) { return new (t || BulkOperationsStepperContainerComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.BulkOperationsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
BulkOperationsStepperContainerComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationsStepperContainerComponent, selectors: [["bulk-operations-stepper-container"]], viewQuery: function BulkOperationsStepperContainerComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5, core.ViewContainerRef);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
    } }, decls: 7, vars: 3, consts: [[1, "card", "card--grid--fullpage"], [1, "d-contents"], ["container", ""]], template: function BulkOperationsStepperContainerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 0);
        ɵngcc0.ɵɵelementStart(4, "div", 1);
        ɵngcc0.ɵɵelement(5, "template", null, 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 1, "New bulk operation"));
    } }, directives: [ɵngcc3.TitleComponent], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsStepperContainerComponent, [{
        type: core.Component,
        args: [{
                selector: 'bulk-operations-stepper-container',
                template: "<c8y-title>{{ 'New bulk operation' | translate }}</c8y-title>\n<div class=\"card card--grid--fullpage\">\n  <div class=\"d-contents\">\n    <template #container></template>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.BulkOperationsService }, { type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc0.ChangeDetectorRef }]; }, { container: [{
            type: core.ViewChild,
            args: ['container', { static: false, read: core.ViewContainerRef }]
        }] }); })();
        return BulkOperationsStepperContainerComponent;
    }());
    BulkOperationsStepperContainerComponent.ctorParameters = function () { return [
        { type: router.ActivatedRoute },
        { type: bulkOperationsService.BulkOperationsService },
        { type: core.ComponentFactoryResolver },
        { type: core.ChangeDetectorRef }
    ]; };
    BulkOperationsStepperContainerComponent.propDecorators = {
        container: [{ type: core.ViewChild, args: ['container', { static: false, read: core.ViewContainerRef },] }]
    };

    var bulkOperatioStepperRoutes = [
        {
            path: 'devicecontrol/bulk/creation/:type',
            component: BulkOperationsStepperContainerComponent
        }
    ];
    /**
     * Container module for the new bulk operation steppers. Use this module in your application
     * to provide a step by step guided process for scheduling of bulk operations.
     * The module reads the bulk operation type from the URL and displays the corresponding stepper.
     * You can provide stepper for your custom bulk operation type by registering it with the
     * HOOK_LIST_BULK_TYPE injection token.
     */
    var BulkOperationsStepperContainerModule = /** @class */ (function () {
        function BulkOperationsStepperContainerModule() {
        }
BulkOperationsStepperContainerModule.ɵfac = function BulkOperationsStepperContainerModule_Factory(t) { return new (t || BulkOperationsStepperContainerModule)(); };
BulkOperationsStepperContainerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationsStepperContainerModule });
BulkOperationsStepperContainerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[ngxComponents.CoreModule, router.RouterModule.forChild(bulkOperatioStepperRoutes)]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsStepperContainerModule, [{
        type: core.NgModule,
        args: [{
                declarations: [BulkOperationsStepperContainerComponent],
                imports: [ngxComponents.CoreModule, router.RouterModule.forChild(bulkOperatioStepperRoutes)],
                entryComponents: [BulkOperationsStepperContainerComponent]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationsStepperContainerModule, { declarations: [BulkOperationsStepperContainerComponent], imports: [ɵngcc3.CoreModule, ɵngcc1.RouterModule] }); })();
        return BulkOperationsStepperContainerModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.BulkOperationsStepperContainerComponent = BulkOperationsStepperContainerComponent;
    exports.BulkOperationsStepperContainerModule = BulkOperationsStepperContainerModule;
    exports.bulkOperatioStepperRoutes = bulkOperatioStepperRoutes;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
