// @c8y/ngx-components/esm2015/connectivity/impact/impact-node.factory.js
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, toObservable } from '@c8y/ngx-components';
import { map } from 'rxjs/operators';
import { ImpactTabFactory } from './impact-tab.factory';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './impact-tab.factory';
export class ImpactNodeFactory {
    constructor(tabFactory) {
        this.tabFactory = tabFactory;
        this.connectivityNode = new NavigatorNode({
            parent: gettext('Settings'),
            label: gettext('Connectivity'),
            path: 'connectivitySettings',
            icon: 'plug',
            preventDuplicates: true
        });
    }
    get() {
        return toObservable(this.tabFactory.tab()).pipe(map(tab => {
            return !!tab ? [this.connectivityNode] : [];
        }));
    }
}
ImpactNodeFactory.ɵfac = function ImpactNodeFactory_Factory(t) { return new (t || ImpactNodeFactory)(ɵngcc0.ɵɵinject(ɵngcc1.ImpactTabFactory)); };
ImpactNodeFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImpactNodeFactory, factory: ImpactNodeFactory.ɵfac });
ImpactNodeFactory.ctorParameters = () => [
    { type: ImpactTabFactory }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactNodeFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ImpactTabFactory }]; }, null); })();
