// @c8y/ngx-components/esm2015/connectivity/impact/impact-connectivity.service.js
import { Injectable } from '@angular/core';
import { FetchClient } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ImpactConnectivityService {
    constructor(client) {
        this.client = client;
        this.microserviceUrl = '/service/impact/connectivity';
        this.header = { 'Content-Type': 'application/json' };
    }
    getOptions() {
        const options = {
            method: 'GET',
            headers: this.header
        };
        return this.client.fetch(`${this.microserviceUrl}`, options);
    }
    updateOptions(impactOptions) {
        const options = {
            method: 'POST',
            headers: this.header,
            body: JSON.stringify(impactOptions)
        };
        return this.client.fetch(`${this.microserviceUrl}`, options);
    }
    deleteOptions() {
        const options = {
            method: 'DELETE',
            headers: this.header
        };
        return this.client.fetch(`${this.microserviceUrl}`, options);
    }
}
ImpactConnectivityService.ɵfac = function ImpactConnectivityService_Factory(t) { return new (t || ImpactConnectivityService)(ɵngcc0.ɵɵinject(ɵngcc1.FetchClient)); };
ImpactConnectivityService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImpactConnectivityService, factory: ImpactConnectivityService.ɵfac });
ImpactConnectivityService.ctorParameters = () => [
    { type: FetchClient }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactConnectivityService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.FetchClient }]; }, null); })();
