// @c8y/ngx-components/esm2015/connectivity/impact/impact.module.js
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, FormsModule, HOOK_NAVIGATOR_NODES, HOOK_TABS } from '@c8y/ngx-components';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { ImpactConnectivityService } from './impact-connectivity.service';
import { ImpactNodeFactory } from './impact-node.factory';
import { ImpactProviderSettingsComponent } from './impact-provider-settings.component';
import { ImpactTabFactory } from './impact-tab.factory';
import { ImpactGuard } from './impact.guard';
import { StatusDisplayComponent } from './status-display.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
const impactRoutes = [
    {
        path: 'connectivitySettings/impact_provider_settings',
        component: ImpactProviderSettingsComponent,
        canActivate: [ImpactGuard]
    }
];
export class ImpactModule {
}
ImpactModule.ɵfac = function ImpactModule_Factory(t) { return new (t || ImpactModule)(); };
ImpactModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ImpactModule });
ImpactModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ImpactConnectivityService,
        ImpactGuard,
        ImpactTabFactory,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: ImpactNodeFactory,
            multi: true
        },
        {
            provide: HOOK_TABS,
            useClass: ImpactTabFactory,
            multi: true
        }
    ], imports: [[
            CoreModule,
            CommonModule,
            FormsModule,
            RouterModule.forChild(impactRoutes),
            PopoverModule,
            ReactiveFormsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactModule, [{
        type: NgModule,
        args: [{
                declarations: [ImpactProviderSettingsComponent, StatusDisplayComponent],
                exports: [ImpactProviderSettingsComponent],
                imports: [
                    CoreModule,
                    CommonModule,
                    FormsModule,
                    RouterModule.forChild(impactRoutes),
                    PopoverModule,
                    ReactiveFormsModule
                ],
                entryComponents: [ImpactProviderSettingsComponent],
                providers: [
                    ImpactConnectivityService,
                    ImpactGuard,
                    ImpactTabFactory,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: ImpactNodeFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_TABS,
                        useClass: ImpactTabFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ImpactModule, { declarations: function () { return [ImpactProviderSettingsComponent, StatusDisplayComponent]; }, imports: function () { return [CoreModule,
        CommonModule,
        FormsModule, ɵngcc1.RouterModule, PopoverModule,
        ReactiveFormsModule]; }, exports: function () { return [ImpactProviderSettingsComponent]; } }); })();
