// @c8y/ngx-components/esm2015/ecosystem/microservices/add-microservice.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { gettext, ModalService, Status } from '@c8y/ngx-components';
import { ApplicationType, TenantService } from '@c8y/client';
import { EcosystemService } from '../ecosystem.service';
import { ERROR_TYPE } from '../ecosystem.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '../shared/add-application.component';
export class AddMicroserviceComponent {
    constructor(ecosystemService, modal, tenantService) {
        this.ecosystemService = ecosystemService;
        this.modal = modal;
        this.tenantService = tenantService;
        this.headerText = gettext('Add microservice');
        this.successText = gettext('Microservice created');
        this.createMicroserviceApplicationHandler = (f) => this.createMicroserviceApplication(f);
        this.uploadMicroserviceHandler = (f, app) => this.uploadMicroservice(f, app);
    }
    createMicroserviceApplication(file) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.isValidAppType(file, ApplicationType.MICROSERVICE);
            return this.ecosystemService.createAppForArchive(file);
        });
    }
    uploadMicroservice(file, microservice) {
        return __awaiter(this, void 0, void 0, function* () {
            const subscribeToCurrentTenant = yield this.askIfActivationAfterUploadNeeded();
            yield this.ecosystemService.uploadArchiveToApp(file, microservice);
            yield this.subscribeMicroservice(microservice, subscribeToCurrentTenant);
        });
    }
    askIfActivationAfterUploadNeeded() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modal.confirm(gettext('Subscribe to microservice'), gettext('You are about to subscribe to the microservice after upload. Do you want to subscribe to it?'), Status.INFO, { ok: gettext('Subscribe'), cancel: gettext("Don't subscribe") });
                return true;
            }
            catch (ex) {
                return false;
            }
        });
    }
    subscribeMicroservice(app, subscribeToCurrentTenant) {
        return __awaiter(this, void 0, void 0, function* () {
            const tenant = (yield this.tenantService.current()).data;
            const applications = tenant.applications.references;
            const isSubscribed = applications.some(({ application }) => application.id === app.id);
            if (!isSubscribed && subscribeToCurrentTenant) {
                try {
                    return yield this.tenantService.subscribeApplication(tenant, app);
                }
                catch (res) {
                    if (res.status === 409) {
                        throw Error(ERROR_TYPE.ALREADY_SUBSCRIBED);
                    }
                }
            }
            else if (isSubscribed && !subscribeToCurrentTenant) {
                return this.tenantService.unsubscribeApplication(tenant, app);
            }
        });
    }
}
AddMicroserviceComponent.ɵfac = function AddMicroserviceComponent_Factory(t) { return new (t || AddMicroserviceComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.TenantService)); };
AddMicroserviceComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddMicroserviceComponent, selectors: [["c8y-add-microservice"]], decls: 1, vars: 5, consts: [[3, "headerText", "headerIcon", "successText", "createApplicationHandler", "uploadApplicationHandler"]], template: function AddMicroserviceComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-add-application", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("headerText", ctx.headerText)("headerIcon", "microchip")("successText", ctx.successText)("createApplicationHandler", ctx.createMicroserviceApplicationHandler)("uploadApplicationHandler", ctx.uploadMicroserviceHandler);
    } }, directives: [ɵngcc4.AddApplicationComponent], encapsulation: 2 });
AddMicroserviceComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: ModalService },
    { type: TenantService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddMicroserviceComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-microservice',
                template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'microchip'"
    [successText]="successText"
    [createApplicationHandler]="createMicroserviceApplicationHandler"
    [uploadApplicationHandler]="uploadMicroserviceHandler"
  ></c8y-add-application>`
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.ModalService }, { type: ɵngcc3.TenantService }]; }, null); })();
