// @c8y/ngx-components/esm2015/core/common/number.pipe.js
import { DecimalPipe as NgDecimalPipe } from '@angular/common';
import { Inject, LOCALE_ID, Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export const NUMBER_FORMAT_REGEXP = /^(\d+)?\.((\d+)(-(\d+))?)?$/;
/**
 * Formats a value according to digit options and locale rules.
 * Extends the behavior of Angular's <code>number</code> pipe by
 * providing an option to define if a truncated value should be
 * rounded up or down (with the <code>number</code> pipe the
 * value will be rounded using the "to-nearest" method).
 *
 * ```html
 * <span>{{ 3.45612 | c8yNumber: 'ceil':'1.1-3' }}</span> <!-- 3.457 -->
 * <span>{{ 3.46 | c8yNumber: 'ceil':'1.3-3' }}</span> <!-- 3.460 -->
 * <span>{{ 3.45685 | c8yNumber: 'floor':'1.1-3' }}</span> <!-- 3.456 -->```
 */
export class NumberPipe extends NgDecimalPipe {
    constructor(locale) {
        super(locale);
    }
    /**
     * @param value The value to be formatted.
     * @param round When the value needs to be truncated defines if it will be rounded up (='ceil') or down (='floor')
     * @param digitsInfo Sets digit and decimal representation.
     * @param locale Specifies what locale format rules to use.
     */
    transform(value, round, digitsInfo, locale) {
        if (round) {
            let maxFractionPart = 0;
            if (digitsInfo) {
                const parts = digitsInfo.match(NUMBER_FORMAT_REGEXP);
                if (parts === null) {
                    throw new Error(`${digitsInfo} is not a valid digit info`);
                }
                maxFractionPart = Number.parseInt(parts[5], 10);
            }
            const multiplicator = Math.pow(10, maxFractionPart);
            value = strToNumber(value);
            value = Math[round](value * multiplicator) / multiplicator;
        }
        return super.transform(value, digitsInfo, locale);
    }
}
NumberPipe.ɵfac = function NumberPipe_Factory(t) { return new (t || NumberPipe)(ɵngcc0.ɵɵdirectiveInject(LOCALE_ID, 16)); };
NumberPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "c8yNumber", type: NumberPipe, pure: true });
NumberPipe.ctorParameters = () => [
    { type: String, decorators: [{ type: Inject, args: [LOCALE_ID,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NumberPipe, [{
        type: Pipe,
        args: [{ name: 'c8yNumber' }]
    }], function () { return [{ type: String, decorators: [{
                type: Inject,
                args: [LOCALE_ID]
            }] }]; }, null); })();
/**
 * Transforms a string into a number (if needed).
 */
function strToNumber(value) {
    // Convert strings to numbers
    if (typeof value === 'string' && !isNaN(Number(value) - parseFloat(value))) {
        return Number(value);
    }
    if (typeof value !== 'number') {
        throw new Error(`${value} is not a number`);
    }
    return value;
}
