// @c8y/ngx-components/esm2015/core/common/progress-bar.component.js
import { Component, Input } from '@angular/core';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../i18n/c8y-translate.pipe';

function ProgressBarComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵelement(1, "div", 5);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("width", ctx_r0.progress + "%");
    ɵngcc0.ɵɵattribute("aria-valuenow", ctx_r0.progress);
} }
function ProgressBarComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 6);
} }
function ProgressBarComponent_p_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r2.message), " ");
} }
export class ProgressBarComponent {
    constructor() {
        this.message = gettext('Loading…');
        this.progress = null;
    }
}
ProgressBarComponent.ɵfac = function ProgressBarComponent_Factory(t) { return new (t || ProgressBarComponent)(); };
ProgressBarComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ProgressBarComponent, selectors: [["c8y-progress-bar"]], inputs: { message: "message", progress: "progress" }, decls: 4, vars: 3, consts: [[1, "bg-inherit"], ["class", "progress progress-striped active", 4, "ngIf"], ["class", "spinner-snake", 4, "ngIf"], [4, "ngIf"], [1, "progress", "progress-striped", "active"], ["role", "progressbar", "aria-valuemin", "0", "aria-valuemax", "100", 1, "progress-bar"], [1, "spinner-snake"]], template: function ProgressBarComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ProgressBarComponent_div_1_Template, 2, 3, "div", 1);
        ɵngcc0.ɵɵtemplate(2, ProgressBarComponent_div_2_Template, 1, 0, "div", 2);
        ɵngcc0.ɵɵtemplate(3, ProgressBarComponent_p_3_Template, 3, 3, "p", 3);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.progress !== null);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.progress === null);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.message);
    } }, directives: [ɵngcc1.NgIf], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ProgressBarComponent.propDecorators = {
    message: [{ type: Input }],
    progress: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ProgressBarComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-progress-bar',
                template: "<div class=\"bg-inherit\">\n  <div class=\"progress progress-striped active\" *ngIf=\"progress !== null\">\n    <div\n    class=\"progress-bar\"\n    role=\"progressbar\"\n    [attr.aria-valuenow]=\"progress\"\n    aria-valuemin=\"0\"\n    aria-valuemax=\"100\"\n    [style.width]=\"progress + '%'\"\n    ></div>\n  </div>\n  <div class=\"spinner-snake\" *ngIf=\"progress === null\"></div>\n  <p *ngIf=\"message\">\n    {{ message | translate }}\n  </p>\n</div>\n"
            }]
    }], function () { return []; }, { message: [{
            type: Input
        }], progress: [{
            type: Input
        }] }); })();
