// @c8y/ngx-components/esm2015/core/common/status.model.js
import { OperationStatus } from '@c8y/client';
export var Status;
(function (Status) {
    Status["SUCCESS"] = "success";
    Status["WARNING"] = "warning";
    Status["DANGER"] = "danger";
    Status["INFO"] = "info";
})(Status || (Status = {}));
export const statusIcons = {
    [Status.SUCCESS]: 'check-circle',
    [Status.WARNING]: 'exclamation-triangle',
    [Status.DANGER]: 'exclamation-circle',
    [Status.INFO]: 'info-circle'
};
export const statusClasses = {
    [Status.SUCCESS]: `text-${Status.SUCCESS}`,
    [Status.WARNING]: `text-${Status.WARNING}`,
    [Status.DANGER]: `text-${Status.DANGER}`,
    [Status.INFO]: `text-${Status.INFO}`
};
export const statusAlert = {
    [Status.SUCCESS]: `alert-${Status.SUCCESS}`,
    [Status.WARNING]: `alert-${Status.WARNING}`,
    [Status.DANGER]: `alert-${Status.DANGER}`,
    [Status.INFO]: `alert-${Status.INFO}`
};
export const operationStatusIcons = {
    [OperationStatus.PENDING]: 'clock-o',
    [OperationStatus.EXECUTING]: 'refresh',
    [OperationStatus.SUCCESSFUL]: 'check-circle',
    [OperationStatus.FAILED]: 'exclamation-circle'
};
export const operationStatusClasses = {
    [OperationStatus.PENDING]: 'text-info',
    [OperationStatus.EXECUTING]: 'text-info',
    [OperationStatus.SUCCESSFUL]: 'text-success',
    [OperationStatus.FAILED]: 'text-danger'
};
