// @c8y/ngx-components/esm2015/core/common/textarea-autoresize.directive.js
import { Directive, HostListener, ElementRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class TextareaAutoresizeDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.focused = false;
        this.resizeObserver = new window.ResizeObserver(() => !this.focused && this.resize());
    }
    onInput() {
        this.resize();
    }
    onFocus() {
        this.focused = true;
    }
    onBlur() {
        this.focused = false;
    }
    ngOnInit() {
        var _a;
        this.resizeObserver.observe((_a = this.elementRef) === null || _a === void 0 ? void 0 : _a.nativeElement);
        setTimeout(() => this.resize());
    }
    ngOnDestroy() {
        var _a;
        this.resizeObserver.unobserve((_a = this.elementRef) === null || _a === void 0 ? void 0 : _a.nativeElement);
    }
    resize() {
        var _a;
        this.elementRef.nativeElement.style.height = '32px';
        this.elementRef.nativeElement.style.height = ((_a = this.elementRef) === null || _a === void 0 ? void 0 : _a.nativeElement.scrollHeight) + 'px';
    }
}
TextareaAutoresizeDirective.ɵfac = function TextareaAutoresizeDirective_Factory(t) { return new (t || TextareaAutoresizeDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
TextareaAutoresizeDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: TextareaAutoresizeDirective, selectors: [["", "c8y-textarea-autoresize", ""]], hostBindings: function TextareaAutoresizeDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("input", function TextareaAutoresizeDirective_input_HostBindingHandler() { return ctx.onInput(); })("focus", function TextareaAutoresizeDirective_focus_HostBindingHandler() { return ctx.onFocus(); })("blur", function TextareaAutoresizeDirective_blur_HostBindingHandler() { return ctx.onBlur(); });
    } } });
TextareaAutoresizeDirective.ctorParameters = () => [
    { type: ElementRef }
];
TextareaAutoresizeDirective.propDecorators = {
    onInput: [{ type: HostListener, args: [':input',] }],
    onFocus: [{ type: HostListener, args: [':focus',] }],
    onBlur: [{ type: HostListener, args: [':blur',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TextareaAutoresizeDirective, [{
        type: Directive,
        args: [{
                selector: '[c8y-textarea-autoresize]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { onInput: [{
            type: HostListener,
            args: [':input']
        }], onFocus: [{
            type: HostListener,
            args: [':focus']
        }], onBlur: [{
            type: HostListener,
            args: [':blur']
        }] }); })();
