// @c8y/ngx-components/esm2015/core/realtime/realtime-button.component.js
import { Component, Input } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../i18n/c8y-translate.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';

const _c0 = function (a0, a1) { return { active: a0, inactive: a1 }; };
export class RealtimeButtonComponent {
    get active() {
        return !!this.service && this.service.active;
    }
    toggle() {
        if (this.service) {
            this.active ? this.service.stop() : this.service.start();
        }
    }
}
RealtimeButtonComponent.ɵfac = function RealtimeButtonComponent_Factory(t) { return new (t || RealtimeButtonComponent)(); };
RealtimeButtonComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: RealtimeButtonComponent, selectors: [["c8y-realtime-btn"]], inputs: { service: "service" }, decls: 5, vars: 7, consts: [["type", "button", 1, "btn", "btn-link", "c8y-realtime", 3, "title", "click"], [1, "c8y-pulse", 3, "ngClass"], ["translate", ""]], template: function RealtimeButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("click", function RealtimeButtonComponent_Template_button_click_0_listener() { return ctx.toggle(); });
        ɵngcc0.ɵɵpipe(1, "translate");
        ɵngcc0.ɵɵelement(2, "span", 1);
        ɵngcc0.ɵɵelementStart(3, "span", 2);
        ɵngcc0.ɵɵtext(4, "Realtime");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Toggle realtime"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c0, ctx.active, !ctx.active));
    } }, directives: [ɵngcc1.NgClass, ɵngcc2.C8yTranslateDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
RealtimeButtonComponent.propDecorators = {
    service: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RealtimeButtonComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-realtime-btn',
                template: "<button\n  type=\"button\"\n  class=\"btn btn-link c8y-realtime\"\n  title=\"{{ 'Toggle realtime' | translate }}\"\n  (click)=\"toggle()\"\n>\n  <span class=\"c8y-pulse\" [ngClass]=\"{ active: active, inactive: !active }\"></span>\n  <span translate>Realtime</span>\n</button>\n"
            }]
    }], null, { service: [{
            type: Input
        }] }); })();
