// @c8y/ngx-components/esm2015/core/realtime/realtime.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { RealtimeButtonComponent } from './realtime-button.component';
import * as ɵngcc0 from '@angular/core';
export class RealtimeModule {
}
RealtimeModule.ɵfac = function RealtimeModule_Factory(t) { return new (t || RealtimeModule)(); };
RealtimeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: RealtimeModule });
RealtimeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RealtimeModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [RealtimeButtonComponent],
                exports: [RealtimeButtonComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RealtimeModule, { declarations: function () { return [RealtimeButtonComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [RealtimeButtonComponent]; } }); })();
