// @c8y/ngx-components/esm2015/core/search/search.component.js
import { Component, Input, Output, ViewChild, EventEmitter } from '@angular/core';
import { SearchService } from './search.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './search.service';

const _c0 = ["template"];
function SearchComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c1 = ["*"];
export class SearchComponent {
    constructor(searchService) {
        this.searchService = searchService;
        this.icon = 'search';
        this.priority = 0;
        this.search = new EventEmitter();
        this.term = '';
    }
    onSearch() {
        this.search.emit(this);
    }
    ngAfterViewInit() {
        this.template = undefined;
        this.viewInitTimeout = setTimeout(() => this.searchService.add(this));
    }
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.searchService.remove(this);
    }
}
SearchComponent.ɵfac = function SearchComponent_Factory(t) { return new (t || SearchComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.SearchService)); };
SearchComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchComponent, selectors: [["c8y-search"]], viewQuery: function SearchComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { icon: "icon", priority: "priority", term: "term", name: "name" }, outputs: { search: "search" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""]], template: function SearchComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, SearchComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
SearchComponent.ctorParameters = () => [
    { type: SearchService }
];
SearchComponent.propDecorators = {
    name: [{ type: Input }],
    icon: [{ type: Input }],
    priority: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: false },] }],
    search: [{ type: Output }],
    term: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search',
                template: "<ng-template #template>\n  <ng-content></ng-content>\n</ng-template>"
            }]
    }], function () { return [{ type: ɵngcc1.SearchService }]; }, { icon: [{
            type: Input
        }], priority: [{
            type: Input
        }], search: [{
            type: Output
        }], term: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }], name: [{
            type: Input
        }] }); })();
