// @c8y/ngx-components/esm2015/core/search/search.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { SearchOutletComponent } from './search-outlet.component';
import { SearchComponent } from './search.component';
import { FormsModule } from '@angular/forms';
import { HighlightComponent } from './highlight.component';
import { SearchResultEmptyComponent } from './search-result-empty.component';
/**
 * The angular module definition for the search.
 * @exports SearchComponent
 * @exports SearchOutletComponent
 * @exports HighlightComponent A component to highlight search results.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class SearchModule {
}
SearchModule.ɵfac = function SearchModule_Factory(t) { return new (t || SearchModule)(); };
SearchModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SearchModule });
SearchModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            BsDropdownModule,
            CommonModule,
            RouterModule.forChild([])
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    SearchOutletComponent,
                    SearchComponent,
                    HighlightComponent,
                    SearchResultEmptyComponent
                ],
                imports: [
                    FormsModule,
                    BsDropdownModule,
                    CommonModule,
                    RouterModule.forChild([])
                ],
                exports: [
                    SearchOutletComponent,
                    SearchComponent,
                    HighlightComponent,
                    SearchResultEmptyComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SearchModule, { declarations: function () { return [SearchOutletComponent,
        SearchComponent,
        HighlightComponent,
        SearchResultEmptyComponent]; }, imports: function () { return [FormsModule,
        BsDropdownModule,
        CommonModule, ɵngcc1.RouterModule]; }, exports: function () { return [SearchOutletComponent,
        SearchComponent,
        HighlightComponent,
        SearchResultEmptyComponent]; } }); })();
