// @c8y/ngx-components/esm2015/core/search/search-outlet.component.js
import { Component, Input } from '@angular/core';
import { sortBy } from 'lodash';
import { AlertService } from '../alert/alert.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../alert/alert.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '../common/outlet.directive';

function SearchOutletComponent_ng_container_0_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function SearchOutletComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, SearchOutletComponent_ng_container_0_ng_container_1_Template, 1, 0, "ng-container", 1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r1 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yOutlet", item_r1.template);
} }
export class SearchOutletComponent {
    constructor(alertService) {
        this.alertService = alertService;
        this.search = [];
    }
    getByPriority() {
        return sortBy(this.search, this.byPriority);
    }
    byPriority(item) {
        return -item.priority;
    }
}
SearchOutletComponent.ɵfac = function SearchOutletComponent_Factory(t) { return new (t || SearchOutletComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
SearchOutletComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchOutletComponent, selectors: [["c8y-search-outlet"]], inputs: { search: "search" }, decls: 1, vars: 1, consts: [[4, "ngFor", "ngForOf"], [4, "c8yOutlet"]], template: function SearchOutletComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SearchOutletComponent_ng_container_0_Template, 2, 1, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngForOf", ctx.getByPriority());
    } }, directives: [ɵngcc2.NgForOf, ɵngcc3.OutletDirective], encapsulation: 2 });
SearchOutletComponent.ctorParameters = () => [
    { type: AlertService }
];
SearchOutletComponent.propDecorators = {
    search: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchOutletComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search-outlet',
                template: "<ng-container *ngFor=\"let item of getByPriority()\">\n  <ng-container *c8yOutlet=\"item.template\"></ng-container>\n</ng-container>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }]; }, { search: [{
            type: Input
        }] }); })();
