// @c8y/ngx-components/esm2015/child-devices/child-devices.guard.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { InventoryService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ChildDevicesTabGuard {
    constructor(inventory) {
        this.inventory = inventory;
    }
    canActivate(route) {
        return __awaiter(this, void 0, void 0, function* () {
            const contextData = route.data.contextData;
            const filter = {
                pageSize: 1
            };
            if (!contextData) {
                return true;
            }
            const device = yield this.inventory.childDevicesList(contextData.id, filter);
            const hasChildren = device.data.length;
            const isVendme = route.data.contextData.type === 'com_nsn_startups_vendme_VendingMachine';
            return hasChildren && !isVendme;
        });
    }
}
ChildDevicesTabGuard.ɵfac = function ChildDevicesTabGuard_Factory(t) { return new (t || ChildDevicesTabGuard)(ɵngcc0.ɵɵinject(ɵngcc1.InventoryService)); };
ChildDevicesTabGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ChildDevicesTabGuard, factory: ChildDevicesTabGuard.ɵfac });
ChildDevicesTabGuard.ctorParameters = () => [
    { type: InventoryService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ChildDevicesTabGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.InventoryService }]; }, null); })();
