// @c8y/ngx-components/esm2015/core/list-group/list-item-body.component.js
import { Component, Input } from '@angular/core';
/**
 * The li-body component can be used to define the body.
 * Usually it doesn't need to be used, as it is the default
 * content projected to a <c8y-li>.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function ListItemBodyComponent_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.body);
} }
const _c0 = ["*"];
export class ListItemBodyComponent {
}
ListItemBodyComponent.ɵfac = function ListItemBodyComponent_Factory(t) { return new (t || ListItemBodyComponent)(); };
ListItemBodyComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListItemBodyComponent, selectors: [["c8y-list-item-body"], ["c8y-li-body"]], inputs: { body: "body" }, ngContentSelectors: _c0, decls: 2, vars: 1, consts: [[4, "ngIf"]], template: function ListItemBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ListItemBodyComponent_span_0_Template, 2, 1, "span", 0);
        ɵngcc0.ɵɵprojection(1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.body);
    } }, directives: [ɵngcc1.NgIf], encapsulation: 2 });
ListItemBodyComponent.propDecorators = {
    body: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListItemBodyComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-item-body, c8y-li-body',
                template: "<span *ngIf=\"body\">{{body}}</span>\n<ng-content></ng-content>\n"
            }]
    }], null, { body: [{
            type: Input
        }] }); })();
