// @c8y/ngx-components/esm2015/core/list-group/list-group.component.js
import { Component, HostBinding, ContentChild } from '@angular/core';
import { ListItemTimelineComponent } from './list-item-timeline.component';
/**
 * List groups are flexible and powerful components for displaying
 * both simple lists of elements and complex ones with custom content.
 *
 * @example
 * ```html
 * <c8y-list-group>
 *   <c8y-li>
 *     Plain <strong>Text</strong>
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-icon [icon]="'rocket'"></c8y-li-icon>
 *     Icon
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-radio></c8y-li-radio>
 *     Radio
 *   </c8y-li>
 *   <c8y-li>
 *     <c8y-li-checkbox></c8y-li-checkbox>
 *     Checkbox
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-footer>
 *       Footer
 *     </c8y-li-footer>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-action icon="rocket">
 *       Launch
 *     </c8y-li-action>
 *   </c8y-li>
 *   <c8y-li>
 *     Content
 *     <c8y-li-collapse>
 *       Collapsed content
 *     </c8y-li-collapse>
 *   </c8y-li>
 * </c8y-list-group>
 * ```
 */
import * as ɵngcc0 from '@angular/core';

const _c0 = ["*"];
export class ListGroupComponent {
    constructor() {
        /**
         * @ignore
         */
        this.timeline = false;
    }
    /**
     * @ignore
     */
    ngAfterContentChecked() {
        this.timeline = !!this.timelineComponent;
    }
}
ListGroupComponent.ɵfac = function ListGroupComponent_Factory(t) { return new (t || ListGroupComponent)(); };
ListGroupComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ListGroupComponent, selectors: [["c8y-list-group"]], contentQueries: function ListGroupComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, ListItemTimelineComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.timelineComponent = _t.first);
    } }, hostAttrs: [1, "c8y-list__group"], hostVars: 2, hostBindings: function ListGroupComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("c8y-list--timeline", ctx.timeline);
    } }, ngContentSelectors: _c0, decls: 1, vars: 0, template: function ListGroupComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
    } }, encapsulation: 2 });
ListGroupComponent.propDecorators = {
    timeline: [{ type: HostBinding, args: ['class.c8y-list--timeline',] }],
    timelineComponent: [{ type: ContentChild, args: [ListItemTimelineComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ListGroupComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-list-group',
                template: "<ng-content></ng-content>\n\n",
                host: {
                    class: 'c8y-list__group'
                }
            }]
    }], function () { return []; }, { timeline: [{
            type: HostBinding,
            args: ['class.c8y-list--timeline']
        }], timelineComponent: [{
            type: ContentChild,
            args: [ListItemTimelineComponent, { static: false }]
        }] }); })();
