// @c8y/ngx-components/esm2015/auth-configuration/basic-auth-settings.component.js
import { Component, Input } from '@angular/core';
import { TenantLoginOptionsService, TenantLoginOptionType } from '@c8y/client';
import { defaults, isEmpty, omitBy } from 'lodash-es';
import { TenantUiService } from '@c8y/ngx-components';
import { ControlContainer, NgForm } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/forms';

function BasicAuthSettingsComponent_div_0_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "You are about to forbid browsers from using Basic authentication. This will prevent users from using web applications on your tenant because you are going to set Basic authentication as the preferred login mode."), " ");
} }
function BasicAuthSettingsComponent_div_0_div_23_button_7_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 16);
    ɵngcc0.ɵɵlistener("click", function BasicAuthSettingsComponent_div_0_div_23_button_7_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(3); return ctx_r8.add(ctx_r8.forbiddenUserAgents); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Add"));
} }
function BasicAuthSettingsComponent_div_0_div_23_Template(rf, ctx) { if (rf & 1) {
    const _r11 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "input", 14);
    ɵngcc0.ɵɵlistener("ngModelChange", function BasicAuthSettingsComponent_div_0_div_23_Template_input_ngModelChange_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r11); const i_r5 = restoredCtx.index; const ctx_r10 = ɵngcc0.ɵɵnextContext(2); return ctx_r10.forbiddenUserAgentsChanged({ event: $event, index: i_r5 }); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 15);
    ɵngcc0.ɵɵelementStart(4, "button", 16);
    ɵngcc0.ɵɵlistener("click", function BasicAuthSettingsComponent_div_0_div_23_Template_button_click_4_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r11); const i_r5 = restoredCtx.index; const ctx_r12 = ɵngcc0.ɵɵnextContext(2); return ctx_r12.forbiddenUserAgentsRemove(i_r5); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, BasicAuthSettingsComponent_div_0_div_23_button_7_Template, 3, 3, "button", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const forbiddenUserAgent_r4 = ctx.$implicit;
    const i_r5 = ctx.index;
    const isLast_r6 = ctx.last;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(2, 5, "e.g."), " forbidden-agent");
    ɵngcc0.ɵɵproperty("name", "forbiddenUserAgent" + i_r5)("ngModel", forbiddenUserAgent_r4.value);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 7, "Remove"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", isLast_r6);
} }
function BasicAuthSettingsComponent_div_0_div_29_button_7_Template(rf, ctx) { if (rf & 1) {
    const _r18 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 16);
    ɵngcc0.ɵɵlistener("click", function BasicAuthSettingsComponent_div_0_div_29_button_7_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r18); const ctx_r17 = ɵngcc0.ɵɵnextContext(3); return ctx_r17.add(ctx_r17.trustedUserAgents); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 19);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 1, "Add"));
} }
function BasicAuthSettingsComponent_div_0_div_29_Template(rf, ctx) { if (rf & 1) {
    const _r20 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "input", 20);
    ɵngcc0.ɵɵlistener("ngModelChange", function BasicAuthSettingsComponent_div_0_div_29_Template_input_ngModelChange_1_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r20); const i_r14 = restoredCtx.index; const ctx_r19 = ɵngcc0.ɵɵnextContext(2); return ctx_r19.trustedUserAgentsChanged({ event: $event, index: i_r14 }); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 15);
    ɵngcc0.ɵɵelementStart(4, "button", 16);
    ɵngcc0.ɵɵlistener("click", function BasicAuthSettingsComponent_div_0_div_29_Template_button_click_4_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r20); const i_r14 = restoredCtx.index; const ctx_r21 = ɵngcc0.ɵɵnextContext(2); return ctx_r21.trustedUserAgentsRemove(i_r14); });
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelement(6, "i", 17);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, BasicAuthSettingsComponent_div_0_div_29_button_7_Template, 3, 3, "button", 18);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const trustedUserAgent_r13 = ctx.$implicit;
    const i_r14 = ctx.index;
    const isLast_r15 = ctx.last;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(2, 5, "e.g."), " trusted-agent");
    ɵngcc0.ɵɵproperty("name", "trustedUserAgent" + i_r14)("ngModel", trustedUserAgent_r13.value);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 7, "Remove"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", isLast_r15);
} }
function BasicAuthSettingsComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r23 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵelementStart(2, "h4", 3);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 4);
    ɵngcc0.ɵɵelementStart(6, "div", 5);
    ɵngcc0.ɵɵelementStart(7, "div", 6);
    ɵngcc0.ɵɵelementStart(8, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(9, "label", 7);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementStart(11, "input", 8);
    ɵngcc0.ɵɵlistener("ngModelChange", function BasicAuthSettingsComponent_div_0_Template_input_ngModelChange_11_listener($event) { ɵngcc0.ɵɵrestoreView(_r23); const ctx_r22 = ɵngcc0.ɵɵnextContext(); return ctx_r22.forbiddenWebBrowsers = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(12, "span");
    ɵngcc0.ɵɵelementStart(13, "span");
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(16, BasicAuthSettingsComponent_div_0_div_16_Template, 3, 3, "div", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(17, "div", 5);
    ɵngcc0.ɵɵelementStart(18, "div", 6);
    ɵngcc0.ɵɵelementStart(19, "label", 10);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵtext(21);
    ɵngcc0.ɵɵpipe(22, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(23, BasicAuthSettingsComponent_div_0_div_23_Template, 8, 9, "div", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "div", 6);
    ɵngcc0.ɵɵelementStart(25, "label", 10);
    ɵngcc0.ɵɵpipe(26, "translate");
    ɵngcc0.ɵɵtext(27);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(29, BasicAuthSettingsComponent_div_0_div_29_Template, 8, 9, "div", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 11, "Basic Auth restrictions"));
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 13, "Forbidden for web browsers"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngModel", ctx_r0.forbiddenWebBrowsers);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(15, 15, "Forbidden for web browsers"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.preferredLoginOptionType === ctx_r0.tenantLoginOptionTypeEnum.BASIC && ctx_r0.forbiddenWebBrowsers);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 17, "Forbidden user agents"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(22, 19, "Forbidden user agents"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.forbiddenUserAgents);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(26, 21, "Trusted user agents"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(28, 23, "Trusted user agents"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r0.trustedUserAgents);
} }
export class BasicAuthSettingsComponent {
    constructor(tenantLoginOptionsService, tenantUiService) {
        this.tenantLoginOptionsService = tenantLoginOptionsService;
        this.tenantUiService = tenantUiService;
        this.preferredLoginOptionType = TenantLoginOptionType.BASIC;
        this.tenantLoginOptionTypeEnum = TenantLoginOptionType;
    }
    ngOnChanges(changes) {
        if (changes.authConfiguration && changes.authConfiguration.currentValue) {
            const { loginOptions } = changes.authConfiguration.currentValue;
            const basic = loginOptions.find(this.tenantUiService.isBasic) || {};
            basic.authenticationRestrictions = defaults({}, omitBy(basic.authenticationRestrictions, isEmpty), {
                forbiddenUserAgents: [''],
                trustedUserAgents: [''],
                forbiddenClients: []
            });
            this.forbiddenUserAgents = basic.authenticationRestrictions.forbiddenUserAgents.map(val => ({ value: val }));
            this.trustedUserAgents = basic.authenticationRestrictions.trustedUserAgents.map(val => ({ value: val }));
            this.preferredLoginOptionType = changes.authConfiguration.currentValue.preferredLoginOptionType;
        }
    }
    ngDoCheck() {
        if (this.preferredLoginOptionType !== this.authConfiguration.preferredLoginOptionType) {
            this.preferredLoginOptionType = this.authConfiguration.preferredLoginOptionType;
        }
    }
    get forbiddenWebBrowsers() {
        return this.authenticationRestrictions.forbiddenClients.includes('WEB_BROWSERS');
    }
    set forbiddenWebBrowsers(value) {
        this.authenticationRestrictions.forbiddenClients = value ? ['WEB_BROWSERS'] : [];
    }
    forbiddenUserAgentsChanged({ event, index }) {
        this.forbiddenUserAgents[index].value = event;
        this.authenticationRestrictions.forbiddenUserAgents[index] = event;
    }
    forbiddenUserAgentsRemove(index) {
        this.authenticationRestrictions.forbiddenUserAgents.splice(index, 1);
        this.remove(this.forbiddenUserAgents, index);
    }
    trustedUserAgentsChanged({ event, index }) {
        this.trustedUserAgents[index].value = event;
        this.authenticationRestrictions.trustedUserAgents[index] = event;
    }
    trustedUserAgentsRemove(index) {
        this.authenticationRestrictions.trustedUserAgents.splice(index, 1);
        this.remove(this.trustedUserAgents, index);
    }
    get authenticationRestrictions() {
        return this.authConfiguration.loginOptions.find(this.tenantUiService.isBasic)
            .authenticationRestrictions;
    }
    add(collection) {
        collection.push({ value: '' });
    }
    remove(collection, index) {
        collection.splice(index, 1);
        if (!collection.length) {
            collection.push({ value: '' });
        }
    }
}
BasicAuthSettingsComponent.ɵfac = function BasicAuthSettingsComponent_Factory(t) { return new (t || BasicAuthSettingsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TenantLoginOptionsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.TenantUiService)); };
BasicAuthSettingsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BasicAuthSettingsComponent, selectors: [["c8y-basic-auth-settings"]], inputs: { authConfiguration: "authConfiguration" }, features: [ɵngcc0.ɵɵProvidersFeature([], [{ provide: ControlContainer, useExisting: NgForm }]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [["class", "card-block separator-top", 4, "ngIf"], [1, "card-block", "separator-top"], [1, "col-sm-2"], [1, "text-right"], [1, "col-sm-9"], [1, "row"], [1, "col-sm-6"], [1, "c8y-switch", 3, "title"], ["type", "checkbox", "name", "forbiddenWebBrowsers", 3, "ngModel", "ngModelChange"], ["class", "alert alert-warning", 4, "ngIf"], [3, "title"], ["class", "input-group m-t-8", 4, "ngFor", "ngForOf"], [1, "alert", "alert-warning"], [1, "input-group", "m-t-8"], ["type", "text", 1, "form-control", 3, "name", "ngModel", "placeholder", "ngModelChange"], [1, "input-group-btn", "col-sm-2"], ["type", "button", 1, "btn", "btn-clean", "text-primary", 3, "title", "click"], [1, "dlt-c8y-icon-minus-circle", "text-danger"], ["type", "button", "class", "btn btn-clean text-primary", 3, "title", "click", 4, "ngIf"], [1, "dlt-c8y-icon-plus-circle"], ["type", "text", 1, "form-control", 3, "name", "placeholder", "ngModel", "ngModelChange"]], template: function BasicAuthSettingsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, BasicAuthSettingsComponent_div_0_Template, 30, 25, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.authConfiguration);
    } }, directives: [ɵngcc3.NgIf, ɵngcc2.FormGroupComponent, ɵngcc4.CheckboxControlValueAccessor, ɵngcc4.NgControlStatus, ɵngcc4.NgModel, ɵngcc3.NgForOf, ɵngcc4.DefaultValueAccessor], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
BasicAuthSettingsComponent.ctorParameters = () => [
    { type: TenantLoginOptionsService },
    { type: TenantUiService }
];
BasicAuthSettingsComponent.propDecorators = {
    authConfiguration: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BasicAuthSettingsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-basic-auth-settings',
                template: "<div class=\"card-block separator-top\" *ngIf=\"authConfiguration\">\n  <div class=\"col-sm-2\">\n    <h4 class=\"text-right\">{{ 'Basic Auth restrictions' | translate }}</h4>\n  </div>\n\n  <div class=\"col-sm-9\">\n    <div class=\"row\">\n      <div class=\"col-sm-6\">\n        <c8y-form-group>\n          <label class=\"c8y-switch\" title=\"{{ 'Forbidden for web browsers' | translate }}\">\n            <input\n              type=\"checkbox\"\n              name=\"forbiddenWebBrowsers\"\n              [(ngModel)]=\"forbiddenWebBrowsers\"\n            />\n            <span></span>\n            <span>{{ 'Forbidden for web browsers' | translate }}</span>\n          </label>\n          <div class=\"alert alert-warning\" *ngIf=\"preferredLoginOptionType === tenantLoginOptionTypeEnum.BASIC && forbiddenWebBrowsers\">\n            {{ 'You are about to forbid browsers from using Basic authentication. This will prevent users from using web applications on your tenant because you are going to set Basic authentication as the preferred login mode.' | translate }}\n          </div>\n        </c8y-form-group>\n      </div>\n    </div>\n\n    <div class=\"row\">\n      <div class=\"col-sm-6\">\n        <label title=\"{{ 'Forbidden user agents' | translate }}\">{{ 'Forbidden user agents' | translate }}</label>\n        <div\n          class=\"input-group m-t-8\"\n          *ngFor=\"\n            let forbiddenUserAgent of forbiddenUserAgents;\n            index as i;\n            last as isLast\n          \"\n        >\n          <input\n            type=\"text\"\n            [name]=\"'forbiddenUserAgent' + i\"\n            [ngModel]=\"forbiddenUserAgent.value\"\n            (ngModelChange)=\"forbiddenUserAgentsChanged({event: $event, index: i})\"\n            class=\"form-control\"\n            placeholder=\"{{ 'e.g.' | translate }} forbidden-agent\"\n          />\n          <div class=\"input-group-btn col-sm-2\">\n            <button\n              title=\"{{ 'Remove' | translate }}\"\n              type=\"button\"\n              (click)=\"forbiddenUserAgentsRemove(i)\"\n              class=\"btn btn-clean text-primary\"\n            >\n              <i class=\"dlt-c8y-icon-minus-circle text-danger\"></i>\n            </button>\n            <button\n              title=\"{{ 'Add' | translate }}\"\n              type=\"button\"\n              class=\"btn btn-clean text-primary\"\n              (click)=\"add(forbiddenUserAgents)\"\n              *ngIf=\"isLast\"\n            >\n              <i class=\"dlt-c8y-icon-plus-circle\"></i>\n            </button>\n          </div>\n        </div>\n      </div>\n      <div class=\"col-sm-6\">\n        <label title=\"{{ 'Trusted user agents' | translate }}\">{{ 'Trusted user agents' | translate }}</label>\n        <div\n          class=\"input-group m-t-8\"\n          *ngFor=\"\n            let trustedUserAgent of trustedUserAgents;\n            index as i;\n            last as isLast\n          \"\n        >\n          <input\n            type=\"text\"\n            [name]=\"'trustedUserAgent' + i\"\n            class=\"form-control\"\n            placeholder=\"{{ 'e.g.' | translate }} trusted-agent\"\n            [ngModel]=\"trustedUserAgent.value\"\n            (ngModelChange)=\"trustedUserAgentsChanged({event: $event, index: i})\"\n          />\n          <div class=\"input-group-btn col-sm-2\">\n            <button\n              title=\"{{ 'Remove' | translate }}\"\n              type=\"button\"\n              (click)=\"trustedUserAgentsRemove(i)\"\n              class=\"btn btn-clean text-primary\"\n            >\n              <i class=\"dlt-c8y-icon-minus-circle text-danger\"></i>\n            </button>\n            <button\n              title=\"{{ 'Add' | translate }}\"\n              type=\"button\"\n              class=\"btn btn-clean text-primary\"\n              (click)=\"add(trustedUserAgents)\"\n              *ngIf=\"isLast\"\n            >\n              <i class=\"dlt-c8y-icon-plus-circle\"></i>\n            </button>\n          </div>\n        </div>\n      </div>\n    </div>\n  </div>\n</div>\n",
                viewProviders: [{ provide: ControlContainer, useExisting: NgForm }]
            }]
    }], function () { return [{ type: ɵngcc1.TenantLoginOptionsService }, { type: ɵngcc2.TenantUiService }]; }, { authConfiguration: [{
            type: Input
        }] }); })();
