// @c8y/ngx-components/esm2015/core/tabs/tabs.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule } from '../common/common.module';
import { TabComponent } from './tab.component';
import { TabsOutletComponent } from './tabs-outlet.component';
import { ConditionalTabsOutletComponent } from './conditional-tabs-outlet.component';
import { HeaderModule } from '../header/header.module';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class TabsModule {
}
TabsModule.ɵfac = function TabsModule_Factory(t) { return new (t || TabsModule)(); };
TabsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: TabsModule });
TabsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            RouterModule.forChild([]),
            HeaderModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ConditionalTabsOutletComponent,
                    TabsOutletComponent,
                    TabComponent
                ],
                imports: [
                    CommonModule,
                    RouterModule.forChild([]),
                    HeaderModule
                ],
                exports: [
                    ConditionalTabsOutletComponent,
                    TabsOutletComponent,
                    TabComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TabsModule, { declarations: function () { return [ConditionalTabsOutletComponent,
        TabsOutletComponent,
        TabComponent]; }, imports: function () { return [CommonModule, ɵngcc1.RouterModule, HeaderModule]; }, exports: function () { return [ConditionalTabsOutletComponent,
        TabsOutletComponent,
        TabComponent]; } }); })();
