// @c8y/ngx-components/esm2015/core/select-modal/select-modal-filter.pipe.js
import { Pipe } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class SelectModalFilterPipe {
    transform(items, filterTerm) {
        return filterTerm.trim().length === 0
            ? items
            : items.filter((item) => {
                const allValues = this.extractValues(item);
                const filteredValues = allValues.filter((value) => this.filterContainString(value, filterTerm));
                return filteredValues.length > 0;
            });
    }
    filterContainString(value, filterTerm) {
        const term = filterTerm.toLowerCase().trim();
        return value && value.toLowerCase().indexOf(term) > -1;
    }
    extractValues(item) {
        let values = [];
        if (item.value) {
            values.push(item.value);
        }
        Object.keys(item).forEach((key) => {
            if (Array.isArray(item[key])) {
                item[key].forEach(obj => values.concat(this.extractValues(obj)));
            }
            if (item[key] instanceof Object) {
                values = values.concat(this.extractValues(item[key]));
            }
        });
        return values;
    }
}
SelectModalFilterPipe.ɵfac = function SelectModalFilterPipe_Factory(t) { return new (t || SelectModalFilterPipe)(); };
SelectModalFilterPipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "selectModalFilterPipe", type: SelectModalFilterPipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectModalFilterPipe, [{
        type: Pipe,
        args: [{
                name: 'selectModalFilterPipe'
            }]
    }], null, null); })();
