// @c8y/ngx-components/esm2015/core/select-modal/select-modal.component.js
import { __awaiter } from "tslib";
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { gettext } from '../i18n/gettext';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { ModalSelectionMode } from './select-modal.model';
import { assign } from 'lodash-es';
import { Subject } from 'rxjs';
import { debounceTime } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../common/icon.directive';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../list-group/list-group.component';
import * as ɵngcc5 from '../forms/filter-input.component';
import * as ɵngcc6 from '../i18n/c8y-translate.directive';
import * as ɵngcc7 from '../list-group/list-item.component';
import * as ɵngcc8 from '../list-group/list-item-icon.component';
import * as ɵngcc9 from '../list-group/list-item-body.component';
import * as ɵngcc10 from '../list-group/list-item-collapse.component';
import * as ɵngcc11 from '../search/highlight.component';
import * as ɵngcc12 from '../list-group/list-item-radio.component';
import * as ɵngcc13 from '../i18n/c8y-translate.pipe';
import * as ɵngcc14 from './select-modal-filter.pipe';

function SelectModalComponent_c8y_filter_10_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-filter", 15);
    ɵngcc0.ɵɵlistener("onSearch", function SelectModalComponent_c8y_filter_10_Template_c8y_filter_onSearch_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.updatePipe($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("icon", "search");
} }
function SelectModalComponent_div_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵelementStart(1, "div", 17);
    ɵngcc0.ɵɵelement(2, "h1", 18);
    ɵngcc0.ɵɵelementStart(3, "p", 19);
    ɵngcc0.ɵɵtext(4, "No items to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("c8yIcon", "", ctx_r1.icon, " ");
} }
function SelectModalComponent_c8y_li_15_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const bodyPart_r11 = ctx.$implicit;
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", bodyPart_r11.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", bodyPart_r11.value)("pattern", ctx_r8.filterTerm)("text", bodyPart_r11.value);
} }
function SelectModalComponent_c8y_li_15_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelementStart(1, "div", 24);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", item_r7.additionalInformation.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", item_r7.additionalInformation.value, " ");
} }
function SelectModalComponent_c8y_li_15_c8y_li_9_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 24);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 29);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const optionPart_r15 = ctx.$implicit;
    const ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngClass", optionPart_r15.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("pattern", ctx_r14.filterTerm)("text", optionPart_r15.value);
} }
function SelectModalComponent_c8y_li_15_c8y_li_9_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-radio", 27);
    ɵngcc0.ɵɵlistener("onSelect", function SelectModalComponent_c8y_li_15_c8y_li_9_Template_c8y_li_radio_onSelect_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r19); const option_r13 = restoredCtx.$implicit; const item_r7 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r17 = ɵngcc0.ɵɵnextContext(); return ctx_r17.updateChoice({ item: item_r7, id: option_r13.obj.id }); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(2, "c8y-li-body", 28);
    ɵngcc0.ɵɵtemplate(3, SelectModalComponent_c8y_li_15_c8y_li_9_div_3_Template, 2, 3, "div", 22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const option_r13 = ctx.$implicit;
    const item_r7 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r10 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", ctx_r10.mode === "single" ? "single" : item_r7.groupId)("disabled", option_r13.disabled)("selected", option_r13.selected);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", option_r13.body);
} }
function SelectModalComponent_c8y_li_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 20);
    ɵngcc0.ɵɵelementStart(4, "div", 21);
    ɵngcc0.ɵɵtemplate(5, SelectModalComponent_c8y_li_15_div_5_Template, 2, 4, "div", 22);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(6, SelectModalComponent_c8y_li_15_div_6_Template, 3, 2, "div", 23);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "c8y-li-collapse");
    ɵngcc0.ɵɵelementStart(8, "c8y-list-group");
    ɵngcc0.ɵɵtemplate(9, SelectModalComponent_c8y_li_15_c8y_li_9_Template, 4, 4, "c8y-li", 10);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r7 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", ctx_r2.icon);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", item_r7.body);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r7.additionalInformation);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngForOf", item_r7.options);
} }
function SelectModalComponent_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 30);
    ɵngcc0.ɵɵtext(2, " Some entries might not be shown. Try narrowing search criteria. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function SelectModalComponent_button_19_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 31);
    ɵngcc0.ɵɵlistener("click", function SelectModalComponent_button_19_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.dismiss(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r4.labels.cancel));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r4.labels.cancel), " ");
} }
export class SelectModalComponent {
    constructor(bsModalRef) {
        this.bsModalRef = bsModalRef;
        this.subTitle = gettext('Select from the list of items matching the device type');
        this.mode = ModalSelectionMode.MULTI;
        this.disableSelected = true;
        this.showFilter = true;
        this.areMoreEntries = false;
        this.result = new EventEmitter();
        this.search = new EventEmitter();
        this.selected = false;
        this.filterTerm = '';
        this.listItems = [];
        this.debouncer = new Subject();
        this._labels = { ok: gettext('Confirm'), cancel: gettext('Cancel') };
        this.debouncer.pipe(debounceTime(500)).subscribe(value => {
            this.search.emit(value);
        });
    }
    set labels(labels) {
        const { ok = this.labels.ok, cancel = this.labels.cancel } = labels || {};
        this._labels = { ok, cancel };
    }
    get labels() {
        return this._labels;
    }
    ngOnChanges(changes) {
        return __awaiter(this, void 0, void 0, function* () {
            if (changes.items && changes.items.currentValue) {
                const itemsPromise = changes.items.currentValue.map((item) => __awaiter(this, void 0, void 0, function* () {
                    item.options = yield item.options;
                    const selected = item.options.find(option => option.selected);
                    if (selected) {
                        item.selectedId = selected.obj.id;
                        if (this.disableSelected) {
                            item.options.map(option => assign(option, { disabled: true }));
                        }
                    }
                    return item;
                }));
                this.listItems = yield Promise.all(itemsPromise);
            }
        });
    }
    updatePipe(filterTerm) {
        this.debouncer.next(filterTerm);
        this.filterTerm = filterTerm;
    }
    updateChoice({ item, id }) {
        if (this.mode === 'single') {
            this.listItems.map(value => (value.selectedId = undefined));
        }
        item.selectedId = id;
        this.selected = true;
    }
    dismiss() {
        this.bsModalRef.hide();
    }
    select() {
        this.result.emit(this.getOutput());
        this.bsModalRef.hide();
    }
    ngOnDestroy() {
        this.debouncer.complete();
        this.result.complete();
        this.search.complete();
    }
    getOutput() {
        return this.listItems
            .filter(item => item.selectedId)
            .map(item => item.options.find(option => item.selectedId === option.obj.id))
            .filter(option => !option.selected)
            .map(selectedOption => selectedOption.obj);
    }
}
SelectModalComponent.ɵfac = function SelectModalComponent_Factory(t) { return new (t || SelectModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef)); };
SelectModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SelectModalComponent, selectors: [["c8y-select-modal"]], inputs: { subTitle: "subTitle", mode: "mode", disableSelected: "disableSelected", showFilter: "showFilter", areMoreEntries: "areMoreEntries", labels: "labels", icon: "icon", title: "title", items: "items" }, outputs: { result: "result", search: "search" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 24, vars: 22, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [3, "c8yIcon"], [1, "text-uppercase"], [1, "p-16", "text-center", "separator-bottom"], [1, "m-b-8"], [3, "icon", "onSearch", 4, "ngIf"], [1, "modal-inner-scroll"], [1, "p-l-16", "p-r-16"], ["class", "panel m-t-8 m-b-8", 4, "ngIf"], [4, "ngFor", "ngForOf"], [4, "ngIf"], [1, "modal-footer"], ["class", "btn btn-default", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-primary", 3, "title", "disabled", "click"], [3, "icon", "onSearch"], [1, "panel", "m-t-8", "m-b-8"], [1, "c8y-empty-state", "text-left"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], ["translate", ""], [1, "content-flex-30"], [1, "col-9"], [3, "ngClass", 4, "ngFor", "ngForOf"], ["class", "col-3 text-right", 4, "ngIf"], [3, "ngClass"], [3, "title", "pattern", "text"], [1, "col-3", "text-right"], [3, "name", "disabled", "selected", "onSelect"], [1, "content-flex-50"], [3, "pattern", "text"], ["translate", "", 1, "alert", "alert-info", "m-t-16", "m-r-8", "m-l-8"], [1, "btn", "btn-default", 3, "title", "click"]], template: function SelectModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "span", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "div", 4);
        ɵngcc0.ɵɵelementStart(7, "p", 5);
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(10, SelectModalComponent_c8y_filter_10_Template, 1, 1, "c8y-filter", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵelementStart(12, "div", 8);
        ɵngcc0.ɵɵtemplate(13, SelectModalComponent_div_13_Template, 5, 1, "div", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "c8y-list-group");
        ɵngcc0.ɵɵtemplate(15, SelectModalComponent_c8y_li_15_Template, 10, 4, "c8y-li", 10);
        ɵngcc0.ɵɵpipe(16, "selectModalFilterPipe");
        ɵngcc0.ɵɵtemplate(17, SelectModalComponent_div_17_Template, 3, 0, "div", 11);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 12);
        ɵngcc0.ɵɵtemplate(19, SelectModalComponent_button_19_Template, 4, 6, "button", 13);
        ɵngcc0.ɵɵelementStart(20, "button", 14);
        ɵngcc0.ɵɵlistener("click", function SelectModalComponent_Template_button_click_20_listener() { return ctx.select(); });
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵtext(22);
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("c8yIcon", ctx.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 11, ctx.title), " ");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(9, 13, ctx.subTitle));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showFilter);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.items || ctx.items.length === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind2(16, 15, ctx.listItems, ctx.filterTerm));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.areMoreEntries);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.labels.cancel);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(21, 18, ctx.labels.ok));
        ɵngcc0.ɵɵproperty("disabled", !ctx.selected);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(23, 20, ctx.labels.ok), " ");
    } }, directives: [ɵngcc2.IconDirective, ɵngcc3.NgIf, ɵngcc4.ListGroupComponent, ɵngcc3.NgForOf, ɵngcc5.FilterInputComponent, ɵngcc6.C8yTranslateDirective, ɵngcc7.ListItemComponent, ɵngcc8.ListItemIconComponent, ɵngcc9.ListItemBodyComponent, ɵngcc10.ListItemCollapseComponent, ɵngcc3.NgClass, ɵngcc11.HighlightComponent, ɵngcc12.ListItemRadioComponent], pipes: [ɵngcc13.C8yTranslatePipe, ɵngcc14.SelectModalFilterPipe], encapsulation: 2 });
SelectModalComponent.ctorParameters = () => [
    { type: BsModalRef }
];
SelectModalComponent.propDecorators = {
    icon: [{ type: Input }],
    title: [{ type: Input }],
    subTitle: [{ type: Input }],
    items: [{ type: Input }],
    mode: [{ type: Input }],
    disableSelected: [{ type: Input }],
    showFilter: [{ type: Input }],
    areMoreEntries: [{ type: Input }],
    labels: [{ type: Input }],
    result: [{ type: Output }],
    search: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-select-modal',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <span c8yIcon=\"{{ icon }}\"></span>\n    <h4 class=\"text-uppercase\">\n      {{ title | translate }}\n    </h4>\n  </div>\n  <div class=\"p-16 text-center separator-bottom\">\n    <p class=\"m-b-8\">{{ subTitle | translate }}</p>\n    <c8y-filter *ngIf=\"showFilter\" [icon]=\"'search'\" (onSearch)=\"updatePipe($event)\"></c8y-filter>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"p-l-16 p-r-16\">\n      <div class=\"panel m-t-8 m-b-8\" *ngIf=\"!items || items.length === 0\">\n        <div class=\"c8y-empty-state text-left\">\n          <h1 c8yIcon=\"{{ icon }} \" class=\"c8y-icon-duocolor\"></h1>\n          <p translate>No items to display.</p>\n        </div>\n      </div>\n    </div>\n    <c8y-list-group>\n      <c8y-li *ngFor=\"let item of listItems | selectModalFilterPipe: filterTerm\">\n        <c8y-li-icon>\n          <i c8yIcon=\"{{ icon }}\"></i>\n        </c8y-li-icon>\n\n        <c8y-li-body class=\"content-flex-30\">\n          <div class=\"col-9\">\n            <div *ngFor=\"let bodyPart of item.body\" [ngClass]=\"bodyPart.class\">\n              <c8y-highlight\n                [title]=\"bodyPart.value\"\n                [pattern]=\"filterTerm\"\n                [text]=\"bodyPart.value\"\n              ></c8y-highlight>\n            </div>\n          </div>\n\n          <div class=\"col-3 text-right\" *ngIf=\"item.additionalInformation\">\n            <div [ngClass]=\"item.additionalInformation.class\">\n              {{ item.additionalInformation.value }}\n            </div>\n          </div>\n        </c8y-li-body>\n\n        <c8y-li-collapse>\n          <c8y-list-group>\n            <c8y-li *ngFor=\"let option of item.options\">\n              <c8y-li-radio\n                [name]=\"mode === 'single' ? 'single' : item.groupId\"\n                (onSelect)=\"updateChoice({ item: item, id: option.obj.id })\"\n                [disabled]=\"option.disabled\"\n                [selected]=\"option.selected\"\n              >\n              </c8y-li-radio>\n              <c8y-li-body class=\"content-flex-50\">\n                <div\n                  *ngFor=\"let optionPart of option.body; let i = index\"\n                  [ngClass]=\"optionPart.class\"\n                >\n                  <c8y-highlight [pattern]=\"filterTerm\" [text]=\"optionPart.value\"></c8y-highlight>\n                </div>\n              </c8y-li-body>\n            </c8y-li>\n          </c8y-list-group>\n        </c8y-li-collapse>\n      </c8y-li>\n      <div *ngIf=\"areMoreEntries\">\n        <div class=\"alert alert-info m-t-16 m-r-8 m-l-8\" translate>\n          Some entries might not be shown. Try narrowing search criteria.\n        </div>\n      </div>\n    </c8y-list-group>\n  </div>\n\n  <div class=\"modal-footer\">\n    <button\n      title=\"{{ labels.cancel | translate }}\"\n      *ngIf=\"labels.cancel\"\n      class=\"btn btn-default\"\n      (click)=\"dismiss()\"\n    >\n      {{ labels.cancel | translate }}\n    </button>\n    <button\n      title=\"{{ labels.ok | translate }}\"\n      class=\"btn btn-primary\"\n      (click)=\"select()\"\n      [disabled]=\"!selected\"\n    >\n      {{ labels.ok | translate }}\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }]; }, { subTitle: [{
            type: Input
        }], mode: [{
            type: Input
        }], disableSelected: [{
            type: Input
        }], showFilter: [{
            type: Input
        }], areMoreEntries: [{
            type: Input
        }], result: [{
            type: Output
        }], search: [{
            type: Output
        }], labels: [{
            type: Input
        }], icon: [{
            type: Input
        }], title: [{
            type: Input
        }], items: [{
            type: Input
        }] }); })();
