// @c8y/ngx-components/esm2015/ecosystem/application-properties/application-properties.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { ApplicationService, ApplicationType, BillingMode, InventoryService, Isolation } from '@c8y/client';
import { packageProperties } from '../ecosystem.constants';
import { EcosystemService } from '../ecosystem.service';
import { FormBuilder, Validators } from '@angular/forms';
import { AlertService, gettext, ViewContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '../ecosystem.service';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from '@c8y/client';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from 'ngx-bootstrap/tooltip';
import * as ɵngcc8 from '../shared/upload-archive.component';
import * as ɵngcc9 from '../activity-log/activity-log.component';

function ApplicationPropertiesComponent_c8y_breadcrumb_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-breadcrumb");
    ɵngcc0.ɵɵelement(1, "c8y-breadcrumb-item", 4);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "c8y-breadcrumb-item", 5);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelement(5, "c8y-breadcrumb-item", 5);
    ɵngcc0.ɵɵelement(6, "c8y-breadcrumb-item", 6);
    ɵngcc0.ɵɵpipe(7, "async");
    ɵngcc0.ɵɵpipe(8, "humanizeAppName");
    ɵngcc0.ɵɵelement(9, "c8y-breadcrumb-item", 6);
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(2, 10, "Ecosystem"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(4, 12, "Applications"))("path", "ecosystem/applications");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("icon", ctx_r0.breadcrumbConfig == null ? null : ctx_r0.breadcrumbConfig.icon)("label", ctx_r0.breadcrumbConfig == null ? null : ctx_r0.breadcrumbConfig.label)("path", ctx_r0.breadcrumbConfig == null ? null : ctx_r0.breadcrumbConfig.path);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(7, 14, ɵngcc0.ɵɵpipeBind1(8, 16, ctx_r0.application)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(10, 18, "Properties"));
} }
function ApplicationPropertiesComponent_c8y_breadcrumb_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-breadcrumb");
    ɵngcc0.ɵɵelement(1, "c8y-breadcrumb-item", 4);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "c8y-breadcrumb-item", 5);
    ɵngcc0.ɵɵelement(4, "c8y-breadcrumb-item", 6);
    ɵngcc0.ɵɵpipe(5, "async");
    ɵngcc0.ɵɵpipe(6, "humanizeAppName");
    ɵngcc0.ɵɵelement(7, "c8y-breadcrumb-item", 6);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(2, 7, "Ecosystem"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("icon", ctx_r1.breadcrumbConfig == null ? null : ctx_r1.breadcrumbConfig.icon)("label", ctx_r1.breadcrumbConfig == null ? null : ctx_r1.breadcrumbConfig.label)("path", ctx_r1.breadcrumbConfig == null ? null : ctx_r1.breadcrumbConfig.path);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(5, 9, ɵngcc0.ɵɵpipeBind1(6, 11, ctx_r1.application)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(8, 13, "Properties"));
} }
function ApplicationPropertiesComponent_div_8_div_2_c8y_app_icon_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-app-icon", 39);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("app", ctx_r5.application)("contextPath", ctx_r5.application.contextPath)("name", ctx_r5.application.name);
} }
function ApplicationPropertiesComponent_div_8_div_2_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 40);
} }
function ApplicationPropertiesComponent_div_8_div_2_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 41);
} }
function ApplicationPropertiesComponent_div_8_div_2_i_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 42);
} }
function ApplicationPropertiesComponent_div_8_div_2_i_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 43);
} }
function ApplicationPropertiesComponent_div_8_div_2_p_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵelementStart(1, "em", 44);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r10.application.description || ɵngcc0.ɵɵpipeBind1(3, 1, "No description available."), " ");
} }
const _c0 = function (a0) { return { updated: a0 }; };
function ApplicationPropertiesComponent_div_8_div_2_div_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 45);
    ɵngcc0.ɵɵelementStart(1, "label", 46);
    ɵngcc0.ɵɵelement(2, "textarea", 47);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c0, ctx_r11.formGroup == null ? null : ctx_r11.formGroup.controls == null ? null : ctx_r11.formGroup.controls.description == null ? null : ctx_r11.formGroup.controls.description.dirty));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(3, 2, "No description available."));
} }
function ApplicationPropertiesComponent_div_8_div_2_div_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 48);
    ɵngcc0.ɵɵelementStart(1, "label", 49);
    ɵngcc0.ɵɵtext(2, "Version:");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "small", 50);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ctx_r12.application.manifest == null ? null : ctx_r12.application.manifest.version);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 48);
    ɵngcc0.ɵɵelementStart(1, "label", 49);
    ɵngcc0.ɵɵtext(2, "Creation time:");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "small", 50);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(5, 1, ctx_r13.binaryMo == null ? null : ctx_r13.binaryMo.creationTime));
} }
function ApplicationPropertiesComponent_div_8_div_2_button_22_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 51);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_2_button_22_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r29); const ctx_r28 = ɵngcc0.ɵɵnextContext(3); ctx_r28.openApp(ctx_r28.application); return $event.stopPropagation(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 52);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r14 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 4, "Open"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r14.disableOpenInBrowser);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "external-link");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 6, "Open"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "small", 53);
    ɵngcc0.ɵɵtext(2, "The application is overwritten by a custom application sharing the same path");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_span_24_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r33 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 55);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_2_span_24_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r33); const ctx_r32 = ɵngcc0.ɵɵnextContext(4); return ctx_r32.subscribe(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 52);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Subscribe"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "check-circle-o");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 5, "Subscribe"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_span_24_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r35 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 55);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_2_span_24_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r35); const ctx_r34 = ɵngcc0.ɵɵnextContext(4); return ctx_r34.unsubscribe(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 52);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Unsubscribe"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "minus-circle");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 5, "Unsubscribe"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_span_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵtemplate(1, ApplicationPropertiesComponent_div_8_div_2_span_24_button_1_Template, 5, 7, "button", 54);
    ɵngcc0.ɵɵtemplate(2, ApplicationPropertiesComponent_div_8_div_2_span_24_button_2_Template, 5, 7, "button", 54);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r16 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r16.isSubscribed);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r16.isSubscribed);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_27_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 56);
    ɵngcc0.ɵɵelement(1, "c8y-properties-list", 57);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r17 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("data", ctx_r17.application.manifest)("properties", ctx_r17.packageProperties)("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Package details"));
} }
function ApplicationPropertiesComponent_div_8_div_2_div_28_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 56);
    ɵngcc0.ɵɵelement(1, "c8y-properties-list", 57);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r18 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("data", ctx_r18.package.manifest)("properties", ctx_r18.packageProperties)("title", ɵngcc0.ɵɵpipeBind1(2, 3, "Source package information"));
} }
function ApplicationPropertiesComponent_div_8_div_2_c8y_message_43_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_c8y_message_53_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_60_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "p", 59);
    ɵngcc0.ɵɵelement(2, "i", 52);
    ɵngcc0.ɵɵelementStart(3, "span");
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r21 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r21.iconMap[ctx_r21.application.type]);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 2, ctx_r21.application.type), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_62_c8y_message_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_62_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 60);
    ɵngcc0.ɵɵelementStart(6, "span", 61);
    ɵngcc0.ɵɵtext(7, "/apps/");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(8, "input", 62);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "c8y-messages");
    ɵngcc0.ɵɵtemplate(11, ApplicationPropertiesComponent_div_8_div_2_div_62_c8y_message_11_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r22 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 4, "Path"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(9, 6, "e.g. my-application`used in URL`"));
    ɵngcc0.ɵɵproperty("readOnly", ctx_r22.application.id || !ctx_r22.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r22.formGroup == null ? null : ctx_r22.formGroup.controls == null ? null : ctx_r22.formGroup.controls.contextPath == null ? null : ctx_r22.formGroup.controls.contextPath.errors == null ? null : ctx_r22.formGroup.controls.contextPath.errors.required);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_c8y_message_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_13_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 66);
    ɵngcc0.ɵɵelementStart(1, "label");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 59);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r38 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 2, "Version"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r38.application.manifest.version, " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelement(1, "i", 67);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-enterprise");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 2, "Single tenant"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelement(1, "i", 67);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yIcon", "c8y-sub-tenants");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 2, "Multi tenant"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 66);
    ɵngcc0.ɵɵelementStart(1, "label");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 59);
    ɵngcc0.ɵɵtemplate(5, ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_span_5_Template, 4, 4, "span", 0);
    ɵngcc0.ɵɵtemplate(6, ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_span_6_Template, 4, 4, "span", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r39 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 3, "Isolation"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r39.singleTenant);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r39.singleTenant);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 69);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("tooltip", ɵngcc0.ɵɵpipeBind1(1, 2, "Resources usage assigned to: Owner"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Subscription"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 69);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("tooltip", "Resources usage assigned to: Subscriber | translate");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 2, "Resources"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 69);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("tooltip", ɵngcc0.ɵɵpipeBind1(1, 2, "Resources usage assigned to: Owner"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Resources"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 66);
    ɵngcc0.ɵɵelementStart(1, "label");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 59);
    ɵngcc0.ɵɵtemplate(5, ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_5_Template, 4, 6, "span", 68);
    ɵngcc0.ɵɵtemplate(6, ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_6_Template, 3, 4, "span", 68);
    ɵngcc0.ɵɵtemplate(7, ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_span_7_Template, 4, 6, "span", 68);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r40 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 4, "Billing mode"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r40.subscription);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r40.subscription && ctx_r40.singleTenant);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r40.subscription && !ctx_r40.singleTenant);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 70);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Provider"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 72);
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 59);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r48 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 2, "Name"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r48.application.manifest.provider.name, " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 72);
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 59);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r49 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 2, "Domain"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r49.application.manifest.provider.domain, " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 72);
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "p", 59);
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r50 = ɵngcc0.ɵɵnextContext(5);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 2, "Support"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r50.application.manifest.provider.support, " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 71);
    ɵngcc0.ɵɵtemplate(1, ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_1_Template, 7, 4, "div", 0);
    ɵngcc0.ɵɵtemplate(2, ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_2_Template, 7, 4, "div", 0);
    ɵngcc0.ɵɵtemplate(3, ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_div_3_Template, 7, 4, "div", 0);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r42 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r42.application.manifest.provider.name);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r42.application.manifest.provider.domain);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r42.application.manifest.provider.support);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_63_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 60);
    ɵngcc0.ɵɵelementStart(6, "span", 61);
    ɵngcc0.ɵɵtext(7, "/service/");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(8, "input", 62);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "c8y-messages");
    ɵngcc0.ɵɵtemplate(11, ApplicationPropertiesComponent_div_8_div_2_div_63_c8y_message_11_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "div", 1);
    ɵngcc0.ɵɵtemplate(13, ApplicationPropertiesComponent_div_8_div_2_div_63_div_13_Template, 6, 4, "div", 63);
    ɵngcc0.ɵɵtemplate(14, ApplicationPropertiesComponent_div_8_div_2_div_63_div_14_Template, 7, 5, "div", 63);
    ɵngcc0.ɵɵtemplate(15, ApplicationPropertiesComponent_div_8_div_2_div_63_div_15_Template, 8, 6, "div", 63);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(16, ApplicationPropertiesComponent_div_8_div_2_div_63_div_16_Template, 3, 3, "div", 64);
    ɵngcc0.ɵɵtemplate(17, ApplicationPropertiesComponent_div_8_div_2_div_63_div_17_Template, 4, 3, "div", 65);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r23 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 9, "Path"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(9, 11, "e.g. my-microservice`used in URL`"));
    ɵngcc0.ɵɵproperty("readOnly", ctx_r23.application.id || !ctx_r23.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.formGroup == null ? null : ctx_r23.formGroup.controls == null ? null : ctx_r23.formGroup.controls.contextPath == null ? null : ctx_r23.formGroup.controls.contextPath.errors == null ? null : ctx_r23.formGroup.controls.contextPath.errors.required);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.application.manifest.version);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.application.manifest.isolation);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.application.manifest.isolation);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.application.manifest.provider);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r23.application.manifest.provider);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_64_c8y_message_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_64_c8y_message_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_64_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 60);
    ɵngcc0.ɵɵelementStart(6, "span", 61);
    ɵngcc0.ɵɵtext(7, " /apps/ ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(8, "input", 73);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "c8y-messages");
    ɵngcc0.ɵɵtemplate(11, ApplicationPropertiesComponent_div_8_div_2_div_64_c8y_message_11_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(12, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(13, "label");
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(16, "input", 74);
    ɵngcc0.ɵɵpipe(17, "translate");
    ɵngcc0.ɵɵelementStart(18, "c8y-messages");
    ɵngcc0.ɵɵtemplate(19, ApplicationPropertiesComponent_div_8_div_2_div_64_c8y_message_19_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(20, "div", 75);
    ɵngcc0.ɵɵelementStart(21, "label");
    ɵngcc0.ɵɵtext(22);
    ɵngcc0.ɵɵpipe(23, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(24, "input", 76);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 75);
    ɵngcc0.ɵɵelementStart(26, "label");
    ɵngcc0.ɵɵtext(27);
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(29, "input", 77);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r24 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 9, "Path"));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(9, 11, "e.g. my-application`used in URL`"));
    ɵngcc0.ɵɵproperty("readOnly", ctx_r24.application.id || !ctx_r24.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r24.formGroup == null ? null : ctx_r24.formGroup.controls == null ? null : ctx_r24.formGroup.controls.contextPath == null ? null : ctx_r24.formGroup.controls.contextPath.errors == null ? null : ctx_r24.formGroup.controls.contextPath.errors.required);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(15, 13, "Server URL"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(17, 15, "e.g. my-application`used in URL`"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r24.formGroup == null ? null : ctx_r24.formGroup.controls == null ? null : ctx_r24.formGroup.controls.resourcesUrl == null ? null : ctx_r24.formGroup.controls.resourcesUrl.errors == null ? null : ctx_r24.formGroup.controls.resourcesUrl.errors.required);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(23, 17, "Username"), "");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(28, 19, "Password"));
} }
function ApplicationPropertiesComponent_div_8_div_2_div_65_c8y_message_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_65_c8y_message_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 58);
    ɵngcc0.ɵɵtext(1, " Valid URL required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_2_div_65_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(2, "label");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "input", 78);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementStart(7, "c8y-messages");
    ɵngcc0.ɵɵtemplate(8, ApplicationPropertiesComponent_div_8_div_2_div_65_c8y_message_8_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵtemplate(9, ApplicationPropertiesComponent_div_8_div_2_div_65_c8y_message_9_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r25 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 6, "External URL"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate1("placeholder", "", ɵngcc0.ɵɵpipeBind1(6, 8, "e.g."), " http://www.example.com/");
    ɵngcc0.ɵɵproperty("pattern", "^(?!javascript:).+")("readOnly", !ctx_r25.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r25.formGroup == null ? null : ctx_r25.formGroup.controls == null ? null : ctx_r25.formGroup.controls.externalUrl == null ? null : ctx_r25.formGroup.controls.externalUrl.errors == null ? null : ctx_r25.formGroup.controls.externalUrl.errors.required);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r25.formGroup == null ? null : ctx_r25.formGroup.controls == null ? null : ctx_r25.formGroup.controls.externalUrl == null ? null : ctx_r25.formGroup.controls.externalUrl.errors == null ? null : ctx_r25.formGroup.controls.externalUrl.errors.pattern);
} }
function ApplicationPropertiesComponent_div_8_div_2_ng_container_66_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r57 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelementStart(1, "c8y-upload-archive", 79);
    ɵngcc0.ɵɵlistener("applicationChange", function ApplicationPropertiesComponent_div_8_div_2_ng_container_66_div_1_Template_c8y_upload_archive_applicationChange_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r57); const ctx_r56 = ɵngcc0.ɵɵnextContext(4); return ctx_r56.application = $event; })("refresh", function ApplicationPropertiesComponent_div_8_div_2_ng_container_66_div_1_Template_c8y_upload_archive_refresh_1_listener() { ɵngcc0.ɵɵrestoreView(_r57); const ctx_r58 = ɵngcc0.ɵɵnextContext(4); return ctx_r58.onNewArchive(); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r55 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("application", ctx_r55.application);
} }
function ApplicationPropertiesComponent_div_8_div_2_ng_container_66_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, ApplicationPropertiesComponent_div_8_div_2_ng_container_66_div_1_Template, 2, 1, "div", 9);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r26 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r26.isLoading);
} }
function ApplicationPropertiesComponent_div_8_div_2_div_67_button_5_Template(rf, ctx) { if (rf & 1) {
    const _r61 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 84);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_2_div_67_button_5_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r61); const ctx_r60 = ɵngcc0.ɵɵnextContext(4); return ctx_r60.delete(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Delete"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, "Delete"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_div_67_Template(rf, ctx) { if (rf & 1) {
    const _r63 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 80);
    ɵngcc0.ɵɵelementStart(1, "button", 81);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_2_div_67_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r63); const ctx_r62 = ɵngcc0.ɵɵnextContext(3); return ctx_r62.cancel(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, ApplicationPropertiesComponent_div_8_div_2_div_67_button_5_Template, 4, 6, "button", 82);
    ɵngcc0.ɵɵelementStart(6, "button", 83);
    ɵngcc0.ɵɵpipe(7, "translate");
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r27 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 6, "Cancel"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 8, "Cancel"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r27.canDelete);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 10, "Save"));
    ɵngcc0.ɵɵproperty("disabled", !ctx_r27.application.type || ctx_r27.formGroup.invalid || ctx_r27.formGroup.pristine);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 12, "Save"), " ");
} }
function ApplicationPropertiesComponent_div_8_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelementStart(1, "div", 12);
    ɵngcc0.ɵɵelementStart(2, "div", 13);
    ɵngcc0.ɵɵtemplate(3, ApplicationPropertiesComponent_div_8_div_2_c8y_app_icon_3_Template, 1, 3, "c8y-app-icon", 14);
    ɵngcc0.ɵɵtemplate(4, ApplicationPropertiesComponent_div_8_div_2_i_4_Template, 1, 0, "i", 15);
    ɵngcc0.ɵɵtemplate(5, ApplicationPropertiesComponent_div_8_div_2_i_5_Template, 1, 0, "i", 16);
    ɵngcc0.ɵɵtemplate(6, ApplicationPropertiesComponent_div_8_div_2_i_6_Template, 1, 0, "i", 17);
    ɵngcc0.ɵɵtemplate(7, ApplicationPropertiesComponent_div_8_div_2_i_7_Template, 1, 0, "i", 18);
    ɵngcc0.ɵɵelementStart(8, "div", 19);
    ɵngcc0.ɵɵelementStart(9, "h4", 20);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "async");
    ɵngcc0.ɵɵpipe(12, "humanizeAppName");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(13, ApplicationPropertiesComponent_div_8_div_2_p_13_Template, 4, 3, "p", 0);
    ɵngcc0.ɵɵtemplate(14, ApplicationPropertiesComponent_div_8_div_2_div_14_Template, 4, 6, "div", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(15, "div", 22);
    ɵngcc0.ɵɵelementStart(16, "span", 23);
    ɵngcc0.ɵɵtext(17);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(19, ApplicationPropertiesComponent_div_8_div_2_div_19_Template, 5, 1, "div", 24);
    ɵngcc0.ɵɵtemplate(20, ApplicationPropertiesComponent_div_8_div_2_div_20_Template, 6, 3, "div", 24);
    ɵngcc0.ɵɵelementStart(21, "div", 25);
    ɵngcc0.ɵɵtemplate(22, ApplicationPropertiesComponent_div_8_div_2_button_22_Template, 5, 8, "button", 26);
    ɵngcc0.ɵɵtemplate(23, ApplicationPropertiesComponent_div_8_div_2_div_23_Template, 3, 0, "div", 0);
    ɵngcc0.ɵɵtemplate(24, ApplicationPropertiesComponent_div_8_div_2_span_24_Template, 3, 2, "span", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(25, "div", 27);
    ɵngcc0.ɵɵelementStart(26, "div", 28);
    ɵngcc0.ɵɵtemplate(27, ApplicationPropertiesComponent_div_8_div_2_div_27_Template, 3, 5, "div", 29);
    ɵngcc0.ɵɵtemplate(28, ApplicationPropertiesComponent_div_8_div_2_div_28_Template, 3, 5, "div", 29);
    ɵngcc0.ɵɵelementStart(29, "div", 1);
    ɵngcc0.ɵɵelementStart(30, "div", 30);
    ɵngcc0.ɵɵelementStart(31, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(32, "label");
    ɵngcc0.ɵɵtext(33, "ID");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(34, "input", 31);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(35, "div", 32);
    ɵngcc0.ɵɵelementStart(36, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(37, "label");
    ɵngcc0.ɵɵtext(38);
    ɵngcc0.ɵɵpipe(39, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(40, "input", 33);
    ɵngcc0.ɵɵpipe(41, "translate");
    ɵngcc0.ɵɵelementStart(42, "c8y-messages");
    ɵngcc0.ɵɵtemplate(43, ApplicationPropertiesComponent_div_8_div_2_c8y_message_43_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(44, "div", 1);
    ɵngcc0.ɵɵelementStart(45, "div", 30);
    ɵngcc0.ɵɵelementStart(46, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(47, "label");
    ɵngcc0.ɵɵtext(48);
    ɵngcc0.ɵɵpipe(49, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(50, "input", 35);
    ɵngcc0.ɵɵpipe(51, "translate");
    ɵngcc0.ɵɵelementStart(52, "c8y-messages");
    ɵngcc0.ɵɵtemplate(53, ApplicationPropertiesComponent_div_8_div_2_c8y_message_53_Template, 2, 0, "c8y-message", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(54, "div", 32);
    ɵngcc0.ɵɵelementStart(55, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(56, "label");
    ɵngcc0.ɵɵtext(57);
    ɵngcc0.ɵɵpipe(58, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(59, "div");
    ɵngcc0.ɵɵtemplate(60, ApplicationPropertiesComponent_div_8_div_2_div_60_Template, 6, 4, "div", 0);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(61, "div", 36);
    ɵngcc0.ɵɵtemplate(62, ApplicationPropertiesComponent_div_8_div_2_div_62_Template, 12, 8, "div", 37);
    ɵngcc0.ɵɵtemplate(63, ApplicationPropertiesComponent_div_8_div_2_div_63_Template, 18, 13, "div", 37);
    ɵngcc0.ɵɵtemplate(64, ApplicationPropertiesComponent_div_8_div_2_div_64_Template, 30, 21, "div", 37);
    ɵngcc0.ɵɵtemplate(65, ApplicationPropertiesComponent_div_8_div_2_div_65_Template, 10, 10, "div", 37);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(66, ApplicationPropertiesComponent_div_8_div_2_ng_container_66_Template, 2, 1, "ng-container", 0);
    ɵngcc0.ɵɵtemplate(67, ApplicationPropertiesComponent_div_8_div_2_div_67_Template, 10, 14, "div", 38);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.isPackage && !ctx_r3.isFeature && !ctx_r3.isMicroservice && !ctx_r3.isExternal);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isPackage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isFeature);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isMicroservice);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isExternal);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 35, ɵngcc0.ɵɵpipeBind1(12, 37, ctx_r3.application)));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.isOwner);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isOwner);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r3.appState == null ? null : ctx_r3.appState.class);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(18, 39, ctx_r3.appState == null ? null : ctx_r3.appState.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", (ctx_r3.application.manifest == null ? null : ctx_r3.application.manifest.version) && (ctx_r3.isPackage || ctx_r3.isMicroservice));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.binaryMo == null ? null : ctx_r3.binaryMo.creationTime);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.canOpenInBrowser);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.canOpenInBrowser && ctx_r3.disableOpenInBrowser);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isCustomMicroservice);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isPackage);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.package);
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵproperty("readonly", true);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(39, 41, "Name"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(41, 43, "e.g. My application"));
    ɵngcc0.ɵɵproperty("readonly", !ctx_r3.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.formGroup == null ? null : ctx_r3.formGroup.controls == null ? null : ctx_r3.formGroup.controls.name == null ? null : ctx_r3.formGroup.controls.name.errors == null ? null : ctx_r3.formGroup.controls.name.errors.required);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(49, 45, "Application key"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(51, 47, "e.g. my-application-key"));
    ɵngcc0.ɵɵproperty("readonly", ctx_r3.application.id || !ctx_r3.isOwner);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.formGroup == null ? null : ctx_r3.formGroup.controls == null ? null : ctx_r3.formGroup.controls.key == null ? null : ctx_r3.formGroup.controls.key.errors == null ? null : ctx_r3.formGroup.controls.key.errors.required);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(58, 49, "Type"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.application.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitch", ctx_r3.application.type);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "HOSTED");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "MICROSERVICE");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "REPOSITORY");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngSwitchCase", "EXTERNAL");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.isCustomMicroservice);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.application && !!ctx_r3.isOwner);
} }
function ApplicationPropertiesComponent_div_8_div_3_div_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 93);
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesComponent_div_8_div_3_c8y_activity_log_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-activity-log", 94);
} if (rf & 2) {
    const ctx_r65 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("application", ctx_r65.application);
} }
const _c1 = function (a0) { return { "icon-spin": a0 }; };
function ApplicationPropertiesComponent_div_8_div_3_Template(rf, ctx) { if (rf & 1) {
    const _r67 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 85);
    ɵngcc0.ɵɵelementStart(1, "div", 86);
    ɵngcc0.ɵɵelementStart(2, "h4", 87);
    ɵngcc0.ɵɵtext(3, "Activity log");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 88);
    ɵngcc0.ɵɵelementStart(5, "button", 89);
    ɵngcc0.ɵɵlistener("click", function ApplicationPropertiesComponent_div_8_div_3_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r67); const ctx_r66 = ɵngcc0.ɵɵnextContext(2); return ctx_r66.load(); });
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelement(7, "i", 90);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(10, ApplicationPropertiesComponent_div_8_div_3_div_10_Template, 2, 0, "div", 91);
    ɵngcc0.ɵɵtemplate(11, ApplicationPropertiesComponent_div_8_div_3_c8y_activity_log_11_Template, 1, 1, "c8y-activity-log", 92);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 5, "Reload"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(9, _c1, ctx_r4.isLoading));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 7, "Reload"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.isLoading);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.isLoading);
} }
const _c2 = function (a0) { return { "d-grid grid__col--7-5--md": a0 }; };
function ApplicationPropertiesComponent_div_8_Template(rf, ctx) { if (rf & 1) {
    const _r69 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵelementStart(1, "form", 8);
    ɵngcc0.ɵɵlistener("ngSubmit", function ApplicationPropertiesComponent_div_8_Template_form_ngSubmit_1_listener() { ɵngcc0.ɵɵrestoreView(_r69); const ctx_r68 = ɵngcc0.ɵɵnextContext(); return ctx_r68.formGroup.valid && ctx_r68.save(ctx_r68.formGroup.value); });
    ɵngcc0.ɵɵtemplate(2, ApplicationPropertiesComponent_div_8_div_2_Template, 68, 51, "div", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, ApplicationPropertiesComponent_div_8_div_3_Template, 12, 11, "div", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c2, ctx_r2.isActivityLogSupported));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("formGroup", ctx_r2.formGroup);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.isLoading);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.isActivityLogSupported);
} }
const _c3 = function (a0, a1) { return { "col-md-8": a0, "col-md-12": a1 }; };
const MICROSERVICES_BASE_PATH = '/microservices';
const APPLICATIONS_BASE_PATH = '/ecosystem';
export class ApplicationPropertiesComponent {
    constructor(activatedRoute, ecosystemService, router, formBuilder, applicationService, alertService, inventoryService) {
        this.activatedRoute = activatedRoute;
        this.ecosystemService = ecosystemService;
        this.router = router;
        this.formBuilder = formBuilder;
        this.applicationService = applicationService;
        this.alertService = alertService;
        this.inventoryService = inventoryService;
        this.singleTenant = false;
        this.subscription = false;
        this.iconMap = {
            HOSTED: 'cloud',
            EXTERNAL: 'external-link-square',
            MICROSERVICE: 'microchip'
        };
        this.isLoading = true;
        this.packageProperties = [...packageProperties];
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.load();
            this.isUnpacked = this.ecosystemService.isUnpacked(this.application);
            this.isPackage = this.ecosystemService.isPackage(this.application);
            this.isFeature = this.ecosystemService.isFeature(this.application);
            this.isExternal = this.ecosystemService.isExternal(this.application);
            this.isMicroservice = this.ecosystemService.isMicroservice(this.application);
            this.appState = this.ecosystemService.getAppState(this.application);
            if (this.isUnpacked) {
                yield this.resolveSourcePackage();
            }
            this.setBreadcrumbConifg();
            if (this.isCustomMicroservice) {
                this.loadBinaryMo();
            }
        });
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            this.isLoading = true;
            this.initForm();
            yield this.loadApplication();
            this.isLoading = false;
        });
    }
    onApplication(app) {
        if (app.manifest) {
            this.singleTenant = app.manifest.isolation === Isolation.PER_TENANT;
            this.subscription = app.manifest.billingMode === BillingMode.SUBSCRIPTION;
        }
    }
    cancel() {
        if (this.application.type === ApplicationType.MICROSERVICE) {
            this.router.navigateByUrl(MICROSERVICES_BASE_PATH);
        }
        else {
            this.router.navigateByUrl(APPLICATIONS_BASE_PATH);
        }
    }
    openApp(app) {
        window.open(this.applicationService.getHref(app));
    }
    getPackage(entityOrId) {
        return this.applicationService.detail(entityOrId);
    }
    delete() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.ecosystemService.deleteApp(this.application);
                if (this.application.type === ApplicationType.MICROSERVICE) {
                    this.router.navigateByUrl(MICROSERVICES_BASE_PATH);
                }
                else {
                    this.router.navigateByUrl(APPLICATIONS_BASE_PATH);
                }
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
    subscribe() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.subscribeApp(this.application);
            this.loadApplication();
        });
    }
    unsubscribe() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.unsubscribeApp(this.application);
            this.loadApplication();
        });
    }
    loadApplication() {
        return __awaiter(this, void 0, void 0, function* () {
            const { id } = this.activatedRoute.snapshot.parent.data.contextData;
            this.application = yield this.ecosystemService.getApplication(id);
            if (this.application.type === ApplicationType.MICROSERVICE) {
                this.formGroup.get('name').disable();
            }
            this.formGroup.patchValue(Object.assign({}, this.application));
            this.canOpenInBrowser = this.ecosystemService.canOpenAppInBrowser(this.application);
            this.disableOpenInBrowser =
                this.canOpenInBrowser &&
                    (yield this.ecosystemService.isOverwrittenByCustomApp(this.application));
            this.canDelete = this.ecosystemService.canDeleteApp(this.application);
            this.isOwner = this.ecosystemService.isOwner(this.application);
            this.isActivityLogSupported = this.isActivityLogSupportedByApp(this.application);
            this.isCustomMicroservice = this.ecosystemService.isCustomMicroservice(this.application);
            this.isSubscribed = yield this.ecosystemService.checkIfSubscribed(this.application);
            this.onApplication(this.application);
        });
    }
    save(app) {
        return __awaiter(this, void 0, void 0, function* () {
            app.id = this.application.id;
            try {
                yield this.ecosystemService.updateApp(app);
                this.alertService.success(gettext('Application saved.'));
                this.formGroup.reset();
                this.loadApplication();
            }
            catch (ex) {
                // do nothing
            }
        });
    }
    onNewArchive() {
        this.loadBinaryMo();
    }
    resolveSourcePackage() {
        return __awaiter(this, void 0, void 0, function* () {
            this.package = (yield this.getPackage(this.application.manifest.source)).data;
            if (this.package) {
                this.packageProperties.push({
                    label: gettext('Source package'),
                    value: this.package.name,
                    type: 'link',
                    action: () => this.router.navigateByUrl(ViewContext.Application.replace(':id', this.package.id.toString()))
                });
            }
        });
    }
    initForm() {
        this.formGroup = this.formBuilder.group({
            id: [{ value: '' }],
            name: [Validators.required, Validators.maxLength(120)],
            key: [Validators.required, Validators.maxLength(120)],
            contextPath: [Validators.required, Validators.maxLength(120)],
            description: ['', Validators.maxLength(200)],
            resourcesUrl: [Validators.required],
            username: [Validators.required],
            password: [Validators.required],
            externalUrl: [Validators.required]
        });
    }
    isActivityLogSupportedByApp(app) {
        return (this.ecosystemService.isOwner(app) &&
            app.type !== ApplicationType.MICROSERVICE &&
            app.type !== ApplicationType.EXTERNAL);
    }
    setBreadcrumbConifg() {
        this.breadcrumbConfig = {
            icon: this.isMicroservice ? 'microchip' : this.isFeature ? 'tab' : 'c8y-modules',
            label: this.isMicroservice
                ? gettext('Microservices')
                : this.isFeature
                    ? gettext('Features')
                    : gettext('All applications'),
            path: this.isMicroservice
                ? 'microservices'
                : this.isFeature
                    ? 'ecosystem/features'
                    : 'ecosystem/applications'
        };
    }
    loadBinaryMo() {
        return __awaiter(this, void 0, void 0, function* () {
            this.binaryMo = (yield this.inventoryService.detail(this.application.activeVersionId)).data;
        });
    }
}
ApplicationPropertiesComponent.ɵfac = function ApplicationPropertiesComponent_Factory(t) { return new (t || ApplicationPropertiesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.FormBuilder), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.InventoryService)); };
ApplicationPropertiesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ApplicationPropertiesComponent, selectors: [["c8y-application-properties"]], decls: 9, vars: 12, consts: [[4, "ngIf"], [1, "row"], [3, "ngClass"], ["class", "card content-fullpage", 3, "ngClass", 4, "ngIf"], [3, "icon", "label"], [3, "icon", "label", "path"], [3, "label"], [1, "card", "content-fullpage", 3, "ngClass"], ["novalidate", "", 1, "d-flex", "d-col", "content-fullpage", 3, "formGroup", "ngSubmit"], ["class", "d-contents", 4, "ngIf"], ["class", "content-fullpage d-flex d-col bg-gray-white", 4, "ngIf"], [1, "d-contents"], [1, "card-block", "separator-bottom", "large-padding", "flex-no-shrink"], [1, "d-flex-md", "a-i-start", "text-center", "text-left-md"], ["class", "icon-48", 3, "app", "contextPath", "name", 4, "ngIf"], ["c8yIcon", "big-parcel", "class", "icon-48", 4, "ngIf"], ["c8yIcon", "tab", "class", "icon-48", 4, "ngIf"], ["c8yIcon", "microchip", "class", "icon-48", 4, "ngIf"], ["c8yIcon", "globe1", "class", "icon-48", 4, "ngIf"], [1, "p-t-md-16", "p-l-md-16", "p-r-md-32", "flex-grow"], [1, "text-medium", "m-b-8"], ["class", "form-group m-b-0", 4, "ngIf"], [1, "text-right-md", "m-t-4"], [1, "label", 3, "ngClass"], ["class", "fit-w m-t-2", 4, "ngIf"], [1, "m-t-8"], ["class", "btn btn-default btn-xs", 3, "disabled", "title", "click", 4, "ngIf"], [1, "inner-scroll", "bg-white", "flex-grow"], [1, "card-block", "large-padding"], ["class", "row p-16", 4, "ngIf"], [1, "col-sm-5"], ["autocomplete", "off", "formControlName", "id", "name", "id", "type", "text", 1, "form-control", 3, "readonly"], [1, "col-sm-7"], ["formControlName", "name", "name", "name", "required", "", "type", "text", 1, "form-control", 3, "readonly", "placeholder"], ["translate", "", 4, "ngIf"], ["name", "key", "formControlName", "key", "required", "", "type", "text", 1, "form-control", 3, "readonly", "placeholder"], [3, "ngSwitch"], [4, "ngSwitchCase"], ["class", "card-footer separator", 4, "ngIf"], [1, "icon-48", 3, "app", "contextPath", "name"], ["c8yIcon", "big-parcel", 1, "icon-48"], ["c8yIcon", "tab", 1, "icon-48"], ["c8yIcon", "microchip", 1, "icon-48"], ["c8yIcon", "globe1", 1, "icon-48"], [1, "text-muted"], [1, "form-group", "m-b-0"], [1, "editable", 3, "ngClass"], ["c8y-textarea-autoresize", "", "formControlName", "description", "name", "description", 1, "form-control", "no-resize", 3, "placeholder"], [1, "fit-w", "m-t-2"], ["translate", "", 1, "text-label-small"], [1, "p-l-4", "text-bold"], [1, "btn", "btn-default", "btn-xs", 3, "disabled", "title", "click"], [3, "c8yIcon"], ["translate", "", 1, "text-muted"], ["class", "btn btn-default btn-xs", 3, "title", "click", 4, "ngIf"], [1, "btn", "btn-default", "btn-xs", 3, "title", "click"], [1, "row", "p-16"], ["icon", "info", 3, "data", "properties", "title"], ["translate", ""], [1, "form-control-static"], [1, "input-group"], [1, "input-group-addon"], ["name", "contextPath", "required", "", "type", "text", "formControlName", "contextPath", 1, "form-control", 3, "readOnly", "placeholder"], ["class", "col-sm-4 m-b-16 flex-auto", 4, "ngIf"], ["class", "legend form-block m-t-40", 4, "ngIf"], ["class", "list-inline", 4, "ngIf"], [1, "col-sm-4", "m-b-16", "flex-auto"], [1, "c8y-icon-duocolor", "h4", 3, "c8yIcon"], [3, "tooltip", 4, "ngIf"], [3, "tooltip"], [1, "legend", "form-block", "m-t-40"], [1, "list-inline"], [1, "col-sm-4", "m-b-16"], ["formControlName", "contextPath", "name", "contextPath", "required", "", "type", "text", 1, "form-control", 3, "readOnly", "placeholder"], ["name", "resourcesUrl", "formControlName", "resourcesUrl", "required", "", "type", "url", 1, "form-control", 3, "placeholder"], [1, "form-group"], ["formControlName", "username", "name", "username", "type", "text", 1, "form-control"], ["formControlName", "password", "name", "password", "type", "text", 1, "form-control"], ["formControlName", "externalUrl", "name", "externalUrl", "required", "", "type", "url", 1, "form-control", 3, "pattern", "readOnly", "placeholder"], [3, "application", "applicationChange", "refresh"], [1, "card-footer", "separator"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["class", "btn btn-danger", "type", "button", 3, "title", "click", 4, "ngIf"], ["type", "submit", 1, "btn", "btn-primary", "btn-form", 3, "disabled", "title"], ["type", "button", 1, "btn", "btn-danger", 3, "title", "click"], [1, "content-fullpage", "d-flex", "d-col", "bg-gray-white"], [1, "card-header", "separator"], ["translate", "", 1, "card-title"], [1, "m-l-auto"], ["type", "button", 1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["class", "p-16 text-center", 4, "ngIf"], ["class", "d-contents", 3, "application", 4, "ngIf"], [1, "p-16", "text-center"], [1, "d-contents", 3, "application"]], template: function ApplicationPropertiesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵpipe(3, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(4, ApplicationPropertiesComponent_c8y_breadcrumb_4_Template, 11, 20, "c8y-breadcrumb", 0);
        ɵngcc0.ɵɵtemplate(5, ApplicationPropertiesComponent_c8y_breadcrumb_5_Template, 9, 15, "c8y-breadcrumb", 0);
        ɵngcc0.ɵɵelementStart(6, "div", 1);
        ɵngcc0.ɵɵelementStart(7, "div", 2);
        ɵngcc0.ɵɵtemplate(8, ApplicationPropertiesComponent_div_8_Template, 4, 6, "div", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 5, ɵngcc0.ɵɵpipeBind1(3, 7, ctx.application)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isMicroservice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isMicroservice);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(9, _c3, !ctx.isActivityLogSupported, ctx.isActivityLogSupported));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.application);
    } }, directives: [ɵngcc5.TitleComponent, ɵngcc6.NgIf, ɵngcc6.NgClass, ɵngcc5.BreadcrumbComponent, ɵngcc5.BreadcrumbItemComponent, ɵngcc3.ɵNgNoValidate, ɵngcc3.NgControlStatusGroup, ɵngcc3.FormGroupDirective, ɵngcc5.FormGroupComponent, ɵngcc3.DefaultValueAccessor, ɵngcc3.NgControlStatus, ɵngcc3.FormControlName, ɵngcc5.RequiredInputPlaceholderDirective, ɵngcc3.RequiredValidator, ɵngcc5.MessagesComponent, ɵngcc6.NgSwitch, ɵngcc6.NgSwitchCase, ɵngcc5.AppIconComponent, ɵngcc5.IconDirective, ɵngcc5.ɵd, ɵngcc5.C8yTranslateDirective, ɵngcc5.PropertiesListComponent, ɵngcc5.MessageDirective, ɵngcc7.TooltipDirective, ɵngcc3.PatternValidator, ɵngcc8.UploadArchiveComponent, ɵngcc5.LoadingComponent, ɵngcc9.ActivityLogComponent], pipes: [ɵngcc6.AsyncPipe, ɵngcc5.HumanizeAppNamePipe, ɵngcc5.C8yTranslatePipe, ɵngcc5.DatePipe], encapsulation: 2 });
ApplicationPropertiesComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: EcosystemService },
    { type: Router },
    { type: FormBuilder },
    { type: ApplicationService },
    { type: AlertService },
    { type: InventoryService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApplicationPropertiesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-application-properties',
                template: "<c8y-title>{{ application | humanizeAppName | async }}</c8y-title>\n\n<c8y-breadcrumb *ngIf=\"!isMicroservice\">\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"breadcrumbConfig?.icon\"\n    [label]=\"breadcrumbConfig?.label\"\n    [path]=\"breadcrumbConfig?.path\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"application | humanizeAppName | async\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"'Properties' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-breadcrumb *ngIf=\"isMicroservice\">\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"breadcrumbConfig?.icon\"\n    [label]=\"breadcrumbConfig?.label\"\n    [path]=\"breadcrumbConfig?.path\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"application | humanizeAppName | async\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"'Properties' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<div class=\"row\">\n  <div [ngClass]=\"{ 'col-md-8': !isActivityLogSupported, 'col-md-12': isActivityLogSupported }\">\n    <div\n      *ngIf=\"application\"\n      [ngClass]=\"{ 'd-grid grid__col--7-5--md': isActivityLogSupported }\"\n      class=\"card content-fullpage\"\n    >\n      <form\n        (ngSubmit)=\"formGroup.valid && save(formGroup.value)\"\n        [formGroup]=\"formGroup\"\n        class=\"d-flex d-col content-fullpage\"\n        novalidate\n      >\n        <div *ngIf=\"!isLoading\" class=\"d-contents\">\n          <div class=\"card-block separator-bottom large-padding flex-no-shrink\">\n            <div class=\"d-flex-md a-i-start text-center text-left-md\">\n              <c8y-app-icon\n                *ngIf=\"!isPackage && !isFeature && !isMicroservice && !isExternal\"\n                [app]=\"application\"\n                [contextPath]=\"application.contextPath\"\n                [name]=\"application.name\"\n                class=\"icon-48\"\n              ></c8y-app-icon>\n              <i *ngIf=\"isPackage\" c8yIcon=\"big-parcel\" class=\"icon-48\"></i>\n              <i *ngIf=\"isFeature\" c8yIcon=\"tab\" class=\"icon-48\"></i>\n              <i *ngIf=\"isMicroservice\" c8yIcon=\"microchip\" class=\"icon-48\"></i>\n              <i *ngIf=\"isExternal\" c8yIcon=\"globe1\" class=\"icon-48\"></i>\n\n              <div class=\"p-t-md-16 p-l-md-16 p-r-md-32 flex-grow\">\n                <h4 class=\"text-medium m-b-8\">{{ application | humanizeAppName | async }}</h4>\n                <p *ngIf=\"!isOwner\">\n                  <em class=\"text-muted\">\n                    {{ application.description || ('No description available.' | translate) }}\n                  </em>\n                </p>\n                <div *ngIf=\"isOwner\" class=\"form-group m-b-0\">\n                  <label\n                    [ngClass]=\"{ updated: formGroup?.controls?.description?.dirty }\"\n                    class=\"editable\"\n                  >\n                    <textarea\n                      c8y-textarea-autoresize\n                      class=\"form-control no-resize\"\n                      formControlName=\"description\"\n                      name=\"description\"\n                      placeholder=\"{{ 'No description available.' | translate }}\"\n                    ></textarea>\n                  </label>\n                </div>\n              </div>\n              <div class=\"text-right-md m-t-4\">\n                <span [ngClass]=\"appState?.class\" class=\"label\">{{\n                  appState?.label | translate\n                }}</span>\n                <div\n                  *ngIf=\"application.manifest?.version && (isPackage || isMicroservice)\"\n                  class=\"fit-w m-t-2\"\n                >\n                  <label class=\"text-label-small\" translate>Version:</label>\n                  <small class=\"p-l-4 text-bold\">{{ application.manifest?.version }}</small>\n                </div>\n                <div *ngIf=\"binaryMo?.creationTime\" class=\"fit-w m-t-2\">\n                  <label class=\"text-label-small\" translate>Creation time:</label>\n                  <small class=\"p-l-4 text-bold\">{{ binaryMo?.creationTime | c8yDate }}</small>\n                </div>\n                <div class=\"m-t-8\">\n                  <button\n                    (click)=\"openApp(application); $event.stopPropagation()\"\n                    [disabled]=\"disableOpenInBrowser\"\n                    *ngIf=\"canOpenInBrowser\"\n                    class=\"btn btn-default btn-xs\"\n                    title=\"{{ 'Open' | translate }}\"\n                  >\n                    <i [c8yIcon]=\"'external-link'\"></i>\n                    {{ 'Open' | translate }}\n                  </button>\n                  <div *ngIf=\"canOpenInBrowser && disableOpenInBrowser\">\n                    <small class=\"text-muted\" translate\n                      >The application is overwritten by a custom application sharing the same\n                      path</small\n                    >\n                  </div>\n                  <span *ngIf=\"isCustomMicroservice\">\n                    <button\n                      (click)=\"subscribe()\"\n                      *ngIf=\"!isSubscribed\"\n                      class=\"btn btn-default btn-xs\"\n                      title=\"{{ 'Subscribe' | translate }}\"\n                    >\n                      <i [c8yIcon]=\"'check-circle-o'\"></i>\n                      {{ 'Subscribe' | translate }}\n                    </button>\n                    <button\n                      (click)=\"unsubscribe()\"\n                      *ngIf=\"isSubscribed\"\n                      class=\"btn btn-default btn-xs\"\n                      title=\"{{ 'Unsubscribe' | translate }}\"\n                    >\n                      <i [c8yIcon]=\"'minus-circle'\"></i>\n                      {{ 'Unsubscribe' | translate }}\n                    </button>\n                  </span>\n                </div>\n              </div>\n            </div>\n          </div>\n          <div class=\"inner-scroll bg-white flex-grow\">\n            <div class=\"card-block large-padding\">\n              <div *ngIf=\"isPackage\" class=\"row p-16\">\n                <c8y-properties-list\n                  [data]=\"application.manifest\"\n                  [properties]=\"packageProperties\"\n                  [title]=\"'Package details' | translate\"\n                  icon=\"info\"\n                ></c8y-properties-list>\n              </div>\n              <div *ngIf=\"package\" class=\"row p-16\">\n                <c8y-properties-list\n                  [data]=\"package.manifest\"\n                  [properties]=\"packageProperties\"\n                  [title]=\"'Source package information' | translate\"\n                  icon=\"info\"\n                ></c8y-properties-list>\n              </div>\n              <div class=\"row\">\n                <div class=\"col-sm-5\">\n                  <c8y-form-group>\n                    <label>ID</label>\n                    <input\n                      [readonly]=\"true\"\n                      autocomplete=\"off\"\n                      class=\"form-control\"\n                      formControlName=\"id\"\n                      name=\"id\"\n                      type=\"text\"\n                    />\n                  </c8y-form-group>\n                </div>\n                <div class=\"col-sm-7\">\n                  <c8y-form-group>\n                    <label>{{ 'Name' | translate }}</label>\n                    <input\n                      [readonly]=\"!isOwner\"\n                      class=\"form-control\"\n                      formControlName=\"name\"\n                      name=\"name\"\n                      placeholder=\"{{ 'e.g. My application' | translate }}\"\n                      required\n                      type=\"text\"\n                    />\n                    <c8y-messages>\n                      <c8y-message *ngIf=\"formGroup?.controls?.name?.errors?.required\" translate>\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n                </div>\n              </div>\n\n              <div class=\"row\">\n                <div class=\"col-sm-5\">\n                  <c8y-form-group>\n                    <label>{{ 'Application key' | translate }}</label>\n                    <input\n                      [readonly]=\"application.id || !isOwner\"\n                      class=\"form-control\"\n                      name=\"key\"\n                      formControlName=\"key\"\n                      placeholder=\"{{ 'e.g. my-application-key' | translate }}\"\n                      required\n                      type=\"text\"\n                    />\n                    <c8y-messages>\n                      <c8y-message *ngIf=\"formGroup?.controls?.key?.errors?.required\" translate>\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n                </div>\n\n                <div class=\"col-sm-7\">\n                  <c8y-form-group>\n                    <label>{{ 'Type' | translate }}</label>\n                    <div>\n                      <div *ngIf=\"application.id\">\n                        <p class=\"form-control-static\">\n                          <i [c8yIcon]=\"iconMap[application.type]\"></i>\n                          <span>\n                            {{ application.type | translate }}\n                          </span>\n                        </p>\n                      </div>\n                    </div>\n                  </c8y-form-group>\n                </div>\n              </div>\n\n              <div [ngSwitch]=\"application.type\">\n                <div *ngSwitchCase=\"'HOSTED'\">\n                  <c8y-form-group>\n                    <label>{{ 'Path' | translate }}</label>\n                    <div class=\"input-group\">\n                      <span class=\"input-group-addon\">/apps/</span>\n                      <input\n                        [readOnly]=\"application.id || !isOwner\"\n                        class=\"form-control\"\n                        name=\"contextPath\"\n                        placeholder=\"{{ 'e.g. my-application`used in URL`' | translate }}\"\n                        required\n                        type=\"text\"\n                        formControlName=\"contextPath\"\n                      />\n                    </div>\n                    <c8y-messages>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.contextPath?.errors?.required\"\n                        translate\n                      >\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n                </div>\n\n                <div *ngSwitchCase=\"'MICROSERVICE'\">\n                  <c8y-form-group>\n                    <label>{{ 'Path' | translate }}</label>\n                    <div class=\"input-group\">\n                      <span class=\"input-group-addon\">/service/</span>\n                      <input\n                        [readOnly]=\"application.id || !isOwner\"\n                        class=\"form-control\"\n                        name=\"contextPath\"\n                        placeholder=\"{{ 'e.g. my-microservice`used in URL`' | translate }}\"\n                        required\n                        type=\"text\"\n                        formControlName=\"contextPath\"\n                      />\n                    </div>\n                    <c8y-messages>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.contextPath?.errors?.required\"\n                        translate\n                      >\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n                  <div class=\"row\">\n                    <div *ngIf=\"application.manifest.version\" class=\"col-sm-4 m-b-16 flex-auto\">\n                      <label>{{ 'Version' | translate }}</label>\n                      <p class=\"form-control-static\">\n                        {{ application.manifest.version }}\n                      </p>\n                    </div>\n                    <div *ngIf=\"application.manifest.isolation\" class=\"col-sm-4 m-b-16 flex-auto\">\n                      <label>{{ 'Isolation' | translate }}</label>\n                      <p class=\"form-control-static\">\n                        <span *ngIf=\"singleTenant\">\n                          <i [c8yIcon]=\"'c8y-enterprise'\" class=\"c8y-icon-duocolor h4\"></i>\n                          {{ 'Single tenant' | translate }}\n                        </span>\n                        <span *ngIf=\"!singleTenant\">\n                          <i [c8yIcon]=\"'c8y-sub-tenants'\" class=\"c8y-icon-duocolor h4\"></i>\n                          {{ 'Multi tenant' | translate }}\n                        </span>\n                      </p>\n                    </div>\n                    <div *ngIf=\"application.manifest.isolation\" class=\"col-sm-4 m-b-16 flex-auto\">\n                      <label>{{ 'Billing mode' | translate }}</label>\n                      <p class=\"form-control-static\">\n                        <span\n                          *ngIf=\"subscription\"\n                          [tooltip]=\"'Resources usage assigned to: Owner' | translate\"\n                        >\n                          {{ 'Subscription' | translate }}\n                        </span>\n                        <span\n                          *ngIf=\"!subscription && singleTenant\"\n                          [tooltip]=\"'Resources usage assigned to: Subscriber | translate'\"\n                        >\n                          {{ 'Resources' | translate }}\n                        </span>\n                        <span\n                          *ngIf=\"!subscription && !singleTenant\"\n                          [tooltip]=\"'Resources usage assigned to: Owner' | translate\"\n                        >\n                          {{ 'Resources' | translate }}\n                        </span>\n                      </p>\n                    </div>\n                  </div>\n\n                  <div *ngIf=\"application.manifest.provider\" class=\"legend form-block m-t-40\">\n                    {{ 'Provider' | translate }}\n                  </div>\n                  <div class=\"list-inline\" *ngIf=\"application.manifest.provider\">\n                    <div *ngIf=\"application.manifest.provider.name\">\n                      <div class=\"col-sm-4 m-b-16\">\n                        <label>{{ 'Name' | translate }}</label>\n                        <p class=\"form-control-static\">\n                          {{ application.manifest.provider.name }}\n                        </p>\n                      </div>\n                    </div>\n                    <div *ngIf=\"application.manifest.provider.domain\">\n                      <div class=\"col-sm-4 m-b-16\">\n                        <label>{{ 'Domain' | translate }}</label>\n                        <p class=\"form-control-static\">\n                          {{ application.manifest.provider.domain }}\n                        </p>\n                      </div>\n                    </div>\n                    <div *ngIf=\"application.manifest.provider.support\">\n                      <div class=\"col-sm-4 m-b-16\">\n                        <label>{{ 'Support' | translate }}</label>\n                        <p class=\"form-control-static\">\n                          {{ application.manifest.provider.support }}\n                        </p>\n                      </div>\n                    </div>\n                  </div>\n                </div>\n\n                <div *ngSwitchCase=\"'REPOSITORY'\">\n                  <c8y-form-group>\n                    <label>{{ 'Path' | translate }}</label>\n                    <div class=\"input-group\">\n                      <span class=\"input-group-addon\"> /apps/ </span>\n                      <input\n                        [readOnly]=\"application.id || !isOwner\"\n                        class=\"form-control\"\n                        formControlName=\"contextPath\"\n                        name=\"contextPath\"\n                        placeholder=\"{{ 'e.g. my-application`used in URL`' | translate }}\"\n                        required\n                        type=\"text\"\n                      />\n                    </div>\n                    <c8y-messages>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.contextPath?.errors?.required\"\n                        translate\n                      >\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n\n                  <c8y-form-group>\n                    <label>{{ 'Server URL' | translate }}</label>\n                    <input\n                      class=\"form-control\"\n                      name=\"resourcesUrl\"\n                      formControlName=\"resourcesUrl\"\n                      placeholder=\"{{ 'e.g. my-application`used in URL`' | translate }}\"\n                      required\n                      type=\"url\"\n                    />\n                    <c8y-messages>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.resourcesUrl?.errors?.required\"\n                        translate\n                      >\n                        This field is required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n\n                  <div class=\"form-group\">\n                    <label> {{ 'Username' | translate }}</label>\n                    <input\n                      class=\"form-control\"\n                      formControlName=\"username\"\n                      name=\"username\"\n                      type=\"text\"\n                    />\n                  </div>\n\n                  <div class=\"form-group\">\n                    <label>{{ 'Password' | translate }}</label>\n                    <input\n                      class=\"form-control\"\n                      formControlName=\"password\"\n                      name=\"password\"\n                      type=\"text\"\n                    />\n                  </div>\n                </div>\n\n                <div *ngSwitchCase=\"'EXTERNAL'\">\n                  <c8y-form-group>\n                    <label>{{ 'External URL' | translate }}</label>\n                    <input\n                      [pattern]=\"'^(?!javascript:).+'\"\n                      [readOnly]=\"!isOwner\"\n                      class=\"form-control\"\n                      formControlName=\"externalUrl\"\n                      name=\"externalUrl\"\n                      placeholder=\"{{ 'e.g.' | translate }} http://www.example.com/\"\n                      required\n                      type=\"url\"\n                    />\n                    <c8y-messages>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.externalUrl?.errors?.required\"\n                        translate\n                      >\n                        This field is required.\n                      </c8y-message>\n                      <c8y-message\n                        *ngIf=\"formGroup?.controls?.externalUrl?.errors?.pattern\"\n                        translate\n                      >\n                        Valid URL required.\n                      </c8y-message>\n                    </c8y-messages>\n                  </c8y-form-group>\n                </div>\n              </div>\n            </div>\n          </div>\n          <ng-container *ngIf=\"isCustomMicroservice\">\n            <div *ngIf=\"!isLoading\" class=\"d-contents\">\n              <c8y-upload-archive\n                [(application)]=\"application\"\n                (refresh)=\"onNewArchive()\"\n              ></c8y-upload-archive>\n            </div>\n          </ng-container>\n          <div *ngIf=\"application && !!isOwner\" class=\"card-footer separator\">\n            <button\n              (click)=\"cancel()\"\n              class=\"btn btn-default\"\n              title=\"{{ 'Cancel' | translate }}\"\n              type=\"button\"\n            >\n              {{ 'Cancel' | translate }}\n            </button>\n            <button\n              (click)=\"delete()\"\n              *ngIf=\"canDelete\"\n              class=\"btn btn-danger\"\n              title=\"{{ 'Delete' | translate }}\"\n              type=\"button\"\n            >\n              {{ 'Delete' | translate }}\n            </button>\n            <button\n              [disabled]=\"!application.type || formGroup.invalid || formGroup.pristine\"\n              class=\"btn btn-primary btn-form\"\n              title=\"{{ 'Save' | translate }}\"\n              type=\"submit\"\n            >\n              {{ 'Save' | translate }}\n            </button>\n          </div>\n        </div>\n      </form>\n\n      <div *ngIf=\"isActivityLogSupported\" class=\"content-fullpage d-flex d-col bg-gray-white\">\n        <div class=\"card-header separator\">\n          <h4 class=\"card-title\" translate>Activity log</h4>\n          <div class=\"m-l-auto\">\n            <button\n              (click)=\"load()\"\n              class=\"btn btn-link\"\n              title=\"{{ 'Reload' | translate }}\"\n              type=\"button\"\n            >\n              <i [ngClass]=\"{ 'icon-spin': isLoading }\" c8yIcon=\"refresh\"></i>\n              {{ 'Reload' | translate }}\n            </button>\n          </div>\n        </div>\n        <div *ngIf=\"isLoading\" class=\"p-16 text-center\">\n          <c8y-loading></c8y-loading>\n        </div>\n        <c8y-activity-log\n          *ngIf=\"!isLoading\"\n          [application]=\"application\"\n          class=\"d-contents\"\n        ></c8y-activity-log>\n      </div>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.EcosystemService }, { type: ɵngcc1.Router }, { type: ɵngcc3.FormBuilder }, { type: ɵngcc4.ApplicationService }, { type: ɵngcc5.AlertService }, { type: ɵngcc4.InventoryService }]; }, null); })();
