// @c8y/ngx-components/esm2015/sub-assets/groups.component.js
import { Component, EventEmitter } from '@angular/core';
import { Permissions } from '@c8y/ngx-components';
import { SubAssetsService } from './sub-assets.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './sub-assets.service';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from './sub-assets-grid.component';
import * as ɵngcc5 from './add-group/add-group.component';

function GroupsComponent_c8y_add_group_11_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-add-group", 8);
    ɵngcc0.ɵɵlistener("onCancel", function GroupsComponent_c8y_add_group_11_Template_c8y_add_group_onCancel_0_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.showAddGroup = false; });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("refresh", ctx_r0.refresh);
} }
const _c0 = function (a0) { return { drawerOpen: a0 }; };
export class GroupsComponent {
    constructor(permissions, subAssetsService) {
        this.permissions = permissions;
        this.subAssetsService = subAssetsService;
        this.refresh = new EventEmitter();
        this.showAddGroup = false;
        this.filterable = false;
        this.sortable = true;
        this.shouldDisableAddGroup = false;
    }
    ngOnInit() {
        if (!this.permissions.hasRole('ROLE_INVENTORY_READ')) {
            this.sortable = false;
        }
        this.shouldDisableAddGroup = !this.subAssetsService.canCreateGroup();
    }
}
GroupsComponent.ɵfac = function GroupsComponent_Factory(t) { return new (t || GroupsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Permissions), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SubAssetsService)); };
GroupsComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: GroupsComponent, selectors: [["c8y-groups-name"]], decls: 13, vars: 15, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "disabled", "click"], ["c8yIcon", "plus-circle", 1, "m-r-4"], ["translate", ""], [3, "ngClass"], [1, "bottom-drawer"], [3, "refresh", "onCancel", 4, "ngIf"], [1, "content-fullpage", "d-flex", "d-col", 3, "refresh", "filterable", "sortable"], [3, "refresh", "onCancel"]], template: function GroupsComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(4, "button", 1);
        ɵngcc0.ɵɵlistener("click", function GroupsComponent_Template_button_click_4_listener() { return ctx.showAddGroup = !ctx.showAddGroup; });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "i", 2);
        ɵngcc0.ɵɵelementStart(7, "span", 3);
        ɵngcc0.ɵɵtext(8, "Add group");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "div", 4);
        ɵngcc0.ɵɵelementStart(10, "div", 5);
        ɵngcc0.ɵɵtemplate(11, GroupsComponent_c8y_add_group_11_Template, 1, 1, "c8y-add-group", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(12, "c8y-sub-assets-grid", 7);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 9, "Groups"), "\n");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 11, "Add group"));
        ɵngcc0.ɵɵproperty("disabled", ctx.shouldDisableAddGroup);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(13, _c0, ctx.showAddGroup));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showAddGroup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("refresh", ctx.refresh)("filterable", ctx.filterable)("sortable", ctx.sortable);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc1.ActionBarItemComponent, ɵngcc1.IconDirective, ɵngcc1.C8yTranslateDirective, ɵngcc3.NgClass, ɵngcc3.NgIf, ɵngcc4.SubAssetsGridComponent, ɵngcc5.AddGroupComponent], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
GroupsComponent.ctorParameters = () => [
    { type: Permissions },
    { type: SubAssetsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GroupsComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-groups-name',
                template: "<c8y-title>\n  {{ 'Groups' | translate }}\n</c8y-title>\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button\n    class=\"btn btn-link\"\n    title=\"{{ 'Add group' | translate }}\"\n    (click)=\"showAddGroup = !showAddGroup\"\n    [disabled]=\"shouldDisableAddGroup\"\n  >\n    <i c8yIcon=\"plus-circle\" class=\"m-r-4\"></i><span translate>Add group</span>\n  </button>\n</c8y-action-bar-item>\n <!-- TODO: uncomment when help is available \n  <c8y-help src=\"/users-guide/device-management/#grouping-devices\"></c8y-help> \n-->\n\n<div [ngClass]=\"{ drawerOpen: showAddGroup }\">\n  <div class=\"bottom-drawer\">\n    <c8y-add-group\n      [refresh]=\"refresh\"\n      (onCancel)=\"showAddGroup = false\"\n      *ngIf=\"showAddGroup\"\n    ></c8y-add-group>\n  </div>\n</div>\n<c8y-sub-assets-grid\n  [refresh]=\"refresh\"\n  [filterable]=\"filterable\"\n  [sortable]=\"sortable\"\n  class=\"content-fullpage d-flex d-col\"\n></c8y-sub-assets-grid>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: ɵngcc2.SubAssetsService }]; }, null); })();
