// @c8y/ngx-components/esm2015/core/dynamic-component/dynamic-component.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild, ComponentFactoryResolver, ViewContainerRef } from '@angular/core';
import { DynamicComponentService } from './dynamic-component.service';
import { of, isObservable } from 'rxjs';
import { isUndefined } from 'lodash-es';
/**
 * C8y dynamic component.
 * ## Example:
 *
 * register component in HOOK in module:
 * ```typescript
 *  import { HOOK_COMPONENT } from '@c8y/ngx-components';
 *
 * @NgModule({
 *  ...,
 *  providers: [{
 *      provide: HOOK_COMPONENT,
 *      multi: true,
 *      useValue: [{
 *          id: 'test-component',
 *          label: 'My test component',
 *          description: 'this is test component',
 *          component: TestComponent
 *      }],
 *  ...
 *  }]
 *
 * ```
 * Showing dynamic component:
 * ```html
 * <c8y-dynamic-component [componentId]="'test-component'" [config]="config"></c8y-dynamic-component>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dynamic-component.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from 'ngx-bootstrap/collapse';
import * as ɵngcc4 from '../i18n/c8y-translate.directive';
import * as ɵngcc5 from '../i18n/c8y-translate.pipe';

const _c0 = ["host"];
function DynamicComponentComponent_ng_template_0_Template(rf, ctx) { }
function DynamicComponentComponent_div_2_span_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1, "Show details");
    ɵngcc0.ɵɵelementEnd();
} }
function DynamicComponentComponent_div_2_span_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 9);
    ɵngcc0.ɵɵtext(1, "Hide details");
    ɵngcc0.ɵɵelementEnd();
} }
function DynamicComponentComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 2);
    ɵngcc0.ɵɵelementStart(1, "strong", 3);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 4);
    ɵngcc0.ɵɵelementStart(5, "button", 5);
    ɵngcc0.ɵɵlistener("click", function DynamicComponentComponent_div_2_Template_button_click_5_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.expandErrorDetails = !ctx_r5.expandErrorDetails; });
    ɵngcc0.ɵɵelement(6, "i", 6);
    ɵngcc0.ɵɵtemplate(7, DynamicComponentComponent_div_2_span_7_Template, 2, 0, "span", 7);
    ɵngcc0.ɵɵtemplate(8, DynamicComponentComponent_div_2_span_8_Template, 2, 0, "span", 7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 8);
    ɵngcc0.ɵɵelementStart(10, "pre");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate2(" ", ɵngcc0.ɵɵpipeBind1(3, 7, "This widget cannot be rendered because the current application does not support the following component:"), " ", ctx_r2.componentId, ". ");
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r2.expandErrorDetails);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.expandErrorDetails);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("collapse", !ctx_r2.expandErrorDetails)("isAnimated", true);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1("      ", ctx_r2.error, "\n    ");
} }
export class DynamicComponentComponent {
    /**
     * @ignore only DI
     */
    constructor(componentFactoryResolver, dynamicComponentService) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.dynamicComponentService = dynamicComponentService;
        /**
         * DynamicComponents can have two modes, an edit (config) and an view (component) mode.
         * By default it is shown in the component mode.
         */
        this.mode = 'component';
        /**
         * Disable this to hide the error that is shown if the component was not found.
         */
        this.notFoundError = true;
        /**
         * @ignore
         */
        this.expandErrorDetails = false;
    }
    /**
     * Calls the dynamic component life cycle hook. Currently only
     * supporting onBeforeSave, a hook which is called before a config component
     * is saved.
     */
    callLifeCycleHooks() {
        return this.callOnBeforeSaveHook();
    }
    /**
     * @ignore
     */
    ngOnChanges() {
        return __awaiter(this, void 0, void 0, function* () {
            const cmp = yield this.dynamicComponentService.getById(this.componentId);
            this.loadComponent(cmp);
        });
    }
    loadComponent(dynamicComponent) {
        try {
            this.error = undefined;
            const componentFactory = this.componentFactoryResolver.resolveComponentFactory(this.mode === 'component' ? dynamicComponent.component : dynamicComponent.configComponent);
            this.host.clear();
            this.componentRef = this.host.createComponent(componentFactory);
            this.componentRef.instance.config = this.config;
        }
        catch (ex) {
            this.error = ex;
        }
    }
    callOnBeforeSaveHook() {
        if (!this.componentRef) {
            return of(true);
        }
        const hook = this.componentRef.instance.onBeforeSave;
        if (hook) {
            const result = hook.call(this.componentRef.instance, this.config);
            if (isUndefined(result)) {
                return of(true);
            }
            return isObservable(result) ? result : of(result);
        }
        return of(true);
    }
}
DynamicComponentComponent.ɵfac = function DynamicComponentComponent_Factory(t) { return new (t || DynamicComponentComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DynamicComponentService)); };
DynamicComponentComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DynamicComponentComponent, selectors: [["c8y-dynamic-component"]], viewQuery: function DynamicComponentComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.host = _t.first);
    } }, inputs: { mode: "mode", notFoundError: "notFoundError", componentId: "componentId", config: "config" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 3, vars: 1, consts: [["host", ""], ["class", "alert alert-warning m-8", "role", "alert", 4, "ngIf"], ["role", "alert", 1, "alert", "alert-warning", "m-8"], [1, "message"], [1, "text-muted", "m-t-8"], [1, "btn", "btn-clean", 3, "click"], ["c8yIcon", "chevron-down"], ["translate", "", 4, "ngIf"], [3, "collapse", "isAnimated"], ["translate", ""]], template: function DynamicComponentComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DynamicComponentComponent_ng_template_0_Template, 0, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(2, DynamicComponentComponent_div_2_Template, 12, 9, "div", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.notFoundError && ctx.error);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.CollapseDirective, ɵngcc4.C8yTranslateDirective], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
DynamicComponentComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: DynamicComponentService }
];
DynamicComponentComponent.propDecorators = {
    componentId: [{ type: Input }],
    config: [{ type: Input }],
    mode: [{ type: Input }],
    notFoundError: [{ type: Input }],
    host: [{ type: ViewChild, args: ['host', { read: ViewContainerRef, static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicComponentComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-dynamic-component',
                template: "<ng-template #host></ng-template>\n\n<div class=\"alert alert-warning m-8\" role=\"alert\" *ngIf=\"notFoundError && error\">\n  <strong class=\"message\">\n    {{\n      'This widget cannot be rendered because the current application does not support the following component:'\n        | translate\n    }}\n    {{ componentId }}.\n  </strong>\n  <p class=\"text-muted m-t-8\">\n    <button class=\"btn btn-clean\" (click)=\"expandErrorDetails = !expandErrorDetails\">\n      <i c8yIcon=\"chevron-down\"></i>\n      <span *ngIf=\"!expandErrorDetails\" translate>Show details</span>\n      <span *ngIf=\"expandErrorDetails\" translate>Hide details</span>\n    </button>\n  </p>\n  <div [collapse]=\"!expandErrorDetails\" [isAnimated]=\"true\">\n    <pre>\n      {{ error }}\n    </pre>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc1.DynamicComponentService }]; }, { mode: [{
            type: Input
        }], notFoundError: [{
            type: Input
        }], componentId: [{
            type: Input
        }], config: [{
            type: Input
        }], host: [{
            type: ViewChild,
            args: ['host', { read: ViewContainerRef, static: true }]
        }] }); })();
