// @c8y/ngx-components/esm2015/protocol-impact/impact-protocol.module.js
import { NgModule } from '@angular/core';
import { CoreModule, HOOK_ACTION_BAR } from '@c8y/ngx-components';
import { ImpactActionFactory } from './impact-action.factory';
import { ImpactSubscriptionService } from './impact-subscription.service';
import { RefreshActionComponent } from './refresh-action.component';
import * as ɵngcc0 from '@angular/core';
export class ImpactProtocolModule {
}
ImpactProtocolModule.ɵfac = function ImpactProtocolModule_Factory(t) { return new (t || ImpactProtocolModule)(); };
ImpactProtocolModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ImpactProtocolModule });
ImpactProtocolModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ImpactSubscriptionService,
        { provide: HOOK_ACTION_BAR, useClass: ImpactActionFactory, multi: true }
    ], imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactProtocolModule, [{
        type: NgModule,
        args: [{
                declarations: [RefreshActionComponent],
                imports: [CoreModule],
                providers: [
                    ImpactSubscriptionService,
                    { provide: HOOK_ACTION_BAR, useClass: ImpactActionFactory, multi: true }
                ],
                entryComponents: [RefreshActionComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ImpactProtocolModule, { declarations: function () { return [RefreshActionComponent]; }, imports: function () { return [CoreModule]; } }); })();
