// @c8y/ngx-components/esm2015/protocol-impact/impact-action.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { get } from 'lodash-es';
import { RefreshActionComponent } from './refresh-action.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ImpactActionFactory {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    get(activeRoute) {
        return __awaiter(this, void 0, void 0, function* () {
            const actions = [];
            const data = !activeRoute.parent || activeRoute.snapshot.data.context
                ? activeRoute.snapshot.data
                : activeRoute.parent.snapshot.data;
            const { contextData } = data;
            const isDeviceInfoTab = get(activeRoute, 'snapshot.url[0].path') === 'device-info';
            const showRefreshActionButton = isDeviceInfoTab &&
                contextData &&
                contextData.c8y_ImpactResourceInfo &&
                /* call application service only for relevant devices to reduce number of service calls! */
                (yield this.applicationService.isAvailable(ImpactActionFactory.applicationName)).data;
            if (showRefreshActionButton) {
                actions.push({
                    priority: 500,
                    placement: 'right',
                    template: RefreshActionComponent
                });
            }
            return actions;
        });
    }
}
ImpactActionFactory.ɵfac = function ImpactActionFactory_Factory(t) { return new (t || ImpactActionFactory)(ɵngcc0.ɵɵinject(ɵngcc1.ApplicationService)); };
ImpactActionFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImpactActionFactory, factory: ImpactActionFactory.ɵfac });
ImpactActionFactory.applicationName = 'impact';
ImpactActionFactory.ctorParameters = () => [
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactActionFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ApplicationService }]; }, null); })();
