// @c8y/ngx-components/esm2015/protocol-impact/impact-subscription.service.js
import { Injectable } from '@angular/core';
import { FetchClient } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ImpactSubscriptionService {
    constructor(client) {
        this.client = client;
        this.microserviceUrl = '/service/impact';
        this.header = { 'Content-Type': 'application/json' };
    }
    refreshDeviceResources(deviceId) {
        const options = {
            method: 'PUT',
            headers: this.header
        };
        return this.client.fetch(`${this.microserviceUrl}/refresh/${deviceId}`, options);
    }
}
ImpactSubscriptionService.ɵfac = function ImpactSubscriptionService_Factory(t) { return new (t || ImpactSubscriptionService)(ɵngcc0.ɵɵinject(ɵngcc1.FetchClient)); };
ImpactSubscriptionService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImpactSubscriptionService, factory: ImpactSubscriptionService.ɵfac });
ImpactSubscriptionService.ctorParameters = () => [
    { type: FetchClient }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactSubscriptionService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.FetchClient }]; }, null); })();
