// @c8y/ngx-components/esm2015/core/search/highlight.component.js
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { escape, escapeRegExp } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["text"];
export class HighlightComponent {
    constructor() {
        this.pattern = '';
        this.text = '';
        this.elementClass = 'text-info';
        this.highlightedText = '';
    }
    ngOnChanges() {
        const textNoHtml = escape(this.text || '');
        const patternNoHtml = escape(this.pattern || '');
        const patternForRegex = escapeRegExp(patternNoHtml).replace(/\\\*/g, '.*');
        this.highlightedText = textNoHtml.replace(new RegExp(patternForRegex, 'i'), `<strong class="${this.elementClass}">$&</strong>`);
    }
}
HighlightComponent.ɵfac = function HighlightComponent_Factory(t) { return new (t || HighlightComponent)(); };
HighlightComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: HighlightComponent, selectors: [["c8y-highlight"]], viewQuery: function HighlightComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7, ElementRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
    } }, inputs: { pattern: "pattern", text: "text", elementClass: "elementClass" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 1, vars: 1, consts: [[3, "innerHTML"]], template: function HighlightComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "span", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("innerHTML", ctx.highlightedText, ɵngcc0.ɵɵsanitizeHtml);
    } }, encapsulation: 2 });
HighlightComponent.propDecorators = {
    container: [{ type: ViewChild, args: ['text', { read: ElementRef, static: true },] }],
    pattern: [{ type: Input }],
    text: [{ type: Input }],
    elementClass: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HighlightComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-highlight',
                template: '<span [innerHTML]="highlightedText"></span>'
            }]
    }], function () { return []; }, { pattern: [{
            type: Input
        }], text: [{
            type: Input
        }], elementClass: [{
            type: Input
        }], container: [{
            type: ViewChild,
            args: ['text', { read: ElementRef, static: true }]
        }] }); })();
