// @c8y/ngx-components/esm2015/core/search/search-result-empty.component.js
import { Component, Input } from '@angular/core';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/icon.directive';
import * as ɵngcc2 from '../i18n/c8y-translate.directive';
import * as ɵngcc3 from '../i18n/c8y-translate.pipe';
export class SearchResultEmptyComponent {
    constructor() {
        this.NO_RESULT_FOUND = gettext('No results found.');
        this.msg = this.NO_RESULT_FOUND;
    }
    set _message(msg) {
        if (msg && msg.length > 0) {
            this.msg = msg;
        }
        else {
            this.msg = this.NO_RESULT_FOUND;
        }
    }
    get message() {
        return this.msg;
    }
}
SearchResultEmptyComponent.ɵfac = function SearchResultEmptyComponent_Factory(t) { return new (t || SearchResultEmptyComponent)(); };
SearchResultEmptyComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SearchResultEmptyComponent, selectors: [["c8y-search-result-empty"]], inputs: { _message: ["message", "_message"] }, decls: 10, vars: 3, consts: [[1, "no-results"], [1, "c8y-empty-state"], ["c8yIcon", "search"], [1, "p-t-8"], ["translate", ""]], template: function SearchResultEmptyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "h1", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵelementStart(4, "p");
        ɵngcc0.ɵɵelementStart(5, "strong");
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "small", 4);
        ɵngcc0.ɵɵtext(9, "Refine your search terms.");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(7, 1, ctx.message));
    } }, directives: [ɵngcc1.IconDirective, ɵngcc2.C8yTranslateDirective], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
SearchResultEmptyComponent.propDecorators = {
    _message: [{ type: Input, args: ['message',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SearchResultEmptyComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-search-result-empty',
                template: "<div class=\"no-results\">\n  <div class=\"c8y-empty-state\">\n    <h1 c8yIcon=\"search\"></h1>\n    <div class=\"p-t-8\">\n      <p><strong>{{ message | translate }}</strong></p>\n      <small translate>Refine your search terms.</small>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return []; }, { _message: [{
            type: Input,
            args: ['message']
        }] }); })();
