// @c8y/ngx-components/esm2015/repository/software-device-tab/software-device-tab.guard.js
import { Injectable } from '@angular/core';
import { get, has } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
export class SoftwareDeviceTabGuard {
    constructor() {
        this.operations = ['c8y_SoftwareUpdate', 'c8y_SoftwareList', 'c8y_Software'];
    }
    canActivate(route) {
        const device = get(route, 'data.contextData') || get(route, 'parent.data.contextData');
        return (this.isAnyOperationSupported(device) ||
            has(device, 'c8y_Software') ||
            has(device, 'c8y_SoftwareList') ||
            has(device, 'c8y_SupportedSoftwareTypes'));
    }
    isAnyOperationSupported(device) {
        const supportedOperations = (device === null || device === void 0 ? void 0 : device.c8y_SupportedOperations) || [];
        return this.operations.some(operation => supportedOperations.indexOf(operation) >= 0);
    }
}
SoftwareDeviceTabGuard.ɵfac = function SoftwareDeviceTabGuard_Factory(t) { return new (t || SoftwareDeviceTabGuard)(); };
SoftwareDeviceTabGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SoftwareDeviceTabGuard, factory: SoftwareDeviceTabGuard.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SoftwareDeviceTabGuard, [{
        type: Injectable
    }], function () { return []; }, null); })();
