// @c8y/ngx-components/esm2015/core/dynamic-forms/object/object.type.component.js
import { Component } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '@ngx-formly/core';

function ObjectTypeComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("id", ctx_r0.id + "-fieldset");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.to.label);
} }
function ObjectTypeComponent_p_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.to.description);
} }
function ObjectTypeComponent_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 6);
    ɵngcc0.ɵɵelement(1, "formly-validation-message", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("field", ctx_r2.field);
} }
function ObjectTypeComponent_formly_field_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "formly-field", 7);
} if (rf & 2) {
    const f_r4 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("field", f_r4);
} }
export class ObjectTypeComponent extends FieldType {
}
ObjectTypeComponent.ɵfac = /*@__PURE__*/ function () { let ɵObjectTypeComponent_BaseFactory; return function ObjectTypeComponent_Factory(t) { return (ɵObjectTypeComponent_BaseFactory || (ɵObjectTypeComponent_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(ObjectTypeComponent)))(t || ObjectTypeComponent); }; }();
ObjectTypeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ObjectTypeComponent, selectors: [["c8y-object-type"]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 5, consts: [["role", "group"], ["class", "legend form-block", 3, "id", 4, "ngIf"], [4, "ngIf"], ["class", "form-control-feedback-message has-error", "role", "alert", 4, "ngIf"], [3, "field", 4, "ngFor", "ngForOf"], [1, "legend", "form-block", 3, "id"], ["role", "alert", 1, "form-control-feedback-message", "has-error"], [3, "field"]], template: function ObjectTypeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, ObjectTypeComponent_div_1_Template, 2, 2, "div", 1);
        ɵngcc0.ɵɵtemplate(2, ObjectTypeComponent_p_2_Template, 2, 1, "p", 2);
        ɵngcc0.ɵɵtemplate(3, ObjectTypeComponent_div_3_Template, 2, 1, "div", 3);
        ɵngcc0.ɵɵtemplate(4, ObjectTypeComponent_formly_field_4_Template, 1, 1, "formly-field", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-labelledby", ctx.id + "-fieldset");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showError && ctx.formControl.errors);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.field.fieldGroup);
    } }, directives: [ɵngcc1.NgIf, ɵngcc1.NgForOf, ɵngcc2.ɵc, ɵngcc2.FormlyField], encapsulation: 2 });
ObjectTypeComponent.CONFIG = {
    types: [{ name: 'object', component: ObjectTypeComponent }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ObjectTypeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-object-type',
                template: "<div role=\"group\"\n  [attr.aria-labelledby]=\"id + '-fieldset'\" \n>\n  <div id=\"{{id + '-fieldset'}}\" \n    class=\"legend form-block\" \n    *ngIf=\"to.label\"\n  >{{ to.label }}</div>\n  <p *ngIf=\"to.description\">{{ to.description }}</p>\n  <div class=\"form-control-feedback-message has-error\" \n    role=\"alert\" \n    *ngIf=\"showError && formControl.errors\"\n  >\n    <formly-validation-message [field]=\"field\"></formly-validation-message>\n  </div>\n  <formly-field *ngFor=\"let f of field.fieldGroup\" [field]=\"f\"></formly-field>\n</div>\n"
            }]
    }], null, null); })();
