// @c8y/ngx-components/esm2015/upgrade/action-bar.factory.js
import { Injectable } from '@angular/core';
import { BridgeService } from './bridge.service';
import { switchMap } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './bridge.service';
export class Ng1ActionBarFactoryService {
    constructor(bridge) {
        this.bridge = bridge;
        this.routeChanges$ = bridge.$routeChanges;
        this.$ng1RouteChangeSuccess = bridge.$ng1RouteChangeSuccess;
    }
    get() {
        return this.routeChanges$.pipe(switchMap(e => {
            return this.bridge.getActionBarItems();
        }));
    }
}
Ng1ActionBarFactoryService.ɵfac = function Ng1ActionBarFactoryService_Factory(t) { return new (t || Ng1ActionBarFactoryService)(ɵngcc0.ɵɵinject(ɵngcc1.BridgeService)); };
Ng1ActionBarFactoryService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Ng1ActionBarFactoryService, factory: Ng1ActionBarFactoryService.ɵfac });
Ng1ActionBarFactoryService.ctorParameters = () => [
    { type: BridgeService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng1ActionBarFactoryService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BridgeService }]; }, null); })();
