// @c8y/ngx-components/esm2015/upgrade/action.factory.js
import { Injectable } from '@angular/core';
import { TabsService } from '@c8y/ngx-components';
import { BridgeService } from './bridge.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './bridge.service';
import * as ɵngcc2 from '@c8y/ngx-components';
export class Ng1ActionFactoryService {
    constructor(bridge, tabs) {
        this.bridge = bridge;
        this.tabs = tabs;
        this.routeChanges$ = bridge.$routeChanges;
        this.$location = bridge.injector.get('$location');
        this.tabs.items$.subscribe((newTabs) => this.handleTabsRedirect(newTabs));
    }
    handleTabsRedirect(tabs) {
        /**
         * This function is doing the same process as function redirect in the file
         * modules/core/ui/navigation/tabs.provider.js
         * That function  is not run because bridge.service.ts overrides the method addTab where the redirect() was called.
         */
        const redirectedTab = tabs.find((tab) => tab.redirectedTo);
        const [topPriorityTab] = tabs;
        if (redirectedTab && !topPriorityTab.redirectedTo) {
            this.$location.replace();
            this.$location.path(topPriorityTab.path);
            topPriorityTab.redirectedTo = true;
            redirectedTab.redirectedTo = false;
        }
    }
    get() {
        return this.bridge.getActions();
    }
}
Ng1ActionFactoryService.ɵfac = function Ng1ActionFactoryService_Factory(t) { return new (t || Ng1ActionFactoryService)(ɵngcc0.ɵɵinject(ɵngcc1.BridgeService), ɵngcc0.ɵɵinject(ɵngcc2.TabsService)); };
Ng1ActionFactoryService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Ng1ActionFactoryService, factory: Ng1ActionFactoryService.ɵfac });
Ng1ActionFactoryService.ctorParameters = () => [
    { type: BridgeService },
    { type: TabsService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng1ActionFactoryService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BridgeService }, { type: ɵngcc2.TabsService }]; }, null); })();
