// @c8y/ngx-components/esm2015/upgrade/breadcrumb.factory.js
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { debounceTime, switchMap } from 'rxjs/operators';
import { BridgeService } from './bridge.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './bridge.service';
export class Ng1BreadcrumbFactoryService {
    constructor(bridge) {
        this.bridge = bridge;
        this.trigger = new ReplaySubject(1);
        this.breadcrumbs = this.trigger.pipe(debounceTime(100), switchMap(() => {
            return this.bridge.getBreadcrumbs();
        }));
    }
    get() {
        this.trigger.next();
        return this.breadcrumbs;
    }
}
Ng1BreadcrumbFactoryService.ɵfac = function Ng1BreadcrumbFactoryService_Factory(t) { return new (t || Ng1BreadcrumbFactoryService)(ɵngcc0.ɵɵinject(ɵngcc1.BridgeService)); };
Ng1BreadcrumbFactoryService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: Ng1BreadcrumbFactoryService, factory: Ng1BreadcrumbFactoryService.ɵfac });
Ng1BreadcrumbFactoryService.ctorParameters = () => [
    { type: BridgeService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng1BreadcrumbFactoryService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BridgeService }]; }, null); })();
