// @c8y/ngx-components/esm2015/core/tabs/tab.component.js
import { Component, ViewChild, Input } from '@angular/core';
import { TabsService } from './tabs.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './tabs.service';

const _c0 = ["tpl"];
function TabComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵprojection(0);
} }
const _c1 = ["*"];
export class TabComponent {
    constructor(tabs) {
        this.tabs = tabs;
        this.label = '';
    }
    ngAfterContentInit() {
        setTimeout(() => this.tabs.add(this));
    }
    ngOnDestroy() {
        this.tabs.remove(this);
    }
}
TabComponent.ɵfac = function TabComponent_Factory(t) { return new (t || TabComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.TabsService)); };
TabComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TabComponent, selectors: [["c8y-tab"]], viewQuery: function TabComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { label: "label", path: "path", icon: "icon", priority: "priority", orientation: "orientation" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["tpl", ""]], template: function TabComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, TabComponent_ng_template_0_Template, 1, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
TabComponent.ctorParameters = () => [
    { type: TabsService }
];
TabComponent.propDecorators = {
    template: [{ type: ViewChild, args: ['tpl', { static: false },] }],
    path: [{ type: Input }],
    label: [{ type: Input }],
    icon: [{ type: Input }],
    priority: [{ type: Input }],
    orientation: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TabComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-tab',
                template: '<ng-template #tpl><ng-content></ng-content></ng-template>'
            }]
    }], function () { return [{ type: ɵngcc1.TabsService }]; }, { label: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['tpl', { static: false }]
        }], path: [{
            type: Input
        }], icon: [{
            type: Input
        }], priority: [{
            type: Input
        }], orientation: [{
            type: Input
        }] }); })();
