// @c8y/ngx-components/esm2015/device-profile/device-profile-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
export class DeviceProfileNavigationFactory {
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.nodeItem) {
                this.nodeItem = new NavigatorNode({
                    label: gettext('Device profiles'),
                    path: '/device-profiles',
                    icon: 'c8y-device-profile',
                    parent: gettext('Management')
                });
            }
            return this.nodeItem;
        });
    }
}
DeviceProfileNavigationFactory.ɵfac = function DeviceProfileNavigationFactory_Factory(t) { return new (t || DeviceProfileNavigationFactory)(); };
DeviceProfileNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: DeviceProfileNavigationFactory, factory: DeviceProfileNavigationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceProfileNavigationFactory, [{
        type: Injectable
    }], null, null); })();
