// @c8y/ngx-components/esm2015/device-profile/device-profile.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule, CoreModule, HOOK_NAVIGATOR_NODES, HOOK_ROUTE, gettext, ViewContext } from '@c8y/ngx-components';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { ReactiveFormsModule } from '@angular/forms';
import { DeviceProfileComponent } from './device-profile.component';
import { DeviceProfileListComponent } from './device-profile-list.component';
import { DeviceProfileNavigationFactory } from './device-profile-navigation.factory';
import { DeviceProfileService } from './device-profile.service';
import { AddDeviceProfileComponent } from './add-device-profile.component';
import { DeviceProfileGuard } from './device-profile.guard';
import { RepositoryModule } from '@c8y/ngx-components/repository';
import { SelectConfigurationModalComponent } from './select-configuration-modal.component';
import { DeviceTabProfileComponent } from './device-tab-profile/device-tab-profile.component';
import { DeviceTabProfileDetailComponent } from './device-tab-profile/device-tab-profile-detail.component';
import { DeviceProfileItemListComponent } from './device-tab-profile/device-profile-item-list.component';
import { SingleOperationDetailsModule } from '@c8y/ngx-components/operations/single-operation-details';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from 'ngx-bootstrap/dropdown';
const deviceProfilesRoutes = [
    {
        path: 'device-profiles/:id',
        component: DeviceProfileComponent
    },
    {
        path: 'device-profiles',
        component: DeviceProfileListComponent
    }
];
const deviceTabProfileRoutes = [
    {
        context: ViewContext.Device,
        path: 'device-profile',
        component: DeviceTabProfileComponent,
        label: gettext('Device profile'),
        icon: 'c8y-device-profile',
        canActivate: [DeviceProfileGuard]
    }
];
const ɵ0 = deviceTabProfileRoutes;
export class DeviceProfileModule {
}
DeviceProfileModule.ɵfac = function DeviceProfileModule_Factory(t) { return new (t || DeviceProfileModule)(); };
DeviceProfileModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceProfileModule });
DeviceProfileModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: DeviceProfileNavigationFactory,
            multi: true
        },
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        },
        DeviceProfileService,
        DeviceProfileGuard
    ], imports: [[
            CoreModule,
            CommonModule,
            RepositoryModule,
            RouterModule.forChild(deviceProfilesRoutes),
            BsDropdownModule.forRoot(),
            TooltipModule,
            ReactiveFormsModule,
            ButtonsModule,
            PopoverModule,
            SingleOperationDetailsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceProfileModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    DeviceProfileComponent,
                    DeviceProfileListComponent,
                    AddDeviceProfileComponent,
                    SelectConfigurationModalComponent,
                    DeviceTabProfileComponent,
                    DeviceTabProfileDetailComponent,
                    DeviceProfileItemListComponent
                ],
                exports: [],
                imports: [
                    CoreModule,
                    CommonModule,
                    RepositoryModule,
                    RouterModule.forChild(deviceProfilesRoutes),
                    BsDropdownModule.forRoot(),
                    TooltipModule,
                    ReactiveFormsModule,
                    ButtonsModule,
                    PopoverModule,
                    SingleOperationDetailsModule
                ],
                entryComponents: [
                    AddDeviceProfileComponent,
                    SelectConfigurationModalComponent,
                    DeviceTabProfileComponent
                ],
                providers: [
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: DeviceProfileNavigationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    },
                    DeviceProfileService,
                    DeviceProfileGuard
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceProfileModule, { declarations: function () { return [DeviceProfileComponent,
        DeviceProfileListComponent,
        AddDeviceProfileComponent,
        SelectConfigurationModalComponent,
        DeviceTabProfileComponent,
        DeviceTabProfileDetailComponent,
        DeviceProfileItemListComponent]; }, imports: function () { return [CoreModule,
        CommonModule,
        RepositoryModule, ɵngcc1.RouterModule, ɵngcc2.BsDropdownModule, TooltipModule,
        ReactiveFormsModule,
        ButtonsModule,
        PopoverModule,
        SingleOperationDetailsModule]; } }); })();
export { ɵ0 };
