// @c8y/ngx-components/esm2015/repository/repository-navigation-factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
export class RepositoryNavigationFactory {
    constructor() {
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Firmware repository'),
                    path: 'firmware',
                    icon: 'c8y-firmware',
                    parent: gettext('Management'),
                    priority: 1000
                }));
                this.navs.push(new NavigatorNode({
                    label: gettext('Software repository'),
                    path: 'software',
                    icon: 'c8y-tools',
                    parent: gettext('Management'),
                    priority: 900
                }));
                this.navs.push(new NavigatorNode({
                    label: gettext('Configuration repository'),
                    path: 'configuration',
                    icon: 'gears',
                    parent: gettext('Management'),
                    priority: 800
                }));
            }
            return this.navs;
        });
    }
}
RepositoryNavigationFactory.ɵfac = function RepositoryNavigationFactory_Factory(t) { return new (t || RepositoryNavigationFactory)(); };
RepositoryNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: RepositoryNavigationFactory, factory: RepositoryNavigationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RepositoryNavigationFactory, [{
        type: Injectable
    }], function () { return []; }, null); })();
