// @c8y/ngx-components/esm2015/repository/repository.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule, FormsModule, gettext, HOOK_NAVIGATOR_NODES, HOOK_ROUTE, OperationRealtimeService, ViewContext } from '@c8y/ngx-components';
import { SingleOperationDetailsModule } from '@c8y/ngx-components/operations/single-operation-details';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { TabsModule } from 'ngx-bootstrap/tabs';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { ConfigurationFilterPipe } from './configuration-device-tab/configuration-filter.pipe';
import { ConfigurationPreviewComponent } from './configuration-device-tab/configuration-preview.component';
import { DeviceConfigurationListComponent } from './configuration-device-tab/device-configuration-list.component';
import { DeviceConfigurationComponent } from './configuration-device-tab/device-configuration.component';
import { DeviceConfigurationGuard } from './configuration-device-tab/device-configuration.guard';
import { DeviceConfigurationService } from './configuration-device-tab/device-configuration.service';
import { SaveToRepositoryComponent } from './configuration-device-tab/save-to-repository.component';
import { SourceCodePreviewComponent } from './configuration-device-tab/source-code-preview.component';
import { TextBasedConfigurationComponent } from './configuration-device-tab/text-based-configuration.component';
import { ConfigurationDetailComponent } from './configuration/configuration-detail.component';
import { ConfigurationListComponent } from './configuration/configuration-list.component';
import { FileDownloadComponent } from './file-download/file-download.component';
import { FirmwareDeviceTabComponent } from './firmware-device-tab/firmware-device-tab.component';
import { FirmwareDeviceTabGuard } from './firmware-device-tab/firmware-device-tab.guard';
import { AddFirmwareModalComponent } from './firmware/add-firmware-modal.component';
import { AddFirmwarePatchModalComponent } from './firmware/add-firmware-patch-modal.component';
import { FirmwareDetailsComponent } from './firmware/firmware-details.component';
import { FirmwareListComponent } from './firmware/firmware-list.component';
import { RepositoryNavigationFactory } from './repository-navigation-factory';
import { RepositoryService } from './repository.service';
import { RepositorySelectModalComponent } from './select-modal/repository-select-modal.component';
import { AdvancedSoftwareService } from './software-device-tab/advanced-software.service';
import { DeviceSoftwareChangesComponent } from './software-device-tab/device-software-changes.component';
import { DeviceSoftwareListComponent } from './software-device-tab/device-software-list.component';
import { InstalledSoftwareComponent } from './software-device-tab/installed-software.component';
import { SoftwareDeviceTabComponent } from './software-device-tab/software-device-tab.component';
import { SoftwareDeviceTabGuard } from './software-device-tab/software-device-tab.guard';
import { AddSoftwareModalComponent } from './software/add-software-modal.component';
import { DeviceTypeCellRendererComponent } from './software/columns/device-type.cell-renderer.component';
import { NameCellRendererComponent } from './software/columns/name.cell-renderer.component';
import { SoftwareTypeCellRendererComponent } from './software/columns/software-type.cell-renderer.component';
import { SoftwareTypeFilteringFormRendererComponent } from './software/columns/software-type.filtering-form-renderer.component';
import { VersionsCellRendererComponent } from './software/columns/versions.cell-renderer.component';
import { SoftwareDetailsComponent } from './software/software-details.component';
import { SoftwareListComponent } from './software/software-list.component';
import { SoftwareTypeComponent } from './software/software-type.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
import * as ɵngcc2 from 'ngx-bootstrap/tabs';
const ɵ0 = [
    {
        path: 'firmware',
        component: FirmwareListComponent
    },
    {
        path: 'firmware/:id',
        component: FirmwareDetailsComponent
    },
    {
        context: ViewContext.Device,
        path: 'firmware',
        component: FirmwareDeviceTabComponent,
        label: gettext('Firmware'),
        icon: 'c8y-firmware',
        priority: 500,
        canActivate: [FirmwareDeviceTabGuard]
    },
    {
        path: 'software',
        component: SoftwareListComponent
    },
    {
        path: 'software/:id',
        component: SoftwareDetailsComponent
    },
    {
        context: ViewContext.Device,
        path: 'software',
        component: SoftwareDeviceTabComponent,
        label: gettext('Software'),
        icon: 'c8y-tools',
        priority: 499,
        canActivate: [SoftwareDeviceTabGuard]
    },
    {
        path: 'configuration',
        component: ConfigurationListComponent
    },
    {
        context: ViewContext.Device,
        path: 'device-configuration',
        component: DeviceConfigurationComponent,
        label: gettext('Configuration'),
        icon: 'gears',
        priority: 600,
        canActivate: [DeviceConfigurationGuard]
    }
];
export class RepositoryModule {
}
RepositoryModule.ɵfac = function RepositoryModule_Factory(t) { return new (t || RepositoryModule)(); };
RepositoryModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: RepositoryModule });
RepositoryModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        RepositoryService,
        OperationRealtimeService,
        DeviceConfigurationService,
        FirmwareDeviceTabGuard,
        SoftwareDeviceTabGuard,
        DeviceConfigurationGuard,
        AdvancedSoftwareService,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: RepositoryNavigationFactory,
            multi: true
        },
        {
            provide: HOOK_ROUTE,
            useValue: ɵ0,
            multi: true
        }
    ], imports: [[
            CoreModule,
            FormsModule,
            RouterModule,
            SingleOperationDetailsModule,
            BsDropdownModule.forRoot(),
            TooltipModule,
            PopoverModule,
            TabsModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RepositoryModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CoreModule,
                    FormsModule,
                    RouterModule,
                    SingleOperationDetailsModule,
                    BsDropdownModule.forRoot(),
                    TooltipModule,
                    PopoverModule,
                    TabsModule.forRoot()
                ],
                exports: [SoftwareTypeComponent, FileDownloadComponent],
                declarations: [
                    FirmwareListComponent,
                    FirmwareDetailsComponent,
                    AddFirmwareModalComponent,
                    AddFirmwarePatchModalComponent,
                    SoftwareListComponent,
                    SoftwareDetailsComponent,
                    AddSoftwareModalComponent,
                    SoftwareTypeComponent,
                    FirmwareDeviceTabComponent,
                    SoftwareDeviceTabComponent,
                    InstalledSoftwareComponent,
                    RepositorySelectModalComponent,
                    DeviceSoftwareListComponent,
                    DeviceSoftwareChangesComponent,
                    ConfigurationListComponent,
                    ConfigurationDetailComponent,
                    DeviceConfigurationComponent,
                    DeviceConfigurationListComponent,
                    SourceCodePreviewComponent,
                    ConfigurationPreviewComponent,
                    ConfigurationFilterPipe,
                    SaveToRepositoryComponent,
                    TextBasedConfigurationComponent,
                    FileDownloadComponent,
                    NameCellRendererComponent,
                    SoftwareTypeCellRendererComponent,
                    SoftwareTypeFilteringFormRendererComponent,
                    DeviceTypeCellRendererComponent,
                    VersionsCellRendererComponent
                ],
                entryComponents: [
                    FirmwareListComponent,
                    FirmwareDetailsComponent,
                    AddFirmwareModalComponent,
                    AddFirmwarePatchModalComponent,
                    SoftwareListComponent,
                    SoftwareDetailsComponent,
                    AddSoftwareModalComponent,
                    FirmwareDeviceTabComponent,
                    SoftwareDeviceTabComponent,
                    DeviceSoftwareListComponent,
                    RepositorySelectModalComponent,
                    DeviceSoftwareChangesComponent,
                    ConfigurationListComponent,
                    ConfigurationDetailComponent,
                    DeviceConfigurationComponent,
                    SaveToRepositoryComponent,
                    TextBasedConfigurationComponent
                ],
                providers: [
                    RepositoryService,
                    OperationRealtimeService,
                    DeviceConfigurationService,
                    FirmwareDeviceTabGuard,
                    SoftwareDeviceTabGuard,
                    DeviceConfigurationGuard,
                    AdvancedSoftwareService,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: RepositoryNavigationFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_ROUTE,
                        useValue: ɵ0,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RepositoryModule, { declarations: function () { return [FirmwareListComponent,
        FirmwareDetailsComponent,
        AddFirmwareModalComponent,
        AddFirmwarePatchModalComponent,
        SoftwareListComponent,
        SoftwareDetailsComponent,
        AddSoftwareModalComponent,
        SoftwareTypeComponent,
        FirmwareDeviceTabComponent,
        SoftwareDeviceTabComponent,
        InstalledSoftwareComponent,
        RepositorySelectModalComponent,
        DeviceSoftwareListComponent,
        DeviceSoftwareChangesComponent,
        ConfigurationListComponent,
        ConfigurationDetailComponent,
        DeviceConfigurationComponent,
        DeviceConfigurationListComponent,
        SourceCodePreviewComponent,
        ConfigurationPreviewComponent,
        ConfigurationFilterPipe,
        SaveToRepositoryComponent,
        TextBasedConfigurationComponent,
        FileDownloadComponent,
        NameCellRendererComponent,
        SoftwareTypeCellRendererComponent,
        SoftwareTypeFilteringFormRendererComponent,
        DeviceTypeCellRendererComponent,
        VersionsCellRendererComponent]; }, imports: function () { return [CoreModule,
        FormsModule,
        RouterModule,
        SingleOperationDetailsModule, ɵngcc1.BsDropdownModule, TooltipModule,
        PopoverModule, ɵngcc2.TabsModule]; }, exports: function () { return [SoftwareTypeComponent, FileDownloadComponent]; } }); })();
export { ɵ0 };
