// @c8y/ngx-components/esm2015/connectivity/actility/actility.guard.js
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ActilityGuard {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    canActivate(route, state) {
        if (this.showActilityProvider === undefined) {
            this.showActilityProvider = this.applicationService
                .isAvailable(ActilityGuard.applicationName)
                .then(result => result && result.data);
        }
        return this.showActilityProvider;
    }
}
ActilityGuard.ɵfac = function ActilityGuard_Factory(t) { return new (t || ActilityGuard)(ɵngcc0.ɵɵinject(ɵngcc1.ApplicationService)); };
ActilityGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ActilityGuard, factory: ActilityGuard.ɵfac });
ActilityGuard.applicationName = 'actility';
ActilityGuard.ctorParameters = () => [
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ApplicationService }]; }, null); })();
