// @c8y/ngx-components/esm2015/connectivity/actility/actility.module.js
import { NgModule } from '@angular/core';
import { HOOK_NAVIGATOR_NODES, HOOK_TABS } from '@c8y/ngx-components';
import { ActilityNodeFactory } from './actility-node.factory';
import { ActilityTabFactory } from './actility-tab.factory';
import { ActilityGuard } from './actility.guard';
import * as ɵngcc0 from '@angular/core';
export class ActilityModule {
}
ActilityModule.ɵfac = function ActilityModule_Factory(t) { return new (t || ActilityModule)(); };
ActilityModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ActilityModule });
ActilityModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        ActilityGuard,
        ActilityTabFactory,
        {
            provide: HOOK_NAVIGATOR_NODES,
            useClass: ActilityNodeFactory,
            multi: true
        },
        {
            provide: HOOK_TABS,
            useClass: ActilityTabFactory,
            multi: true
        }
    ] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActilityModule, [{
        type: NgModule,
        args: [{
                providers: [
                    ActilityGuard,
                    ActilityTabFactory,
                    {
                        provide: HOOK_NAVIGATOR_NODES,
                        useClass: ActilityNodeFactory,
                        multi: true
                    },
                    {
                        provide: HOOK_TABS,
                        useClass: ActilityTabFactory,
                        multi: true
                    }
                ]
            }]
    }], null, null); })();
