// @c8y/ngx-components/esm2015/ecosystem/applications/install-from-package/install-from-package.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { WizardComponent } from '@c8y/ngx-components';
import { EcosystemService } from '../../ecosystem.service';
import { ApplicationPropertiesFormComponent } from '../../shared/application-properties-form.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../ecosystem.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../../shared/application-properties-form.component';

function InstallFromPackageComponent_ng_container_5_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 10);
    ɵngcc0.ɵɵelementStart(1, "p", 11);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "No packages to display."));
} }
function InstallFromPackageComponent_ng_container_5_div_6_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 14);
    ɵngcc0.ɵɵlistener("click", function InstallFromPackageComponent_ng_container_5_div_6_button_1_Template_button_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r10); const package_r8 = restoredCtx.$implicit; const ctx_r9 = ɵngcc0.ɵɵnextContext(3); return ctx_r9.selectPackage(package_r8); });
    ɵngcc0.ɵɵelement(1, "i", 15);
    ɵngcc0.ɵɵelement(2, "span", 16);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const package_r8 = ctx.$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", package_r8.name);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("innerText", package_r8.name);
} }
function InstallFromPackageComponent_ng_container_5_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵtemplate(1, InstallFromPackageComponent_ng_container_5_div_6_button_1_Template, 3, 2, "button", 13);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r6.packages);
} }
function InstallFromPackageComponent_ng_container_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "div", 6);
    ɵngcc0.ɵɵelementStart(2, "p", 7);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, InstallFromPackageComponent_ng_container_5_div_5_Template, 4, 3, "div", 8);
    ɵngcc0.ɵɵtemplate(6, InstallFromPackageComponent_ng_container_5_div_6_Template, 2, 1, "div", 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, "Select from available packages"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !(ctx_r0.packages == null ? null : ctx_r0.packages.length));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.packages == null ? null : ctx_r0.packages.length);
} }
function InstallFromPackageComponent_ng_container_6_c8y_application_properties_form_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-application-properties-form", 21);
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("application", ctx_r11.newAppConfig);
} }
function InstallFromPackageComponent_ng_container_6_c8y_progress_bar_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-progress-bar", 22);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("message", ɵngcc0.ɵɵpipeBind1(1, 1, "Installing\u2026"));
} }
function InstallFromPackageComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "p", 17);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "div", 18);
    ɵngcc0.ɵɵtemplate(5, InstallFromPackageComponent_ng_container_6_c8y_application_properties_form_5_Template, 1, 1, "c8y-application-properties-form", 19);
    ɵngcc0.ɵɵtemplate(6, InstallFromPackageComponent_ng_container_6_c8y_progress_bar_6_Template, 2, 3, "c8y-progress-bar", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 3, "Provide application details"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r1.inProgress);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.inProgress);
} }
function InstallFromPackageComponent_ng_container_7_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 26);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, "Application created"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function InstallFromPackageComponent_ng_container_7_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelement(1, "c8y-operation-result", 27);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("text", ɵngcc0.ɵɵpipeBind1(2, 3, "Application creation failed"));
    ɵngcc0.ɵɵproperty("size", 84)("vertical", true);
} }
function InstallFromPackageComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, InstallFromPackageComponent_ng_container_7_div_1_Template, 3, 5, "div", 23);
    ɵngcc0.ɵɵtemplate(2, InstallFromPackageComponent_ng_container_7_ng_template_2_Template, 3, 5, "ng-template", null, 24, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const _r14 = ɵngcc0.ɵɵreference(3);
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.deployedWithSuccess)("ngIfElse", _r14);
} }
function InstallFromPackageComponent_button_9_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 28);
    ɵngcc0.ɵɵlistener("click", function InstallFromPackageComponent_button_9_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.selectedPackage ? ctx_r16.selectedPackage = null : ctx_r16.back(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2, " Back ");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, "Back"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r3.inProgress);
} }
function InstallFromPackageComponent_button_16_Template(rf, ctx) { if (rf & 1) {
    const _r19 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 29);
    ɵngcc0.ɵɵlistener("click", function InstallFromPackageComponent_button_16_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r19); const ctx_r18 = ɵngcc0.ɵɵnextContext(); return ctx_r18.save(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 3, "Install"));
    ɵngcc0.ɵɵproperty("disabled", ctx_r4.inProgress);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 5, "Install"), " ");
} }
export class InstallFromPackageComponent {
    constructor(ecosystemService, applicationService, wizardComponent) {
        this.ecosystemService = ecosystemService;
        this.applicationService = applicationService;
        this.wizardComponent = wizardComponent;
        this.deployedWithSuccess = false;
        this.isDeployed = false;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.loadPackages();
        });
    }
    back() {
        this.wizardComponent.reset();
    }
    cancel() {
        this.wizardComponent.close();
    }
    save() {
        return __awaiter(this, void 0, void 0, function* () {
            const formGroupValue = this.applicationPropertiesForm.formGroup.getRawValue();
            this.inProgress = true;
            const { data: clonedPkg } = yield this.applicationService.clone(this.selectedPackage);
            const config = this.ecosystemService.createConfig(clonedPkg, formGroupValue);
            try {
                const { data: updatedApp } = yield this.ecosystemService.updateApp(config, true);
                if (updatedApp) {
                    yield this.ecosystemService.updateAppManifest(updatedApp, this.selectedPackage);
                }
                this.deployedWithSuccess = true;
            }
            catch (error) {
                this.inProgress = false;
                yield this.applicationService.delete(clonedPkg.id);
            }
            this.markAsDeployed();
        });
    }
    selectPackage(selectedPackage) {
        return __awaiter(this, void 0, void 0, function* () {
            const apps = yield this.ecosystemService.getWebApplications();
            this.newAppConfig = this.ecosystemService.getUniqueAppConfig(selectedPackage, apps);
            this.selectedPackage = selectedPackage;
        });
    }
    /* The 'select package version' option will be added later when BE support for multiple app versions has been implemented. MTM-40278
    // onSelectedArchive(archive) {
    //   this.formGroup.controls['packageVersion'].setValue(archive.name);
    // } */
    markAsDeployed() {
        this.isDeployed = true;
        this.inProgress = false;
    }
    loadPackages() {
        return __awaiter(this, void 0, void 0, function* () {
            const applications = yield this.ecosystemService.getApplications();
            this.packages = applications.data.filter((app) => this.ecosystemService.isApplicationPackage(app));
        });
    }
}
InstallFromPackageComponent.ɵfac = function InstallFromPackageComponent_Factory(t) { return new (t || InstallFromPackageComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.WizardComponent)); };
InstallFromPackageComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: InstallFromPackageComponent, selectors: [["c8y-install-from-package"]], viewQuery: function InstallFromPackageComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ApplicationPropertiesFormComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.applicationPropertiesForm = _t.first);
    } }, decls: 17, vars: 16, consts: [[3, "c8yIcon"], ["translate", ""], [4, "ngIf"], ["class", "btn btn-default", "translate", "", "type", "button", 3, "title", "disabled", "click", 4, "ngIf"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"], ["class", "btn btn-primary", "type", "button", 3, "title", "disabled", "click", 4, "ngIf"], [1, "modal-inner-scroll"], [1, "p-16", "text-medium", "text-center", "separator-bottom", "sticky-top", "bg-white"], ["class", "c8y-empty-state text-center", 4, "ngIf"], ["class", "c8y-wizard-list-nav", "style", "min-height: 257px", 4, "ngIf"], [1, "c8y-empty-state", "text-center"], [1, "text-center"], [1, "c8y-wizard-list-nav", 2, "min-height", "257px"], ["class", "list-group-item text-truncate", "type", "button", 3, "title", "click", 4, "ngFor", "ngForOf"], ["type", "button", 1, "list-group-item", "text-truncate", 3, "title", "click"], ["c8yIcon", "big-parcel", 1, "list-group-icon"], [3, "innerText"], [1, "p-16", "text-center", "text-medium", "separator-bottom", "sticky-top", "bg-white"], [1, "d-flex", "d-col", "a-i-center", "j-c-center", 2, "min-height", "257px"], ["class", "d-block fit-w", 3, "application", 4, "ngIf"], ["class", "text-center d-block", 3, "message", 4, "ngIf"], [1, "d-block", "fit-w", 3, "application"], [1, "text-center", "d-block", 3, "message"], ["class", "d-flex a-i-center j-c-center", "style", "min-height: 257px", 4, "ngIf", "ngIfElse"], ["failedDeploy", ""], [1, "d-flex", "a-i-center", "j-c-center", 2, "min-height", "257px"], ["type", "success", 1, "lead", 3, "text", "size", "vertical"], ["type", "error", 1, "lead", 3, "text", "size", "vertical"], ["translate", "", "type", "button", 1, "btn", "btn-default", 3, "title", "disabled", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "title", "disabled", "click"]], template: function InstallFromPackageComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-wizard-header");
        ɵngcc0.ɵɵelement(1, "i", 0);
        ɵngcc0.ɵɵelementStart(2, "h4", 1);
        ɵngcc0.ɵɵtext(3, "Install from package");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-wizard-body");
        ɵngcc0.ɵɵtemplate(5, InstallFromPackageComponent_ng_container_5_Template, 7, 5, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(6, InstallFromPackageComponent_ng_container_6_Template, 7, 5, "ng-container", 2);
        ɵngcc0.ɵɵtemplate(7, InstallFromPackageComponent_ng_container_7_Template, 4, 2, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "c8y-wizard-footer");
        ɵngcc0.ɵɵtemplate(9, InstallFromPackageComponent_button_9_Template, 3, 4, "button", 3);
        ɵngcc0.ɵɵelementStart(10, "button", 4);
        ɵngcc0.ɵɵlistener("click", function InstallFromPackageComponent_Template_button_click_10_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵtext(13);
        ɵngcc0.ɵɵpipe(14, "translate");
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(16, InstallFromPackageComponent_button_16_Template, 4, 7, "button", 5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("c8yIcon", "big-parcel");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.selectedPackage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDeployed && ctx.selectedPackage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDeployed);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDeployed);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ctx.isDeployed && ctx.deployedWithSuccess ? ɵngcc0.ɵɵpipeBind1(11, 8, "Close") : ɵngcc0.ɵɵpipeBind1(12, 10, "Cancel"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.isDeployed && ctx.deployedWithSuccess ? ɵngcc0.ɵɵpipeBind1(14, 12, "Close") : ɵngcc0.ɵɵpipeBind1(15, 14, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isDeployed);
    } }, directives: [ɵngcc3.WizardHeaderComponent, ɵngcc3.IconDirective, ɵngcc3.C8yTranslateDirective, ɵngcc3.WizardBodyComponent, ɵngcc4.NgIf, ɵngcc3.WizardFooterComponent, ɵngcc4.NgForOf, ɵngcc5.ApplicationPropertiesFormComponent, ɵngcc3.ProgressBarComponent, ɵngcc3.ɵe], pipes: [ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
InstallFromPackageComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: ApplicationService },
    { type: WizardComponent }
];
InstallFromPackageComponent.propDecorators = {
    applicationPropertiesForm: [{ type: ViewChild, args: [ApplicationPropertiesFormComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(InstallFromPackageComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-install-from-package',
                template: "<c8y-wizard-header>\n  <i [c8yIcon]=\"'big-parcel'\"></i>\n  <h4 translate>Install from package</h4>\n</c8y-wizard-header>\n<c8y-wizard-body>\n  <ng-container *ngIf=\"!selectedPackage\">\n    <div class=\"modal-inner-scroll\">\n      <p class=\"p-16 text-medium text-center separator-bottom sticky-top bg-white\">\n        {{ 'Select from available packages' | translate }}\n      </p>\n      <div *ngIf=\"!packages?.length\" class=\"c8y-empty-state text-center\">\n        <p class=\"text-center\">{{ 'No packages to display.' | translate }}</p>\n      </div>\n      <div *ngIf=\"packages?.length\" class=\"c8y-wizard-list-nav\" style=\"min-height: 257px\">\n        <button\n          class=\"list-group-item text-truncate\"\n          *ngFor=\"let package of packages\"\n          (click)=\"selectPackage(package)\"\n          title=\"{{ package.name }}\"\n          type=\"button\"\n        >\n          <i c8yIcon=\"big-parcel\" class=\"list-group-icon\"></i>\n          <span [innerText]=\"package.name\"></span>\n        </button>\n      </div>\n    </div>\n  </ng-container>\n  <ng-container *ngIf=\"!isDeployed && selectedPackage\">\n    <p class=\"p-16 text-center text-medium separator-bottom sticky-top bg-white\">\n      {{ 'Provide application details' | translate }}\n    </p>\n    <div class=\"d-flex d-col a-i-center j-c-center\" style=\"min-height: 257px\">\n      <c8y-application-properties-form\n        *ngIf=\"!inProgress\"\n        [application]=\"newAppConfig\"\n        class=\"d-block fit-w\"\n      ></c8y-application-properties-form>\n      <c8y-progress-bar\n        [message]=\"'Installing\u2026' | translate\"\n        class=\"text-center d-block\"\n        *ngIf=\"inProgress\"\n      ></c8y-progress-bar>\n    </div>\n  </ng-container>\n\n  <ng-container *ngIf=\"isDeployed\">\n    <div\n      *ngIf=\"deployedWithSuccess; else failedDeploy\"\n      class=\"d-flex a-i-center j-c-center\"\n      style=\"min-height: 257px\"\n    >\n      <c8y-operation-result\n        text=\"{{ 'Application created' | translate }}\"\n        [size]=\"84\"\n        [vertical]=\"true\"\n        type=\"success\"\n        class=\"lead\"\n      ></c8y-operation-result>\n    </div>\n    <ng-template #failedDeploy>\n      <div class=\"d-flex a-i-center j-c-center\" style=\"min-height: 257px\">\n        <c8y-operation-result\n          text=\"{{ 'Application creation failed' | translate }}\"\n          [size]=\"84\"\n          [vertical]=\"true\"\n          type=\"error\"\n          class=\"lead\"\n        ></c8y-operation-result>\n      </div>\n    </ng-template>\n  </ng-container>\n</c8y-wizard-body>\n\n<c8y-wizard-footer>\n  <button\n    *ngIf=\"!isDeployed\"\n    (click)=\"selectedPackage ? (selectedPackage = null) : back()\"\n    class=\"btn btn-default\"\n    title=\"{{ 'Back' | translate }}\"\n    translate\n    [disabled]=\"inProgress\"\n    type=\"button\"\n  >\n    Back\n  </button>\n  <button\n    title=\"{{ isDeployed && deployedWithSuccess ? ('Close' | translate) : ('Cancel' | translate) }}\"\n    class=\"btn btn-default\"\n    type=\"button\"\n    (click)=\"cancel()\"\n  >\n    {{ isDeployed && deployedWithSuccess ? ('Close' | translate) : ('Cancel' | translate) }}\n  </button>\n\n  <button\n    title=\"{{ 'Install' | translate }}\"\n    class=\"btn btn-primary\"\n    type=\"button\"\n    (click)=\"save()\"\n    [disabled]=\"inProgress\"\n    *ngIf=\"!isDeployed\"\n  >\n    {{ 'Install' | translate }}\n  </button>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.ApplicationService }, { type: ɵngcc3.WizardComponent }]; }, { applicationPropertiesForm: [{
            type: ViewChild,
            args: [ApplicationPropertiesFormComponent]
        }] }); })();
