// @c8y/ngx-components/esm2015/core/select-modal/select-modal.module.js
import { NgModule } from '@angular/core';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ListGroupModule } from '../list-group/list-group.module';
import { FormsModule } from '../forms/forms.module';
import { SelectModalComponent } from './select-modal.component';
import { SelectModalFilterPipe } from './select-modal-filter.pipe';
import { SearchModule } from '../search/search.module';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
export class SelectModalModule {
}
SelectModalModule.ɵfac = function SelectModalModule_Factory(t) { return new (t || SelectModalModule)(); };
SelectModalModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: SelectModalModule });
SelectModalModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            BsDropdownModule.forRoot(),
            CommonModule,
            ListGroupModule,
            SearchModule,
            FormsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SelectModalModule, [{
        type: NgModule,
        args: [{
                imports: [
                    BsDropdownModule.forRoot(),
                    CommonModule,
                    ListGroupModule,
                    SearchModule,
                    FormsModule
                ],
                exports: [
                    SelectModalComponent,
                    SelectModalFilterPipe
                ],
                declarations: [
                    SelectModalComponent,
                    SelectModalFilterPipe
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SelectModalModule, { declarations: function () { return [SelectModalComponent,
        SelectModalFilterPipe]; }, imports: function () { return [ɵngcc1.BsDropdownModule, CommonModule,
        ListGroupModule,
        SearchModule,
        FormsModule]; }, exports: function () { return [SelectModalComponent,
        SelectModalFilterPipe]; } }); })();
