// @c8y/ngx-components/esm2015/ecosystem/packages/package-archives/package-archives.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { EcosystemService } from '../../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '../../ecosystem.service';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '../../activity-log/activity-log.component';

function PackageArchivesComponent_c8y_activity_log_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-activity-log", 9);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("application", ctx_r0.application);
} }
export class PackageArchivesComponent {
    constructor(activatedRoute, ecosystemService) {
        this.activatedRoute = activatedRoute;
        this.ecosystemService = ecosystemService;
        this.isLoading = false;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.loadApplication();
        });
    }
    loadApplication() {
        return __awaiter(this, void 0, void 0, function* () {
            const { id } = this.activatedRoute.snapshot.parent.data.contextData;
            this.isLoading = true;
            this.application = yield this.ecosystemService.getApplication(id);
            this.isLoading = false;
        });
    }
}
PackageArchivesComponent.ɵfac = function PackageArchivesComponent_Factory(t) { return new (t || PackageArchivesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService)); };
PackageArchivesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: PackageArchivesComponent, selectors: [["c8y-package-archives"]], decls: 24, vars: 31, consts: [[3, "icon", "label"], [3, "icon", "label", "path"], [3, "label"], [1, "row"], [1, "col-lg-8"], [1, "card", "content-fullpage", "d-flex", "d-col"], [1, "card-header", "large-padding", "separator"], [1, ""], ["class", "d-contents", 3, "application", 4, "ngIf"], [1, "d-contents", 3, "application"]], template: function PackageArchivesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵpipe(3, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(5, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelement(7, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(8, "translate");
        ɵngcc0.ɵɵelement(9, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(10, "translate");
        ɵngcc0.ɵɵelement(11, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵpipe(13, "humanizeAppName");
        ɵngcc0.ɵɵelement(14, "c8y-breadcrumb-item", 2);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 3);
        ɵngcc0.ɵɵelementStart(17, "div", 4);
        ɵngcc0.ɵɵelementStart(18, "div", 5);
        ɵngcc0.ɵɵelementStart(19, "div", 6);
        ɵngcc0.ɵɵelementStart(20, "h4", 7);
        ɵngcc0.ɵɵtext(21);
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(23, PackageArchivesComponent_c8y_activity_log_23_Template, 1, 1, "c8y-activity-log", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 13, ɵngcc0.ɵɵpipeBind1(3, 15, ctx.application)));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(6, 17, "Ecosystem"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(8, 19, "Applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "big-parcel")("label", ɵngcc0.ɵɵpipeBind1(10, 21, "Packages"))("path", "ecosystem/package-list");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(12, 23, ɵngcc0.ɵɵpipeBind1(13, 25, ctx.application)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("label", ɵngcc0.ɵɵpipeBind1(15, 27, "Archives"));
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(22, 29, "Archives"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
    } }, directives: [ɵngcc3.TitleComponent, ɵngcc3.BreadcrumbComponent, ɵngcc3.BreadcrumbItemComponent, ɵngcc4.NgIf, ɵngcc5.ActivityLogComponent], pipes: [ɵngcc4.AsyncPipe, ɵngcc3.HumanizeAppNamePipe, ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
PackageArchivesComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: EcosystemService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PackageArchivesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-package-archives',
                template: "<c8y-title>{{ application | humanizeAppName | async }}</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'big-parcel'\"\n    [label]=\"'Packages' | translate\"\n    [path]=\"'ecosystem/package-list'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"application | humanizeAppName | async\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [label]=\"'Archives' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<div class=\"row\">\n  <div class=\"col-lg-8\">\n    <div class=\"card content-fullpage d-flex d-col\">\n      <div class=\"card-header large-padding separator\">\n        <h4 class>{{ 'Archives' | translate }}</h4>\n      </div>\n      <c8y-activity-log class=\"d-contents\" *ngIf=\"!isLoading\" [application]=\"application\">\n      </c8y-activity-log>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }, { type: ɵngcc2.EcosystemService }]; }, null); })();
