// @c8y/ngx-components/esm2015/device-grid/schema-form.component.js
import { Directive, ElementRef, Injector, Input } from '@angular/core';
import { UpgradeComponent } from '@angular/upgrade/static';
// TODO: MTM-31785: find replacement for schema-form ng1 library
import * as ɵngcc0 from '@angular/core';
export class SchemaFormComponent extends UpgradeComponent {
    constructor(elementRef, injector) {
        super('c8ySchemaForm', elementRef, injector);
    }
}
SchemaFormComponent.ɵfac = function SchemaFormComponent_Factory(t) { return new (t || SchemaFormComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
SchemaFormComponent.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: SchemaFormComponent, selectors: [["c8y-schema-form"]], inputs: { schema: "schema", model: "model", inputForm: ["form", "inputForm"], inputOptions: ["options", "inputOptions"] }, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
SchemaFormComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Injector }
];
SchemaFormComponent.propDecorators = {
    schema: [{ type: Input }],
    model: [{ type: Input }],
    inputForm: [{ type: Input, args: ['form',] }],
    inputOptions: [{ type: Input, args: ['options',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SchemaFormComponent, [{
        type: Directive,
        args: [{
                selector: 'c8y-schema-form'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Injector }]; }, { schema: [{
            type: Input
        }], model: [{
            type: Input
        }], inputForm: [{
            type: Input,
            args: ['form']
        }], inputOptions: [{
            type: Input,
            args: ['options']
        }] }); })();
