// @c8y/client/lib/esm2015/src/tenant-login-options/ITenantLoginOptions.js
export var TenantLoginOptionType;
(function (TenantLoginOptionType) {
    /**
     * OAI-Secure: authentication information is stored in a HttpOnly cookie
     */
    TenantLoginOptionType["OAUTH2_INTERNAL"] = "OAUTH2_INTERNAL";
    /**
     * SSO redirect: allows a user to login with a single 3rd-party authorization server using the OAuth2 protocol.
     */
    TenantLoginOptionType["OAUTH2"] = "OAUTH2";
    /**
     * Basic Auth: authentication information is stored in a session storage
     */
    TenantLoginOptionType["BASIC"] = "BASIC";
})(TenantLoginOptionType || (TenantLoginOptionType = {}));
export var GrantType;
(function (GrantType) {
    GrantType["AUTHORIZATION_CODE"] = "AUTHORIZATION_CODE";
    GrantType["PASSWORD"] = "PASSWORD";
})(GrantType || (GrantType = {}));
export var UserManagementSource;
(function (UserManagementSource) {
    UserManagementSource["REMOTE"] = "REMOTE";
    UserManagementSource["INTERNAL"] = "INTERNAL";
})(UserManagementSource || (UserManagementSource = {}));
