// @c8y/ngx-components/esm2015/core/common/dropdown-direction.directive.js
import { Directive, ElementRef, HostListener, Renderer2 } from '@angular/core';
import { forEach } from 'lodash-es';
/**
 * A directive to determine if a dropdown should open to the top or as usual to the bottom.
 * This directive has a direct dependency to the bootstrap dropdown (!)
 *
 * Preconditions:
 *  - Bootstrap dropdown is used
 *  - Bootstrap dropdown is not used with container="body"
 *  - Bootstrap dropdown is used like it is intended and documented for e.g.
 * ```html
 * <div dropdown class="dropdown">
 *  <button class="dropdown-toggle" dropdownToggle>Toggle</button>
 *  <ul class="dropdown-menu" *dropdownMenu>
 *    <!-- Content here -->
 *  </ul>
 * </div>
 * ```
 *
 * Example:
 * ```html
 * <div dropdown c8yBsDropdownDirection class="dropdown">
 *  <button class="dropdown-toggle" dropdownToggle>Toggle</button>
 *  <ul class="dropdown-menu" *dropdownMenu>
 *    <!-- Content here -->
 *  </ul>
 * </div>
 * ```
 * You make use of the bootstrap dropdown as usual and additionally add the direction directive to it.
 */
import * as ɵngcc0 from '@angular/core';
export class DropdownDirectionDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.MARGIN_BOTTOM = 10;
        this.VISIBILITY = 'visibility';
        this.DROPUP = 'dropup';
    }
    onClick() {
        const open = this.elementRef.nativeElement.classList.contains('open');
        if (open) {
            let menu;
            let toggle;
            forEach(this.elementRef.nativeElement.children, (element) => {
                if (element.classList.contains('dropdown-menu')) {
                    menu = element;
                }
                if (element.classList.contains('dropdown-toggle')) {
                    toggle = element;
                }
            });
            if (menu && toggle) {
                this.renderer.setStyle(menu, this.VISIBILITY, 'hidden');
                this.shouldDropup(menu, toggle);
                this.renderer.removeStyle(menu, this.VISIBILITY);
            }
        }
    }
    shouldDropup(container, position) {
        const containerHeight = container.getBoundingClientRect().height;
        const containerDropPosition = position.getBoundingClientRect().bottom;
        const layoutViewport = window.innerHeight;
        const sum = containerHeight + containerDropPosition + this.MARGIN_BOTTOM;
        if (sum >= layoutViewport) {
            this.renderer.addClass(this.elementRef.nativeElement, this.DROPUP);
        }
        else if (this.elementRef.nativeElement.classList.contains(this.DROPUP)) {
            this.renderer.removeClass(this.elementRef.nativeElement, this.DROPUP);
        }
    }
}
DropdownDirectionDirective.ɵfac = function DropdownDirectionDirective_Factory(t) { return new (t || DropdownDirectionDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
DropdownDirectionDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: DropdownDirectionDirective, selectors: [["", "dropdown", "", "c8yBsDropdownDirection", ""], ["", "dropdown", "", "c8yDropdownDirection", ""]], hostBindings: function DropdownDirectionDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function DropdownDirectionDirective_click_HostBindingHandler() { return ctx.onClick(); });
    } } });
DropdownDirectionDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
DropdownDirectionDirective.propDecorators = {
    onClick: [{ type: HostListener, args: ['click',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DropdownDirectionDirective, [{
        type: Directive,
        args: [{ selector: '[dropdown][c8yBsDropdownDirection],[dropdown][c8yDropdownDirection]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { onClick: [{
            type: HostListener,
            args: ['click']
        }] }); })();
