// @c8y/ngx-components/esm2015/core/common/humanize.pipe.js
import { Pipe } from '@angular/core';
import { upperFirst } from 'lodash-es';
/*
 * Humanize any string
 * Usage:
 *   'something' | humanize
*/
import * as ɵngcc0 from '@angular/core';
export class HumanizePipe {
    static humanize(str = '') {
        str = str
            .replace(/\./g, ' ')
            .replace(/c8y[\s,_]*/gi, '')
            .replace(/_/g, ' ')
            .replace(/([A-Z][a-z])/g, ' $1')
            .replace(/^\$\.?/, '')
            .replace(/^\s*/, '')
            .replace(/\s*$/, '')
            .replace(/\s+/g, ' ');
        return upperFirst(str);
    }
    transform(str) {
        return HumanizePipe.humanize(str);
    }
}
HumanizePipe.ɵfac = function HumanizePipe_Factory(t) { return new (t || HumanizePipe)(); };
HumanizePipe.ɵpipe = /*@__PURE__*/ ɵngcc0.ɵɵdefinePipe({ name: "humanize", type: HumanizePipe, pure: true });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HumanizePipe, [{
        type: Pipe,
        args: [{ name: 'humanize' }]
    }], null, null); })();
