// @c8y/ngx-components/esm2015/core/range-display/range-display.module.js
import { NgModule } from '@angular/core';
import { CommonModule as C8yCommonModule } from '../common/common.module';
import { RangeDisplayComponent } from './range-display.component';
/**
 * The angular module definition for a range-display.
 * @exports RangeDisplayModule
 */
import * as ɵngcc0 from '@angular/core';
export class RangeDisplayModule {
}
RangeDisplayModule.ɵfac = function RangeDisplayModule_Factory(t) { return new (t || RangeDisplayModule)(); };
RangeDisplayModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: RangeDisplayModule });
RangeDisplayModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[C8yCommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(RangeDisplayModule, [{
        type: NgModule,
        args: [{
                imports: [C8yCommonModule],
                exports: [RangeDisplayComponent],
                declarations: [RangeDisplayComponent],
                entryComponents: [RangeDisplayComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RangeDisplayModule, { declarations: function () { return [RangeDisplayComponent]; }, imports: function () { return [C8yCommonModule]; }, exports: function () { return [RangeDisplayComponent]; } }); })();
