// @c8y/ngx-components/esm2015/connectivity/impact/impact.guard.js
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
export class ImpactGuard {
    constructor(applicationService) {
        this.applicationService = applicationService;
    }
    canActivate(route, state) {
        if (this.showImpactProvider === undefined) {
            this.showImpactProvider = this.applicationService
                .isAvailable(ImpactGuard.applicationName)
                .then(result => result && result.data);
        }
        return this.showImpactProvider;
    }
}
ImpactGuard.ɵfac = function ImpactGuard_Factory(t) { return new (t || ImpactGuard)(ɵngcc0.ɵɵinject(ɵngcc1.ApplicationService)); };
ImpactGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: ImpactGuard, factory: ImpactGuard.ɵfac });
ImpactGuard.applicationName = 'impact';
ImpactGuard.ctorParameters = () => [
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ImpactGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.ApplicationService }]; }, null); })();
