// @c8y/ngx-components/esm2015/core/bootstrap/cookie-banner/cookie-banner-preferences-modal/cookie-banner-preferences-modal.component.js
import { __awaiter } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
import { Status } from '../../../common/status.model';
import { gettext } from '../../../i18n/gettext';
import { CookieBannerService } from '../cookie-banner.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../cookie-banner.service';
import * as ɵngcc2 from '../../../modal/modal-confirm.component';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/forms';
import * as ɵngcc5 from '../../../common/humanize.pipe';
import * as ɵngcc6 from '../../../i18n/c8y-translate.pipe';

const _c0 = ["modalRef"];
function CookieBannerPreferencesModalComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵelementStart(1, "div", 4);
    ɵngcc0.ɵɵelementStart(2, "div", 5);
    ɵngcc0.ɵɵelementStart(3, "label", 6);
    ɵngcc0.ɵɵelementStart(4, "input", 7);
    ɵngcc0.ɵɵlistener("ngModelChange", function CookieBannerPreferencesModalComponent_div_2_Template_input_ngModelChange_4_listener($event) { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r4); const item_r2 = restoredCtx.$implicit; return item_r2.value = $event; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(5, "span");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 8);
    ɵngcc0.ɵɵelementStart(7, "h4", 9);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "humanize");
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "p");
    ɵngcc0.ɵɵtext(12);
    ɵngcc0.ɵɵpipe(13, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const item_r2 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("name", item_r2.name);
    ɵngcc0.ɵɵproperty("ngModel", item_r2.value)("disabled", item_r2.isReadonly);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 5, ɵngcc0.ɵɵpipeBind1(10, 7, item_r2.name)), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(13, 9, ctx_r1.getCookieDescription(item_r2.name)), " ");
} }
export class CookieBannerPreferencesModalComponent {
    constructor(cookieBannerService) {
        this.cookieBannerService = cookieBannerService;
        this.closeSubject = new Subject();
        this.title = gettext('Cookie preferences');
        this.labels = { ok: gettext('Save settings'), cancel: gettext('Cancel') };
        this.status = Status.INFO;
    }
    ngOnInit() {
        this.configCookiePreferencesList = this.cookieBannerService.transformCookiePreferencesToList(this.configCookiePreferences);
    }
    ngAfterViewInit() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.modalRef.result;
                this.onClose();
            }
            catch (error) {
                this.onDismiss();
            }
        });
    }
    onClose() {
        const cookiePref = this.cookieBannerService.transformCookiePreferencesListToCookiePreferences(this.configCookiePreferencesList);
        this.closeSubject.next(cookiePref);
        this.closeSubject.complete();
    }
    onDismiss() {
        this.closeSubject.next(undefined);
        this.closeSubject.complete();
    }
    getCookieDescription(cookieType) {
        return this.cookieBannerService.getCookieDescription(cookieType);
    }
}
CookieBannerPreferencesModalComponent.ɵfac = function CookieBannerPreferencesModalComponent_Factory(t) { return new (t || CookieBannerPreferencesModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CookieBannerService)); };
CookieBannerPreferencesModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: CookieBannerPreferencesModalComponent, selectors: [["c8y-cookie-banner-preferences-modal"]], viewQuery: function CookieBannerPreferencesModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.modalRef = _t.first);
    } }, decls: 3, vars: 4, consts: [[3, "title", "status", "labels"], ["modalRef", ""], ["class", "row p-t-8 p-b-8 separator-top", 4, "ngFor", "ngForOf"], [1, "row", "p-t-8", "p-b-8", "separator-top"], [1, "col-xs-2"], [1, "text-center"], [1, "c8y-switch"], ["type", "checkbox", 3, "name", "ngModel", "disabled", "ngModelChange"], [1, "col-xs-10"], [1, "text-medium"]], template: function CookieBannerPreferencesModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-confirm-modal", 0, 1);
        ɵngcc0.ɵɵtemplate(2, CookieBannerPreferencesModalComponent_div_2_Template, 14, 11, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("title", ctx.title)("status", ctx.status)("labels", ctx.labels);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.configCookiePreferencesList);
    } }, directives: [ɵngcc2.ConfirmModalComponent, ɵngcc3.NgForOf, ɵngcc4.CheckboxControlValueAccessor, ɵngcc4.NgControlStatus, ɵngcc4.NgModel], pipes: [ɵngcc5.HumanizePipe, ɵngcc6.C8yTranslatePipe], encapsulation: 2 });
CookieBannerPreferencesModalComponent.ctorParameters = () => [
    { type: CookieBannerService }
];
CookieBannerPreferencesModalComponent.propDecorators = {
    modalRef: [{ type: ViewChild, args: ['modalRef', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CookieBannerPreferencesModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-cookie-banner-preferences-modal',
                template: "<c8y-confirm-modal [title]=\"title\" [status]=\"status\" [labels]=\"labels\" #modalRef>\n\n  <div *ngFor=\"let item of configCookiePreferencesList\" class=\"row p-t-8 p-b-8 separator-top\">\n    <div class=\"col-xs-2\">\n      <div class=\"text-center\">\n        <label class=\"c8y-switch\">\n          <input\n            name=\"{{ item.name }}\"\n            type=\"checkbox\"\n            [(ngModel)]=\"item.value\"\n            [disabled]=\"item.isReadonly\"\n          />\n          <span></span>\n        </label>\n      </div>\n    </div>\n    <div class=\"col-xs-10\">\n      <h4 class=\"text-medium\">\n        {{ item.name | translate | humanize }}\n      </h4>\n      <p>\n        {{ getCookieDescription(item.name) | translate }}\n      </p>\n    </div>\n  </div>\n</c8y-confirm-modal>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CookieBannerService }]; }, { modalRef: [{
            type: ViewChild,
            args: ['modalRef', { static: false }]
        }] }); })();
