// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation.component.js
import { Component, Input } from '@angular/core';
import { operationStatusClasses, operationStatusIcons } from '@c8y/ngx-components';
import { SingleOperationDetailsService } from './single-operation-details.service';
/**
 * Single operation component displays a single operation's status, description and creationTime.
 *
 * Example:
 *
 * ```html
 * <c8y-single-operation [operation]="operation"></c8y-single-operation>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './single-operation-details.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from 'ngx-bootstrap/tooltip';

function SingleOperationComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "a", 2);
    ɵngcc0.ɵɵelementStart(2, "div", 3);
    ɵngcc0.ɵɵelementStart(3, "div", 4);
    ɵngcc0.ɵɵelement(4, "i", 5);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 6);
    ɵngcc0.ɵɵelementStart(7, "div", 7);
    ɵngcc0.ɵɵelementStart(8, "div", 8);
    ɵngcc0.ɵɵelementStart(9, "button", 9);
    ɵngcc0.ɵɵlistener("click", function SingleOperationComponent_div_0_Template_button_click_9_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.openDetails(); });
    ɵngcc0.ɵɵpipe(10, "translate");
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "div", 10);
    ɵngcc0.ɵɵelementStart(14, "small", 11);
    ɵngcc0.ɵɵelement(15, "i", 12);
    ɵngcc0.ɵɵelementStart(16, "span");
    ɵngcc0.ɵɵtext(17);
    ɵngcc0.ɵɵpipe(18, "c8yDate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r0.statusIcons[ctx_r0.operation.status])("ngClass", ctx_r0.statusClasses[ctx_r0.operation.status])("tooltip", ɵngcc0.ɵɵpipeBind1(5, 6, ctx_r0.operation.status == null ? null : ctx_r0.operation.status.toString()));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(10, 8, ctx_r0.operation.description));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 10, ctx_r0.operation.description), " ");
    ɵngcc0.ɵɵadvance(6);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(18, 12, ctx_r0.operation.creationTime));
} }
export class SingleOperationComponent {
    constructor(singleOperationDetailsService) {
        this.singleOperationDetailsService = singleOperationDetailsService;
    }
    ngOnInit() {
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
    }
    openDetails() {
        this.singleOperationDetailsService.openDetails(this.operation);
    }
}
SingleOperationComponent.ɵfac = function SingleOperationComponent_Factory(t) { return new (t || SingleOperationComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.SingleOperationDetailsService)); };
SingleOperationComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SingleOperationComponent, selectors: [["c8y-single-operation"]], inputs: { operation: "operation" }, decls: 1, vars: 1, consts: [["class", "c8y-list--group", 4, "ngIf"], [1, "c8y-list--group"], [1, "c8y-list__item"], [1, "c8y-list__item__block"], [1, "c8y-list__item__icon"], ["placement", "right", "container", "body", 3, "c8yIcon", "ngClass", "tooltip"], [1, "c8y-list__item__body", "p-r-16"], [1, "content-flex-40"], [1, "col-8"], [1, "btn-clean", "text-truncate", 3, "title", "click"], [1, "col-4", "text-muted"], [1, "icon-flex"], ["c8yIcon", "calendar", 1, "m-r-4"]], template: function SingleOperationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SingleOperationComponent_div_0_Template, 19, 14, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.operation);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.IconDirective, ɵngcc2.NgClass, ɵngcc4.TooltipDirective], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc3.DatePipe], encapsulation: 2 });
SingleOperationComponent.ctorParameters = () => [
    { type: SingleOperationDetailsService }
];
SingleOperationComponent.propDecorators = {
    operation: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SingleOperationComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-single-operation',
                template: "<div class=\"c8y-list--group\" *ngIf=\"operation\">\n  <a class=\"c8y-list__item\">\n    <div class=\"c8y-list__item__block\">\n      <div class=\"c8y-list__item__icon\">\n        <i\n          [c8yIcon]=\"statusIcons[operation.status]\"\n          [ngClass]=\"statusClasses[operation.status]\"\n          [tooltip]=\"operation.status?.toString() | translate\"\n          placement=\"right\"\n          container=\"body\"\n        ></i>\n      </div>\n      <div class=\"c8y-list__item__body p-r-16\">\n        <div class=\"content-flex-40\">\n          <div class=\"col-8\">\n            <button\n              class=\"btn-clean text-truncate\"\n              (click)=\"openDetails()\"\n              title=\"{{ operation.description | translate }}\"\n            >\n              {{ operation.description | translate }}\n            </button>\n          </div>\n          <div class=\"col-4 text-muted\">\n            <small class=\"icon-flex\">\n              <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n              <span>{{ operation.creationTime | c8yDate }}</span>\n            </small>\n          </div>\n        </div>\n      </div>\n    </div>\n  </a>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.SingleOperationDetailsService }]; }, { operation: [{
            type: Input
        }] }); })();
