// @c8y/ngx-components/esm2015/operations/single-operation-details/single-operation-list-item.component.js
import { __awaiter } from "tslib";
import { Component, Input, ViewChild } from '@angular/core';
import { Router } from '@angular/router';
import { assign, get } from 'lodash-es';
import { OperationService, OperationStatus } from '@c8y/client';
import { AlertService, gettext, operationStatusClasses, operationStatusIcons } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from './single-operation-details.component';

const _c0 = ["listItem"];
const _c1 = ["details"];
const _c2 = ["historyOfChanges"];
function SingleOperationListItemComponent_c8y_li_action_28_Template(rf, ctx) { if (rf & 1) {
    const _r9 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 23);
    ɵngcc0.ɵɵlistener("click", function SingleOperationListItemComponent_c8y_li_action_28_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r9); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.cancel(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(1, 1, "Cancel`operation`"));
} }
const _c3 = function (a0) { return { active: a0 }; };
function SingleOperationListItemComponent_li_32_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 3);
    ɵngcc0.ɵɵelementStart(1, "button", 24);
    ɵngcc0.ɵɵlistener("click", function SingleOperationListItemComponent_li_32_Template_button_click_1_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r12); const tab_r10 = restoredCtx.$implicit; const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.selectedTab = tab_r10; });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 25);
    ɵngcc0.ɵɵelementStart(4, "span", 26);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const tab_r10 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c3, ctx_r2.selectedTab === tab_r10));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 4, tab_r10.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", tab_r10.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 6, tab_r10.label));
} }
function SingleOperationListItemComponent_ng_container_33_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function SingleOperationListItemComponent_ng_template_34_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-single-operation-details", 27);
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("operation", ctx_r5.operation)("readOnly", ctx_r5.readOnly);
} }
function SingleOperationListItemComponent_ng_template_36_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-audit-log", 28);
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("source", ctx_r7.operation.id);
} }
const _c4 = function (a0) { return { "c8y-list__item--no-expand": a0 }; };
const _c5 = function (a0, a1) { return { "content-flex-80": a0, "content-flex-50": a1 }; };
export class SingleOperationListItemComponent {
    constructor(router, operationService, alertService) {
        this.router = router;
        this.operationService = operationService;
        this.alertService = alertService;
        this.collapsed = true;
        this.readOnly = false;
        this.tabs = [];
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
        this.OperationStatus = OperationStatus;
    }
    get displayedDescription() {
        const commandText = get(this.operation, 'c8y_Command.text');
        const opDescription = get(this.operation, 'description');
        return commandText || opDescription || gettext('(no command text or description available)');
    }
    ngOnInit() {
        this.statusIcons = operationStatusIcons;
        this.statusClasses = operationStatusClasses;
        this.tabs = [
            {
                label: gettext('Details'),
                icon: 'asterisk',
                template: this.detailsTabTemplate
            },
            {
                label: gettext('History of changes'),
                icon: 'archive',
                template: this.historyOfChangesTabTemplate
            }
        ];
        this.selectedTab = this.tabs[0];
    }
    scheduleAsBulkOperation() {
        this.router.navigateByUrl(`/devicecontrol/single/create-bulk/${this.operation.id}`);
    }
    cancel() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const operationAfterUpdate = (yield this.operationService.update({
                    id: this.operation.id,
                    status: OperationStatus.FAILED,
                    failureReason: gettext('Operation cancelled by user.')
                })).data;
                assign(this.operation, operationAfterUpdate);
                this.alertService.success(gettext('Operation cancelled.'));
            }
            catch (ex) {
                this.alertService.addServerFailure(ex);
            }
        });
    }
}
SingleOperationListItemComponent.ɵfac = function SingleOperationListItemComponent_Factory(t) { return new (t || SingleOperationListItemComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.OperationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
SingleOperationListItemComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SingleOperationListItemComponent, selectors: [["c8y-single-operation-list-item"]], viewQuery: function SingleOperationListItemComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
        ɵngcc0.ɵɵviewQuery(_c1, 7);
        ɵngcc0.ɵɵviewQuery(_c2, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.listItem = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.detailsTabTemplate = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.historyOfChangesTabTemplate = _t.first);
    } }, inputs: { collapsed: "collapsed", readOnly: "readOnly", operation: "operation" }, decls: 38, vars: 28, consts: [[1, "c8y-list__item--double-actions", 3, "ngClass", "collapsed", "id"], ["listItem", ""], [3, "c8yIcon", "ngClass"], [3, "ngClass"], [1, "col-6"], [1, "text-truncate", 3, "title"], [1, "flex-grow"], [1, "col-3"], [1, "m-r-16"], [1, "icon-flex"], ["c8yIcon", "calendar", 1, "m-r-4"], [1, "c8y-list__item__footer"], ["c8yIcon", "exchange", 1, "m-r-4"], [3, "href"], ["icon", "c8y-energy", 3, "label", "click"], ["icon", "times", 3, "label", "click", 4, "ngIf"], [1, "m-b-16"], [1, "tabContainer"], [1, "nav", "nav-tabs", "nav-tabsc8y"], [3, "ngClass", 4, "ngFor", "ngForOf"], [4, "ngTemplateOutlet"], ["details", ""], ["historyOfChanges", ""], ["icon", "times", 3, "label", "click"], [1, "btn-clean", 3, "title", "click"], [3, "c8yIcon"], [1, "txt"], [3, "operation", "readOnly"], [3, "source"]], template: function SingleOperationListItemComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-li", 0, 1);
        ɵngcc0.ɵɵelementStart(2, "c8y-li-icon");
        ɵngcc0.ɵɵelement(3, "i", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "span", 5);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵtext(8);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵelementStart(12, "div", 8);
        ɵngcc0.ɵɵelementStart(13, "small", 9);
        ɵngcc0.ɵɵelement(14, "i", 10);
        ɵngcc0.ɵɵelementStart(15, "span");
        ɵngcc0.ɵɵtext(16);
        ɵngcc0.ɵɵpipe(17, "c8yDate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 11);
        ɵngcc0.ɵɵelementStart(19, "div", 8);
        ɵngcc0.ɵɵelementStart(20, "small", 9);
        ɵngcc0.ɵɵelement(21, "i", 12);
        ɵngcc0.ɵɵelementStart(22, "span");
        ɵngcc0.ɵɵelementStart(23, "a", 13);
        ɵngcc0.ɵɵtext(24);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerStart(25);
        ɵngcc0.ɵɵelementStart(26, "c8y-li-action", 14);
        ɵngcc0.ɵɵlistener("click", function SingleOperationListItemComponent_Template_c8y_li_action_click_26_listener() { return ctx.scheduleAsBulkOperation(); });
        ɵngcc0.ɵɵpipe(27, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(28, SingleOperationListItemComponent_c8y_li_action_28_Template, 2, 3, "c8y-li-action", 15);
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵelementStart(29, "c8y-li-collapse", 16);
        ɵngcc0.ɵɵelementStart(30, "div", 17);
        ɵngcc0.ɵɵelementStart(31, "ul", 18);
        ɵngcc0.ɵɵtemplate(32, SingleOperationListItemComponent_li_32_Template, 7, 10, "li", 19);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(33, SingleOperationListItemComponent_ng_container_33_Template, 1, 0, "ng-container", 20);
        ɵngcc0.ɵɵtemplate(34, SingleOperationListItemComponent_ng_template_34_Template, 1, 2, "ng-template", null, 21, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(36, SingleOperationListItemComponent_ng_template_36_Template, 1, 1, "ng-template", null, 22, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate1("id", "single-operation-", ctx.operation.id, "");
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(23, _c4, !ctx.collapsed))("collapsed", ctx.collapsed);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.statusIcons[ctx.operation.status])("ngClass", ctx.statusClasses[ctx.operation.status]);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(25, _c5, !ctx.readOnly, ctx.readOnly));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(7, 15, ctx.displayedDescription));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 17, ctx.displayedDescription), " ");
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(17, 19, ctx.operation.creationTime), " ");
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate1("href", "#/device/", ctx.operation.deviceId, "", ɵngcc0.ɵɵsanitizeUrl);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.operation.deviceName);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(27, 21, "Schedule as bulk operation"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.readOnly && ctx.operation.status === ctx.OperationStatus.PENDING);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.tabs);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx.selectedTab.template);
    } }, directives: [ɵngcc3.ListItemComponent, ɵngcc4.NgClass, ɵngcc3.ListItemIconComponent, ɵngcc3.IconDirective, ɵngcc3.ListItemActionComponent, ɵngcc4.NgIf, ɵngcc3.ListItemCollapseComponent, ɵngcc4.NgForOf, ɵngcc4.NgTemplateOutlet, ɵngcc5.SingleOperationDetailsComponent, ɵngcc3.ɵbe], pipes: [ɵngcc3.C8yTranslatePipe, ɵngcc3.DatePipe], encapsulation: 2 });
SingleOperationListItemComponent.ctorParameters = () => [
    { type: Router },
    { type: OperationService },
    { type: AlertService }
];
SingleOperationListItemComponent.propDecorators = {
    operation: [{ type: Input }],
    collapsed: [{ type: Input }],
    readOnly: [{ type: Input }],
    listItem: [{ type: ViewChild, args: ['listItem', { static: true },] }],
    detailsTabTemplate: [{ type: ViewChild, args: ['details', { static: true },] }],
    historyOfChangesTabTemplate: [{ type: ViewChild, args: ['historyOfChanges', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SingleOperationListItemComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-single-operation-list-item',
                template: "<c8y-li\n  class=\"c8y-list__item--double-actions\"\n  [ngClass]=\"{ 'c8y-list__item--no-expand': !collapsed }\"\n  [collapsed]=\"collapsed\"\n  #listItem\n  id=\"single-operation-{{ operation.id }}\"\n>\n  <c8y-li-icon>\n    <i [c8yIcon]=\"statusIcons[operation.status]\" [ngClass]=\"statusClasses[operation.status]\"></i>\n  </c8y-li-icon>\n  <div [ngClass]=\"{ 'content-flex-80': !readOnly, 'content-flex-50': readOnly }\">\n    <div class=\"col-6\">\n      <span class=\"text-truncate\" title=\"{{ displayedDescription | translate }}\">\n        {{ displayedDescription | translate }}\n      </span>\n    </div>\n    <div class=\"flex-grow\"></div>\n    <div class=\"col-3\">\n      <div class=\"m-r-16\">\n        <small class=\"icon-flex\">\n          <i c8yIcon=\"calendar\" class=\"m-r-4\"></i>\n          <span>\n            {{ operation.creationTime | c8yDate }}\n          </span>\n        </small>\n      </div>\n    </div>\n  </div>\n  <div class=\"c8y-list__item__footer\">\n    <div class=\"m-r-16\">\n      <small class=\"icon-flex\">\n        <i c8yIcon=\"exchange\" class=\"m-r-4\"></i>\n        <span>\n          <a href=\"#/device/{{ operation.deviceId }}\">{{ operation.deviceName }}</a>\n        </span>\n      </small>\n    </div>\n  </div>\n  <ng-container>\n    <c8y-li-action\n      label=\"{{ 'Schedule as bulk operation' | translate }}\"\n      (click)=\"scheduleAsBulkOperation()\"\n      icon=\"c8y-energy\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      *ngIf=\"!readOnly && operation.status === OperationStatus.PENDING\"\n      label=\"{{ 'Cancel`operation`' | translate }}\"\n      (click)=\"cancel()\"\n      icon=\"times\"\n    >\n    </c8y-li-action>\n  </ng-container>\n  <c8y-li-collapse class=\"m-b-16\">\n    <div class=\"tabContainer\">\n      <ul class=\"nav nav-tabs nav-tabsc8y\">\n        <li [ngClass]=\"{ active: selectedTab === tab }\" *ngFor=\"let tab of tabs\">\n          <button [title]=\"tab.label | translate\" (click)=\"selectedTab = tab\" class=\"btn-clean\">\n            <i [c8yIcon]=\"tab.icon\"></i>\n            <span class=\"txt\">{{ tab.label | translate }}</span>\n          </button>\n        </li>\n      </ul>\n    </div>\n\n    <ng-container *ngTemplateOutlet=\"selectedTab.template\"></ng-container>\n\n    <ng-template #details>\n      <c8y-single-operation-details [operation]=\"operation\" [readOnly]=\"readOnly\">\n      </c8y-single-operation-details>\n    </ng-template>\n\n    <ng-template #historyOfChanges>\n      <c8y-audit-log [source]=\"operation.id\"></c8y-audit-log>\n    </ng-template>\n  </c8y-li-collapse>\n</c8y-li>\n"
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: ɵngcc2.OperationService }, { type: ɵngcc3.AlertService }]; }, { collapsed: [{
            type: Input
        }], readOnly: [{
            type: Input
        }], operation: [{
            type: Input
        }], listItem: [{
            type: ViewChild,
            args: ['listItem', { static: true }]
        }], detailsTabTemplate: [{
            type: ViewChild,
            args: ['details', { static: true }]
        }], historyOfChangesTabTemplate: [{
            type: ViewChild,
            args: ['historyOfChanges', { static: true }]
        }] }); })();
