// @c8y/ngx-components/esm2015/repository/software/columns/name.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@angular/router';

const _c0 = function (a0) { return [a0]; };
export class NameCellRendererComponent {
    constructor(context) {
        this.context = context;
    }
}
NameCellRendererComponent.ɵfac = function NameCellRendererComponent_Factory(t) { return new (t || NameCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext)); };
NameCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NameCellRendererComponent, selectors: [["ng-component"]], decls: 2, vars: 5, consts: [[1, "interact", 3, "routerLink", "title"]], template: function NameCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "a", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("routerLink", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx.context.item.id))("title", ctx.context.item.name);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.context.item.name, " ");
    } }, directives: [ɵngcc2.RouterLinkWithHref], encapsulation: 2 });
NameCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NameCellRendererComponent, [{
        type: Component,
        args: [{
                template: `
    <a [routerLink]="[context.item.id]" [title]="context.item.name" class="interact">
      {{ context.item.name }}
    </a>
  `
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }]; }, null); })();
