// @c8y/ngx-components/esm2015/ecosystem/packages/add-package.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ApplicationType } from '@c8y/client';
import { EcosystemService } from '../ecosystem.service';
import { gettext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../ecosystem.service';
import * as ɵngcc2 from '../shared/add-application.component';
export class AddPackageComponent {
    constructor(ecosystemService) {
        this.ecosystemService = ecosystemService;
        this.headerText = gettext('Add package');
        this.successText = gettext('Package created');
        this.createPackageAppHandler = (f) => this.createPackageApp(f);
        this.uploadPackageHandler = (f, app) => this.uploadPackage(f, app);
    }
    createPackageApp(file) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.ecosystemService.isValidAppType(file, ApplicationType.HOSTED);
            return this.ecosystemService.createAppForArchive(file, true);
        });
    }
    uploadPackage(file, pckg) {
        return this.ecosystemService.uploadArchiveToApp(file, pckg);
    }
}
AddPackageComponent.ɵfac = function AddPackageComponent_Factory(t) { return new (t || AddPackageComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService)); };
AddPackageComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AddPackageComponent, selectors: [["c8y-add-package"]], decls: 1, vars: 5, consts: [[3, "headerText", "headerIcon", "successText", "createApplicationHandler", "uploadApplicationHandler"]], template: function AddPackageComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "c8y-add-application", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("headerText", ctx.headerText)("headerIcon", "big-parcel")("successText", ctx.successText)("createApplicationHandler", ctx.createPackageAppHandler)("uploadApplicationHandler", ctx.uploadPackageHandler);
    } }, directives: [ɵngcc2.AddApplicationComponent], encapsulation: 2 });
AddPackageComponent.ctorParameters = () => [
    { type: EcosystemService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AddPackageComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-add-package',
                template: `<c8y-add-application
    [headerText]="headerText"
    [headerIcon]="'big-parcel'"
    [successText]="successText"
    [createApplicationHandler]="createPackageAppHandler"
    [uploadApplicationHandler]="uploadPackageHandler"
  ></c8y-add-application>`
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }]; }, null); })();
