// @c8y/ngx-components/esm2015/core/wizard/wizard-body.component.js
import { Component, ViewChild } from '@angular/core';
import { WizardComponent } from './wizard.component';
import { WizardService } from './wizard.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './wizard.component';
import * as ɵngcc2 from './wizard.service';

const _c0 = ["bodyContent"];
function WizardBodyComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 1);
    ɵngcc0.ɵɵelementStart(1, "div", 2);
    ɵngcc0.ɵɵprojection(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = ["*"];
export class WizardBodyComponent {
    constructor(wizard, wizardService) {
        this.wizard = wizard;
        this.wizardService = wizardService;
    }
    ngOnInit() {
        setTimeout(() => {
            this.wizardService.bodyTemplate$.next({
                template: this.bodyContent,
                wizardId: this.wizard.id
            });
        });
    }
}
WizardBodyComponent.ɵfac = function WizardBodyComponent_Factory(t) { return new (t || WizardBodyComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.WizardComponent), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.WizardService)); };
WizardBodyComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: WizardBodyComponent, selectors: [["c8y-wizard-body"]], viewQuery: function WizardBodyComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.bodyContent = _t.first);
    } }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["bodyContent", ""], [1, "modal-inner-scroll", "animated", "fadeIn"], [1, "modal-body"]], template: function WizardBodyComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, WizardBodyComponent_ng_template_0_Template, 3, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
WizardBodyComponent.ctorParameters = () => [
    { type: WizardComponent },
    { type: WizardService }
];
WizardBodyComponent.propDecorators = {
    bodyContent: [{ type: ViewChild, args: ['bodyContent', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WizardBodyComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-wizard-body',
                template: `
    <ng-template #bodyContent>
      <div class="modal-inner-scroll animated fadeIn">
        <div class="modal-body">
          <ng-content></ng-content>
        </div>
      </div>
    </ng-template>
  `
            }]
    }], function () { return [{ type: ɵngcc1.WizardComponent }, { type: ɵngcc2.WizardService }]; }, { bodyContent: [{
            type: ViewChild,
            args: ['bodyContent', { static: true }]
        }] }); })();
