// @c8y/client/lib/cjs/src/tenant-login-options/ITenantLoginOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserManagementSource = exports.GrantType = exports.TenantLoginOptionType = void 0;
var TenantLoginOptionType;
(function (TenantLoginOptionType) {
    /**
     * OAI-Secure: authentication information is stored in a HttpOnly cookie
     */
    TenantLoginOptionType["OAUTH2_INTERNAL"] = "OAUTH2_INTERNAL";
    /**
     * SSO redirect: allows a user to login with a single 3rd-party authorization server using the OAuth2 protocol.
     */
    TenantLoginOptionType["OAUTH2"] = "OAUTH2";
    /**
     * Basic Auth: authentication information is stored in a session storage
     */
    TenantLoginOptionType["BASIC"] = "BASIC";
})(TenantLoginOptionType = exports.TenantLoginOptionType || (exports.TenantLoginOptionType = {}));
var GrantType;
(function (GrantType) {
    GrantType["AUTHORIZATION_CODE"] = "AUTHORIZATION_CODE";
    GrantType["PASSWORD"] = "PASSWORD";
})(GrantType = exports.GrantType || (exports.GrantType = {}));
var UserManagementSource;
(function (UserManagementSource) {
    UserManagementSource["REMOTE"] = "REMOTE";
    UserManagementSource["INTERNAL"] = "INTERNAL";
})(UserManagementSource = exports.UserManagementSource || (exports.UserManagementSource = {}));
