// @c8y/ngx-components/esm2015/core/audit-log/audit-log.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { AuditLogComponent } from './audit-log.component';
/** Audit module allows to show audits list. */
import * as ɵngcc0 from '@angular/core';
export class AuditLogModule {
}
AuditLogModule.ɵfac = function AuditLogModule_Factory(t) { return new (t || AuditLogModule)(); };
AuditLogModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AuditLogModule });
AuditLogModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AuditLogModule, [{
        type: NgModule,
        args: [{
                declarations: [AuditLogComponent],
                imports: [CommonModule],
                exports: [AuditLogComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AuditLogModule, { declarations: function () { return [AuditLogComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [AuditLogComponent]; } }); })();
