// @c8y/ngx-components/esm2015/operations/bulk-operations-list/modals/bulk-operations-modal.component.js
import { Component, ViewContainerRef, ViewChild, ComponentFactoryResolver } from '@angular/core';
import { BsModalRef } from 'ngx-bootstrap/modal';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';

const _c0 = ["container"];
export class BulkOperationsModalComponent {
    constructor(bsModalRef, resolver) {
        this.bsModalRef = bsModalRef;
        this.resolver = resolver;
        this.showComponent = false;
    }
    cancel() {
        this.bsModalRef.hide();
    }
    createComponent(item) {
        if (item && item.component) {
            this.showComponent = true;
            const factory = this.resolver.resolveComponentFactory(item.component);
            this.container.createComponent(factory);
        }
    }
}
BulkOperationsModalComponent.ɵfac = function BulkOperationsModalComponent_Factory(t) { return new (t || BulkOperationsModalComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver)); };
BulkOperationsModalComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: BulkOperationsModalComponent, selectors: [["c8y-bulk-operations-modal"]], viewQuery: function BulkOperationsModalComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5, ViewContainerRef);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.container = _t.first);
    } }, decls: 23, vars: 5, consts: [[1, "viewport-modal"], [1, "modal-header", "dialog-header"], [1, "c8y-icon", "c8y-icon-energy"], ["translate", ""], [1, "modal-inner-scroll"], [1, "modal-body", "text-center", 3, "hidden"], ["c8y-guide-docs", ""], ["c8y-guide-href", "users-guide/device-management/#bulk-operation-wizard"], [3, "onCustomType"], ["container", ""], [1, "modal-footer", 3, "hidden"], ["type", "button", 1, "btn", "btn-default", 3, "title", "click"]], template: function BulkOperationsModalComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "h1", 2);
        ɵngcc0.ɵɵelementStart(3, "h4", 3);
        ɵngcc0.ɵɵtext(4, "New bulk operation");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "div", 5);
        ɵngcc0.ɵɵelementStart(7, "h4", 3);
        ɵngcc0.ɵɵtext(8, "Select one of the available options");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "p", 6);
        ɵngcc0.ɵɵelementStart(10, "small", 3);
        ɵngcc0.ɵɵdisableBindings();
        ɵngcc0.ɵɵtext(11, " Check the ");
        ɵngcc0.ɵɵelementStart(12, "a", 7);
        ɵngcc0.ɵɵtext(13, "User guide`KEEP_ORIGINAL`");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(14, " for more details.");
        ɵngcc0.ɵɵenableBindings();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "c8y-bulk-operation-type-list", 8);
        ɵngcc0.ɵɵlistener("onCustomType", function BulkOperationsModalComponent_Template_c8y_bulk_operation_type_list_onCustomType_15_listener($event) { return ctx.createComponent($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(16, "template", null, 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 10);
        ɵngcc0.ɵɵelementStart(19, "button", 11);
        ɵngcc0.ɵɵlistener("click", function BulkOperationsModalComponent_Template_button_click_19_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(20, "translate");
        ɵngcc0.ɵɵelementStart(21, "span", 3);
        ɵngcc0.ɵɵtext(22, "Cancel");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("hidden", ctx.showComponent);
        ɵngcc0.ɵɵadvance(12);
        ɵngcc0.ɵɵproperty("hidden", ctx.showComponent);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(20, 3, "Cancel"));
    } }, encapsulation: 2 });
BulkOperationsModalComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: ComponentFactoryResolver }
];
BulkOperationsModalComponent.propDecorators = {
    container: [{ type: ViewChild, args: ['container', { static: false, read: ViewContainerRef },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsModalComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-bulk-operations-modal',
                template: "<div class=\"viewport-modal\">\n  <div class=\"modal-header dialog-header\">\n    <h1 class=\"c8y-icon c8y-icon-energy\"></h1>\n    <h4 translate>New bulk operation</h4>\n  </div>\n  <div class=\"modal-inner-scroll\">\n    <div class=\"modal-body text-center\" [hidden]=\"showComponent\">\n      <h4 translate>Select one of the available options</h4>\n      <p c8y-guide-docs>\n        <small translate ngNonBindable>\n          Check the\n          <a c8y-guide-href=\"users-guide/device-management/#bulk-operation-wizard\"\n            >User guide`KEEP_ORIGINAL`</a\n          >\n          for more details.</small\n        >\n      </p>\n    </div>\n\n    <c8y-bulk-operation-type-list\n      (onCustomType)=\"createComponent($event)\"\n    ></c8y-bulk-operation-type-list>\n    <template #container></template>\n  </div>\n  <div class=\"modal-footer\" [hidden]=\"showComponent\">\n    <button\n      (click)=\"cancel()\"\n      type=\"button\"\n      class=\"btn btn-default\"\n      title=\"{{ 'Cancel' | translate }}\"\n    >\n      <span translate>Cancel</span>\n    </button>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc0.ComponentFactoryResolver }]; }, { container: [{
            type: ViewChild,
            args: ['container', { static: false, read: ViewContainerRef }]
        }] }); })();
