// @c8y/ngx-components/esm2015/operations/bulk-operations-list/modals/bulk-operation-modals.service.js
import { Injectable } from '@angular/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { BulkOperationsModalComponent } from './bulk-operations-modal.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
export class BulkOperationModalsService {
    constructor(bsModalService) {
        this.bsModalService = bsModalService;
    }
    showNewBulkOperationModal() {
        this.bsModalRef = this.bsModalService.show(BulkOperationsModalComponent, {
            backdrop: 'static',
            class: 'modal-sm'
        });
    }
    hideNewBulkOperationModal() {
        if (this.bsModalRef) {
            this.bsModalRef.hide();
        }
    }
}
BulkOperationModalsService.ɵfac = function BulkOperationModalsService_Factory(t) { return new (t || BulkOperationModalsService)(ɵngcc0.ɵɵinject(ɵngcc1.BsModalService)); };
BulkOperationModalsService.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: BulkOperationModalsService, factory: BulkOperationModalsService.ɵfac });
BulkOperationModalsService.ctorParameters = () => [
    { type: BsModalService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationModalsService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BsModalService }]; }, null); })();
