// @c8y/ngx-components/esm2015/device-profile/device-tab-profile/device-tab-profile.component.js
import { __awaiter } from "tslib";
import { Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { pipe } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { DeviceProfileService } from '../device-profile.service';
import { AlertService, gettext, ManagedObjectRealtimeService, OperationRealtimeService } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '../device-profile.service';
import * as ɵngcc3 from '@angular/router';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from '@c8y/ngx-components/operations/single-operation-details';
import * as ɵngcc7 from './device-tab-profile-detail.component';

function DeviceTabProfileComponent_c8y_li_20_Template(rf, ctx) { if (rf & 1) {
    const _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 24);
    ɵngcc0.ɵɵlistener("click", function DeviceTabProfileComponent_c8y_li_20_Template_c8y_li_click_0_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r4); const profile_r2 = restoredCtx.$implicit; const ctx_r3 = ɵngcc0.ɵɵnextContext(); ctx_r3.selectProfile(profile_r2); return ctx_r3.setPipe(""); });
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 25);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const profile_r2 = ctx.$implicit;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", profile_r2.name || "--")("pattern", ctx_r1.pattern);
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
export class DeviceTabProfileComponent {
    constructor(deviceRealtime, deviceProfileService, route, operationRealtime, alertService) {
        this.deviceRealtime = deviceRealtime;
        this.deviceProfileService = deviceProfileService;
        this.route = route;
        this.operationRealtime = operationRealtime;
        this.alertService = alertService;
        this.firmwareItems = [];
        this.softwareItems = [];
        this.configurationItems = [];
        this.pattern = '';
        this.reloading = false;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.device = this.route.snapshot.parent.data.contextData;
            this.getDeviceProfilesAndUpdateProfileItems();
            this.subscribeToManagedObjects();
        });
    }
    getDeviceProfilesAndUpdateProfileItems() {
        return __awaiter(this, void 0, void 0, function* () {
            this.reloading = true;
            this.deviceProfiles = yield this.deviceProfileService.getDeviceProfilesByDeviceType(this.device.type);
            if (this.device.c8y_Profile) {
                const profileId = this.device.c8y_Profile.profileId;
                this.selectedProfile = this.deviceProfiles.data.find(mo => mo.id === profileId);
            }
            this.updateProfileItems(this.device, this.selectedProfile);
            this.operation = yield this.deviceProfileService.getProfileOperation(this.device.id);
            this.subscribeToOperations();
            this.reloading = false;
        });
    }
    selectProfile(mo) {
        this.selectedProfile = mo;
        this.updateProfileItems(this.device, this.selectedProfile);
    }
    createOperation() {
        return __awaiter(this, void 0, void 0, function* () {
            this.operation = yield this.deviceProfileService.createProfileOperation(this.device, this.selectedProfile);
        });
    }
    setPipe(filterStr) {
        this.pattern = filterStr;
        this.filterPipe = pipe(map((data) => {
            return data.filter((mo) => mo.name && mo.name.toLowerCase().indexOf(filterStr.toLowerCase()) > -1);
        }));
    }
    ngOnDestroy() {
        this.operationsSubscription.unsubscribe();
        this.moOnUpdateSubscription.unsubscribe();
        this.moOnDeleteSubscription.unsubscribe();
    }
    updateProfileItems(device, profile) {
        this.firmwareItems = this.deviceProfileService.getFirmwareItems(device, profile);
        this.softwareItems = this.deviceProfileService.getSoftwareItems(device, profile);
        this.configurationItems = this.deviceProfileService.getConfigurationItems(device, profile);
    }
    subscribeToManagedObjects() {
        this.moOnUpdateSubscription = this.deviceRealtime
            .onUpdate$(this.device.id)
            .subscribe((managedObject) => {
            this.updateProfileItems(managedObject, this.selectedProfile);
        });
        this.moOnDeleteSubscription = this.deviceRealtime.onDelete$(this.device.id).subscribe(() => {
            this.alertService.danger(gettext('This device has just been deleted. You will be redirected to "All devices" page now.'));
            window.location.href = '#/device';
        });
    }
    subscribeToOperations() {
        this.operationsSubscription = this.operationRealtime
            .onAll$(this.device.id)
            .pipe(map(({ data }) => data), filter(operation => operation.c8y_DeviceProfile))
            .subscribe(operation => {
            this.operation = operation;
        });
    }
}
DeviceTabProfileComponent.ɵfac = function DeviceTabProfileComponent_Factory(t) { return new (t || DeviceTabProfileComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ManagedObjectRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceProfileService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.OperationRealtimeService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService)); };
DeviceTabProfileComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DeviceTabProfileComponent, selectors: [["c8y-device-tab-profile"]], features: [ɵngcc0.ɵɵProvidersFeature([ManagedObjectRealtimeService])], decls: 50, vars: 64, consts: [[3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], [3, "service"], [1, "card", "card--grid--fullpage", "card--grid--fullpage", "card--grid", "grid__row--2-10--md"], [1, "card--grid", "grid__col--6-6--md"], [1, "bg-white"], [1, "card-header", "separator"], ["translate", "", 1, "card-title"], [1, "p-16"], ["deviceProfileForm", "ngForm"], [1, "input-group"], ["name", "selectProfile", 1, "flex-grow", 3, "ngModel", "placeholder", "allowFreeEntries", "ngModelChange", "onSearch"], ["class", "p-l-8 p-r-8 c8y-list__item--link", 3, "click", 4, "c8yFor", "c8yForOf", "c8yForPipe"], [1, "input-group-btn"], ["translate", "", 1, "btn", "btn-primary", 3, "title", "disabled", "click"], [1, "bg-gray-white"], [1, "card-block"], [3, "operation"], [1, "card--grid__inner-scroll", "flex-col", "no-align-items"], [1, "d-contents"], [1, "d-contents", 3, "sectionTitle", "sectionIcon", "emptyStateText", "emptyStateDetails", "isProfileSelected", "items", "isEmpty"], [1, "d-contents", 3, "sectionTitle", "sectionIcon", "emptyStateText", "emptyStateDetails", "isProfileSelected", "items", "isEmpty", "showTextLabel"], [1, "card--grid", "grid__col--6-6--md", "flex-grow"], [1, "p-l-8", "p-r-8", "c8y-list__item--link", 3, "click"], [3, "text", "pattern"]], template: function DeviceTabProfileComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function DeviceTabProfileComponent_Template_button_click_1_listener() { return ctx.getDeviceProfilesAndUpdateProfileItems(); });
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelement(3, "i", 2);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelement(7, "c8y-realtime-btn", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "div", 4);
        ɵngcc0.ɵɵelementStart(9, "div", 5);
        ɵngcc0.ɵɵelementStart(10, "div", 6);
        ɵngcc0.ɵɵelementStart(11, "div", 7);
        ɵngcc0.ɵɵelementStart(12, "h4", 8);
        ɵngcc0.ɵɵtext(13, "Device profile");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "div", 9);
        ɵngcc0.ɵɵelementStart(15, "form", null, 10);
        ɵngcc0.ɵɵelementStart(17, "div", 11);
        ɵngcc0.ɵɵelementStart(18, "c8y-typeahead", 12);
        ɵngcc0.ɵɵlistener("ngModelChange", function DeviceTabProfileComponent_Template_c8y_typeahead_ngModelChange_18_listener($event) { return ctx.selectedProfile = $event; })("onSearch", function DeviceTabProfileComponent_Template_c8y_typeahead_onSearch_18_listener($event) { return ctx.setPipe($event); });
        ɵngcc0.ɵɵpipe(19, "translate");
        ɵngcc0.ɵɵtemplate(20, DeviceTabProfileComponent_c8y_li_20_Template, 2, 2, "c8y-li", 13);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "div", 14);
        ɵngcc0.ɵɵelementStart(22, "button", 15);
        ɵngcc0.ɵɵlistener("click", function DeviceTabProfileComponent_Template_button_click_22_listener() { return ctx.createOperation(); });
        ɵngcc0.ɵɵpipe(23, "translate");
        ɵngcc0.ɵɵtext(24, " Assign device profile ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(25, "div", 16);
        ɵngcc0.ɵɵelementStart(26, "div", 7);
        ɵngcc0.ɵɵelementStart(27, "h4", 8);
        ɵngcc0.ɵɵtext(28, "Currently installed");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(29, "div", 17);
        ɵngcc0.ɵɵelement(30, "c8y-single-operation", 18);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(31, "div", 19);
        ɵngcc0.ɵɵelementStart(32, "div", 20);
        ɵngcc0.ɵɵelement(33, "c8y-device-tab-profile-detail", 21);
        ɵngcc0.ɵɵpipe(34, "translate");
        ɵngcc0.ɵɵpipe(35, "translate");
        ɵngcc0.ɵɵpipe(36, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(37, "div", 20);
        ɵngcc0.ɵɵelement(38, "c8y-device-tab-profile-detail", 21);
        ɵngcc0.ɵɵpipe(39, "translate");
        ɵngcc0.ɵɵpipe(40, "translate");
        ɵngcc0.ɵɵpipe(41, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(42, "div", 20);
        ɵngcc0.ɵɵelement(43, "c8y-device-tab-profile-detail", 22);
        ɵngcc0.ɵɵpipe(44, "translate");
        ɵngcc0.ɵɵpipe(45, "translate");
        ɵngcc0.ɵɵpipe(46, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(47, "div", 23);
        ɵngcc0.ɵɵelement(48, "div", 6);
        ɵngcc0.ɵɵelement(49, "div", 16);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 36, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(62, _c0, ctx.reloading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 38, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("service", ctx.deviceRealtime);
        ɵngcc0.ɵɵadvance(11);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(19, 40, "Select device profile"));
        ɵngcc0.ɵɵproperty("ngModel", ctx.selectedProfile)("allowFreeEntries", false);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("c8yForOf", ctx.deviceProfiles)("c8yForPipe", ctx.filterPipe);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(23, 42, "Assign device profile"));
        ɵngcc0.ɵɵproperty("disabled", !(ctx.selectedProfile == null ? null : ctx.selectedProfile.id));
        ɵngcc0.ɵɵadvance(8);
        ɵngcc0.ɵɵproperty("operation", ctx.operation);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("sectionTitle", ɵngcc0.ɵɵpipeBind1(34, 44, "Firmware"))("sectionIcon", "c8y-firmware")("emptyStateText", ɵngcc0.ɵɵpipeBind1(35, 46, "No firmware to display."))("emptyStateDetails", ɵngcc0.ɵɵpipeBind1(36, 48, "No firmware assigned."))("isProfileSelected", !!ctx.selectedProfile)("items", ctx.firmwareItems)("isEmpty", !(ctx.selectedProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile.firmware == null ? null : ctx.selectedProfile.c8y_DeviceProfile.firmware.name));
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("sectionTitle", ɵngcc0.ɵɵpipeBind1(39, 50, "Software"))("sectionIcon", "c8y-tools")("emptyStateText", ɵngcc0.ɵɵpipeBind1(40, 52, "No software to display."))("emptyStateDetails", ɵngcc0.ɵɵpipeBind1(41, 54, "No software assigned."))("isProfileSelected", !!ctx.selectedProfile)("items", ctx.softwareItems)("isEmpty", !(ctx.selectedProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile.software) || (ctx.selectedProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile.software == null ? null : ctx.selectedProfile.c8y_DeviceProfile.software.length) === 0);
        ɵngcc0.ɵɵadvance(5);
        ɵngcc0.ɵɵproperty("sectionTitle", ɵngcc0.ɵɵpipeBind1(44, 56, "Configuration"))("sectionIcon", "gears")("emptyStateText", ɵngcc0.ɵɵpipeBind1(45, 58, "No configuration to display"))("emptyStateDetails", ɵngcc0.ɵɵpipeBind1(46, 60, "No configuration assigned"))("isProfileSelected", !!ctx.selectedProfile)("items", ctx.configurationItems)("isEmpty", !(ctx.selectedProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile.configuration) || (ctx.selectedProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile == null ? null : ctx.selectedProfile.c8y_DeviceProfile.configuration == null ? null : ctx.selectedProfile.c8y_DeviceProfile.configuration.length) === 0)("showTextLabel", false);
    } }, directives: [ɵngcc1.ActionBarItemComponent, ɵngcc1.IconDirective, ɵngcc4.NgClass, ɵngcc1.RealtimeButtonComponent, ɵngcc1.C8yTranslateDirective, ɵngcc5.ɵNgNoValidate, ɵngcc5.NgControlStatusGroup, ɵngcc5.NgForm, ɵngcc1.TypeaheadComponent, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc1.ForOfDirective, ɵngcc6.SingleOperationComponent, ɵngcc7.DeviceTabProfileDetailComponent, ɵngcc1.ListItemComponent, ɵngcc1.HighlightComponent], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
DeviceTabProfileComponent.ctorParameters = () => [
    { type: ManagedObjectRealtimeService },
    { type: DeviceProfileService },
    { type: ActivatedRoute },
    { type: OperationRealtimeService },
    { type: AlertService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceTabProfileComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-device-tab-profile',
                template: "<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\" (click)=\"getDeviceProfilesAndUpdateProfileItems()\">\n    <i c8yIcon=\"refresh\" [ngClass]=\"{ 'icon-spin': reloading }\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n<c8y-action-bar-item [placement]=\"'right'\">\n  <c8y-realtime-btn [service]=\"deviceRealtime\"></c8y-realtime-btn>\n</c8y-action-bar-item>\n\n<div class=\"card card--grid--fullpage card--grid--fullpage card--grid grid__row--2-10--md\">\n  <div class=\"card--grid grid__col--6-6--md\">\n    <!-- AVAILABLE PROFILES -->\n    <div class=\"bg-white\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Device profile</h4>\n      </div>\n      <div class=\"p-16\">\n        <form #deviceProfileForm=\"ngForm\">\n          <div class=\"input-group\">\n            <c8y-typeahead\n              class=\"flex-grow\"\n              name=\"selectProfile\"\n              [(ngModel)]=\"selectedProfile\"\n              placeholder=\"{{ 'Select device profile' | translate }}\"\n              (onSearch)=\"setPipe($event)\"\n              [allowFreeEntries]=\"false\"\n            >\n              <c8y-li\n                *c8yFor=\"let profile of deviceProfiles; pipe: filterPipe\"\n                class=\"p-l-8 p-r-8 c8y-list__item--link\"\n                (click)=\"selectProfile(profile); setPipe('')\"\n              >\n                <c8y-highlight\n                  [text]=\"profile.name || '&#45;&#45;'\"\n                  [pattern]=\"pattern\"\n                ></c8y-highlight>\n              </c8y-li>\n            </c8y-typeahead>\n            <div class=\"input-group-btn\">\n              <button\n                class=\"btn btn-primary\"\n                (click)=\"createOperation()\"\n                title=\"{{ 'Assign device profile' | translate }}\"\n                [disabled]=\"!selectedProfile?.id\"\n                translate\n              >\n                Assign device profile\n              </button>\n            </div>\n          </div>\n        </form>\n      </div>\n    </div>\n\n    <!-- INSTALL PROFILE OPERATION -->\n    <div class=\"bg-gray-white\">\n      <div class=\"card-header separator\">\n        <h4 class=\"card-title\" translate>Currently installed</h4>\n      </div>\n      <div class=\"card-block\">\n        <c8y-single-operation [operation]=\"operation\"></c8y-single-operation>\n      </div>\n    </div>\n  </div>\n  <div class=\"card--grid__inner-scroll flex-col no-align-items\">\n    <div class=\"d-contents\">\n      <!-- FIRMWARE -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Firmware' | translate\"\n        [sectionIcon]=\"'c8y-firmware'\"\n        [emptyStateText]=\"'No firmware to display.' | translate\"\n        [emptyStateDetails]=\"'No firmware assigned.' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"firmwareItems\"\n        [isEmpty]=\"!selectedProfile?.c8y_DeviceProfile?.firmware?.name\"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <div class=\"d-contents\">\n      <!-- SOFTWARE -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Software' | translate\"\n        [sectionIcon]=\"'c8y-tools'\"\n        [emptyStateText]=\"'No software to display.' | translate\"\n        [emptyStateDetails]=\"'No software assigned.' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"softwareItems\"\n        [isEmpty]=\"\n          !selectedProfile?.c8y_DeviceProfile?.software ||\n          selectedProfile?.c8y_DeviceProfile?.software?.length === 0\n        \"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <div class=\"d-contents\">\n      <!-- CONFIGURATION -->\n      <c8y-device-tab-profile-detail\n        [sectionTitle]=\"'Configuration' | translate\"\n        [sectionIcon]=\"'gears'\"\n        [emptyStateText]=\"'No configuration to display' | translate\"\n        [emptyStateDetails]=\"'No configuration assigned' | translate\"\n        [isProfileSelected]=\"!!selectedProfile\"\n        [items]=\"configurationItems\"\n        [isEmpty]=\"\n          !selectedProfile?.c8y_DeviceProfile?.configuration ||\n          selectedProfile?.c8y_DeviceProfile?.configuration?.length === 0\n        \"\n        [showTextLabel]=\"false\"\n        class=\"d-contents\"\n      ></c8y-device-tab-profile-detail>\n    </div>\n    <!-- fill in the remanining vertical space when empty -->\n    <div class=\"card--grid grid__col--6-6--md flex-grow\">\n      <div class=\"bg-white\"></div>\n      <div class=\"bg-gray-white\"></div>\n    </div>\n  </div>\n</div>\n",
                providers: [ManagedObjectRealtimeService]
            }]
    }], function () { return [{ type: ɵngcc1.ManagedObjectRealtimeService }, { type: ɵngcc2.DeviceProfileService }, { type: ɵngcc3.ActivatedRoute }, { type: ɵngcc1.OperationRealtimeService }, { type: ɵngcc1.AlertService }]; }, null); })();
