// @c8y/ngx-components/esm2015/core/data-grid/data-grid.module.js
import { CdkTableModule } from '@angular/cdk/table';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { PaginationModule } from 'ngx-bootstrap/pagination';
import { CommonModule as C8YCommonModule } from '../common/common.module';
import { DataGridComponent } from './data-grid.component';
import { ColumnDirective } from './column/column.directive';
import { CellRendererDefDirective, HeaderCellRendererDefDirective } from './column/cell-renderer';
import { CellRendererComponent } from './column/cell-renderer.component';
import { FilteringFormRendererDefDirective } from './column/filtering-form-renderer';
import { FilteringFormRendererComponent } from './column/filtering-form-renderer.component';
import { IsControlVisiblePipe } from './is-control-visible.pipe';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/dropdown';
import * as ɵngcc2 from 'ngx-bootstrap/popover';
import * as ɵngcc3 from 'ngx-bootstrap/tooltip';
import * as ɵngcc4 from 'ngx-bootstrap/pagination';
export class DataGridModule {
}
DataGridModule.ɵfac = function DataGridModule_Factory(t) { return new (t || DataGridModule)(); };
DataGridModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DataGridModule });
DataGridModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            CdkTableModule,
            DragDropModule,
            FormsModule,
            ReactiveFormsModule,
            C8YCommonModule,
            BsDropdownModule.forRoot(),
            PopoverModule.forRoot(),
            TooltipModule.forRoot(),
            PaginationModule.forRoot()
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DataGridModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    CdkTableModule,
                    DragDropModule,
                    FormsModule,
                    ReactiveFormsModule,
                    C8YCommonModule,
                    BsDropdownModule.forRoot(),
                    PopoverModule.forRoot(),
                    TooltipModule.forRoot(),
                    PaginationModule.forRoot()
                ],
                declarations: [
                    CellRendererDefDirective,
                    HeaderCellRendererDefDirective,
                    CellRendererComponent,
                    FilteringFormRendererDefDirective,
                    FilteringFormRendererComponent,
                    ColumnDirective,
                    DataGridComponent,
                    IsControlVisiblePipe
                ],
                exports: [
                    CellRendererDefDirective,
                    HeaderCellRendererDefDirective,
                    FilteringFormRendererDefDirective,
                    ColumnDirective,
                    DataGridComponent,
                    IsControlVisiblePipe
                ],
                entryComponents: [DataGridComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DataGridModule, { declarations: function () { return [CellRendererDefDirective,
        HeaderCellRendererDefDirective,
        CellRendererComponent,
        FilteringFormRendererDefDirective,
        FilteringFormRendererComponent,
        ColumnDirective,
        DataGridComponent,
        IsControlVisiblePipe]; }, imports: function () { return [CommonModule,
        CdkTableModule,
        DragDropModule,
        FormsModule,
        ReactiveFormsModule,
        C8YCommonModule, ɵngcc1.BsDropdownModule, ɵngcc2.PopoverModule, ɵngcc3.TooltipModule, ɵngcc4.PaginationModule]; }, exports: function () { return [CellRendererDefDirective,
        HeaderCellRendererDefDirective,
        FilteringFormRendererDefDirective,
        ColumnDirective,
        DataGridComponent,
        IsControlVisiblePipe]; } }); })();
