// @c8y/ngx-components/esm2015/auth-configuration/factories/navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode, Permissions } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class NavigationFactory {
    constructor(permissions) {
        this.permissions = permissions;
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            const userHasPermission = this.permissions.hasAnyRole([
                'ROLE_TENANT_ADMIN',
                'ROLE_TENANT_MANAGEMENT_ADMIN'
            ]);
            if (this.navs.length === 0 && userHasPermission) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Authentication'),
                    icon: 'c8y-shield',
                    path: '/auth-configuration',
                    parent: gettext('Settings'),
                    priority: 2000
                }));
            }
            return this.navs;
        });
    }
}
NavigationFactory.ɵfac = function NavigationFactory_Factory(t) { return new (t || NavigationFactory)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions)); };
NavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: NavigationFactory, factory: NavigationFactory.ɵfac });
NavigationFactory.ctorParameters = () => [
    { type: Permissions }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavigationFactory, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Permissions }]; }, null); })();
