// @c8y/ngx-components/esm2015/core/quicklink/quick-link.component.js
import { Component, Input } from '@angular/core';
import { kebabCase } from 'lodash-es';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../common/icon.directive';
import * as ɵngcc2 from '../i18n/c8y-translate.pipe';
export class QuickLinkComponent {
    constructor() {
        this.classes = '';
    }
    ngOnInit() {
        this.classes = `card-block text-center ${kebabCase(this.label)}`;
    }
}
QuickLinkComponent.ɵfac = function QuickLinkComponent_Factory(t) { return new (t || QuickLinkComponent)(); };
QuickLinkComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: QuickLinkComponent, selectors: [["c8y-quick-link"]], inputs: { icon: "icon", label: "label" }, decls: 7, vars: 9, consts: [[3, "title"], [1, "text-gray-darker"], [1, "c8y-icon-duocolor", 3, "c8yIcon"], [1, "text-muted"]], template: function QuickLinkComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵpipe(1, "translate");
        ɵngcc0.ɵɵelementStart(2, "h1", 1);
        ɵngcc0.ɵɵelement(3, "i", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "p", 3);
        ɵngcc0.ɵɵtext(5);
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.classes);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 5, ctx.label));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("c8yIcon", ctx.icon);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 7, ctx.label));
    } }, directives: [ɵngcc1.IconDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
QuickLinkComponent.propDecorators = {
    icon: [{ type: Input }],
    label: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(QuickLinkComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-quick-link',
                template: "<div [class]=\"classes\" title=\"{{ label | translate }}\">\r\n  <h1 class=\"text-gray-darker\">\r\n    <i [c8yIcon]=\"icon\" class=\"c8y-icon-duocolor\"></i>\r\n  </h1>\r\n  <p class=\"text-muted\">{{ label | translate }}</p>\r\n</div>\r\n"
            }]
    }], function () { return []; }, { icon: [{
            type: Input
        }], label: [{
            type: Input
        }] }); })();
