// @c8y/ngx-components/esm2015/core/quicklink/quick-link.module.js
import { CoreModule } from '../core.module';
import { NgModule } from '@angular/core';
import { QuickLinkComponent } from './quick-link.component';
import * as ɵngcc0 from '@angular/core';
export class QuickLinkModule {
}
QuickLinkModule.ɵfac = function QuickLinkModule_Factory(t) { return new (t || QuickLinkModule)(); };
QuickLinkModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: QuickLinkModule });
QuickLinkModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [], imports: [[CoreModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(QuickLinkModule, [{
        type: NgModule,
        args: [{
                declarations: [QuickLinkComponent],
                imports: [CoreModule],
                exports: [QuickLinkComponent],
                providers: []
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(QuickLinkModule, { declarations: function () { return [QuickLinkComponent]; }, imports: function () { return [CoreModule]; }, exports: function () { return [QuickLinkComponent]; } }); })();
