// @c8y/ngx-components/esm2015/core/action/action.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { CommonModule } from '../common/common.module';
import { ActionOutletComponent } from './action-outlet.component';
import { ActionComponent } from './action.component';
/**
 * The angular module definition for actions.
 * @exports ActionComponent
 * @exports ActionOutletComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class ActionModule {
}
ActionModule.ɵfac = function ActionModule_Factory(t) { return new (t || ActionModule)(); };
ActionModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: ActionModule });
ActionModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            BsDropdownModule,
            RouterModule.forChild([])
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ActionOutletComponent,
                    ActionComponent
                ],
                imports: [
                    CommonModule,
                    BsDropdownModule,
                    RouterModule.forChild([])
                ],
                exports: [
                    ActionOutletComponent,
                    ActionComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ActionModule, { declarations: function () { return [ActionOutletComponent,
        ActionComponent]; }, imports: function () { return [CommonModule,
        BsDropdownModule, ɵngcc1.RouterModule]; }, exports: function () { return [ActionOutletComponent,
        ActionComponent]; } }); })();
