// @c8y/ngx-components/esm2015/core/login/login.module.js
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { FormsModule as C8yFormsModule } from '../forms/forms.module';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
import { ChangePasswordComponent } from './change-password.component';
import { CredentialsComponent } from './credentials.component';
import { LoginComponent } from './login.component';
import { LoginService } from './login.service';
import { RecoverPasswordComponent } from './recover-password.component';
import { AuthenticationModule } from '../authentication/authentication.module';
import { AlertModule } from '../alert/alert.module';
import { TotpAuthComponent } from './totp-auth.component';
import { PasswordStrengthValidatorDirective } from './password-strength-validator.directive';
import { TenantIdSetupComponent } from './tenant-id-setup.component';
/**
 * The angular module definition for the login.
 * @exports LoginComponent
 */
import * as ɵngcc0 from '@angular/core';
export class LoginModule {
    static forRoot() {
        return {
            ngModule: LoginModule,
            providers: LoginModule.providers()
        };
    }
    static providers() {
        return [LoginService];
    }
}
LoginModule.ɵfac = function LoginModule_Factory(t) { return new (t || LoginModule)(); };
LoginModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: LoginModule });
LoginModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            FormsModule,
            TooltipModule,
            CommonModule,
            AuthenticationModule,
            AlertModule,
            C8yFormsModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LoginModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LoginComponent,
                    RecoverPasswordComponent,
                    ChangePasswordComponent,
                    CredentialsComponent,
                    TotpAuthComponent,
                    PasswordStrengthValidatorDirective,
                    TenantIdSetupComponent
                ],
                imports: [
                    FormsModule,
                    TooltipModule,
                    CommonModule,
                    AuthenticationModule,
                    AlertModule,
                    C8yFormsModule
                ],
                exports: [LoginComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(LoginModule, { declarations: function () { return [LoginComponent,
        RecoverPasswordComponent,
        ChangePasswordComponent,
        CredentialsComponent,
        TotpAuthComponent,
        PasswordStrengthValidatorDirective,
        TenantIdSetupComponent]; }, imports: function () { return [FormsModule,
        TooltipModule,
        CommonModule,
        AuthenticationModule,
        AlertModule,
        C8yFormsModule]; }, exports: function () { return [LoginComponent]; } }); })();
