// @c8y/ngx-components/esm2015/core/login/password-strength-validator.directive.js
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { StrengthValidatorService } from './strength-validator-service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './strength-validator-service';
export class PasswordStrengthValidatorDirective {
    constructor(passwordService) {
        this.passwordService = passwordService;
    }
    set passwordStrengthEnforced(value) {
        this.forced = value;
    }
    validate(control) {
        const strengthFulfilled = this.passwordService.isStrong(control.value || '');
        const enforcementForcedAndNotFulfilled = this.forced && !strengthFulfilled;
        return enforcementForcedAndNotFulfilled ? { passwordStrength: true } : null;
    }
}
PasswordStrengthValidatorDirective.ɵfac = function PasswordStrengthValidatorDirective_Factory(t) { return new (t || PasswordStrengthValidatorDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.StrengthValidatorService)); };
PasswordStrengthValidatorDirective.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: PasswordStrengthValidatorDirective, selectors: [["", "passwordStrengthEnforced", ""]], inputs: { passwordStrengthEnforced: "passwordStrengthEnforced" }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: NG_VALIDATORS, useExisting: PasswordStrengthValidatorDirective, multi: true }])] });
PasswordStrengthValidatorDirective.ctorParameters = () => [
    { type: StrengthValidatorService }
];
PasswordStrengthValidatorDirective.propDecorators = {
    passwordStrengthEnforced: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(PasswordStrengthValidatorDirective, [{
        type: Directive,
        args: [{
                selector: '[passwordStrengthEnforced]',
                providers: [{ provide: NG_VALIDATORS, useExisting: PasswordStrengthValidatorDirective, multi: true }]
            }]
    }], function () { return [{ type: ɵngcc1.StrengthValidatorService }]; }, { passwordStrengthEnforced: [{
            type: Input
        }] }); })();
