// @c8y/ngx-components/esm2015/core/login/totp-auth.component.js
import { __awaiter } from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { UserService } from '@c8y/client';
import { AlertService } from '../alert/alert.service';
import { LoginService } from './login.service';
import { LoginViews } from './login.model';
import { gettext } from '../i18n/gettext';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './login.service';
import * as ɵngcc2 from '@c8y/client';
import * as ɵngcc3 from '../alert/alert.service';
import * as ɵngcc4 from '../i18n/c8y-translate.directive';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from '../authentication/totp-challenge.component';
import * as ɵngcc7 from '../authentication/totp-setup.component';
import * as ɵngcc8 from '../i18n/c8y-translate.pipe';

function TotpAuthComponent_c8y_totp_setup_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-totp-setup");
} }
export class TotpAuthComponent {
    constructor(loginService, userService, alert) {
        this.loginService = loginService;
        this.userService = userService;
        this.alert = alert;
        this.onCancel = new EventEmitter();
        this.LOGIN_VIEWS = LoginViews;
        this.loading = false;
        this.hasError = false;
        this.isSetup = false;
    }
    /**
     * In case of a setup we need the user to be authorized
     * first.
     */
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.view === this.LOGIN_VIEWS.TotpSetup) {
                yield this.loginService.switchLoginMode(this.credentials);
                this.isSetup = true;
            }
        });
    }
    onTotpSuccess(code) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.loading = true;
                this.hasError = false;
                this.credentials.tfa = code;
                if (this.isSetup) {
                    yield this.userService.activateTotp();
                }
                yield this.loginService.switchLoginMode(this.credentials);
                yield this.loginService.verifyAppAccess();
                yield this.loginService.authFulfilled();
            }
            catch (e) {
                this.alert.removeLastDanger();
                if (e.data && e.data.message === 'Access is denied') {
                    this.alert.addServerFailure(e);
                }
                if (e.data && e.data.message === 'Authentication failed! : User account is locked') {
                    this.alert.warning(gettext('Authentication failed due to: user account is locked.'));
                }
                else {
                    this.hasError = true;
                }
            }
            finally {
                this.loading = false;
            }
        });
    }
}
TotpAuthComponent.ɵfac = function TotpAuthComponent_Factory(t) { return new (t || TotpAuthComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.LoginService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.UserService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService)); };
TotpAuthComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TotpAuthComponent, selectors: [["c8y-totp-auth"]], inputs: { credentials: "credentials", view: "view" }, outputs: { onCancel: "onCancel" }, decls: 8, vars: 7, consts: [["translate", "", 1, "legend", "form-block", "center"], [4, "ngIf"], [3, "loading", "hasError", "verify", "onSuccess"], [1, "text-center", "m-t-8"], ["translate", "", 1, "btn", "btn-link", "btn-sm", 3, "title", "click"]], template: function TotpAuthComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtext(1, " Two-factor authentication\n");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(2, TotpAuthComponent_c8y_totp_setup_2_Template, 1, 0, "c8y-totp-setup", 1);
        ɵngcc0.ɵɵelementStart(3, "c8y-totp-challenge", 2);
        ɵngcc0.ɵɵlistener("onSuccess", function TotpAuthComponent_Template_c8y_totp_challenge_onSuccess_3_listener($event) { return ctx.onTotpSuccess($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "a", 4);
        ɵngcc0.ɵɵlistener("click", function TotpAuthComponent_Template_a_click_5_listener() { return ctx.onCancel.emit(); });
        ɵngcc0.ɵɵpipe(6, "translate");
        ɵngcc0.ɵɵtext(7, " Cancel ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSetup);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("loading", ctx.loading)("hasError", ctx.hasError)("verify", ctx.view === ctx.LOGIN_VIEWS.TotpSetup);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(6, 5, "Cancel"));
    } }, directives: [ɵngcc4.C8yTranslateDirective, ɵngcc5.NgIf, ɵngcc6.TotpChallengeComponent, ɵngcc7.TotpSetupComponent], pipes: [ɵngcc8.C8yTranslatePipe], encapsulation: 2 });
TotpAuthComponent.ctorParameters = () => [
    { type: LoginService },
    { type: UserService },
    { type: AlertService }
];
TotpAuthComponent.propDecorators = {
    credentials: [{ type: Input }],
    view: [{ type: Input }],
    onCancel: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TotpAuthComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-totp-auth',
                template: "<div\n  class=\"legend form-block center\"\n  translate\n>\n  Two-factor authentication\n</div>\n\n<c8y-totp-setup *ngIf=\"isSetup\">\n</c8y-totp-setup>\n<c8y-totp-challenge\n  [loading]=\"loading\"\n  [hasError]=\"hasError\"\n  [verify]=\"view === LOGIN_VIEWS.TotpSetup\"\n  (onSuccess)=\"onTotpSuccess($event)\"\n></c8y-totp-challenge>\n\n<div class=\"text-center m-t-8\">\n  <a\n    title=\"{{ 'Cancel' | translate }}\"\n    class=\"btn btn-link btn-sm\"\n    (click)=\"onCancel.emit()\"\n    translate\n  >\n    Cancel\n  </a>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.LoginService }, { type: ɵngcc2.UserService }, { type: ɵngcc3.AlertService }]; }, { onCancel: [{
            type: Output
        }], credentials: [{
            type: Input
        }], view: [{
            type: Input
        }] }); })();
