// @c8y/ngx-components/esm2015/ecosystem/shared/application-properties-form.component.js
import { __awaiter } from "tslib";
import { Component, Input } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/forms';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';

function ApplicationPropertiesFormComponent_c8y_message_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 11);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesFormComponent_c8y_message_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 11);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
function ApplicationPropertiesFormComponent_c8y_message_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-message", 11);
    ɵngcc0.ɵɵtext(1, " This field is required. ");
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function () { return { example: "1.0.0" }; };
function ApplicationPropertiesFormComponent_c8y_form_group_25_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-form-group");
    ɵngcc0.ɵɵelementStart(1, "label", 12);
    ɵngcc0.ɵɵtext(2, "Use package version");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 7);
    ɵngcc0.ɵɵelement(4, "input", 13);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind2(5, 1, "e.g. {{ example }}", ɵngcc0.ɵɵpureFunction0(4, _c0)));
} }
export class ApplicationPropertiesFormComponent {
    constructor(formBuilder) {
        this.formBuilder = formBuilder;
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.initForm();
            this.setForm(this.application);
        });
    }
    initForm() {
        this.formGroup = this.formBuilder.group({
            name: [Validators.required, Validators.maxLength(120)],
            key: [Validators.required, Validators.maxLength(120)],
            contextPath: [Validators.required, Validators.maxLength(120)]
        });
    }
    setForm(applicationConfig) {
        this.formGroup.patchValue(Object.assign({}, applicationConfig));
    }
}
ApplicationPropertiesFormComponent.ɵfac = function ApplicationPropertiesFormComponent_Factory(t) { return new (t || ApplicationPropertiesFormComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormBuilder)); };
ApplicationPropertiesFormComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ApplicationPropertiesFormComponent, selectors: [["c8y-application-properties-form"]], inputs: { application: "application", isPackage: "isPackage" }, decls: 26, vars: 14, consts: [["name", "appDeployForm", "novalidate", "", 1, "c8y-wizard-form", "text-left", 3, "formGroup", "ngSubmit"], ["for", "name", "translate", ""], ["formControlName", "name", "name", "name", "required", "", "type", "text", 1, "form-control", 3, "placeholder"], ["translate", "", 4, "ngIf"], ["for", "key", "translate", ""], ["formControlName", "key", "name", "key", "required", "", "type", "text", 1, "form-control", 3, "placeholder"], ["for", "contextPath", "translate", ""], [1, "input-group"], [1, "input-group-addon"], ["formControlName", "contextPath", "name", "contextPath", "required", "", "type", "text", 1, "form-control", 3, "placeholder"], [4, "ngIf"], ["translate", ""], ["for", "version", "translate", ""], ["disabled", "", "formControlName", "version", "name", "version", "type", "text", 1, "form-control", 3, "placeholder"]], template: function ApplicationPropertiesFormComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0);
        ɵngcc0.ɵɵlistener("ngSubmit", function ApplicationPropertiesFormComponent_Template_form_ngSubmit_0_listener() { return ctx.formGroup.valid; });
        ɵngcc0.ɵɵelementStart(1, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(2, "label", 1);
        ɵngcc0.ɵɵtext(3, "Name");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(4, "input", 2);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelementStart(6, "c8y-messages");
        ɵngcc0.ɵɵtemplate(7, ApplicationPropertiesFormComponent_c8y_message_7_Template, 2, 0, "c8y-message", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(9, "label", 4);
        ɵngcc0.ɵɵtext(10, "Application key");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(11, "input", 5);
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelementStart(13, "c8y-messages");
        ɵngcc0.ɵɵtemplate(14, ApplicationPropertiesFormComponent_c8y_message_14_Template, 2, 0, "c8y-message", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "c8y-form-group");
        ɵngcc0.ɵɵelementStart(16, "label", 6);
        ɵngcc0.ɵɵtext(17, "Path");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 7);
        ɵngcc0.ɵɵelementStart(19, "span", 8);
        ɵngcc0.ɵɵtext(20, "/apps/");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(21, "input", 9);
        ɵngcc0.ɵɵpipe(22, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(23, "c8y-messages");
        ɵngcc0.ɵɵtemplate(24, ApplicationPropertiesFormComponent_c8y_message_24_Template, 2, 0, "c8y-message", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(25, ApplicationPropertiesFormComponent_c8y_form_group_25_Template, 6, 5, "c8y-form-group", 10);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("formGroup", ctx.formGroup);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(5, 8, "e.g. My application"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.name == null ? null : ctx.formGroup.controls.name.errors == null ? null : ctx.formGroup.controls.name.errors.required);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(12, 10, "e.g. my-application-key"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.key == null ? null : ctx.formGroup.controls.key.errors == null ? null : ctx.formGroup.controls.key.errors.required);
        ɵngcc0.ɵɵadvance(7);
        ɵngcc0.ɵɵpropertyInterpolate("placeholder", ɵngcc0.ɵɵpipeBind1(22, 12, "e.g. my-application"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.formGroup == null ? null : ctx.formGroup.controls == null ? null : ctx.formGroup.controls.contextPath == null ? null : ctx.formGroup.controls.contextPath.errors == null ? null : ctx.formGroup.controls.contextPath.errors.required);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isPackage);
    } }, directives: [ɵngcc1.ɵNgNoValidate, ɵngcc1.NgControlStatusGroup, ɵngcc1.FormGroupDirective, ɵngcc2.FormGroupComponent, ɵngcc2.C8yTranslateDirective, ɵngcc1.DefaultValueAccessor, ɵngcc2.RequiredInputPlaceholderDirective, ɵngcc1.NgControlStatus, ɵngcc1.FormControlName, ɵngcc1.RequiredValidator, ɵngcc2.MessagesComponent, ɵngcc3.NgIf, ɵngcc2.MessageDirective], pipes: [ɵngcc2.C8yTranslatePipe], encapsulation: 2 });
ApplicationPropertiesFormComponent.ctorParameters = () => [
    { type: FormBuilder }
];
ApplicationPropertiesFormComponent.propDecorators = {
    application: [{ type: Input }],
    isPackage: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApplicationPropertiesFormComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-application-properties-form',
                template: "<form\n  (ngSubmit)=\"(formGroup.valid)\"\n  [formGroup]=\"formGroup\"\n  class=\"c8y-wizard-form text-left\"\n  name=\"appDeployForm\"\n  novalidate\n>\n  <c8y-form-group>\n    <label for=\"name\" translate>Name</label>\n    <input\n      class=\"form-control\"\n      formControlName=\"name\"\n      name=\"name\"\n      placeholder=\"{{ 'e.g. My application' | translate }}\"\n      required\n      type=\"text\"\n    />\n    <c8y-messages>\n      <c8y-message *ngIf=\"formGroup?.controls?.name?.errors?.required\" translate>\n        This field is required.\n      </c8y-message>\n    </c8y-messages>\n  </c8y-form-group>\n  <c8y-form-group>\n    <label for=\"key\" translate>Application key</label>\n    <input\n      class=\"form-control\"\n      formControlName=\"key\"\n      name=\"key\"\n      placeholder=\"{{ 'e.g. my-application-key' | translate }}\"\n      required\n      type=\"text\"\n    />\n    <c8y-messages>\n      <c8y-message *ngIf=\"formGroup?.controls?.key?.errors?.required\" translate>\n        This field is required.\n      </c8y-message>\n    </c8y-messages>\n  </c8y-form-group>\n  <c8y-form-group>\n    <label for=\"contextPath\" translate>Path</label>\n    <div class=\"input-group\">\n      <span class=\"input-group-addon\">/apps/</span>\n      <input\n        class=\"form-control\"\n        formControlName=\"contextPath\"\n        name=\"contextPath\"\n        placeholder=\"{{ 'e.g. my-application' | translate }}\"\n        required\n        type=\"text\"\n      />\n    </div>\n    <c8y-messages>\n      <c8y-message *ngIf=\"formGroup?.controls?.contextPath?.errors?.required\" translate>\n        This field is required.\n      </c8y-message>\n    </c8y-messages>\n  </c8y-form-group>\n  <c8y-form-group *ngIf=\"isPackage\">\n    <label for=\"version\" translate>Use package version</label>\n    <div class=\"input-group\">\n      <input\n        class=\"form-control\"\n        disabled\n        formControlName=\"version\"\n        name=\"version\"\n        placeholder=\"{{ 'e.g. \\{\\{ example \\}\\}' | translate:{ example: '1.0.0' } }}\"\n        type=\"text\"\n      />\n    </div>\n  </c8y-form-group>\n</form>\n"
            }]
    }], function () { return [{ type: ɵngcc1.FormBuilder }]; }, { application: [{
            type: Input
        }], isPackage: [{
            type: Input
        }] }); })();
