// @c8y/ngx-components/esm2015/core/drop-area/drop-area.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '../common/common.module';
import { DropAreaComponent } from './drop-area.component';
/**
 * The angular module definition for a drop-area.
 * @exports DropAreaComponent
 */
import * as ɵngcc0 from '@angular/core';
export class DropAreaModule {
}
DropAreaModule.ɵfac = function DropAreaModule_Factory(t) { return new (t || DropAreaModule)(); };
DropAreaModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DropAreaModule });
DropAreaModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DropAreaModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                ],
                exports: [
                    DropAreaComponent,
                ],
                declarations: [
                    DropAreaComponent,
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DropAreaModule, { declarations: function () { return [DropAreaComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [DropAreaComponent]; } }); })();
