// @c8y/ngx-components/esm2015/connectivity/sim/sim.guard.js
import { Injectable } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { Permissions } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/client';
export class SimGuard {
    constructor(permissions, applicationService) {
        this.permissions = permissions;
        this.applicationService = applicationService;
    }
    canActivate(route, state) {
        if (this.showSimProvider === undefined) {
            const hasRole = Promise.resolve(this.permissions.hasRole('ROLE_CONNECTIVITY_READ'));
            this.showSimProvider = hasRole.then((hasPermissions) => hasPermissions &&
                this.applicationService
                    .isAvailable(SimGuard.applicationName)
                    .then(result => result && result.data));
        }
        return this.showSimProvider;
    }
}
SimGuard.ɵfac = function SimGuard_Factory(t) { return new (t || SimGuard)(ɵngcc0.ɵɵinject(ɵngcc1.Permissions), ɵngcc0.ɵɵinject(ɵngcc2.ApplicationService)); };
SimGuard.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: SimGuard, factory: SimGuard.ɵfac });
SimGuard.applicationName = 'connectivity';
SimGuard.ctorParameters = () => [
    { type: Permissions },
    { type: ApplicationService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SimGuard, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Permissions }, { type: ɵngcc2.ApplicationService }]; }, null); })();
