// @c8y/ngx-components/esm2015/repository/configuration/configuration-list.component.js
import { __awaiter, __decorate } from "tslib";
import { Component, ViewChild } from '@angular/core';
import { InventoryBinaryService, InventoryService } from '@c8y/client';
import { AlertService, FilterInputComponent, gettext, memoize, ModalService, Status } from '@c8y/ngx-components';
import { TranslateService } from '@ngx-translate/core';
import { BsModalService } from 'ngx-bootstrap/modal';
import { of, pipe } from 'rxjs';
import { map } from 'rxjs/operators';
import { RepositoryType } from '../repository.model';
import { RepositoryService } from '../repository.service';
import { ConfigurationDetailComponent } from './configuration-detail.component';
import { property } from 'lodash-es';
import { saveAs } from 'file-saver';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '../repository.service';
import * as ɵngcc3 from 'ngx-bootstrap/modal';
import * as ɵngcc4 from '@ngx-translate/core';
import * as ɵngcc5 from '@c8y/client';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from 'ngx-bootstrap/tooltip';

function ConfigurationListComponent_small_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "async");
    ɵngcc0.ɵɵelementStart(3, "span", 0);
    ɵngcc0.ɵɵtext(4, " snapshots ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, ctx_r0.configurations$).data.length, " ");
} }
function ConfigurationListComponent_small_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 13);
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementStart(4, "span", 0);
    ɵngcc0.ɵɵtext(5, "snapshots");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("tooltip", ɵngcc0.ɵɵpipeBind1(1, 2, "More data available. Scroll to the bottom of the list to load it."));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r1.configurations$).paging.pageSize, "+ ");
} }
function ConfigurationListComponent_small_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small");
    ɵngcc0.ɵɵelementStart(1, "span", 0);
    ɵngcc0.ɵɵtext(2, " Search results for ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1("\u00A0\" ", ctx_r2.filterTerm, "\" ");
} }
function ConfigurationListComponent_div_22_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelement(1, "h1", 15);
    ɵngcc0.ɵɵelementStart(2, "h3", 0);
    ɵngcc0.ɵɵtext(3, "There are no configuration snapshots defined");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 0);
    ɵngcc0.ɵɵtext(5, "Add a configuration snapshot first.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div");
    ɵngcc0.ɵɵelementStart(7, "button", 16);
    ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_div_22_Template_button_click_7_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.add(); });
    ɵngcc0.ɵɵtext(8, " Add configuration snapshot ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "p", 17);
    ɵngcc0.ɵɵelementStart(10, "small", 0);
    ɵngcc0.ɵɵdisableBindings();
    ɵngcc0.ɵɵtext(11, " Find out more in the ");
    ɵngcc0.ɵɵelementStart(12, "a", 18);
    ɵngcc0.ɵɵtext(13, " User guide`KEEP_ORIGINAL` ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtext(14, ". ");
    ɵngcc0.ɵɵenableBindings();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationListComponent_div_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 19);
    ɵngcc0.ɵɵelement(1, "h1", 20);
    ɵngcc0.ɵɵelementStart(2, "h3", 0);
    ɵngcc0.ɵɵtext(3, " No results to display. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 0);
    ɵngcc0.ɵɵtext(5, " Refine your search terms or check your spelling. ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-li", 24);
    ɵngcc0.ɵɵelementStart(1, "c8y-li-icon");
    ɵngcc0.ɵɵelement(2, "i", 25);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "c8y-li-body", 26);
    ɵngcc0.ɵɵelementStart(4, "div", 27);
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "div", 28);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(10, "div", 28);
    ɵngcc0.ɵɵtext(11);
    ɵngcc0.ɵɵpipe(12, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(13, "div", 27);
    ɵngcc0.ɵɵtext(14);
    ɵngcc0.ɵɵpipe(15, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(16, "div", 27);
    ɵngcc0.ɵɵtext(17);
    ɵngcc0.ɵɵpipe(18, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(6, 5, "Configuration"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 7, "Description"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 9, "File"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 11, "Device type"), " ");
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(18, 13, "Configuration type"), " ");
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_small_10_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 47);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 32);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r10.description || "-")("pattern", ctx_r11.filterTerm);
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_small_15_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 48);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "async");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r12.getBinaryName(configuration_r10)));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(3, 4, ctx_r12.getBinaryName(configuration_r10)), " ");
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 48);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵpropertyInterpolate("title", configuration_r10.url);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", configuration_r10.url, " ");
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_span_23_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span");
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 32);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r10.deviceType || "-")("pattern", ctx_r15.filterTerm);
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_span_26_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 49);
    ɵngcc0.ɵɵelement(1, "c8y-highlight", 32);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r16 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r10.configurationType)("pattern", ctx_r16.filterTerm);
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_c8y_li_action_31_Template(rf, ctx) { if (rf & 1) {
    const _r29 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li-action", 50);
    ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_c8y_li_action_31_Template_c8y_li_action_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r29); const configuration_r10 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r27 = ɵngcc0.ɵɵnextContext(2); return ctx_r27.download(configuration_r10); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(1, 1, "Download"));
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_32_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 51);
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "Undefined"));
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_34_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 51);
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(3, 1, "No description"));
} }
function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_Template(rf, ctx) { if (rf & 1) {
    const _r31 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-li");
    ɵngcc0.ɵɵelement(1, "c8y-li-icon", 29);
    ɵngcc0.ɵɵelementStart(2, "div", 26);
    ɵngcc0.ɵɵelementStart(3, "button", 30);
    ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_Template_button_click_3_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r31); const configuration_r10 = restoredCtx.$implicit; const ctx_r30 = ɵngcc0.ɵɵnextContext(2); return ctx_r30.edit(configuration_r10); });
    ɵngcc0.ɵɵelementStart(4, "span", 31);
    ɵngcc0.ɵɵelement(5, "c8y-highlight", 32);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(6, "div", 28);
    ɵngcc0.ɵɵelementStart(7, "div", 33);
    ɵngcc0.ɵɵtext(8);
    ɵngcc0.ɵɵpipe(9, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(10, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_small_10_Template, 2, 2, "small", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(11, "div", 28);
    ɵngcc0.ɵɵelementStart(12, "span", 35);
    ɵngcc0.ɵɵelementStart(13, "span", 36);
    ɵngcc0.ɵɵtext(14, " File ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(15, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_small_15_Template, 4, 6, "small", 37);
    ɵngcc0.ɵɵtemplate(16, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_16_Template, 2, 2, "ng-template", null, 38, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(18, "div", 27);
    ɵngcc0.ɵɵelementStart(19, "div", 31);
    ɵngcc0.ɵɵpipe(20, "translate");
    ɵngcc0.ɵɵelementStart(21, "span", 39);
    ɵngcc0.ɵɵtext(22, " Device type ");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(23, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_span_23_Template, 2, 2, "span", 40);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(24, "div", 27);
    ɵngcc0.ɵɵelementStart(25, "div", 31);
    ɵngcc0.ɵɵtemplate(26, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_span_26_Template, 2, 2, "span", 41);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(27, "c8y-li-action", 42);
    ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_Template_c8y_li_action_click_27_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r31); const configuration_r10 = restoredCtx.$implicit; const ctx_r32 = ɵngcc0.ɵɵnextContext(2); return ctx_r32.edit(configuration_r10); });
    ɵngcc0.ɵɵpipe(28, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(29, "c8y-li-action", 43);
    ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_c8y_list_group_26_c8y_li_3_Template_c8y_li_action_click_29_listener() { const restoredCtx = ɵngcc0.ɵɵrestoreView(_r31); const configuration_r10 = restoredCtx.$implicit; const ctx_r33 = ɵngcc0.ɵɵnextContext(2); return ctx_r33.delete(configuration_r10); });
    ɵngcc0.ɵɵpipe(30, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(31, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_c8y_li_action_31_Template, 2, 3, "c8y-li-action", 44);
    ɵngcc0.ɵɵtemplate(32, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_32_Template, 4, 3, "ng-template", null, 45, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(34, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_ng_template_34_Template, 4, 3, "ng-template", null, 46, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const configuration_r10 = ctx.$implicit;
    const _r13 = ɵngcc0.ɵɵreference(17);
    const _r18 = ɵngcc0.ɵɵreference(33);
    const _r20 = ɵngcc0.ɵɵreference(35);
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate("title", configuration_r10.name || "-");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("text", configuration_r10.name || "-")("pattern", ctx_r9.filterTerm);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(9, 18, "Description"), " ");
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", configuration_r10.description)("ngIfElse", _r20);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.isBinaryFile(configuration_r10))("ngIfElse", _r13);
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵpropertyInterpolate2("title", "", ɵngcc0.ɵɵpipeBind1(20, 20, "Device type"), ": ", configuration_r10.deviceType || "-", "");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵproperty("ngIf", configuration_r10.deviceType)("ngIfElse", _r18);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", configuration_r10.configurationType);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", configuration_r10.configurationType)("ngIfElse", _r18);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(28, 22, "Edit"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("label", ɵngcc0.ɵɵpipeBind1(30, 24, "Delete"));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r9.isBinaryFile(configuration_r10));
} }
const _c0 = function (a0) { return { "dd-low": a0 }; };
function ConfigurationListComponent_c8y_list_group_26_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "c8y-list-group", 21);
    ɵngcc0.ɵɵpipe(1, "async");
    ɵngcc0.ɵɵtemplate(2, ConfigurationListComponent_c8y_list_group_26_c8y_li_2_Template, 19, 15, "c8y-li", 22);
    ɵngcc0.ɵɵtemplate(3, ConfigurationListComponent_c8y_list_group_26_c8y_li_3_Template, 36, 26, "c8y-li", 23);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    let tmp_0_0;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(6, _c0, ctx_r5.data && ctx_r5.data.length ? ctx_r5.data.length < 10 : ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(1, 4, ctx_r5.configurations$)) == null ? null : tmp_0_0.data.length) < 10));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r5.shouldShowEmptyState());
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("c8yForOf", ctx_r5.configurations$)("c8yForPipe", ctx_r5.filterPipe);
} }
const _c1 = function (a0) { return { "icon-spin": a0 }; };
export class ConfigurationListComponent {
    constructor(alert, repositoryService, bsModalService, modalService, translateService, inventoryBinaryService, inventoryService) {
        this.alert = alert;
        this.repositoryService = repositoryService;
        this.bsModalService = bsModalService;
        this.modalService = modalService;
        this.translateService = translateService;
        this.inventoryBinaryService = inventoryBinaryService;
        this.inventoryService = inventoryService;
        this.filterTerm = '';
        this.reloading = false;
        this.DELETED_SUCCESS_MSG = gettext('Configuration deleted.');
    }
    ngOnInit() {
        this.loadConfigurations();
    }
    loadConfigurations() {
        return __awaiter(this, void 0, void 0, function* () {
            this.reloading = true;
            this.configurations$ = of(yield this.repositoryService.listRepositoryEntries(RepositoryType.CONFIGURATION));
            this.reloading = false;
            this.reset();
        });
    }
    add() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.bsModalService.show(ConfigurationDetailComponent, {
                    class: 'modal-sm',
                    ignoreBackdropClick: true
                }).content.result;
                yield this.loadConfigurations();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    edit(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            const fileBinary = yield this.repositoryService.getBinaryFile(configuration.url, {
                allowExternal: false
            });
            try {
                const modal = this.bsModalService.show(ConfigurationDetailComponent, {
                    class: 'modal-sm',
                    ignoreBackdropClick: true,
                    initialState: {
                        selected: configuration,
                        version: configuration.name,
                        deviceType: configuration.deviceType,
                        description: configuration.description,
                        binary: { file: fileBinary, url: configuration.url }
                    }
                }).content;
                modal.mo = configuration;
                yield modal.result;
                yield this.loadConfigurations();
            }
            catch (ex) {
                // intended empty
            }
        });
    }
    isBinaryFile(configuration) {
        return configuration.url ? !!this.inventoryBinaryService.getIdFromUrl(configuration.url) : false;
    }
    getBinaryName(configuration) {
        return this.repositoryService.getBinaryName$(configuration.url);
    }
    download(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            const fileBinary = yield this.repositoryService.getBinaryFile(configuration.url, {
                allowExternal: false
            });
            saveAs(fileBinary);
        });
    }
    delete(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const title = gettext('Delete configuration snapshot');
                const confirmationText = gettext('You are about to delete the configuration snapshot {{ name }}.');
                const hint = gettext('This operation is irreversible.');
                const proceed = gettext('Do you want to proceed?');
                const body = [
                    this.translateService.instant(confirmationText, {
                        name: configuration.name
                    }),
                    this.translateService.instant(hint),
                    this.translateService.instant(proceed)
                ].join(' ');
                const labels = {
                    ok: gettext('Delete')
                };
                yield this.modalService.confirm(title, body, Status.DANGER, labels);
                yield this.repositoryService.delete(configuration);
                this.alert.success(this.DELETED_SUCCESS_MSG);
                yield this.loadConfigurations();
            }
            catch (ex) {
                if (ex) {
                    this.alert.addServerFailure(ex);
                }
            }
        });
    }
    setPipe(filterTerm) {
        this.filterTerm = filterTerm;
        this.filterPipe = pipe(map((data) => {
            this.data =
                filterTerm.trim().length === 0
                    ? data
                    : data.filter((mo) => this.filterContainString(mo.name, filterTerm) ||
                        this.filterContainString(mo.configurationType, filterTerm) ||
                        this.filterContainString(mo.deviceType, filterTerm) ||
                        this.filterContainString(mo.description, filterTerm));
            return this.data;
        }));
    }
    shouldShowEmptyState() {
        return !(this.data && this.data.length > 0);
    }
    reset() {
        this.filter.filterTerm = '';
        this.setPipe('');
    }
    filterContainString(name, filterTerm) {
        const term = filterTerm.toLowerCase().trim();
        return name && name.toLowerCase().indexOf(term) > -1;
    }
}
ConfigurationListComponent.ɵfac = function ConfigurationListComponent_Factory(t) { return new (t || ConfigurationListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.RepositoryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.BsModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ModalService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TranslateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.InventoryBinaryService), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.InventoryService)); };
ConfigurationListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ConfigurationListComponent, selectors: [["c8y-configuration-list"]], viewQuery: function ConfigurationListComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(FilterInputComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.filter = _t.first);
    } }, decls: 28, vars: 31, consts: [["translate", ""], [4, "ngIf"], ["container", "body", 3, "tooltip", 4, "ngIf"], ["itemClass", "navbar-form"], [3, "icon", "onSearch"], [3, "placement"], [1, "btn", "btn-link", 3, "click"], ["c8yIcon", "plus-circle"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["class", "c8y-empty-state text-center", 4, "ngIf"], ["class", "c8y-empty-state", 4, "ngIf"], ["class", "m-b-24", 3, "ngClass", 4, "ngIf"], ["container", "body", 3, "tooltip"], [1, "c8y-empty-state", "text-center"], ["c8yIcon", "gears"], ["translate", "", 1, "btn", "btn-primary", 3, "click"], ["c8y-guide-docs", ""], ["c8y-guide-href", "users-guide/device-management/#configuration-repository"], [1, "c8y-empty-state"], ["c8yIcon", "search", 1, "c8y-icon-duocolor"], [1, "m-b-24", 3, "ngClass"], ["class", "page-sticky-header c8y-list__item--empty-actions hidden-xs", 4, "ngIf"], [4, "c8yFor", "c8yForOf", "c8yForPipe"], [1, "page-sticky-header", "c8y-list__item--empty-actions", "hidden-xs"], [1, "p-l-24"], [1, "content-flex-60"], [1, "col-2"], [1, "col-3"], ["icon", "gears"], [1, "btn-clean", "col-2", 3, "click"], [1, "text-truncate", 3, "title"], ["elementClass", "text-info", 3, "text", "pattern"], [1, "text-label-small", "visible-xs-inline", "m-r-4"], ["class", "text-truncate", "title", "configuration.description", 4, "ngIf", "ngIfElse"], [1, "text-truncate"], ["translate", "", 1, "text-label-small", "m-r-4", "visible-xs-inline"], [3, "title", 4, "ngIf", "ngIfElse"], ["noFile", ""], ["translate", "", 1, "text-label-small", "visible-xs-inline", "m-r-4"], [4, "ngIf", "ngIfElse"], ["class", "label label-primary", 4, "ngIf", "ngIfElse"], ["icon", "pencil", 3, "label", "click"], ["icon", "trash-o", 3, "label", "click"], ["icon", "download", 3, "label", "click", 4, "ngIf"], ["emptyText", ""], ["emptyDescription", ""], ["title", "configuration.description", 1, "text-truncate"], [3, "title"], [1, "label", "label-primary"], ["icon", "download", 3, "label", "click"], [1, "text-muted"]], template: function ConfigurationListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵelementStart(1, "span", 0);
        ɵngcc0.ɵɵtext(2, " Configuration snapshots repository ");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtext(3, "\u00A0 ");
        ɵngcc0.ɵɵtemplate(4, ConfigurationListComponent_small_4_Template, 5, 3, "small", 1);
        ɵngcc0.ɵɵpipe(5, "async");
        ɵngcc0.ɵɵtemplate(6, ConfigurationListComponent_small_6_Template, 6, 6, "small", 2);
        ɵngcc0.ɵɵpipe(7, "async");
        ɵngcc0.ɵɵtemplate(8, ConfigurationListComponent_small_8_Template, 4, 1, "small", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "c8y-action-bar-item", 3);
        ɵngcc0.ɵɵelementStart(10, "c8y-filter", 4);
        ɵngcc0.ɵɵlistener("onSearch", function ConfigurationListComponent_Template_c8y_filter_onSearch_10_listener($event) { return ctx.setPipe($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "c8y-action-bar-item", 5);
        ɵngcc0.ɵɵelementStart(12, "button", 6);
        ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_Template_button_click_12_listener() { return ctx.add(); });
        ɵngcc0.ɵɵelement(13, "i", 7);
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-action-bar-item", 5);
        ɵngcc0.ɵɵelementStart(17, "button", 8);
        ɵngcc0.ɵɵlistener("click", function ConfigurationListComponent_Template_button_click_17_listener() { return ctx.loadConfigurations(); });
        ɵngcc0.ɵɵpipe(18, "translate");
        ɵngcc0.ɵɵelement(19, "i", 9);
        ɵngcc0.ɵɵtext(20);
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(22, ConfigurationListComponent_div_22_Template, 15, 0, "div", 10);
        ɵngcc0.ɵɵpipe(23, "async");
        ɵngcc0.ɵɵtemplate(24, ConfigurationListComponent_div_24_Template, 6, 0, "div", 11);
        ɵngcc0.ɵɵpipe(25, "async");
        ɵngcc0.ɵɵtemplate(26, ConfigurationListComponent_c8y_list_group_26_Template, 4, 8, "c8y-list-group", 12);
        ɵngcc0.ɵɵpipe(27, "async");
    } if (rf & 2) {
        let tmp_0_0;
        let tmp_1_0;
        let tmp_10_0;
        let tmp_11_0;
        let tmp_12_0;
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_0_0 = ɵngcc0.ɵɵpipeBind1(5, 13, ctx.configurations$)) == null ? null : tmp_0_0.paging.totalPages) === 1 && !ctx.filterTerm);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_1_0 = ɵngcc0.ɵɵpipeBind1(7, 15, ctx.configurations$)) == null ? null : tmp_1_0.paging.totalPages) > 1 && !ctx.filterTerm);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.filterTerm);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "search");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 17, "Add configuration snapshot"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(18, 19, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(29, _c1, ctx.reloading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(21, 21, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.filterTerm && ((tmp_10_0 = ɵngcc0.ɵɵpipeBind1(23, 23, ctx.configurations$)) == null ? null : tmp_10_0.data.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.shouldShowEmptyState() && ((tmp_11_0 = ɵngcc0.ɵɵpipeBind1(25, 25, ctx.configurations$)) == null ? null : tmp_11_0.data.length) > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_12_0 = ɵngcc0.ɵɵpipeBind1(27, 27, ctx.configurations$)) == null ? null : tmp_12_0.data.length) > 0);
    } }, directives: [ɵngcc1.TitleComponent, ɵngcc1.C8yTranslateDirective, ɵngcc6.NgIf, ɵngcc1.ActionBarItemComponent, ɵngcc1.FilterInputComponent, ɵngcc1.IconDirective, ɵngcc6.NgClass, ɵngcc7.TooltipDirective, ɵngcc1.GuideDocsComponent, ɵngcc1.GuideHrefDirective, ɵngcc1.ListGroupComponent, ɵngcc1.ForOfDirective, ɵngcc1.ListItemComponent, ɵngcc1.ListItemIconComponent, ɵngcc1.ListItemBodyComponent, ɵngcc1.HighlightComponent, ɵngcc1.ListItemActionComponent], pipes: [ɵngcc6.AsyncPipe, ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
ConfigurationListComponent.ctorParameters = () => [
    { type: AlertService },
    { type: RepositoryService },
    { type: BsModalService },
    { type: ModalService },
    { type: TranslateService },
    { type: InventoryBinaryService },
    { type: InventoryService }
];
ConfigurationListComponent.propDecorators = {
    filter: [{ type: ViewChild, args: [FilterInputComponent, { static: false },] }]
};
__decorate([
    memoize(property('id'))
], ConfigurationListComponent.prototype, "getBinaryName", null);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ConfigurationListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-configuration-list',
                template: "<c8y-title>\n  <span translate>\n    Configuration snapshots repository\n  </span>&nbsp;\n  <small\n    *ngIf=\"(configurations$ | async)?.paging.totalPages === 1 && !filterTerm\"\n  >\n    {{ (configurations$ | async).data.length }}\n    <span translate>\n      snapshots\n    </span>\n  </small>\n  <small\n    *ngIf=\"(configurations$ | async)?.paging.totalPages > 1 && !filterTerm\"\n    [tooltip]=\"'More data available. Scroll to the bottom of the list to load it.' | translate\"\n    container=\"body\"\n  >\n    {{ (configurations$ | async).paging.pageSize }}+\n    <span translate>snapshots</span>\n  </small>\n  <small *ngIf=\"filterTerm\">\n    <span translate>\n      Search results for\n    </span>&nbsp;\"\n      {{ this.filterTerm }}\"\n    </small>\n</c8y-title>\n\n<c8y-action-bar-item itemClass=\"navbar-form\">\n  <c8y-filter\n    [icon]=\"'search'\"\n    (onSearch)=\"setPipe($event)\"\n  ></c8y-filter>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    (click)=\"add()\"\n  >\n    <i c8yIcon=\"plus-circle\"></i>\n    {{ 'Add configuration snapshot' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button class=\"btn btn-link\"\n    title=\"{{ 'Reload' | translate }}\"\n    (click)=\"loadConfigurations()\"\n  >\n    <i [ngClass]=\"{ 'icon-spin': reloading }\"\n      c8yIcon=\"refresh\"\n    ></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<!-- empty state -->\n<div class=\"c8y-empty-state text-center\"\n  *ngIf=\"!filterTerm && (configurations$ | async)?.data.length === 0\"\n>\n  <h1 c8yIcon=\"gears\"></h1>\n  <h3 translate>There are no configuration snapshots defined</h3>\n  <p translate>Add a configuration snapshot first.</p>\n  <div>\n    <button (click)=\"add()\" class=\"btn btn-primary\" translate>\n      Add configuration snapshot\n    </button>\n  </div>\n  <p c8y-guide-docs>\n    <small translate ngNonBindable>\n      Find out more in the\n      <a c8y-guide-href=\"users-guide/device-management/#configuration-repository\">\n        User guide`KEEP_ORIGINAL`\n        </a>.\n    </small>\n  </p>\n</div>\n\n<!-- no results empty state -->\n<div class=\"c8y-empty-state\"\n  *ngIf=\"shouldShowEmptyState() && (configurations$ | async)?.data.length > 0\"\n>\n  <h1 class=\"c8y-icon-duocolor\"\n    c8yIcon=\"search\"\n  ></h1>\n  <h3 translate>\n    No results to display.\n  </h3>\n  <p translate>\n    Refine your search terms or check your spelling.\n  </p>\n</div>\n\n<c8y-list-group class=\"m-b-24\"\n  *ngIf=\"(configurations$ | async)?.data.length > 0\"\n  [ngClass]=\"{ 'dd-low': data && data.length ? data.length < 10 : (configurations$ | async)?.data.length < 10 }\"\n\n>\n  <c8y-li class=\"page-sticky-header c8y-list__item--empty-actions hidden-xs\"\n    *ngIf=\"!shouldShowEmptyState()\"\n  >\n    <c8y-li-icon>\n      <i class=\"p-l-24\"></i>\n    </c8y-li-icon>\n    <c8y-li-body class=\"content-flex-60\">\n      <div class=\"col-2\">\n        {{ 'Configuration' | translate }}\n      </div>\n      <div class=\"col-3\">\n        {{ 'Description' | translate }}\n      </div>\n      <div class=\"col-3\">\n        {{ 'File' | translate }}\n      </div>\n      <div class=\"col-2\">\n        {{ 'Device type' | translate }}\n      </div>\n      <div class=\"col-2\">\n        {{ 'Configuration type' | translate }}\n      </div>\n    </c8y-li-body>\n  </c8y-li>\n\n  <c8y-li *c8yFor=\"let configuration of configurations$; pipe: filterPipe\">\n    <c8y-li-icon icon=\"gears\"></c8y-li-icon>\n    <div class=\"content-flex-60\">\n      <button class=\"btn-clean col-2\"\n        (click)=\"edit(configuration)\"\n      >\n        <span class=\"text-truncate\"\n          title=\"{{ configuration.name || '-' }}\"\n        >\n          <c8y-highlight\n            [text]=\"configuration.name || '-'\"\n            elementClass=\"text-info\"\n            [pattern]=\"filterTerm\"\n          ></c8y-highlight>\n        </span>\n      </button>\n      <div class=\"col-3\">\n        <div class=\"text-label-small visible-xs-inline m-r-4\">\n          {{ 'Description' | translate }}\n        </div>\n        <small class=\"text-truncate\"\n          *ngIf=\"configuration.description; else emptyDescription\"\n          title=\"configuration.description\"\n        >\n          <c8y-highlight\n            [text]=\"configuration.description || '-'\"\n            elementClass=\"text-info\"\n            [pattern]=\"filterTerm\">\n          </c8y-highlight>\n        </small>\n      </div>\n      <div class=\"col-3\">\n        <span class=\"text-truncate\">\n          <span class=\"text-label-small m-r-4 visible-xs-inline\" translate>\n            File\n          </span>\n          <small\n            *ngIf=\"isBinaryFile(configuration); else noFile\"\n            title=\"{{ getBinaryName(configuration) | async }}\"\n          >\n            {{ getBinaryName(configuration) | async }}\n          </small>\n          <ng-template #noFile>\n            <small title=\"{{ configuration.url }}\">\n              {{ configuration.url }}\n            </small>\n          </ng-template>\n        </span>\n      </div>\n      <div class=\"col-2\">\n        <div class=\"text-truncate\"\n          title=\"{{ 'Device type' | translate }}: {{ configuration.deviceType || '-' }}\"\n        >\n          <span class=\"text-label-small visible-xs-inline m-r-4\" translate>\n            Device type\n          </span>\n          <span *ngIf=\"configuration.deviceType; else emptyText\">\n            <c8y-highlight\n              [text]=\"configuration.deviceType || '-'\"\n              elementClass=\"text-info\"\n              [pattern]=\"filterTerm\"\n            >\n            </c8y-highlight>\n          </span>\n        </div>\n      </div>\n      <div class=\"col-2\">\n        <div class=\"text-truncate\"\n          title=\"{{ configuration.configurationType }}\"\n        >\n          <span class=\"label label-primary\"\n            *ngIf=\"configuration.configurationType; else emptyText\"\n          >\n            <c8y-highlight\n              [text]=\"configuration.configurationType\"\n              elementClass=\"text-info\"\n              [pattern]=\"filterTerm\"\n            >\n            </c8y-highlight>\n          </span>\n        </div>\n      </div>\n    </div>\n    <c8y-li-action\n      (click)=\"edit(configuration)\"\n      icon=\"pencil\"\n      label=\"{{ 'Edit' | translate }}\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      (click)=\"delete(configuration)\"\n      icon=\"trash-o\"\n      label=\"{{ 'Delete' | translate }}\"\n    >\n    </c8y-li-action>\n    <c8y-li-action\n      *ngIf=\"isBinaryFile(configuration)\"\n      (click)=\"download(configuration)\"\n      icon=\"download\"\n      label=\"{{ 'Download' | translate }}\"\n    >\n    </c8y-li-action>\n    <ng-template #emptyText>\n      <small class=\"text-muted\">\n        <em>{{ 'Undefined' | translate }}</em>\n      </small>\n    </ng-template>\n    <ng-template #emptyDescription>\n      <small class=\"text-muted\">\n        <em>{{ 'No description' | translate }}</em>\n      </small>\n    </ng-template>\n  </c8y-li>\n</c8y-list-group>\n"
            }]
    }], function () { return [{ type: ɵngcc1.AlertService }, { type: ɵngcc2.RepositoryService }, { type: ɵngcc3.BsModalService }, { type: ɵngcc1.ModalService }, { type: ɵngcc4.TranslateService }, { type: ɵngcc5.InventoryBinaryService }, { type: ɵngcc5.InventoryService }]; }, { filter: [{
            type: ViewChild,
            args: [FilterInputComponent, { static: false }]
        }] }); })();
