// @c8y/ngx-components/esm2015/ecosystem/shared/application-card.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Router } from '@angular/router';
import { ApplicationService } from '@c8y/client';
import { AlertService, gettext } from '@c8y/ngx-components';
import { EcosystemService } from './../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/client';
import * as ɵngcc2 from './../ecosystem.service';
import * as ɵngcc3 from '@c8y/ngx-components';
import * as ɵngcc4 from '@angular/router';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from 'ngx-bootstrap/dropdown';

function ApplicationCardComponent_c8y_app_icon_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-app-icon", 20);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("app", ctx_r0.app)("contextPath", ctx_r0.app.contextPath)("name", ctx_r0.app.name);
} }
function ApplicationCardComponent_i_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 21);
} }
function ApplicationCardComponent_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 22);
} }
function ApplicationCardComponent_i_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 23);
} }
function ApplicationCardComponent_i_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 24);
} }
function ApplicationCardComponent_small_14_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "small", 25);
    ɵngcc0.ɵɵelementStart(1, "em");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r5.app.manifest.version);
} }
function ApplicationCardComponent_div_17_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 26);
    ɵngcc0.ɵɵelement(1, "div", 27);
    ɵngcc0.ɵɵelementStart(2, "p", 28);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(4, 1, ctx_r6.app.description));
} }
function ApplicationCardComponent_button_24_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 29);
    ɵngcc0.ɵɵlistener("click", function ApplicationCardComponent_button_24_Template_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); ctx_r9.openApp(); return $event.stopPropagation(); });
    ɵngcc0.ɵɵpipe(1, "translate");
    ɵngcc0.ɵɵelement(2, "i", 30);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("disabled", ctx_r7.disableOpenInBrowser)("title", ɵngcc0.ɵɵpipeBind1(1, 3, ctx_r7.openButtonTitle));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 5, "Open"), " ");
} }
function ApplicationCardComponent_div_26_ul_5_li_1_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "button", 38);
    ɵngcc0.ɵɵlistener("click", function ApplicationCardComponent_div_26_ul_5_li_1_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r15); const ctx_r14 = ɵngcc0.ɵɵnextContext(3); return ctx_r14.detail(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 39);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Edit"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Edit"), " ");
} }
function ApplicationCardComponent_div_26_ul_5_li_2_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵelementStart(1, "button", 40);
    ɵngcc0.ɵɵlistener("click", function ApplicationCardComponent_div_26_ul_5_li_2_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(3); return ctx_r16.delete(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 41);
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵpipe(5, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(2, 2, "Delete"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(5, 4, "Delete"), " ");
} }
function ApplicationCardComponent_div_26_ul_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 36);
    ɵngcc0.ɵɵtemplate(1, ApplicationCardComponent_div_26_ul_5_li_1_Template, 6, 6, "li", 37);
    ɵngcc0.ɵɵtemplate(2, ApplicationCardComponent_div_26_ul_5_li_2_Template, 6, 6, "li", 37);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.canEdit);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r11.canDelete);
} }
function ApplicationCardComponent_div_26_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 31);
    ɵngcc0.ɵɵelementStart(1, "div", 32);
    ɵngcc0.ɵɵelementStart(2, "button", 33);
    ɵngcc0.ɵɵpipe(3, "translate");
    ɵngcc0.ɵɵelement(4, "i", 34);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(5, ApplicationCardComponent_div_26_ul_5_Template, 3, 2, "ul", 35);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 1, "Settings"));
} }
export class ApplicationCardComponent {
    constructor(applicationService, ecosystemService, alertService, router) {
        this.applicationService = applicationService;
        this.ecosystemService = ecosystemService;
        this.alertService = alertService;
        this.router = router;
        this.canEdit = true;
        this.onAppDeleted = new EventEmitter();
    }
    get openButtonTitle() {
        return !this.disableOpenInBrowser ?
            gettext('Open') :
            gettext('This application is overridden.');
    }
    ngOnInit() {
        return __awaiter(this, void 0, void 0, function* () {
            this.appState = this.ecosystemService.getAppState(this.app);
            const contextPath = this.app.contextPath;
            this.canOpenInBrowser = this.ecosystemService.canOpenAppInBrowser(this.app);
            this.disableOpenInBrowser =
                this.canOpenInBrowser && (yield this.ecosystemService.isOverwrittenByCustomApp(this.app));
            this.canDelete = this.ecosystemService.canDeleteApp(this.app);
            this.isPackage = this.ecosystemService.isPackage(this.app);
            this.isFeature = this.ecosystemService.isFeature(this.app);
            this.isMicroservice = this.ecosystemService.isMicroservice(this.app);
            this.isExternal = this.ecosystemService.isExternal(this.app);
        });
    }
    detail() {
        this.isMicroservice
            ? this.router.navigateByUrl(`/microservices/${this.app.id}`)
            : this.router.navigateByUrl(`/ecosystem/application/${this.app.id}`);
    }
    openApp() {
        window.open(this.applicationService.getHref(this.app));
    }
    delete() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.ecosystemService.deleteApp(this.app);
                this.onAppDeleted.emit();
            }
            catch (ex) {
                if (ex) {
                    this.alertService.addServerFailure(ex);
                }
            }
        });
    }
}
ApplicationCardComponent.ɵfac = function ApplicationCardComponent_Factory(t) { return new (t || ApplicationCardComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ApplicationService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.AlertService), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.Router)); };
ApplicationCardComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ApplicationCardComponent, selectors: [["c8y-application-card"]], inputs: { canEdit: "canEdit", app: "app" }, outputs: { onAppDeleted: "onAppDeleted" }, decls: 27, vars: 23, consts: [[1, "card", "pointer", "card--d-col"], [1, "card-block", "text-center", "card-column-40", 3, "click"], [1, "card-appicon"], [3, "app", "contextPath", "name", 4, "ngIf"], ["c8yIcon", "big-parcel", 4, "ngIf"], ["c8yIcon", "tab", 4, "ngIf"], ["c8yIcon", "microchip", 4, "ngIf"], ["c8yIcon", "globe1", 4, "ngIf"], [1, "e2e-appCardName", "text-medium", 3, "title"], ["class", "text-muted", 4, "ngIf"], [1, "card-block", "p-0", "no-min-height", "card--d-col", "card-column-80", "flex-grow", 3, "click"], [1, "card-block", "no-min-height", "p-t-0", "p-b-0", "card-column-80"], ["class", "text-center-grid", 4, "ngIf"], [1, "card-block", "no-min-height", "p-t-0", "p-b-0", "card-column-20", "text-center-grid"], [1, "label", 3, "ngClass"], [1, "card-footer", "no-min-height", "p-t-0", "p-b-0", "card-column-20", "text-center-grid"], [1, "p-b-16", "card-hidden-list"], ["class", "btn btn-xs btn-default", 3, "disabled", "title", "click", 4, "ngIf"], [1, "p-b-32", "card-hidden-list"], ["class", "card-actions m-t-0", 4, "ngIf"], [3, "app", "contextPath", "name"], ["c8yIcon", "big-parcel"], ["c8yIcon", "tab"], ["c8yIcon", "microchip"], ["c8yIcon", "globe1"], [1, "text-muted"], [1, "text-center-grid"], [1, "p-b-8", "card-hidden-grid"], [1, "small", "l-h-tight", "p-b-8", "p-l-0", "p-r-0"], [1, "btn", "btn-xs", "btn-default", 3, "disabled", "title", "click"], ["c8yIcon", "external-link", 1, "m-r-4"], [1, "card-actions", "m-t-0"], ["dropdown", "", 1, "dropdown"], ["dropdownToggle", "", 1, "dropdown-toggle", "c8y-dropdown", 3, "title"], ["c8yIcon", "ellipsis-v"], ["class", "dropdown-menu dropdown-menu-right", 4, "dropdownMenu"], [1, "dropdown-menu", "dropdown-menu-right"], [4, "ngIf"], [3, "title", "click"], ["c8yIcon", "pencil", 1, "m-r-4"], ["href", "", 3, "title", "click"], ["c8yIcon", "trash", 1, "m-r-4"]], template: function ApplicationCardComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵlistener("click", function ApplicationCardComponent_Template_div_click_1_listener() { return ctx.detail(); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵtemplate(3, ApplicationCardComponent_c8y_app_icon_3_Template, 1, 3, "c8y-app-icon", 3);
        ɵngcc0.ɵɵtemplate(4, ApplicationCardComponent_i_4_Template, 1, 0, "i", 4);
        ɵngcc0.ɵɵtemplate(5, ApplicationCardComponent_i_5_Template, 1, 0, "i", 5);
        ɵngcc0.ɵɵtemplate(6, ApplicationCardComponent_i_6_Template, 1, 0, "i", 6);
        ɵngcc0.ɵɵtemplate(7, ApplicationCardComponent_i_7_Template, 1, 0, "i", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "p", 8);
        ɵngcc0.ɵɵpipe(9, "async");
        ɵngcc0.ɵɵpipe(10, "humanizeAppName");
        ɵngcc0.ɵɵtext(11);
        ɵngcc0.ɵɵpipe(12, "async");
        ɵngcc0.ɵɵpipe(13, "humanizeAppName");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(14, ApplicationCardComponent_small_14_Template, 3, 1, "small", 9);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(15, "div", 10);
        ɵngcc0.ɵɵlistener("click", function ApplicationCardComponent_Template_div_click_15_listener() { return ctx.detail(); });
        ɵngcc0.ɵɵelementStart(16, "div", 11);
        ɵngcc0.ɵɵtemplate(17, ApplicationCardComponent_div_17_Template, 5, 3, "div", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "div", 13);
        ɵngcc0.ɵɵelementStart(19, "span", 14);
        ɵngcc0.ɵɵtext(20);
        ɵngcc0.ɵɵpipe(21, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(22, "div", 15);
        ɵngcc0.ɵɵelement(23, "div", 16);
        ɵngcc0.ɵɵtemplate(24, ApplicationCardComponent_button_24_Template, 5, 7, "button", 17);
        ɵngcc0.ɵɵelement(25, "div", 18);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(26, ApplicationCardComponent_div_26_Template, 6, 3, "div", 19);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isPackage && !ctx.isFeature && !ctx.isMicroservice && !ctx.isExternal);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isPackage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isFeature);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isMicroservice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isExternal);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 13, ɵngcc0.ɵɵpipeBind1(10, 15, ctx.app)));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(12, 17, ɵngcc0.ɵɵpipeBind1(13, 19, ctx.app)), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.app.manifest == null ? null : ctx.app.manifest.version) && (ctx.isPackage || ctx.isMicroservice));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.app.description);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.appState.class);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(21, 21, ctx.appState.label));
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.canOpenInBrowser);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.canEdit || ctx.canOpenInBrowser || ctx.canDelete);
    } }, directives: [ɵngcc5.NgIf, ɵngcc5.NgClass, ɵngcc3.AppIconComponent, ɵngcc3.IconDirective, ɵngcc6.BsDropdownDirective, ɵngcc6.BsDropdownToggleDirective, ɵngcc6.BsDropdownMenuDirective], pipes: [ɵngcc5.AsyncPipe, ɵngcc3.HumanizeAppNamePipe, ɵngcc3.C8yTranslatePipe], encapsulation: 2 });
ApplicationCardComponent.ctorParameters = () => [
    { type: ApplicationService },
    { type: EcosystemService },
    { type: AlertService },
    { type: Router }
];
ApplicationCardComponent.propDecorators = {
    app: [{ type: Input }],
    canEdit: [{ type: Input }],
    onAppDeleted: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ApplicationCardComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-application-card',
                template: "<div class=\"card pointer card--d-col\">\n  <div (click)=\"detail()\" class=\"card-block text-center card-column-40\">\n    <div class=\"card-appicon\">\n      <c8y-app-icon\n        *ngIf=\"!isPackage && !isFeature && !isMicroservice && !isExternal\"\n        [app]=\"app\"\n        [contextPath]=\"app.contextPath\"\n        [name]=\"app.name\"\n      ></c8y-app-icon>\n      <i *ngIf=\"isPackage\" c8yIcon=\"big-parcel\"></i>\n      <i *ngIf=\"isFeature\" c8yIcon=\"tab\"></i>\n      <i *ngIf=\"isMicroservice\" c8yIcon=\"microchip\"></i>\n      <i *ngIf=\"isExternal\" c8yIcon=\"globe1\"></i>\n    </div>\n    <p class=\"e2e-appCardName text-medium\" title=\"{{ app | humanizeAppName | async }}\">\n      {{ app | humanizeAppName | async }}\n    </p>\n    <small *ngIf=\"app.manifest?.version && (isPackage || isMicroservice)\" class=\"text-muted\">\n      <em>{{ app.manifest.version }}</em>\n    </small>\n  </div>\n  <div (click)=\"detail()\" class=\"card-block p-0 no-min-height card--d-col card-column-80 flex-grow\">\n    <div class=\"card-block no-min-height p-t-0 p-b-0 card-column-80\">\n      <div *ngIf=\"app.description\" class=\"text-center-grid\">\n        <div class=\"p-b-8 card-hidden-grid\"></div>\n        <p class=\"small l-h-tight p-b-8 p-l-0 p-r-0\">{{ app.description | translate }}</p>\n      </div>\n    </div>\n    <div class=\"card-block no-min-height p-t-0 p-b-0 card-column-20 text-center-grid\">\n      <span [ngClass]=\"appState.class\" class=\"label\">{{ appState.label | translate }}</span>\n    </div>\n    <div class=\"card-footer no-min-height p-t-0 p-b-0 card-column-20 text-center-grid\">\n      <div class=\"p-b-16 card-hidden-list\"></div>\n      <button\n        (click)=\"openApp(); $event.stopPropagation()\"\n        *ngIf=\"canOpenInBrowser\"\n        [disabled]=\"disableOpenInBrowser\"\n        class=\"btn btn-xs btn-default\"\n        [title]=\"openButtonTitle | translate\"\n      >\n        <i c8yIcon=\"external-link\" class=\"m-r-4\"></i>\n        {{ 'Open' | translate }}\n      </button>\n      <div class=\"p-b-32 card-hidden-list\"></div>\n    </div>\n  </div>\n\n  <div *ngIf=\"canEdit || canOpenInBrowser || canDelete\" class=\"card-actions m-t-0\">\n    <div class=\"dropdown\" dropdown>\n      <button\n        class=\"dropdown-toggle c8y-dropdown\"\n        dropdownToggle\n        title=\"{{ 'Settings' | translate }}\"\n      >\n        <i c8yIcon=\"ellipsis-v\"></i>\n      </button>\n      <ul *dropdownMenu class=\"dropdown-menu dropdown-menu-right\">\n        <li *ngIf=\"canEdit\">\n          <button (click)=\"detail()\" title=\"{{ 'Edit' | translate }}\">\n            <i c8yIcon=\"pencil\" class=\"m-r-4\"></i>\n            {{ 'Edit' | translate }}\n          </button>\n        </li>\n        <li *ngIf=\"canDelete\">\n          <button\n            (click)=\"delete()\"\n            href=\"\"\n            title=\"{{ 'Delete' | translate }}\"\n          >\n            <i c8yIcon=\"trash\" class=\"m-r-4\"></i>\n            {{ 'Delete' | translate }}\n          </button>\n        </li>\n      </ul>\n    </div>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ApplicationService }, { type: ɵngcc2.EcosystemService }, { type: ɵngcc3.AlertService }, { type: ɵngcc4.Router }]; }, { canEdit: [{
            type: Input
        }], onAppDeleted: [{
            type: Output
        }], app: [{
            type: Input
        }] }); })();
