// @c8y/ngx-components/esm2015/ecosystem/shared/upload-archive.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { AlertService, gettext } from '@c8y/ngx-components';
import { EcosystemService } from '../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
export class UploadArchiveComponent {
    constructor(ecosystemService, alertService) {
        this.ecosystemService = ecosystemService;
        this.alertService = alertService;
        this.applicationChange = new EventEmitter();
        this.refresh = new EventEmitter();
        this.archives = [];
        this.uploadInProgress = false;
        this.canReactivate = false;
    }
    get uploadProgress() {
        return this.ecosystemService.progress;
    }
    onFileDroppedEvent(event) {
        if (event && event.length > 0) {
            const file = event[0].file;
            this.onFile(file);
        }
    }
    onFile(file) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.ecosystemService.isValidAppType(file, this.application.type);
                const isCustomMicroservice = this.ecosystemService.isCustomMicroservice(this.application);
                this.archives = isCustomMicroservice
                    ? []
                    : yield this.ecosystemService.listArchives(this.application.id);
                this.archives.sort((a, b) => {
                    return new Date(b.created) - new Date(a.created);
                });
                this.uploadInProgress = true;
                this.uploadProgress.next(0);
                if (this.archives.length === 6 && !isCustomMicroservice) {
                    yield this.ecosystemService.removeOldestArchive(this.application, this.archives);
                }
                const app = yield this.ecosystemService.uploadArchiveToApp(file, this.application);
                this.applicationChange.emit(app);
                this.uploadProgress.next(100);
                this.uploadInProgress = false;
                this.refresh.emit();
                this.alertService.success(gettext('Archive uploaded and activated.'));
            }
            catch (ex) {
                this.uploadInProgress = false;
                this.alertService.addServerFailure(ex);
            }
        });
    }
}
UploadArchiveComponent.ɵfac = function UploadArchiveComponent_Factory(t) { return new (t || UploadArchiveComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AlertService)); };
UploadArchiveComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: UploadArchiveComponent, selectors: [["c8y-upload-archive"]], inputs: { application: "application" }, outputs: { applicationChange: "applicationChange", refresh: "refresh" }, decls: 3, vars: 11, consts: [[1, "drop-area-sm", 3, "accept", "files", "loading", "maxAllowedFiles", "message", "progress", "forceHideList", "dropped"]], template: function UploadArchiveComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-drop-area", 0);
        ɵngcc0.ɵɵlistener("dropped", function UploadArchiveComponent_Template_c8y_drop_area_dropped_0_listener($event) { return ctx.onFileDroppedEvent($event); });
        ɵngcc0.ɵɵpipe(1, "translate");
        ɵngcc0.ɵɵpipe(2, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("accept", ".zip")("files", ctx.fileList)("loading", ctx.uploadInProgress)("maxAllowedFiles", 1)("message", ɵngcc0.ɵɵpipeBind1(1, 7, "Upload a *.zip file"))("progress", ɵngcc0.ɵɵpipeBind1(2, 9, ctx.uploadProgress))("forceHideList", true);
    } }, directives: [ɵngcc2.DropAreaComponent], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.AsyncPipe], encapsulation: 2 });
UploadArchiveComponent.ctorParameters = () => [
    { type: EcosystemService },
    { type: AlertService }
];
UploadArchiveComponent.propDecorators = {
    application: [{ type: Input }],
    applicationChange: [{ type: Output }],
    refresh: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UploadArchiveComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-upload-archive',
                template: "<c8y-drop-area\n  (dropped)=\"onFileDroppedEvent($event)\"\n  [accept]=\"'.zip'\"\n  [files]=\"fileList\"\n  [loading]=\"uploadInProgress\"\n  [maxAllowedFiles]=\"1\"\n  [message]=\"'Upload a *.zip file' | translate\"\n  [progress]=\"uploadProgress | async\"\n  [forceHideList]=\"true\"\n  class=\"drop-area-sm\"\n>\n</c8y-drop-area>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }, { type: ɵngcc2.AlertService }]; }, { applicationChange: [{
            type: Output
        }], refresh: [{
            type: Output
        }], application: [{
            type: Input
        }] }); })();
