// @c8y/ngx-components/esm2015/ecosystem/applications/duplicate-application/duplicate-application-properties/duplicate-application-properties.component.js
import { __awaiter } from "tslib";
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ApplicationService } from '@c8y/client';
import { BsModalRef } from 'ngx-bootstrap/modal';
import { EcosystemService } from '../../../ecosystem.service';
import { ApplicationPropertiesFormComponent } from '../../../shared/application-properties-form.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/modal';
import * as ɵngcc2 from '../../../ecosystem.service';
import * as ɵngcc3 from '@c8y/client';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from '@c8y/ngx-components';
import * as ɵngcc6 from '../../../shared/application-properties-form.component';

function DuplicateApplicationPropertiesComponent_p_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p", 6);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(2, 1, "Provide application details"), "\n");
} }
function DuplicateApplicationPropertiesComponent_c8y_application_properties_form_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-application-properties-form", 7);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("application", ctx_r1.newAppConfig);
} }
function DuplicateApplicationPropertiesComponent_c8y_progress_bar_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "c8y-progress-bar", 8);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("message", ɵngcc0.ɵɵpipeBind1(1, 1, "Duplicating\u2026"));
} }
export class DuplicateApplicationPropertiesComponent {
    constructor(bsModalRef, ecosystemService, applicationService) {
        this.bsModalRef = bsModalRef;
        this.ecosystemService = ecosystemService;
        this.applicationService = applicationService;
        this.duplicatedApp = new EventEmitter();
        this.inProgress = false;
    }
    ngOnInit() {
        this.newAppConfig = this.ecosystemService.getUniqueAppConfig(this.selectedApp.value, this.existingApps);
    }
    duplicateApp() {
        return __awaiter(this, void 0, void 0, function* () {
            const formGroupValue = this.applicationPropertiesForm.formGroup.getRawValue();
            this.inProgress = true;
            const clonedApp = (yield this.applicationService.clone(this.selectedApp.value)).data;
            Object.assign(clonedApp, formGroupValue);
            delete clonedApp.type;
            yield this.updateApp(clonedApp);
            this.inProgress = false;
        });
    }
    cancel() {
        this.bsModalRef.hide();
    }
    back() {
        this.stepper.previous();
    }
    updateApp(clonedAppConfig) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const app = (yield this.ecosystemService.updateApp(clonedAppConfig, true)).data;
                this.duplicatedApp.emit(app);
                this.stepper.next();
            }
            catch (err) {
                this.inProgress = false;
            }
        });
    }
}
DuplicateApplicationPropertiesComponent.ɵfac = function DuplicateApplicationPropertiesComponent_Factory(t) { return new (t || DuplicateApplicationPropertiesComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BsModalRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.EcosystemService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.ApplicationService)); };
DuplicateApplicationPropertiesComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: DuplicateApplicationPropertiesComponent, selectors: [["c8y-duplicate-application-properties"]], viewQuery: function DuplicateApplicationPropertiesComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(ApplicationPropertiesFormComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.applicationPropertiesForm = _t.first);
    } }, inputs: { stepper: "stepper", existingApps: "existingApps", selectedApp: "selectedApp" }, outputs: { duplicatedApp: "duplicatedApp" }, decls: 16, vars: 23, consts: [["class", "p-16 text-center text-medium separator-bottom sticky-top bg-white fit-w", 4, "ngIf"], ["class", "d-block fit-w", 3, "application", 4, "ngIf"], ["class", "text-center d-block", 3, "message", 4, "ngIf"], ["type", "button", 1, "btn", "btn-default", 3, "disabled", "title", "click"], [1, "btn", "btn-default", 3, "title", "click"], ["type", "button", 1, "btn", "btn-primary", 3, "disabled", "title", "click"], [1, "p-16", "text-center", "text-medium", "separator-bottom", "sticky-top", "bg-white", "fit-w"], [1, "d-block", "fit-w", 3, "application"], [1, "text-center", "d-block", 3, "message"]], template: function DuplicateApplicationPropertiesComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, DuplicateApplicationPropertiesComponent_p_0_Template, 3, 3, "p", 0);
        ɵngcc0.ɵɵtemplate(1, DuplicateApplicationPropertiesComponent_c8y_application_properties_form_1_Template, 1, 1, "c8y-application-properties-form", 1);
        ɵngcc0.ɵɵtemplate(2, DuplicateApplicationPropertiesComponent_c8y_progress_bar_2_Template, 2, 3, "c8y-progress-bar", 2);
        ɵngcc0.ɵɵelementStart(3, "c8y-wizard-footer");
        ɵngcc0.ɵɵelementStart(4, "button", 3);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_Template_button_click_4_listener() { return ctx.back(); });
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "button", 4);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_Template_button_click_8_listener() { return ctx.cancel(); });
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵpipe(11, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "button", 5);
        ɵngcc0.ɵɵlistener("click", function DuplicateApplicationPropertiesComponent_Template_button_click_12_listener() { return ctx.duplicateApp(); });
        ɵngcc0.ɵɵpipe(13, "translate");
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", !ctx.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.inProgress);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.inProgress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(5, 11, "Back"));
        ɵngcc0.ɵɵproperty("disabled", ctx.inProgress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(7, 13, "Back"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(9, 15, "Cancel"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(11, 17, "Cancel"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(13, 19, "Duplicate"));
        ɵngcc0.ɵɵproperty("disabled", ctx.inProgress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 21, "Duplicate"), " ");
    } }, directives: [ɵngcc4.NgIf, ɵngcc5.WizardFooterComponent, ɵngcc6.ApplicationPropertiesFormComponent, ɵngcc5.ProgressBarComponent], pipes: [ɵngcc5.C8yTranslatePipe], encapsulation: 2 });
DuplicateApplicationPropertiesComponent.ctorParameters = () => [
    { type: BsModalRef },
    { type: EcosystemService },
    { type: ApplicationService }
];
DuplicateApplicationPropertiesComponent.propDecorators = {
    stepper: [{ type: Input }],
    existingApps: [{ type: Input }],
    selectedApp: [{ type: Input }],
    duplicatedApp: [{ type: Output }],
    applicationPropertiesForm: [{ type: ViewChild, args: [ApplicationPropertiesFormComponent,] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DuplicateApplicationPropertiesComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-duplicate-application-properties',
                template: "<p\n  class=\"p-16 text-center text-medium separator-bottom sticky-top bg-white fit-w\"\n  *ngIf=\"!inProgress\"\n>\n  {{ 'Provide application details' | translate }}\n</p>\n<c8y-application-properties-form\n  *ngIf=\"!inProgress\"\n  [application]=\"newAppConfig\"\n  class=\"d-block fit-w\"\n></c8y-application-properties-form>\n<c8y-progress-bar\n  *ngIf=\"inProgress\"\n  [message]=\"'Duplicating\u2026' | translate\"\n  class=\"text-center d-block\"\n></c8y-progress-bar>\n\n<c8y-wizard-footer>\n  <button\n    (click)=\"back()\"\n    [disabled]=\"inProgress\"\n    class=\"btn btn-default\"\n    type=\"button\"\n    title=\"{{ 'Back' | translate }}\"\n  >\n    {{ 'Back' | translate }}\n  </button>\n  <button (click)=\"cancel()\" class=\"btn btn-default\" title=\"{{ 'Cancel' | translate }}\">\n    {{ 'Cancel' | translate }}\n  </button>\n  <button\n    (click)=\"duplicateApp()\"\n    [disabled]=\"inProgress\"\n    class=\"btn btn-primary\"\n    type=\"button\"\n    title=\"{{ 'Duplicate' | translate }}\"\n  >\n    {{ 'Duplicate' | translate }}\n  </button>\n</c8y-wizard-footer>\n"
            }]
    }], function () { return [{ type: ɵngcc1.BsModalRef }, { type: ɵngcc2.EcosystemService }, { type: ɵngcc3.ApplicationService }]; }, { duplicatedApp: [{
            type: Output
        }], stepper: [{
            type: Input
        }], existingApps: [{
            type: Input
        }], selectedApp: [{
            type: Input
        }], applicationPropertiesForm: [{
            type: ViewChild,
            args: [ApplicationPropertiesFormComponent]
        }] }); })();
