// @c8y/ngx-components/esm2015/core/docs/docs.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DocsService } from './docs.service';
import { GuideHrefDirective } from './guide-href.directive';
import { GuideDocsComponent } from './guide-docs.component';
import * as ɵngcc0 from '@angular/core';
export class DocsModule {
}
DocsModule.ɵfac = function DocsModule_Factory(t) { return new (t || DocsModule)(); };
DocsModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DocsModule });
DocsModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [DocsService], imports: [[CommonModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DocsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    GuideHrefDirective,
                    GuideDocsComponent
                ],
                imports: [CommonModule],
                exports: [
                    GuideHrefDirective,
                    GuideDocsComponent
                ],
                providers: [DocsService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DocsModule, { declarations: function () { return [GuideHrefDirective,
        GuideDocsComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [GuideHrefDirective,
        GuideDocsComponent]; } }); })();
