// @c8y/ngx-components/esm2015/core/action/action.component.js
import { Component, Input, ViewChild } from '@angular/core';
import { ActionService } from './action.service';
/**
 * A <c8y-action> can be added with content projection.
 *
 * @example
 * ```html
 * <c8y-action>
 *   <button class="btn" (click)="myAction()">My action</button>
 * </c8y-action>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './action.service';

const _c0 = ["template"];
function ActionComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li");
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = ["*"];
export class ActionComponent {
    constructor(actionService) {
        this.actionService = actionService;
        /**
         * Is this action disabled.
         */
        this.disabled = false;
        /**
         * The priority of this action (the higher the number, the higher the action will be displayed; optional)
         */
        this.priority = 0;
    }
    /**
     * After page view is initialized, component adds itself to global action list.
     */
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.actionService.add(this));
    }
    /**
     * Action is removed from global action list, if component is not used anymore.
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        this.actionService.remove(this);
    }
}
ActionComponent.ɵfac = function ActionComponent_Factory(t) { return new (t || ActionComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActionService)); };
ActionComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: ActionComponent, selectors: [["c8y-action"]], viewQuery: function ActionComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { disabled: "disabled", priority: "priority", label: "label", icon: "icon" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""]], template: function ActionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, ActionComponent_ng_template_0_Template, 2, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
ActionComponent.ctorParameters = () => [
    { type: ActionService }
];
ActionComponent.propDecorators = {
    disabled: [{ type: Input }],
    label: [{ type: Input }],
    priority: [{ type: Input }],
    icon: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ActionComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-action',
                template: "<ng-template #template>\n  <li>\n    <ng-content></ng-content>\n  </li>\n</ng-template>"
            }]
    }], function () { return [{ type: ɵngcc1.ActionService }]; }, { disabled: [{
            type: Input
        }], priority: [{
            type: Input
        }], label: [{
            type: Input
        }], icon: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: false }]
        }] }); })();
