// @c8y/ngx-components/esm2015/sub-assets/columns/asset-type.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext, gettext } from '@c8y/ngx-components';
import { DeviceGroupService } from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from '@c8y/ngx-components/assets-navigator';
import * as ɵngcc3 from '@angular/common';

function AssetTypeCellRendererComponent_device_status_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "device-status", 2);
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("mo", ctx_r0.context.item);
} }
function AssetTypeCellRendererComponent_i_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 3);
    ɵngcc0.ɵɵpipe(1, "translate");
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(1, 2, ctx_r1.iconTitle));
    ɵngcc0.ɵɵproperty("c8yIcon", ctx_r1.groupIcon);
} }
export class AssetTypeCellRendererComponent {
    constructor(context, deviceGroupService) {
        this.context = context;
        this.deviceGroupService = deviceGroupService;
    }
    ngOnInit() {
        this.isAnyGroup = this.deviceGroupService.isAnyGroup(this.context.item);
        this.isDevice = this.deviceGroupService.isDevice(this.context.item);
        this.groupIcon = this.deviceGroupService.icon(this.context.item);
        this.iconTitle = this.getIconTitle(this.groupIcon);
    }
    getIconTitle(icon) {
        switch (icon) {
            case 'c8y-group':
                return gettext('Group');
            case 'c8y-group-smart':
                return gettext('Smart group');
            case 'c8y-group-remote-inactive':
                return 'Inactive remote group';
            case 'c8y-group-remote':
                return gettext('Remote group');
            default:
                return gettext('Child device');
        }
    }
}
AssetTypeCellRendererComponent.ɵfac = function AssetTypeCellRendererComponent_Factory(t) { return new (t || AssetTypeCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DeviceGroupService)); };
AssetTypeCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: AssetTypeCellRendererComponent, selectors: [["ng-component"]], decls: 2, vars: 2, consts: [["class", "m-l-8", 3, "mo", 4, "ngIf"], ["class", "m-l-8 c8y-icon-duocolor", 3, "c8yIcon", "title", 4, "ngIf"], [1, "m-l-8", 3, "mo"], [1, "m-l-8", "c8y-icon-duocolor", 3, "c8yIcon", "title"]], template: function AssetTypeCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, AssetTypeCellRendererComponent_device_status_0_Template, 1, 1, "device-status", 0);
        ɵngcc0.ɵɵtemplate(1, AssetTypeCellRendererComponent_i_1_Template, 2, 4, "i", 1);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isDevice);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isAnyGroup || !ctx.isDevice);
    } }, directives: [ɵngcc3.NgIf, ɵngcc1.DeviceStatusComponent, ɵngcc1.IconDirective], pipes: [ɵngcc1.C8yTranslatePipe], encapsulation: 2 });
AssetTypeCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext },
    { type: DeviceGroupService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetTypeCellRendererComponent, [{
        type: Component,
        args: [{
                template: "<device-status class=\"m-l-8\" [mo]=\"context.item\" *ngIf=\"isDevice\"></device-status>\n<i\n  [c8yIcon]=\"groupIcon\"\n  title=\"{{ iconTitle | translate }}\"\n  class=\"m-l-8 c8y-icon-duocolor\"\n  *ngIf=\"isAnyGroup || !isDevice\"\n></i>\n"
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }, { type: ɵngcc2.DeviceGroupService }]; }, null); })();
