// @c8y/ngx-components/esm2015/core/dynamic-forms/json-schema/c8y-json-schema.service.js
import { Injectable } from '@angular/core';
import { FormlyJsonschema } from '@ngx-formly/core/json-schema';
import * as i0 from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class C8yJSONSchema extends FormlyJsonschema {
    toFieldConfig(schema, options) {
        const config = super.toFieldConfig(schema, this.c8yOptions((options || {}).map));
        if (Array.isArray(config.fieldGroup)) {
            config.fieldGroup.forEach((field) => {
                if (field.type === 'array' && field.templateOptions.required) {
                    field.fieldArray.templateOptions.required = true;
                }
            });
        }
        return config;
    }
    c8yOptions(map) {
        return {
            map(mappedField, mapSource) {
                let result = mappedField;
                /** `const` type is represented as hidden input */
                if (mapSource.const) {
                    result = Object.assign(Object.assign({}, result), { defaultValue: mapSource.const, templateOptions: Object.assign(Object.assign({}, result.templateOptions), { type: 'hidden' }) });
                }
                /** `writeOnly` field represented as password input */
                if (mapSource.writeOnly) {
                    result = Object.assign(Object.assign({}, result), { templateOptions: Object.assign(Object.assign({}, result.templateOptions), { type: 'password', autocomplete: 'off' }) });
                }
                /** `examples` values used as placeholder */
                if (mapSource.examples) {
                    result = Object.assign(Object.assign({}, result), { templateOptions: Object.assign(Object.assign({}, result.templateOptions), { placeholder: mapSource.examples.join(', ') }) });
                }
                /** `enum` field represented as radio button group */
                if (mapSource.enum) {
                    result = Object.assign(Object.assign({}, result), { type: 'radio' });
                }
                /** fields with `contentMediaType` or `contentEncoding` are represented as file inputs */
                if (mapSource.contentMediaType || mapSource.contentEncoding) {
                    result = Object.assign(Object.assign({}, result), { type: 'file', templateOptions: Object.assign(Object.assign({}, result.templateOptions), { accept: mapSource.contentMediaType, maxAllowedFiles: 1, contentEncoding: mapSource.contentEncoding }) });
                }
                return map ? map(result, mapSource) : result;
            }
        };
    }
}
C8yJSONSchema.ɵfac = /*@__PURE__*/ function () { let ɵC8yJSONSchema_BaseFactory; return function C8yJSONSchema_Factory(t) { return (ɵC8yJSONSchema_BaseFactory || (ɵC8yJSONSchema_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(C8yJSONSchema)))(t || C8yJSONSchema); }; }();
C8yJSONSchema.ɵprov = i0.ɵɵdefineInjectable({ factory: function C8yJSONSchema_Factory() { return new C8yJSONSchema(); }, token: C8yJSONSchema, providedIn: "root" });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(C8yJSONSchema, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], null, null); })();
