// @c8y/ngx-components/esm2015/operations/bulk-operations-stepper-container/bulk-operations-stepper-container.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CoreModule } from '@c8y/ngx-components';
import { BulkOperationsStepperContainerComponent } from './bulk-operations-stepper-container.component';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export const bulkOperatioStepperRoutes = [
    {
        path: 'devicecontrol/bulk/creation/:type',
        component: BulkOperationsStepperContainerComponent
    }
];
/**
 * Container module for the new bulk operation steppers. Use this module in your application
 * to provide a step by step guided process for scheduling of bulk operations.
 * The module reads the bulk operation type from the URL and displays the corresponding stepper.
 * You can provide stepper for your custom bulk operation type by registering it with the
 * HOOK_LIST_BULK_TYPE injection token.
 */
export class BulkOperationsStepperContainerModule {
}
BulkOperationsStepperContainerModule.ɵfac = function BulkOperationsStepperContainerModule_Factory(t) { return new (t || BulkOperationsStepperContainerModule)(); };
BulkOperationsStepperContainerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: BulkOperationsStepperContainerModule });
BulkOperationsStepperContainerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BulkOperationsStepperContainerModule, [{
        type: NgModule,
        args: [{
                declarations: [BulkOperationsStepperContainerComponent],
                imports: [CoreModule, RouterModule.forChild(bulkOperatioStepperRoutes)],
                entryComponents: [BulkOperationsStepperContainerComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BulkOperationsStepperContainerModule, { declarations: function () { return [BulkOperationsStepperContainerComponent]; }, imports: function () { return [CoreModule, ɵngcc1.RouterModule]; } }); })();
