// @c8y/ngx-components/esm2015/core/header/title/title.component.js
import { Component, ViewChild, Input } from '@angular/core';
import { HeaderService } from '../header.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../header.service';

const _c0 = ["template"];
function TitleComponent_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "h1", 1);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
const _c1 = ["*"];
export class TitleComponent {
    constructor(headerService) {
        this.headerService = headerService;
        this.pageTitleUpdate = true;
    }
    ngAfterViewInit() {
        this.viewInitTimeout = setTimeout(() => this.headerService.changeTitle(this.template, this.pageTitleUpdate));
    }
    ngOnDestroy() {
        if (this.headerService.state.title === this.template) {
            clearTimeout(this.viewInitTimeout);
            this.headerService.changeTitle();
        }
    }
}
TitleComponent.ɵfac = function TitleComponent_Factory(t) { return new (t || TitleComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.HeaderService)); };
TitleComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: TitleComponent, selectors: [["c8y-title"]], viewQuery: function TitleComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.template = _t.first);
    } }, inputs: { pageTitleUpdate: "pageTitleUpdate" }, ngContentSelectors: _c1, decls: 2, vars: 0, consts: [["template", ""], [1, "text-truncate"]], template: function TitleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵtemplate(0, TitleComponent_ng_template_0_Template, 2, 0, "ng-template", null, 0, ɵngcc0.ɵɵtemplateRefExtractor);
    } }, encapsulation: 2 });
TitleComponent.ctorParameters = () => [
    { type: HeaderService }
];
TitleComponent.propDecorators = {
    pageTitleUpdate: [{ type: Input }],
    template: [{ type: ViewChild, args: ['template', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TitleComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-title',
                template: "<ng-template #template>\n  <h1 class=\"text-truncate\">\n    <ng-content></ng-content>\n  </h1>\n</ng-template>\n\n"
            }]
    }], function () { return [{ type: ɵngcc1.HeaderService }]; }, { pageTitleUpdate: [{
            type: Input
        }], template: [{
            type: ViewChild,
            args: ['template', { static: true }]
        }] }); })();
