// @c8y/ngx-components/esm2015/trusted-certificates/trusted-certificates-navigation.factory.js
import { __awaiter } from "tslib";
import { Injectable } from '@angular/core';
import { gettext, NavigatorNode } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
export class TrustedCertificatesNavigationFactory {
    constructor() {
        this.navs = [];
    }
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.navs.length === 0) {
                this.navs.push(new NavigatorNode({
                    label: gettext('Trusted certificates'),
                    icon: 'certificate',
                    path: '/trusted-certificates',
                    parent: gettext('Management'),
                    priority: 100
                }));
            }
            return this.navs;
        });
    }
}
TrustedCertificatesNavigationFactory.ɵfac = function TrustedCertificatesNavigationFactory_Factory(t) { return new (t || TrustedCertificatesNavigationFactory)(); };
TrustedCertificatesNavigationFactory.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: TrustedCertificatesNavigationFactory, factory: TrustedCertificatesNavigationFactory.ɵfac });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(TrustedCertificatesNavigationFactory, [{
        type: Injectable
    }], function () { return []; }, null); })();
