// @c8y/ngx-components/esm2015/app-logs/app-logs-auto-refresh.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@c8y/ngx-components';
import { ButtonsModule } from 'ngx-bootstrap/buttons';
import { AppLogsAutoRefreshComponent } from './app-logs-auto-refresh.component';
import { AppLogsService } from './app-logs.service';
import * as ɵngcc0 from '@angular/core';
export class AppLogsAutoRefreshModule {
}
AppLogsAutoRefreshModule.ɵfac = function AppLogsAutoRefreshModule_Factory(t) { return new (t || AppLogsAutoRefreshModule)(); };
AppLogsAutoRefreshModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AppLogsAutoRefreshModule });
AppLogsAutoRefreshModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [AppLogsService], imports: [[CommonModule, ButtonsModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AppLogsAutoRefreshModule, [{
        type: NgModule,
        args: [{
                declarations: [AppLogsAutoRefreshComponent],
                exports: [AppLogsAutoRefreshComponent],
                imports: [CommonModule, ButtonsModule],
                entryComponents: [AppLogsAutoRefreshComponent],
                providers: [AppLogsService]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AppLogsAutoRefreshModule, { declarations: function () { return [AppLogsAutoRefreshComponent]; }, imports: function () { return [CommonModule, ButtonsModule]; }, exports: function () { return [AppLogsAutoRefreshComponent]; } }); })();
