// @c8y/ngx-components/esm2015/device-grid/columns/serial-number.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import { ColumnUtilService } from './column-util.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
import * as ɵngcc2 from './column-util.service';
export class SerialNumberCellRendererComponent {
    constructor(context, columnUtilService) {
        this.context = context;
        this.columnUtilService = columnUtilService;
    }
}
SerialNumberCellRendererComponent.ɵfac = function SerialNumberCellRendererComponent_Factory(t) { return new (t || SerialNumberCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.ColumnUtilService)); };
SerialNumberCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: SerialNumberCellRendererComponent, selectors: [["ng-component"]], decls: 1, vars: 1, template: function SerialNumberCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtext(0);
    } if (rf & 2) {
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.columnUtilService.getSerialNumber(ctx.context.item), " ");
    } }, encapsulation: 2 });
SerialNumberCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext },
    { type: ColumnUtilService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SerialNumberCellRendererComponent, [{
        type: Component,
        args: [{
                template: `
    {{ columnUtilService.getSerialNumber(context.item) }}
  `
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }, { type: ɵngcc2.ColumnUtilService }]; }, null); })();
