// @c8y/ngx-components/esm2015/device-grid/columns/status.cell-renderer.component.js
import { Component } from '@angular/core';
import { CellRendererContext } from '@c8y/ngx-components';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@c8y/ngx-components';
export class StatusCellRendererComponent {
    constructor(context) {
        this.context = context;
    }
}
StatusCellRendererComponent.ɵfac = function StatusCellRendererComponent_Factory(t) { return new (t || StatusCellRendererComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CellRendererContext)); };
StatusCellRendererComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: StatusCellRendererComponent, selectors: [["ng-component"]], decls: 1, vars: 1, consts: [[3, "mo"]], template: function StatusCellRendererComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "device-status", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("mo", ctx.context.item);
    } }, directives: [ɵngcc1.DeviceStatusComponent], encapsulation: 2 });
StatusCellRendererComponent.ctorParameters = () => [
    { type: CellRendererContext }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(StatusCellRendererComponent, [{
        type: Component,
        args: [{
                template: `
    <device-status [mo]="context.item"></device-status>
  `
            }]
    }], function () { return [{ type: ɵngcc1.CellRendererContext }]; }, null); })();
