// @c8y/ngx-components/esm2015/operations/bulk-operation-stepper/custom-step.directive.js
import { Directive, Input, TemplateRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class CustomStep {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.buttonsDisabled = false;
        this.onNext = ({ stepper, step }) => {
            // steps without own `onNext` handler, e.g. preview steps need to mark themselves as `completed`,
            // otherwise stepper will not allow to move forth from them as soon as the user navigates back
            // and the `c8y-stepper-buttons` component marks the step as incomplete.
            step.completed = true;
            stepper.next();
        };
    }
}
CustomStep.ɵfac = function CustomStep_Factory(t) { return new (t || CustomStep)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.TemplateRef)); };
CustomStep.ɵdir = /*@__PURE__*/ ɵngcc0.ɵɵdefineDirective({ type: CustomStep, selectors: [["", "customStep", ""]], inputs: { buttonsDisabled: ["customStepButtonsDisabled", "buttonsDisabled"], onNext: ["customStepOnNext", "onNext"], label: ["customStep", "label"], completed: ["customStepCompleted", "completed"] } });
CustomStep.ctorParameters = () => [
    { type: TemplateRef }
];
CustomStep.propDecorators = {
    label: [{ type: Input, args: ['customStep',] }],
    completed: [{ type: Input, args: ['customStepCompleted',] }],
    buttonsDisabled: [{ type: Input, args: ['customStepButtonsDisabled',] }],
    onNext: [{ type: Input, args: ['customStepOnNext',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CustomStep, [{
        type: Directive,
        args: [{ selector: '[customStep]' }]
    }], function () { return [{ type: ɵngcc0.TemplateRef }]; }, { buttonsDisabled: [{
            type: Input,
            args: ['customStepButtonsDisabled']
        }], onNext: [{
            type: Input,
            args: ['customStepOnNext']
        }], label: [{
            type: Input,
            args: ['customStep']
        }], completed: [{
            type: Input,
            args: ['customStepCompleted']
        }] }); })();
