// @c8y/ngx-components/esm2015/assets-navigator/asset-selector/asset-selector.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@c8y/ngx-components';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { GroupNodeService } from './group-node.service';
import { AssetSelectorNodeComponent } from './asset-selector-node.component';
import { AssetSelectorComponent } from './asset-selector.component';
import * as ɵngcc0 from '@angular/core';
export class AssetSelectorModule {
}
AssetSelectorModule.ɵfac = function AssetSelectorModule_Factory(t) { return new (t || AssetSelectorModule)(); };
AssetSelectorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: AssetSelectorModule });
AssetSelectorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [GroupNodeService], imports: [[CommonModule, CollapseModule, BsDropdownModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(AssetSelectorModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, CollapseModule, BsDropdownModule],
                exports: [AssetSelectorNodeComponent, AssetSelectorComponent],
                providers: [GroupNodeService],
                declarations: [AssetSelectorNodeComponent, AssetSelectorComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AssetSelectorModule, { declarations: function () { return [AssetSelectorNodeComponent, AssetSelectorComponent]; }, imports: function () { return [CommonModule, CollapseModule, BsDropdownModule]; }, exports: function () { return [AssetSelectorNodeComponent, AssetSelectorComponent]; } }); })();
ɵngcc0.ɵɵsetComponentScope(AssetSelectorComponent, [AssetSelectorNodeComponent], []);
