// @c8y/ngx-components/esm2015/core/navigator/navigator.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { CommonModule } from '../common/common.module';
import { NavigatorItemComponent } from './navigator-entry.component';
import { NavigatorIconComponent } from './navigator-icon.component';
import { NavigatorNodeComponent } from './navigator-node.component';
import { NavigatorOutletComponent } from './navigator-outlet.component';
import { ModalModule } from '../modal/modal.module';
/**
 * The angular module definition for the navigator.
 * @exports NavigatorOutletComponent
 * @exports NavigatorItemComponent
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ngx-bootstrap/collapse';
export class NavigatorModule {
}
NavigatorModule.ɵfac = function NavigatorModule_Factory(t) { return new (t || NavigatorModule)(); };
NavigatorModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NavigatorModule });
NavigatorModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            RouterModule,
            CollapseModule.forRoot(),
            CommonModule,
            ModalModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavigatorModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    NavigatorOutletComponent,
                    NavigatorNodeComponent,
                    NavigatorItemComponent,
                    NavigatorIconComponent
                ],
                imports: [
                    RouterModule,
                    CollapseModule.forRoot(),
                    CommonModule,
                    ModalModule
                ],
                exports: [
                    NavigatorOutletComponent,
                    NavigatorItemComponent
                ],
                entryComponents: [
                    NavigatorIconComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NavigatorModule, { declarations: function () { return [NavigatorOutletComponent,
        NavigatorNodeComponent,
        NavigatorItemComponent,
        NavigatorIconComponent]; }, imports: function () { return [RouterModule, ɵngcc1.CollapseModule, CommonModule,
        ModalModule]; }, exports: function () { return [NavigatorOutletComponent,
        NavigatorItemComponent]; } }); })();
