// @c8y/ngx-components/esm2015/core/navigator/navigator-node.component.js
// tslint:disable-next-line:max-line-length
import { Component, ComponentFactoryResolver, Input, ViewChild, ViewChildren, ViewContainerRef, Output, EventEmitter } from '@angular/core';
import { RouterLinkActive } from '@angular/router';
import { NavigatorIconComponent } from './navigator-icon.component';
import { NavigatorService } from './navigator.service';
import { PopoverConfirmComponent } from '../modal/popover-confirm.component';
/**
 * Navigator node renderer.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './navigator.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from '@angular/router';
import * as ɵngcc4 from 'ngx-bootstrap/collapse';
import * as ɵngcc5 from '../modal/popover-confirm.component';
import * as ɵngcc6 from '../common/icon.directive';
import * as ɵngcc7 from '../i18n/c8y-translate.pipe';

const _c0 = ["icon"];
function NavigatorNodeComponent_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function NavigatorNodeComponent_div_5_c8y_navigator_node_display_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-navigator-node-display", 7);
    ɵngcc0.ɵɵlistener("nodeClick", function NavigatorNodeComponent_div_5_c8y_navigator_node_display_1_Template_c8y_navigator_node_display_nodeClick_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(2); return ctx_r6.nodeClick.emit($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const childNode_r5 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("node", childNode_r5);
} }
function NavigatorNodeComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵtemplate(1, NavigatorNodeComponent_div_5_c8y_navigator_node_display_1_Template, 1, 1, "c8y-navigator-node-display", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("collapse", !ctx_r1.node.open)("isAnimated", true);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r1.node.children);
} }
function NavigatorNodeComponent_ng_template_6_i_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 12);
} if (rf & 2) {
    ɵngcc0.ɵɵproperty("c8yIcon", "circle-o-notch");
} }
function NavigatorNodeComponent_ng_template_6_i_7_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "i", 13);
    ɵngcc0.ɵɵlistener("click", function NavigatorNodeComponent_ng_template_6_i_7_Template_i_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(2); return ctx_r11.click("expander", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵpropertyInterpolate("title", ctx_r10.expandTitle);
    ɵngcc0.ɵɵproperty("c8yIcon", "chevron-down");
} }
function NavigatorNodeComponent_ng_template_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, NavigatorNodeComponent_ng_template_6_i_0_Template, 1, 1, "i", 8);
    ɵngcc0.ɵɵelementStart(1, "div");
    ɵngcc0.ɵɵelementContainer(2, null, 9);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "span");
    ɵngcc0.ɵɵtext(5);
    ɵngcc0.ɵɵpipe(6, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(7, NavigatorNodeComponent_ng_template_6_i_7_Template, 1, 2, "i", 10);
    ɵngcc0.ɵɵelement(8, "c8y-popover-confirm", 11);
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.node.loading && !ctx_r3.isRoot);
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(6, 3, ctx_r3.node.label));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.node.hasChildren);
} }
const _c1 = function (a0, a1) { return { dragged: a0, disabled: a1 }; };
const _c2 = function (a0) { return { exact: a0 }; };
const _c3 = function (a0, a1, a2, a3) { return { "root-link": a0, open: a1, parent: a2, "dragged-hover": a3 }; };
export class NavigatorNodeComponent {
    constructor(componentFactoryResolver, navigator) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.navigator = navigator;
        /**
         * Event emitter responsible for broadcasting one of the following events: "icon", "expander" or "link" as string value.
         *
         * The type of event depends on where you click on the navigator node:
         * * clicking the icon will broadcast the event "icon",
         * * clicking the expander will broadcast the event "expander",
         * * clicking the label will broadcast the event "link".
         */
        this.nodeClick = new EventEmitter();
    }
    /**
     * @ignore
     */
    ngAfterViewInit() {
        this.childrenNodes.forEach(n => (n.parentNode = this));
        this.viewInitTimeout = setTimeout(() => {
            this.updateIcon();
            this.openOnFirstUrl();
            this.expandParentIfActive();
            this.node.confirm = this.confirm;
        });
    }
    /**
     * Expands the navigator node when the URL matches the path provided in the node.
     */
    openOnFirstUrl() {
        if (this.navigator.openOnFirstUrl(this.node) && !this.node.open) {
            this.expandCollapse(true);
        }
    }
    /**
     * Returns the expander title.
     */
    get expandTitle() {
        return !this.node.open ? 'Expand' : 'Collapse';
    }
    /**
     * Click event handler.
     *
     * @param {string} from Source of the click event.
     * @param event DOM event.
     */
    click(from = 'link', $event) {
        let open;
        switch (from) {
            case "expander" /* EXPANDER */:
                if ($event) {
                    $event.preventDefault();
                    $event.stopPropagation();
                }
                open = !this.node.open;
                break;
            case "icon" /* ICON */:
            case "link" /* LINK */:
                open = this.routerLinkActive && this.routerLinkActive.isActive ? true : !this.node.open;
                break;
        }
        this.handleExpandCollapse(open, from, $event);
        this.nodeClick.emit(from);
    }
    /**
     * Expands or collapses the navigator node and its children recursively.
     *
     * @param {boolean} open Open or close the node.
     * @param {boolean} recursive Open or close the children.
     * @param {ClickOptions} forNode Click options.
     */
    expandCollapse(open, recursive, forNode = {}) {
        this.node.open = open;
        if (recursive) {
            this.childrenNodes.forEach(n => n.expandCollapse(open, recursive));
        }
        forNode.open = open;
        this.node.click(forNode);
    }
    /**
     * @ignore
     */
    ngOnDestroy() {
        clearTimeout(this.viewInitTimeout);
        if (this.node) {
            this.node.destroy();
        }
    }
    handleExpandCollapse(open, from, $event) {
        if (open === this.node.open ||
            (this.node.path && !open && from !== "expander" /* EXPANDER */)) {
            return;
        }
        this.expandCollapse(open, false, { [from]: true, $event });
    }
    expandParentIfActive() {
        if (this.routerLinkActive && this.routerLinkActive.isActive && this.parentNode) {
            this.parentNode.expandCollapse(true);
        }
    }
    updateIcon() {
        if (this.node.iconTemplate) {
            this.updateIconTemplate();
        }
        else {
            this.updateIconComponent();
        }
    }
    updateIconComponent() {
        let componentType = NavigatorIconComponent;
        if (this.node && this.node.iconComponent) {
            componentType = this.node.iconComponent;
        }
        this.iconSlot.clear();
        const iconComponentFactory = this.componentFactoryResolver.resolveComponentFactory(componentType);
        const componentRef = this.iconSlot.createComponent(iconComponentFactory);
        componentRef.instance.node = this.node;
    }
    updateIconTemplate() {
        this.iconSlot.clear();
        this.iconSlot.createEmbeddedView(this.node.iconTemplate);
    }
}
NavigatorNodeComponent.ɵfac = function NavigatorNodeComponent_Factory(t) { return new (t || NavigatorNodeComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ComponentFactoryResolver), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NavigatorService)); };
NavigatorNodeComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: NavigatorNodeComponent, selectors: [["c8y-navigator-node-display"]], viewQuery: function NavigatorNodeComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 5, ViewContainerRef);
        ɵngcc0.ɵɵviewQuery(RouterLinkActive, 5);
        ɵngcc0.ɵɵviewQuery(PopoverConfirmComponent, 5);
        ɵngcc0.ɵɵviewQuery(NavigatorNodeComponent, 5);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.iconSlot = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.routerLinkActive = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.confirm = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.childrenNodes = _t);
    } }, inputs: { node: "node", isRoot: "isRoot" }, outputs: { nodeClick: "nodeClick" }, decls: 8, vars: 23, consts: [[1, "slot", 3, "hidden", "draggable", "ngClass", "dragstart", "dragend", "drop"], ["draggable", "false", 1, "link", 3, "title", "routerLink", "routerLinkActive", "routerLinkActiveOptions", "ngClass", "dragover", "dragenter", "dragleave", "click"], [4, "ngTemplateOutlet"], ["class", "children panel-expand expand", 3, "collapse", "isAnimated", 4, "ngIf"], ["inner", ""], [1, "children", "panel-expand", "expand", 3, "collapse", "isAnimated"], [3, "node", "nodeClick", 4, "ngFor", "ngForOf"], [3, "node", "nodeClick"], ["class", "icon-spin loadingIndicator", 3, "c8yIcon", 4, "ngIf"], ["icon", ""], ["class", "expander", 3, "c8yIcon", "title", "click", 4, "ngIf"], ["outsideClick", "true", "containerClass", "navigator-popover"], [1, "icon-spin", "loadingIndicator", 3, "c8yIcon"], [1, "expander", 3, "c8yIcon", "title", "click"]], template: function NavigatorNodeComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("dragstart", function NavigatorNodeComponent_Template_div_dragstart_0_listener($event) { return ctx.node.dragStart($event); })("dragend", function NavigatorNodeComponent_Template_div_dragend_0_listener($event) { return ctx.node.dragEnd($event); })("drop", function NavigatorNodeComponent_Template_div_drop_0_listener($event) { return ctx.node.drop($event); });
        ɵngcc0.ɵɵelementContainerStart(1);
        ɵngcc0.ɵɵelementStart(2, "a", 1);
        ɵngcc0.ɵɵlistener("dragover", function NavigatorNodeComponent_Template_a_dragover_2_listener($event) { return ctx.node.canDrop && $event.preventDefault(); })("dragenter", function NavigatorNodeComponent_Template_a_dragenter_2_listener($event) { return ctx.node.canDrop && ctx.node.dragEnter($event); })("dragleave", function NavigatorNodeComponent_Template_a_dragleave_2_listener($event) { return ctx.node.canDrop && ctx.node.dragLeave($event); })("click", function NavigatorNodeComponent_Template_a_click_2_listener($event) { return ctx.click(ctx.node.canNavigate ? "link" : "expander", $event); });
        ɵngcc0.ɵɵpipe(3, "translate");
        ɵngcc0.ɵɵtemplate(4, NavigatorNodeComponent_ng_container_4_Template, 1, 0, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementContainerEnd();
        ɵngcc0.ɵɵtemplate(5, NavigatorNodeComponent_div_5_Template, 2, 3, "div", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, NavigatorNodeComponent_ng_template_6_Template, 9, 5, "ng-template", null, 4, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        const _r2 = ɵngcc0.ɵɵreference(7);
        ɵngcc0.ɵɵproperty("hidden", ctx.node.hidden)("draggable", ctx.node.draggable)("ngClass", ɵngcc0.ɵɵpureFunction2(13, _c1, ctx.node.dragged, ctx.node.loading));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(3, 11, ctx.node.label));
        ɵngcc0.ɵɵproperty("routerLink", ctx.node.canNavigate ? ctx.node.path : undefined)("routerLinkActive", ctx.node.canNavigate ? "active" : "")("routerLinkActiveOptions", ɵngcc0.ɵɵpureFunction1(16, _c2, ctx.node.routerLinkExact))("ngClass", ɵngcc0.ɵɵpureFunction4(18, _c3, ctx.isRoot, ctx.node.open && ctx.node.hasChildren, ctx.node.hasChildren, ctx.node.draggedHover && !ctx.node.dragged));
        ɵngcc0.ɵɵattribute("id", ctx.isRoot ? ctx.node.id : undefined);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r2);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.node.children.length);
    } }, directives: [ɵngcc2.NgClass, ɵngcc3.RouterLinkWithHref, ɵngcc3.RouterLinkActive, ɵngcc2.NgTemplateOutlet, ɵngcc2.NgIf, ɵngcc4.CollapseDirective, ɵngcc2.NgForOf, NavigatorNodeComponent, ɵngcc5.PopoverConfirmComponent, ɵngcc6.IconDirective], pipes: [ɵngcc7.C8yTranslatePipe], encapsulation: 2 });
NavigatorNodeComponent.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: NavigatorService }
];
NavigatorNodeComponent.propDecorators = {
    iconSlot: [{ type: ViewChild, args: ['icon', { read: ViewContainerRef, static: false },] }],
    node: [{ type: Input }],
    isRoot: [{ type: Input }],
    routerLinkActive: [{ type: ViewChild, args: [RouterLinkActive, { static: false },] }],
    childrenNodes: [{ type: ViewChildren, args: [NavigatorNodeComponent,] }],
    nodeClick: [{ type: Output }],
    confirm: [{ type: ViewChild, args: [PopoverConfirmComponent, { static: false },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NavigatorNodeComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-navigator-node-display',
                template: "<div\n  class=\"slot\"\n  [hidden]=\"node.hidden\"\n  (dragstart)=\"node.dragStart($event)\"\n  (dragend)=\"node.dragEnd($event)\"\n  (drop)=\"node.drop($event)\"\n  [draggable]=\"node.draggable\"\n  [ngClass]=\"{ dragged: node.dragged, disabled: node.loading }\"\n>\n  <ng-container>\n    <a\n      class=\"link\"\n      draggable=\"false\"\n      (dragover)=\"node.canDrop && $event.preventDefault()\"\n      (dragenter)=\"node.canDrop && node.dragEnter($event)\"\n      (dragleave)=\"node.canDrop && node.dragLeave($event)\"\n      title=\"{{node.label | translate}}\"\n      [attr.id]=\"isRoot ? node.id : undefined\"\n      [routerLink]=\"node.canNavigate ? node.path : undefined\"\n      [routerLinkActive]=\"node.canNavigate ? 'active' : ''\"\n      [routerLinkActiveOptions]=\"{exact:node.routerLinkExact}\"\n      (click)=\"click(node.canNavigate ? 'link' : 'expander', $event)\"\n      [ngClass]=\"{'root-link':isRoot, open: node.open && node.hasChildren, parent: node.hasChildren, 'dragged-hover': node.draggedHover && !node.dragged}\"\n    >\n      <ng-container *ngTemplateOutlet=\"inner\"></ng-container>\n    </a>\n  </ng-container>\n\n  <div\n    class=\"children panel-expand expand\"\n    *ngIf=\"node.children.length\"\n    [collapse]=\"!node.open\"\n    [isAnimated]=\"true\"\n  >\n    <c8y-navigator-node-display\n      *ngFor=\"let childNode of node.children\"\n      [node]=\"childNode\"\n      (nodeClick)=\"nodeClick.emit($event)\"\n    >\n    </c8y-navigator-node-display>\n  </div>\n</div>\n\n<ng-template #inner>\n  <!-- loader -->\n  <i\n    class=\"icon-spin loadingIndicator\"\n    [c8yIcon]=\"'circle-o-notch'\"\n    *ngIf=\"node.loading && !isRoot\"\n  >\n  </i>\n\n  <!-- icon -->\n  <div>\n    <ng-container #icon></ng-container>\n  </div>\n\n  <!--title  -->\n  <span>{{node.label | translate}}</span>\n\n  <i\n    class=\"expander\"\n    [c8yIcon]=\"'chevron-down'\"\n    *ngIf=\"node.hasChildren\"\n    (click)=\"click('expander', $event)\"\n    title=\"{{expandTitle}}\"\n  >\n  </i>\n\n  <c8y-popover-confirm\n    outsideClick=\"true\"\n    containerClass=\"navigator-popover\"\n  ></c8y-popover-confirm>\n</ng-template>\n"
            }]
    }], function () { return [{ type: ɵngcc0.ComponentFactoryResolver }, { type: ɵngcc1.NavigatorService }]; }, { nodeClick: [{
            type: Output
        }], iconSlot: [{
            type: ViewChild,
            args: ['icon', { read: ViewContainerRef, static: false }]
        }], node: [{
            type: Input
        }], isRoot: [{
            type: Input
        }], routerLinkActive: [{
            type: ViewChild,
            args: [RouterLinkActive, { static: false }]
        }], childrenNodes: [{
            type: ViewChildren,
            args: [NavigatorNodeComponent]
        }], confirm: [{
            type: ViewChild,
            args: [PopoverConfirmComponent, { static: false }]
        }] }); })();
