// @c8y/ngx-components/esm2015/ecosystem/features/feature-list.component.js
import { Component } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { shareReplay, switchMap, tap } from 'rxjs/operators';
import { EcosystemService } from '../ecosystem.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../ecosystem.service';
import * as ɵngcc2 from '@c8y/ngx-components';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../shared/application-card.component';

function FeatureListComponent_div_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 11);
    ɵngcc0.ɵɵelement(1, "h1", 12);
    ɵngcc0.ɵɵelementStart(2, "h3", 13);
    ɵngcc0.ɵɵtext(3, "No features to display.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "p", 13);
    ɵngcc0.ɵɵtext(5, "No additional features are subscribed to the tenant.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function FeatureListComponent_div_22_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 14);
    ɵngcc0.ɵɵelementStart(1, "div", 15);
    ɵngcc0.ɵɵelement(2, "div", 16);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵpipe(4, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div", 17);
    ɵngcc0.ɵɵelementStart(6, "div", 18);
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵpipe(8, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(9, "div", 19);
    ɵngcc0.ɵɵtext(10);
    ɵngcc0.ɵɵpipe(11, "translate");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(12, "div", 19);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(4, 3, "Feature"), " ");
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(8, 5, "Description"));
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(11, 7, "Type"));
} }
function FeatureListComponent_div_24_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 20);
    ɵngcc0.ɵɵelement(1, "c8y-application-card", 21);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const feature_r3 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("app", feature_r3)("canEdit", false);
} }
const _c0 = function (a0) { return { "icon-spin": a0 }; };
export class FeatureListComponent {
    constructor(ecosystemService) {
        this.ecosystemService = ecosystemService;
        this.reloading = false;
        this.reload$ = new BehaviorSubject(null);
        this.features$ = this.reload$.pipe(tap(() => (this.reloading = true)), switchMap(() => this.ecosystemService.getFeatureApplications()), tap(() => (this.reloading = false)), shareReplay());
    }
    ngOnInit() {
        this.loadFeatures();
    }
    loadFeatures() {
        this.reload$.next();
    }
}
FeatureListComponent.ɵfac = function FeatureListComponent_Factory(t) { return new (t || FeatureListComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.EcosystemService)); };
FeatureListComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FeatureListComponent, selectors: [["c8y-feature-list"]], decls: 26, vars: 40, consts: [[3, "icon", "label"], [3, "icon", "label", "path"], [3, "placement"], [1, "btn", "btn-link", 3, "title", "click"], ["c8yIcon", "refresh", 3, "ngClass"], ["itemClass", "navbar-form hidden-xs", 3, "placement"], [3, "listLength", "onListClassChange"], ["class", "c8y-empty-state text-center", 4, "ngIf"], [1, "card-group", 3, "ngClass"], ["class", "page-sticky-header hidden-xs flex-row", 4, "ngIf"], ["class", "col-xs-12 col-sm-4 col-md-3", 4, "ngFor", "ngForOf"], [1, "c8y-empty-state", "text-center"], [1, "c8y-icon", "c8y-icon-modules", "c8y-icon-duocolor"], ["translate", ""], [1, "page-sticky-header", "hidden-xs", "flex-row"], [1, "card-block", "card-column-40"], [1, "card-appicon", "p-l-32", "p-r-16", "m-r-0", "m-l-4"], [1, "card-block", "p-0", "card-column-80"], [1, "card-block", "card-column-80"], [1, "card-block", "card-column-20"], [1, "col-xs-12", "col-sm-4", "col-md-3"], [1, "d-contents", 3, "app", "canEdit"]], template: function FeatureListComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "c8y-title");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵpipe(2, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "c8y-breadcrumb");
        ɵngcc0.ɵɵelement(4, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(5, "translate");
        ɵngcc0.ɵɵelement(6, "c8y-breadcrumb-item", 1);
        ɵngcc0.ɵɵpipe(7, "translate");
        ɵngcc0.ɵɵelement(8, "c8y-breadcrumb-item", 0);
        ɵngcc0.ɵɵpipe(9, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(10, "c8y-action-bar-item", 2);
        ɵngcc0.ɵɵelementStart(11, "button", 3);
        ɵngcc0.ɵɵlistener("click", function FeatureListComponent_Template_button_click_11_listener() { return ctx.reload$.next(); });
        ɵngcc0.ɵɵpipe(12, "translate");
        ɵngcc0.ɵɵelement(13, "i", 4);
        ɵngcc0.ɵɵtext(14);
        ɵngcc0.ɵɵpipe(15, "translate");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "c8y-action-bar-item", 5);
        ɵngcc0.ɵɵelementStart(17, "c8y-list-display-switch", 6);
        ɵngcc0.ɵɵlistener("onListClassChange", function FeatureListComponent_Template_c8y_list_display_switch_onListClassChange_17_listener($event) { return ctx.listClass = $event; });
        ɵngcc0.ɵɵpipe(18, "async");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(19, FeatureListComponent_div_19_Template, 6, 0, "div", 7);
        ɵngcc0.ɵɵpipe(20, "async");
        ɵngcc0.ɵɵelementStart(21, "div", 8);
        ɵngcc0.ɵɵtemplate(22, FeatureListComponent_div_22_Template, 13, 9, "div", 9);
        ɵngcc0.ɵɵpipe(23, "async");
        ɵngcc0.ɵɵtemplate(24, FeatureListComponent_div_24_Template, 2, 2, "div", 10);
        ɵngcc0.ɵɵpipe(25, "async");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        let tmp_13_0;
        let tmp_14_0;
        let tmp_16_0;
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind1(2, 18, "Features"));
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("icon", "c8y-atom")("label", ɵngcc0.ɵɵpipeBind1(5, 20, "Ecosystem"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "c8y-modules")("label", ɵngcc0.ɵɵpipeBind1(7, 22, "Applications"))("path", "ecosystem/applications");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("icon", "tab")("label", ɵngcc0.ɵɵpipeBind1(9, 24, "Features"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "right");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵpropertyInterpolate("title", ɵngcc0.ɵɵpipeBind1(12, 26, "Reload"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(38, _c0, ctx.reloading));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ɵngcc0.ɵɵpipeBind1(15, 28, "Reload"), " ");
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("placement", "left");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("listLength", (tmp_13_0 = ɵngcc0.ɵɵpipeBind1(18, 30, ctx.features$)) == null ? null : tmp_13_0.length);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_14_0 = ɵngcc0.ɵɵpipeBind1(20, 32, ctx.features$)) == null ? null : tmp_14_0.length) === 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ctx.listClass);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ((tmp_16_0 = ɵngcc0.ɵɵpipeBind1(23, 34, ctx.features$)) == null ? null : tmp_16_0.length) > 0);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(25, 36, ctx.features$));
    } }, directives: [ɵngcc2.TitleComponent, ɵngcc2.BreadcrumbComponent, ɵngcc2.BreadcrumbItemComponent, ɵngcc2.ActionBarItemComponent, ɵngcc2.IconDirective, ɵngcc3.NgClass, ɵngcc2.ɵbh, ɵngcc3.NgIf, ɵngcc3.NgForOf, ɵngcc2.C8yTranslateDirective, ɵngcc4.ApplicationCardComponent], pipes: [ɵngcc2.C8yTranslatePipe, ɵngcc3.AsyncPipe], encapsulation: 2 });
FeatureListComponent.ctorParameters = () => [
    { type: EcosystemService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FeatureListComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-feature-list',
                template: "<c8y-title>{{ 'Features' | translate }}</c8y-title>\n\n<c8y-breadcrumb>\n  <c8y-breadcrumb-item [icon]=\"'c8y-atom'\" [label]=\"'Ecosystem' | translate\"></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item\n    [icon]=\"'c8y-modules'\"\n    [label]=\"'Applications' | translate\"\n    [path]=\"'ecosystem/applications'\"\n  ></c8y-breadcrumb-item>\n  <c8y-breadcrumb-item [icon]=\"'tab'\" [label]=\"'Features' | translate\"></c8y-breadcrumb-item>\n</c8y-breadcrumb>\n\n<c8y-action-bar-item [placement]=\"'right'\">\n  <button (click)=\"reload$.next()\" class=\"btn btn-link\" title=\"{{ 'Reload' | translate }}\">\n    <i [ngClass]=\"{ 'icon-spin': reloading }\" c8yIcon=\"refresh\"></i>\n    {{ 'Reload' | translate }}\n  </button>\n</c8y-action-bar-item>\n\n<c8y-action-bar-item [placement]=\"'left'\" itemClass=\"navbar-form hidden-xs\">\n  <c8y-list-display-switch\n    (onListClassChange)=\"listClass = $event\"\n    [listLength]=\"(features$ | async)?.length\"\n  ></c8y-list-display-switch>\n</c8y-action-bar-item>\n\n<div *ngIf=\"(features$ | async)?.length === 0\" class=\"c8y-empty-state text-center\">\n  <h1 class=\"c8y-icon c8y-icon-modules c8y-icon-duocolor\"></h1>\n  <h3 translate>No features to display.</h3>\n  <p translate>No additional features are subscribed to the tenant.</p>\n</div>\n\n<div [ngClass]=\"listClass\" class=\"card-group\">\n  <div class=\"page-sticky-header hidden-xs flex-row\" *ngIf=\"(features$ | async)?.length > 0\">\n    <div class=\"card-block card-column-40\">\n      <div class=\"card-appicon p-l-32 p-r-16 m-r-0 m-l-4\"></div>\n      {{ 'Feature' | translate }}\n    </div>\n    <div class=\"card-block p-0 card-column-80\">\n      <div class=\"card-block card-column-80\">{{ 'Description' | translate }}</div>\n      <div class=\"card-block card-column-20\">{{ 'Type' | translate }}</div>\n      <div class=\"card-block card-column-20\"></div>\n    </div>\n  </div>\n  <div *ngFor=\"let feature of features$ | async\" class=\"col-xs-12 col-sm-4 col-md-3\">\n    <c8y-application-card\n      [app]=\"feature\"\n      [canEdit]=\"false\"\n      class=\"d-contents\"\n    ></c8y-application-card>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.EcosystemService }]; }, null); })();
