// @c8y/ngx-components/esm2015/core/dynamic-forms/file/file.type.component.js
import { ChangeDetectionStrategy, Component, ViewChild } from '@angular/core';
import { FieldType } from '@ngx-formly/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';
import * as ɵngcc2 from '../../drop-area/drop-area.component';
import * as ɵngcc3 from '@angular/forms';
import * as ɵngcc4 from 'ngx-bootstrap/popover';
import * as ɵngcc5 from '../../common/icon.directive';
import * as ɵngcc6 from '@ngx-formly/core';

const _c0 = ["dropArea"];
function FieldFile_div_1_button_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "button", 7);
    ɵngcc0.ɵɵelement(1, "i", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("popover", ctx_r3.to.description);
} }
function FieldFile_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 5);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, FieldFile_div_1_button_2_Template, 2, 1, "button", 6);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵpropertyInterpolate("id", ctx_r0.id + "-fieldset");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r0.to.label, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.to.description);
} }
function FieldFile_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵelement(1, "formly-validation-message", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("field", ctx_r2.field);
} }
export class FieldFile extends FieldType {
    constructor() {
        super(...arguments);
        this.dropAreaInputs = [
            'title',
            'message',
            'icon',
            'loadingMessage',
            'alwaysShow',
            'clickToOpen',
            'loading',
            'progress',
            'maxAllowedFiles',
            'files',
            'accept'
        ];
    }
    ngOnInit() {
        this.dropAreaInputs
            .filter(input => this.to[input] !== undefined)
            .forEach(input => (this.dropArea[input] = this.to[input]));
    }
}
FieldFile.ɵfac = /*@__PURE__*/ function () { let ɵFieldFile_BaseFactory; return function FieldFile_Factory(t) { return (ɵFieldFile_BaseFactory || (ɵFieldFile_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FieldFile)))(t || FieldFile); }; }();
FieldFile.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: FieldFile, selectors: [["c8y-field-file"]], viewQuery: function FieldFile_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.dropArea = _t.first);
    } }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 5, vars: 4, consts: [["role", "group", 1, "form-group"], ["class", "legend form-block", 3, "id", 4, "ngIf"], [1, "drop-area-sm", 3, "formControl"], ["dropArea", ""], ["class", "form-control-feedback-message has-error", "role", "alert", 4, "ngIf"], [1, "legend", "form-block", 3, "id"], ["class", "btn-clean text-14 m-l-4", "placement", "right", "outsideClick", "true", "type", "button", 3, "popover", 4, "ngIf"], ["placement", "right", "outsideClick", "true", "type", "button", 1, "btn-clean", "text-14", "m-l-4", 3, "popover"], ["c8yIcon", "question-circle-o", 1, "text-primary"], ["role", "alert", 1, "form-control-feedback-message", "has-error"], [3, "field"]], template: function FieldFile_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, FieldFile_div_1_Template, 3, 3, "div", 1);
        ɵngcc0.ɵɵelement(2, "c8y-drop-area", 2, 3);
        ɵngcc0.ɵɵtemplate(4, FieldFile_div_4_Template, 2, 1, "div", 4);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵattribute("aria-labelledby", ctx.id + "-fieldset");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.to.label || ctx.to.description);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("formControl", ctx.formControl);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showError && ctx.formControl.errors);
    } }, directives: [ɵngcc1.NgIf, ɵngcc2.DropAreaComponent, ɵngcc3.NgControlStatus, ɵngcc3.FormControlDirective, ɵngcc4.PopoverDirective, ɵngcc5.IconDirective, ɵngcc6.ɵc], encapsulation: 2, changeDetection: 0 });
FieldFile.CONFIG = {
    types: [
        {
            name: 'file',
            component: FieldFile
        }
    ]
};
FieldFile.propDecorators = {
    dropArea: [{ type: ViewChild, args: ['dropArea', { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FieldFile, [{
        type: Component,
        args: [{
                selector: 'c8y-field-file',
                template: "<div class=\"form-group\" role=\"group\" [attr.aria-labelledby]=\"id + '-fieldset'\">\n  <div id=\"{{ id + '-fieldset' }}\" class=\"legend form-block\" *ngIf=\"to.label || to.description\">\n    {{ to.label }}\n    <button\n      *ngIf=\"to.description\"\n      class=\"btn-clean text-14 m-l-4\"\n      [popover]=\"to.description\"\n      placement=\"right\"\n      outsideClick=\"true\"\n      type=\"button\"\n    >\n      <i c8yIcon=\"question-circle-o\" class=\"text-primary\"></i>\n    </button>\n  </div>\n\n  <c8y-drop-area #dropArea class=\"drop-area-sm\" [formControl]=\"formControl\"> </c8y-drop-area>\n\n  <div\n    class=\"form-control-feedback-message has-error\"\n    role=\"alert\"\n    *ngIf=\"showError && formControl.errors\"\n  >\n    <formly-validation-message [field]=\"field\"></formly-validation-message>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], null, { dropArea: [{
            type: ViewChild,
            args: ['dropArea', { static: true }]
        }] }); })();
