// @c8y/ngx-components/esm2015/core/help/help.component.js
import { Component, Input } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { AppStateService } from '../common/ui-state.service';
import { DocsService } from '../docs/docs.service';
/**
 * A component which shows a context help in
 * the action bar.
 *
 * @example
 * ```html
 * <c8y-help src="/users-guide/cockpit/#dashboards"></c8y-help>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../docs/docs.service';
import * as ɵngcc2 from '../common/ui-state.service';
import * as ɵngcc3 from '@ngx-translate/core';
import * as ɵngcc4 from '@angular/common';
import * as ɵngcc5 from 'ngx-bootstrap/collapse';
import * as ɵngcc6 from '../action-bar/action-bar-item.component';
import * as ɵngcc7 from '../common/icon.directive';
import * as ɵngcc8 from '../common/loading.component';
import * as ɵngcc9 from '../i18n/c8y-translate.directive';
import * as ɵngcc10 from '../i18n/c8y-translate.pipe';

function HelpComponent_c8y_action_bar_item_0_Template(rf, ctx) { if (rf & 1) {
    const _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "c8y-action-bar-item", 6);
    ɵngcc0.ɵɵelementStart(1, "button", 7);
    ɵngcc0.ɵɵlistener("click", function HelpComponent_c8y_action_bar_item_0_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r6); const ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.toggle(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵelement(3, "i", 8);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("placement", "right")("priority", ctx_r0.priority);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 5, "Open help"));
    ɵngcc0.ɵɵattribute("aria-expanded", !ctx_r0.isCollapsed);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", "question-circle-o");
} }
function HelpComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "c8y-loading");
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_5_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "h4", 13);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(3, "div", 14);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r7.title);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("innerHTML", ctx_r7.sectionContent, ɵngcc0.ɵɵsanitizeHtml);
} }
function HelpComponent_div_5_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵelementStart(1, "h4", 15);
    ɵngcc0.ɵɵtext(2, "Sorry, that didn't work");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 16);
    ɵngcc0.ɵɵelementStart(4, "p", 17);
    ɵngcc0.ɵɵtext(5, "The content couldn't be loaded.");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
const _c0 = function (a0, a1) { return { "text-warning": a0, "text-muted": a1 }; };
function HelpComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "div", 9);
    ɵngcc0.ɵɵelement(2, "i", 10);
    ɵngcc0.ɵɵtemplate(3, HelpComponent_div_5_div_3_Template, 4, 2, "div", 11);
    ɵngcc0.ɵɵtemplate(4, HelpComponent_div_5_div_4_Template, 6, 0, "div", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("c8yIcon", !ctx_r3.hasError ? ctx_r3.icon : "unlink")("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c0, ctx_r3.hasError, !ctx_r3.hasError));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r3.hasError);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r3.hasError);
} }
function HelpComponent_div_6_span_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 17);
    ɵngcc0.ɵɵtext(1, " Open the ");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3, "User guide`KEEP_ORIGINAL`");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_6_span_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 17);
    ɵngcc0.ɵɵtext(1, " Check the ");
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3, "User guide`KEEP_ORIGINAL`");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} }
function HelpComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    const _r12 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 18);
    ɵngcc0.ɵɵelementStart(1, "button", 19);
    ɵngcc0.ɵɵlistener("click", function HelpComponent_div_6_Template_button_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r12); const ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.toggle(); });
    ɵngcc0.ɵɵpipe(2, "translate");
    ɵngcc0.ɵɵtext(3, "Close");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "a", 20);
    ɵngcc0.ɵɵtemplate(5, HelpComponent_div_6_span_5_Template, 4, 0, "span", 21);
    ɵngcc0.ɵɵtemplate(6, HelpComponent_div_6_span_6_Template, 4, 0, "span", 21);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("title", ɵngcc0.ɵɵpipeBind1(2, 5, "Close help"));
    ɵngcc0.ɵɵattribute("aria-expanded", !ctx_r4.isCollapsed);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵpropertyInterpolate("href", ctx_r4.hasError ? ctx_r4.getUrl() : ctx_r4.getUrl(ctx_r4.src), ɵngcc0.ɵɵsanitizeUrl);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r4.hasError);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.hasError);
} }
export class HelpComponent {
    /**
     * @ignore Only private DI
     */
    constructor(docsService, appState, translateService) {
        this.docsService = docsService;
        this.appState = appState;
        this.translateService = translateService;
        /**
         * The source of the documentation. Used to link to the documentation as well as
         * to parse the source to display.
         */
        this.src = '';
        /**
         * Indicates if the help dialog is collapsed.
         */
        this.isCollapsed = true;
        /**
         * The priority where the help icon should be shown in the action bar
         */
        this.priority = Infinity;
        /**
         * Currrent version of the UI which will be used to get the correct data from
         * the docs. If the UI version is equal to the latest doc version, the value
         * must be ''.
         */
        this.version = this.parseVersion(this.appState.uiVersion);
        /**
         * An title. Set in open by passing the source.
         */
        this.title = '';
        /**
         * The section heading in the doc which is going to be displayed.
         */
        this.sectionHeading = '';
        /**
         * The section content in the doc which is going to be displayed.
         */
        this.sectionContent = '';
        /**
         * Indicates if the component is loading.
         */
        this.isLoading = true;
        /**
         * Indicates if the component failed loading the source.
         */
        this.hasError = false;
        this.SUPPORTED_LANGUAGES = ['en'];
    }
    /**
     * Identifies if the current user language is supported
     * Currently only English is supported.
     */
    get isSupportedLanguage() {
        return this.SUPPORTED_LANGUAGES.indexOf(this.translateService.currentLang) > -1;
    }
    /**
     * Builds the URL based on the src. The Base URL can be set in the application options docBaseUrl.
     * @param src The source of the help on the guide.
     * @param index This flag is used to call the index.json content of a guide. For example, "https://www.cumulocity.com/guides/users-guide/cockpit/index.json".
     */
    getUrl(src = '', index = false) {
        const docsUrl = new URL(this.docsService.getBaseUrl());
        const srcUrl = new URL(`${docsUrl}${src}`);
        const [url, hashFragment] = src.split('#');
        this.sectionHeading = hashFragment;
        if (index) {
            src = `${url}index.json`;
        }
        if (this.version === '') {
            return index ? `${docsUrl.href + src}` : srcUrl.href;
        }
        return `${docsUrl.href}/${this.version}${src}`;
    }
    /**
     * Toggles the visibility of the help dialog.
     */
    toggle() {
        if (this.isCollapsed) {
            this.open();
            return;
        }
        this.close();
    }
    /**
     * Closes the help dialog.
     */
    close() {
        this.isCollapsed = true;
        this.clean();
    }
    /**
     * Opens the help dialog.
     */
    open() {
        this.isCollapsed = false;
        this.isLoading = true;
        this.requestContent();
        if (!this.icon) {
            this.icon = this.resolveIcon();
        }
    }
    requestContent() {
        const req = new XMLHttpRequest();
        req.onreadystatechange = () => this.render(req);
        req.addEventListener('load', () => this.render(req));
        req.open('GET', this.getUrl(this.src, true));
        req.responseType = 'json';
        req.setRequestHeader('Accept', 'text/html');
        req.send();
    }
    clean() {
        this.title = '';
        this.isLoading = true;
        this.hasError = false;
        this.sectionContent = '';
    }
    parseVersion(uiVersion) {
        const version = uiVersion.split('.')[0];
        const majorNumber = Math.floor(parseInt(version, 10) / 100);
        const minorNumber = parseInt(version, 10) - majorNumber * 100;
        return `${majorNumber}.${minorNumber}.0`;
    }
    resolveIcon() {
        try {
            const icon = Array.from(document.querySelector('nav .active i').classList).find(classes => classes.startsWith('c8y-icon-') || classes.startsWith('dlt-c8y-icon-'));
            return icon.replace('dlt-c8y-icon-', '').replace('c8y-icon-', 'c8y-');
        }
        catch (ex) {
            return 'life-saver';
        }
    }
    render(req) {
        if (req.readyState === 4) {
            this.isLoading = false;
            if (req.status === 404 && this.version !== '') {
                this.version = '';
                this.open();
            }
            if (req.status === 200) {
                this.hasError = false;
                const sectionData = req.response[this.sectionHeading];
                if (sectionData) {
                    this.title = sectionData.title;
                    this.sectionContent = sectionData.helpcontent;
                }
            }
            else {
                this.hasError = true;
            }
        }
    }
}
HelpComponent.ɵfac = function HelpComponent_Factory(t) { return new (t || HelpComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DocsService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.AppStateService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.TranslateService)); };
HelpComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: HelpComponent, selectors: [["c8y-help"]], inputs: { src: "src", isCollapsed: "isCollapsed", priority: "priority", icon: "icon" }, decls: 7, vars: 6, consts: [["itemClass", "pull-right", 3, "placement", "priority", 4, "ngIf"], ["id", "collapseHelp", 1, "c8y-help-drawer", 3, "collapse", "isAnimated"], [1, "c8y-help-drawer-block"], ["docOutlet", ""], [4, "ngIf"], ["class", "c8y-help-drawer-footer", 4, "ngIf"], ["itemClass", "pull-right", 3, "placement", "priority"], ["aria-controls", "collapseHelp", 1, "btn", "btn-help", 3, "title", "click"], [1, "text-info", 3, "c8yIcon"], [1, "d-flex"], [1, "c8y-icon-duocolor", "icon-48", 3, "c8yIcon", "ngClass"], ["class", "p-l-16 p-t-16 flex-grow", 4, "ngIf"], [1, "p-l-16", "p-t-16", "flex-grow"], [1, "text-bold", "text-primary", "m-b-16"], ["id", "helpContent", 1, "help-content", 3, "innerHTML"], ["translate", "", 1, "text-bold", "m-b-16"], [1, "help-content"], ["translate", ""], [1, "c8y-help-drawer-footer"], ["aria-controls", "collapseHelp", "translate", "", 1, "btn", "btn-default", 3, "title", "click"], ["target", "_blank", 1, "btn", "btn-primary", 3, "href"], ["translate", "", 4, "ngIf"]], template: function HelpComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, HelpComponent_c8y_action_bar_item_0_Template, 4, 7, "c8y-action-bar-item", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2, 3);
        ɵngcc0.ɵɵtemplate(4, HelpComponent_div_4_Template, 2, 0, "div", 4);
        ɵngcc0.ɵɵtemplate(5, HelpComponent_div_5_Template, 5, 7, "div", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(6, HelpComponent_div_6_Template, 7, 7, "div", 5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isSupportedLanguage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("collapse", ctx.isCollapsed)("isAnimated", true);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLoading);
    } }, directives: [ɵngcc4.NgIf, ɵngcc5.CollapseDirective, ɵngcc6.ActionBarItemComponent, ɵngcc7.IconDirective, ɵngcc8.LoadingComponent, ɵngcc4.NgClass, ɵngcc9.C8yTranslateDirective], pipes: [ɵngcc10.C8yTranslatePipe], encapsulation: 2 });
HelpComponent.ctorParameters = () => [
    { type: DocsService },
    { type: AppStateService },
    { type: TranslateService }
];
HelpComponent.propDecorators = {
    src: [{ type: Input }],
    isCollapsed: [{ type: Input }],
    priority: [{ type: Input }],
    icon: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(HelpComponent, [{
        type: Component,
        args: [{
                selector: 'c8y-help',
                template: "<c8y-action-bar-item\n  [placement]=\"'right'\"\n  itemClass=\"pull-right\"\n  [priority]=\"priority\"\n  *ngIf=\"isSupportedLanguage\"\n>\n  <button\n    class=\"btn btn-help\"\n    [title]=\"'Open help' | translate\"\n    (click)=\"toggle()\"\n    [attr.aria-expanded]=\"!isCollapsed\"\n    aria-controls=\"collapseHelp\"\n  >\n    <i\n      [c8yIcon]=\"'question-circle-o'\"\n      class=\"text-info\"\n    ></i>\n  </button>\n</c8y-action-bar-item>\n\n<div\n  id=\"collapseHelp\"\n  class=\"c8y-help-drawer\"\n  [collapse]=\"isCollapsed\"\n  [isAnimated]=\"true\"\n>\n  <div\n    class=\"c8y-help-drawer-block\"\n    #docOutlet\n  >\n    <div\n      *ngIf=\"isLoading\"\n    >\n      <c8y-loading></c8y-loading>\n    </div>\n\n    <div *ngIf=\"!isLoading\">\n      <div class=\"d-flex\">\n        <i\n          [c8yIcon]=\"!hasError ? icon : 'unlink'\"\n          [ngClass]=\"{ 'text-warning': hasError, 'text-muted': !hasError }\"\n          class=\"c8y-icon-duocolor icon-48\"\n        ></i>\n        <div\n          class=\"p-l-16 p-t-16 flex-grow\"\n          *ngIf=\"!hasError\"\n        >\n          <h4 class=\"text-bold text-primary m-b-16\">{{title}}</h4>\n          <div\n            id=\"helpContent\"\n            class=\"help-content\"\n            [innerHTML] = \"sectionContent\"\n          ></div>\n          \n        </div>\n\n        <div\n          class=\"p-l-16 p-t-16 flex-grow\"\n          *ngIf=\"hasError\"\n        >\n          <h4\n            class=\"text-bold m-b-16\"\n            translate\n          >Sorry, that didn't work</h4>\n          <div class=\"help-content\">\n            <p translate>The content couldn't be loaded.</p>\n          </div>\n        </div>\n      </div>\n    </div>\n\n  </div>\n  <div class=\"c8y-help-drawer-footer\" *ngIf=\"!isLoading\">\n    <button\n      class=\"btn btn-default\"\n      (click)=\"toggle()\"\n      [title]=\"'Close help' | translate\"\n      [attr.aria-expanded]=\"!isCollapsed\"\n      aria-controls=\"collapseHelp\"\n      translate\n    >Close</button>\n    <a\n      href=\"{{ hasError ? getUrl() : getUrl(src) }}\"\n      class=\"btn btn-primary\"\n      target=\"_blank\"\n    >\n    <span translate *ngIf=\"!hasError\">\n      Open the <span>User guide`KEEP_ORIGINAL`</span>\n    </span>\n    <span translate *ngIf=\"hasError\">\n      Check the <span>User guide`KEEP_ORIGINAL`</span>\n    </span>\n    </a>\n  </div>\n</div>\n"
            }]
    }], function () { return [{ type: ɵngcc1.DocsService }, { type: ɵngcc2.AppStateService }, { type: ɵngcc3.TranslateService }]; }, { src: [{
            type: Input
        }], isCollapsed: [{
            type: Input
        }], priority: [{
            type: Input
        }], icon: [{
            type: Input
        }] }); })();
