// @c8y/ngx-components/esm2015/core/device-status/device-status.module.js
import { NgModule } from '@angular/core';
import { DeviceStatusComponent } from './device-status.component';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { CommonModule } from '../common/common.module';
/**
 * The angular module definition for a device-status.
 * @exports DeviceStatusComponent
 */
import * as ɵngcc0 from '@angular/core';
export class DeviceStatusModule {
}
DeviceStatusModule.ɵfac = function DeviceStatusModule_Factory(t) { return new (t || DeviceStatusModule)(); };
DeviceStatusModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DeviceStatusModule });
DeviceStatusModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            TooltipModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DeviceStatusModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    TooltipModule
                ],
                exports: [
                    DeviceStatusComponent
                ],
                declarations: [
                    DeviceStatusComponent
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DeviceStatusModule, { declarations: function () { return [DeviceStatusComponent]; }, imports: function () { return [CommonModule,
        TooltipModule]; }, exports: function () { return [DeviceStatusComponent]; } }); })();
