// @c8y/ngx-components/esm2015/core/dynamic-component/dynamic-component.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { DynamicComponentComponent } from './dynamic-component.component';
import { DynamicComponentService } from './dynamic-component.service';
import { I18nModule } from '../i18n/i18n.module';
import { RESOLVING_COMPONENT_WAIT_TIME } from './dynamic-component.model';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class DynamicComponentModule {
}
DynamicComponentModule.ɵfac = function DynamicComponentModule_Factory(t) { return new (t || DynamicComponentModule)(); };
DynamicComponentModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: DynamicComponentModule });
DynamicComponentModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [DynamicComponentService, { provide: RESOLVING_COMPONENT_WAIT_TIME, useValue: 5000 }], imports: [[CommonModule, CollapseModule, RouterModule.forChild([]), I18nModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicComponentModule, [{
        type: NgModule,
        args: [{
                declarations: [DynamicComponentComponent],
                imports: [CommonModule, CollapseModule, RouterModule.forChild([]), I18nModule],
                exports: [DynamicComponentComponent],
                providers: [DynamicComponentService, { provide: RESOLVING_COMPONENT_WAIT_TIME, useValue: 5000 }]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DynamicComponentModule, { declarations: function () { return [DynamicComponentComponent]; }, imports: function () { return [CommonModule, CollapseModule, ɵngcc1.RouterModule, I18nModule]; }, exports: function () { return [DynamicComponentComponent]; } }); })();
