// @c8y/ngx-components/esm2015/core/dynamic-component/dynamic-component.service.js
import { Inject, Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { BehaviorSubject, Subject, timer } from 'rxjs';
import { filter, mergeMap, shareReplay, take, takeUntil } from 'rxjs/operators';
import { fromTriggerOnce, stateToFactory } from '../common/extension-hooks';
import { StateService } from '../common/state-service.abstract';
import { HOOK_COMPONENTS, RESOLVING_COMPONENT_WAIT_TIME } from './dynamic-component.model';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "./dynamic-component.model";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
export class DynamicComponentService extends StateService {
    constructor(injector, router, waitTimeout) {
        super();
        this.injector = injector;
        this.refreshTrigger = new Subject();
        this.state$ = new BehaviorSubject(new Set());
        /**
         * Additional factories that can be added by plugins.
         */
        this.factories = [];
        this.items$ = fromTriggerOnce(router, this.refreshTrigger, [
            () => this.injector.get(HOOK_COMPONENTS, []),
            () => this.factories,
            stateToFactory(this.state$)
        ]).pipe(takeUntil(timer(waitTimeout)), shareReplay(1));
    }
    refresh() {
        this.refreshTrigger.next(1);
    }
    get state() {
        return this.state$.value;
    }
    /**
     * Returns a component by it's id as an observable.
     * @param id The id of the component.
     */
    getById(id) {
        return this.items$
            .pipe(mergeMap((cmps) => cmps), filter((cmp) => cmp.id === id), take(1))
            .toPromise();
    }
    add(component) {
        this.state.add(component);
        this.emitNewState();
    }
    remove(component) {
        this.state.delete(component);
        this.emitNewState();
    }
}
DynamicComponentService.ɵfac = function DynamicComponentService_Factory(t) { return new (t || DynamicComponentService)(ɵngcc0.ɵɵinject(ɵngcc0.Injector), ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(RESOLVING_COMPONENT_WAIT_TIME)); };
DynamicComponentService.ɵprov = i0.ɵɵdefineInjectable({ factory: function DynamicComponentService_Factory() { return new DynamicComponentService(i0.ɵɵinject(i0.INJECTOR), i0.ɵɵinject(i1.Router), i0.ɵɵinject(i2.RESOLVING_COMPONENT_WAIT_TIME)); }, token: DynamicComponentService, providedIn: "root" });
DynamicComponentService.ctorParameters = () => [
    { type: Injector },
    { type: Router },
    { type: Number, decorators: [{ type: Inject, args: [RESOLVING_COMPONENT_WAIT_TIME,] }] }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DynamicComponentService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc0.Injector }, { type: ɵngcc1.Router }, { type: Number, decorators: [{
                type: Inject,
                args: [RESOLVING_COMPONENT_WAIT_TIME]
            }] }]; }, null); })();
